// @postman/pm-bin-linux-x64/bin/lib/commands/flows/api/ai-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("./request");
const constants_1 = require("../../../postman-entities/api/constants");
const FLOWS_AI_SERVICE = 'flows-ai';
const flowsAiModule = {
    async actorChat(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: FLOWS_AI_SERVICE,
                debug: options.debug,
                body: {
                    model: options.model,
                    messages: options.messages,
                    ...(options.tools !== undefined && { tools: options.tools }),
                    ...(options.outputSchema !== undefined && { outputSchema: options.outputSchema })
                }
            };
            return (await (0, request_1.makeApiRequest)('/actor/chat', requestOptions)).data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async agentExecute(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: FLOWS_AI_SERVICE,
                debug: options.debug,
                body: {
                    model: options.model,
                    systemPrompt: options.systemPrompt,
                    ...(options.context !== undefined && { context: options.context }),
                    ...(options.mcpServers !== undefined && { mcpServers: options.mcpServers }),
                    ...(options.outputSchema !== undefined && { outputSchema: options.outputSchema })
                }
            };
            return (await (0, request_1.makeApiRequest)('/agent/execute', requestOptions)).data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async generateText(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: FLOWS_AI_SERVICE,
                debug: options.debug,
                body: {
                    prompt: options.prompt,
                    model: options.model,
                    ...(options.predictability !== undefined && { predictability: options.predictability }),
                    ...(options.creativity !== undefined && { creativity: options.creativity })
                }
            };
            return (await (0, request_1.makeApiRequest)('/generate/text', requestOptions)).data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async generateAuto(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: FLOWS_AI_SERVICE,
                debug: options.debug,
                body: {
                    prompt: options.prompt,
                    model: options.model,
                    ...(options.predictability !== undefined && { predictability: options.predictability }),
                    ...(options.creativity !== undefined && { creativity: options.creativity })
                }
            };
            return (await (0, request_1.makeApiRequest)('/generate/auto', requestOptions)).data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    }
};
exports.default = flowsAiModule;
