// @postman/pm-bin-linux-x64/bin/lib/commands/flows/api/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("./request");
const constants_1 = require("../../../postman-entities/api/constants");
const either_1 = require("../types/either");
const flowsModule = {
    async listRuns(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.workspaceId) {
            throw new Error('Workspace ID is required');
        }
        try {
            const filter = {
                time: {
                    start_time_ms: options.startMs,
                    end_time_ms: options.endMs
                }
            };
            if (options.flowId) {
                filter.action_id = options.flowId;
            }
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: {
                    workspaceId: options.workspaceId,
                    options: { filter }
                }
            };
            const response = await (0, request_1.makeApiRequest)('/rpc/actions.executionHistory.v3', requestOptions);
            return {
                history: response.data?.history || [],
                meta: response.data?.meta || { actions: {}, done: true }
            };
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async getRun(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.runId) {
            throw new Error('Run ID is required');
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: { session: options.runId }
            };
            const response = await (0, request_1.makeApiRequest)('/rpc/actions.getExecution.v3', requestOptions);
            return response;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async listFlows(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.workspace) {
            throw new Error('Workspace ID is required');
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug
            };
            const response = await (0, request_1.makeApiRequest)(`/flows/workspace/${options.workspace}/get`, requestOptions);
            const flows = response.data.map((item) => {
                const activeTargets = item.attributes?.activeTargets;
                const isDeployed = Array.isArray(activeTargets) && activeTargets.length > 0;
                return {
                    id: item.id,
                    name: item.attributes.name,
                    type: item.attributes.type,
                    updatedAt: item.attributes.updatedOn,
                    isDeployed
                };
            });
            return flows;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async createTarget(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.flowId) {
            throw new Error('Flow ID is required');
        }
        if (!options.path) {
            throw new Error('Path is required');
        }
        try {
            const requestBody = {
                action: options.flowId,
                config: {
                    path: options.path,
                    settings: {}
                }
            };
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: requestBody
            };
            const result = await (0, request_1.apiRequestWithStatus)('/rpc/target.create.v1', requestOptions);
            if ((0, either_1.isLeft)(result)) {
                if (result.left.statusCode === 409) {
                    return (0, either_1.left)({ conflict: true });
                }
                throw new Error(result.left.error.message || 'Failed to create target');
            }
            return (0, either_1.right)(result.right.data);
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async getTarget(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.flowId) {
            throw new Error('Flow ID is required');
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: {
                    action: options.flowId
                }
            };
            const response = await (0, request_1.makeApiRequest)('/rpc/target.get.v1', requestOptions);
            return response.data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async updateTargetPath(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.targetId) {
            throw new Error('Target ID is required');
        }
        if (!options.path) {
            throw new Error('Path is required');
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                body: {
                    target: options.targetId,
                    path: options.path
                }
            };
            await (0, request_1.makeApiRequest)('/rpc/target.updatePath.v1', requestOptions);
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async deployTarget(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.targetId) {
            throw new Error('Target ID is required');
        }
        if (!options.flowId) {
            throw new Error('Flow ID is required');
        }
        try {
            const requestBody = {
                target: options.targetId,
                description: options.description || '',
                configuration: {
                    source: {
                        action: options.flowId
                    },
                    settings: {
                        httpSessionTimeout: options.timeout || 10000
                    }
                }
            };
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: requestBody
            };
            const response = await (0, request_1.makeApiRequest)('/rpc/target.deploy.v1', requestOptions);
            return response.data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async enableAuth(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.flowId) {
            throw new Error('Flow ID is required');
        }
        if (!options.targetId) {
            throw new Error('Target ID is required');
        }
        try {
            const requestBody = {
                action: options.flowId,
                target: options.targetId
            };
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: requestBody
            };
            const response = await (0, request_1.makeApiRequest)('/rpc/target.authEnable.v1', requestOptions);
            return response.data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async maintenance(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.targetId) {
            throw new Error('Target ID is required');
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: {
                    target: options.targetId,
                    enabled: options.enabled
                }
            };
            await (0, request_1.makeApiRequest)('/rpc/target.maintenance.v1', requestOptions);
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async disableAuth(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.targetId) {
            throw new Error('Target ID is required');
        }
        try {
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: {
                    target: options.targetId
                }
            };
            await (0, request_1.makeApiRequest)('/rpc/target.authDisable.v1', requestOptions);
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    },
    async decryptAuth(options) {
        if (!options.accessToken) {
            throw new Error(constants_1.NO_AUTHORIZATION_DATA);
        }
        if (!options.targetId) {
            throw new Error('Target ID is required');
        }
        if (!options.encryptedAuthToken) {
            throw new Error('Encrypted auth token is required');
        }
        try {
            const requestBody = {
                target: options.targetId,
                encryptedAuthToken: options.encryptedAuthToken
            };
            const requestOptions = {
                method: 'post',
                accessToken: options.accessToken,
                service: options.service,
                debug: options.debug,
                body: requestBody
            };
            const response = await (0, request_1.makeApiRequest)('/rpc/target.authDecrypt.v1', requestOptions);
            return response.data;
        }
        catch (error) {
            throw (0, request_1.canonicalRequestErrorHandler)(error);
        }
    }
};
exports.default = flowsModule;
