// @postman/pm-bin-linux-x64/bin/lib/commands/flows/api/request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getClient = getClient;
exports.makeApiRequest = makeApiRequest;
exports.apiRequestWithStatus = apiRequestWithStatus;
exports.canonicalRequestErrorHandler = canonicalRequestErrorHandler;
const print = __importStar(require("../../../framework/print"));
const _ = __importStar(require("lodash"));
const request_1 = __importDefault(require("@postman/platform-req/lib/request"));
const liquid_json_1 = __importDefault(require("liquid-json"));
const util = __importStar(require("../../../util"));
const either_1 = require("../types/either");
function getClient() {
    const options = {
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: {
                    enabled: true
                }
            }
        }
    };
    return request_1.default.defaults(options);
}
function apiRequestWithStatus(endpoint, options) {
    return new Promise((resolve, reject) => {
        if (!options.accessToken) {
            reject(new Error('Postman access token is required'));
            return;
        }
        const client = getClient();
        const searchParams = new URLSearchParams();
        Object.entries(options.queryParams || {}).forEach(([key, value]) => {
            if (!_.isNil(value)) {
                if (Array.isArray(value)) {
                    searchParams.append(key, value.join(','));
                }
                else {
                    searchParams.append(key, value);
                }
            }
        });
        const queryString = searchParams.toString();
        const baseUrl = util.POSTMAN_GATEWAY_BASE_URL();
        const url = `${baseUrl}${endpoint}${queryString ? '?' + queryString : ''}`;
        const requestOptions = {
            url: url,
            headers: {
                'x-access-token': options.accessToken,
                'x-pstmn-req-service': options.service,
                'Content-Type': 'application/json',
                'User-Agent': util.userAgent,
                Accept: 'application/json'
            },
            body: options.body ? JSON.stringify(options.body) : undefined
        };
        if (options.debug) {
            print.lf({
                ...requestOptions,
                headers: {
                    ...requestOptions.headers,
                    'x-access-token': '[REDACTED]'
                }
            });
        }
        client[options.method || 'get'](requestOptions, (err, response, body) => {
            if (err) {
                return reject(new Error(err.message || err));
            }
            try {
                if (options.debug) {
                    print.lf('Response body:', body);
                }
                if (_.isString(body)) {
                    body = liquid_json_1.default.parse(body.trim());
                }
            }
            catch {
                return reject(new Error('Invalid JSON response from Flows API'));
            }
            if (!(/2../).test(response.statusCode)) {
                const apiResponse = body;
                const errorMessage = _.get(apiResponse, 'error.message') ||
                    _.get(apiResponse, 'message') ||
                    `API request failed with status ${response.statusCode}`;
                if ((/40./).test(response.statusCode)) {
                    return resolve((0, either_1.left)({
                        statusCode: response.statusCode,
                        error: body.error || { message: errorMessage }
                    }));
                }
                return reject(new Error(errorMessage));
            }
            const data = _.get(body, 'data', body);
            const cursor = _.get(body, 'meta.nextCursor', undefined);
            resolve((0, either_1.right)({ data, cursor }));
        });
    });
}
async function makeApiRequest(endpoint, options) {
    const result = await apiRequestWithStatus(endpoint, options);
    if ((0, either_1.isLeft)(result)) {
        const error = result.left.error;
        throw new Error(error.message || 'API request failed');
    }
    return result.right;
}
function canonicalRequestErrorHandler(error) {
    if (error.message.includes('401') || error.message.includes('Unauthorized')) {
        return new Error('Invalid access token. Please check a valid access token and try again.');
    }
    if (error.message.includes('403') || error.message.includes('Forbidden')) {
        return new Error('Access denied. Please check your permissions for the specified resource.');
    }
    if (error.message.includes('404')) {
        return new Error('Resource not found.');
    }
    return error;
}
