// @postman/pm-bin-linux-x64/bin/lib/postman-entities/file-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSON_AND_YAML_EXTENSIONS = exports.YAML_EXTENSIONS = exports.JSON_EXTENSION = void 0;
exports.hasExtension = hasExtension;
exports.isYamlFile = isYamlFile;
exports.isJsonFile = isJsonFile;
exports.scanDirectoryForFiles = scanDirectoryForFiles;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
exports.JSON_EXTENSION = '.json';
exports.YAML_EXTENSIONS = ['.yaml', '.yml'];
exports.JSON_AND_YAML_EXTENSIONS = [exports.JSON_EXTENSION, ...exports.YAML_EXTENSIONS];
function hasExtension(filePath, extensions) {
    const lower = filePath.toLowerCase();
    return extensions.some((ext) => { return lower.endsWith(ext); });
}
function isYamlFile(filePath) {
    return hasExtension(filePath, exports.YAML_EXTENSIONS);
}
function isJsonFile(filePath) {
    return hasExtension(filePath, [exports.JSON_EXTENSION]);
}
function scanDirectoryForFiles(directory, extensions) {
    return promises_1.default.readdir(directory)
        .then((files) => {
        return files
            .filter((file) => { return hasExtension(file, extensions); })
            .map((file) => { return path_1.default.join(directory, file); });
    })
        .catch((err) => {
        if (err.code === 'ENOENT') {
            return [];
        }
        throw err;
    });
}
