// @postman/pm-bin-linux-x64/bin/lib/commands/runner/services/run-service.js
const _ = require('lodash'), os = require('os'), fs = require('fs'), uuid = require('uuid'), util = require('../../../util'), version = require('../../../../package.json').version, X_Cluster_Id_Header = 'x-cluster-id', X_Cluster_Key_Header = 'x-api-key', Accept_Header = 'accept', Content_Type_Header = 'content-type', User_Agent_Header = 'User-Agent', X_Agent_Version_Header = 'x-agent-version';
let request = require('postman-request');
let isProxified = false;
function buildInstanceId() {
    const TOTAL_LENGTH = 25, shortName = 'postman-cli';
    let appId = shortName.substr(0, 1), hostname = _.last(_.split(os.hostname(), '-')).replace(/\W/g, '').substr(0, 17), pid = process.pid, random = uuid.v4().replace(/-/g, ''), instanceId = `${appId}${hostname}${pid}_${random}`;
    return instanceId.substr(0, TOTAL_LENGTH);
}
const fileErrorMessages = {
    deletedResourceError: 'File used in the associated collection is not accessible as it has been deleted by the ' +
        'team admin.',
    planLimitExceeded: 'File used in the associated collection could not be retrieved from Postman as you have ' +
        'reached the limit for data file retrievals. To continue accessing your uploaded files, upgrade your plan.'
};
const agentId = buildInstanceId();
const uniqueId = uuid.v4().replace(/-/g, '').substring(0, 10);
function buildInstanceInfo({ disabledFields = '' }) {
    let fieldsDisabled = {};
    if (disabledFields && typeof disabledFields === 'string') {
        disabledFields.split(',').map((field) => {
            return field.trim();
        }).forEach((key) => { fieldsDisabled[key] = false; });
    }
    const memoryUsage = process.memoryUsage();
    const cpuUsage = process.cpuUsage();
    const instanceInfo = {
        hostname: fieldsDisabled.hostname ? null : os.hostname(),
        pid: fieldsDisabled.pid ? null : process.pid,
        uniqueId,
        osType: fieldsDisabled.osType ? null : os.type(),
        osPlatform: fieldsDisabled.osPlatform ? null : os.platform(),
        arch: fieldsDisabled.arch ? null : os.arch(),
        cliVersion: fieldsDisabled.cliVersion ? null : version,
        memoryUsage: fieldsDisabled.memoryUsage ? null : {
            rss: memoryUsage.rss,
            heapTotal: memoryUsage.heapTotal,
            heapUsed: memoryUsage.heapUsed,
            external: memoryUsage.external,
            arrayBuffers: memoryUsage.arrayBuffers
        },
        cpuUsage: fieldsDisabled.cpuUsage ? null : {
            user: cpuUsage.user,
            system: cpuUsage.system
        }
    };
    return instanceInfo;
}
function getApiUrl() {
    const envApiUrl = process.env.POSTMAN_RUNNER_URL;
    if (envApiUrl) {
        return envApiUrl;
    }
    return util.POSTMAN_GATEWAY_BASE_URL();
}
function createHeaders(runnerId, runnerKey, includeContentType = false) {
    const headers = {
        [Accept_Header]: 'application/json',
        [User_Agent_Header]: 'Postman CLI/' + version,
        [X_Cluster_Id_Header]: runnerId,
        [X_Cluster_Key_Header]: runnerKey,
        [X_Agent_Version_Header]: 'Postman CLI/' + version,
        'x-pstmn-req-service': 'monitoring-orchestrator'
    };
    if (includeContentType) {
        headers[Content_Type_Header] = 'application/json';
    }
    return headers;
}
function handleResponseError(response, operation) {
    let data = response.body;
    if (typeof response.body === 'string') {
        try {
            data = JSON.parse(response.body);
        }
        catch (parseError) {
            data = { message: response.body };
        }
    }
    const message = _.get(data, 'error.details.message') || _.get(data, 'error') || _.get(data, 'message');
    const errorObj = new Error(`Failed to ${operation}: ${response.statusCode} - ${message}`);
    errorObj.statusCode = response.statusCode;
    return errorObj;
}
function proxifyRequest(connectionOptions = {}) {
    if (isProxified) {
        return;
    }
    const requestDefaultOptions = {};
    if (connectionOptions.proxyOptions) {
        requestDefaultOptions.proxy = {
            hostname: connectionOptions.proxyOptions.host,
            port: connectionOptions.proxyOptions.port,
            auth: `${connectionOptions.proxyOptions.username}:${connectionOptions.proxyOptions.password}`
        };
    }
    if (connectionOptions.caCertPath) {
        requestDefaultOptions.extraCA = fs.readFileSync(connectionOptions.caCertPath);
    }
    if (connectionOptions.tunnel) {
        requestDefaultOptions.tunnel = connectionOptions.tunnel;
    }
    if (Object.keys(requestDefaultOptions).length === 0) {
        return;
    }
    request = request.defaults(requestDefaultOptions);
    isProxified = true;
}
function makeRequest(options) {
    const { method, url, headers, json, qs, operation, responseHandler, responseErrorHandler } = options;
    return new Promise((resolve, reject) => {
        const requestOptions = {
            url,
            headers
        };
        if (json) {
            requestOptions.json = json;
        }
        if (qs) {
            requestOptions.qs = qs;
        }
        request[method.toLowerCase()](requestOptions, (err, response) => {
            if (err) {
                reject(err);
                return;
            }
            if (response && !(response.statusCode >= 200 && response.statusCode < 300)) {
                if (responseErrorHandler) {
                    reject(responseErrorHandler(response, operation));
                }
                else {
                    reject(handleResponseError(response, operation));
                }
                return;
            }
            if (responseHandler) {
                resolve(responseHandler(response));
            }
            else {
                resolve(response.body);
            }
        });
    });
}
function fetchRun(runnerId, runnerKey, disableTelemetry) {
    const apiUrl = getApiUrl(), runUrl = `${apiUrl}/runs/fetch`, headers = createHeaders(runnerId, runnerKey), instanceInfo = buildInstanceInfo({ disabledFields: disableTelemetry });
    return makeRequest({
        method: 'POST',
        url: runUrl,
        headers,
        json: {
            cluster: runnerId,
            agentId: agentId,
            instanceInfo
        },
        operation: 'fetch run'
    });
}
function finishRun({ run, runLog, runnerId, runnerKey }) {
    const apiUrl = getApiUrl(), finishUrl = `${apiUrl}/runs/${run.id}/finish`;
    return makeRequest({
        method: 'POST',
        url: finishUrl,
        headers: createHeaders(runnerId, runnerKey, true),
        json: { runLog },
        operation: 'finish run',
        responseHandler: (response) => {
            return {
                success: true,
                result: response.body
            };
        }
    });
}
function releaseRun({ run, runLog, runnerId, runnerKey }) {
    const apiUrl = getApiUrl(), releaseUrl = `${apiUrl}/runs/${run.id}/release`;
    return makeRequest({
        method: 'POST',
        url: releaseUrl,
        headers: createHeaders(runnerId, runnerKey, true),
        json: { runLog },
        operation: 'release run',
        responseHandler: (response) => {
            return {
                success: true,
                result: response.body
            };
        }
    });
}
function extractFileName(contentDisposition) {
    if (!contentDisposition || typeof contentDisposition !== 'string') {
        return 'unknown-file';
    }
    try {
        const filenameMatch = contentDisposition.match(/filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/);
        if (filenameMatch && filenameMatch[1]) {
            return filenameMatch[1].replace(/^['"]|['"]$/g, '').trim();
        }
        const simpleSplit = contentDisposition.split('filename=')[1];
        if (simpleSplit) {
            return simpleSplit.replace(/^['"]|['"]$/g, '').trim();
        }
    }
    catch (error) {
    }
    return 'unknown-file';
}
function getDataFile({ fileId, runId, runnerId, runnerKey }) {
    const apiUrl = getApiUrl(), fetchUrl = `${apiUrl}/runs/${runId}/file`;
    return makeRequest({
        method: 'GET',
        url: fetchUrl,
        headers: createHeaders(runnerId, runnerKey, true),
        qs: { file: fileId },
        operation: 'fetch file',
        responseHandler: (response) => {
            const fileName = extractFileName(response.headers['content-disposition']);
            const fileSize = response.headers['content-length'] ||
                (response.body ? Buffer.byteLength(response.body) : 0);
            return {
                data: response.body,
                fileName,
                fileSize
            };
        },
        responseErrorHandler: (response, operation) => {
            let data = response.body;
            if (typeof data === 'string') {
                try {
                    data = JSON.parse(data);
                }
                catch (parseError) {
                    return response.body;
                }
            }
            if (data.type === 'deletedResourceError') {
                const errorObj = new Error(fileErrorMessages.deletedResourceError);
                errorObj.status = response.statusCode;
                return errorObj;
            }
            else if (data.type === 'planLimitExceeded') {
                const errorObj = new Error(fileErrorMessages.planLimitExceeded);
                errorObj.status = response.statusCode;
                return errorObj;
            }
            return handleResponseError(response, operation);
        }
    });
}
function fetchPackages({ packageId, runId, runnerId, runnerKey }) {
    const apiUrl = getApiUrl(), fetchUrl = `${apiUrl}/runs/${runId}/package`;
    return makeRequest({
        method: 'GET',
        url: fetchUrl,
        headers: createHeaders(runnerId, runnerKey, true),
        qs: { package: packageId },
        operation: 'fetch packages',
        responseHandler: (response) => {
            let parsedBody = response.body;
            if (typeof response.body === 'string') {
                try {
                    parsedBody = JSON.parse(response.body);
                }
                catch (parseError) {
                    return response.body;
                }
            }
            return parsedBody ? parsedBody.contents : null;
        },
        responseErrorHandler: (response, operation) => {
            let data = response.body;
            if (typeof data === 'string') {
                try {
                    data = JSON.parse(data);
                }
                catch (parseError) {
                    return response.body;
                }
            }
            if (data.name === 'instanceNotFoundError') {
                const errorObj = new Error('Package with id ' + packageId + ' not found');
                errorObj.status = response.statusCode;
                return errorObj;
            }
            return handleResponseError(response, operation);
        }
    });
}
function fetchReferencedRequest({ requestId, runId, runnerId, runnerKey }) {
    const apiUrl = getApiUrl(), fetchUrl = `${apiUrl}/runs/${runId}/referenced-request`, queryParams = { requestId };
    return makeRequest({
        method: 'GET',
        url: fetchUrl,
        headers: createHeaders(runnerId, runnerKey, true),
        qs: queryParams,
        operation: 'fetch referenced request',
        responseHandler: (response) => {
            let parsedBody = response.body;
            if (typeof response.body === 'string') {
                try {
                    parsedBody = JSON.parse(response.body);
                }
                catch (parseError) {
                    return response.body;
                }
            }
            return parsedBody;
        },
        responseErrorHandler: (response, operation) => {
            let data = response.body;
            if (typeof data === 'string') {
                try {
                    data = JSON.parse(data);
                }
                catch (parseError) {
                    return handleResponseError(response, operation);
                }
            }
            if (data && data.name === 'instanceNotFoundError') {
                const errorObj = new Error('Referenced request with id ' + requestId + ' not found');
                errorObj.status = response.statusCode;
                return errorObj;
            }
            return handleResponseError(response, operation);
        }
    });
}
module.exports = {
    proxifyRequest,
    fetchRun,
    finishRun,
    releaseRun,
    getDataFile,
    fetchPackages,
    fetchReferencedRequest
};
