// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/aws-codebuild.js
const PROVIDER_NAME = 'aws-codebuild';
function detect(env) {
    return Boolean(env.CODEBUILD_BUILD_ID);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.CODEBUILD_WEBHOOK_HEAD_REF?.replace('refs/heads/', ''),
        commit: {
            sha: env.CODEBUILD_RESOLVED_SOURCE_VERSION,
            sourceVersion: env.CODEBUILD_SOURCE_VERSION,
            previousCommit: env.CODEBUILD_WEBHOOK_PREV_COMMIT,
            mergeCommit: env.CODEBUILD_WEBHOOK_MERGE_COMMIT
        },
        build: {
            id: env.CODEBUILD_BUILD_ID,
            arn: env.CODEBUILD_BUILD_ARN,
            number: env.CODEBUILD_BUILD_NUMBER,
            image: env.CODEBUILD_BUILD_IMAGE,
            succeeding: env.CODEBUILD_BUILD_SUCCEEDING === '1',
            startTime: env.CODEBUILD_START_TIME ? parseInt(env.CODEBUILD_START_TIME, 10) : undefined
        },
        initiator: env.CODEBUILD_INITIATOR,
        region: env.AWS_DEFAULT_REGION || env.AWS_REGION,
        accountId: env.AWS_ACCOUNT_ID,
        webhook: env.CODEBUILD_WEBHOOK_EVENT ? {
            event: env.CODEBUILD_WEBHOOK_EVENT,
            trigger: env.CODEBUILD_WEBHOOK_TRIGGER,
            actorAccountId: env.CODEBUILD_WEBHOOK_ACTOR_ACCOUNT_ID,
            baseRef: env.CODEBUILD_WEBHOOK_BASE_REF,
            headRef: env.CODEBUILD_WEBHOOK_HEAD_REF
        } : undefined,
        pipeline: {
            id: env.CODEBUILD_BUILD_ID,
            arn: env.CODEBUILD_BUILD_ARN,
            number: env.CODEBUILD_BUILD_NUMBER,
            url: env.CODEBUILD_PUBLIC_BUILD_URL
        },
        repository: {
            url: env.CODEBUILD_SOURCE_REPO_URL
        },
        batch: env.CODEBUILD_BATCH_BUILD_IDENTIFIER ? {
            identifier: env.CODEBUILD_BATCH_BUILD_IDENTIFIER
        } : undefined,
        paths: {
            logPath: env.CODEBUILD_LOG_PATH,
            srcDir: env.CODEBUILD_SRC_DIR,
            home: env.HOME
        }
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
