// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/azure-pipelines.js
const PROVIDER_NAME = 'azure-pipelines';
function detect(env) {
    return Boolean(env.BUILD_BUILDID);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.BUILD_SOURCEBRANCHNAME,
        commit: {
            sha: env.BUILD_SOURCEVERSION,
            message: env.BUILD_SOURCEVERSION
        },
        job: {
            name: env.SYSTEM_JOBNAME,
            displayName: env.SYSTEM_JOBDISPLAYNAME,
            id: env.SYSTEM_JOBID,
            attempt: env.SYSTEM_JOBATTEMPT ? parseInt(env.SYSTEM_JOBATTEMPT, 10) : undefined,
            identifier: env.SYSTEM_JOBIDENTIFIER,
            status: env.AGENT_JOBSTATUS
        },
        stage: {
            name: env.SYSTEM_STAGENAME,
            displayName: env.SYSTEM_STAGEDISPLAYNAME,
            id: env.SYSTEM_STAGEID,
            attempt: env.SYSTEM_STAGEATTEMPT ? parseInt(env.SYSTEM_STAGEATTEMPT, 10) : undefined
        },
        phase: {
            name: env.SYSTEM_PHASENAME
        },
        build: {
            id: env.BUILD_BUILDID,
            number: env.BUILD_BUILDNUMBER,
            uri: env.BUILD_BUILDURI,
            definitionName: env.BUILD_DEFINITIONNAME,
            definitionVersion: env.BUILD_DEFINITIONVERSION,
            queuedBy: env.BUILD_QUEUEDBY,
            queuedById: env.BUILD_QUEUEDBYID,
            requestedFor: env.BUILD_REQUESTEDFOR,
            requestedForId: env.BUILD_REQUESTEDFORID,
            reason: env.BUILD_REASON
        },
        pipeline: {
            id: env.BUILD_BUILDID,
            number: env.BUILD_BUILDNUMBER,
            url: env.SYSTEM_COLLECTIONURI && env.SYSTEM_TEAMPROJECTID && env.BUILD_BUILDID ?
                `${env.SYSTEM_COLLECTIONURI}${env.SYSTEM_TEAMPROJECTID}/_build/results?` +
                    `buildId=${env.BUILD_BUILDID}` : undefined,
            startTime: env.SYSTEM_PIPELINESTARTTIME
        },
        repository: {
            id: env.BUILD_REPOSITORY_ID,
            name: env.BUILD_REPOSITORY_NAME,
            url: env.BUILD_REPOSITORY_URI,
            provider: env.BUILD_REPOSITORY_PROVIDER,
            branch: env.BUILD_SOURCEBRANCH,
            localPath: env.BUILD_REPOSITORY_LOCALPATH,
            clean: env.BUILD_REPOSITORY_CLEAN,
            gitSubmoduleCheckout: env.BUILD_REPOSITORY_GIT_SUBMODULE_CHECKOUT === 'true',
            tfvcWorkspace: env.BUILD_REPOSITORY_TFVC_WORKSPACE
        },
        pullRequest: env.SYSTEM_PULLREQUEST_PULLREQUESTID ? {
            id: env.SYSTEM_PULLREQUEST_PULLREQUESTID,
            number: env.SYSTEM_PULLREQUEST_PULLREQUESTNUMBER || env.SYSTEM_PULLREQUEST_PULLREQUESTID,
            sourceBranch: env.SYSTEM_PULLREQUEST_SOURCEBRANCH,
            sourceCommit: env.SYSTEM_PULLREQUEST_SOURCECOMMITID,
            sourceRepositoryUrl: env.SYSTEM_PULLREQUEST_SOURCEREPOSITORYURI,
            targetBranch: env.SYSTEM_PULLREQUEST_TARGETBRANCH,
            isFork: env.SYSTEM_PULLREQUEST_ISFORK === 'True'
        } : undefined,
        runner: {
            name: env.AGENT_NAME,
            machineName: env.AGENT_MACHINENAME,
            version: env.AGENT_VERSION,
            jobName: env.AGENT_JOBNAME,
            os: env.AGENT_OS,
            arch: env.AGENT_OSARCHITECTURE,
            workFolder: env.AGENT_WORKFOLDER,
            tempDirectory: env.AGENT_TEMPDIRECTORY,
            toolsDirectory: env.AGENT_TOOLSDIRECTORY
        },
        system: {
            teamFoundationCollectionUri: env.SYSTEM_TEAMFOUNDATIONCOLLECTIONURI,
            collectionUri: env.SYSTEM_COLLECTIONURI,
            teamProject: env.SYSTEM_TEAMPROJECT,
            teamProjectId: env.SYSTEM_TEAMPROJECTID,
            definitionId: env.SYSTEM_DEFINITIONID,
            hostType: env.SYSTEM_HOSTTYPE,
            defaultWorkingDirectory: env.SYSTEM_DEFAULTWORKINGDIRECTORY,
            taskInstanceId: env.SYSTEM_TASKINSTANCEID,
            taskDefinitionsUri: env.SYSTEM_TASKDEFINITIONSURI,
            totalJobsInProgress: env.SYSTEM_TOTALJOBUNINPROGRESS ?
                parseInt(env.SYSTEM_TOTALJOBUNINPROGRESS, 10) : undefined,
            parallelExecutionType: env.SYSTEM_PARALLELEXECUTIONTYPE
        },
        directories: {
            binaries: env.BUILD_BINARIESDIRECTORY,
            artifactStaging: env.BUILD_ARTIFACTSTAGINGDIRECTORY,
            sources: env.BUILD_SOURCESDIRECTORY,
            staging: env.BUILD_STAGINGDIRECTORY
        },
        environment: env.ENVIRONMENT_NAME ? {
            name: env.ENVIRONMENT_NAME,
            id: env.ENVIRONMENT_ID,
            resourceName: env.ENVIRONMENT_RESOURCENAME,
            resourceId: env.ENVIRONMENT_RESOURCEID
        } : undefined,
        release: env.RELEASE_RELEASENAME ? {
            definitionName: env.RELEASE_DEFINITIONNAME,
            definitionId: env.RELEASE_DEFINITIONID,
            releaseName: env.RELEASE_RELEASENAME,
            releaseId: env.RELEASE_RELEASEID
        } : undefined
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
