// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/bitbucket.js
const PROVIDER_NAME = 'bitbucket';
function detect(env) {
    return Boolean(env.BITBUCKET_BUILD_NUMBER);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.BITBUCKET_BRANCH,
        tag: env.BITBUCKET_TAG,
        bookmark: env.BITBUCKET_BOOKMARK,
        commit: {
            sha: env.BITBUCKET_COMMIT,
            author: env.BITBUCKET_COMMIT_AUTHOR,
            message: env.BITBUCKET_COMMIT_MESSAGE
        },
        pipeline: {
            id: env.BITBUCKET_BUILD_NUMBER,
            uuid: env.BITBUCKET_PIPELINE_UUID,
            url: env.BITBUCKET_GIT_HTTP_ORIGIN ?
                `${env.BITBUCKET_GIT_HTTP_ORIGIN}/pipelines/results/${env.BITBUCKET_BUILD_NUMBER}` : undefined
        },
        step: {
            uuid: env.BITBUCKET_STEP_UUID,
            triggererUuid: env.BITBUCKET_STEP_TRIGGERER_UUID
        },
        repository: {
            name: env.BITBUCKET_REPO_SLUG,
            fullName: env.BITBUCKET_REPO_FULL_NAME,
            id: env.BITBUCKET_REPO_UUID,
            owner: env.BITBUCKET_REPO_OWNER,
            url: env.BITBUCKET_GIT_HTTP_ORIGIN,
            isPrivate: env.BITBUCKET_REPO_IS_PRIVATE === 'true'
        },
        workspace: env.BITBUCKET_WORKSPACE,
        project: {
            key: env.BITBUCKET_PROJECT_KEY,
            uuid: env.BITBUCKET_PROJECT_UUID
        },
        pullRequest: env.BITBUCKET_PR_ID ? {
            id: env.BITBUCKET_PR_ID,
            number: env.BITBUCKET_PR_ID,
            targetBranch: env.BITBUCKET_PR_DESTINATION_BRANCH,
            targetCommit: env.BITBUCKET_PR_DESTINATION_COMMIT
        } : undefined,
        deployment: env.BITBUCKET_DEPLOYMENT_ENVIRONMENT ? {
            environment: env.BITBUCKET_DEPLOYMENT_ENVIRONMENT
        } : undefined,
        parallel: env.BITBUCKET_PARALLEL_STEP ? {
            step: env.BITBUCKET_PARALLEL_STEP,
            stepCount: env.BITBUCKET_PARALLEL_STEP_COUNT
        } : undefined,
        cloneDir: env.BITBUCKET_CLONE_DIR,
        exitCode: env.BITBUCKET_EXIT_CODE ? parseInt(env.BITBUCKET_EXIT_CODE, 10) : undefined
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
