// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/buildkite.js
const PROVIDER_NAME = 'buildkite';
function detect(env) {
    return Boolean(env.BUILDKITE);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.BUILDKITE_BRANCH,
        tag: env.BUILDKITE_TAG,
        commit: {
            sha: env.BUILDKITE_COMMIT,
            message: env.BUILDKITE_MESSAGE
        },
        job: {
            id: env.BUILDKITE_JOB_ID,
            label: env.BUILDKITE_LABEL,
            parallelJob: env.BUILDKITE_PARALLEL_JOB ?
                parseInt(env.BUILDKITE_PARALLEL_JOB, 10) : undefined,
            parallelJobCount: env.BUILDKITE_PARALLEL_JOB_COUNT ?
                parseInt(env.BUILDKITE_PARALLEL_JOB_COUNT, 10) : undefined,
            retryCount: env.BUILDKITE_RETRY_COUNT ?
                parseInt(env.BUILDKITE_RETRY_COUNT, 10) : undefined,
            timeout: env.BUILDKITE_TIMEOUT
        },
        build: {
            id: env.BUILDKITE_BUILD_ID,
            number: env.BUILDKITE_BUILD_NUMBER,
            url: env.BUILDKITE_BUILD_URL,
            author: env.BUILDKITE_BUILD_AUTHOR,
            creator: env.BUILDKITE_BUILD_CREATOR,
            creatorTeams: env.BUILDKITE_BUILD_CREATOR_TEAMS ?
                env.BUILDKITE_BUILD_CREATOR_TEAMS.split(',') : undefined,
            checkoutPath: env.BUILDKITE_BUILD_CHECKOUT_PATH,
            path: env.BUILDKITE_BUILD_PATH
        },
        organization: env.BUILDKITE_ORGANIZATION_SLUG,
        source: env.BUILDKITE_SOURCE,
        triggeredFrom: env.BUILDKITE_TRIGGERED_FROM_BUILD_ID,
        rebuiltFrom: env.BUILDKITE_REBUILT_FROM_BUILD_ID,
        pipeline: {
            id: env.BUILDKITE_BUILD_ID,
            number: env.BUILDKITE_BUILD_NUMBER,
            url: env.BUILDKITE_BUILD_URL,
            slug: env.BUILDKITE_PIPELINE_SLUG,
            provider: env.BUILDKITE_PIPELINE_PROVIDER,
            name: env.BUILDKITE_PIPELINE_NAME,
            defaultBranch: env.BUILDKITE_PIPELINE_DEFAULT_BRANCH,
            teams: env.BUILDKITE_PIPELINE_TEAMS ?
                env.BUILDKITE_PIPELINE_TEAMS.split(',') : undefined
        },
        repository: {
            url: env.BUILDKITE_REPO,
            sshHost: env.BUILDKITE_REPO_SSH_HOST
        },
        runner: {
            name: env.BUILDKITE_AGENT_NAME,
            id: env.BUILDKITE_AGENT_ID,
            pid: env.BUILDKITE_AGENT_PID ? parseInt(env.BUILDKITE_AGENT_PID, 10) : undefined,
            metaData: (() => {
                const metaData = {};
                Object.keys(env).forEach((key) => {
                    if (key.startsWith('BUILDKITE_AGENT_META_DATA_')) {
                        const metaKey = key.replace('BUILDKITE_AGENT_META_DATA_', '').toLowerCase();
                        metaData[metaKey] = env[key];
                    }
                });
                return Object.keys(metaData).length > 0 ? metaData : undefined;
            })()
        },
        pullRequest: env.BUILDKITE_PULL_REQUEST !== 'false' ? {
            number: env.BUILDKITE_PULL_REQUEST,
            repoUrl: env.BUILDKITE_PULL_REQUEST_REPO,
            targetBranch: env.BUILDKITE_PULL_REQUEST_BASE_BRANCH
        } : undefined,
        group: env.BUILDKITE_GROUP_ID ? {
            id: env.BUILDKITE_GROUP_ID,
            key: env.BUILDKITE_GROUP_KEY,
            label: env.BUILDKITE_GROUP_LABEL
        } : undefined,
        command: env.BUILDKITE_COMMAND ? {
            command: env.BUILDKITE_COMMAND,
            eval: env.BUILDKITE_COMMAND_EVAL === 'true',
            exitStatus: env.BUILDKITE_COMMAND_EXIT_STATUS ?
                parseInt(env.BUILDKITE_COMMAND_EXIT_STATUS, 10) : undefined
        } : undefined,
        paths: {
            artifactPaths: env.BUILDKITE_ARTIFACT_PATHS,
            binPath: env.BUILDKITE_BIN_PATH,
            pluginsPath: env.BUILDKITE_PLUGINS_PATH,
            hooksPath: env.BUILDKITE_HOOKS_PATH
        }
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
