// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/circleci.js
const PROVIDER_NAME = 'circleci';
function detect(env) {
    return Boolean(env.CIRCLECI);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.CIRCLE_BRANCH,
        tag: env.CIRCLE_TAG,
        commit: {
            sha: env.CIRCLE_SHA1
        },
        job: {
            name: env.CIRCLE_JOB
        },
        workflow: {
            id: env.CIRCLE_WORKFLOW_ID,
            workspaceId: env.CIRCLE_WORKFLOW_WORKSPACE_ID
        },
        pipeline: {
            id: env.CIRCLE_PIPELINE_ID,
            number: env.CIRCLE_PIPELINE_NUMBER,
            buildNum: env.CIRCLE_BUILD_NUM,
            url: env.CIRCLE_BUILD_URL,
            compareUrl: env.CIRCLE_COMPARE_URL,
            repoUrl: env.CIRCLE_PIPELINE_REPO_URL,
            repoFullName: env.CIRCLE_PIPELINE_REPO_FULL_NAME,
            triggerLogin: env.CIRCLE_PIPELINE_TRIGGER_LOGIN
        },
        project: {
            id: env.CIRCLE_PROJECT_ID,
            name: env.CIRCLE_PROJECT_REPONAME,
            owner: env.CIRCLE_PROJECT_USERNAME
        },
        repository: {
            url: env.CIRCLE_REPOSITORY_URL,
            name: env.CIRCLE_PROJECT_REPONAME,
            owner: env.CIRCLE_PROJECT_USERNAME
        },
        pullRequest: env.CIRCLE_PULL_REQUEST || env.CIRCLE_PR_NUMBER ? {
            url: env.CIRCLE_PULL_REQUEST,
            number: env.CIRCLE_PR_NUMBER,
            repoName: env.CIRCLE_PR_REPONAME,
            username: env.CIRCLE_PR_USERNAME,
            allUrls: env.CIRCLE_PULL_REQUESTS ?
                env.CIRCLE_PULL_REQUESTS.split(',') : undefined
        } : undefined,
        parallel: env.CIRCLE_NODE_TOTAL ? {
            index: env.CIRCLE_NODE_INDEX,
            total: env.CIRCLE_NODE_TOTAL
        } : undefined,
        username: env.CIRCLE_USERNAME,
        previousBuildNum: env.CIRCLE_PREVIOUS_BUILD_NUM,
        workingDirectory: env.CIRCLE_WORKING_DIRECTORY,
        stage: env.CIRCLE_STAGE
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
