// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/github-actions.js
const fs = require('fs'), path = require('path'), CodeOwners = require('codeowners'), PROVIDER_NAME = 'github-actions';
function detect(env) {
    return Boolean(env.GITHUB_ACTION);
}
function getCodeOwner(repoRoot = process.cwd()) {
    try {
        const codeOwnerFileCandidates = [
            path.join(repoRoot, '.github', 'CODEOWNERS'),
            path.join(repoRoot, 'CODEOWNERS'),
            path.join(repoRoot, 'docs', 'CODEOWNERS')
        ], selectedCodeOwnerFilePath = codeOwnerFileCandidates.find(fs.existsSync);
        if (!selectedCodeOwnerFilePath) {
            return null;
        }
        let repo = new CodeOwners();
        return repo.getOwner(selectedCodeOwnerFilePath)[0] || null;
    }
    catch {
        return null;
    }
}
function getGitHubEvent(env) {
    try {
        if (!env.GITHUB_EVENT_PATH || !fs.existsSync(env.GITHUB_EVENT_PATH)) {
            return null;
        }
        const eventData = fs.readFileSync(env.GITHUB_EVENT_PATH, 'utf8');
        return JSON.parse(eventData);
    }
    catch {
        return null;
    }
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.GITHUB_REF_NAME || env.GITHUB_HEAD_REF || env.GITHUB_REF?.replace('refs/heads/', ''),
        tag: env.GITHUB_REF_TYPE === 'tag' ? env.GITHUB_REF_NAME : undefined,
        commit: {
            sha: env.GITHUB_SHA,
            shortSha: env.GITHUB_SHA?.substring(0, 7),
            headSha: env.GITHUB_HEAD_SHA
        },
        workflow: {
            name: env.GITHUB_WORKFLOW,
            ref: env.GITHUB_REF,
            refType: env.GITHUB_REF_TYPE,
            workflowRef: env.GITHUB_WORKFLOW_REF,
            workflowSha: env.GITHUB_WORKFLOW_SHA
        },
        job: {
            name: env.GITHUB_JOB
        },
        actor: env.GITHUB_ACTOR,
        triggeredBy: env.GITHUB_EVENT_NAME,
        triggeringActor: env.GITHUB_TRIGGERING_ACTOR,
        pipeline: {
            id: env.GITHUB_RUN_ID,
            number: env.GITHUB_RUN_NUMBER,
            attempt: env.GITHUB_RUN_ATTEMPT,
            url: env.GITHUB_SERVER_URL && env.GITHUB_REPOSITORY && env.GITHUB_RUN_ID ?
                `${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}/actions/runs/${env.GITHUB_RUN_ID}` : undefined
        },
        repository: {
            fullName: env.GITHUB_REPOSITORY,
            name: env.GITHUB_REPOSITORY?.split('/')[1],
            owner: env.GITHUB_REPOSITORY_OWNER,
            id: env.GITHUB_REPOSITORY_ID,
            ownerId: env.GITHUB_REPOSITORY_OWNER_ID,
            code_owner: getCodeOwner()
        },
        pullRequest: env.GITHUB_EVENT_NAME === 'pull_request' || env.GITHUB_HEAD_REF ? (() => {
            const githubEvent = getGitHubEvent(env), prData = githubEvent?.pull_request;
            return {
                number: prData?.number || env.GITHUB_PR_NUMBER || env.GITHUB_REF?.match(/refs\/pull\/(\d+)\//)?.[1],
                sourceBranch: env.GITHUB_HEAD_REF || prData?.head?.ref,
                targetBranch: env.GITHUB_BASE_REF || prData?.base?.ref,
                baseRef: env.GITHUB_BASE_REF || prData?.base?.ref,
                headRef: env.GITHUB_HEAD_REF || prData?.head?.ref,
                headSha: env.GITHUB_HEAD_SHA || prData?.head?.sha,
                baseSha: prData?.base?.sha,
                title: prData?.title || env.GITHUB_PR_TITLE,
                status: prData?.state || 'unknown',
                draft: prData?.draft || false,
                merged: prData?.merged || false,
                mergeable: prData?.mergeable,
                user: prData?.user?.login || env.GITHUB_ACTOR,
                url: prData?.html_url
            };
        })() : undefined,
        runner: {
            os: env.RUNNER_OS,
            arch: env.RUNNER_ARCH,
            name: env.RUNNER_NAME,
            environment: env.RUNNER_ENVIRONMENT,
            temp: env.RUNNER_TEMP,
            toolCache: env.RUNNER_TOOL_CACHE,
            workspace: env.RUNNER_WORKSPACE,
            debug: env.RUNNER_DEBUG === '1'
        },
        server: {
            url: env.GITHUB_SERVER_URL,
            apiUrl: env.GITHUB_API_URL,
            graphqlUrl: env.GITHUB_GRAPHQL_URL
        },
        action: env.GITHUB_ACTION ? {
            name: env.GITHUB_ACTION,
            path: env.GITHUB_ACTION_PATH,
            repository: env.GITHUB_ACTION_REPOSITORY
        } : undefined,
        retentionDays: env.GITHUB_RETENTION_DAYS ? parseInt(env.GITHUB_RETENTION_DAYS, 10) : undefined,
        paths: {
            env: env.GITHUB_ENV,
            path: env.GITHUB_PATH,
            output: env.GITHUB_OUTPUT,
            state: env.GITHUB_STATE,
            stepSummary: env.GITHUB_STEP_SUMMARY
        }
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
