// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/gitlab.js
const PROVIDER_NAME = 'gitlab';
function detect(env) {
    return Boolean(env.GITLAB_CI);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.CI_COMMIT_BRANCH || env.CI_COMMIT_REF_NAME,
        tag: env.CI_COMMIT_TAG,
        commit: {
            sha: env.CI_COMMIT_SHA,
            shortSha: env.CI_COMMIT_SHORT_SHA,
            beforeSha: env.CI_COMMIT_BEFORE_SHA,
            message: env.CI_COMMIT_MESSAGE,
            title: env.CI_COMMIT_TITLE,
            description: env.CI_COMMIT_DESCRIPTION,
            author: env.CI_COMMIT_AUTHOR,
            timestamp: env.CI_COMMIT_TIMESTAMP,
            refName: env.CI_COMMIT_REF_NAME,
            refSlug: env.CI_COMMIT_REF_SLUG,
            refProtected: env.CI_COMMIT_REF_PROTECTED === 'true'
        },
        job: {
            id: env.CI_JOB_ID,
            name: env.CI_JOB_NAME,
            stage: env.CI_JOB_STAGE,
            url: env.CI_JOB_URL,
            startedAt: env.CI_JOB_STARTED_AT,
            status: env.CI_JOB_STATUS,
            manual: env.CI_JOB_MANUAL === 'true'
        },
        pipeline: {
            id: env.CI_PIPELINE_ID,
            iid: env.CI_PIPELINE_IID,
            url: env.CI_PIPELINE_URL,
            source: env.CI_PIPELINE_SOURCE,
            createdAt: env.CI_PIPELINE_CREATED_AT,
            triggered: env.CI_PIPELINE_TRIGGERED === 'true'
        },
        project: {
            id: env.CI_PROJECT_ID,
            name: env.CI_PROJECT_NAME,
            title: env.CI_PROJECT_TITLE,
            path: env.CI_PROJECT_PATH,
            pathSlug: env.CI_PROJECT_PATH_SLUG,
            namespace: env.CI_PROJECT_NAMESPACE,
            rootNamespace: env.CI_PROJECT_ROOT_NAMESPACE,
            url: env.CI_PROJECT_URL,
            dir: env.CI_PROJECT_DIR,
            visibility: env.CI_PROJECT_VISIBILITY,
            repositoryLanguages: env.CI_PROJECT_REPOSITORY_LANGUAGES ?
                env.CI_PROJECT_REPOSITORY_LANGUAGES.split(',') : undefined
        },
        repository: {
            url: env.CI_REPOSITORY_URL
        },
        mergeRequest: env.CI_MERGE_REQUEST_ID ? {
            id: env.CI_MERGE_REQUEST_ID,
            iid: env.CI_MERGE_REQUEST_IID,
            number: env.CI_MERGE_REQUEST_IID,
            title: env.CI_MERGE_REQUEST_TITLE,
            description: env.CI_MERGE_REQUEST_DESCRIPTION,
            targetBranch: env.CI_MERGE_REQUEST_TARGET_BRANCH_NAME,
            sourceBranch: env.CI_MERGE_REQUEST_SOURCE_BRANCH_NAME,
            refPath: env.CI_MERGE_REQUEST_REF_PATH,
            projectId: env.CI_MERGE_REQUEST_PROJECT_ID,
            projectPath: env.CI_MERGE_REQUEST_PROJECT_PATH,
            projectUrl: env.CI_MERGE_REQUEST_PROJECT_URL,
            approved: env.CI_MERGE_REQUEST_APPROVED === 'true',
            assignees: env.CI_MERGE_REQUEST_ASSIGNEES ?
                env.CI_MERGE_REQUEST_ASSIGNEES.split(',') : undefined,
            labels: env.CI_MERGE_REQUEST_LABELS ?
                env.CI_MERGE_REQUEST_LABELS.split(',') : undefined,
            milestone: env.CI_MERGE_REQUEST_MILESTONE,
            eventType: env.CI_MERGE_REQUEST_EVENT_TYPE
        } : undefined,
        externalPullRequest: env.CI_EXTERNAL_PULL_REQUEST_IID ? {
            iid: env.CI_EXTERNAL_PULL_REQUEST_IID,
            sourceBranch: env.CI_EXTERNAL_PULL_REQUEST_SOURCE_BRANCH_NAME,
            sourceBranchSha: env.CI_EXTERNAL_PULL_REQUEST_SOURCE_BRANCH_SHA,
            targetBranch: env.CI_EXTERNAL_PULL_REQUEST_TARGET_BRANCH_NAME,
            targetBranchSha: env.CI_EXTERNAL_PULL_REQUEST_TARGET_BRANCH_SHA
        } : undefined,
        environment: env.CI_ENVIRONMENT_NAME ? {
            name: env.CI_ENVIRONMENT_NAME,
            slug: env.CI_ENVIRONMENT_SLUG,
            url: env.CI_ENVIRONMENT_URL,
            action: env.CI_ENVIRONMENT_ACTION,
            tier: env.CI_ENVIRONMENT_TIER
        } : undefined,
        deploy: env.CI_DEPLOY_FREEZE ? {
            user: env.CI_DEPLOY_USER,
            freeze: env.CI_DEPLOY_FREEZE === 'true'
        } : undefined,
        release: env.CI_RELEASE_DESCRIPTION ? {
            description: env.CI_RELEASE_DESCRIPTION
        } : undefined,
        runner: {
            id: env.CI_RUNNER_ID,
            description: env.CI_RUNNER_DESCRIPTION,
            tags: env.CI_RUNNER_TAGS ? env.CI_RUNNER_TAGS.split(',') : undefined,
            version: env.CI_RUNNER_VERSION,
            revision: env.CI_RUNNER_REVISION,
            arch: env.CI_RUNNER_EXECUTABLE_ARCH
        },
        server: {
            name: env.CI_SERVER_NAME,
            host: env.CI_SERVER_HOST,
            port: env.CI_SERVER_PORT,
            protocol: env.CI_SERVER_PROTOCOL,
            url: env.CI_SERVER_URL,
            revision: env.CI_SERVER_REVISION,
            version: env.CI_SERVER_VERSION,
            versionMajor: env.CI_SERVER_VERSION_MAJOR,
            versionMinor: env.CI_SERVER_VERSION_MINOR,
            versionPatch: env.CI_SERVER_VERSION_PATCH
        },
        user: env.GITLAB_USER_LOGIN ? {
            id: env.GITLAB_USER_ID,
            login: env.GITLAB_USER_LOGIN,
            name: env.GITLAB_USER_NAME
        } : undefined
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
