// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/google-cloud-build.js
const PROVIDER_NAME = 'google-cloud-build';
function detect(env) {
    return Boolean(env.PROJECT_ID) && Boolean(env.BUILD_ID &&
        (env.BUILDER_OUTPUT || env.COMMIT_SHA || env.LOCATION || env.TRIGGER_NAME));
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.BRANCH_NAME,
        tag: env.TAG_NAME,
        commit: {
            sha: env.COMMIT_SHA,
            shortSha: env.SHORT_SHA
        },
        revision: env.REVISION_ID,
        project: {
            id: env.PROJECT_ID,
            number: env.PROJECT_NUMBER
        },
        location: env.LOCATION,
        trigger: env.TRIGGER_NAME ? {
            name: env.TRIGGER_NAME,
            buildConfigPath: env.TRIGGER_BUILD_CONFIG_PATH
        } : undefined,
        build: {
            id: env.BUILD_ID,
            status: env.BUILD_STATUS
        },
        pipeline: {
            id: env.BUILD_ID
        },
        repository: {
            name: env.REPO_NAME,
            fullName: env.REPO_FULL_NAME
        },
        pullRequest: env._PR_NUMBER ? {
            number: env._PR_NUMBER,
            sourceBranch: env._HEAD_BRANCH,
            targetBranch: env._BASE_BRANCH,
            sourceRepoUrl: env._HEAD_REPO_URL
        } : undefined,
        serviceAccount: env.SERVICE_ACCOUNT,
        substitutions: {
            repoUrl: env._REPO_URL,
            headBranch: env._HEAD_BRANCH,
            baseBranch: env._BASE_BRANCH,
            headRepoUrl: env._HEAD_REPO_URL,
            prNumber: env._PR_NUMBER,
            branchName: env._BRANCH_NAME,
            tagName: env._TAG_NAME,
            buildId: env._BUILD_ID
        }
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
