// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/jenkins.js
const PROVIDER_NAME = 'jenkins';
function detect(env) {
    return Boolean(env.JENKINS_URL);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.BRANCH_NAME || env.GIT_BRANCH,
        commit: {
            sha: env.GIT_COMMIT,
            previousSha: env.GIT_PREVIOUS_COMMIT,
            previousSuccessfulSha: env.GIT_PREVIOUS_SUCCESSFUL_COMMIT,
            author: env.GIT_AUTHOR_NAME ? {
                name: env.GIT_AUTHOR_NAME
            } : undefined,
            committer: env.GIT_COMMITTER_NAME ? {
                name: env.GIT_COMMITTER_NAME
            } : undefined
        },
        job: {
            name: env.JOB_NAME,
            baseName: env.JOB_BASE_NAME,
            displayUrl: env.JOB_DISPLAY_URL,
            url: env.JOB_URL,
            tag: env.BUILD_TAG
        },
        build: {
            id: env.BUILD_ID,
            number: env.BUILD_NUMBER,
            displayName: env.BUILD_DISPLAY_NAME,
            tag: env.BUILD_TAG,
            url: env.BUILD_URL
        },
        pipeline: {
            id: env.BUILD_ID,
            number: env.BUILD_NUMBER,
            url: env.BUILD_URL
        },
        repository: {
            url: env.GIT_URL,
            localBranch: env.GIT_LOCAL_BRANCH,
            urls: (() => {
                const urls = [];
                let i = 1;
                while (env[`GIT_URL_${i}`]) {
                    urls.push(env[`GIT_URL_${i}`]);
                    i++;
                }
                return urls.length > 0 ? urls : undefined;
            })()
        },
        pullRequest: env.CHANGE_ID ? {
            id: env.CHANGE_ID,
            number: env.CHANGE_ID,
            sourceBranch: env.CHANGE_BRANCH,
            targetBranch: env.CHANGE_TARGET,
            fork: env.CHANGE_FORK,
            title: env.CHANGE_TITLE,
            url: env.CHANGE_URL,
            author: env.CHANGE_AUTHOR,
            authorDisplayName: env.CHANGE_AUTHOR_DISPLAY_NAME
        } : undefined,
        runner: {
            name: env.NODE_NAME,
            labels: env.NODE_LABELS ? env.NODE_LABELS.split(' ') : undefined
        },
        executor: {
            number: env.EXECUTOR_NUMBER
        },
        workspace: env.WORKSPACE ? {
            path: env.WORKSPACE,
            tmp: env.WORKSPACE_TMP
        } : undefined,
        server: {
            url: env.JENKINS_URL,
            home: env.JENKINS_HOME
        },
        run: env.RUN_DISPLAY_URL ? {
            displayUrl: env.RUN_DISPLAY_URL,
            artifactsDisplayUrl: env.RUN_ARTIFACTS_DISPLAY_URL,
            changesDisplayUrl: env.RUN_CHANGES_DISPLAY_URL,
            testsDisplayUrl: env.RUN_TESTS_DISPLAY_URL
        } : undefined
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
