// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-providers/travis.js
const PROVIDER_NAME = 'travis';
function detect(env) {
    return Boolean(env.TRAVIS);
}
function getInfo(env) {
    return {
        provider: PROVIDER_NAME,
        branch: env.TRAVIS_BRANCH,
        tag: env.TRAVIS_TAG,
        commit: {
            sha: env.TRAVIS_COMMIT,
            message: env.TRAVIS_COMMIT_MESSAGE,
            range: env.TRAVIS_COMMIT_RANGE
        },
        job: {
            id: env.TRAVIS_JOB_ID,
            name: env.TRAVIS_JOB_NAME,
            number: env.TRAVIS_JOB_NUMBER,
            url: env.TRAVIS_JOB_WEB_URL
        },
        build: {
            id: env.TRAVIS_BUILD_ID,
            number: env.TRAVIS_BUILD_NUMBER,
            url: env.TRAVIS_BUILD_WEB_URL,
            dir: env.TRAVIS_BUILD_DIR,
            stageName: env.TRAVIS_BUILD_STAGE_NAME
        },
        eventType: env.TRAVIS_EVENT_TYPE,
        pipeline: {
            id: env.TRAVIS_BUILD_ID,
            number: env.TRAVIS_BUILD_NUMBER,
            url: env.TRAVIS_BUILD_WEB_URL
        },
        repository: {
            id: env.TRAVIS_REPO_ID,
            fullName: env.TRAVIS_REPO_SLUG,
            name: env.TRAVIS_REPO_NAME
        },
        pullRequest: env.TRAVIS_PULL_REQUEST !== 'false' ? {
            number: env.TRAVIS_PULL_REQUEST,
            sourceBranch: env.TRAVIS_PULL_REQUEST_BRANCH,
            repoSlug: env.TRAVIS_PULL_REQUEST_SLUG,
            sha: env.TRAVIS_PULL_REQUEST_SHA
        } : undefined,
        os: env.TRAVIS_OS_NAME,
        osVersion: env.TRAVIS_OS_VERSION,
        dist: env.TRAVIS_DIST,
        language: env.TRAVIS_LANGUAGE,
        sudo: env.TRAVIS_SUDO === 'true',
        secureEnvVars: env.TRAVIS_SECURE_ENV_VARS === 'true',
        testResult: env.TRAVIS_TEST_RESULT ? parseInt(env.TRAVIS_TEST_RESULT, 10) : undefined,
        allowFailure: env.TRAVIS_ALLOW_FAILURE === 'true'
    };
}
module.exports = {
    name: PROVIDER_NAME,
    detect: detect,
    getInfo: getInfo
};
