// @postman/pm-bin-linux-x64/bin/lib/commands/performance/output/blessed-output.js
const colors = require('colors/safe');
class BlessedOutput {
    constructor(options = {}) {
        this.options = options;
        this.silent = options.silent || false;
        this.verbose = options.verbose || false;
        this.stats = {
            startTime: Date.now(),
            requests: 0,
            errors: 0,
            avgResponseTime: 0,
            currentVUs: 0,
            progress: 0
        };
        this.lastUpdate = 0;
        this.updateInterval = 500;
        this.logBuffer = [];
        this.maxLogs = 5;
        if (!this.silent) {
            this.initScreen();
        }
    }
    initScreen() {
        const blessed = require('blessed');
        this.screen = blessed.screen({
            smartCSR: true,
            title: 'Postman Performance Test',
            fullUnicode: true
        });
        this.headerBox = blessed.box({
            top: 0,
            left: 0,
            width: '100%',
            height: 8,
            content: '',
            tags: true
        });
        this.progressBox = blessed.box({
            top: 7,
            left: 0,
            width: '100%',
            height: 2,
            content: '',
            tags: true
        });
        this.statsBox = blessed.box({
            top: 9,
            left: 0,
            width: '100%',
            height: 6,
            content: '',
            tags: true
        });
        this.logsBox = blessed.box({
            top: 14,
            left: 0,
            width: '100%',
            height: 15,
            content: '',
            tags: true,
            scrollable: true,
            alwaysScroll: true
        });
        this.screen.append(this.logsBox);
        this.screen.append(this.headerBox);
        this.screen.append(this.progressBox);
        this.screen.append(this.statsBox);
        this.screen.key(['escape', 'q', 'C-c'], () => {
            this.destroy();
            process.exit(0);
        });
        this.screen.render();
    }
    printHeader(config) {
        if (this.silent)
            return;
        const { collectionId, loadProfile, vuCount, duration, environment, globals, dataFile } = config;
        let content = `{bold}Performance Test{/bold}\n\n`;
        content += `{bold}Collection:{/bold}  ${collectionId}\n`;
        content += `{bold}Load Profile:{/bold} ${loadProfile}\n`;
        content += `{bold}Peak Virtual Users:{/bold} ${vuCount}\n`;
        content += `{bold}Duration:{/bold}      ${duration} minute(s)\n`;
        if (dataFile) {
            content += `{bold}Data File:{/bold}     ${dataFile}\n`;
        }
        if (environment) {
            content += `{bold}Environment:{/bold}   ${environment.substring(0, 50)}...\n`;
        }
        if (globals) {
            content += `{bold}Globals:{/bold}       ${globals.substring(0, 50)}...\n`;
        }
        this.headerBox.setContent(content);
        const headerLines = content.split('\n').length;
        this.headerBox.height = headerLines + 1;
        this.progressBox.top = headerLines + 1;
        this.statsBox.top = headerLines + 3;
        this.logsBox.top = headerLines + 8;
        this.screen.render();
    }
    printInit() {
        if (this.silent)
            return;
        if (this.logsBox) {
            this.addLog('Initializing performance test...', 'info');
        }
    }
    printStart() {
        if (this.silent)
            return;
        if (this.logsBox) {
            this.addLog('Performance test started', 'info');
        }
    }
    updateProgress(metrics = {}) {
        if (this.silent)
            return;
        const now = Date.now();
        if (now - this.lastUpdate < this.updateInterval) {
            return;
        }
        this.lastUpdate = now;
        this.stats.requests = metrics.totalRequests || this.stats.requests;
        this.stats.errors = metrics.errorCount || this.stats.errors;
        this.stats.avgResponseTime = metrics.avgResponseTime || this.stats.avgResponseTime;
        this.stats.currentVUs = metrics.currentVUs || this.stats.currentVUs;
        this.stats.progress = metrics.progress || this.stats.progress;
        this.renderProgressBar();
        this.renderStats();
        this.screen.render();
    }
    renderProgressBar() {
        const progress = Math.floor(this.stats.progress);
        const barWidth = 50;
        const filled = Math.floor((progress / 100) * barWidth);
        const empty = barWidth - filled;
        const bar = '{green-fg}' + '█'.repeat(filled) + '{/green-fg}' +
            '{gray-fg}' + '░'.repeat(empty) + '{/gray-fg}';
        const content = `  ${bar}  {bold}{white-fg}${progress}%{/white-fg}{/bold}`;
        this.progressBox.setContent(content);
    }
    renderStats() {
        const elapsed = this.stats.startTime ?
            Math.floor((Date.now() - this.stats.startTime) / 1000) : 0;
        const mins = Math.floor(elapsed / 60);
        const secs = elapsed % 60;
        const timeStr = `${String(mins).padStart(2, '0')}:${String(secs).padStart(2, '0')}`;
        const reqsPerSec = elapsed > 0 ? (this.stats.requests / elapsed).toFixed(2) : '0.00';
        const errorRate = this.stats.requests > 0 ?
            ((this.stats.errors / this.stats.requests) * 100).toFixed(2) : '0.00';
        let content = '';
        content += `{bold}Requests Per Second:{/bold} ${reqsPerSec}\n`;
        content += `{bold}Total Requests:{/bold} ${this.stats.requests}\n`;
        if (this.stats.errors > 0) {
            content += `{bold}Errors:{/bold}    {red-fg}${this.stats.errors}{/red-fg}  {gray-fg}(${errorRate}%){/gray-fg}\n`;
        }
        else {
            content += `{bold}Errors:{/bold}    ${this.stats.errors}  {gray-fg}(${errorRate}%){/gray-fg}\n`;
        }
        content += `{bold}Avg Response Time:{/bold} ${Math.floor(this.stats.avgResponseTime)}ms  {gray-fg}VUs: ${this.stats.currentVUs}{/gray-fg}\n`;
        content += `{bold}Elapsed:{/bold}   ${timeStr}`;
        this.statsBox.setContent(content);
    }
    printLog(message, level = 'info') {
        if (this.silent || !this.logsBox)
            return;
        this.addLog(message, level);
    }
    addLog(message, level = 'info') {
        if (!this.logsBox)
            return;
        if (!message || message.trim() === '')
            return;
        if (!this.verbose && level === 'debug') {
            return;
        }
        if (message.includes('collectionUID')) {
            return;
        }
        const timestamp = new Date().toLocaleTimeString();
        const logEntry = `[${timestamp}] ${message}`;
        this.logBuffer.push(logEntry);
        if (this.logBuffer.length > 100) {
            this.logBuffer.shift();
        }
        this.renderLogs();
    }
    renderLogs() {
        if (!this.logsBox)
            return;
        const content = this.logBuffer.join('\n');
        this.logsBox.setContent(content);
        this.logsBox.setScrollPerc(100);
        if (this.screen) {
            this.screen.render();
        }
    }
    printSystemMetricsBreached(data) {
        if (this.silent)
            return;
        const { metric, value, threshold, type } = data;
        if (type === 'warning') {
            this.printLog(`System ${metric}: ${value} (threshold: ${threshold})`, 'warn');
        }
        else if (type === 'critical') {
            this.printLog(`CRITICAL: ${metric} = ${value} (threshold: ${threshold})`, 'error');
            this.printLog('Stopping performance test...', 'error');
        }
    }
    printComplete() {
        if (this.silent)
            return;
        this.destroy();
    }
    printError(error) {
        if (this.silent)
            return;
        this.destroy();
        console.log(colors.red('\n✗ Performance test failed'));
        console.log(colors.red(`  ${error.message || error}\n`));
    }
    destroy() {
        if (this.screen) {
            this.screen.destroy();
            this.screen = null;
        }
    }
}
module.exports = BlessedOutput;
