// @postman/pm-bin-linux-x64/bin/lib/commands/performance/output/simple-output.js
const colors = require('colors/safe');
function clearCurrentLine() {
    if (process.stdout.isTTY) {
        process.stdout.write('\r\x1b[K');
    }
}
class SimpleOutput {
    constructor(options = {}) {
        this.options = options;
        this.silent = options.silent || false;
        this.verbose = options.verbose || false;
        this.stats = {
            startTime: Date.now(),
            requests: 0,
            errors: 0,
            avgResponseTime: 0,
            currentVUs: 0,
            progress: 0
        };
        this.lastUpdate = 0;
        this.updateInterval = 500;
        this.updateTimer = null;
        this.sigintHandler = null;
        this.headerPrinted = false;
        this.setupSignalHandlers();
        if (!this.silent) {
            this.initScreen();
        }
    }
    setupSignalHandlers() {
        this.sigintHandler = () => {
            this.destroy();
            process.exit(0);
        };
        process.on('SIGINT', this.sigintHandler);
    }
    initScreen() {
        this.startUpdate();
    }
    startUpdate() {
        if (this.updateTimer) {
            clearInterval(this.updateTimer);
        }
        this.updateTimer = setInterval(() => {
            this.updateDisplay();
        }, this.updateInterval);
    }
    updateDisplay() {
        if (!process.stdout.isTTY || !this.headerPrinted) {
            return;
        }
        const elapsed = this.stats.startTime ?
            Math.floor((Date.now() - this.stats.startTime) / 1000) : 0, mins = Math.floor(elapsed / 60), secs = elapsed % 60, timeStr = `${String(mins).padStart(2, '0')}:${String(secs).padStart(2, '0')}`;
        process.stdout.write(`\x1b[3A\r\x1b[KElapsed: ${timeStr}\x1b[3B`);
    }
    printHeader(config) {
        if (this.silent) {
            return;
        }
        const { collectionId, loadProfile, vuCount, duration, environment, globals, dataFile } = config;
        console.log(colors.bold('\nPerformance Test'));
        console.log(`Collection: ${collectionId}`);
        console.log(`Load Profile: ${loadProfile}`);
        console.log(`Peak Virtual Users: ${vuCount}`);
        console.log(`Duration: ${duration} minute(s)`);
        if (dataFile) {
            console.log(`Data File: ${dataFile}`);
        }
        if (environment) {
            const envDisplay = environment.length > 50 ? `${environment.substring(0, 50)}...` : environment;
            console.log(`Environment: ${envDisplay}`);
        }
        if (globals) {
            const globalsDisplay = globals.length > 50 ? `${globals.substring(0, 50)}...` : globals;
            console.log(`Globals: ${globalsDisplay}`);
        }
        console.log('Elapsed: 00:00');
        console.log('\n' + colors.bold('Performance testing in progress...'));
        this.headerPrinted = true;
    }
    printInit() {
    }
    printStart() {
    }
    updateProgress(metrics = {}) {
        if (this.silent) {
            return;
        }
        const now = Date.now();
        if (now - this.lastUpdate < this.updateInterval) {
            return;
        }
        this.lastUpdate = now;
        this.stats.requests = metrics.totalRequests || this.stats.requests;
        this.stats.errors = metrics.errorCount || this.stats.errors;
        this.stats.avgResponseTime = metrics.avgResponseTime || this.stats.avgResponseTime;
        this.stats.currentVUs = metrics.currentVUs || this.stats.currentVUs;
        this.stats.progress = metrics.progress || this.stats.progress;
    }
    printLog() {
    }
    printSystemMetricsBreached(data) {
        if (this.silent) {
            return;
        }
        const { metric, value, threshold, type } = data;
        if (type === 'warning') {
            this.printLog(`System ${metric}: ${value} (threshold: ${threshold})`, 'warn');
        }
        else if (type === 'critical') {
            this.printLog(`CRITICAL: ${metric} = ${value} (threshold: ${threshold})`, 'error');
            this.printLog('Stopping performance test...', 'error');
        }
    }
    printComplete() {
        if (this.updateTimer) {
            clearInterval(this.updateTimer);
            this.updateTimer = null;
        }
        if (this.headerPrinted && process.stdout.isTTY) {
            process.stdout.write('\x1b[3A\r\x1b[K\x1b[3B');
        }
    }
    printError(error) {
        if (this.silent) {
            return;
        }
        if (this.updateTimer) {
            clearInterval(this.updateTimer);
            this.updateTimer = null;
        }
        if (this.headerPrinted && process.stdout.isTTY) {
            process.stdout.write('\x1b[3A\r\x1b[K\x1b[3B');
        }
        console.log(colors.red('\n✗ Performance test failed'));
        console.log(colors.red(`  ${error.message || error}\n`));
    }
    destroy() {
        if (this.updateTimer) {
            clearInterval(this.updateTimer);
            this.updateTimer = null;
        }
        if (this.sigintHandler) {
            process.removeListener('SIGINT', this.sigintHandler);
            this.sigintHandler = null;
        }
        if (process.stdout.isTTY) {
            try {
                process.stdout.write('\x1b[2J\x1b[H');
            }
            catch (err) {
                clearCurrentLine();
            }
        }
    }
}
module.exports = SimpleOutput;
