// @postman/pm-bin-linux-x64/bin/lib/commands/performance/pass-if.js
const VALID_FUNCTIONS = ['less_than', 'greater_than', 'less_than_eq', 'greater_than_eq'];
const VALID_METRICS = ['avg', 'p90', 'p95', 'p99', 'error_rate', 'rps'];
const OPERATOR_MAP = {
    'less_than': 'lt',
    'greater_than': 'gt',
    'less_than_eq': 'lte',
    'greater_than_eq': 'gte'
};
const METRIC_NAME_MAP = {
    'avg': 'responseTime.average',
    'p90': 'responseTime.p90',
    'p95': 'responseTime.p95',
    'p99': 'responseTime.p99',
    'error_rate': 'errorRate',
    'failure_rate': 'testFailures.percentage',
    'rps': 'throughput'
};
function parsePassIfCondition(condition) {
    if (!condition || typeof condition !== 'string') {
        throw new Error('Pass-if condition must be a non-empty string');
    }
    const trimmed = condition.trim();
    const match = trimmed.match(/^(\w+)\((\w+),\s*(\d+(?:\.\d+)?)\)$/);
    if (!match) {
        throw new Error(`Invalid pass-if condition format: "${condition}". ` +
            'Expected format: function(metric, value). ' +
            'Example: less_than(p95, 500)');
    }
    const [, func, metric, value] = match;
    if (!VALID_FUNCTIONS.includes(func)) {
        throw new Error(`Invalid function: "${func}". ` +
            `Valid functions: ${VALID_FUNCTIONS.join(', ')}`);
    }
    if (!VALID_METRICS.includes(metric)) {
        throw new Error(`Invalid metric: "${metric}". ` +
            `Valid metrics: ${VALID_METRICS.join(', ')}`);
    }
    return {
        function: func,
        metric: metric,
        value: parseFloat(value)
    };
}
function getMetricValue(metrics, metricName) {
    if (!metrics) {
        return null;
    }
    if (['avg', 'p90', 'p95', 'p99'].includes(metricName)) {
        if (!metrics.responseTime) {
            return null;
        }
        const mapping = {
            'avg': 'avg',
            'p90': 'percentile90',
            'p95': 'percentile95',
            'p99': 'percentile99'
        };
        const key = mapping[metricName];
        return metrics.responseTime[key] !== undefined ? metrics.responseTime[key] : null;
    }
    if (metricName === 'error_rate') {
        return metrics.errors?.percentage !== undefined ? metrics.errors.percentage : null;
    }
    if (metricName === 'failure_rate') {
        return metrics.testFailures?.percentage !== undefined ? metrics.testFailures.percentage : null;
    }
    if (metricName === 'rps') {
        return metrics.requestThroughput?.rate !== undefined ? metrics.requestThroughput.rate : null;
    }
    return null;
}
function evaluatePassIfCondition(condition, metrics) {
    const actualValue = getMetricValue(metrics, condition.metric);
    if (actualValue === null) {
        return {
            passed: false,
            actualValue: null,
            expectedValue: condition.value,
            metric: condition.metric,
            operator: condition.function,
            message: `Metric ${condition.metric} not found in results`
        };
    }
    let passed = false;
    let operatorSymbol = '';
    switch (condition.function) {
        case 'less_than':
            passed = actualValue < condition.value;
            operatorSymbol = '<';
            break;
        case 'less_than_eq':
            passed = actualValue <= condition.value;
            operatorSymbol = '<=';
            break;
        case 'greater_than':
            passed = actualValue > condition.value;
            operatorSymbol = '>';
            break;
        case 'greater_than_eq':
            passed = actualValue >= condition.value;
            operatorSymbol = '>=';
            break;
        default:
            return {
                passed: false,
                actualValue,
                expectedValue: condition.value,
                metric: condition.metric,
                operator: condition.function,
                message: `Unknown function: ${condition.function}`
            };
    }
    let unit = '';
    if (['avg', 'p90', 'p95', 'p99'].includes(condition.metric)) {
        unit = 'ms';
    }
    else if (['error_rate', 'failure_rate'].includes(condition.metric)) {
        unit = '%';
    }
    else if (condition.metric === 'rps') {
        unit = ' req/s';
    }
    const message = passed
        ? `✓ ${condition.metric} ${operatorSymbol} ${condition.value}${unit} (actual: ${actualValue.toFixed(2)}${unit})`
        : `✗ ${condition.metric} ${operatorSymbol} ${condition.value}${unit} (actual: ${actualValue.toFixed(2)}${unit})`;
    return {
        passed,
        actualValue,
        expectedValue: condition.value,
        metric: condition.metric,
        operator: operatorSymbol,
        message
    };
}
function buildChecksPayload(condition) {
    return {
        aggregateMetricsThresholds: [
            {
                metricName: METRIC_NAME_MAP[condition.metric] || condition.metric,
                operator: OPERATOR_MAP[condition.function] || condition.function,
                threshold: condition.value
            }
        ]
    };
}
module.exports = {
    parsePassIfCondition,
    evaluatePassIfCondition,
    buildChecksPayload,
    VALID_FUNCTIONS,
    VALID_METRICS
};
