// @postman/pm-bin-linux-x64/bin/lib/commands/performance/run.js
const asyncWaterfall = require('async/waterfall'), colors = require('colors/safe'), fs = require('fs'), path = require('path'), cmdUtil = require('../cmd-util'), libUtil = require('../../util'), print = require('../../framework/print'), config = require('../../config'), perftest = require('../../perftest'), PerformanceOutput = require('./output/'), { parsePassIfCondition } = require('./pass-if'), { getFixedLoadProfile, getRampUpLoadProfile, getSpikeLoadProfile, getPeakLoadProfile, csvToDataFileContent } = require('../../perftest/utils');
module.exports = function (collectionId, command) {
    let options;
    command.alias = 'default';
    asyncWaterfall([
        (next) => { return config.get(command, { command: 'performance' }, next); },
        (command, next) => {
            options = cmdUtil.commanderToObject(command);
            return next(null, options);
        },
        (options, next) => {
            libUtil.extractPostmanApiKey(options.postmanApiKey, (err, postmanApiKey) => {
                if (!postmanApiKey && !options.accessToken) {
                    print.lf('No authorization data found. Please use the `postman login` command.\nRefer:' +
                        'https://learning.postman.com/docs/postman-cli/postman-cli-options/#postman-login\n');
                }
                options.postmanApiKey = postmanApiKey;
                return next(err, options);
            });
        }
    ], (err, options) => {
        if (err) {
            console.error(colors.red(`Error: ${err.message || err}\n`));
            process.exit(1);
        }
        if (options.passIf) {
            try {
                options.passIfCondition = parsePassIfCondition(options.passIf);
            }
            catch (parseError) {
                console.error(colors.red(`Error: ${parseError.message}\n`));
                process.exit(1);
            }
        }
        const validLoadProfiles = ['fixed', 'ramp-up', 'spike', 'peak'];
        if (!validLoadProfiles.includes(options.loadProfile)) {
            console.error(colors.red(`Error: Invalid load profile "${options.loadProfile}". Must be one of: ${validLoadProfiles.join(', ')}`));
            process.exit(1);
        }
        options.vuCount = parseInt(options.vuCount || 20, 10);
        options.duration = parseInt(options.duration || 10, 10);
        let loadProfile;
        switch (options.loadProfile) {
            case 'fixed':
                loadProfile = getFixedLoadProfile(options.vuCount, options.duration);
                break;
            case 'ramp-up':
                loadProfile = getRampUpLoadProfile(options.vuCount, options.duration);
                break;
            case 'spike':
                loadProfile = getSpikeLoadProfile(options.vuCount, options.duration);
                break;
            case 'peak':
                loadProfile = getPeakLoadProfile(options.vuCount, options.duration);
                break;
        }
        const performanceConfig = {
            collectionId: collectionId,
            loadProfile: loadProfile
        };
        if (options.environment) {
            performanceConfig.environment = options.environment;
        }
        if (options.globals) {
            performanceConfig.globals = options.globals;
        }
        if (options.dataFile) {
            try {
                const dataFilePath = path.resolve(options.dataFile);
                if (!fs.existsSync(dataFilePath)) {
                    console.error(colors.red(`Error: Data file not found: ${options.dataFile}`));
                    process.exit(1);
                }
                const fileContent = fs.readFileSync(dataFilePath, 'utf8');
                const fileExtension = path.extname(dataFilePath).toLowerCase();
                let dataContents;
                if (fileExtension === '.json') {
                    dataContents = JSON.parse(fileContent);
                }
                else if (fileExtension === '.csv') {
                    dataContents = csvToDataFileContent(fileContent);
                }
                else {
                    console.error(colors.red(`Error: Unsupported data file format. Only JSON and CSV files are supported.`));
                    process.exit(1);
                }
                if (fileExtension === '.json') {
                    if (!Array.isArray(dataContents)) {
                        console.error(colors.red(`Error: JSON data file must contain an array of objects.`));
                        process.exit(1);
                    }
                }
                performanceConfig.dataFileConfig = {
                    distributionType: 'random',
                    contents: dataContents
                };
                if (options.verbose) {
                    console.log(colors.gray(`Data file loaded: ${options.dataFile}`));
                }
            }
            catch (error) {
                console.error(colors.red(`Error reading data file: ${error.message}`));
                process.exit(1);
            }
        }
        const outputHandler = new PerformanceOutput({
            silent: options.silent,
            verbose: options.verbose,
            disableUnicode: options.disableUnicode,
            color: options.color
        });
        outputHandler.printHeader({
            collectionId,
            loadProfile: options.loadProfile,
            vuCount: options.vuCount,
            duration: options.duration,
            environment: options.environment,
            globals: options.globals,
            dataFile: options.dataFile
        });
        perftest(performanceConfig, options, outputHandler, function (err, aggregatedMetrics) {
            if (err) {
                outputHandler.printError(err);
                process.exit(1);
            }
            outputHandler.printComplete();
            process.exit(0);
        });
    });
};
