// @postman/pm-bin-linux-x64/bin/lib/commands/flows/cmd/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuthForCommand = getAuthForCommand;
exports.promptForMore = promptForMore;
exports.formatDateTime = formatDateTime;
exports.filterByName = filterByName;
const cmdUtil = __importStar(require("../../../commands/cmd-util"));
const libUtil = __importStar(require("../../../util"));
const config = __importStar(require("../../../config"));
const readline = __importStar(require("readline"));
const constants_1 = require("../../../constants");
function getAuthForCommand(command, commandName, apiKeyRequired = false) {
    return new Promise((resolve, reject) => {
        config.get(command, { command: commandName }, (err, options) => {
            if (err) {
                return reject(new Error(`Error getting configuration: ${err.message || err}`));
            }
            options = { ...options, ...cmdUtil.commanderToObject(command) };
            libUtil.extractPostmanApiKey(options.postmanApiKey, (err, postmanApiKey) => {
                if (err || (!postmanApiKey && !options.accessToken && apiKeyRequired)) {
                    return reject(new Error(constants_1.MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE));
                }
                options.postmanApiKey = postmanApiKey || options.accessToken;
                options.accessToken = postmanApiKey || options.accessToken;
                resolve(options);
            });
        });
    });
}
function promptForMore(container) {
    return new Promise((resolve) => {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.question(`Show more ${container}? (y/N): `, (answer) => {
            rl.close();
            resolve(answer.toLowerCase() === 'y' || answer.toLowerCase() === 'yes');
        });
    });
}
function formatDateTime(date) {
    const year = date.getFullYear(), month = String(date.getMonth() + 1).padStart(2, '0'), day = String(date.getDate()).padStart(2, '0'), hours = String(date.getHours()).padStart(2, '0'), minutes = String(date.getMinutes()).padStart(2, '0'), seconds = String(date.getSeconds()).padStart(2, '0'), formattedTime = `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
    return formattedTime;
}
function filterByName(items, filter, nameProperty = 'name') {
    if (!filter) {
        return items;
    }
    const regexMetaChars = /[^a-zA-Z0-9\s-_]/, isRegexPattern = regexMetaChars.test(filter);
    if (isRegexPattern) {
        try {
            const filterRegex = new RegExp(filter, 'i');
            return items.filter((item) => { return filterRegex.test(item[nameProperty]); });
        }
        catch {
            throw new Error(`Invalid filter pattern: ${filter}`);
        }
    }
    else {
        return items.filter((item) => {
            return item[nameProperty].toLowerCase().startsWith(filter.toLowerCase());
        });
    }
}
