// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/formatters/jsonFormatter.js
const { formatRange } = require('./util');
function outputResultsInJson(allResults) {
    const syntaxResults = allResults.syntax || [], governanceResults = allResults.governance || [], allViolations = [
        ...syntaxResults.map((result) => {
            return {
                file: result.file || '',
                'line number': formatRange(result) || '',
                path: result.path ? result.path.join('.') : '',
                severity: result.severity || '',
                issue: result.message || '',
                'issue type': 'Syntax'
            };
        }),
        ...governanceResults.map((violation) => {
            return {
                file: violation.file || '',
                'line number': formatRange(violation) || '',
                path: violation.path ? violation.path.join('.') : '',
                severity: violation.severity || '',
                issue: violation.message || '',
                'issue type': 'Governance'
            };
        })
    ];
    return {
        violations: allViolations
    };
}
module.exports = outputResultsInJson;
