// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/formatters/tableFormatter.js
const Table = require('cli-table3');
const colors = require('colors/safe');
const { formatRange, getFormattedSeverity } = require('./util');
const { SEVERITY_COLOR_MAP, LINE_BREAK, SPECIFICATION_INPUT_TYPES } = require('../constants');
const { POSTMAN_ARTEMIS_BASE_URL } = require('../../../util');
function formatResults(specification, allResults) {
    let output = `${LINE_BREAK}`, strColor, table, issuesCounter = {
        error: 0,
        warning: 0,
        info: 0,
        hint: 0
    };
    const syntaxResults = allResults.syntax || [], governanceResults = allResults.governance || [], totalIssues = syntaxResults.length + governanceResults.length, allViolations = [
        ...syntaxResults.map((r) => {
            return { ...r, issueType: 'Syntax' };
        }),
        ...governanceResults.map((r) => {
            return { ...r, issueType: 'Governance' };
        })
    ];
    output += `${colors.bold('SPECIFICATION DETAILS')}${LINE_BREAK}`;
    output += `${colors.bold('Name:')} ${specification.name} ${LINE_BREAK}`;
    output += specification.inputType === SPECIFICATION_INPUT_TYPES.ID ?
        `${colors.bold('URL:')} ${POSTMAN_ARTEMIS_BASE_URL()}/specification/${specification.id}${LINE_BREAK}` :
        `${colors.bold('PATH:')} ${specification.path}${LINE_BREAK}`;
    output += `${colors.bold('Type:')} ${specification.type}${LINE_BREAK}`;
    output += `${LINE_BREAK}`;
    if (totalIssues === 0) {
        output += `${colors.green('No issues found.')}${LINE_BREAK}`;
        return output;
    }
    const headers = [
        colors.bold('File'),
        colors.bold('Line Number'),
        colors.bold('Path'),
        colors.bold('Severity'),
        colors.bold('Issue'),
        colors.bold('Issue Type')
    ], colWidths = [22, 22, 50, 20, 50, 22];
    table = new Table({
        head: headers,
        style: {
            head: ['white'],
            border: []
        },
        colWidths: colWidths,
        wordWrap: true,
        wrapOnWordBoundary: false
    });
    allViolations.forEach((result) => {
        const row = [
            result.file || '',
            `${formatRange(result)}`,
            result.path ? result.path.join('.') : '',
            getFormattedSeverity(result.severity),
            result.message,
            result.issueType
        ];
        table.push(row);
        issuesCounter[result.severity.toLowerCase()]++;
    });
    output += table.toString();
    output += `${LINE_BREAK}${LINE_BREAK}`;
    if (issuesCounter.error > 0) {
        strColor = SEVERITY_COLOR_MAP.ERROR;
    }
    else if (issuesCounter.warning > 0) {
        strColor = SEVERITY_COLOR_MAP.WARNING;
    }
    else if (issuesCounter.info > 0) {
        strColor = SEVERITY_COLOR_MAP.INFO;
    }
    else {
        strColor = SEVERITY_COLOR_MAP.HINT;
    }
    output += `${colors.bold[strColor](`✖ ${totalIssues} problems (${issuesCounter.error} errors, ` +
        `${issuesCounter.warning} warnings, ${issuesCounter.info} infos, ` +
        `${issuesCounter.hint} hints)`)}${LINE_BREAK}`;
    return output;
}
module.exports = formatResults;
