// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/formatters/tableFormatter.js
const _ = require('lodash');
const Table = require('cli-table3');
const colors = require('colors/safe');
const { SCHEMA_TYPES, SEVERITY_COLOR_MAP, LF, COMMAND_MODE } = require('../constants');
const { getRangeAndFileForConfigMode } = require('./utils/getRangeAndFileForConfigMode');
function getRange(result) {
    return {
        range: result.range,
        prefix: ''
    };
}
function formatRange(result, api) {
    const { range, prefix } = getRange(result, api);
    if (!range) {
        return '';
    }
    let str = '';
    if (prefix) {
        str = `${prefix} -> `;
    }
    return str + `${range.start.line + 1}:${range.start.character + 1}`;
}
function formatRangeFromFileConfig(result, api) {
    let range, file, str = '';
    if (api.allFiles && api.allFiles.length > 1) {
        let details = getRangeAndFileForConfigMode(result, api);
        range = details.range;
        file = details.file;
    }
    else {
        range = result.range;
    }
    if (!range) {
        return '';
    }
    if (file) {
        str = `${file} -> `;
    }
    return str + `${range.start.line + 1}:${range.start.character + 1}`;
}
function getFormattedSeverity(severity) {
    return colors[SEVERITY_COLOR_MAP[severity]](severity);
}
function formatResults(api, commandMode, categorizedResults) {
    let output = `${LF}`, issuesFound = false, totalIssues = 0, issuesCounter = {
        error: 0,
        warn: 0,
        info: 0,
        hint: 0
    };
    output += `${colors.bold('API DETAILS')}${LF}`;
    output += `${colors.bold('Name:')} ${api.name} ${LF}`;
    output += `${colors.bold('Type:')} ${SCHEMA_TYPES[api.type]}${LF}`;
    output += `${colors.bold('Language:')} ${api.language}${LF}`;
    output += `${LF}`;
    for (let [category, results] of Object.entries(categorizedResults)) {
        if (results.length === 0) {
            continue;
        }
        output += `${colors.bold('Validation Type:')} ${_.capitalize(category)}${LF}`;
        let table = new Table({
            head: [
                colors.bold('Range'),
                colors.bold('Severity'),
                colors.bold('Description'),
                colors.bold('Path')
            ],
            style: {
                head: ['white'],
                border: []
            },
            colWidths: [20, 20, 50, 50],
            wordWrap: true,
            wrapOnWordBoundary: false
        });
        results.forEach((result) => {
            let getRangeFormatterFn = commandMode !== COMMAND_MODE.CONFIG ? formatRange : formatRangeFromFileConfig;
            table.push([
                `${getRangeFormatterFn(result, api)}`,
                getFormattedSeverity(result.severity),
                result.message,
                result.path ? result.path.join('.') : ''
            ]);
            issuesCounter[result.severity.toLowerCase()]++;
        });
        totalIssues += results.length;
        issuesFound = true;
        output += table.toString();
        output += `${LF}${LF}`;
    }
    if (!issuesFound) {
        output += `${LF}${colors.green('No issues found.')}${LF}`;
    }
    else {
        let strColor;
        if (issuesCounter.error > 0) {
            strColor = SEVERITY_COLOR_MAP.ERROR;
        }
        else if (issuesCounter.warn > 0) {
            strColor = SEVERITY_COLOR_MAP.WARN;
        }
        else if (issuesCounter.info > 0) {
            strColor = SEVERITY_COLOR_MAP.INFO;
        }
        else {
            strColor = SEVERITY_COLOR_MAP.HINT;
        }
        output += `${colors.bold[strColor](`✖ ${totalIssues} problems (${issuesCounter.error} errors, ${issuesCounter.warn} warnings, ${issuesCounter.info} infos, ${issuesCounter.hint} hints)`)}${LF}`;
        output += `${LF}`;
    }
    return output;
}
module.exports = formatResults;
