// @postman/pm-bin-linux-x64/bin/lib/region-util.js
const _ = require('lodash'), colors = require('colors/safe'), request = require('postman-request'), console = require('./framework/console'), { REGIONS } = require('./constants'), cliEnvironment = require('./config/cli-environment'), getIapubBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].iapub,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].iapub
    };
}, getPackmanBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].packman,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].packman
    };
}, getRuntimeAgentBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].runtimeAgent,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].runtimeAgent
    };
}, getGatewayBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].gateway,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].gateway
    };
}, getApiBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].api,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].api
    };
}, getArtemisBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].artemis,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].artemis
    };
}, getRunnerApiBaseUrls = function () {
    const settings = cliEnvironment.getSettings();
    return {
        [REGIONS.US]: settings.baseUrls[REGIONS.US].runnerApi,
        [REGIONS.EU]: settings.baseUrls[REGIONS.EU].runnerApi
    };
}, getSDKGenBaseUrls = function () {
    return getGatewayBaseUrls();
}, getIntegrationsBaseUrls = function () {
    return getGatewayBaseUrls();
}, regionUtil = {
    ERROR_MESSAGES: {
        unsupportedRegion(region) {
            return `The ${region ? '' : 'current '}region ${region ? region : ''} is not supported.`;
        }
    },
    _cachedRegion: null,
    validateRegion: function (region, { defaultRegion = REGIONS.US, errorLevel = 'warn' } = {}) {
        const regionUrls = getApiBaseUrls();
        if (!Object.values(REGIONS).includes(region) || !Object.keys(regionUrls).includes(region)) {
            if (errorLevel === 'warn') {
                console.warn(colors.yellow(this.ERROR_MESSAGES.unsupportedRegion(region) + '\nUsing default region (\'us\').'));
                return defaultRegion;
            }
            else if (errorLevel === 'error') {
                throw new Error(this.ERROR_MESSAGES.unsupportedRegion(JSON.stringify(region)));
            }
        }
        return region;
    },
    setCurrentRegionForProcess: function (region, { warnOnNonDefaultRegion = false } = {}) {
        if (!region) {
            throw new Error('Region value required');
        }
        this.validateRegion(region, { errorLevel: 'error' });
        this._cachedRegion = region;
        if (warnOnNonDefaultRegion && region !== REGIONS.US) {
            console.warn(colors.yellow(`Postman CLI set to use the "${region}" region`));
        }
    },
    getCurrentRegion: function () {
        if (this._cachedRegion) {
            return this._cachedRegion;
        }
        const rcfile = require('./config/rc-file'), config = rcfile.loadSync(['home']), getRegionOfAlias = (alias, fallback) => {
            const defaultProfile = _.find(_.get(config, 'login._profiles', []), { alias });
            return defaultProfile && defaultProfile.region ? defaultProfile.region : fallback;
        }, regionOfDefaultAlias = config && getRegionOfAlias('default'), region = regionOfDefaultAlias ? regionOfDefaultAlias : REGIONS.US;
        this.setCurrentRegionForProcess(region);
        return this._cachedRegion;
    },
    getGatewayBaseURL: function () {
        if (process.env.POSTMAN_GATEWAY_BASE_URL) {
            return process.env.POSTMAN_GATEWAY_BASE_URL;
        }
        const region = this.getCurrentRegion(), regionUrls = getGatewayBaseUrls();
        return regionUrls[region] || regionUrls[REGIONS.US];
    },
    getPackmanBaseURL: function () {
        if (process.env.POSTMAN_PACKMAN_BASE_URL) {
            return process.env.POSTMAN_PACKMAN_BASE_URL;
        }
        const region = this.getCurrentRegion(), regionUrls = getPackmanBaseUrls();
        return regionUrls[region] || regionUrls[REGIONS.US];
    },
    getRuntimeAgentBaseURL: function () {
        if (process.env.POSTMAN_RUNTIME_AGENT_BASE_URL) {
            return process.env.POSTMAN_RUNTIME_AGENT_BASE_URL;
        }
        const region = this.getCurrentRegion(), regionUrls = getRuntimeAgentBaseUrls();
        return regionUrls[region] || regionUrls[REGIONS.US];
    },
    getAPIBaseURL: function () {
        if (process.env.POSTMAN_API_BASE_URL) {
            return process.env.POSTMAN_API_BASE_URL;
        }
        const region = this.getCurrentRegion(), regionUrls = getApiBaseUrls();
        return regionUrls[region] || regionUrls[REGIONS.US];
    },
    getARTEMISBaseURL: function () {
        if (process.env.POSTMAN_ARTEMIS_BASE_URL) {
            return process.env.POSTMAN_ARTEMIS_BASE_URL;
        }
        const region = this.getCurrentRegion(), artemisRegionUrls = getArtemisBaseUrls();
        return artemisRegionUrls[region] || artemisRegionUrls[REGIONS.US];
    },
    getIAPUBBaseURL: function (region) {
        if (process.env.POSTMAN_IAPUB_BASE_URL) {
            return process.env.POSTMAN_IAPUB_BASE_URL;
        }
        const iapubRegionUrls = getIapubBaseUrls();
        if (!iapubRegionUrls[region]) {
            throw new Error(`Unsupported region: ${region}`);
        }
        return iapubRegionUrls[region];
    },
    getRunnerAPIBaseURL: function () {
        if (process.env.POSTMAN_RUNNER_API_BASE_URL) {
            return process.env.POSTMAN_RUNNER_API_BASE_URL;
        }
        const region = this.getCurrentRegion(), regionUrls = getRunnerApiBaseUrls();
        return regionUrls[region] || regionUrls[REGIONS.US];
    },
    getComponentLibraryBaseURL: function () {
        if (process.env.COMPONENT_LIBRARY_BASE_URL) {
            return process.env.COMPONENT_LIBRARY_BASE_URL;
        }
        const region = this.getCurrentRegion(), settings = cliEnvironment.getSettings();
        return settings.baseUrls[region].componentLibrary ||
            settings.baseUrls[REGIONS.US].componentLibrary;
    },
    getSDKGenBaseURL: function () {
        if (process.env.POSTMAN_SDK_GEN_BASE_URL) {
            return process.env.POSTMAN_SDK_GEN_BASE_URL;
        }
        return this.getGatewayBaseURL();
    },
    getIntegrationsBaseURL: function () {
        if (process.env.POSTMAN_INTEGRATIONS_BASE_URL) {
            return process.env.POSTMAN_INTEGRATIONS_BASE_URL;
        }
        return this.getGatewayBaseURL();
    },
    REGION_DETECTION_ERROR: 'An error occurred while trying to detect the region for the provided API key. ' +
        '\nPlease ensure that the API key is valid and try again.',
    BAD_API_KEY_ERROR: 'The provided API key is invalid. Please check the API key and try again.',
    REGION_VALIDATION_ERROR: 'Unable to validate API key for the specified region. Please check your network connection and try again.',
    validateApiKeyForRegion: function (key, region, requiredFields, callback) {
        const util = require('./util'), regionUrls = getApiBaseUrls(), requestConfig = {
            url: `${regionUrls[region]}/me`,
            headers: {
                ...util.defaultHeaders,
                'x-api-key': key
            }
        };
        request.get(requestConfig, (err, response, body) => {
            if (err) {
                logger.error(`Error validating API key for region ${region}: ${err.message || err}`);
                return callback(util.augmentError(new Error(this.REGION_VALIDATION_ERROR), `\nError validating API key for region ${region}: ${err.message || err}`));
            }
            let parsedBody;
            try {
                parsedBody = body && JSON.parse(body);
            }
            catch (e) {
                logger.error(`Error parsing response body while validating API key for region ${region}: ${e.message}`);
                return callback(util.augmentError(new Error(this.REGION_VALIDATION_ERROR), `\nError parsing response for region ${region}: ${e.message}`));
            }
            const isValidKey = _.inRange(response.statusCode, 200, 300);
            if (!isValidKey) {
                if (_.inRange(response.statusCode, 400, 500)) {
                    return callback(util.augmentError(new Error(this.BAD_API_KEY_ERROR, { cause: this.BAD_API_KEY_ERROR }), `\nInvalid Postman API key was provided. (Status code: ${response.statusCode})`));
                }
                return callback(util.augmentError(new Error(this.REGION_VALIDATION_ERROR), `\nUnexpected response received (Status code: ${response.statusCode})`));
            }
            return callback(null, {
                isValidKey: true,
                region: region,
                authInfo: _.pick(parsedBody, requiredFields || [])
            });
        });
    }
};
module.exports = {
    ...regionUtil,
    getApiBaseUrls,
    getGatewayBaseUrls,
    getIapubBaseUrls,
    getArtemisBaseUrls,
    getPackmanBaseUrls,
    getRuntimeAgentBaseUrls,
    getRunnerApiBaseUrls,
    getSDKGenBaseUrls,
    getIntegrationsBaseUrls
};
