// @postman/pm-bin-linux-x64/bin/lib/user-type.js
const requestFactory = require('@postman/platform-req/lib/request');
const { jwtDecode } = require('jwt-decode');
const util = require('./util');
function getClient() {
    return requestFactory.defaults({
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: { enabled: true }
            }
        }
    });
}
function getUserType(postmanApiKey, accessToken) {
    if (!postmanApiKey && !accessToken) {
        return Promise.reject(new Error('Authentication required. Please run `postman login` to authenticate.'));
    }
    const headers = { 'User-Agent': util.userAgent }, client = getClient(), url = `${util.POSTMAN_RUNTIME_AGENT_BASE_URL()}/handshake/token?source=packman`;
    if (postmanApiKey) {
        headers['x-api-key'] = postmanApiKey;
    }
    else {
        headers['x-access-token'] = accessToken;
    }
    return new Promise((resolve, reject) => {
        client.get({
            url: url,
            headers: headers,
            json: true
        }, (err, response) => {
            if (err) {
                return reject(new Error(`Failed to verify user: ${err.message}`));
            }
            if (response.statusCode !== 200) {
                return reject(new Error(`Failed to verify user. Status: ${response.statusCode}`));
            }
            if (!response.body || !response.body.token) {
                return reject(new Error('Invalid response: missing auth token'));
            }
            try {
                const decoded = jwtDecode(response.body.token), resolvedType = decoded && decoded.userType ? decoded.userType : '';
                return resolve(resolvedType);
            }
            catch (decodeErr) {
                return reject(new Error(`Failed to verify user: ${decodeErr.message}`));
            }
        });
    });
}
module.exports = { getUserType };
