// @postman/pm-bin-linux-x64/bin/lib/util.js
const fs = require('fs'), os = require('os'), regionUtil = require('./region-util'), _ = require('lodash'), chardet = require('chardet'), { filesize } = require('filesize'), prettyms = require('pretty-ms'), liquidJSON = require('liquid-json'), request = require('postman-request'), version = require('../package.json').version, { getAPIKeyFromCLIArguments, validatePostmanAPIKey } = require('./run-upload/helpers/util'), cliUtils = require('./reporters/cli/cli-utils'), SEP = ' / ', FILESIZE_OPTIONS = { spacer: ' ', standard: 'jedec' }, CHARDET_BUFF_MAP = {
    ASCII: 'ascii',
    'UTF-8': 'utf8',
    'UTF-16LE': 'utf16le',
    'ISO-8859-1': 'latin1'
}, USER_AGENT_VALUE = 'Postman CLI/' + version, DEFAULT_HEADERS = {
    'User-Agent': USER_AGENT_VALUE
}, POSTMAN_API_URL_REGEX = !process.env.NEWMAN_TEST_ENV ? /^https?:\/\/api\.(eu\.)?(get)?postman\.com.*/ :
    /^https?:\/\/(localhost:\d+|api\.(eu\.)?(get)?postman\.com).*/, POSTMAN_ID_REGEX = /^([0-9A-Z]+-|)[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}$/i, POSTMAN_ID_REGEX_UNANCHORED = /([0-9A-Z]+-|)[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}/i, util = {
    POSTMAN_API_BASE_URL: regionUtil.getAPIBaseURL.bind(regionUtil),
    POSTMAN_ARTEMIS_BASE_URL: regionUtil.getARTEMISBaseURL.bind(regionUtil),
    POSTMAN_IAPUB_BASE_URL: regionUtil.getIAPUBBaseURL.bind(regionUtil),
    POSTMAN_GATEWAY_BASE_URL: regionUtil.getGatewayBaseURL.bind(regionUtil),
    POSTMAN_PACKMAN_BASE_URL: regionUtil.getPackmanBaseURL.bind(regionUtil),
    POSTMAN_RUNTIME_AGENT_BASE_URL: regionUtil.getRuntimeAgentBaseURL.bind(regionUtil),
    POSTMAN_RUNNER_API_BASE_URL: regionUtil.getRunnerAPIBaseURL.bind(regionUtil),
    COMPONENT_LIBRARY_BASE_URL: regionUtil.getComponentLibraryBaseURL.bind(regionUtil),
    POSTMAN_SDK_GEN_BASE_URL: regionUtil.getSDKGenBaseURL.bind(regionUtil),
    POSTMAN_INTEGRATIONS_BASE_URL: regionUtil.getIntegrationsBaseURL.bind(regionUtil),
    version: version,
    userAgent: USER_AGENT_VALUE,
    defaultHeaders: DEFAULT_HEADERS,
    signal: (message) => {
        process.send && process.send(message);
    },
    prettyms: function (ms) {
        if (ms < 1) {
            return `${parseInt(ms * 1000, 10)} ${cliUtils.usePlainText() ? 'us' : 'µs'}`;
        }
        return (ms < 1998) ? `${parseInt(ms, 10)} ms` : prettyms(ms || 0).replace(/(\d+)([a-zA-Z]+)/, '$1 $2');
    },
    beautifyTime: function (obj) {
        return _.forEach(obj, (value, key) => {
            value && (obj[key] = this.prettyms(value));
        });
    },
    filesize: function (bytes) {
        let fn = filesize;
        if (typeof filesize !== 'function' && typeof filesize.default === 'function') {
            fn = filesize.default;
        }
        return fn(bytes || 0, FILESIZE_OPTIONS);
    },
    getFullName: function (item, separator) {
        if (_.isEmpty(item) || !_.isFunction(item.parent) || !_.isFunction(item.forEachParent)) {
            return;
        }
        var chain = [];
        item.forEachParent(function (parent) { chain.unshift(parent.name || parent.id); });
        item.parent() && chain.push(item.name || item.id);
        return chain.join(_.isString(separator) ? separator : SEP);
    },
    getFullPath: function (item) {
        if (!item || _.isEmpty(item) || !_.isFunction(item.parent) || !_.isFunction(item.forEachParent)) {
            return [];
        }
        var chain = [];
        item.forEachParent(function (parent) {
            chain.unshift(parent.name || parent.id);
        });
        if (item.parent()) {
            chain.push(item.name || item.id);
        }
        return chain;
    },
    detectEncoding: function (buff) {
        return CHARDET_BUFF_MAP[chardet.detect(buff)];
    },
    fetchJson: function (location, type, callback) {
        if (!callback && _.isFunction(type)) {
            callback = type;
            type = 'resource';
        }
        return (/^https?:\/\/.*/).test(location) ?
            request.get({
                url: location,
                json: true,
                headers: DEFAULT_HEADERS
            }, (err, response, body) => {
                if (err) {
                    return callback(_.set(err, 'help', `unable to fetch data from url "${location}"`));
                }
                try {
                    _.isString(body) && (body = liquidJSON.parse(body.trim()));
                }
                catch (e) {
                    return callback(_.set(e, 'help', `the url "${location}" did not provide valid JSON data`));
                }
                var error;
                if (response.statusCode !== 200) {
                    error = new Error(_.get(body, 'error.message', `Error fetching ${type}, the Postman server returned status code: ${response.statusCode}`));
                    return callback(_.assign(error, {
                        name: _.get(body, 'error.name', _.capitalize(type) + 'FetchError'),
                        help: `Error: ${type} couldn't be loaded from Postman servers. Make sure the ID and API key are correct and try again.`
                    }));
                }
                return callback(null, body);
            }) :
            fs.readFile(location, function (err, value) {
                if (err) {
                    return callback(_.set(err, 'help', `unable to read data from file "${location}"`));
                }
                try {
                    value = liquidJSON.parse(value.toString(util.detectEncoding(value)).trim());
                }
                catch (e) {
                    return callback(_.set(e, 'help', `the file at "${location}" does not contain valid JSON data`));
                }
                return callback(null, value);
            });
    },
    fetch: function (location, options, callback) {
        !callback && _.isFunction(options) && (callback = options, options = {});
        return (/^https?:\/\/.*/).test(location) ?
            request.get({
                url: location,
                headers: DEFAULT_HEADERS
            }, (err, response, body) => {
                if (err) {
                    return callback(err);
                }
                return callback(null, body);
            }) :
            fs.readFile(String(location), function (err, value) {
                if (err) {
                    return callback(err);
                }
                return callback(null, value.toString(util.detectEncoding(value)));
            });
    },
    extractPostmanApiKey: function (postmanApiKey, callback) {
        if (!postmanApiKey) {
            return callback();
        }
        if (_.isString(postmanApiKey)) {
            return callback(null, postmanApiKey);
        }
        postmanApiKey = postmanApiKey.postmanApiKey;
        return callback(null, postmanApiKey);
    },
    NO_REGION_VALID_FOR_API_KEY_ERROR: 'The given API key did not work. ' +
        '\nPlease ensure that the API key is valid and try again.',
    INVALID_REGION_DETERMINER_ERROR: 'Invalid region determiner provided. ' +
        '\nIt should be a Postman API key or an object containing profile details.',
    isPublicAPIUrl: function (location) {
        return POSTMAN_API_URL_REGEX.test(location);
    },
    extractResourceIDFromPublicAPIUrl: function (location) {
        if (!_.isString(location)) {
            return null;
        }
        let match = location.match(POSTMAN_ID_REGEX_UNANCHORED);
        if (!match) {
            return null;
        }
        return match[0];
    },
    isPostmanResource: function (location) {
        return _.isString(location) && !fs.existsSync(location) &&
            (this.isPublicAPIUrl(location) || POSTMAN_ID_REGEX.test(location));
    },
    isV1Collection: function (object) {
        return Boolean(object && object.name && object.order && object.requests);
    },
    isInt: function (value) {
        return (/^(-|\+)?([1-9]+[0-9]*)$/).test(value);
    },
    isFloat: function (value) {
        return (value - parseFloat(value) + 1) >= 0;
    },
    commanderToObject: (command) => {
        return _.reduce(command, (result, value, key) => {
            const validProp = !_.startsWith(key, '_') && !_.includes(['commands', 'options', 'parent'], key);
            validProp && (result[key] = value);
            return result;
        }, {});
    },
    getArch: () => {
        let platform = os.platform(), arch = os.arch();
        if (_.includes(platform, 'win32')) {
            if (_.includes(arch, 'x32') || _.includes(arch, 'ia32')) {
                return '32';
            }
            return '64';
        }
        if (_.includes(platform, 'linux-arm')) {
            return 'arm64';
        }
        if (_.includes(platform, 'linux')) {
            if (_.includes(arch, '64')) {
                return '64';
            }
            return '32';
        }
        if (_.includes(platform, 'darwin')) {
            if (_.includes(arch, 'arm64')) {
                return 'arm64';
            }
            return '64';
        }
    },
    getPlatform: () => {
        let platform = process.platform, arch = process.arch;
        if (platform === 'win32') {
            if (arch === 'x32' || arch === 'ia32') {
                return 'win32';
            }
            return 'win64';
        }
        if (_.includes(platform, 'linux')) {
            return 'linux';
        }
        return 'osx';
    },
    removeCloudFilesFolder: (workingDir) => {
        if (!fs.existsSync(`${workingDir}/.cloudFiles`)) {
            return;
        }
        fs.rmSync(`${workingDir}/.cloudFiles`, {
            recursive: true,
            force: true
        });
    },
    getValidPostmanAPIKey: (options) => {
        let processArgs = process.argv, postmanApiKey = (validatePostmanAPIKey(options.postmanApiKey) && options.postmanApiKey) ||
            (validatePostmanAPIKey(process.env.POSTMAN_API_KEY) && process.env.POSTMAN_API_KEY);
        if (!postmanApiKey) {
            const apiKey = getAPIKeyFromCLIArguments(processArgs);
            validatePostmanAPIKey(apiKey) && (postmanApiKey = apiKey);
        }
        return postmanApiKey;
    },
    augmentError: (error, message) => {
        if (!error || !_.isString(message)) {
            return error;
        }
        if (process.env.POSTMAN_CLI_VERBOSE_ERRORS === 'true') {
            return new Error(`${error.message}\n\n${message}`, { cause: error.cause });
        }
        return error;
    }
};
module.exports = util;
