// @postman/pm-bin-linux-x64/bin/lib/perftest/schemas/performance-run-parameters.js
const POSTMAN_ID_PATTERN = '^([0-9A-Z]+-)?[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}$';
module.exports = {
    $schema: 'http://json-schema.org/draft-07/schema#',
    $id: 'performance-run-parameters-schema',
    title: 'Performance Run Parameter Schema',
    description: 'Schema for validating performance run parameters',
    type: 'object',
    required: ['version', 'runs'],
    properties: {
        version: {
            enum: ['0.1.0'],
            description: 'Schema version for the performance run parameters'
        },
        metadata: {
            type: 'object',
            description: 'Metadata about the performance test suite',
            properties: {
                name: {
                    type: 'string',
                    description: 'Name of the performance test suite'
                },
                description: {
                    type: 'string',
                    description: 'Description of what this performance test suite tests'
                },
                tags: {
                    type: 'array',
                    description: 'Tags for categorizing this test suite',
                    items: {
                        type: 'string'
                    }
                },
                infrastructure: {
                    type: 'object',
                    description: 'Infrastructure-related metadata (unstructured)',
                    additionalProperties: true
                }
            },
            additionalProperties: true
        },
        defaults: {
            type: 'object',
            description: 'Default configurations inherited by all runs unless overridden',
            properties: {
                loadProfile: {
                    $ref: '#/definitions/loadProfile'
                },
                assertions: {
                    $ref: '#/definitions/assertions'
                }
            },
            additionalProperties: false
        },
        runs: {
            type: 'array',
            description: 'List of performance test runs, each with specific associations and configurations',
            minItems: 1,
            items: {
                type: 'object',
                required: ['name', 'associations'],
                properties: {
                    name: {
                        type: 'string',
                        description: 'Name of this specific run'
                    },
                    associations: {
                        type: 'object',
                        description: 'Resources associated with this run',
                        properties: {
                            collections: {
                                type: 'array',
                                description: 'Collection IDs for this run',
                                items: {
                                    type: 'string',
                                    pattern: POSTMAN_ID_PATTERN
                                },
                                maxItems: 1,
                                minItems: 1
                            },
                            environments: {
                                type: 'array',
                                description: 'Environment IDs for this run',
                                items: {
                                    type: 'string',
                                    pattern: POSTMAN_ID_PATTERN
                                },
                                maxItems: 1,
                                minItems: 0
                            },
                            dataFiles: {
                                type: 'array',
                                description: 'Data file paths for this run',
                                items: {
                                    type: 'string'
                                },
                                maxItems: 1,
                                minItems: 0
                            }
                        },
                        additionalProperties: false
                    },
                    loadProfile: {
                        $ref: '#/definitions/loadProfile',
                        description: 'Load profile overrides for this run'
                    },
                    assertions: {
                        $ref: '#/definitions/assertions',
                        description: 'Assertions overrides for this run'
                    }
                },
                additionalProperties: false
            }
        }
    },
    additionalProperties: false,
    definitions: {
        loadProfile: {
            type: 'object',
            description: 'Load profile configuration defining the test load pattern',
            required: ['duration', 'peakVUs', 'load'],
            properties: {
                duration: {
                    type: 'integer',
                    description: 'Total test duration in seconds',
                    minimum: 1
                },
                peakVUs: {
                    type: 'integer',
                    description: 'Maximum number of virtual users',
                    minimum: 1
                },
                load: {
                    description: 'Load shape definition - can be a ramp object or custom stages',
                    oneOf: [
                        {
                            type: 'object',
                            required: ['fixed'],
                            additionalProperties: false
                        },
                        {
                            type: 'object',
                            required: ['ramp'],
                            properties: {
                                ramp: {
                                    type: 'object',
                                    required: ['up'],
                                    properties: {
                                        up: {
                                            type: 'integer',
                                            description: 'Ramp up duration in seconds',
                                            minimum: 0
                                        }
                                    },
                                    additionalProperties: false
                                }
                            },
                            additionalProperties: false
                        },
                        {
                            type: 'object',
                            required: ['stages'],
                            properties: {
                                stages: {
                                    type: 'array',
                                    description: 'Custom load stages',
                                    minItems: 1,
                                    items: {
                                        type: 'object',
                                        required: ['duration', 'vus'],
                                        properties: {
                                            name: {
                                                type: 'string',
                                                description: 'Optional name for this stage'
                                            },
                                            duration: {
                                                type: 'integer',
                                                description: 'Duration of this stage in seconds',
                                                minimum: 1
                                            },
                                            vus: {
                                                type: 'integer',
                                                description: 'Target VUs for this stage',
                                                minimum: 0
                                            }
                                        },
                                        additionalProperties: false
                                    }
                                }
                            },
                            additionalProperties: false
                        }
                    ]
                }
            },
            additionalProperties: false
        },
        assertions: {
            type: 'object',
            description: 'Performance assertions and thresholds',
            properties: {
                errorRateThreshold: {
                    type: 'number',
                    description: 'Maximum acceptable error rate (0.0 to 1.0)',
                    minimum: 0,
                    maximum: 1
                },
                failureRateThreshold: {
                    type: 'number',
                    description: 'Maximum acceptable failure rate (0.0 to 1.0)',
                    minimum: 0,
                    maximum: 1
                },
                responseTime: {
                    type: 'object',
                    description: 'Response time thresholds in milliseconds',
                    properties: {
                        average: {
                            type: 'number',
                            description: 'Maximum acceptable average response time',
                            minimum: 0
                        },
                        p90: {
                            type: 'number',
                            description: 'Maximum acceptable 90th percentile response time',
                            minimum: 0
                        },
                        p95: {
                            type: 'number',
                            description: 'Maximum acceptable 95th percentile response time',
                            minimum: 0
                        },
                        p99: {
                            type: 'number',
                            description: 'Maximum acceptable 99th percentile response time',
                            minimum: 0
                        }
                    },
                    additionalProperties: false
                },
                statusCodePercentages: {
                    type: 'object',
                    description: 'Expected status code distribution',
                    patternProperties: {
                        '^[1-5][0-9]{2}$': {
                            type: 'object',
                            properties: {
                                min: {
                                    type: 'number',
                                    minimum: 0,
                                    maximum: 100
                                },
                                max: {
                                    type: 'number',
                                    minimum: 0,
                                    maximum: 100
                                }
                            }
                        }
                    },
                    additionalProperties: false
                },
                throughputThreshold: {
                    type: 'number',
                    description: 'Minimum acceptable throughput (requests per second)',
                    minimum: 0
                },
                failureAction: {
                    type: 'string',
                    description: 'Action to take when thresholds are breached',
                    enum: ['continue', 'stop', 'warn']
                }
            },
            additionalProperties: false
        }
    }
};
