// @postman/pm-bin-linux-x64/bin/lib/commands/runner/core/lifecycle.js
const runService = require('../services/run-service'), Runner = require('./runner'), { getLogger } = require('../logging/logger-factory');
const logger = getLogger();
const { RunnerProxyManagerInstance: ProxyManager } = require('../proxy');
const { parseUserProxyOptions } = require('../utils/proxy-options-parser');
const ErrorHandler = {
    getError(operation, statusCode, runnerId, originalError = null) {
        const messages = {
            401: 'Authentication failed: Please verify your runner ID and key.',
            403: 'Access forbidden: Insufficient permissions for this operation.',
            404: `${operation} failed: Resource not found or no longer available.`,
            429: `${operation} failed: Too many requests. Please try again later.`,
            500: `${operation} failed: Internal server error.`,
            502: `${operation} failed: Service temporarily unavailable.`,
            503: `${operation} failed: Service unavailable.`,
            504: `${operation} failed: Request timeout.`
        };
        let message = messages[statusCode] || `${operation} failed with status ${statusCode}.`;
        if (statusCode >= 500 && originalError?.message) {
            message += ` Details: ${originalError.message}`;
        }
        const error = new Error(message);
        error.statusCode = statusCode;
        error.runnerId = runnerId;
        error.operation = operation;
        return error;
    }
};
async function fetchRun({ runnerId, runnerKey, disableTelemetry }) {
    try {
        logger.info('Checking for available runs', {
            operation: 'fetchRun',
            runnerId: runnerId
        });
        const runDataResponse = await runService.fetchRun(runnerId, runnerKey, disableTelemetry);
        if (!runDataResponse.runData) {
            logger.info('No runs available at this time', {
                operation: 'fetchRun',
                runnerId: runnerId
            });
            return null;
        }
        logger.info('Run found, starting execution', {
            operation: 'fetchRun',
            runnerId: runnerId,
            runId: runDataResponse.runData.run?.id
        });
        return {
            runData: runDataResponse.runData,
            eventIngestion: runDataResponse.eventIngestion
        };
    }
    catch (error) {
        if (error.statusCode) {
            throw ErrorHandler.getError('Fetch run', error.statusCode, runnerId, error);
        }
        throw error;
    }
}
async function executeRun({ runData, eventIngestion, runnerId, runnerKey, connectionOptions, workingDir, reportEvents }) {
    try {
        const resolvedWorkingDir = workingDir || process.cwd();
        const { collection, environment, jobInput, runConfig = {}, data, job, run, trace } = runData, runner = new Runner({
            job: job,
            run: run,
            collection: collection,
            environment: environment,
            jobInput: jobInput,
            runConfig: runConfig,
            eventIngestion: eventIngestion,
            data: data,
            runTrace: trace,
            runnerId: runnerId,
            runnerKey: runnerKey,
            workingDir: resolvedWorkingDir,
            connectionOptions: connectionOptions,
            reportEvents: reportEvents
        });
        const runLog = await runner.exec();
        return { runLog, run };
    }
    catch (error) {
        if (error.statusCode) {
            throw ErrorHandler.getError('Execute run', error.statusCode, runnerId, error);
        }
        logger.error('Error executing run', {
            operation: 'executeRun',
            runnerId: runnerId,
            errorMessage: error.message,
            errorStack: error.stack
        });
        throw error;
    }
}
async function finishRun({ run, runLog, runnerId, runnerKey }) {
    try {
        const finishRunResponse = await runService.finishRun({
            run,
            runLog,
            runnerId,
            runnerKey
        });
        if (finishRunResponse.success) {
            logger.info('Run execution completed successfully', {
                operation: 'finishRun',
                runnerId: runnerId,
                runId: run.id
            });
            return true;
        }
        logger.warn('Run execution failed to complete', {
            operation: 'finishRun',
            runnerId: runnerId,
            runId: run.id
        });
        return false;
    }
    catch (error) {
        if (error.statusCode) {
            throw ErrorHandler.getError('Finish run', error.statusCode, runnerId, error);
        }
        logger.error('Error finishing run execution', {
            operation: 'finishRun',
            runnerId: runnerId,
            runId: run.id,
            errorMessage: error.message,
            errorStack: error.stack
        });
        throw error;
    }
}
async function releaseRun({ run, runLog, runnerId, runnerKey }) {
    try {
        await runService.releaseRun({
            run,
            runLog,
            runnerId,
            runnerKey
        });
        logger.info('Run released successfully', {
            operation: 'releaseRun',
            runnerId: runnerId,
            runId: run.id
        });
    }
    catch (error) {
        if (error.statusCode) {
            throw ErrorHandler.getError('Release run', error.statusCode, runnerId, error);
        }
        logger.error('Error releasing run', {
            operation: 'releaseRun',
            runnerId: runnerId,
            runId: run?.id,
            errorMessage: error.message,
            errorStack: error.stack
        });
        throw error;
    }
}
async function getCaCertPath(cliOptions) {
    if (cliOptions.sslExtraCaCerts) {
        return cliOptions.sslExtraCaCerts;
    }
    if (ProxyManager._isRunning) {
        const certInfo = await ProxyManager.retrieveGeneratedCertificateInfo();
        return certInfo?.caCertPath;
    }
    return null;
}
function getProxyManagerOptions() {
    if (ProxyManager._isRunning) {
        return {
            host: ProxyManager.getProxyHost(),
            port: ProxyManager.getProxyPort(),
            tunnel: true
        };
    }
    return null;
}
let cachedConnectionOptions = null;
let initializationPromise = null;
function initializeConnectionOptions(cliOptions = {}) {
    if (cachedConnectionOptions) {
        return Promise.resolve(cachedConnectionOptions);
    }
    if (initializationPromise) {
        return initializationPromise;
    }
    initializationPromise = (async () => {
        try {
            const proxyOptions = parseUserProxyOptions(cliOptions) || getProxyManagerOptions();
            const caCertPath = await getCaCertPath(cliOptions);
            const connectionOptions = { proxyOptions, caCertPath };
            if (proxyOptions) {
                runService.proxifyRequest(connectionOptions);
            }
            cachedConnectionOptions = connectionOptions;
            return connectionOptions;
        }
        catch (error) {
            initializationPromise = null;
            throw error;
        }
    })();
    return initializationPromise;
}
async function runLifecycle({ runnerId, runnerKey, disableTelemetry, cliOptions = {}, workingDir = null }) {
    const connectionOptions = await initializeConnectionOptions(cliOptions);
    try {
        const fetchResult = await fetchRun({ runnerId, runnerKey, disableTelemetry });
        if (!fetchResult) {
            return false;
        }
        const { runData, eventIngestion } = fetchResult;
        const reportEvents = cliOptions.reportEvents;
        const { runLog, run } = await executeRun({
            runData, eventIngestion, runnerId, runnerKey, connectionOptions, workingDir, reportEvents
        });
        const finishSuccess = await finishRun({ run, runLog, runnerId, runnerKey });
        if (!finishSuccess) {
            logger.info('Releasing run', {
                operation: 'runLifecycle',
                runId: run.id,
                runnerId: runnerId
            });
            await releaseRun({ run, runLog, runnerId, runnerKey });
        }
        return true;
    }
    catch (error) {
        logger.error('Failed to execute runner lifecycle', {
            operation: 'runLifecycle',
            runnerId: runnerId,
            errorMessage: error.message,
            errorStack: error.stack
        });
        throw error;
    }
}
module.exports = {
    runLifecycle,
    initializeConnectionOptions
};
