// @postman/pm-bin-linux-x64/bin/lib/commands/runner/core/runner.js
const PackageResolver = require('../../../run/package-resolver');
const RunnerPackageLoader = require('../data/package-loader');
const RequestResolver = require('../../../run/request-resolver');
const RunnerRequestLoader = require('../data/request-loader');
const SecureFS = require('../../../run/secure-fs');
const DataFileLoader = require('../data/data-file-loader');
const logRuntimeEvent = require('../logging/runtime-event-logger');
const { Collection, ProxyConfigList } = require('postman-collection');
const { getLogger } = require('../logging/logger-factory');
const RunLog = require('@postman/newman-remote-run-log'), runtime = require('postman-runtime'), _ = require('lodash'), runnerConfig = require('./runner-config'), analytics = require('../analytics'), NETWORK_ERRORS = ['ECONNRESET'];
const logger = getLogger();
const RESTRICTED_ADDRESSES = {
    localhost: true,
    '127.0.0.1': true,
    '0.0.0.0': true,
    127.1: true,
    0: true,
    '::': true,
    '[::]': true,
    '::0': true,
    '[::0]': true,
    '::1': true,
    '[::1]': true,
    '0:0:0:0:0:0:0:0': true,
    '[0:0:0:0:0:0:0:0]': true,
    '0:0:0:0:0:0:0:1': true,
    '[0:0:0:0:0:0:0:1]': true,
    '': true,
    '127.0.0.1/8': true,
    '::1/128': true,
    '169.254.169.254': true,
    '100.100.100.200': true
};
function failsafe(promise, timeout) {
    return new Promise((resolve, reject) => {
        const failsafeTimeout = timeout + 1000, timeoutId = setTimeout(() => {
            reject(new Error(`Postman Runtime did not respond within ${failsafeTimeout}ms`));
        }, failsafeTimeout);
        promise.then((result) => {
            clearTimeout(timeoutId);
            resolve(result);
        }, (error) => {
            clearTimeout(timeoutId);
            reject(error);
        });
    });
}
function createCallbacks(callback) {
    return {
        start(error, cursor) {
            callback('start', { error, cursor });
        },
        beforeIteration(error, cursor) {
            callback('beforeIteration', { error, cursor });
        },
        beforeItem(error, cursor, item) {
            callback('beforeItem', { error, cursor, item });
        },
        beforePrerequest(error, cursor, events, item) {
            callback('beforePrerequest', { error, cursor, events, item });
        },
        prerequest(error, cursor, scriptResults, item) {
            callback('prerequest', { error, cursor, scriptResults, item });
        },
        beforeRequest(error, cursor, request, item) {
            callback('beforeRequest', { error, cursor, request, item });
        },
        request(error, cursor, response, request, item) {
            callback('request', { error, cursor, response, request, item });
        },
        responseStart(error, cursor, response, request, item) {
            callback('responseStart', { error, cursor, response, request, item });
        },
        response(error, cursor, response, request, item) {
            callback('response', { error, cursor, response, request, item });
        },
        io(error, cursor, trace) {
            callback('io', { error, cursor, trace });
        },
        beforeTest(error, cursor, events, item) {
            callback('beforeTest', { error, cursor, events, item });
        },
        beforeScript(error, cursor, script, event, item) {
            callback('beforeScript', { error, cursor, script, event, item });
        },
        script(error, cursor, result, script, event, item) {
            callback('script', { error, cursor, result, event, item });
        },
        assertion(cursor, assertion) {
            callback('assertion', { cursor, assertion });
        },
        exception(cursor, error) {
            callback('exception', { cursor, error });
        },
        console(cursor, level, ...message) {
            callback('console', { cursor, level, message });
        },
        test(error, cursor, scriptResults, item) {
            callback('test', { error, cursor, scriptResults, item });
        },
        item(error, cursor, item) {
            callback('item', { error, cursor, item });
        },
        iteration(error, cursor) {
            callback('iteration', { error, cursor });
        },
        pause(error, cursor) {
            callback('pause', { error, cursor });
        },
        resume(error, cursor) {
            callback('resume', { error, cursor });
        },
        error(error) {
            callback('error', { error });
        },
        done(error, cursor) {
            callback('done', { error, cursor });
        }
    };
}
function sanitizeQueryParamsForEvent(args) {
    const clonedArgs = _.cloneDeep(args);
    if (clonedArgs.request) {
        clonedArgs.request.url.query.clear();
    }
    if (clonedArgs.result && clonedArgs.result.request) {
        clonedArgs.result.request.url.query.clear();
    }
    if (clonedArgs.item && clonedArgs.item.request) {
        clonedArgs.item.request.url.query.clear();
    }
    if (clonedArgs.scriptResults && clonedArgs.scriptResults.length) {
        for (const scriptResult of clonedArgs.scriptResults) {
            if (scriptResult.result && scriptResult.result.request) {
                scriptResult.result.request.url.clear();
            }
        }
    }
    return clonedArgs;
}
function checkForFailedTests(runLog, entry) {
    const assertions = runLog.getAssertions([entry]);
    for (const assertion of assertions) {
        if (!assertion.passed) {
            return new Error('This monitor encountered an error after running successfully for a while. To avoid ' +
                'false error reports, the Monitoring platform will automatically re-run the monitor when it encounters ' +
                'an error. The run will be re-attempted twice at most, and it will be marked as a failure if the third ' +
                'and final attempt also fails. In the current run, the following assertion failed:' +
                assertion.name);
        }
    }
}
function doRun({ runnerId, job, run, runLog, collection, options, runConfig, sanitizeQueryParams, eventStreamer }) {
    return new Promise((resolve, reject) => {
        const postmanRunner = new runtime.Runner();
        postmanRunner.run(collection, options, (initializationError, postmanRun) => {
            if (initializationError) {
                return reject(initializationError);
            }
            logger.info('Starting run execution', {
                operation: 'runExecution',
                runId: run.id,
                jobId: job.id,
                runnerId: runnerId
            });
            let abortedError;
            const fastFail = (eventName, args, entry) => {
                if (options && run.attemptCount >= options.maxRunAttempts) {
                    return;
                }
                if (abortedError) {
                    return;
                }
                let error = args.error, previousRunsWereSuccessful = job.consecutiveCount >= options.healthyConsecutiveCount;
                if (error && NETWORK_ERRORS.includes(error.code)) {
                    abortedError = error;
                }
                else if (error && previousRunsWereSuccessful) {
                    abortedError = error;
                }
                else if (previousRunsWereSuccessful) {
                    error = checkForFailedTests(runLog, entry);
                    if (error) {
                        abortedError = error;
                    }
                }
                if (abortedError) {
                    postmanRun.abort();
                }
            }, processEvent = (eventName, args) => {
                logRuntimeEvent(eventName, args, {
                    runnerId: runnerId,
                    runId: run.id,
                    jobId: job.id
                });
                let eventArgs = args;
                if (sanitizeQueryParams || runConfig.sanitizeQueryParams) {
                    eventArgs = sanitizeQueryParamsForEvent(args);
                }
                const entry = runLog.push(eventName, eventArgs), testError = checkForFailedTests(runLog, entry);
                if (eventStreamer) {
                    try {
                        eventStreamer.onEvent(eventName, eventArgs);
                    }
                    catch (err) {
                    }
                }
                fastFail(eventName, args, entry);
                if (testError) {
                    logger.error('Test failed', {
                        operation: 'runExecution',
                        runId: run.id,
                        errorMessage: testError.message
                    });
                }
                if (eventName === 'error' && args.error.message === 'callback timed out') {
                    logger.error('Run timed out', {
                        operation: 'runExecution',
                        runId: run.id,
                        eventName: eventName
                    });
                }
                if (eventName === 'done') {
                    if (abortedError) {
                        reject(abortedError);
                    }
                    else {
                        resolve();
                    }
                }
            };
            postmanRun.start(createCallbacks(processEvent));
        });
    });
}
class Runner {
    constructor({ job, run, collection, environment, jobInput, runConfig, eventIngestion, data, runTrace, runnerId, runnerKey, workingDir, connectionOptions = {}, reportEvents }) {
        if (!runnerId || typeof runnerId !== 'string') {
            throw new Error('Runner ID is required and must be a string');
        }
        if (!runnerKey || typeof runnerKey !== 'string') {
            throw new Error('Runner Key is required and must be a string');
        }
        this.job = job;
        this.run = run;
        this.finished = false;
        this.runTrace = runTrace;
        this.environment = environment;
        this.data = data;
        this.collection = collection;
        this.jobInput = jobInput;
        this.runConfig = runConfig;
        this.eventIngestion = eventIngestion;
        this.runnerId = runnerId;
        this.runnerKey = runnerKey;
        this.sanitizeQueryParams = _.get(runConfig, 'sanitizeQueryParams');
        this.config = null;
        this.timeout = null;
        this.options = null;
        this.workingDir = workingDir || process.cwd();
        this.insecureFileRead = false;
        this.connectionOptions = connectionOptions;
        this.reportEvents = reportEvents;
        this.eventStreamer = null;
        this.collection = new Collection(collection);
        this.runLog = new RunLog({ job, run });
        if (jobInput) {
            this.runLog.push('input', { request: jobInput });
        }
    }
    async init() {
        this.config = runnerConfig.get();
        const executionTimeout = _.get(this.runConfig, 'executionTimeout'), maxTimeout = _.get(this.config, 'timeout');
        if (!isNaN(executionTimeout)) {
            if (executionTimeout <= maxTimeout) {
                this.timeout = executionTimeout;
            }
            else {
                logger.error('Execution timeout exceeds maximum allowed timeout', {
                    operation: 'runInitialization',
                    executionTimeout: executionTimeout,
                    maxTimeout: maxTimeout,
                    runId: this.run.id
                });
            }
        }
        this.timeout = this.timeout || maxTimeout;
        this.options = await this.getRunnerOptions(this.runConfig);
        this.eventStreamer = analytics.create({
            runId: this.run.id,
            jobId: this.job.id,
            runnerId: this.runnerId,
            collectionId: this.collection?.id || '',
            collectionName: this.collection?.name || '',
            workspaceId: this.eventIngestion?.monitorWorkspaceId || '',
            environmentId: this.environment?.id || '',
            environmentName: this.environment?.name || '',
            trigger: this.job.trigger || '',
            orgId: this.job?.team || this.run?.team || '',
            monitorId: this.job?.jobTemplate || '',
            eventIngestion: this.eventIngestion,
            reportEvents: this.reportEvents
        });
        if (this.eventStreamer) {
            this.eventStreamer.setRunLog(this.runLog);
        }
    }
    getRunnerOptions(runConfig) {
        const retryConfig = _.get(runConfig, 'retryConfig'), maxRunAttempts = _.get(this.config, 'maxRunAttempts'), healthyConsecutiveCount = _.get(this.config, 'healthyConsecutiveCount'), useWhatWGUrlParser = _.get(this.config, 'useWhatWGUrlParser');
        const options = {
            stopOnError: false,
            maxRunAttempts: _.get(retryConfig, 'attempts') ? retryConfig.attempts + 1 : maxRunAttempts,
            healthyConsecutiveCount: _.get(retryConfig, 'retryAfter', healthyConsecutiveCount),
            timeout: {
                global: this.timeout,
                request: this.job.requestTimeout,
                script: 0
            },
            requester: {
                useWhatWGUrlParser: useWhatWGUrlParser,
                strictSSL: Boolean(this.job.strictSSL),
                followRedirects: Boolean(this.job.followRedirects),
                systemHeaders: {
                    'x-postman-monitor-trace': this.runTrace
                },
                ...(this.connectionOptions.caCertPath ? { extendedRootCA: this.connectionOptions.caCertPath } : {})
            },
            network: {
                restrictedAddresses: RESTRICTED_ADDRESSES
            },
            delay: {
                item: this.job.requestDelay
            },
            environment: {
                name: _.get(this.environment, 'name', ''),
                values: this.environment ? this.environment.values : []
            },
            globals: {
                values: [{
                        key: 'previousRequest',
                        value: JSON.stringify(this.jobInput)
                    }]
            },
            fileResolver: new SecureFS({
                workingDir: this.workingDir,
                insecureFileRead: this.insecureFileRead,
                fileLoader: new DataFileLoader({
                    runnerId: this.runnerId,
                    runnerKey: this.runnerKey,
                    runId: this.run.id
                })
            }),
            script: {
                packageResolver: (() => {
                    const packageLoader = new RunnerPackageLoader({
                        runnerId: this.runnerId,
                        runnerKey: this.runnerKey,
                        runId: this.run.id
                    });
                    const resolver = new PackageResolver({ packageLoader, logger });
                    return resolver.resolvePackages;
                })(),
                requestResolver: (() => {
                    const requestLoader = new RunnerRequestLoader({
                        runnerId: this.runnerId,
                        runnerKey: this.runnerKey,
                        runId: this.run.id
                    });
                    const resolver = new RequestResolver({ requestLoader });
                    return resolver.resolveRequest;
                })()
            },
            ...(this.connectionOptions.proxyOptions ? { proxies: new ProxyConfigList({}, [this.connectionOptions.proxyOptions]) } : {})
        };
        if (runConfig && runConfig.iterationCount) {
            options.iterationCount = runConfig.iterationCount;
        }
        if (this.data) {
            options.data = this.data;
            if (!options.iterationCount) {
                options.iterationCount = this.data.length < _.get(this.config, 'limits.maxIterationCount') ?
                    this.data.length :
                    _.get(this.config, 'limits.maxIterationCount');
            }
        }
        if (runConfig && runConfig.requestSelection) {
            const execute = _.map(runConfig.requestSelection, 'id'), lookupStrategy = 'followOrder';
            options.entrypoint = {
                execute,
                lookupStrategy
            };
        }
        return options;
    }
    async exec() {
        try {
            await this.init();
            const promise = doRun({
                runnerId: this.runnerId,
                job: this.job,
                run: this.run,
                runLog: this.runLog,
                collection: this.collection,
                options: this.options,
                runConfig: this.runConfig,
                sanitizeQueryParams: this.sanitizeQueryParams,
                eventStreamer: this.eventStreamer
            });
            await failsafe(promise, this.timeout);
            this.finished = true;
        }
        catch (error) {
            logger.error('Error executing run', {
                operation: 'runExecution',
                runId: this.run.id,
                errorMessage: error.message,
                errorStack: error.stack
            });
            this.runLog.push('error', { error });
            if (this.eventStreamer) {
                try {
                    this.eventStreamer.onEvent('error', { error });
                }
                catch (err) {
                }
            }
            if (this.options && this.run.attemptCount >= this.options.maxRunAttempts) {
                this.finished = true;
            }
            else {
                this.finished = false;
            }
        }
        finally {
            if (this.eventStreamer) {
                try {
                    await this.eventStreamer.flush();
                    this.eventStreamer.destroy();
                }
                catch (err) {
                }
            }
        }
        return this.runLog;
    }
}
module.exports = Runner;
