// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/Util/OpenAPI.js
const fs = require('fs'), _ = require('lodash'), openAPI = require('openapi-to-postmanv2'), { resolveRelativePath, convertToPosixPath } = require('./FilePath');
async function detectMissingFilePaths(rootFileData, schemaFilesData, specificationVersion) {
    try {
        const relatedFiles = await openAPI.detectRelatedFiles({
            origin: 'browser',
            type: 'multiFile',
            specificationVersion: specificationVersion,
            rootFiles: [rootFileData],
            data: schemaFilesData
        }), allResults = _.get(relatedFiles, 'output.data', []);
        let allMissingRelatedFiles = [];
        _.forEach(allResults, (result) => {
            const missingRelatedFiles = _.get(result, 'missingRelatedFiles', []), missingRelatedFilesFormatted = _.map(missingRelatedFiles, ({ path: filePath }) => {
                return convertToPosixPath(filePath);
            });
            allMissingRelatedFiles = _.concat(allMissingRelatedFiles, missingRelatedFilesFormatted);
        });
        return allMissingRelatedFiles;
    }
    catch (err) {
        const errObj = {
            error: {
                message: 'Failed to read provided spec files.'
            }
        };
        throw errObj;
    }
}
async function getAllRelatedFiles(rootFilePath, specificationVersion) {
    const resolvedRootFilePath = resolveRelativePath(rootFilePath), rootFileContent = fs.readFileSync(resolvedRootFilePath, 'utf8'), rootFileData = {
        path: convertToPosixPath(resolvedRootFilePath),
        content: rootFileContent
    }, schemaFilesData = [{
            path: convertToPosixPath(resolvedRootFilePath),
            content: rootFileContent
        }];
    let schemaFilesDetected = [], schemaFilesToBeDetected = await detectMissingFilePaths(rootFileData, schemaFilesData, specificationVersion);
    do {
        for (const schemaFile of schemaFilesToBeDetected) {
            let schemaFileContent = '';
            const resolvedSchemaFilePath = resolveRelativePath(schemaFile);
            try {
                schemaFileContent = fs.readFileSync(resolvedSchemaFilePath, 'utf8');
            }
            catch (e) {
                const errObj = {
                    error: {
                        message: `Could not find the schema file at ${schemaFile}.` +
                            '\nYour spec might be referencing files that do not exist.'
                    }
                };
                throw errObj;
            }
            schemaFilesData.push({
                path: convertToPosixPath(resolvedSchemaFilePath),
                content: schemaFileContent
            });
        }
        schemaFilesDetected = schemaFilesToBeDetected;
        schemaFilesToBeDetected = await detectMissingFilePaths(rootFileData, schemaFilesData, specificationVersion) || [];
    } while (!_.isEmpty(schemaFilesToBeDetected) || !_.isEqual(schemaFilesToBeDetected, schemaFilesDetected));
    return schemaFilesData;
}
async function detectRootFiles(files, version) {
    const result = await openAPI.detectRootFiles({
        type: 'multiFile',
        specificationVersion: version,
        data: files
    });
    if (result.result && result.output && result.output.data && result.output.data.length > 0) {
        return { ...result, version };
    }
    return null;
}
module.exports = {
    getAllRelatedFiles,
    detectRootFiles
};
