// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/analytics.js
const { v4: uuidv4 } = require('uuid'), { createClient, context } = require('../../framework/clickhouse'), TABLE = 'postman_spec_lint_events_v3';
async function sendLintEvents({ specId, workspaceId, violations, lintPassed }) {
    const client = createClient(), hasAuth = await client.checkAuth();
    if (!hasAuth) {
        return;
    }
    const now = new Date().toISOString(), eventSpan = uuidv4(), cliVersion = require('../../../package.json').version, rows = [];
    await context.insert(client, {
        eventSpan: eventSpan,
        executorType: 'spec-lint',
        workspaceId: workspaceId,
        cliVersion: cliVersion
    });
    if (Array.isArray(violations) && violations.length) {
        for (const v of violations) {
            let lineNumber = '';
            if (v?.range?.start?.line >= 0) {
                lineNumber = String(v.range.start.line + 1);
            }
            rows.push({
                timestamp: now,
                spec_id: specId || '',
                lint_pass: Boolean(lintPassed),
                workspace_id: workspaceId || '',
                file: v.file || '',
                line_number: lineNumber,
                path: Array.isArray(v.path) ? v.path.join('.') : (v.path || ''),
                severity: v.severity || '',
                issue: v.message || ''
            });
        }
    }
    else {
        rows.push({
            timestamp: now,
            spec_id: specId || '',
            lint_pass: Boolean(lintPassed),
            workspace_id: workspaceId || '',
            file: '',
            line_number: '',
            path: '',
            severity: lintPassed ? '' : 'ERROR',
            issue: lintPassed ? '' : 'Validation failed'
        });
    }
    try {
        await client.insert(TABLE, rows);
    }
    catch {
    }
}
module.exports = { sendLintEvents: sendLintEvents };
