// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/constants.js
const LINE_BREAK = '\n', BATCH_SIZE = 5, FILE_TYPES = {
    ROOT: 'ROOT'
}, SYNTAX_ERROR_CODE = {
    PARSER: 'parser'
}, OUTPUT_FORMATS = {
    JSON: 'JSON',
    CSV: 'CSV'
}, FAILED_SEVERITY_DEFAULT_VALUE = 'ERROR', SEVERITY_MAP = {
    0: 'ERROR',
    1: 'WARNING',
    2: 'INFO',
    3: 'HINT'
}, SEVERITY_COLOR_MAP = {
    ERROR: 'red',
    WARNING: 'yellow',
    INFO: 'cyan',
    HINT: 'white'
}, EXIT_CODE = {
    OK: 0,
    ERROR: 1
}, SEVERITY_FAIL_ORDER = {
    ERROR: 4,
    WARNING: 3,
    INFO: 2,
    HINT: 1
}, SPECTRAL_VERSION_6 = 'SPECTRAL 6.0', SPECIFICATION_INPUT_TYPES = {
    ID: 'ID',
    FILE_PATH: 'FILE_PATH'
}, VALIDATION_TYPES = {
    SYNTAX: 'SYNTAX',
    GOVERNANCE: 'GOVERNANCE'
};
module.exports = {
    LINE_BREAK,
    SEVERITY_COLOR_MAP,
    EXIT_CODE,
    SEVERITY_FAIL_ORDER,
    SPECTRAL_VERSION_6,
    SEVERITY_MAP,
    BATCH_SIZE,
    FILE_TYPES,
    OUTPUT_FORMATS,
    FAILED_SEVERITY_DEFAULT_VALUE,
    SYNTAX_ERROR_CODE,
    SPECIFICATION_INPUT_TYPES,
    VALIDATION_TYPES
};
