// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/index.js
require('../../../lib/logger').init();
const _ = require('lodash');
const fs = require('fs');
const path = require('path');
const colors = require('colors/safe');
const GovernanceSandbox = require('@postman/governance-sandbox');
const { RuleUtils } = require('@postman/api-governance-utils');
const SpecificationService = require('./Service/SpecificationService');
const RuleService = require('./Service/RuleService');
const SyntaxParserService = require('./Service/SyntaxParserService');
const tableFormatter = require('./formatters/tableFormatter');
const outputResultsInCSV = require('./formatters/csvFormatter');
const outputResultsInJson = require('./formatters/jsonFormatter');
const print = require('../../framework/print');
const authUtils = require('../util');
const cliUtils = require('../../util');
const util = require('util');
const { SPECTRAL_VERSION_6, EXIT_CODE, LINE_BREAK, SEVERITY_FAIL_ORDER, SEVERITY_MAP, OUTPUT_FORMATS, FILE_TYPES, SPECIFICATION_INPUT_TYPES, VALIDATION_TYPES } = require('./constants');
const { v4 } = require('uuid');
const { sendLintEvents } = require('./analytics');
function filterSpectralConfigByGovernanceTag(spectralRules) {
    const filteredSpectralRules = {};
    Object.keys(spectralRules.rules).forEach((ruleKey) => {
        const rule = spectralRules.rules[ruleKey], hasGovernanceTag = rule &&
            rule.config &&
            Array.isArray(rule.config.tags) &&
            rule.config.tags.includes('governance');
        if (hasGovernanceTag && spectralRules.spectralConfig.rules[ruleKey]) {
            filteredSpectralRules[ruleKey] = spectralRules.spectralConfig.rules[ruleKey];
        }
    });
    return {
        rules: filteredSpectralRules
    };
}
function resolveWorkspaceIdFromLocalConfig(existingWorkspaceId) {
    return new Promise((resolve) => {
        if (existingWorkspaceId) {
            return resolve(existingWorkspaceId);
        }
        const workingDir = process.cwd(), configPath = path.join(workingDir, '.postman', 'config.json');
        fs.readFile(configPath, 'utf8', (err, data) => {
            if (err) {
                return resolve(existingWorkspaceId);
            }
            try {
                const parsed = JSON.parse(data), workspaceId = _.get(parsed, 'workspace.id') || _.get(parsed, 'workspace_id');
                return resolve(workspaceId || existingWorkspaceId);
            }
            catch (e) {
                return resolve(existingWorkspaceId);
            }
        });
    });
}
function extractRelevantSyntaxInfo(syntaxResults) {
    const syntaxInfo = [];
    syntaxResults.forEach((result) => {
        syntaxInfo.push({
            message: result.message,
            path: result.path,
            range: result.range,
            severity: result.severity,
            file: result.file,
            code: result.code,
            type: VALIDATION_TYPES.SYNTAX
        });
    });
    return syntaxInfo;
}
function extractRelevantInfoFromGovernanceViolation(governanceVoilation) {
    const governanceVoilationInfo = [];
    governanceVoilation.forEach((violation) => {
        const { message, path, range, severity, source } = violation;
        governanceVoilationInfo.push({
            message: message,
            path: path,
            range: range,
            severity: SEVERITY_MAP[severity],
            file: source,
            type: VALIDATION_TYPES.GOVERNANCE
        });
    });
    return governanceVoilationInfo;
}
function checkForFailSeverity(governanceViolationResults, failSeverity) {
    return governanceViolationResults.some((result) => {
        if (!failSeverity || !result.severity) {
            return false;
        }
        return SEVERITY_FAIL_ORDER[failSeverity] <= SEVERITY_FAIL_ORDER[result.severity.toUpperCase()];
    });
}
function runGovernanceValidation({ specificationFileDetails, spectralGovernanceRules, customFunctionsBundle }) {
    const files = {
        rootFile: specificationFileDetails.find((file) => { return file.type === FILE_TYPES.ROOT; }),
        nonRootFiles: specificationFileDetails
    };
    return new Promise((resolve, reject) => {
        const id = v4(), code = `
                let global = {};
                // custom functions bundle
                ${customFunctionsBundle}
                let files = ${JSON.stringify({ rootFile: files.rootFile, nonRootFiles: files.nonRootFiles })};
                let rules = ${JSON.stringify(spectralGovernanceRules)};
                // api validator function
                apiValidator.validateMultiFileSpec(files, rules, global).then((warnings)=>{
                    apiValidator.result({ 'type': 'warnings', 'warnings': warnings, 'id': '${id}' });
                }).catch((error)=>{
                    apiValidator.result({ 'type': 'error', error, 'id': '${id}' });
                });
            `;
        GovernanceSandbox.createContext({}, (err, ctx) => {
            if (err) {
                logger.error(`Error while creating context for governance sandbox - ${JSON.stringify(err)}`);
                return reject(new Error('Failed to initialize the validation environment. Please try again.'));
            }
            ctx.on('apiValidator.result', (message) => {
                if (!message || !message.type) {
                    logger.error(`Invalid message received from sandbox - ${JSON.stringify(message)}`);
                    return reject(new Error('An unexpected issue occurred during validation.'));
                }
                switch (message.type) {
                    case 'error':
                        logger.error(`Error while validating spec - ${JSON.stringify(message)}`);
                        return reject(new Error('Validation failed due to an internal error. Please try again later.'));
                    case 'warnings':
                        return resolve(message.warnings);
                    default:
                        logger.error(`Sandbox: unhandled result type:${message.type} received`);
                        return reject(new Error('Unexpected validation result. Please contact support if the issue persists.'));
                }
            });
            ctx.on('error', (err) => {
                err && logger.error(`Error thrown by sandbox - ${JSON.stringify(err)}`);
                return reject(new Error('An error occurred during validation. Please try again.'));
            });
            ctx.execute(code, (err) => {
                if (err) {
                    logger.error(`Error thrown by sandbox while executing - ${JSON.stringify(err)}`);
                    return reject(new Error('Validation could not be completed due to an internal error.'));
                }
            });
        });
    });
}
function transformComponentViolations(violations) {
    for (const violation of violations) {
        if (violation.file.startsWith('CommonComponents_')) {
            violation.file = 'Component: ' + violation.file.slice('CommonComponents_'.length, -'.yaml'.length);
        }
    }
}
module.exports = {
    validateSpec: async (specificationInput, command) => {
        if (!specificationInput) {
            throw new Error('Specification input is not provided');
        }
        try {
            const options = await util.promisify(authUtils.addApiKeyToOptions)(command, {}), commanderOptions = cliUtils.commanderToObject(command), { failSeverity, output: outputFormat, workspaceId: rawWorkspaceId } = commanderOptions, workspaceId = await resolveWorkspaceIdFromLocalConfig(rawWorkspaceId), specificationInputType = cliUtils.isPostmanResource(specificationInput) ?
                SPECIFICATION_INPUT_TYPES.ID : SPECIFICATION_INPUT_TYPES.FILE_PATH, [rulesets, specificationDetails] = await Promise.all([
                RuleService.fetchGovernanceRulesets(specificationInput, specificationInputType, { workspaceId }, options),
                SpecificationService.fetchSpecificationDetails(specificationInput, specificationInputType, options)
            ]), spectralRules = RuleUtils.getSpectralRules(rulesets, {
                isEnabled: true,
                engine: SPECTRAL_VERSION_6,
                entityConfig: { type: specificationDetails.type }
            }), customFunctionsBundle = _.get(rulesets, 'customFunctionsBundle.value', '');
            if (!spectralRules || !spectralRules.rules || !spectralRules.spectralConfig) {
                print(`${colors.yellow('No rules found for linting.')}${LINE_BREAK}`);
                return { code: EXIT_CODE.OK };
            }
            const spectralGovernanceRules = filterSpectralConfigByGovernanceTag(spectralRules), governanceRuleCodes = Object.keys(spectralRules.rules);
            const syntaxResults = await SyntaxParserService.parseSyntax({ specificationFileDetails: specificationDetails.files }), syntaxViolationResults = extractRelevantSyntaxInfo(syntaxResults);
            const governanceValidationResults = await runGovernanceValidation({
                specificationFileDetails: specificationDetails.files,
                spectralGovernanceRules: spectralGovernanceRules,
                customFunctionsBundle: customFunctionsBundle
            });
            const filteredGovernanceResults = governanceValidationResults.filter((result) => { return governanceRuleCodes.includes(String(result.code)); }), governanceViolationResults = extractRelevantInfoFromGovernanceViolation(filteredGovernanceResults);
            const allResults = {
                syntax: syntaxViolationResults,
                governance: governanceViolationResults
            };
            const allViolations = [...syntaxViolationResults, ...governanceViolationResults];
            const isFailSeverityFound = checkForFailSeverity(allViolations, failSeverity);
            transformComponentViolations(allViolations);
            if (outputFormat === OUTPUT_FORMATS.CSV) {
                const violationsInCSVFormat = outputResultsInCSV(allResults);
                print(violationsInCSVFormat);
            }
            else if (outputFormat === OUTPUT_FORMATS.JSON) {
                const violationsInJsonFormat = outputResultsInJson(allResults);
                print(JSON.stringify(violationsInJsonFormat, null, 2));
                print(LINE_BREAK);
            }
            else {
                specificationDetails.inputType = specificationInputType;
                const violationsInTableFormat = tableFormatter(specificationDetails, allResults);
                print(violationsInTableFormat);
            }
            if (commanderOptions.reportEvents) {
                try {
                    const specIdForAnalytics = specificationInputType === SPECIFICATION_INPUT_TYPES.ID ?
                        (specificationDetails.id || '') : (specificationDetails.path || ''), lintPassed = !isFailSeverityFound;
                    await sendLintEvents({
                        specId: specIdForAnalytics,
                        workspaceId: workspaceId,
                        violations: allViolations,
                        lintPassed: lintPassed
                    });
                }
                catch (e) { }
            }
            return {
                code: isFailSeverityFound ? EXIT_CODE.ERROR : EXIT_CODE.OK
            };
        }
        catch (err) {
            logger.error(`Error in validateSpec: ${JSON.stringify(err)}`);
            throw new Error(`${err.message || err}`);
        }
    }
};
