// @postman/pm-bin-linux-x64/bin/lib/logger/index.js
const winston = require('winston'), fs = require('fs'), path = require('path'), os = require('os'), console = require('../framework/console'), cliEnvironment = require('../config/cli-environment'), V1_API_KEY_REGEX = /[a-f0-9]{32}/g, V2_API_KEY_REGEX = /(PMAK-[a-f0-9]{24}-[a-f0-9]{34})/g, logger = {
    init() {
        const settings = cliEnvironment.getSettings(), postmanHomeDir = settings.postmanHomeDir || '.postman';
        let origin = 'postman-cli', sessionId = process.pid, homeDir = os.homedir(), logPath = path.join(homeDir, postmanHomeDir, 'logs');
        if (!homeDir || homeDir === '/') {
            homeDir = process.cwd();
            logPath = path.join(homeDir, postmanHomeDir, 'logs');
        }
        try {
            if (!fs.existsSync(logPath)) {
                fs.mkdirSync(logPath, { recursive: true });
            }
        }
        catch (error) {
            console.warn(`Warning: Could not create log directory at ${logPath}: ${error.message}`);
            logPath = path.join(process.cwd(), postmanHomeDir, 'logs');
            try {
                if (!fs.existsSync(logPath)) {
                    fs.mkdirSync(logPath, { recursive: true });
                }
            }
            catch (fallbackError) {
                console.warn('Warning: Could not create fallback log directory at ' +
                    `${logPath}: ${fallbackError.message}`);
                logPath = null;
            }
        }
        global.logger = winston.createLogger({
            level: 'info',
            format: winston.format.printf((info) => {
                info.message = (info.message).replace(V2_API_KEY_REGEX, '[[REDACTED_API_KEY]]');
                info.message = (info.message).replace(V1_API_KEY_REGEX, '[[REDACTED_API_KEY]]');
                return `[${sessionId}][${new Date().getTime()}][${origin}][${info.level}][${info.message}]`;
            }),
            transports: [
                ...(logPath ? [new winston.transports.File({
                        filename: path.join(logPath, 'postman-cli.log'),
                        level: 'info',
                        maxFiles: 5,
                        maxsize: 5 * 1024 * 1024
                    })] : [])
            ]
        });
    }
};
module.exports = logger;
