// @postman/pm-bin-linux-x64/bin/lib/api/ec-resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ecProxyRequest = ecProxyRequest;
exports.patchECCollection = patchECCollection;
exports.getECItems = getECItems;
exports.deleteECItem = deleteECItem;
exports.createECItem = createECItem;
exports.createECItemsRecursive = createECItemsRecursive;
exports.updateExtensibleCollection = updateExtensibleCollection;
const requestFactory = require('@postman/platform-req/lib/request');
const util = require('../util');
function getClient() {
    return requestFactory.defaults({
        timeout: 300000
    });
}
function buildHeaders(options) {
    const headers = {
        'Content-Type': 'application/json'
    };
    if (options.postmanApiKey) {
        headers['X-Api-Key'] = options.postmanApiKey;
    }
    else if (options.accessToken) {
        headers['x-access-token'] = options.accessToken;
    }
    return headers;
}
function ecProxyRequest(method, ecPath, body, options) {
    return new Promise((resolve, reject) => {
        const client = getClient(), gatewayUrl = util.POSTMAN_GATEWAY_BASE_URL(), url = `${gatewayUrl}${ecPath}`, headers = buildHeaders(options);
        headers['x-pstmn-req-service'] = 'collection-api';
        const requestOptions = { url, headers };
        if (body !== undefined) {
            requestOptions.body = JSON.stringify(body);
        }
        client[method](requestOptions, (err, response, responseBody) => {
            if (err) {
                reject(new Error(`EC request failed: ${err.message}`));
                return;
            }
            let parsed = responseBody;
            try {
                if (typeof parsed === 'string') {
                    parsed = JSON.parse(parsed.trim());
                }
            }
            catch {
                reject(new Error('EC service returned invalid JSON'));
                return;
            }
            if (response && !(/2../).test(String(response.statusCode))) {
                const msg = parsed?.error?.message || parsed?.message || `HTTP ${response.statusCode}`;
                reject(new Error(`EC request failed (${method.toUpperCase()} ${ecPath}): ${msg}`));
                return;
            }
            resolve(parsed?.data ?? parsed);
        });
    });
}
async function patchECCollection(collectionId, patch, options) {
    await ecProxyRequest('patch', `/collections/${collectionId}`, patch, options);
}
async function getECItems(collectionId, options) {
    const result = await ecProxyRequest('get', `/collections/${collectionId}/items/`, undefined, options);
    return Array.isArray(result) ? result : [];
}
async function deleteECItem(collectionId, itemId, options) {
    await ecProxyRequest('delete', `/collections/${collectionId}/items/${itemId}`, undefined, options);
}
function createECItem(collectionId, itemData, options) {
    return ecProxyRequest('post', `/collections/${collectionId}/items/`, itemData, options);
}
async function createECItemsRecursive(collectionId, items, parentId, options) {
    let previousSiblingId;
    for (const item of items) {
        const newItem = {
            type: item.type,
            title: item.title ?? '',
            position: {
                parent: parentId,
                ...(previousSiblingId ? { previousSibling: previousSiblingId } : {})
            },
            payload: item.payload ?? {},
            extensions: item.extensions ?? {}
        };
        const created = await createECItem(collectionId, newItem, options);
        const createdId = created?.id;
        if (!createdId) {
            continue;
        }
        previousSiblingId = createdId;
        if (item.children && item.children.length > 0) {
            await createECItemsRecursive(collectionId, item.children, createdId, options);
        }
    }
}
async function updateExtensibleCollection(cloudCollectionId, extensible, collectionTitle, options) {
    const collectionPatch = [
        { op: 'replace', path: '/title', value: extensible.title ?? collectionTitle },
        { op: 'replace', path: '/payload', value: extensible.payload ?? {} },
        { op: 'replace', path: '/extensions', value: extensible.extensions ?? {} }
    ];
    await patchECCollection(cloudCollectionId, collectionPatch, options);
    const serverItems = await getECItems(cloudCollectionId, options);
    for (const item of serverItems) {
        try {
            await deleteECItem(cloudCollectionId, item.id, options);
        }
        catch {
        }
    }
    await createECItemsRecursive(cloudCollectionId, extensible.children ?? [], cloudCollectionId, options);
}
