// @postman/pm-bin-linux-x64/bin/lib/api/externalPackages.js
const requestFactory = require('@postman/platform-req/lib/request');
const { minimatch } = require('minimatch');
const util = require('../util');
const externalPackageErrorHandler = require('./externalPackageErrorHandler');
const { jwtDecode } = require('jwt-decode');
const SUPPORTED_REGISTRIES = ['npm', 'jsr'];
const MAX_PACKAGE_REQUIRED_LENGTH = 256;
const AccessScope = {
    ALL: 'all',
    NONE: 'none',
    RESTRICTED: 'restricted'
};
function getClient() {
    return requestFactory.defaults({
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: {
                    enabled: true
                }
            }
        }
    });
}
function platformRequestWrapper({ url, headers, method, json = false }) {
    const client = getClient();
    return new Promise((resolve, reject) => {
        client[method]({
            url,
            headers,
            json
        }, (err, response) => {
            if (err) {
                reject(err);
                return;
            }
            resolve(response);
        });
    });
}
class ExternalPackageManager {
    constructor() {
        this._token = null;
        this._userType = null;
        this._fetchTokenPromise = null;
        this._accessDetails = null;
        this._fetchAccessDetailsPromise = null;
    }
    static isExternalPackage(packageId) {
        const [packageRegistry, packageIdentifier] = packageId.split(':');
        if (!packageIdentifier || !SUPPORTED_REGISTRIES.includes(packageRegistry)) {
            return false;
        }
        return true;
    }
    _fetchAuthToken(postmanApiKey, accessToken) {
        if (this._token) {
            return Promise.resolve(this._token);
        }
        if (this._fetchTokenPromise) {
            return this._fetchTokenPromise;
        }
        this._fetchTokenPromise = (async () => {
            const headers = {
                'User-Agent': util.userAgent
            };
            if (postmanApiKey) {
                headers['x-api-key'] = postmanApiKey;
            }
            else {
                headers['x-access-token'] = accessToken;
            }
            try {
                const response = await platformRequestWrapper({
                    method: 'get',
                    url: `${util.POSTMAN_RUNTIME_AGENT_BASE_URL()}/handshake/token?source=packman`,
                    headers: headers,
                    json: true
                });
                if (response.statusCode !== 200) {
                    throw new Error('Failed to get auth token. Status: ' + response.status);
                }
                if (!response.body.token) {
                    throw new Error('Invalid response: Missing auth token');
                }
                this._token = response.body.token;
                const decoded = jwtDecode(this._token);
                this._userType = decoded.userType;
                return this._token;
            }
            finally {
                this._fetchTokenPromise = null;
            }
        })();
        return this._fetchTokenPromise;
    }
    _fetchPackageAccessDetails(token) {
        if (this._accessDetails) {
            return Promise.resolve(this._accessDetails);
        }
        if (this._fetchAccessDetailsPromise) {
            return this._fetchAccessDetailsPromise;
        }
        this._fetchAccessDetailsPromise = (async () => {
            try {
                const response = await platformRequestWrapper({
                    method: 'get',
                    url: `${util.POSTMAN_PACKMAN_BASE_URL()}/v1/package-access`,
                    headers: {
                        Authorization: `Bearer ${token}`
                    },
                    json: true
                });
                if (response.statusCode === 404) {
                    return null;
                }
                if (response.statusCode !== 200) {
                    throw new Error('Failed to fetch package access details. Status: ' + response.statusCode);
                }
                if (!response.body.access) {
                    throw new Error('Invalid response: Missing access');
                }
                this._accessDetails = response.body.access;
                return this._accessDetails;
            }
            finally {
                this._fetchAccessDetailsPromise = null;
            }
        })();
        return this._fetchAccessDetailsPromise;
    }
    _isPackageAllowed(packageId) {
        if (!this._accessDetails) {
            return true;
        }
        const { scope, packages } = this._accessDetails;
        if (!scope) {
            return true;
        }
        if (scope === AccessScope.ALL) {
            return true;
        }
        if (scope === AccessScope.NONE) {
            return false;
        }
        if (scope === AccessScope.RESTRICTED && Array.isArray(packages)) {
            const normalizedPackageId = ExternalPackageManager.isExternalPackageVersionMissing(packageId) ?
                `${packageId}@` :
                packageId;
            return packages.some((pattern) => { return minimatch(normalizedPackageId, pattern); });
        }
        return false;
    }
    static isExternalPackageVersionMissing(packageId) {
        const packageIdentifier = packageId.includes(':') ?
            packageId.split(':')[1] : packageId;
        const lastAtIndex = packageIdentifier.lastIndexOf('@');
        if (lastAtIndex === -1 || lastAtIndex === 0) {
            return true;
        }
        return false;
    }
    async fetchPackage(packageId, postmanApiKey, accessToken) {
        if (packageId.length > MAX_PACKAGE_REQUIRED_LENGTH) {
            return {
                id: packageId,
                name: packageId,
                error: { message: externalPackageErrorHandler.ErrorMessage.INVALID_PACKAGE_NAME }
            };
        }
        const token = await this._fetchAuthToken(postmanApiKey, accessToken);
        if (this._userType === 'enterprise') {
            await this._fetchPackageAccessDetails(token);
            const isAllowed = await this._isPackageAllowed(packageId);
            if (!isAllowed) {
                return {
                    id: packageId,
                    name: packageId,
                    error: { message: externalPackageErrorHandler.ErrorMessage.PACKAGE_ACCESS_DENIED }
                };
            }
        }
        const [packageRegistry, packageIdentifier] = packageId.split(':');
        const response = await platformRequestWrapper({
            url: `${util.POSTMAN_PACKMAN_BASE_URL()}/v1/bundle/${encodeURIComponent(packageRegistry)}/${encodeURIComponent(packageIdentifier)}`,
            method: 'get',
            headers: {
                Authorization: `Bearer ${token}`
            }
        });
        if (response.statusCode !== 200) {
            const error = JSON.parse(response.body);
            return {
                id: packageId,
                name: packageId,
                error: { message: externalPackageErrorHandler.getErrorMessage(error.code) }
            };
        }
        return {
            id: packageId,
            name: packageId,
            scriptId: packageId,
            sourceCode: response.body
        };
    }
}
module.exports = ExternalPackageManager;
