// @postman/pm-bin-linux-x64/bin/lib/api/externalPackageErrorHandler.js
const ErrorCode = {
    ENTRY_POINT_NOT_FOUND: 'ENTRY_POINT_NOT_FOUND',
    PACKAGE_JSON_NOT_FOUND: 'PACKAGE_JSON_NOT_FOUND',
    BUNDLING_FAILURE: 'BUNDLING_FAILURE',
    INSTALL_MEMORY_LIMIT_EXCEEDED: 'INSTALL_MEMORY_LIMIT_EXCEEDED',
    INSTALL_DISK_LIMIT_EXCEEDED: 'INSTALL_DISK_LIMIT_EXCEEDED',
    INSTALL_TIMEOUT: 'INSTALL_TIMEOUT',
    INSTALL_DISK_MONITOR_ERROR: 'INSTALL_DISK_MONITOR_ERROR',
    INSTALL_MEMORY_MONITOR_ERROR: 'INSTALL_MEMORY_MONITOR_ERROR',
    INSTALL_ERROR: 'INSTALL_ERROR',
    UNSUPPORTED_REGISTRY: 'UNSUPPORTED_REGISTRY',
    INVALID_PACKAGE_SCOPE: 'INVALID_PACKAGE_SCOPE',
    INVALID_PACKAGE_NAME: 'INVALID_PACKAGE_NAME',
    INVALID_PACKAGE_VERSION: 'INVALID_PACKAGE_VERSION',
    INVALID_FILENAME: 'INVALID_FILENAME',
    PACKAGE_DENIED: 'PACKAGE_DENIED',
    PACKAGE_NOT_FOUND: 'PACKAGE_NOT_FOUND',
    PACKAGE_VERSION_NOT_FOUND: 'PACKAGE_VERSION_NOT_FOUND',
    UNEXPECTED_ERROR: 'UNEXPECTED_ERROR',
    RATE_LIMIT_EXCEEDED: 'RATE_LIMIT_EXCEEDED',
    SANDBOX_UNSUPPORTED_PACKAGE: 'SANDBOX_UNSUPPORTED_PACKAGE',
    URI_TOO_LONG: 'URI_TOO_LONG',
    PACKAGE_ACCESS_DENIED: 'PACKAGE_ACCESS_DENIED'
}, ErrorMessage = {
    FAILED_LOADING: 'This package cannot be used in Postman.',
    INVALID_REGISTRY: 'The package registry is not supported.',
    INVALID_PACKAGE_NAME: 'Package name invalid.',
    INVALID_PACKAGE_SCOPE: 'Package scope invalid.',
    INVALID_PACKAGE_VERSION: 'Package version invalid.',
    INVALID_FILENAME: 'Package filename invalid.',
    PACKAGE_NOT_FOUND: 'Couldn\'t find the package.',
    RATE_LIMIT_EXCEEDED: 'Package usage limit reached. Contact Postman support.',
    UNEXPECTED_ERROR: 'Something went wrong. Try again.',
    INVALID_PACKAGE_DETAILS: 'Package invalid.',
    PACKAGE_ACCESS_DENIED: 'Package not allowed in this team. Contact your admin.'
};
function getErrorMessage(errorCode = ErrorCode.UNEXPECTED_ERROR) {
    const loadingErrors = [
        ErrorCode.ENTRY_POINT_NOT_FOUND,
        ErrorCode.PACKAGE_JSON_NOT_FOUND,
        ErrorCode.BUNDLING_FAILURE,
        ErrorCode.INSTALL_ERROR,
        ErrorCode.INSTALL_MEMORY_LIMIT_EXCEEDED,
        ErrorCode.INSTALL_DISK_LIMIT_EXCEEDED,
        ErrorCode.INSTALL_TIMEOUT,
        ErrorCode.INSTALL_DISK_MONITOR_ERROR,
        ErrorCode.INSTALL_MEMORY_MONITOR_ERROR,
        ErrorCode.SANDBOX_UNSUPPORTED_PACKAGE,
        ErrorCode.PACKAGE_DENIED
    ];
    if (loadingErrors.includes(errorCode)) {
        return ErrorMessage.FAILED_LOADING;
    }
    switch (errorCode) {
        case ErrorCode.UNSUPPORTED_REGISTRY:
            return ErrorMessage.INVALID_REGISTRY;
        case ErrorCode.PACKAGE_NOT_FOUND:
            return ErrorMessage.PACKAGE_NOT_FOUND;
        case ErrorCode.RATE_LIMIT_EXCEEDED:
            return ErrorMessage.RATE_LIMIT_EXCEEDED;
        case ErrorCode.INVALID_PACKAGE_SCOPE:
            return ErrorMessage.INVALID_PACKAGE_SCOPE;
        case ErrorCode.INVALID_PACKAGE_NAME:
        case ErrorCode.URI_TOO_LONG:
            return ErrorMessage.INVALID_PACKAGE_NAME;
        case ErrorCode.INVALID_PACKAGE_VERSION:
        case ErrorCode.PACKAGE_VERSION_NOT_FOUND:
            return ErrorMessage.INVALID_PACKAGE_VERSION;
        case ErrorCode.INVALID_FILENAME:
            return ErrorMessage.INVALID_FILENAME;
        case ErrorCode.UNEXPECTED_ERROR:
        default:
            return ErrorMessage.UNEXPECTED_ERROR;
    }
}
module.exports = {
    getErrorMessage,
    ErrorMessage
};
