// @postman/pm-bin-linux-x64/bin/lib/api/integrations-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationsService = exports.SPECIFICATION_INTEGRATION_RULE = exports.INTEGRATION_RULE = void 0;
exports.createIntegrationsService = createIntegrationsService;
const util_1 = __importDefault(require("../util"));
const request_1 = __importDefault(require("@postman/platform-req/lib/request"));
exports.INTEGRATION_RULE = 'collection_sdk_github';
exports.SPECIFICATION_INTEGRATION_RULE = 'spec_sdk_github';
class IntegrationsService {
    client;
    baseUrl;
    accessToken;
    postmanApiKey;
    constructor(options) {
        this.baseUrl = options.baseUrl;
        this.accessToken = options.accessToken;
        this.postmanApiKey = options.postmanApiKey;
        this.client = request_1.default.defaults({
            json: true,
            headers: this.getAuthHeaders(),
            timeout: 30000
        });
    }
    getAuthHeaders() {
        const headers = {
            'Content-Type': 'application/json',
            'x-pstmn-req-service': 'integrations'
        };
        if (this.accessToken) {
            headers['x-access-token'] = this.accessToken;
        }
        else if (this.postmanApiKey) {
            headers['X-Api-Key'] = this.postmanApiKey;
        }
        return headers;
    }
    makeGatewayRequest(method, path, body) {
        return new Promise((resolve, reject) => {
            const url = `${this.baseUrl}${path}`;
            const options = {
                method,
                url,
                headers: this.getAuthHeaders(),
                json: true
            };
            if (body && (method === 'POST' || method === 'PUT')) {
                options.body = body;
            }
            this.client(options, (err, response, responseBody) => {
                if (err) {
                    return reject(err);
                }
                if (response.statusCode >= 400) {
                    let errorMessage = 'Unknown error';
                    if (responseBody) {
                        if (typeof responseBody === 'string') {
                            errorMessage = responseBody;
                        }
                        else if (responseBody.error) {
                            if (typeof responseBody.error === 'string') {
                                errorMessage = responseBody.error;
                            }
                            else if (responseBody.error.message) {
                                errorMessage = responseBody.error.message;
                            }
                            else {
                                errorMessage = JSON.stringify(responseBody.error);
                            }
                        }
                        else if (responseBody.message) {
                            errorMessage = responseBody.message;
                        }
                        else {
                            errorMessage = JSON.stringify(responseBody);
                        }
                    }
                    else if (response.statusMessage) {
                        errorMessage = response.statusMessage;
                    }
                    return reject(new Error(`${errorMessage} (Status: ${response.statusCode})`));
                }
                return resolve(responseBody);
            });
        });
    }
    static extractRateLimitInfo(response) {
        const headers = response.headers || {}, limit = parseInt(headers['x-ratelimit-limit'] || '60', 10), remaining = parseInt(headers['x-ratelimit-remaining'] || '60', 10), reset = parseInt(headers['x-ratelimit-reset'] || '0', 10);
        return { limit, remaining, reset };
    }
    createOAuthSession(service, tokenType) {
        return new Promise((resolve, reject) => {
            const path = `/oauth/session?service=${service}&tokenType=${tokenType}`;
            this.makeGatewayRequest('POST', path)
                .then((body) => {
                const channel = process.env.POSTMAN_CHANNEL;
                const isProduction = !channel || channel === 'prod';
                if (isProduction && body.redirectUrl) {
                    body.redirectUrl = body.redirectUrl.replace(/postman-beta\.com/g, 'postman.com');
                }
                return resolve(body);
            })
                .catch((err) => {
                let errorMessage = 'Unknown error';
                if (err instanceof Error) {
                    errorMessage = err.message;
                }
                else if (typeof err === 'string') {
                    errorMessage = err;
                }
                else if (err && typeof err === 'object') {
                    errorMessage = err.message || JSON.stringify(err);
                }
                const debugInfo = process.env.VERBOSE ?
                    `\n  URL: ${this.baseUrl}\n  Path: ${path}\n  Error: ${errorMessage}` :
                    '';
                return reject(new Error(`Failed to create OAuth session: ${errorMessage}${debugInfo}`));
            });
        });
    }
    getOAuthSession(sessionId) {
        return new Promise((resolve, reject) => {
            const path = `/oauth/session/${sessionId}`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                return resolve(body);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get OAuth session: ${err.message || err}`));
            });
        });
    }
    getGitUser(tokenId) {
        return new Promise((resolve, reject) => {
            const path = `/git/user?tokenId=${encodeURIComponent(tokenId)}`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                return resolve(body);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get Git user: ${err.message || err}`));
            });
        });
    }
    getGitOrgs(tokenId) {
        return new Promise((resolve, reject) => {
            const path = `/git/orgs?tokenId=${encodeURIComponent(tokenId)}`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                return resolve(body.orgs || []);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get Git orgs: ${err.message || err}`));
            });
        });
    }
    getGitRepositories(tokenId, owner) {
        return new Promise((resolve, reject) => {
            const path = `/git/repositories?tokenId=${encodeURIComponent(tokenId)}&owner=${encodeURIComponent(owner)}`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                return resolve(body.repositories || []);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get repositories: ${err.message || err}`));
            });
        });
    }
    getGitBranches(tokenId, owner, repository) {
        return new Promise((resolve, reject) => {
            const path = `/git/repositories/${encodeURIComponent(repository)}/branches?tokenId=${encodeURIComponent(tokenId)}&owner=${encodeURIComponent(owner)}`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                return resolve(body.branches || []);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get branches: ${err.message || err}`));
            });
        });
    }
    createIntegration(workspaceId, config) {
        return new Promise((resolve, reject) => {
            const path = `/integration?workspace=${encodeURIComponent(workspaceId)}`;
            this.makeGatewayRequest('POST', path, config)
                .then((body) => {
                return resolve(body.data || body);
            })
                .catch((err) => {
                return reject(new Error(`Failed to create integration: ${err.message || err}`));
            });
        });
    }
    getIntegrations(id, language, entityType = 'collection') {
        return new Promise((resolve, reject) => {
            const rule = entityType === 'collection' ? exports.INTEGRATION_RULE : exports.SPECIFICATION_INTEGRATION_RULE;
            const path = `/v2/integration?rule=${rule}&entityType=${entityType}&entityId=${encodeURIComponent(id)}&serviceId=github`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                let integrations = body.data || [];
                if (language) {
                    integrations = integrations.filter((integration) => {
                        return integration.variables?.language === language;
                    });
                }
                return resolve(integrations);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get integrations: ${err.message || err}`));
            });
        });
    }
    getIntegrationById(integrationId) {
        return new Promise((resolve, reject) => {
            const path = `/integration/${encodeURIComponent(integrationId)}?populateStatus=1`;
            this.makeGatewayRequest('GET', path)
                .then((body) => {
                const integration = body.data || body;
                if (!integration || !integration.id) {
                    return reject(new Error('Invalid integration response: integration data not found'));
                }
                return resolve(integration);
            })
                .catch((err) => {
                return reject(new Error(`Failed to get integration: ${err.message || err}`));
            });
        });
    }
    deleteIntegration(integrationId) {
        return new Promise((resolve, reject) => {
            const path = `/integration/${encodeURIComponent(integrationId)}`;
            this.makeGatewayRequest('DELETE', path)
                .then(() => {
                return resolve();
            })
                .catch((err) => {
                return reject(new Error(`Failed to delete integration: ${err.message || err}`));
            });
        });
    }
    triggerSDKPush(integrationId, buildId) {
        return new Promise((resolve, reject) => {
            const path = '/sdk/trigger-push';
            this.makeGatewayRequest('POST', path, { integrationId, buildId })
                .then((body) => {
                return resolve(body.prUrl);
            })
                .catch((err) => {
                return reject(new Error(`Failed to trigger SDK push: ${err.message || err}`));
            });
        });
    }
}
exports.IntegrationsService = IntegrationsService;
function createIntegrationsService(options) {
    const baseUrl = util_1.default.POSTMAN_GATEWAY_BASE_URL();
    return new IntegrationsService({
        baseUrl,
        accessToken: options.accessToken,
        postmanApiKey: options.postmanApiKey
    });
}
