// @postman/pm-bin-linux-x64/bin/lib/api/packages.js
const util = require('../util'), requestFactory = require('@postman/platform-req/lib/request'), _ = require('lodash'), USER_FACING_ERROR_MESSAGES = {
    'urn:error:package:planNotSupportedError': 'Packages are not supported in Postman CLI for your plan. ' +
        'To use packages, upgrade your plan by visiting https://www.postman.com/pricing/',
    'urn:error:package:deletedResourceError': 'The package you are trying to access has been deleted.',
    'urn:error:package:instanceNotFoundError': 'The package you are trying to access was not found.',
    'urn:error:package:forbiddenError': 'You need to be a member of this team to access this package'
}, ExternalPackageManager = require('./externalPackages'), externalPackage = new ExternalPackageManager();
function _createFriendlyError(status, errorData = {}) {
    const err = new Error(USER_FACING_ERROR_MESSAGES[errorData.type] ||
        errorData.detail || errorData.title || 'Something went wrong while fetching the package.');
    err.status = status;
    err.type = errorData.type;
    err.name = 'Request Error';
    return err;
}
function _handleRequestError(host, path, error) {
    const loggedError = {
        host: host,
        path: path,
        message: `Unable to send request server: ${error.message}`,
        code: error.code,
        status: _.get(error, 'response.statusCode')
    };
    if (error.response) {
        loggedError.message = `Invalid response from server: ${error.message}`;
    }
    logger.error(JSON.stringify(loggedError));
    throw _createFriendlyError(_.get(error, 'response.statusCode', 500), _.get(error, 'response.body'));
}
function _getRequestHeaders({ postmanApiKey, accessToken }) {
    const requestHeaders = {
        accept: 'application/vnd.postman.v2+json',
        'User-Agent': util.userAgent,
        'x-pstmn-req-service': 'package-library'
    };
    if (postmanApiKey) {
        requestHeaders['x-api-key'] = postmanApiKey;
    }
    else {
        requestHeaders['x-access-token'] = accessToken;
    }
    return requestHeaders;
}
function _getClient() {
    const client = requestFactory.defaults({
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: {
                    enabled: true
                },
                DELETE: {
                    enabled: false
                },
                POST: {
                    enabled: false
                },
                PUT: {
                    enabled: false
                },
                PATCH: {
                    enabled: false
                }
            }
        }
    });
    return client;
}
function _platformRequestWrapper({ url, headers, method }) {
    const client = _getClient();
    return new Promise((resolve, reject) => {
        client[method]({
            url: url,
            headers: headers,
            json: true
        }, (err, response) => {
            if (err) {
                reject(err);
                return;
            }
            if (response && !(/2../).test(response.statusCode)) {
                const error = new Error();
                error.response = response;
                error.code = response.statusMessage;
                error.message = _.get(response, 'body.message') || _.get(response, 'body.error.message');
                reject(error);
                return;
            }
            else if (!(response && response.body && response.body.data)) {
                const error = new Error('packages.assertValidResponseBody: Invalid response.');
                reject(error);
                return;
            }
            resolve(response);
        });
    });
}
async function _request({ uri, headers, method }) {
    const url = util.POSTMAN_GATEWAY_BASE_URL() + uri, response = await _platformRequestWrapper({ url, headers, method }).catch((err) => {
        _handleRequestError(util.POSTMAN_GATEWAY_BASE_URL(), uri, err);
    });
    return response.body.data;
}
async function getPackageMeta({ packageId, postmanApiKey, accessToken }) {
    if (!packageId) {
        throw new Error('Package ID is missing.');
    }
    if (!postmanApiKey && !accessToken) {
        throw new Error('Postman credentials are missing.');
    }
    const response = await _request({
        method: 'get',
        uri: `/packages/${packageId}`,
        headers: _getRequestHeaders({ postmanApiKey, accessToken })
    });
    if (response.deletedAt) {
        throw _createFriendlyError(404, { type: 'urn:error:package:deletedResourceError' });
    }
    return response;
}
async function getScriptContents({ packageId, scriptId, postmanApiKey, accessToken }) {
    if (!packageId || !scriptId) {
        throw new Error('Missing Package or Script ID.');
    }
    if (!postmanApiKey && !accessToken) {
        throw new Error('Postman credentials are missing');
    }
    const response = await _request({
        method: 'get',
        uri: `/packages/${packageId}/scripts/${scriptId}?include=contents`,
        headers: _getRequestHeaders({ postmanApiKey, accessToken })
    });
    return response;
}
async function loadPackageContents({ package, postmanApiKey, accessToken }) {
    try {
        if (ExternalPackageManager.isExternalPackage(package.id)) {
            return externalPackage.fetchPackage(package.id, postmanApiKey, accessToken);
        }
        const packageMeta = await getPackageMeta({ packageId: package.id, postmanApiKey: postmanApiKey,
            accessToken: accessToken }), scriptContents = await getScriptContents({
            packageId: package.id,
            scriptId: packageMeta.indexScript.id,
            postmanApiKey: postmanApiKey,
            accessToken: accessToken
        });
        return {
            id: package.id,
            name: package.name,
            scriptId: packageMeta.indexScript.id,
            sourceCode: scriptContents.contents.src
        };
    }
    catch (error) {
        return {
            id: package.id,
            name: package.name,
            error: error
        };
    }
}
function loadPackageContentsInBulk({ packages, postmanApiKey, accessToken }) {
    return Promise.all(packages.map((item) => {
        return loadPackageContents({ package: item, postmanApiKey: postmanApiKey, accessToken: accessToken });
    }));
}
module.exports = {
    loadPackageContentsInBulk
};
