// @postman/pm-bin-linux-x64/bin/lib/api/postman-resource.js
require('../logger').init();
const console = require('../framework/console');
const colors = require('colors/safe');
let _ = require('lodash'), util = require('../util'), requestFactory = require('@postman/platform-req/lib/request'), liquidJSON = require('liquid-json'), POSTMAN_API_PATH_MAP = {
    collection: 'collection',
    environment: 'environment',
    request: 'request',
    specification: 'specs'
}, DEFAULT_HEADERS = {
    'User-Agent': util.userAgent
}, REQUEST_METHOD_MAP = {
    fetch: 'get',
    sync: 'put',
    delete: 'delete',
    import: 'post',
    deepupdate: 'put',
    check: 'get',
    create: 'post',
    update: 'put'
}, OPERATION_VERB_MAP = {
    fetch: 'fetching',
    sync: 'syncing',
    delete: 'deleting',
    import: 'importing',
    deepupdate: 'updating',
    check: 'checking',
    create: 'creating',
    update: 'updating'
}, getClient = () => {
    const client = requestFactory.defaults({
        timeout: 300000,
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: {
                    enabled: true
                },
                DELETE: {
                    enabled: false
                },
                POST: {
                    enabled: false
                },
                PUT: {
                    enabled: false
                },
                PATCH: {
                    enabled: false
                }
            }
        }
    });
    return client;
}, remoteOperation = (operation, resource, callback) => {
    let { url, apikey, type, data, accessToken } = resource, headers = _.clone(DEFAULT_HEADERS), requestOptions = {
        url,
        headers
    }, requestMethod = REQUEST_METHOD_MAP[operation], client = getClient();
    if (apikey) {
        headers['X-Api-Key'] = apikey;
    }
    else if (accessToken) {
        headers['x-access-token'] = accessToken;
    }
    headers['x-pstmn-req-service'] = 'sync';
    if (operation === 'sync') {
        data = _.set({}, type, data);
    }
    if (['deepupdate', 'create', 'import', 'sync', 'update'].includes(operation)) {
        requestOptions.body = JSON.stringify(data);
        requestOptions.headers['Content-Type'] = 'application/json';
    }
    return client[requestMethod](requestOptions, (err, response, body) => {
        if (err) {
            logger.error(`Unable to ${operation} data from url "${url}": ${err.message || err}`);
            return callback(_.set(err, 'help', `unable to ${operation} data from url "${url}"`));
        }
        if (operation === 'check') {
            const exists = (/2../).test(response.statusCode);
            return callback(null, exists);
        }
        try {
            _.isString(body) && (body = liquidJSON.parse(body.trim()));
        }
        catch (e) {
            logger.error(`Url "${url}" did not provide valid JSON data`);
            return callback(_.set(e, 'help', `the url "${url}" did not provide valid JSON data`));
        }
        if (!(/2../).test(response.statusCode)) {
            console.log(colors.yellow(`Got ERROR code ${response.statusCode} with body :: ${JSON.stringify(body).substring(20)}...`));
            const error = new Error(`Failed ${OPERATION_VERB_MAP[operation]} a ${type} with ERROR code ${response.statusCode}.`);
            _.assign(error, { name: _.get(body, 'error.name', _.capitalize(type) + _.capitalize(operation) + 'Error') });
            return callback(error);
        }
        return callback(null, body.data || body);
    });
};
class PostmanResource {
    constructor(type, location, apikey, accessToken) {
        this.type = type;
        this.apikey = apikey;
        this.accessToken = accessToken;
        this.data = undefined;
        if ((/^https?:\/\/.*/).test(location)) {
            this.url = location;
        }
        else {
            this.url = `${util.POSTMAN_GATEWAY_BASE_URL()}/${POSTMAN_API_PATH_MAP[type]}/${location}`;
            if (type === 'collection') {
                this.url = `${this.url}?format=2.1.0&populate=true`;
            }
            if (type === 'request') {
                this.url = `${this.url}?populate=parent`;
            }
        }
    }
    get(callback) {
        if (this.data) {
            return callback(null, this.data);
        }
        return remoteOperation('fetch', this, (err, response) => {
            if (err) {
                return callback(err);
            }
            this.data = response;
            return callback(null, this.data);
        });
    }
    update(data, callback) {
        this.data = data;
        return remoteOperation('sync', this, callback);
    }
    updateRaw(data, callback) {
        this.data = data;
        return remoteOperation('update', this, callback);
    }
    delete(callback) {
        return remoteOperation('delete', this, callback);
    }
    deepUpdate(data, userId, resourceId, callback) {
        const originalUrl = this.url, gatewayUrl = util.POSTMAN_GATEWAY_BASE_URL();
        this.url = `${gatewayUrl}/collection/deepupdate/${resourceId}?format=2.1.0&user_id=${userId}`;
        this.data = data;
        return remoteOperation('deepupdate', this, (err, result) => {
            this.url = originalUrl;
            return callback(err, result);
        });
    }
    createInWorkspace(data, workspaceId, callback) {
        this.url = `${util.POSTMAN_GATEWAY_BASE_URL()}/${this.type}/import?workspace=${workspaceId}&format=2.1.0`;
        this.data = data;
        return remoteOperation('create', this, callback);
    }
    exists(callback) {
        return remoteOperation('check', this, callback);
    }
}
PostmanResource.getWorkspaceCollections = function (workspaceId, apikey, accessToken, callback) {
    const url = `${util.POSTMAN_GATEWAY_BASE_URL()}/workspaces/${workspaceId}?include=elements`, headers = _.clone(DEFAULT_HEADERS), client = getClient();
    if (apikey) {
        headers['X-Api-Key'] = apikey;
    }
    else if (accessToken) {
        headers['x-access-token'] = accessToken;
    }
    else {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    headers['x-pstmn-req-service'] = 'workspaces';
    headers['Content-Type'] = 'application/json';
    client.get({
        url,
        headers
    }, (err, response, body) => {
        if (err) {
            return callback(new Error(`Failed to get workspace: ${err.message || err}`));
        }
        try {
            _.isString(body) && (body = liquidJSON.parse(body.trim()));
        }
        catch (e) {
            return callback(new Error('Invalid JSON response from API'));
        }
        if (!(/2../).test(response.statusCode)) {
            const errorMessage = _.get(body, 'error.message') || body.message || 'Unknown error';
            return callback(new Error(`Failed to get workspace: ${errorMessage}`));
        }
        const collections = _.get(body, 'data.elements.collections', []);
        return callback(null, collections);
    });
};
PostmanResource.getWorkspaceEnvironments = function (workspaceId, apikey, accessToken, callback) {
    const url = `${util.POSTMAN_GATEWAY_BASE_URL()}/workspaces/${workspaceId}?include=elements`, headers = _.clone(DEFAULT_HEADERS), client = getClient();
    if (apikey) {
        headers['X-Api-Key'] = apikey;
    }
    else if (accessToken) {
        headers['x-access-token'] = accessToken;
    }
    else {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    headers['x-pstmn-req-service'] = 'workspaces';
    headers['Content-Type'] = 'application/json';
    client.get({
        url,
        headers
    }, (err, response, body) => {
        if (err) {
            return callback(new Error(`Failed to get workspace: ${err.message || err}`));
        }
        try {
            _.isString(body) && (body = liquidJSON.parse(body.trim()));
        }
        catch (e) {
            return callback(new Error('Invalid JSON response from API'));
        }
        if (!(/2../).test(response.statusCode)) {
            const errorMessage = _.get(body, 'error.message') || body.message || 'Unknown error';
            return callback(new Error(`Failed to get workspace: ${errorMessage}`));
        }
        const environments = _.get(body, 'data.elements.environments', []);
        return callback(null, environments);
    });
};
PostmanResource.getWorkspaceSpecifications = function (workspaceId, apikey, accessToken, callback) {
    const url = `${util.POSTMAN_GATEWAY_BASE_URL()}/workspaces/${workspaceId}?include=elements`, headers = _.clone(DEFAULT_HEADERS), client = getClient();
    if (apikey) {
        headers['X-Api-Key'] = apikey;
    }
    else if (accessToken) {
        headers['x-access-token'] = accessToken;
    }
    else {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    headers['x-pstmn-req-service'] = 'workspaces';
    headers['Content-Type'] = 'application/json';
    client.get({
        url,
        headers
    }, (err, response, body) => {
        if (err) {
            return callback(new Error(`Failed to get workspace: ${err.message || err}`));
        }
        try {
            _.isString(body) && (body = liquidJSON.parse(body.trim()));
        }
        catch (e) {
            return callback(new Error('Invalid JSON response from API'));
        }
        if (!(/2../).test(response.statusCode)) {
            const errorMessage = _.get(body, 'error.message') || body.message || 'Unknown error';
            return callback(new Error(`Failed to get workspace: ${errorMessage}`));
        }
        const specifications = _.get(body, 'data.elements.specifications', []);
        return callback(null, specifications);
    });
};
PostmanResource.ENVIRONMENT = 'environment';
PostmanResource.COLLECTION = 'collection';
PostmanResource.SPECIFICATION = 'specification';
module.exports = PostmanResource;
