// @postman/pm-bin-linux-x64/bin/lib/api/sdk-generator-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDKGeneratorService = exports.BuildStatus = void 0;
exports.createSDKGeneratorService = createSDKGeneratorService;
const util_1 = __importDefault(require("../util"));
const request_1 = __importDefault(require("@postman/platform-req/lib/request"));
const form_data_1 = __importDefault(require("form-data"));
const fs_1 = __importDefault(require("fs"));
var BuildStatus;
(function (BuildStatus) {
    BuildStatus["RECEIVED"] = "RECEIVED";
    BuildStatus["IN_PROGRESS"] = "IN_PROGRESS";
    BuildStatus["SUCCESS"] = "SUCCESS";
    BuildStatus["FAILURE"] = "FAILURE";
})(BuildStatus || (exports.BuildStatus = BuildStatus = {}));
class SDKGeneratorService {
    baseUrl;
    accessToken;
    postmanApiKey;
    client;
    constructor(authOptions) {
        this.baseUrl = util_1.default.POSTMAN_GATEWAY_BASE_URL();
        this.accessToken = authOptions.accessToken;
        this.postmanApiKey = authOptions.postmanApiKey;
        this.client = request_1.default.defaults({
            retry: {
                enabled: true,
                delay: 50,
                backOffNoise: 100,
                maxAttempts: 2,
                maxDelay: 1000,
                statusCodes: [500, 502, 503, 504],
                methods: {
                    GET: { enabled: true },
                    POST: { enabled: false },
                    PUT: { enabled: false },
                    DELETE: { enabled: false },
                    PATCH: { enabled: false }
                }
            }
        });
    }
    async createBuild(options) {
        const requestBody = {
            languages: options.languages
        };
        if (options.sdkVersion) {
            requestBody.sdkVersion = options.sdkVersion;
        }
        if (options.collectionId) {
            requestBody.collectionId = options.collectionId;
        }
        else if (options.specificationId) {
            requestBody.specificationId = options.specificationId;
        }
        if (options.sdkName) {
            requestBody.sdkName = options.sdkName;
        }
        if (options.sdkConfig) {
            requestBody.sdkConfig = options.sdkConfig;
        }
        const { data, rateLimitInfo } = await this.request('POST', '/v1/builds', requestBody);
        const build = {
            id: data.buildId,
            status: data.status,
            languages: data.languages,
            collectionId: options.collectionId,
            apiSpecId: options.specificationId,
            createdAt: new Date().toISOString()
        };
        return { build, rateLimitInfo };
    }
    async createOneOffBuild(options) {
        return new Promise((resolve, reject) => {
            const form = new form_data_1.default();
            if (!fs_1.default.existsSync(options.filePath)) {
                reject(new Error(`File not found: ${options.filePath}`));
                return;
            }
            const fileStream = fs_1.default.createReadStream(options.filePath);
            if (options.entityType === 'collection') {
                form.append('collection', fileStream);
            }
            else if (options.entityType === 'specification') {
                form.append('specification', fileStream);
            }
            else {
                reject(new Error(`Unsupported file type: ${options.entityType}`));
            }
            form.append('sdkName', options.sdkName);
            if (options.sdkVersion) {
                form.append('sdkVersion', options.sdkVersion);
            }
            options.languages.forEach((language) => {
                form.append('languages', language);
            });
            if (options.sdkConfig) {
                form.append('sdkConfig', JSON.stringify(options.sdkConfig));
            }
            const headers = this.getHeaders(false);
            Object.assign(headers, form.getHeaders());
            const url = `${this.baseUrl}/v1/builds/one-off`;
            this.client({
                method: 'POST',
                url,
                headers,
                body: form,
                json: false
            }, (err, response, body) => {
                if (err) {
                    reject(new Error(`Request failed: ${err.message}`));
                    return;
                }
                let responseBody;
                try {
                    responseBody = typeof body === 'string' ? JSON.parse(body) : body;
                }
                catch (parseErr) {
                    reject(new Error(`Failed to parse response: ${parseErr.message}`));
                    return;
                }
                const rateLimitInfo = this.parseRateLimitHeaders(response);
                if (response.statusCode >= 400) {
                    const apiError = responseBody;
                    if (apiError.error && typeof apiError.error === 'object' && apiError.error.code) {
                        let errorMessage = `${apiError.error.code}: ${apiError.error.message || 'Unknown error'}`;
                        if (apiError.error.details && apiError.error.details.length > 0) {
                            errorMessage += '\nDetails:\n' +
                                apiError.error.details.map((d) => { return `  - ${d.field ? `${d.field}: ` : ''}${d.message}`; }).join('\n');
                        }
                        if (apiError.error.requestId) {
                            errorMessage += `\nRequest ID: ${apiError.error.requestId}`;
                        }
                        const error = new Error(errorMessage);
                        error.code = apiError.error.code;
                        error.statusCode = response.statusCode;
                        error.rateLimitInfo = rateLimitInfo;
                        reject(error);
                        return;
                    }
                    if (apiError.message || apiError.error) {
                        let errorMessage = '';
                        if (typeof apiError.error === 'string') {
                            errorMessage = apiError.error;
                        }
                        else if (apiError.statusCode) {
                            errorMessage = `HTTP ${apiError.statusCode}`;
                        }
                        else {
                            errorMessage = `HTTP ${response.statusCode}`;
                        }
                        if (apiError.message) {
                            if (Array.isArray(apiError.message)) {
                                errorMessage += ': ' + apiError.message.join(', ');
                            }
                            else {
                                errorMessage += ': ' + apiError.message;
                            }
                        }
                        const error = new Error(errorMessage);
                        error.statusCode = apiError.statusCode || response.statusCode;
                        error.rateLimitInfo = rateLimitInfo;
                        reject(error);
                        return;
                    }
                    reject(new Error(`HTTP ${response.statusCode}: ${JSON.stringify(responseBody)}`));
                    return;
                }
                const build = {
                    id: responseBody.buildId,
                    status: responseBody.status,
                    languages: responseBody.languages,
                    createdAt: new Date().toISOString()
                };
                resolve({ build, rateLimitInfo });
            });
        });
    }
    async uploadSDK(zipFilePath, language, collectionId, specificationId) {
        return new Promise((resolve, reject) => {
            if (!collectionId && !specificationId) {
                reject(new Error('Either collectionId or specificationId must be provided'));
                return;
            }
            if (collectionId && specificationId) {
                reject(new Error('Cannot provide both collectionId and specificationId'));
                return;
            }
            if (!fs_1.default.existsSync(zipFilePath)) {
                reject(new Error(`SDK zip file not found: ${zipFilePath}`));
                return;
            }
            const form = new form_data_1.default();
            const fileStream = fs_1.default.createReadStream(zipFilePath);
            form.append('file', fileStream);
            form.append('language', language);
            if (collectionId) {
                form.append('collectionId', collectionId);
            }
            else if (specificationId) {
                form.append('specificationId', specificationId);
            }
            const headers = this.getHeaders(false);
            Object.assign(headers, form.getHeaders());
            const url = `${this.baseUrl}/v1/builds/upload`;
            this.client({
                method: 'POST',
                url,
                headers,
                body: form,
                json: false
            }, (err, response, body) => {
                if (err) {
                    reject(new Error(`SDK upload failed: ${err.message}`));
                    return;
                }
                let responseBody;
                try {
                    responseBody = typeof body === 'string' ? JSON.parse(body) : body;
                }
                catch (parseErr) {
                    reject(new Error(`Failed to parse response: ${parseErr.message}`));
                    return;
                }
                const rateLimitInfo = this.parseRateLimitHeaders(response);
                if (response.statusCode >= 400) {
                    const apiError = responseBody;
                    if (apiError.error && typeof apiError.error === 'object' && apiError.error.code) {
                        let errorMessage = `${apiError.error.code}: ${apiError.error.message || 'Unknown error'}`;
                        if (apiError.error.details && apiError.error.details.length > 0) {
                            errorMessage += '\nDetails:\n' +
                                apiError.error.details.map((d) => { return `  - ${d.field ? `${d.field}: ` : ''}${d.message}`; }).join('\n');
                        }
                        if (apiError.error.requestId) {
                            errorMessage += `\nRequest ID: ${apiError.error.requestId}`;
                        }
                        const error = new Error(errorMessage);
                        error.code = apiError.error.code;
                        error.statusCode = response.statusCode;
                        error.rateLimitInfo = rateLimitInfo;
                        reject(error);
                        return;
                    }
                    reject(new Error(`HTTP ${response.statusCode}: ${body}`));
                    return;
                }
                resolve({ data: responseBody, rateLimitInfo });
            });
        });
    }
    async getBuild(buildId) {
        const { data, rateLimitInfo } = await this.request('GET', `/v1/builds/${buildId}`);
        return { build: data, rateLimitInfo };
    }
    async getBuildSummary(collectionId, specificationId, buildId) {
        if (!collectionId && !specificationId && !buildId) {
            throw new Error('Either buildId, collectionId, or specificationId must be provided');
        }
        const queryParams = [];
        if (buildId) {
            queryParams.push(`buildId=${buildId}`);
        }
        if (collectionId) {
            queryParams.push(`collectionId=${collectionId}`);
        }
        if (specificationId) {
            queryParams.push(`specificationId=${specificationId}`);
        }
        const queryString = queryParams.join('&');
        try {
            const { data, rateLimitInfo } = await this.request('GET', `/v1/builds/summary?${queryString}`);
            return { buildSummary: data, rateLimitInfo };
        }
        catch (err) {
            if (err.statusCode === 404) {
                return { buildSummary: null, rateLimitInfo: err.rateLimitInfo || {} };
            }
            throw err;
        }
    }
    async getSDKDownloadUrl(id, language, idType = 'collection') {
        const queryParamMap = {
            collection: 'collectionId',
            specification: 'specificationId',
            build: 'buildId'
        };
        const queryParam = queryParamMap[idType];
        const { data } = await this.request('GET', `/v1/sdks/download?${queryParam}=${id}&language=${language.toUpperCase()}`);
        return data.downloadUrl;
    }
    async listBuilds(options) {
        const queryParams = [];
        if (options.status) {
            queryParams.push(`status=${options.status}`);
        }
        if (options.collectionId) {
            queryParams.push(`collectionId=${options.collectionId}`);
        }
        if (options.specificationId) {
            queryParams.push(`specificationId=${options.specificationId}`);
        }
        const queryString = `?${queryParams.join('&')}`, response = await this.request('GET', `/v1/builds${queryString}`);
        return {
            data: Array.isArray(response.data) ? response.data : [],
            rateLimitInfo: response.rateLimitInfo
        };
    }
    async pollBuildStatus(buildId, options = {}) {
        const interval = options.interval || 2000, timeout = options.timeout || 300000, startTime = Date.now();
        return new Promise((resolve, reject) => {
            const poll = async () => {
                try {
                    const { build } = await this.getBuild(buildId);
                    if (options.onProgress) {
                        options.onProgress(build);
                    }
                    const isBuildEffectivelyComplete = (b) => {
                        if (!b.artifacts || b.artifacts.length === 0) {
                            return false;
                        }
                        const allSettled = b.artifacts.every((artifact) => {
                            return artifact.status === 'SUCCESS' || artifact.status === 'FAIL';
                        });
                        if (allSettled) {
                            return true;
                        }
                        const settledCount = b.artifacts.filter((a) => {
                            return a.status === 'SUCCESS' || a.status === 'FAIL';
                        }).length;
                        const inProgressCount = b.artifacts.filter((a) => {
                            return a.status === 'IN_PROGRESS';
                        }).length;
                        if (settledCount > 0 && inProgressCount > 0) {
                            const elapsedTime = Date.now() - startTime;
                            if (settledCount / b.artifacts.length >= 0.9 && elapsedTime > 60000) {
                                return true;
                            }
                        }
                        return false;
                    };
                    const isComplete = build.status === BuildStatus.SUCCESS ||
                        build.status === BuildStatus.FAILURE ||
                        isBuildEffectivelyComplete(build);
                    if (isComplete) {
                        return resolve(build);
                    }
                    if (Date.now() - startTime > timeout) {
                        return reject(new Error('Build polling timeout exceeded'));
                    }
                    setTimeout(poll, interval);
                }
                catch (err) {
                    reject(err);
                }
            };
            poll();
        });
    }
    getHeaders(includeContentType = true) {
        const headers = {
            'User-Agent': util_1.default.userAgent,
            'x-pstmn-req-service': 'sdk-gen-service'
        };
        if (includeContentType) {
            headers['Content-Type'] = 'application/json';
        }
        if (this.accessToken) {
            headers['x-access-token'] = this.accessToken;
        }
        else if (this.postmanApiKey) {
            headers['X-Api-Key'] = this.postmanApiKey;
        }
        return headers;
    }
    parseRateLimitHeaders(response) {
        const headers = response.headers || {};
        return {
            limit: headers['x-ratelimit-limit'] ? parseInt(headers['x-ratelimit-limit'], 10) : undefined,
            remaining: headers['x-ratelimit-remaining'] ? parseInt(headers['x-ratelimit-remaining'], 10) : undefined,
            reset: headers['x-ratelimit-reset'] ? parseInt(headers['x-ratelimit-reset'], 10) : undefined
        };
    }
    async request(method, endpoint, body) {
        return new Promise((resolve, reject) => {
            const url = `${this.baseUrl}${endpoint}`, options = {
                method,
                url,
                headers: this.getHeaders(),
                json: true
            };
            if (body && (method === 'POST' || method === 'PUT')) {
                options.body = body;
            }
            this.client(options, (err, response, responseBody) => {
                if (err) {
                    return reject(new Error(`Request failed: ${err.message}`));
                }
                const rateLimitInfo = this.parseRateLimitHeaders(response);
                if (response.statusCode >= 400) {
                    const apiError = responseBody;
                    if (apiError.error && typeof apiError.error === 'object' && apiError.error.code) {
                        let errorMessage = `${apiError.error.code}: ${apiError.error.message || 'Unknown error'}`;
                        if (apiError.error.details && apiError.error.details.length > 0) {
                            errorMessage += '\nDetails:\n' +
                                apiError.error.details.map((d) => { return `  - ${d.field ? `${d.field}: ` : ''}${d.message}`; }).join('\n');
                        }
                        if (apiError.error.requestId) {
                            errorMessage += `\nRequest ID: ${apiError.error.requestId}`;
                        }
                        const error = new Error(errorMessage);
                        error.code = apiError.error.code;
                        error.statusCode = response.statusCode;
                        error.rateLimitInfo = rateLimitInfo;
                        return reject(error);
                    }
                    if (apiError.message || apiError.error) {
                        let errorMessage = '';
                        if (typeof apiError.error === 'string') {
                            errorMessage = apiError.error;
                        }
                        else if (apiError.statusCode) {
                            errorMessage = `HTTP ${apiError.statusCode}`;
                        }
                        else {
                            errorMessage = `HTTP ${response.statusCode}`;
                        }
                        if (apiError.message) {
                            if (Array.isArray(apiError.message)) {
                                errorMessage += ': ' + apiError.message.join(', ');
                            }
                            else {
                                errorMessage += ': ' + apiError.message;
                            }
                        }
                        const error = new Error(errorMessage);
                        error.statusCode = apiError.statusCode || response.statusCode;
                        error.rateLimitInfo = rateLimitInfo;
                        return reject(error);
                    }
                    return reject(new Error(`HTTP ${response.statusCode}: ${JSON.stringify(responseBody)}`));
                }
                resolve({ data: responseBody, rateLimitInfo });
            });
        });
    }
}
exports.SDKGeneratorService = SDKGeneratorService;
function createSDKGeneratorService(authOptions) {
    if (!authOptions.accessToken && !authOptions.postmanApiKey) {
        throw new Error('Authentication required. Please provide accessToken or postmanApiKey.');
    }
    return new SDKGeneratorService(authOptions);
}
exports.default = SDKGeneratorService;
