// @postman/pm-bin-linux-x64/bin/lib/run/unified/proto/api-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createApiClient = createApiClient;
exports.parseJsonBody = parseJsonBody;
exports.assertHttpSuccess = assertHttpSuccess;
exports.getGatewayBaseUrl = getGatewayBaseUrl;
const requestFactory = require('@postman/platform-req/lib/request');
const util = require('../../../util');
const API_TIMEOUT_MS = 60000;
const API_RETRY_CONFIG = {
    enabled: true,
    delay: 50,
    backOffNoise: 100,
    maxAttempts: 2,
    maxDelay: 1000,
    statusCodes: [500, 502, 503, 504],
    methods: { GET: { enabled: true } }
};
const HTTP_SUCCESS_REGEX = /2../;
function createApiClient(auth, serviceHeaders) {
    const headers = { 'User-Agent': util.userAgent, ...serviceHeaders };
    if (auth.postmanApiKey) {
        headers['x-api-key'] = auth.postmanApiKey;
    }
    else if (auth.accessToken) {
        headers['x-access-token'] = auth.accessToken;
    }
    return {
        client: requestFactory.defaults({ timeout: API_TIMEOUT_MS, retry: API_RETRY_CONFIG }),
        headers
    };
}
function parseJsonBody(body) {
    if (typeof body === 'string') {
        return JSON.parse(body);
    }
    return body;
}
function assertHttpSuccess(response, errorPrefix) {
    if (!HTTP_SUCCESS_REGEX.test(String(response.statusCode))) {
        throw new Error(errorPrefix + ' (HTTP ' + response.statusCode + ')');
    }
}
function getGatewayBaseUrl() {
    return util.POSTMAN_GATEWAY_BASE_URL();
}
