// @postman/pm-bin-linux-x64/bin/lib/run/unified/proto/loaders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadFromFile = loadFromFile;
exports.loadFromSpecification = loadFromSpecification;
exports.loadFromApi = loadFromApi;
const api_client_1 = require("./api-client");
const Protobuf = require('@postman/protobufjs');
const { FileDescriptorSet } = require('@postman/protobufjs/ext/descriptor');
const fs = require('fs');
const https = require('https');
const http = require('http');
const PROTO_PARSE_OPTIONS = {
    keepCase: true,
    alternateCommentMode: true
};
function encodeToBase64(root) {
    root.resolveAll();
    return Buffer.from(FileDescriptorSet.encode(root.toDescriptor()).finish()).toString('base64');
}
function parseProtoContent(content) {
    const root = new Protobuf.Root();
    Protobuf.parse(content, root, PROTO_PARSE_OPTIONS);
    return encodeToBase64(root);
}
function isUrl(path) {
    return (/^https?:\/\//i).test(path);
}
function prepareUrl(url) {
    return url.replace(/^(https?:\/\/github\.com\/[^/]+\/[^/]+\/)blob\/(.*)/i, '$1raw/$2');
}
function fetchUrl(url) {
    return new Promise((resolve, reject) => {
        const transport = url.startsWith('https') ? https : http;
        const request = transport.get(url, (res) => {
            if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                return fetchUrl(res.headers.location).then(resolve, reject);
            }
            if (res.statusCode < 200 || res.statusCode >= 300) {
                return reject(new Error('Failed to fetch proto from URL (HTTP ' + res.statusCode + '): ' + url));
            }
            const chunks = [];
            res.on('data', (chunk) => { chunks.push(chunk); });
            res.on('end', () => { resolve(Buffer.concat(chunks).toString('utf8')); });
            res.on('error', reject);
        });
        request.on('error', reject);
        request.setTimeout(30000, () => {
            request.destroy(new Error('Timed out fetching proto from URL: ' + url));
        });
    });
}
function loadProtoFromUrl(rootUrl) {
    const root = new Protobuf.Root();
    const originalResolvePath = root.resolvePath;
    root.resolvePath = (origin, target) => {
        if (isUrl(target)) {
            return prepareUrl(target);
        }
        if (!isUrl(origin)) {
            return originalResolvePath(origin, target);
        }
        const parsed = new URL(origin);
        parsed.pathname = originalResolvePath(parsed.pathname, target);
        return parsed.toString();
    };
    root.fetch = (path, callback) => {
        if (isUrl(path)) {
            fetchUrl(path).then((content) => { callback(null, content); }, (err) => { callback(err); });
        }
        else {
            fs.readFile(path, 'utf8', callback);
        }
    };
    return root.load(prepareUrl(rootUrl), PROTO_PARSE_OPTIONS)
        .then(encodeToBase64);
}
function loadFromFile(protoPath) {
    if (isUrl(protoPath)) {
        return loadProtoFromUrl(protoPath);
    }
    return new Promise((resolve, reject) => {
        if (!fs.existsSync(protoPath)) {
            return reject(new Error('Proto file not found: ' + protoPath));
        }
        Protobuf.load(protoPath, (err, root) => {
            if (err) {
                return reject(err);
            }
            try {
                resolve(encodeToBase64(root));
            }
            catch (e) {
                reject(e);
            }
        });
    });
}
function loadFromSpecification(specificationId, auth) {
    return new Promise((resolve, reject) => {
        const gatewayBaseUrl = (0, api_client_1.getGatewayBaseUrl)();
        const url = gatewayBaseUrl + '/specifications/' + specificationId +
            '/files?fields=id,name,content,type,path';
        const { client, headers } = (0, api_client_1.createApiClient)(auth, {
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api'
        });
        client.get({ url, headers }, (err, response, body) => {
            if (err) {
                return reject(new Error('Failed to fetch specification files: ' + (err.message || err)));
            }
            let parsed;
            try {
                parsed = (0, api_client_1.parseJsonBody)(body);
            }
            catch {
                return reject(new Error('Specification files response did not contain valid JSON'));
            }
            try {
                (0, api_client_1.assertHttpSuccess)(response, 'Failed to fetch specification files');
            }
            catch (e) {
                return reject(e);
            }
            const files = ((parsed && parsed.data) || []);
            if (!files.length) {
                return reject(new Error('No files found for specification ID: ' + specificationId));
            }
            try {
                const root = new Protobuf.Root();
                for (const file of files) {
                    if (file.content) {
                        Protobuf.parse(file.content, root, { keepCase: true });
                    }
                }
                resolve(encodeToBase64(root));
            }
            catch (parseErr) {
                const msg = parseErr instanceof Error ? parseErr.message : String(parseErr);
                reject(new Error('Failed to parse proto from specification: ' + msg));
            }
        });
    });
}
function loadFromApi(apiId, versionId, auth) {
    return new Promise((resolve, reject) => {
        const gatewayBaseUrl = (0, api_client_1.getGatewayBaseUrl)();
        const apiUrl = gatewayBaseUrl + '/apis/' + apiId +
            '?populate=schemas&populate=versions&populate=relations';
        const { client, headers } = (0, api_client_1.createApiClient)(auth, {
            'x-pstmn-req-service': 'api-dev-platform',
            'x-postman-service': 'cloud-api'
        });
        client.get({ url: apiUrl, headers }, (err, response, body) => {
            if (err) {
                return reject(new Error('Failed to fetch API: ' + (err.message || err)));
            }
            let parsed;
            try {
                parsed = (0, api_client_1.parseJsonBody)(body);
            }
            catch {
                return reject(new Error('API response did not contain valid JSON'));
            }
            try {
                (0, api_client_1.assertHttpSuccess)(response, 'Failed to fetch API');
            }
            catch (e) {
                return reject(e);
            }
            const api = ((parsed && parsed.data) || parsed);
            const versions = ((api && api.versions) || []);
            const version = versions.find((v) => { return v.id === versionId; });
            if (!version) {
                return reject(new Error('Version ' + versionId + ' not found in API ' + apiId));
            }
            const schema = (version.schemas && version.schemas[0]) || null;
            if (schema && (schema.content || schema.schema)) {
                try {
                    return resolve(parseProtoContent(schema.content || schema.schema));
                }
                catch (parseErr) {
                    const msg = parseErr instanceof Error ? parseErr.message : String(parseErr);
                    return reject(new Error('Failed to parse proto from API: ' + msg));
                }
            }
            const relations = version.relations || [];
            const apiDefinition = relations.find((r) => { return r.type === 'apiDefinition'; });
            if (apiDefinition) {
                const schemaUrl = gatewayBaseUrl + '/apis/' + apiId + '/versions/' +
                    versionId + '/schemas/' + apiDefinition.id;
                return client.get({ url: schemaUrl, headers }, (schemaErr, schemaResponse, schemaBody) => {
                    if (schemaErr) {
                        return reject(new Error('Failed to fetch API schema: ' + (schemaErr.message || schemaErr)));
                    }
                    let schemaParsed;
                    try {
                        schemaParsed = (0, api_client_1.parseJsonBody)(schemaBody);
                    }
                    catch {
                        return reject(new Error('API schema response did not contain valid JSON'));
                    }
                    try {
                        (0, api_client_1.assertHttpSuccess)(schemaResponse, 'Failed to fetch API schema');
                    }
                    catch (e) {
                        return reject(e);
                    }
                    const fetched = ((schemaParsed && (schemaParsed.data || schemaParsed.schema)) ||
                        schemaParsed);
                    const content = fetched && (fetched.content || fetched.schema);
                    if (!content) {
                        return reject(new Error('No proto content found in API schema'));
                    }
                    try {
                        resolve(parseProtoContent(content));
                    }
                    catch (parseErr) {
                        const msg = parseErr instanceof Error ? parseErr.message : String(parseErr);
                        reject(new Error('Failed to parse proto from API: ' + msg));
                    }
                });
            }
            return reject(new Error('No schema definition found for API ' + apiId +
                ' version ' + versionId +
                '. Git-linked API definitions are not yet supported in the CLI.'));
        });
    });
}
