// @postman/pm-bin-linux-x64/bin/lib/run/unified/proto/resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveCollectionSchemas = resolveCollectionSchemas;
const _ = require("lodash");
const loaders_1 = require("./loaders");
const print = require('../../../framework/print');
const colors = require('colors/safe');
const DEFAULT_GRPC_CONNECTION_TIMEOUT_MS = 30000;
function schemaCacheKey(schema, payload) {
    switch (schema.source) {
        case 'file':
            return schema.location ? 'file:' + schema.location : undefined;
        case 'auto':
            return payload?.url ? 'auto:' + String(payload.url) : undefined;
        case 'specification':
            return schema.specificationId ? 'spec:' + schema.specificationId : undefined;
        case 'api':
            return (schema.apiId && schema.versionId) ?
                'api:' + schema.apiId + ':' + schema.versionId : undefined;
        default:
            return undefined;
    }
}
function resolveGrpcProxyFromEnv(targetUrl) {
    const noGrpcProxy = process.env.no_grpc_proxy || process.env.NO_GRPC_PROXY || '';
    const noProxy = process.env.no_proxy || process.env.NO_PROXY || '';
    const proxyUrl = process.env.grpc_proxy || process.env.GRPC_PROXY ||
        process.env.https_proxy || process.env.HTTPS_PROXY ||
        process.env.http_proxy || process.env.HTTP_PROXY;
    if (!proxyUrl) {
        return undefined;
    }
    let hostname;
    try {
        hostname = new URL(targetUrl.replace(/^grpc:/, 'http:').replace(/^grpcs:/, 'https:')).hostname;
    }
    catch {
        return proxyUrl;
    }
    const exclusions = (noGrpcProxy + ',' + noProxy).split(',').map((s) => { return s.trim(); });
    for (const exclusion of exclusions) {
        if (exclusion === '*') {
            return undefined;
        }
        if (exclusion && hostname.endsWith(exclusion)) {
            return undefined;
        }
    }
    return proxyUrl;
}
function buildReflectionRequest(payload) {
    const metadata = (payload.metadata || []).map((m) => {
        return {
            key: String(m?.key ?? ''),
            value: String(m?.value ?? ''),
            description: String(m?.description ?? ''),
            disabled: Boolean(m?.disabled ?? false)
        };
    });
    const settings = {
        strictSSL: Boolean(_.get(payload, 'settings.strictSSL')),
        secureConnection: Boolean(_.get(payload, 'settings.secureConnection')),
        serverNameOverride: _.get(payload, 'settings.serverNameOverride') || '',
        connectionTimeout: _.get(payload, 'settings.connectionTimeout', DEFAULT_GRPC_CONNECTION_TIMEOUT_MS)
    };
    const proxy = resolveGrpcProxyFromEnv(payload.url);
    return {
        type: 'grpc-request',
        id: payload.id || ('reflection-' + Date.now()),
        title: 'Server Reflection Request',
        createdAt: new Date().toISOString(),
        payload: {
            url: payload.url || '',
            methodPath: '',
            methodDescriptor: '',
            message: { content: '' },
            metadata,
            settings: proxy ? { ...settings, proxy } : settings
        },
        extensions: {}
    };
}
function getSchemaSourceLabel(schema) {
    switch (schema.source) {
        case 'file': return schema.location || 'file';
        case 'auto': return 'server-reflection';
        case 'specification': return 'specification:' + schema.specificationId;
        case 'api': return 'api:' + schema.apiId;
        default: return schema.source;
    }
}
function warnSchemaResolutionFailed(source, message) {
    const symbols = require('../../../reporters/cli/cli-utils-symbols');
    print.lf(colors.yellow('  %s Could not resolve gRPC schema from "%s": %s'), symbols().warning, source, message);
}
const LOADERS = {
    file(schema) {
        return schema.location ? (0, loaders_1.loadFromFile)(schema.location) : Promise.resolve(undefined);
    },
    auto(schema, payload, _auth, reflectionMethod) {
        if (!payload?.url) {
            return Promise.resolve(undefined);
        }
        if (!reflectionMethod) {
            return Promise.resolve(undefined);
        }
        const reflectionRequest = buildReflectionRequest(payload);
        return reflectionMethod(reflectionRequest)
            .then((descriptor) => { return JSON.stringify(descriptor || { nested: {} }); });
    },
    specification(schema, _payload, auth) {
        if (!schema.specificationId || !auth) {
            return Promise.resolve(undefined);
        }
        return (0, loaders_1.loadFromSpecification)(schema.specificationId, auth);
    },
    api(schema, _payload, auth) {
        if (!schema.apiId || !schema.versionId || !auth) {
            return Promise.resolve(undefined);
        }
        return (0, loaders_1.loadFromApi)(schema.apiId, schema.versionId, auth);
    }
};
function resolveSchema(schema, payload, auth, reflectionMethod, cache) {
    const loader = LOADERS[schema.source];
    if (!loader) {
        return Promise.resolve(undefined);
    }
    const cacheKey = schemaCacheKey(schema, payload);
    if (cacheKey) {
        const cached = cache.get(cacheKey);
        if (cached) {
            return cached;
        }
    }
    const promise = loader(schema, payload, auth, reflectionMethod)
        .catch((e) => {
        const message = e instanceof Error ? e.message : String(e);
        warnSchemaResolutionFailed(getSchemaSourceLabel(schema), message);
        if (cacheKey) {
            cache.delete(cacheKey);
        }
        return undefined;
    });
    if (cacheKey) {
        cache.set(cacheKey, promise);
    }
    return promise;
}
async function resolveCollectionSchemas(node, auth, reflectionMethod, cache = new Map()) {
    if (node.type === 'grpc-request') {
        const schema = node.extensions?.schema;
        if (schema) {
            const descriptor = await resolveSchema(schema, node.payload || {}, auth, reflectionMethod, cache);
            if (descriptor && node.payload) {
                node.payload.methodDescriptor = descriptor;
            }
        }
    }
    if (node.children) {
        await Promise.all(node.children.map((child) => {
            return resolveCollectionSchemas(child, auth, reflectionMethod, cache);
        }));
    }
}
