// @postman/pm-bin-linux-x64/bin/lib/reporters/html/handlebarHelpers.js
module.exports = (handlebars) => {
    handlebars.registerHelper('gt', (a, b) => { return a > b; });
    handlebars.registerHelper('formatDate', function (timestamp) {
        const date = new Date(timestamp);
        return date.toLocaleString('en-US', {
            month: 'short',
            day: 'numeric',
            year: 'numeric',
            hour: '2-digit',
            minute: '2-digit',
            second: '2-digit',
            hour12: false
        }).replace(date.getDate() + ', ' + date.getFullYear() + ',', date.getDate() + ', ' + date.getFullYear());
    });
    handlebars.registerHelper('subtract', function (a, b) {
        return a - b;
    });
    handlebars.registerHelper('add', function (a, b) {
        return a + b;
    });
    handlebars.registerHelper('percentage', function (a, b) {
        if (b === 0 || a === 0) {
            return 0;
        }
        return Math.round((b / a) * 100);
    });
    handlebars.registerHelper('methodColor', function (method) {
        const colors = {
            GET: '#007F31',
            POST: '#007bff',
            PUT: '#fd7e14',
            DELETE: '#dc3545',
            PATCH: '#ffc107',
            OPTIONS: '#20c997',
            HEAD: '#6f42c1'
        };
        return method ? colors[method.toUpperCase()] || '#999999' : '#999999';
    });
    handlebars.registerHelper('json', function (context) {
        return JSON.stringify(context)
            .replace(/</g, '\\u003c')
            .replace(/>/g, '\\u003e')
            .replace(/&/g, '\\u0026')
            .replace(/\u2028/g, '\\u2028')
            .replace(/\u2029/g, '\\u2029')
            .replace(/\u0027/g, '\\u0027')
            .replace(/\u0060/g, '\\u0060');
    });
    handlebars.registerHelper('iterationRequestFolders', function (folders) {
        let result = '';
        const folderIcon = `
            <svg width="24px" height="24px" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                <g clip-path="url(#clip0_6171_7237)">
                    <path
                        d="M5 18V6C5 5.44772 5.44772 5 6 5H9.69722C9.89465 5 10.0877 5.05844
                        10.2519 5.16795L12.7481 6.83205C12.9123 6.94156 13.1054 7 13.3028 7H18C18.5523
                        7 19 7.44772 19 8V18C19 18.5523 18.5523 19 18 19H6C5.44772 19 5 18.5523 5 18Z"
                        stroke="#6B6B6B"
                        stroke-linejoin="round"
                    />
                    <path
                        d="M5 10H19"
                        stroke="#6B6B6B"
                        stroke-linecap="round"
                    />
                </g>
                <defs>
                    <clipPath id="clip0_6171_7237">
                        <rect width="16" height="16" fill="white" transform="translate(4 4)"/>
                    </clipPath>
                </defs>
            </svg>`;
        for (const folder of folders) {
            result += `
                <div style="display: flex;gap:2px;align-items: center;">
                    ${folderIcon}
                    <div class="tooltip-container text-overflow-tooltip flex">
                        <span class="folder-name tooltip-trigger">${folder}</span>
                        <div class="tooltip">${folder}</div>
                    </div>
                </div>
                <span class="folder-slash"> / </span>
            `;
        }
        return result;
    });
};
