// @postman/pm-bin-linux-x64/bin/lib/reporters/html/index.js
const _ = require('lodash'), handlebars = require('handlebars'), colors = require('colors/safe'), print = require('../../framework/print'), util = require('../../util'), reporterUtil = require('../util'), htmlReporterUtil = require('./util'), console = require('../../framework/console'), registerHandlebarHelpers = require('./handlebarHelpers'), { SUPPORTED_BODY_FORMATS, UNSUPPORTED_REQUEST_BODY_MSG, UNSUPPORTED_RESPONSE_BODY_MSG, CONTENT_TYPES } = require('./constants'), htmlStyles = require('./styles'), htmlTemplate = require('./template'), htmlScriptContent = require('./script-literal'), getExecutionTestStats = (tests) => {
    const stats = { total: 0, passed: 0, failed: 0, skipped: 0 };
    if (Array.isArray(tests) && tests.length) {
        stats.total = tests.length;
        tests.forEach((test) => {
            if (test.status === 'passed') {
                stats.passed++;
            }
            else if (test.status === 'failed') {
                stats.failed++;
            }
            else if (test.status === 'skipped') {
                stats.skipped++;
            }
        });
    }
    return stats;
}, getTransformedRequest = (request) => {
    const { url, method, body, headers } = request, transformedReq = {
        url: htmlReporterUtil.getRequestUrlString(url),
        method: method,
        body: body,
        header: headers
    };
    if (!body) {
        return transformedReq;
    }
    let { mode } = body;
    switch (mode) {
        case 'file':
            transformedReq.body = `File: ${_.get(body, 'file.src')}`;
            transformedReq.contentType = CONTENT_TYPES.TEXT;
            break;
        case 'raw':
            transformedReq.body = body.raw;
            transformedReq.contentType = body.options ? _.get(body, 'options.raw.language') : CONTENT_TYPES.JSON;
            break;
        case 'urlencoded':
        case 'formdata':
            transformedReq.body = htmlReporterUtil.normalizeFormData(_.get(body[mode], 'members'));
            transformedReq.contentType = CONTENT_TYPES.JSON;
            break;
        case 'graphql':
            transformedReq.body = htmlReporterUtil.normalizeGraphQlData(body.graphql);
            transformedReq.contentType = CONTENT_TYPES.JSON;
            break;
        default:
            transformedReq.contentType = htmlReporterUtil.getContentTypeByHeaders(request.headers);
            break;
    }
    if (!SUPPORTED_BODY_FORMATS.includes(transformedReq.contentType)) {
        transformedReq.body = UNSUPPORTED_REQUEST_BODY_MSG;
        transformedReq.contentType = CONTENT_TYPES.TEXT;
    }
    return transformedReq;
}, getTransformedResponse = (response) => {
    if (!response || !_.isFunction(response.toJSON)) {
        return null;
    }
    const responseJson = response.toJSON(), transformedResponse = {
        id: responseJson.id,
        status: responseJson.status,
        code: responseJson.code,
        header: responseJson.header
    }, contentType = htmlReporterUtil.getContentTypeByHeaders(responseJson.header);
    transformedResponse.contentType = contentType;
    if (responseJson.responseTime && Number.isFinite(responseJson.responseTime)) {
        transformedResponse.responseTime = util.prettyms(responseJson.responseTime);
    }
    if (responseJson.responseSize && Number.isFinite(responseJson.responseSize)) {
        transformedResponse.responseSize = util.filesize(Number(responseJson.responseSize));
    }
    if (contentType && SUPPORTED_BODY_FORMATS.includes(contentType)) {
        transformedResponse.body = Buffer.from(responseJson.stream).toString();
    }
    else {
        transformedResponse.body = UNSUPPORTED_RESPONSE_BODY_MSG;
        transformedResponse.contentType = CONTENT_TYPES.TEXT;
    }
    return transformedResponse;
}, applyCustomReportOptions = (aggregatedExecutions, reporterOptions) => {
    return aggregatedExecutions.map((exeGroup) => {
        const trimmedExecutions = exeGroup.executions.map((execution) => {
            if (reporterOptions.omitAllHeadersAndBody) {
                reporterOptions.omitRequestBodies = true;
                reporterOptions.omitResponseBodies = true;
                reporterOptions.omitHeaders = true;
            }
            if (reporterOptions.omitHeaders) {
                _.unset(execution, 'request.header');
                _.unset(execution, 'response.header');
            }
            if (reporterOptions.omitResponseBodies) {
                _.unset(execution, 'response.body');
            }
            if (reporterOptions.omitRequestBodies) {
                _.unset(execution, 'request.body');
            }
            return execution;
        });
        return {
            ...exeGroup,
            executions: trimmedExecutions
        };
    });
}, groupExecutions = (aggregatedExecutions) => {
    return aggregatedExecutions.reduce((grouped, { parent, ...rest }) => {
        const { iterationIndex, ...parentWithoutIteration } = parent;
        if (!grouped[iterationIndex]) {
            grouped[iterationIndex] = {
                iterationIndex: iterationIndex,
                items: []
            };
        }
        grouped[iterationIndex].items.push({ parent: parentWithoutIteration, ...rest });
        return grouped;
    }, {});
}, getAggregatedFailuresAndStats = (summary, reducedExecutionsMap) => {
    try {
        return htmlReporterUtil.aggregateFailures(summary, reducedExecutionsMap);
    }
    catch (e) {
        console.error('Error aggregating failures:', e.message);
        return { stats: {}, items: [] };
    }
}, HTMLReporter = function (cliEventSource, reporterOptions) {
    registerHandlebarHelpers(handlebars);
    const skippedTests = [];
    cliEventSource.on('assertion', function (err, test) {
        if (err) {
            return;
        }
        if (test.skipped) {
            const parent = test.item.parent(), parentPath = util.getFullPath(parent);
            skippedTests.push({
                iterationIndex: test.cursor.iteration,
                assertion: test.assertion,
                source: {
                    id: test.item.id,
                    name: test.item.name,
                    request: {
                        method: test.item.request && test.item.request.method
                    }
                },
                parent: {
                    id: parent.id,
                    name: parentPath.join(' / '),
                    folders: parentPath
                }
            });
        }
    });
    cliEventSource.on('beforeDone', function (err, data) {
        print.lf(colors.grey('Generating run report [html]...'));
        if (err) {
            print.lf(colors.red(`Could not generate report [html]: ${err.message}`));
            return;
        }
        const reducedExecutions = {}, traversedExecutions = {}, summary = data.summary, executions = _.get(data, 'summary.run.executions', []), iterationsWithErrors = new Set(), iterationWithTestFailures = new Set(), aggregatedExecutions = [];
        try {
            executions.forEach((exec) => {
                const executionId = exec.cursor.ref, iterationIndex = exec.cursor.iteration;
                if (!traversedExecutions[executionId]) {
                    traversedExecutions[executionId] = true;
                    const reducedExecution = _.pick(exec, ['item', 'request', 'response']);
                    reducedExecution.item = {
                        id: reducedExecution.item.id,
                        name: reducedExecution.item.name
                    };
                    reducedExecution.request = getTransformedRequest(reducedExecution.request);
                    reducedExecution.response = getTransformedResponse(reducedExecution.response);
                    reducedExecution.errors = reporterUtil.extractExecutionErrors(exec, summary.run.failures);
                    if (reducedExecution.errors && reducedExecution.errors.length > 0) {
                        iterationsWithErrors.add(iterationIndex);
                    }
                    reducedExecution.tests = reporterUtil.getTransformedAssertions(exec.assertions);
                    reducedExecution.testStats = getExecutionTestStats(reducedExecution.tests);
                    if (reducedExecution.testStats.failed > 0) {
                        iterationWithTestFailures.add(iterationIndex);
                    }
                    reducedExecution.parent = exec.item.parent() || (exec.itemLegacy && exec.itemLegacy.parent()) ||
                        { id: '', name: '' };
                    reducedExecutions[executionId] = reducedExecution;
                }
            });
            executions.forEach((exec) => {
                const executionId = exec.cursor.ref, iterationIndex = exec.cursor.iteration, parent = exec.item.parent() || (exec.itemLegacy && exec.itemLegacy.parent()) ||
                    { id: '', name: '' }, parentPath = util.getFullPath(parent), current = _.omit(reducedExecutions[executionId], ['parent']), last = _.last(aggregatedExecutions);
                if (last && last.parent.id === parent.id && last.parent.iterationIndex === iterationIndex) {
                    last.executions.push(current);
                }
                else {
                    aggregatedExecutions.push({
                        parent: {
                            id: parent.id,
                            name: parentPath.join(' / '),
                            folders: parentPath,
                            iterationIndex: iterationIndex
                        },
                        executions: [current]
                    });
                }
            });
            let aggregatedAndTrimmedExecutions = applyCustomReportOptions(aggregatedExecutions, reporterOptions), groupedExecutions = groupExecutions(aggregatedAndTrimmedExecutions), runStats = reporterUtil.getRunSummaryStats(summary), testStats = {
                total: runStats.tests.executed,
                passed: runStats.tests.executed - runStats.tests.failed - skippedTests.length,
                failed: runStats.tests.failed,
                skipped: skippedTests.length
            }, testFailuresAndErrors = getAggregatedFailuresAndStats(summary, reducedExecutions), skippedTestsWithStats = skippedTests.length > 0 ?
                htmlReporterUtil.getSkippedTestsWithStats(skippedTests) : [], finalData = {
                meta: {
                    collection: {
                        id: summary.collection.id,
                        name: summary.collection.name,
                        description: summary.collection.description ? summary.collection.description : null
                    },
                    environment: {
                        name: summary.environment ? summary.environment.name : null
                    },
                    started: summary.run.timings.started,
                    completed: summary.run.timings.completed,
                    duration: util.prettyms(summary.run.timings.completed - summary.run.timings.started)
                },
                iterations: groupedExecutions,
                failures: testFailuresAndErrors,
                skippedTests: skippedTestsWithStats,
                summary: {
                    ...runStats,
                    tests: testStats,
                    iterations: {
                        ...runStats.iterations,
                        iterationsWithErrors: Array.from(iterationsWithErrors),
                        iterationsWithTestFailures: Array.from(iterationWithTestFailures)
                    }
                },
                runError: summary.run.error,
                srcCode: {
                    styles: htmlStyles,
                    scripts: htmlScriptContent
                }
            }, compiler = handlebars.compile(htmlTemplate);
            print.lf(colors.grey('Saving run report [html]...'));
            this.exports.push({
                name: 'html',
                default: summary.collection.name ?
                    `${summary.collection.name.replaceAll(' ', '-')}.html` : 'postman-cli-run-report.html',
                path: reporterOptions.export,
                content: compiler(finalData)
            });
        }
        catch (err) {
            print.lf(colors.red(`Could not generate report [html]: ${err.message}`));
        }
    });
};
module.exports = HTMLReporter;
module.exports.getExecutionTestStats = getExecutionTestStats;
module.exports.groupExecutions = groupExecutions;
module.exports.applyCustomReportOptions = applyCustomReportOptions;
module.exports.getTransformedResponse = getTransformedResponse;
module.exports.getTransformedRequest = getTransformedRequest;
