// @postman/pm-bin-linux-x64/bin/lib/reporters/html/styles.js
const htmlStyles = `
:root {
    --primary-content-color: #212121;
    --secondary-content-color: #6B6B6B;
    --tertiary-content-color: #A6A6A6;

    --content-color-success: #007F31;
    --content-color-error: #8E1A10;
    --content-color-link: #0265D2;

    --border-color-default: #ededed;
    --border-secondary: #E6E6E6;

    --background-color-error: #ffebe7;
    --background-color-success: #E5FFF1;
    --background-color-secondary: #f9f9f9;
    --background-color-tertiary: #f2f2f2;

    --base-color-brand: #ff6c37;
    --base-color-error: #eb2013;
    --base-color-success: #0cbb52;

    --grey-10: #f2f2f2;
    --red-60: #b02016;
    --red-20: #fbc3ba;
    --red-30: #F79A8E;
    --green-30: #6bdd9a;

    --text-secondary: #6b6b6b;
    --text-muted: #929292;

    --font-weight-regular: 400;
    --font-weight-medium: 500;
    --font-weight-semibold: 600;

    --font-primary: Inter, serif;
    --font-secondary: IBM Plex Mono, monospace;

    --section-border: 1px solid var(--border-secondary);
    --border-radius: 8px;
}

html {
    height: 100vh;
    font-family: var(--font-primary);
    font-size: 16px;
}
* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
    font-family: var(--font-primary);
    color: var(--primary-content-color);
}

#collection-description p,
#collection-description ul,
#collection-description ol,
#collection-description h1,
#collection-description h2,
#collection-description h3,
#collection-description h4,
#collection-description h5,
#collection-description h6 {
    margin: 14px 0;
}

#collection-description ul,
#collection-description li,
#collection-description ol {
    margin-left: 14px;
}

body {
    margin: 0;
    padding: 0;
    height: 100%;
}

a {
    color: #0066cc;
    text-decoration: none;
}

a:hover {
    text-decoration: underline;
}

blockquote {
    border-left: 4px solid #ddd;
    margin: 0;
    padding-left: 12px;
    color: #666;
}

blockquote p {
    color: var(--secondary-content-color);
}

h3{
    font-family: var(--font-primary);
    font-size: 16px;
    font-weight: 600;
    line-height: 20px;
    letter-spacing: -0.11999999731779099px;
    text-align: left;
    text-underline-position: from-font;
    text-decoration-skip-ink: none;
}
/* Common styles*/

/*Divider*/
.divider {
    border-bottom: 1px solid var(--border-color-default);
    margin: 16px 0 16px 0;
}
/*Divider*/

/*--Generic Styled Containers--*/
.skip-styled-container {
    border: 1px solid var(--border-color-default);
    background-color: var(--background-color-secondary);
}

.skip-styled-container.clickable {
    cursor: pointer;
}

.skip-styled-container.clickable:hover {
    border: 1px solid var(--tertiary-content-color);
    background: var(--background-color-tertiary);
}

.pass-styled-container {
    border: 1px solid var(--green-30) !important;
    background-color: #F8FFFB;
}

.fail-styled-container {
    border: 1px solid var(--red-20) !important;
    background: #FFF7F5;
}

.fail-styled-container.clickable {
    cursor: pointer;
}

.fail-styled-container.clickable:hover {
    border: 1px solid var(--base-color-error) !important;
    background: var(--background-color-error);
}

.pass-text{
    color: var(--base-color-success);
}
.fail-text{
    color: var(--base-color-error);
}
.neutral-text{
    color: var(--secondary-content-color);
}
/*--Generic Styled Containers--*/

/*--Generic Tooltip--*/
.tooltip-cell{
    overflow: visible;
}
.tooltip-container {
    position: relative;
    width: fit-content;
}

.tooltip-trigger {
    font-size: 16px;
    font-weight: 400;
    text-align: left;
}
.tooltip-trigger.underline {
    text-decoration-line: underline;
    text-decoration-style: dotted;
    text-underline-offset: auto;
    text-decoration-color: #C6C6C6E3;
    text-decoration-thickness: 13.5%;
    text-underline-position: from-font;
    text-decoration-skip-ink: none;
}
.tooltip {
    visibility: hidden;
    position: absolute;
    opacity: 0;
    z-index: 11;
    transition: opacity 0.3s ease, visibility 0s linear 0.3s;
    width: max-content;
    white-space: wrap;
    max-width: 220px;
    top: calc(100% + 4px);
    left: 0;
    border-radius: 2px;
    background-color: var(--background-color-tertiary);
    border: 1px solid rgba(0, 0, 0, 0.08);
    padding: 6px 8px;
    font-size: 11px;
    line-height: 16px;
    font-weight: 400;
    word-wrap: break-word;
    color: var(--primary-content-color);
    box-shadow: 0 2px 4px rgba(0,0,0,0.1);
}

.tooltip .number {
    font-weight: 600;
}

.tooltip-container:hover .tooltip {
    visibility: visible;
    opacity: 1;
    transition-delay: 0.3s;
}

/*--Generic Tooltip--*/

/*--Generic Tab Classes--*/
.tab {
    display: flex;
    gap: 24px;
    max-width: 1200px;
    margin: 0 auto;
}
.tab-button {
    display: inline-block;
    padding: 12px 0 12px 0;
    font-size: 16px;
    color: var(--tertiary-content-color);
    font-weight: 400;
    border: none;
    background: none;
}
.tab-button:hover {
    cursor: pointer;
    color: var(--secondary-content-color);
}
.tab-button.active {
    color: var(--primary-content-color);
    box-shadow: inset 0 -3px 0 var(--base-color-brand);
    font-weight: 500;
}
.tab-content {
    display: none;
}
.tab-content.tab-summary {
    display: block;
}
/*--Generic Tab Classes--*/



/*--Header--*/
.heading-content{
    display: flex;
    flex-direction: column;
    gap: 15px;
    max-width: 1200px;
    margin: 0 auto;
    width: 100%;
}
.header-container, .content-container {
    padding: 24px 96px 0 96px;
    width: 100%;
    height: 100%;
}

.header-tab-container {
    width: 100%;
    border-bottom: 1px solid var(--border-color-default);
    position: sticky;
    top: 0;
    z-index: 10;
    padding: 0 96px;
}
.heading-content .heading {
    color: var(--secondary-content-color);
    font-weight: 500;
    text-transform: capitalize;
    letter-spacing: 2px;
    font-size: 12px;
}
.heading-content .report-info {
    display: flex;
    flex-direction: column;
    gap: 4px;
    max-width: 100%;
    margin-bottom: 20px;
}

.report-info .collection-name {
    color: var(--primary-content-color);
    font-weight: 600;
    font-size: 28px;
    letter-spacing: -0.4px;
    display: -webkit-box;
    -webkit-line-clamp: 2;
    line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    word-wrap: break-word;
    white-space: normal;
    line-height: 1.3;
    min-height: 0;
}

.report-info .collection-time {
    color: var(--tertiary-content-color);
    font-weight: 400;
    font-size: 16px;
    letter-spacing: -0.4px;
    white-space: nowrap;
    flex-shrink: 0;
    line-height: 1.3;
}

@media (max-width: 768px) {
    .header-container, .content-container {
        width: 100%;
        padding: 24px 8px 0 8px;
    }
    .heading-content {
        align-items: center;
    }
    .tab {
        justify-content: center;
    }
    .heading-content .report-info {
        display: flex;
        flex-direction: column;
    }
    .report-info .collection-name {
        max-width: 250px;
    }
}
/*--Header--*/

/*--Footer--*/
.footer {
    background-color: var(--background-color-secondary);
    padding: 16px;
    border-radius: 8px;
    display: flex;
    gap: 23px;
    align-items: center;
}
.footer .footer-icon-container{
    padding: 0 10px;
}
.footer-icon-container .footer-icon{
    width: 122px;
    height: 118px;
}
.footer .footer-text-container{
    display: flex;
    flex-direction: column;
    gap: 12px;
    width: 637px;
}
.footer .footer-title{
    letter-spacing: -0.12px;
    font-size: 16px;
    font-weight: 600;
}
.footer .footer-content{
    font-size: 14px;
    font-weight: 400;
    line-height: 20px;
}

@media (max-width: 400px) {
    .footer {
        flex-direction: column;
    }
    .footer .footer-icon{
        width: 80px;
        height: 90px;
    }
    .footer .footer-text-container{
        width: 100%;
    }
}
/*--Footer--*/

/*--Main Content--*/
.content-container{
    border: none;
}
.summary-content{
    display: flex;
    flex-direction: column;
    gap: 24px;
    max-width: 1200px;
    margin: 0 auto;
    width: 100%;
}
.all-requests-content{
    display: flex;
    gap: 24px;
    flex-direction: column;
    max-width: 1200px;
    margin: 0 auto;
    width: 100%;
}

@media (max-width: 768px) {
    .summary-content,.all-requests-content{
        padding: 0 16px 0 16px;
    }
}
/*--Main Content--*/

/*--Overview Container*/
.overview-container{
    padding: 16px;
    border: var(--section-border);
    border-radius: var(--border-radius);
}
.run-info-container{
    display: grid;
    grid-template-columns: 25% 25% 25% 25%;
    gap: 4px;
}
.overview-container .overview-data-container{
    display: flex;
    flex-direction: column;
    gap: 8px;
}
.overview-data-container .overview-data-title{
    color: var(--tertiary-content-color);
    font-weight: 400;
    font-size: 12px;
}
.overview-data-container .overview-data-content{
    color: var(--primary-content-color);
    font-size: 14px;
    font-weight: 400;
    padding-right: 10px;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
    max-width: 200px;
}
@media (max-width: 768px) {
    .overview-container .run-info-container {
        display: flex;
        flex-direction: column;
        gap: 20px;
        align-items: center;
    }
    .overview-container .overview-data-container {
        align-items: center;
    }
    .overview-data-container .overview-data-content{
        max-width: 250px;
    }
}
/*--Overview--*/

/*--Collection-Description--*/
#collection-description {
    position: relative;
    font-size: 14px;
}
#collection-description .read-more-btn {
    color: var(--secondary-content-color);
    font-weight: 400;
    font-size: 14px;
    cursor: pointer;
    text-decoration: underline;
    margin-top: 10px;
    display: none;
}
#collection-description .read-more-btn.visible {
    display: inline-block;
}
#collection-description .description-content-wrapper {
    overflow: hidden;
    position: relative;
}
#collection-description .description-content-wrapper.collapsed {
    max-height: 100px;
    position: relative;
}
#collection-description .description-content-wrapper.collapsed::after {
    content: '';
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    height: 75px;
    opacity: 0;
    background: linear-gradient(180deg, rgba(255, 255, 255, 0) 0%, rgba(255, 255, 255, 0.8) 60%, rgba(255, 255, 255, 1) 100%);
    pointer-events: none;
}

#collection-description .description-content-wrapper.collapsed.show-gradient::after {
    opacity: 1;
}
#collection-description .description-content-wrapper.expanded {
    max-height: 400px;
    overflow-y: auto;
}
/*--Collection Description*/

/*--Stats Card--*/
.stats-container {
    display: flex;
    gap: 12px;
    flex-wrap: wrap;
}

.stats-container .stats-data-container {
    flex:1;
    display: flex;
    flex-direction: column;
    padding: 16px 12px 16px 12px;
    width: 200px;
    gap:20px;
    border-radius: 8px;
    border: 1px solid var(--border-color-default);
}

.stats-data-container .stats-data-title {
    font-size: 12px;
    font-weight: 400;
    line-height: 16px;
    text-align: left;
    text-underline-position: from-font;
    text-decoration-skip-ink: none;
}
.stats-data-container .stats-data-content {
    font-family: IBM Plex Mono, serif;
    font-size: 54px;
    font-weight: 400;
    letter-spacing: -1.600000023841858px;
    text-align: left;
    text-underline-position: from-font;
    text-decoration-skip-ink: none;
    overflow-wrap: anywhere;
    line-height: 45px;
}
.stats-data-container .stats-data-metric-container {
    display: flex;
    gap: 10px;
    align-items: baseline
}
.stats-data-metric-container .stats-content-percentage{
    color: var(--secondary-content-color);
    font-size: 14px;
    font-weight: 500;
    line-height: 16px;
    text-align: left;
    text-underline-position: from-font;
    text-decoration-skip-ink: none;
}

@media (max-width: 768px) {
    .stats-container {
        display: grid;
        grid-template-columns: 49% 49%;
    }
    .stats-container .stats-data-container{
        width: auto;
    }
    .stats-data-container .stats-content {
        font-size: 34px;
        overflow: hidden;
        text-overflow: ellipsis;
    }
}
/*--Stats Card--*/

/*--Run Summary--*/
#run-summary-container{
    padding: 16px;
    border-radius: 8px;
    border: var(--section-border);
}
table#summary-table {
    width: 100%;
    border-spacing: 0 10px;
    table-layout: fixed;
    border-radius: 0 !important;
}
table#summary-table td, th {
    border-collapse: collapse;
    border-bottom: 1px solid #EDEDEDB2;
    padding: 4px 0;
    height: 40px;
    text-align: left;
    font-weight: 400;
    line-height: 1.5rem;
    word-wrap: break-word;
    word-break: break-word;
    max-height: 150px;
}

table#summary-table tr:last-child td, tr:last-child th {
    border-bottom: none;
}
table#summary-table .metric-heading{
    text-align:right;
    width: 136px;
}
table#summary-table td .content {
    display: inline-block;
    width: fit-content;
    padding: 0 12px;
    border-radius: 4px;
    min-width: 3rem;
    text-align: center;
}
table#summary-table th {
    color: var(--secondary-content-color);
}

@media (max-width: 768px) {
    table#summary-table .metric-heading{
        width: auto;
    }
}
/*--Run Summary--*/

/*--Failure Tests--*/
.failure-container,.skipped-container{
    padding: 16px;
    border: var(--section-border);
    border-radius: var(--border-radius);
}

.iterations-summary-heading {
    display: flex;
    gap: 12px;
    letter-spacing: -0.12px;
    font-weight: 600;
    font-size: 16px;
    line-height: 20px;
}
.iterations-summary-heading .iterations-summary-heading-data {
    flex: 1;
    display: flex;
    flex-direction: column;
    gap: 8px;
}
.iterations-summary-heading-data .iterations-summary-heading-title {
    padding-top: 3px;
    font-size: 16px;
    font-weight: 600;
    letter-spacing: -0.12px
}
.iterations-summary-heading-data p {
    font-weight: 400;
    font-size: 14px;
    line-height: 22px;
}
.iteration-folder-structure {
    display: flex;
    margin-right: 10px;
    align-items: center;
    gap: 6px;
    flex-wrap: wrap
}
.iteration-folder-structure .folder-name {
    color: var(--primary-content-color);
    font-size: 14px;
    font-weight: 400;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
    display: inline-block;
    max-width: 500px;
}
.iteration-folder-structure .request-method {
    font-family: IBM Plex Mono, monospace;
    font-weight: 600;
    font-size: 12px;
    padding-top: 1px;
}

.folder-slash {
    color: var(--secondary-content-color);
    font-size: 14px;
    font-weight: 400;
    margin: 0 2px;
}
.iterations-summary-action-btn-container {
    display: flex;
    gap: 8px;
}
.iterations-summary-action-btn {
    height: 32px;
    padding: 6px 12px 6px 12px;
    border: none;
    border-radius: 4px;
    background: var(--grey-10);
    font-weight: 400;
    font-size: 12px;
    cursor: pointer;
}

.iterations-summary-action-btn:hover {
    background-color: var(--border-color-default);
}

/*--Iterations List--*/
.iterations-list-container {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    gap: 8px;
}
.iterations-list-container .acc-container {
    font-size: 14px;
    font-weight: 400;
    background: var(--background-color-tertiary);
    padding: 8px;
    border-radius: 8px;
}

.iterations-list-container .acc-container:hover {
    background-color: var(--background-color-secondary);
}

.iterations-list-container .acc-container.open {
    background: var(--background-color-secondary);
}
.iterations-list-container .acc-container .acc-heading {
    padding-right: 16px;
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.iteration-request-title-container {
    display: flex;
}
.iteration-details-title {
    display: flex;
    align-items: center;
    gap: 8px;
    flex-wrap: wrap;
    margin-left: 8px;
}
.iteration-details-title .iteration-number{
    flex-shrink: 0;
    padding-left:3px;
    line-height: 24px;
}
.iterations-list-container .acc-container .acc-heading .view-iteration-request-btn {
    display: none;
    color: var(--secondary-content-color);
    font-family: var(--font-primary);
    font-size: 11px;
    font-weight: 400;
    line-height: 16px;
    text-align: left;
    text-decoration-line: underline;
    text-underline-position: from-font;
    text-decoration-skip-ink: none;
    border: none;
    cursor: pointer;
    background: none;
    flex-shrink: 0;
}

.iterations-list-container .acc-container .acc-heading.active .view-iteration-request-btn {
    display: block;
}

.iterations-list-container .acc-container .acc-heading.active .view-iteration-request-btn:hover {
    color: var(--content-color-link);
}

.iterations-list-container .acc-container .acc-content {
    padding: 0 12px 0 32px;
}

.errors-and-failures-group-container {
    display: flex;
    flex-direction: column;
    gap: 8px;
    margin-top: 8px;
    margin-bottom: 4px;
}

.iteration-request-detail-container {
    font-size: 14px;
    padding: 24px 23px;
    border: 1px solid transparent;
    background: linear-gradient(white, white) padding-box,linear-gradient(221.78deg, rgba(22, 22, 22, 0.04) 71.03%, rgba(198, 198, 198, 0.21) 79.48%),linear-gradient(221.6deg, rgba(198, 198, 198, 0.19) 25.22%, rgba(33, 33, 33, 0.17) 38.45%) border-box;
    border-radius: 8px;
    box-shadow: 0 2px 2px 0 #2121210F, 0 8px 12px -4px #21212114;
}

.iteration-request-detail-container.skipped-tests {
    margin: 10px 0;
}

.iteration-request-error-detail {
    font-family: var(--font-secondary);
    margin-top:12px;
    background: var(--background-color-error);
    color:var(--primary-content-color);
}
.script-request-error-detail {
    padding: 5px 0;
    font-size: 12px;
}
/*--Failure Tests--*/

/*--Generic Accordion Classes--*/
.acc-container .acc-heading {
    cursor: pointer;
}
.acc-container .acc-heading .acc-toggle-btn {
    border: 1px solid var(--border-secondary);
    padding: 4px;
    height: 24px;
    width: 24px;
    border-radius: 4px;
    display: flex;
    align-items: center;
    justify-content: center;
    background: transparent;
    cursor: pointer;
}
.acc-container .acc-heading .acc-toggle-btn .acc-toggle-icon {
    transition: transform 0.3s ease;
}
.acc-container .acc-heading.active .acc-toggle-btn .acc-toggle-icon {
    transform: rotate(90deg);
}
.acc-container .acc-content {
    transition: max-height 0.3s cubic-bezier(0.4, 0, 0.2, 1);
    overflow: hidden;
    max-height: 0;
    padding: 0 0 0 34px;
}
.acc-container .acc-content.active {
    max-height: 100000px;
}
/*--Generic Accordion Classes--*/


@media (max-width: 768px) {
    .acc-container .acc-content {
        padding: 0;
    }
    .iterations-list-container .acc-container .acc-content {
        padding: 0;
    }
    .iterations-summary-heading {
        flex-direction: column;
    }
    .iterations-list-container .acc-container .acc-heading {
        flex-direction: column;
        gap: 10px;
        align-items: start;
    }
    .iterations-list-container .acc-container .acc-heading.active .view-iteration-request-btn{
        align-self: end;
    }
}

/* requests tab css */

/* Typography Utilities */
.text-xs { font-size: 11px; }
.text-sm { font-size: 12px; }
.text-base { font-size: 14px; }
.text-lg { font-size: 16px; }

.font-regular { font-weight: var(--font-weight-regular); }
.font-medium { font-weight: var(--font-weight-medium); }
.font-semibold { font-weight: var(--font-weight-semibold); }

.font-mono { font-family: 'IBM Plex Mono', monospace; }
.font-primary { font-family: var(--font-primary), serif; }

/* Text Colors */
.text-primary { color: var(--primary-content-color); }
.text-secondary { color: var(--secondary-content-color); }
.text-tertiary { color: var(--tertiary-content-color); }
.text-muted { color: var(--text-muted); }
.text-error { color: var(--base-color-error); }
.text-success { color: var(--base-color-success); }

/* Background Colors */
.bg-white { background-color: white; }
.bg-secondary { background-color: var(--background-color-secondary); }
.bg-tertiary { background-color: var(--background-color-tertiary); }
.bg-error { background-color: var(--background-color-error); }
.bg-success-light { background-color: var(--green-30); }

/* Layout Utilities */
.flex { display: flex; }
.flex-col { display: flex; flex-direction: column; }
.flex-row-wrap { display: flex; flex-wrap: wrap; }
.items-center { align-items: center; }
.justify-between { justify-content: space-between; }
.full-width { width: 100%; }
.hidden { visibility: hidden; }

.gap-2 { gap: 2px; }
.gap-4 { gap: 4px; }
.gap-5 { gap: 5px; }
.gap-8 { gap: 8px; }
.gap-12 { gap: 12px; }
.gap-16 { gap: 16px; }

/* Spacing Utilities */
.p-1 { padding: 1px; }
.p-2 { padding: 2px; }
.p-3 { padding: 3px; }
.p-4 { padding: 4px; }
.p-8 { padding: 8px; }
.p-12 { padding: 12px; }
.p-16 { padding: 16px; }
.p-20 { padding: 20px; }
.p-24 { padding: 24px; }

.px-12 { padding-left: 12px; padding-right: 12px; }
.py-12 { padding-top: 12px; padding-bottom: 12px; }
.py-16 { padding-top: 16px; padding-bottom: 16px; }

.mt-16 { margin-top: 16px; }
.mt-24 { margin-top: 24px; }
.mb-8 { margin-bottom: 8px; }
.mr-2 { margin-right: 2px; }

/* Border Utilities */
.border { border: 1px solid; }
.border-default { border-color: var(--border-color-default); }
.border-secondary { border-color: var(--border-secondary); }
.border-error { border-color: var(--red-20); }
.border-success { border-color: var(--green-30); }

.rounded { border-radius: 4px; }
.rounded-lg { border-radius: 8px; }
.rounded-full { border-radius: 9999px; }

.iteration-nav-container {
    display: flex;
    flex-wrap: wrap;
    gap: 5px;
    max-height: 105px;
    overflow-y: auto;
    margin: 12px 0;
}

.iteration-btn {
    min-width: 24px;
    min-height: 24px;
    border: 0.5px solid;
    background-color: white;
    border-radius: 4px;
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 12px;
    font-family: "IBM Plex Mono", serif;
    font-weight: 400;
}

.iteration-btn.hidden-btn {
    display: none;
}

.iteration-btn.small {
    min-width: 24px;
    min-height: 24px;
}
.iteration-btn.passed {
    background-color: #F8FFFB;
    border-color: #A4EEC4;
    color: var(--base-color-success);
}

.iteration-btn.passed:hover {
    background-color: #E5FFF1;
    border-color: #38CC74;
}

.iteration-btn.failed {
    border-color: var(--red-20);
    background-color: #FFF7F5;
    color: var(--base-color-error);
}

.iteration-btn.failed:hover {
    border-color: #F37264;
    background-color: #FFEBE7;
}

.iteration-btn.skipped {
    background-color: #F8FFFB;
    border-color: #A4EEC4;
    color: #22c55e;
}

.iteration-btn.skipped:hover {
    background-color: #E5FFF1;
    border-color: #38CC74;
}

.iteration-btn.selected {
    background-color: #f8f8f8;
    font-weight: bold;
}
.iteration-btn.passed.selected {
    background-color: var(--base-color-success);
    color: white;
}
.iteration-btn.failed.selected {
    background-color: var(--red-60);
    color: white;
}
.iteration-btn.skipped.selected {
    background-color: var(--base-color-success);
    color: white;
}

.iteration-filter-container{
    display: flex;
    flex-wrap: wrap;
    gap: 4px;
    margin: 8px 0 4px 0;
}

.filter-button {
    background-color: var(--background-color-tertiary);
    border: none;
    text-align: center;
    border-radius: 30px;
    padding: 6px 12px;
    font-size: 11px;
    font-weight: 600;
    color: var(--secondary-content-color);
    cursor: pointer;
    line-height: 16px;
    max-height: 28px;
}

.filter-button:hover {
    background-color: var(--border-color-default);
}

.filter-button.selected {
    background-color: #2B2B2B;
    color: #FFF;
}

.requests-filter-container {
    display: flex;
    flex-wrap: wrap;
    gap: 8px;
}

.button {
    padding: 6px 8px;
    cursor: pointer;
    font-size: 12px;
    font-weight: 400;
    line-height: 20px;
    min-width: 100px;
    text-align: center;
    justify-content: center;
    background-color: var(--background-color-tertiary);
    border-radius: 4px;
}

.button:hover {
    background-color: var(--border-color-default);
}

.dynamic-accordion {
    border: 1px solid;
    border-radius: 8px;
    border-color: var(--border-secondary);
}

.dynamic-accordion.parent {
    background-color: var(--background-color-tertiary);
}

.dynamic-accordion.parent.active {
    background-color: var(--background-color-secondary);
}

.dynamic-accordion-content {
    display: none;
}

.dynamic-accordion-content.active {
    display: block;
}

@media (max-width: 768px) {
    .dynamic-accordion-content {
        padding: 8px 12px 8px 16px;
    }
}

.dynamic-accordion.no-border {
    border: 0 !important;
}

.dynamic-accordion-header {
    cursor: pointer;
    display: flex;
    gap: 16px;
    align-items: center;
    flex-wrap: wrap;
}

.dynamic-accordion.parent:hover {
    background-color: var(--background-color-secondary);
}

.dynamic-accordion-header.active .dynamic-accordion-carat {
    transition: transform 0.3s ease;
    transform: rotate(90deg);
}

.dynamic-accordion-header.active .execution-info-section-header {
    border-bottom: 1px solid var(--border-color-default);
}

.iteration-details-container {
    display: flex;
    flex-direction: column;
    gap: 8px;
    min-height: 200px;
}

.execution {
    display: flex;
    gap: 16px;
    padding: 23px;
    width: 100%;
}

@media (max-width: 768px) {
    .execution {
        overflow: hidden;
    }
}

.execution-content {
    display: flex;
    justify-content: space-between;
    align-items: center;
    gap: 12px;
    flex: 1;
}

.http-verb {
    font-family: IBM Plex Mono, monospace;
    font-size: 12px;
    font-weight: 600;
    padding-top: 1px;
}

.execution-metadata {
    display: flex;
    flex-wrap: wrap;
    gap: 16px
}

.execution-name {
    font-size: 14px;
    font-weight: 500;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
    display: inline-block;
    max-width: 575px;
}

.execution-url {
    font-size: 12px;
    color: var(--secondary-content-color);
    font-weight: 400;
    justify-content: flex-start;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    display: inline-block;
    max-width: 600px;
}

.execution-info-section-header {
    padding: 11px;
    font-size: 12px;
    font-style: normal;
    font-weight: 600;
    line-height: 16px;
    color: #212121;
    width: 100%;
    display: flex;
    gap: 10px;
    align-items: center;
}

.execution-folder-container {
    display: flex;
    padding: 8px;
    gap: 8px
}

.execution-folder-name {
    font-size: 14px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 800px;
}

@media (max-width: 1250px) {
    .execution-folder-name {
        max-width: 500px;
    }

    .execution-name {
        max-width: 500px;
    }

    .execution-url {
        max-width: 500px;
    }
}

@media (max-width: 900px) {
    .execution-folder-name {
        max-width: 200px;
    }

    .execution-name {
        max-width: 150px;
    }

    .execution-url {
        max-width: 200px;
    }
}

.dynamic-accordion-content-wrapper {
    display: flex;
    flex-direction: column;
    gap: 8px;
}

.dynamic-accordion-content-wrapper.http-headers {
    overflow-x: auto;
    margin-top: 10px;
    padding: 0 4px 12px 40px;
}

.dynamic-accordion-content-wrapper.error-section {
    padding: 12px 12px 12px 40px;
}

.dynamic-accordion-content-wrapper.tests-section {
    margin-top: 10px;
    padding: 0 2px 8px 40px;
}

.dynamic-accordion-content-wrapper.exec-content {
    padding: 0 24px 24px 61px;
}

.dynamic-accordion-content-wrapper.folder-content {
    padding: 8px 8px 12px 40px;
}

.dynamic-accordion-content-wrapper.iteration-details {
    margin-top: 16px;
    padding-left: 40px;
}

.execution-error-container {
    display: flex;
    align-items: center;
    border: 1px solid var(--border-secondary);
    border-radius: 8px;
    padding: 24px;
}

.execution-bodies-section-content {
    padding-left: 40px;
}

@media (max-width: 768px) {
    .execution-bodies-section-content {
        padding-left: 10px;
    }
}

.execution-error-container > span {
    padding: 12px;
    font-size: 14px;
    font-weight: 400;
    background-color: var(--background-color-error);
    font-family: "IBM Plex Mono", serif;
    word-wrap: break-word;
    color: var(--red-60);
    width: 100%;
}

.total-executions-count {
    display: flex;
    flex: 1;
    justify-content: flex-end;
    font-size: 11px;
    color: var(--tertiary-content-color);
}

@media (max-width: 768px) {
    .dynamic-accordion-content-wrapper {
        padding: 2px !important;
    }
}

.test-stat {
    border-radius: 2px;
    background-color: white;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 4px;
    color: var(--text-muted);
}

.test-stat > span {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    font-size: 12px;
    font-weight: 400;
    aspect-ratio: 1;
    min-width: 10px;
    min-height: 10px;
}

.test-stat.passed {
    background-color: var(--background-color-success);
}

.test-stat.passed > span {
    color: var(--content-color-success);
}

.test-stat.failed {
    background-color: var(--background-color-error);
}

.test-stat.failed.blur {
    background-color: #FFF7F5;
}

.test-stat.failed > span {
    color: var(--base-color-error);
}

.test-stat.failed.blur > span {
    color: var(--red-30);
}

.dot {
    width: 4px;
    height: 4px;
    border: 1px solid;
    border-radius: 8px;
    border-color: var(--border-color-default);
    background-color: var(--border-secondary);
}

.http {
    display: flex;
    padding: 4px 6px;
    border-radius: 4px;
    font-size: 12px;
    font-weight: 500;
    border: 0;
}

.http.success {
    background-color: var(--background-color-success);
    color: var(--content-color-success);
}

.http.error {
    background-color: var(--background-color-error);
    color: var(--content-color-error);
}

.http.redirect {
    background-color: var(--background-color-error);
    color: #8a8483;
}

/* http headers tables start */

table {
    width: 100%;
    border-collapse: collapse;
    table-layout: fixed;
}

.http-headers-table {
    border-collapse: collapse;
}

.http-headers-table th:first-child,
.http-headers-table td:first-child {
    width: 144px;
}

.http-headers-table th,
.http-headers-table td {
    border-bottom: 1px solid var(--border-color-default) !important;
    text-align: left;
    font-size: 12px;
    line-height: 20px !important;
    font-weight: 400;
    color: var(--primary-content-color);
    padding: 8px;
    line-height: 20px;
    word-wrap: break-word;
}

.http-headers-table .cell-content {
    max-height: 100px;
    overflow-y: auto;
}

.http-headers-table tr:last-child td {
    border-bottom: none !important;
}

.http-headers-table th {
    color: var(--secondary-content-color);
    font-weight: 500;
}

/* http headers tables end */

/* execution assertion details */

.assertions-table th,
.assertions-table td {
    text-align: left;
    padding: 8px;
    word-wrap: break-word;
    border-bottom: 1px solid var(--border-color-default) !important;
}

.assertions-table th {
    color: var(--secondary-content-color);
    font-weight: 500;
    font-size: 12px;
    line-height: 20px;
}

.assertions-table th:first-child,
.assertions-table td:first-child {
    width: 75px;
}

.assertions-table tr:last-child td {
    border-bottom: none !important;
}

.assertions-table td {
    font-size: 14px;
    font-weight: 500;
    line-height: 20px;
}

tr.error-row td {
    border-bottom: none !important;
}

.test-status-badge {
    display: inline-flex;
    padding: 0.1rem 0.5rem;
    border-radius: 116px !important;
    font-size: 11px;
    font-weight: 600;
    line-height: 16px;
    background-color: var(--background-color-secondary);
}

.test-status-badge.passed {
    background-color: var(--background-color-success);
    color: var(--content-color-success);
}

.test-status-badge.failed {
    background-color: var(--background-color-error);
    color: var(--content-color-error);
}

.error-row td {
    border-top: none;
}

.tests-row {
    height: 44px;
}

.test-error-data {
    display: block;
    width: 100%;
    font-size: 14px;
    font-weight: 400;
    background-color: var(--background-color-error);
    font-family: "IBM Plex Mono", serif;
}

.error-msg-row td {
    padding-top: 0 !important;
}

.execution-errors-container {
    font-size: 14px;
    padding: 23px;
    border: 1px solid var(--border-secondary);
    border-radius: 8px;
}

.execution-error-detail .hljs {
    font-size: 14px;
    background-color: var(--background-color-error) !important;
    color: var(--primary-content-color) !important;
    text-wrap: wrap;
}


/* execution assertion details end */

/* highlight js customisation start */
.hljs {
    font-family: "IBM Plex Mono", serif;
    background: var(--background-color-secondary);
    padding: 16px 12px 16px 16px;
    font-size: 12px;
    line-height: 20px;
    font-weight: 400;
    overflow-y: auto;
    overflow-x: auto;
    max-height: 600px;
    text-align: left !important;
    display: block !important;
    color: var(--secondary-content-color);
}

.test-error-data .hljs {
    font-size: 14px;
    padding: 12px;
    font-weight: 400;
    text-wrap: wrap;
    background: var(--background-color-error);
    color: var(--primary-content-color);
}

.test-error-data .hljs-number {
    color: var(--red-60);
}

.iteration-request-error-detail .hljs {
    background: none;
    padding: 0;
    color: var(--primary-content-color);
    font-size: 14px;
    text-wrap: wrap;
}

.iteration-request-error-detail .hljs-section, .hljs-title {
    font-weight: var(--font-weight-medium);
    color:var(--red-60);
}

.hljs-attr {
    color: var(--red-60) !important;
}
.hljs-string {
    color: #0053B8;
}

/* highlight js customisation end */

/* execution request or response content/body */
.load-body-placeholder {
    width: 100%;
    display: flex;
    flex-direction: column;
    padding: 20px;
    gap: 12px;
    background-color: var(--background-color-secondary);
}

.load-body-placeholder-text {
    font-size: 11px;
    font-weight: 400;
    color: var(--secondary-content-color);
}

.load-body-btn {
    cursor: pointer;
    font-size: 12px;
    font-weight: 400;
    color: var(--primary-content-color);
    border: 1px solid var(--border-secondary);
    border-radius: 4px;
    padding: 4px 12px;
}

.load-body-btn.loading {
    cursor: not-allowed;
    color: var(--tertiary-content-color);
}

.copyable-container {
    position: relative;
    width: 100%;
}

.copy-button {
    font-size: 14px;
    cursor: pointer;
    display: flex;
    align-items: center;
    gap: 5px;
    color: #555;
    transition: background-color 0.2s;
}


`;
module.exports = htmlStyles;
