// @postman/pm-bin-linux-x64/bin/lib/reporters/html/util.js
const { CONTENT_TYPES, CONTENT_TYPE_HEADER_KEY } = require('./constants'), _ = require('lodash'), commonUtil = require('../../util'), util = {
    getRequestUrlString: function (urlObj) {
        let host = urlObj.host ? urlObj.host : [], protocol = urlObj.protocol ? urlObj.protocol : 'https', url = `${protocol}://${host.join('.')}`;
        if (urlObj.path && urlObj.path.length) {
            url += `/${urlObj.path.filter((val) => { return !_.isNil(val); }).join('/')}`;
        }
        const queryParams = (urlObj.query && urlObj.query.members && urlObj.query.members.length) ?
            urlObj.query.members.map((param) => { return `${param.key}=${param.value}`; }).join('&') : null, variables = urlObj.variable && urlObj.variable.length ?
            urlObj.variable.map((v) => { return `${v.key}=${v.value}`; }).join('&') : null;
        if (queryParams) {
            url += `?${queryParams}`;
        }
        if (variables) {
            url += url.includes('?') ? `&${variables}` : `?${variables}`;
        }
        return url;
    },
    deduceContentType: function (value) {
        if (!value) {
            return CONTENT_TYPES.OTHER;
        }
        const contentTypeMap = {
            'application/json': CONTENT_TYPES.JSON,
            'text/json': CONTENT_TYPES.JSON,
            'application/xml': CONTENT_TYPES.XML,
            'text/xml': CONTENT_TYPES.XML,
            'text/plain': CONTENT_TYPES.TEXT,
            'application/javascript': CONTENT_TYPES.JAVASCRIPT,
            'text/javascript': CONTENT_TYPES.JAVASCRIPT,
            'text/html': CONTENT_TYPES.HTML,
            'application/x-www-form-urlencoded': CONTENT_TYPES.FORM_URLENCODED,
            'multipart/form-data': CONTENT_TYPES.MULTIPART,
            'application/octet-stream': CONTENT_TYPES.BINARY
        };
        for (const [key, type] of Object.entries(contentTypeMap)) {
            if (value.includes(key)) {
                return type;
            }
        }
        return CONTENT_TYPES.OTHER;
    },
    getSkippedTestsWithStats: function (tests) {
        if (!tests || !tests.length) {
            return null;
        }
        const stats = {
            total: tests.length,
            uniqueRequests: new Set(tests.map((test) => {
                return `${test.source.id}-${test.iterationIndex}`;
            })).size
        };
        return {
            stats: stats,
            items: tests
        };
    },
    aggregateFailures: function (summary, executions) {
        if (!summary || !summary.run || !summary.run.failures || !summary.run.failures.length) {
            return { status: {}, items: [] };
        }
        const aggregatedFailures = [], isScriptLevelRequest = ({ at, source = {}, cursor = {} }) => {
            return at === 'request' && !source.name && cursor.scriptId !== undefined;
        };
        summary.run.failures.forEach((failure) => {
            const cursorRef = failure.cursor.ref, associatedExecution = executions[cursorRef], associatedParent = associatedExecution && associatedExecution.parent, associatedItem = associatedExecution && associatedExecution.item, associatedRequest = associatedExecution && associatedExecution.request, parentPath = commonUtil.getFullPath(associatedParent), iterationIndex = failure.cursor.iteration, error = {
                name: failure.error && failure.error.name,
                message: failure.error && failure.error.message,
                test: failure.error && failure.error.test,
                scriptLevelRequest: isScriptLevelRequest(failure)
            }, lastFailure = _.last(aggregatedFailures);
            if (associatedItem && lastFailure && lastFailure.source.id === associatedItem.id &&
                lastFailure.iterationIndex === iterationIndex) {
                lastFailure.errors.push(error);
            }
            else {
                aggregatedFailures.push({
                    iterationIndex: iterationIndex,
                    parent: {
                        id: (associatedParent && associatedParent.id) ? associatedParent.id :
                            (summary.collection && summary.collection.id) || null,
                        name: parentPath.join(' / '),
                        folders: parentPath
                    },
                    source: {
                        id: associatedItem ? associatedItem.id : '',
                        name: associatedItem ? associatedItem.name : '',
                        method: associatedRequest && associatedRequest.method
                    },
                    errors: [error]
                });
            }
        });
        const stats = {
            total: summary.run.failures.length,
            failures: summary.run.failures.filter((f) => { return f.error.name === 'AssertionError'; }).length,
            errors: summary.run.failures.filter((f) => { return f.error.name !== 'AssertionError'; }).length,
            uniqueRequests: new Set(aggregatedFailures.map((f) => {
                return `${f.source.id}-${f.iterationIndex}`;
            })).size
        };
        return { stats: stats, items: aggregatedFailures };
    },
    getContentTypeByHeaders: function (headers) {
        if (!headers || !_.isFunction(headers.find)) {
            return CONTENT_TYPES.MISSING;
        }
        const contentTypeHeader = headers.find((h) => {
            return h.key.toLowerCase() === CONTENT_TYPE_HEADER_KEY;
        });
        if (contentTypeHeader) {
            return this.deduceContentType(contentTypeHeader.value);
        }
        return CONTENT_TYPES.MISSING;
    },
    normalizeFormData: function (members) {
        if (!members || members.length === 0) {
            return '{ "type": "formdata": data: {} }';
        }
        const data = {};
        members.forEach(function (value) {
            if (!value.disabled) {
                if (value.src) {
                    data[value.key] = value.src;
                }
                else {
                    data[value.key] = value.value;
                }
            }
        });
        return JSON.stringify({ type: 'formdata', data: data });
    },
    normalizeGraphQlData: function (graphql) {
        if (!graphql) {
            return '{ "type": "graphql": data: {} }';
        }
        const data = {};
        data.query = graphql.query;
        if (graphql.variables) {
            data.variables = graphql.variables;
        }
        return JSON.stringify({ type: 'graphql', data: data });
    }
};
module.exports = util;
