// @postman/pm-bin-linux-x64/bin/lib/config/cli-environment/beta.js
const { createCliEnvironmentConfig } = require('./factory');
const { REGIONS } = require('../../constants'), settings = {
    channel: 'beta',
    baseUrls: {
        [REGIONS.US]: {
            api: 'https://api.getpostman-beta.com',
            artemis: 'https://go.postman-beta.co',
            iapub: 'https://iapub.postman-beta.co',
            gateway: 'https://gateway.postman-beta.com',
            packman: 'https://packman.pstmn-beta.io',
            runtimeAgent: 'https://ra.gw.postman-beta.com/v1',
            componentLibrary: 'https://components.pstmn-beta.io'
        },
        [REGIONS.EU]: {
            api: 'https://api.eu.postman-alpha.com',
            artemis: 'https://go.eu.postman-alpha.co',
            iapub: 'https://iapub.eu.postman-alpha.co',
            gateway: 'https://gateway.eu.postman-alpha.com',
            componentLibrary: 'https://components.eu.pstmn-alpha.io'
        }
    },
    postmanHomeDir: '.postman-beta',
    logLevel: 'debug',
    enableFeatureFlags: [
        'grpc_protocol_execution_allowed',
        'graphql_v2_protocol_execution_allowed'
    ]
};
module.exports = createCliEnvironmentConfig(settings);
