// @postman/pm-bin-linux-x64/bin/lib/config/cli-environment/default.js
const { createCliEnvironmentConfig } = require('./factory');
const { REGIONS } = require('../../constants'), settings = {
    channel: 'prod',
    baseUrls: {
        [REGIONS.US]: {
            api: 'https://api.getpostman.com',
            artemis: 'https://go.postman.co',
            iapub: 'https://iapub.postman.co',
            gateway: 'https://gateway.postman.com',
            packman: 'https://packman.pstmn.io',
            runtimeAgent: 'https://ra.gw.postman.com/v1',
            componentLibrary: 'https://components.pstmn.io'
        },
        [REGIONS.EU]: {
            api: 'https://api.eu.postman.com',
            artemis: 'https://go.eu.postman.co',
            iapub: 'https://iapub.eu.postman.co',
            gateway: 'https://gateway.eu.postman.com',
            componentLibrary: 'https://components.eu.pstmn.io'
        }
    },
    postmanHomeDir: '.postman',
    logLevel: 'error',
    enableFeatureFlags: [
        'grpc_protocol_execution_allowed',
        'graphql_v2_protocol_execution_allowed'
    ]
};
module.exports = createCliEnvironmentConfig(settings);
