// @postman/pm-bin-linux-x64/bin/lib/config/cli-environment/factory.js
const _ = require('lodash'), BASE_CLI_OPTIONS = {
    default: {
        alias: 'default'
    },
    run: {
        folder: [],
        reporters: ['cli'],
        delayRequest: 0,
        globalVar: [],
        envVar: [],
        color: 'auto',
        timeout: 0,
        timeoutRequest: 0,
        timeoutScript: 0,
        alias: 'default'
    },
    flows: {
        alias: 'default'
    }
}, BASE_COMMAND_KEY_MAP = {
    'collection.run': 'run'
};
function createCliEnvironmentConfig(settings, cliOptionsOverrides = {}) {
    const finalCLIOptions = _.mergeWith({}, BASE_CLI_OPTIONS, cliOptionsOverrides, (objValue, srcValue) => {
        if (_.isArray(objValue)) {
            return srcValue;
        }
    });
    return {
        load: (callback) => {
            return callback(null, finalCLIOptions);
        },
        getSettings: () => {
            return settings;
        }
    };
}
module.exports = {
    createCliEnvironmentConfig,
    BASE_CLI_OPTIONS,
    BASE_COMMAND_KEY_MAP
};
