// @postman/pm-bin-linux-x64/bin/lib/commands/runner/index.js
const { Option } = require('commander'), { getWorkingDirectory } = require('./utils/working-directory'), { RunnerProxyManagerInstance: ProxyManager } = require('./proxy'), { parseUserProxyOptions } = require('./utils/proxy-options-parser'), MetricsServer = require('./metrics'), regionUtil = require('../../region-util'), { REGIONS } = require('../../constants');
const prompts = require('prompts'), util = require('../../util'), { runLifecycle } = require('./core/lifecycle'), runnerConfig = require('./core/runner-config').get(), { getLogger } = require('./logging/logger-factory');
const logger = getLogger();
function sleep(milliseconds) {
    return new Promise((resolve) => {
        setTimeout(resolve, milliseconds);
    });
}
const config = {
    authenticationRetryLimit: 3,
    pingInterval: 5000,
    agentErrorDelay: 100,
    ...runnerConfig
};
let unauthorizedErrorCount = 0;
let metricsServer = null;
let workingDirectory = null;
function cleanUpRun() {
    if (workingDirectory) {
        util.removeCloudFilesFolder(workingDirectory);
    }
    else {
        logger.warn('Working directory not found, skipping cleanup of cloud files');
    }
}
async function cleanUpAll() {
    if (metricsServer) {
        await metricsServer.shutdown();
    }
    await ProxyManager.shutdown();
    if (workingDirectory) {
        util.removeCloudFilesFolder(workingDirectory);
    }
    else {
        logger.warn('Working directory not found, skipping cleanup of cloud files');
    }
}
function setupRegion(region) {
    try {
        const targetRegion = region ? region.toLowerCase() : REGIONS.US, validatedRegion = regionUtil.validateRegion(targetRegion, {
            defaultRegion: REGIONS.US,
            errorLevel: 'error'
        });
        regionUtil.setCurrentRegionForProcess(validatedRegion);
        return validatedRegion;
    }
    catch (error) {
        logger.error(`Error setting up region: ${error.message}`, {
            operation: 'setupRegion',
            error: error
        });
        process.exit(1);
    }
}
async function setupBuiltInProxy(options) {
    if (!options.egressProxy) {
        return;
    }
    const authzUrl = options.egressProxyAuthzUrl || process.env.POSTMAN_RUNNER_AUTHZ_URL;
    if (!authzUrl) {
        logger.error('Runner Authorization Service URL is required when using --egress-proxy. Please provide it via ' +
            '--egress-proxy-authz-url <url> CLI option or set the POSTMAN_RUNNER_AUTHZ_URL environment variable.', { operation: 'setupBuiltInProxy' });
        process.exit(1);
    }
    logger.info('Checking authorization service availability...');
    const AuthorizationService = require('./proxy/authorization/service');
    const authzService = new AuthorizationService(logger, authzUrl);
    const isReachable = await authzService.checkHealth();
    if (!isReachable) {
        logger.error('Authorization service is not reachable. The proxy cannot start without a working authorization service.', { operation: 'setupBuiltInProxy', authzUrl: authzUrl ? 'REDACTED' : 'not provided' });
        logger.info('Please ensure the authorization service is running and accessible.');
        logger.info('To run without the proxy, remove the --egress-proxy flag.');
        process.exit(1);
    }
    try {
        await ProxyManager.start({ authzBaseUrl: authzUrl });
    }
    catch (error) {
        logger.error('Failed to start ProxyManager', {
            operation: 'setupBuiltInProxy',
            error: error
        });
        process.exit(1);
    }
    logger.info('Using Runner Authorization Service', {
        operation: 'setupBuiltInProxy'
    });
}
function setupExternalProxy(options) {
    if (!options.proxy) {
        return;
    }
    let userProxyInfo;
    try {
        userProxyInfo = parseUserProxyOptions(options);
    }
    catch (error) {
        logger.error('Failed to parse proxy URL', {
            operation: 'setupExternalProxy',
            error: error
        });
        process.exit(1);
    }
    if (userProxyInfo) {
        logger.info('Using external proxy configuration', {
            proxyHost: userProxyInfo.host,
            proxyPort: userProxyInfo.port
        });
    }
}
function setupSSLCertificates(options) {
    if (!options.sslExtraCaCerts) {
        return;
    }
    logger.info('Using additional CA certificates', {
        certPath: options.sslExtraCaCerts
    });
}
async function setupMetricsServer(options) {
    if (!options.metrics) {
        return;
    }
    try {
        const metricsPort = parseInt(options.metricsPort, 10);
        metricsServer = new MetricsServer({
            port: metricsPort,
            host: '0.0.0.0',
            runnerId: options.id
        });
        await metricsServer.start();
    }
    catch (error) {
        logger.error('Failed to start metrics server', {
            operation: 'setupMetricsServer',
            error: error
        });
        process.exit(1);
    }
}
function validateProxyOptions(options) {
    if (options.egressProxy && options.proxy) {
        logger.error('Cannot use both --egress-proxy and --proxy together. ' +
            'Use --egress-proxy for built-in MITM proxy or --proxy for external proxy.');
        process.exit(1);
    }
}
async function runRunner({ options, runnerKey }) {
    try {
        const hadRunData = await runLifecycle({
            runnerId: options.id,
            runnerKey: runnerKey,
            disableTelemetry: options.disableTelemetry,
            cliOptions: options,
            workingDir: workingDirectory
        });
        unauthorizedErrorCount = 0;
        return hadRunData;
    }
    catch (error) {
        if (error.statusCode === 401) {
            unauthorizedErrorCount++;
            if (unauthorizedErrorCount >= config.authenticationRetryLimit) {
                logger.error('Authentication failed: Received 3 consecutive 401 errors. ' +
                    'Please verify your runner id, runner key and try again.', {
                    operation: 'runRunner',
                    runnerId: options.id
                });
                await cleanUpAll();
                process.exit(1);
            }
        }
        else {
            unauthorizedErrorCount = 0;
        }
        await sleep(config.agentErrorDelay);
        return false;
    }
}
module.exports = function (program) {
    const runner = program.command('runner')
        .name('runner')
        .addHelpCommand(false)
        .description('Run runners on your own environments for monitoring your APIs');
    runner.command('start')
        .description('Start a runner')
        .requiredOption('--id <id>', 'The ID of the runner to execute')
        .option('--key <key>', 'The secret key for the runner')
        .option('--region <region>', 'Specify the region for the runner. Use "eu" for EU region.')
        .option('--proxy <url>', 'External proxy URL (e.g., http://proxy.com or http://proxy.com:8080)')
        .option('--egress-proxy', 'Enable built-in egress proxy')
        .option('--egress-proxy-authz-url <url>', 'Custom egress proxy authorization service base URL')
        .option('--ssl-extra-ca-certs <path>', 'Additional trusted CA certificates (PEM file, can contain multiple certs)')
        .option('--metrics', 'Enable the metrics server for health checks')
        .option('--metrics-port <port>', 'Port for the metrics server', '9090')
        .addOption(new Option('--report-events', 'Report run events').hideHelp())
        .action(async (options) => {
        const finalRunnerKey = options.key || (await prompts.prompt({
            type: 'password',
            name: 'key',
            message: 'Enter runner key:'
        })).key;
        logger.info('Starting runner', {
            runnerId: options.id
        });
        logger.info('To stop the runner, press Ctrl+C');
        logger.debug('Testing debug log');
        workingDirectory = getWorkingDirectory();
        setupRegion(options.region);
        validateProxyOptions(options);
        await setupBuiltInProxy(options);
        setupExternalProxy(options);
        setupSSLCertificates(options);
        const scheduleNextRun = async () => {
            try {
                const hadRunData = await runRunner({ options: options, runnerKey: finalRunnerKey });
                if (!hadRunData) {
                    setTimeout(scheduleNextRun, config.pingInterval);
                }
                else {
                    cleanUpRun();
                    setImmediate(scheduleNextRun);
                }
            }
            catch (error) {
                logger.error('Error in runner execution', {
                    operation: 'scheduleNextRun',
                    error: error
                });
                setTimeout(scheduleNextRun, config.pingInterval);
            }
        };
        await setupMetricsServer(options);
        scheduleNextRun();
        const gracefulShutdown = async (signal) => {
            logger.info('Graceful shutdown initiated', {
                operation: 'gracefulShutdown',
                signal: signal
            });
            try {
                await cleanUpAll();
                logger.info('Runner stopped successfully!', {
                    operation: 'gracefulShutdown'
                });
                process.exit(0);
            }
            catch (error) {
                logger.error('Error during shutdown', {
                    operation: 'gracefulShutdown',
                    error: error
                });
                process.exit(1);
            }
        };
        process.on('SIGINT', () => {
            gracefulShutdown('SIGINT');
        });
        process.on('SIGHUP', () => {
            gracefulShutdown('SIGHUP');
        });
        process.on('exit', () => {
            gracefulShutdown('exit');
        });
        process.on('SIGTERM', () => {
            gracefulShutdown('SIGTERM');
        });
        process.on('unhandledRejection', async (reason, promise) => {
            logger.error('Unhandled promise rejection', {
                reason: String(reason),
                promise: String(promise)
            });
            try {
                await cleanUpAll();
            }
            catch (cleanupError) {
                logger.error('Error during cleanup after rejection', {
                    cleanupError: String(cleanupError)
                });
            }
            finally {
                process.exit(1);
            }
        });
        process.on('uncaughtException', async (error) => {
            logger.error('Uncaught exception', {
                errorMessage: error.message,
                errorStack: error.stack,
                errorName: error.name
            });
            try {
                await cleanUpAll();
            }
            catch (cleanupError) {
                logger.error('Error during cleanup after exception', {
                    cleanupError: String(cleanupError)
                });
            }
            finally {
                process.exit(1);
            }
        });
    });
};
