// @postman/pm-bin-linux-x64/bin/lib/commands/runner/proxy/cert-manager.js
const fs = require('fs'), https = require('https'), tls = require('tls'), path = require('path'), { getLogger } = require('../logging/logger-factory'), { RUNNER_PROXY_CA_DIR } = require('./constants');
class ProxyCertificateManager {
    constructor() {
        this._logger = getLogger();
        this._originalHttpsAgentCAs = https.globalAgent.options.ca || [];
        this._caCerts = [];
        this._installed = false;
        this._certificatesGenerated = false;
    }
    static getCaCertPath() {
        return path.join(RUNNER_PROXY_CA_DIR, 'certs', 'ca.pem');
    }
    static getCaPrivateKeyPath() {
        return path.join(RUNNER_PROXY_CA_DIR, 'keys', 'ca.private.key');
    }
    static _caCertificatesExist() {
        const certPath = ProxyCertificateManager.getCaCertPath(), keyPath = ProxyCertificateManager.getCaPrivateKeyPath();
        return fs.existsSync(certPath) && fs.existsSync(keyPath);
    }
    _readCaCertificate() {
        try {
            const certPath = ProxyCertificateManager.getCaCertPath();
            if (!fs.existsSync(certPath)) {
                this._logger.debug(`CA certificate not found at: ${certPath}`);
                return null;
            }
            const certContent = fs.readFileSync(certPath, 'utf8');
            this._logger.debug(`Successfully read CA certificate from: ${certPath}`);
            return certContent;
        }
        catch (error) {
            this._logger.error(`Failed to read CA certificate: ${error.message}`);
            return null;
        }
    }
    getCaCert() {
        return this._readCaCertificate();
    }
    getCaKey() {
        try {
            const keyPath = ProxyCertificateManager.getCaPrivateKeyPath();
            if (!fs.existsSync(keyPath)) {
                this._logger.debug(`CA key not found at: ${keyPath}`);
                return null;
            }
            const keyContent = fs.readFileSync(keyPath, 'utf8');
            this._logger.debug(`Successfully read CA key from: ${keyPath}`);
            return keyContent;
        }
        catch (error) {
            this._logger.error(`Failed to read CA key: ${error.message}`);
            return null;
        }
    }
    _installCaCertificate(caCert) {
        if (!this._caCerts.includes(caCert)) {
            this._caCerts.push(caCert);
            this._logger.debug('Added certificate to trusted certificates list');
        }
        https.globalAgent.options.ca = [
            ...tls.rootCertificates,
            ...this._caCerts.map((cert) => { return fs.readFileSync(cert, 'utf8'); })
        ];
    }
    _waitForCertificates(timeout = 10000) {
        const self = this;
        return new Promise((resolve) => {
            const startTime = Date.now(), checkInterval = 100, checkCerts = () => {
                if (ProxyCertificateManager._caCertificatesExist()) {
                    self._certificatesGenerated = true;
                    resolve(true);
                    return;
                }
                if (Date.now() - startTime >= timeout) {
                    self._logger.warn(`Timeout waiting for CA certificates after ${timeout}ms`);
                    resolve(false);
                    return;
                }
                setTimeout(checkCerts, checkInterval);
            };
            checkCerts();
        });
    }
    async install(options = {}) {
        const { timeout = 10000, waitForGeneration = true } = options;
        if (this._installed) {
            this._logger.debug('Certificates already installed');
            return true;
        }
        this._logger.info('Installing certificates...');
        try {
            if (waitForGeneration) {
                this._logger.debug('Waiting for certificates to be generated.');
                const found = await this._waitForCertificates(timeout);
                if (!found) {
                    throw new Error('Certificates were not generated within the timeout period');
                }
            }
            this._installCaCertificate(this.getCertInfo().caCertPath);
            this._installed = true;
            this._logger.info('Certificates installed successfully');
            this._logger.info('HTTPS interception is now enabled');
            return true;
        }
        catch (error) {
            this._logger.error(`Failed to install CA certificates: ${error.message}`);
            return false;
        }
    }
    uninstall() {
        if (!this._installed) {
            this._logger.debug('Certificates not installed, nothing to uninstall');
            return true;
        }
        this._caCerts = [];
        this._installed = false;
        https.globalAgent.options.ca = this._originalHttpsAgentCAs;
        this._logger.info('CA certificates uninstalled successfully');
        return true;
    }
    isInstalled() {
        return this._installed;
    }
    getCertInfo() {
        return {
            certificateCount: this._caCerts.length,
            caCertPath: ProxyCertificateManager.getCaCertPath(),
            caPrivateKeyPath: ProxyCertificateManager.getCaPrivateKeyPath(),
            certificatesExist: ProxyCertificateManager._caCertificatesExist(),
            waitForCertificates: this._waitForCertificates.bind(this)
        };
    }
}
module.exports = ProxyCertificateManager;
