// @postman/pm-bin-linux-x64/bin/lib/commands/runner/proxy/proxy-manager.js
const { Worker } = require('worker_threads'), path = require('path'), fs = require('fs'), ProxyCertificateManager = require('./cert-manager'), { getLogger } = require('../logging/logger-factory');
const { RUNNER_PROXY_HOST, RUNNER_PROXY_PORT } = require('./constants');
let EMBEDDED_WORKER_CONTENT = null;
class RunnerProxyManager {
    constructor() {
        this._worker = null;
        this._isRunning = false;
        this._workerReady = false;
        this._certInstaller = new ProxyCertificateManager();
        this._logger = getLogger();
    }
    _createWorker() {
        return new Promise((resolve, reject) => {
            if (this._worker) {
                resolve();
                return;
            }
            let workerPath;
            if (process.pkg) {
                this._logger.debug('PKG Environment: Using embedded worker content');
                if (!EMBEDDED_WORKER_CONTENT) {
                    const err = new Error('Worker content not embedded. Build may have failed.');
                    this._logger.error(err.message);
                    reject(err);
                    return;
                }
                this._logger.debug(`Worker content available (${EMBEDDED_WORKER_CONTENT.length} bytes)`);
                const os = require('os'), tmpDir = os.tmpdir(), workerTmpPath = path.join(tmpDir, `postman-cli-worker-${Date.now()}.js`);
                try {
                    fs.writeFileSync(workerTmpPath, EMBEDDED_WORKER_CONTENT, 'utf8');
                    this._logger.debug(`Extracted to temp: ${workerTmpPath}`);
                    workerPath = workerTmpPath;
                    const cleanup = () => {
                        try {
                            if (fs.existsSync(workerTmpPath)) {
                                fs.unlinkSync(workerTmpPath);
                            }
                        }
                        catch (err) {
                        }
                    };
                    process.on('exit', cleanup);
                    process.on('SIGINT', cleanup);
                    process.on('SIGTERM', cleanup);
                }
                catch (err) {
                    this._logger.error(`Failed to extract worker: ${err.message}`);
                    reject(err);
                    return;
                }
            }
            else {
                workerPath = path.join(__dirname, 'worker.js');
            }
            this._worker = new Worker(workerPath);
            this._worker.on('message', (message) => {
                this._handleWorkerMessage(message);
            });
            this._worker.on('error', (error) => {
                this._logger.error(`Proxy worker error: ${error.message}`, { trace: error.stack });
                reject(error);
            });
            this._worker.on('exit', (code) => {
                if (code !== 0) {
                    this._logger.error(`Proxy worker exited unexpectedly with exit code: ${code}`, {
                        exitCode: code
                    });
                    this._logger.warn('Proxy worker exited unexpectedly. New requests may fail.');
                }
                else {
                    this._logger.debug('Proxy worker exited cleanly');
                }
                this._worker = null;
                this._isRunning = false;
                this._workerReady = false;
            });
            const readyHandler = (message) => {
                if (message.type === 'ready') {
                    this._workerReady = true;
                    this._worker.off('message', readyHandler);
                    resolve();
                }
            };
            this._worker.on('message', readyHandler);
            setTimeout(() => {
                if (!this._workerReady) {
                    reject(new Error('Worker initialization timeout'));
                }
            }, 10000);
        });
    }
    _handleWorkerMessage(message) {
        const { type, level, message: msg, error, sslCaDir } = message;
        switch (type) {
            case 'ready':
                this._workerReady = true;
                break;
            case 'started':
                this._isRunning = true;
                this._logProxyStarted(sslCaDir);
                this._installCertificates();
                break;
            case 'stopped':
                this._isRunning = false;
                this._logger.info('Runner proxy stopped.');
                break;
            case 'log':
                this._handleLogMessage(level, msg);
                break;
            case 'error':
                this._logger.error(`Runner Proxy error: ${error}`);
                break;
            default:
                break;
        }
    }
    _logProxyStarted(sslCaDir) {
        this._logger.info('Postman Runner Proxy started.');
        this._logger.info('Monitoring HTTP/HTTPS traffic for monitor runs');
        this._logger.info(`Postman Runner Proxy listening on: ${this.getProxyUrl()}`);
        this._logger.debug(`SSL certificates stored in: ${sslCaDir}`);
    }
    _handleLogMessage(level, message) {
        switch (level) {
            case 'success':
                this._logger.info(message);
                break;
            case 'error':
                this._logger.error(message);
                break;
            case 'warn':
                this._logger.warn(message);
                break;
            case 'info':
                this._logger.info(message);
                break;
            case 'debug':
                this._logger.debug(message);
                break;
            default:
                this._logger.info(message);
                break;
        }
    }
    async _installCertificates() {
        try {
            const success = await this._certInstaller.install();
            if (!success) {
                this._logger.warn('Failed to install CA certificates - HTTPS interception may not work properly');
            }
        }
        catch (error) {
            this._logger.warn(`Error installing CA certificates: ${error.message}`);
        }
    }
    async start({ port = RUNNER_PROXY_PORT, authzBaseUrl = null } = {}) {
        if (this._isRunning) {
            return;
        }
        try {
            await this._createWorker();
            return new Promise((resolve, reject) => {
                const messageHandler = (message) => {
                    if (message.type === 'started') {
                        this._worker.off('message', messageHandler);
                        resolve();
                    }
                    else if (message.type === 'error' && message.action === 'start') {
                        this._worker.off('message', messageHandler);
                        reject(new Error(message.error));
                    }
                };
                this._worker.on('message', messageHandler);
                this._worker.postMessage({ action: 'start', port: port, authzBaseUrl: authzBaseUrl });
                setTimeout(() => {
                    this._worker.off('message', messageHandler);
                    reject(new Error('Proxy start timeout'));
                }, 15000);
            });
        }
        catch (error) {
            throw new Error(`Failed to start proxy: ${error.message}`);
        }
    }
    stop() {
        if (!this._isRunning || !this._worker) {
            return Promise.resolve();
        }
        return new Promise((resolve) => {
            const messageHandler = (message) => {
                if (message.type === 'stopped') {
                    this._worker.off('message', messageHandler);
                    this._uninstallCertificates();
                    resolve();
                }
            };
            this._worker.on('message', messageHandler);
            this._worker.postMessage({ action: 'stop' });
            setTimeout(() => {
                this._worker.off('message', messageHandler);
                this._terminateWorker();
                this._uninstallCertificates();
                resolve();
            }, 5000);
        });
    }
    _uninstallCertificates() {
        try {
            this._certInstaller.uninstall();
        }
        catch (error) {
            this._logger.warn(`Error uninstalling CA certificates: ${error.message}`);
        }
    }
    _terminateWorker() {
        if (this._worker) {
            this._worker.terminate();
            this._worker = null;
            this._isRunning = false;
            this._workerReady = false;
            this._uninstallCertificates();
        }
    }
    getProxyConfig() {
        if (!this._isRunning) {
            return null;
        }
        return {
            host: this.getProxyHost(),
            port: this.getProxyPort(),
            protocol: 'http:'
        };
    }
    async retrieveGeneratedCertificateInfo({ generateWaitTimeout = 10000 } = {}) {
        const certInfo = this._certInstaller.getCertInfo(), certsGenerated = await certInfo.waitForCertificates(generateWaitTimeout);
        if (!certsGenerated) {
            throw new Error('Timeout while waiting for proxy certificates to be generated.');
        }
        return certInfo;
    }
    getProxyUrl() {
        if (!this._isRunning) {
            return null;
        }
        return `http://${this.getProxyHost()}:${this.getProxyPort()}`;
    }
    getProxyPort() {
        if (!this._isRunning) {
            return null;
        }
        return RUNNER_PROXY_PORT;
    }
    getProxyHost() {
        if (!this._isRunning) {
            return null;
        }
        return RUNNER_PROXY_HOST;
    }
    async shutdown() {
        await this.stop();
        this._terminateWorker();
    }
}
module.exports = new RunnerProxyManager();
