// @postman/pm-bin-linux-x64/bin/lib/run-upload/helpers/ci-info.js
function getCiInfo(env) {
    let ciInfo = {};
    if (!env) {
        return ciInfo;
    }
    if (env.BITBUCKET_BUILD_NUMBER) {
        ciInfo = {
            branchSource: 'bitbucket',
            branch: env.BITBUCKET_BRANCH,
            pipelineMeta: {
                id: env.BITBUCKET_BUILD_NUMBER,
                source: 'bitbucket',
                url: `${env.BITBUCKET_GIT_HTTP_ORIGIN}/pipelines/results/${env.BITBUCKET_BUILD_NUMBER}`,
                commit: env.BITBUCKET_COMMIT,
                commitUrl: `${env.BITBUCKET_GIT_HTTP_ORIGIN}/commits/${env.BITBUCKET_COMMIT}`
            }
        };
    }
    if (env.GITLAB_CI) {
        ciInfo = {
            branchSource: 'gitlab',
            branch: env.CI_COMMIT_BRANCH,
            pipelineMeta: {
                id: env.CI_PIPELINE_ID,
                source: 'gitlab',
                url: env.CI_PIPELINE_URL,
                commit: env.CI_COMMIT_SHA,
                commitUrl: `${env.CI_PROJECT_URL}/-/commit/${env.CI_COMMIT_SHA}`
            }
        };
    }
    if (env.TRAVIS) {
        ciInfo = {
            branchSource: 'travis',
            branch: env.TRAVIS_BRANCH,
            pipelineMeta: {
                id: env.TRAVIS_BUILD_NUMBER,
                source: 'travis',
                url: env.TRAVIS_BUILD_WEB_URL,
                commit: env.TRAVIS_COMMIT
            }
        };
    }
    if (env.CIRCLECI) {
        ciInfo = {
            branchSource: 'circleci',
            branch: env.CIRCLE_BRANCH,
            pipelineMeta: {
                id: env.CIRCLE_BUILD_NUM,
                source: 'circleci',
                url: env.CIRCLE_BUILD_URL,
                commit: env.CIRCLE_SHA1,
                commitUrl: `${env.CIRCLE_REPOSITORY_URL}/commit/${env.CIRCLE_SHA1}`
            }
        };
    }
    if (env.GITHUB_ACTION) {
        ciInfo = {
            branchSource: 'github',
            branch: env.GITHUB_REF_NAME,
            pipelineMeta: {
                id: env.GITHUB_RUN_NUMBER,
                source: 'github-actions',
                url: `${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}/actions/runs/${env.GITHUB_RUN_ID}`,
                commit: env.GITHUB_SHA,
                commitUrl: `${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}/commit/${env.GITHUB_SHA}`
            }
        };
    }
    if (env.JENKINS_URL) {
        ciInfo = {
            branchSource: 'jenkins',
            branch: env.BRANCH_NAME,
            pipelineMeta: {
                id: env.BUILD_NUMBER,
                url: env.BUILD_URL,
                source: 'jenkins',
                commit: env.GIT_COMMIT,
                commitUrl: `${env.GIT_URL}/commit/${env.GIT_COMMIT}`
            }
        };
    }
    if (env.BUILD_BUILDID) {
        ciInfo = {
            branchSource: 'azure-pipelines',
            branch: env.BUILD_SOURCEBRANCHNAME,
            pipelineMeta: {
                id: env.BUILD_BUILDID,
                url: `${env.SYSTEM_COLLECTIONURI}/${env.SYSTEM_TEAMPROJECTID}/_build/results?buildId=${env.BUILD_BUILDID}`,
                source: 'azure-pipelines',
                commit: env.BUILD_SOURCEVERSION,
                commitUrl: `${env.BUILD_REPOSITORY_URI}/commit/${env.BUILD_SOURCEVERSION}`
            }
        };
    }
    return ciInfo;
}
module.exports = {
    getCiInfo
};
