// @postman/pm-bin-linux-x64/bin/lib/run-upload/helpers/constants.js
module.exports = {
    POSTMAN_CLI_STRING: 'postman-cli',
    RUN_STATUS_FINISHED: 'finished',
    TEST_STATUS_PASS: 'pass',
    TEST_STATUS_FAIL: 'fail',
    TEST_STATUS_SKIPPED: 'skipped',
    FALLBACK_COLLECTION_RUN_NAME: 'Collection Run',
    POSTMAN_API_UPLOAD_PATH: '/collectionruns',
    UPLOAD_AUTHZ_CLIENT_SIDE_ERROR: 'Authorization error occurred while uploading collection run data to Postman. Please ensure you are logged in with a valid API key.\nFor more information, refer: https://go.pstmn.io/postman-cli-faq',
    UPLOAD_AUTHN_CLIENT_SIDE_ERROR: 'Forbidden error occurred while uploading collection run data to Postman. Please ensure you are using a valid API key and have access to add runs for this collection.\nFor more information, refer: https://go.pstmn.io/postman-cli-faq',
    UPLOAD_CLIENT_SIDE_ERROR: 'Error occurred while uploading collection run data to Postman. Please ensure valid postman IDs are being used.\nFor more information, refer: https://go.pstmn.io/postman-cli-faq',
    UPLOAD_SERVER_SIDE_ERROR: 'Error occurred while uploading collection run data to Postman. Please retry.\nFor more information, refer: https://go.pstmn.io/postman-cli-faq',
    RESPONSE_FALLBACK_ERROR_MESSAGE: 'Error occurred while uploading collection run data to Postman. Please retry.',
    V1_API_KEY_REGEX: /^[a-f0-9]{32}$/,
    V2_API_KEY_REGEX: /^(PMAK-[a-f0-9]{24}-[a-f0-9]{34})$/,
    API_KEY_FROM_URL_EXTRACTION_PATTERN: /https?:\/\/api.(?:get)?postman.com\/([a-z]+)s\/([a-z0-9-]+)\?apikey=([a-z0-9A-Z-]+)/
};
