// @postman/pm-bin-linux-x64/bin/lib/run-upload/helpers/run-utils.js
const _ = require('lodash'), uuid = require('uuid'), os = require('os'), { execSync } = require('child_process'), { POSTMAN_CLI_STRING, FALLBACK_COLLECTION_RUN_NAME, RUN_STATUS_FINISHED, TEST_STATUS_PASS, TEST_STATUS_FAIL, TEST_STATUS_SKIPPED } = require('./constants'), { getCiInfo } = require('./ci-info');
function _buildRequestObject(request) {
    if (!request) {
        return {};
    }
    return {
        url: _.invoke(request, 'url.toString', ''),
        method: _.get(request, 'method', ''),
        headers: null,
        body: null
    };
}
function _buildResponseObject(response) {
    if (!response) {
        return {};
    }
    return {
        code: response.code,
        name: response.status,
        time: response.responseTime,
        size: response.responseSize,
        headers: null,
        body: null
    };
}
function _buildTestObject(assertions) {
    const tests = [];
    assertions && assertions.forEach((assert) => {
        let status;
        if (assert.skipped) {
            status = TEST_STATUS_SKIPPED;
        }
        else if (assert.error) {
            status = TEST_STATUS_FAIL;
        }
        else {
            status = TEST_STATUS_PASS;
        }
        tests.push({
            name: assert.assertion,
            error: assert.error ? _.pick(assert.error, ['name', 'message', 'stack']) : null,
            status: status
        });
    });
    return tests;
}
function _extractSkippedTestCountFromRun(runSummary) {
    let skippedTestCount = 0;
    _.forEach(_.get(runSummary, 'run.executions', []), (execution) => {
        _.forEach(_.get(execution, 'assertions', []), (assertion) => {
            if (_.get(assertion, 'skipped')) {
                skippedTestCount++;
            }
        });
    });
    return skippedTestCount;
}
function _executionToIterationConverter(executions, iterationCount) {
    const iterations = [], validIterationCount = _.isSafeInteger(iterationCount) && iterationCount > 0;
    if (!validIterationCount) {
        executions = [executions];
    }
    else {
        executions = _.chunk(executions, (executions.length / iterationCount));
    }
    _.forEach(executions, (iter) => {
        const iteration = [];
        _.forEach(iter, (req) => {
            iteration.push({
                id: req.item.id,
                name: req.item.name || '',
                request: _buildRequestObject(req.request),
                response: _buildResponseObject(req.response),
                error: req.requestError || null,
                tests: _buildTestObject(req.assertions)
            });
        });
        iterations.push(iteration);
    });
    return iterations;
}
function _populateCiRunInformation(collectionRunObj) {
    let env = process.env, ciRunInfo = {};
    try {
        ciRunInfo = getCiInfo(env);
    }
    catch (e) {
        ciRunInfo = {};
    }
    return { ...collectionRunObj, ...ciRunInfo };
}
function _populateLocalRunInformation(collectionRunObj) {
    let localRunInfo = {
        branchSource: 'local'
    };
    try {
        const redirect = os.platform() === 'win32' ? '2>NUL' : '2>/dev/null';
        localRunInfo.branch = execSync(`git rev-parse --abbrev-ref HEAD ${redirect}`).toString().trim();
    }
    catch (e) {
        logger.error('Failed to get git branch');
    }
    return { ...collectionRunObj, ...localRunInfo };
}
function buildCollectionRunObject(collectionRunOptions, runSummary) {
    if (!collectionRunOptions || !runSummary) {
        throw new Error('Cannot build Collection run object without collectionRunOptions or runSummary');
    }
    let failedTestCount = _.get(runSummary, 'run.stats.assertions.failed', 0), skippedTestCount = _extractSkippedTestCountFromRun(runSummary), totalTestCount = _.get(runSummary, 'run.stats.assertions.total', 0), executions = _.get(runSummary, 'run.executions'), iterationCount = _.get(runSummary, 'run.stats.iterations.total', 1), totalRequests = _.get(runSummary, 'run.stats.responses.total', 0), collectionRunObj = {
        id: uuid.v4(),
        collection: _.get(collectionRunOptions, 'collection.id'),
        environment: _.get(collectionRunOptions, 'environment.id'),
        folder: _.get(collectionRunOptions, 'folder.id'),
        name: _.get(collectionRunOptions, 'collection.name', FALLBACK_COLLECTION_RUN_NAME),
        status: RUN_STATUS_FINISHED,
        source: POSTMAN_CLI_STRING,
        delay: collectionRunOptions.delayRequest || 0,
        currentIteration: iterationCount,
        failedTestCount: failedTestCount,
        skippedTestCount: skippedTestCount,
        passedTestCount: (totalTestCount - (failedTestCount + skippedTestCount)),
        totalTestCount: totalTestCount,
        iterations: _executionToIterationConverter(executions, iterationCount),
        totalTime: _.get(runSummary, 'run.stats.responses.totalResponseTime', 0),
        totalRequests: totalRequests,
        startedAt: _.get(runSummary, 'run.timings.started'),
        createdAt: _.get(runSummary, 'run.timings.completed')
    };
    const isCI = require('is-ci');
    if (isCI) {
        collectionRunObj = _populateCiRunInformation(collectionRunObj);
    }
    else {
        collectionRunObj = _populateLocalRunInformation(collectionRunObj);
    }
    collectionRunObj = _.omitBy(collectionRunObj, _.isNil);
    return collectionRunObj;
}
function buildRunOverviewObject(collectionRunOptions, runSummary) {
    if (!runSummary) {
        throw new Error('Cannot build Run Overview object without runSummary');
    }
    const runOverviewObj = {
        collectionName: _.get(collectionRunOptions, 'collection.name', FALLBACK_COLLECTION_RUN_NAME),
        runDurationInMiliseconds: _.get(runSummary, 'run.timings.completed') - _.get(runSummary, 'run.timings.started'),
        averageResponseTimeInMiliseconds: _.get(runSummary, 'run.timings.responseAverage'),
        totalDataReceivedInBytes: _.get(runSummary, 'run.transfers.responseTotal'),
        statistics: _.get(runSummary, 'run.stats')
    };
    return runOverviewObj;
}
module.exports = {
    buildCollectionRunObject,
    buildRunOverviewObject
};
