// @postman/pm-bin-linux-x64/bin/lib/run-upload/helpers/upload-run.js
const _ = require('lodash'), print = require('../../framework/print'), request = require('postman-request'), { POSTMAN_API_UPLOAD_PATH, RESPONSE_FALLBACK_ERROR_MESSAGE, UPLOAD_SERVER_SIDE_ERROR, UPLOAD_AUTHZ_CLIENT_SIDE_ERROR, UPLOAD_AUTHN_CLIENT_SIDE_ERROR, UPLOAD_CLIENT_SIDE_ERROR } = require('./constants'), util = require('../../../lib/util'), { buildCollectionRunObject, buildRunOverviewObject } = require('./run-utils');
function uploadRun(postmanApiKey, accessToken, collectionRunOptions, runSummary, callback) {
    let collectionRunObj, runOverviewObj, requestConfig, integrationId, url = util.POSTMAN_GATEWAY_BASE_URL() + POSTMAN_API_UPLOAD_PATH, qs = new URLSearchParams();
    if (!runSummary) {
        return callback(new Error('runSummary is a required parameter to upload run data'));
    }
    try {
        collectionRunObj = buildCollectionRunObject(collectionRunOptions, runSummary);
        runOverviewObj = buildRunOverviewObject(collectionRunOptions, runSummary);
    }
    catch (error) {
        return callback(new Error('Failed to serialize the run for upload. Please try again.'));
    }
    if (collectionRunOptions.integrationId) {
        integrationId = collectionRunOptions.integrationId;
        qs.append('integrationIdentifier', integrationId);
    }
    if (collectionRunOptions.workspace) {
        qs.append('workspace', collectionRunOptions.workspace);
    }
    if (collectionRunOptions.apiId) {
        qs.append('apiId', collectionRunOptions.apiId);
    }
    requestConfig = {
        url: url = url + '?' + qs.toString(),
        body: JSON.stringify({
            collectionRun: collectionRunObj,
            runOverview: runOverviewObj
        }),
        headers: {
            'content-type': 'application/json',
            accept: 'application/vnd.postman.v2+json',
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'postman-cli-data-ingestor'
        }
    };
    if (postmanApiKey) {
        requestConfig.headers['x-api-key'] = postmanApiKey;
    }
    else if (accessToken) {
        requestConfig.headers['x-access-token'] = accessToken;
    }
    return request.post(requestConfig, (error, response, body) => {
        if (error) {
            return callback(new Error(_.get(error, 'message', RESPONSE_FALLBACK_ERROR_MESSAGE)));
        }
        if (collectionRunOptions.verbose) {
            print.lf();
            print.lf('Response received from postman run publish API');
            print.lf(body);
        }
        if (_.inRange(response.statusCode, 200, 300)) {
            return callback(null, JSON.parse(body));
        }
        else if (response.statusCode === 401) {
            return callback(new Error(UPLOAD_AUTHZ_CLIENT_SIDE_ERROR));
        }
        else if (response.statusCode === 403) {
            return callback(new Error(UPLOAD_AUTHN_CLIENT_SIDE_ERROR));
        }
        else if (_.inRange(response.statusCode, 400, 500)) {
            return callback(new Error(UPLOAD_CLIENT_SIDE_ERROR));
        }
        return callback(new Error(UPLOAD_SERVER_SIDE_ERROR));
    });
}
module.exports = {
    uploadRun
};
