// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/config-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigFile = getConfigFile;
exports.getConfigFilePath = getConfigFilePath;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const node_os_1 = require("node:os");
const lodash_1 = __importDefault(require("lodash"));
const axios_1 = __importDefault(require("axios"));
const ajv_1 = __importDefault(require("ajv"));
const better_ajv_errors_1 = __importDefault(require("better-ajv-errors"));
const liblab_config_schema_1 = __importDefault(require("./schemas/liblab-config.schema"));
const sdk_utils_1 = require("./sdk-utils");
const POSTMAN_CONFIG_FILE_NAME = 'config.json';
const URL_REGEX = /^https?:\/\/[^\s#$./?].\S*$/i;
function isUrl(path) {
    return URL_REGEX.test(path);
}
async function readConfigFile(filePath) {
    if (isUrl(filePath)) {
        (0, sdk_utils_1.validateUrlProtocol)(filePath);
        try {
            const { data } = await axios_1.default.get(filePath, {
                responseType: 'text',
                timeout: 20000
            });
            return data;
        }
        catch (err) {
            throw new Error(`Failed to fetch config from URL: ${err.message}`);
        }
    }
    if (filePath.startsWith('~')) {
        const homeDir = (0, node_os_1.homedir)();
        filePath = node_path_1.default.join(homeDir, filePath.slice(1));
    }
    if (!node_path_1.default.isAbsolute(filePath)) {
        filePath = node_path_1.default.join(process.cwd(), filePath);
    }
    try {
        return node_fs_1.default.readFileSync(filePath, 'utf8');
    }
    catch (err) {
        if (err.code === 'ENOENT') {
            throw new Error(`Config file not found: ${filePath}`);
        }
        throw new Error(`Failed to read config file: ${err.message}`);
    }
}
function discoverConfigFile() {
    const cwd = process.cwd();
    const postmanDir = node_path_1.default.join(cwd, '.postman');
    const postmanConfigPath = node_path_1.default.join(postmanDir, POSTMAN_CONFIG_FILE_NAME);
    if (node_fs_1.default.existsSync(postmanConfigPath)) {
        return postmanConfigPath;
    }
    const cwdLiblabConfigPath = node_path_1.default.join(cwd, 'liblab.config.json');
    if (node_fs_1.default.existsSync(cwdLiblabConfigPath)) {
        return cwdLiblabConfigPath;
    }
    return null;
}
function validateConfig(config) {
    const ajv = new ajv_1.default({
        allErrors: true,
        verbose: true,
        strict: false
    });
    const validationSchema = {
        ...liblab_config_schema_1.default,
        required: [],
        additionalProperties: true
    };
    const validate = ajv.compile(validationSchema);
    const valid = validate(config);
    if (!valid) {
        try {
            const betterErrors = (0, better_ajv_errors_1.default)(validationSchema, config, validate.errors || [], {
                format: 'cli',
                indent: 2
            });
            return {
                valid: false,
                errors: [String(betterErrors)]
            };
        }
        catch {
            const errorMessages = (validate.errors || []).map((error) => {
                return `${error.instancePath || 'root'} ${error.message}`;
            });
            return {
                valid: false,
                errors: errorMessages
            };
        }
    }
    return { valid: true };
}
function extractSdkConfig(filePath, parsed) {
    const normalized = filePath.replace(/\\/g, '/');
    if (normalized.endsWith('/.postman/config.json') || (parsed && typeof parsed === 'object' && 'sdk' in parsed)) {
        return parsed?.sdk ?? null;
    }
    return parsed ?? null;
}
function stripLegacyKeys(value) {
    const legacyKeys = new Set(['liblabVersion']);
    if (lodash_1.default.isNil(value)) {
        return value;
    }
    if (Array.isArray(value)) {
        return value.map((v) => { return stripLegacyKeys(v); });
    }
    if (typeof value === 'object') {
        const out = {};
        Object.keys(value).forEach((k) => {
            if (legacyKeys.has(k)) {
                return;
            }
            out[k] = stripLegacyKeys(value[k]);
        });
        return out;
    }
    return value;
}
function pruneEmptyValues(value) {
    if (lodash_1.default.isNil(value)) {
        return undefined;
    }
    if (typeof value === 'string') {
        return value.trim() ? value : undefined;
    }
    if (Array.isArray(value)) {
        const next = value
            .map((v) => { return pruneEmptyValues(v); })
            .filter((v) => { return v !== undefined; });
        return next;
    }
    if (typeof value === 'object') {
        const out = {};
        Object.keys(value).forEach((k) => {
            const v = pruneEmptyValues(value[k]);
            if (v !== undefined) {
                out[k] = v;
            }
        });
        return Object.keys(out).length ? out : undefined;
    }
    return value;
}
function parseConfig(content) {
    try {
        return JSON.parse(content);
    }
    catch (err) {
        throw new Error(`Failed to parse config file: ${err.message}`);
    }
}
async function getConfigFile(configPath) {
    let filePath;
    if (configPath) {
        filePath = configPath;
    }
    else {
        const discovered = discoverConfigFile();
        if (!discovered) {
            return null;
        }
        filePath = discovered;
    }
    const content = await readConfigFile(filePath);
    const parsed = parseConfig(content);
    const config = extractSdkConfig(filePath, parsed);
    if (!config) {
        return null;
    }
    const stripped = stripLegacyKeys(config);
    const pruned = pruneEmptyValues(stripped);
    if (!pruned) {
        return null;
    }
    const validation = validateConfig(pruned);
    if (!validation.valid) {
        throw new Error(`Config file validation failed:\n${validation.errors?.join('\n')}`);
    }
    return pruned;
}
function getConfigFilePath(configPath) {
    if (configPath) {
        return configPath;
    }
    return discoverConfigFile();
}
