// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/connect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerConnectCommand = registerConnectCommand;
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const waterfall_1 = __importDefault(require("async/waterfall"));
const util_1 = __importDefault(require("../../postman-entities/util"));
const constants_1 = require("../../constants");
const sdk_utils_1 = require("./sdk-utils");
const integrations_service_1 = require("../../api/integrations-service");
const postman_resource_1 = __importDefault(require("../../api/postman-resource"));
const request_1 = __importDefault(require("@postman/platform-req/lib/request"));
const liquid_json_1 = __importDefault(require("liquid-json"));
const util_2 = __importDefault(require("../../util"));
const prompts_1 = __importDefault(require("prompts"));
const open_1 = __importDefault(require("open"));
function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
function createFuzzySearcher(choices) {
    return (input) => {
        if (!input || input.trim().length === 0) {
            return Promise.resolve(choices);
        }
        const searchTerm = input.toLowerCase().trim();
        const filtered = choices.filter((choice) => {
            const titleMatch = choice.title.toLowerCase().includes(searchTerm);
            const descriptionMatch = choice.description && choice.description.toLowerCase().includes(searchTerm);
            return titleMatch || descriptionMatch;
        });
        return Promise.resolve(filtered);
    };
}
function getAllWorkspaces(options) {
    return new Promise((resolve, reject) => {
        const url = `${util_2.default.POSTMAN_GATEWAY_BASE_URL()}/workspaces?archived=false&recentActivity=true&limit=10`;
        const headers = {
            'User-Agent': util_2.default.userAgent,
            'x-pstmn-req-service': 'workspaces',
            'Content-Type': 'application/json'
        };
        if (options.postmanApiKey) {
            headers['X-Api-Key'] = options.postmanApiKey;
        }
        else if (options.accessToken) {
            headers['x-access-token'] = options.accessToken;
        }
        else {
            reject(new Error('No authorization data found. Please use the `postman login` command.'));
            return;
        }
        const client = request_1.default.defaults({
            json: true,
            headers: headers
        });
        client.get({
            url
        }, (err, response, body) => {
            if (err) {
                return reject(new Error(`Failed to fetch workspaces: ${err.message || err}`));
            }
            try {
                if (typeof body === 'string') {
                    body = liquid_json_1.default.parse(body.trim());
                }
            }
            catch {
                return reject(new Error('Invalid JSON response from API'));
            }
            if (!(/2../).test(response.statusCode)) {
                const errorMessage = body?.error?.message ||
                    body?.message ||
                    `HTTP ${response.statusCode}`;
                return reject(new Error(`Failed to fetch workspaces: ${errorMessage}`));
            }
            let workspaces = body?.data?.workspaces ||
                body?.data ||
                body?.workspaces;
            if (workspaces && !Array.isArray(workspaces)) {
                if (workspaces.id || workspaces.workspaceId) {
                    workspaces = [workspaces];
                }
                else {
                    workspaces = Array.isArray(body) ? body : [];
                }
            }
            if (!workspaces || !Array.isArray(workspaces)) {
                workspaces = [];
            }
            const ownedWorkspaces = workspaces.filter((workspace) => {
                if (workspace.type) {
                    return workspace.type === 'personal' || workspace.type === 'team';
                }
                if (workspace.role) {
                    return workspace.role === 'owner' || workspace.role === 'admin';
                }
                return true;
            });
            return resolve(ownedWorkspaces.length > 0 ? ownedWorkspaces : workspaces);
        });
    });
}
function getWorkspaceIdFromId(id, options) {
    return new Promise((resolve, reject) => {
        (0, sdk_utils_1.identifyIdType)(id, options, (err, idType, identifiedId) => {
            if (err || !idType || idType === 'unknown' || !identifiedId) {
                return reject(new Error(`ID "${id}" not found as a collection or specification. ` +
                    'Please verify the ID and your access permissions.'));
            }
            getAllWorkspaces(options)
                .then((workspaces) => {
                if (workspaces.length === 0) {
                    reject(new Error('No workspaces found for this user.'));
                    return Promise.resolve([]);
                }
                const searchPromises = workspaces.map((workspace) => {
                    return new Promise((workspaceResolve) => {
                        const workspaceId = workspace.id || workspace.workspaceId;
                        if (!workspaceId) {
                            workspaceResolve(null);
                            return;
                        }
                        if (idType === 'collection') {
                            postman_resource_1.default.getWorkspaceCollections(workspaceId, options.postmanApiKey, options.accessToken, (err, collections) => {
                                if (err) {
                                    workspaceResolve(null);
                                    return;
                                }
                                const found = collections.some((collection) => {
                                    if (typeof collection === 'string') {
                                        return collection === identifiedId;
                                    }
                                    const collId = collection.id || collection.uid || collection._postman_id;
                                    return collId === identifiedId;
                                });
                                workspaceResolve(found ? workspaceId : null);
                            });
                        }
                        else {
                            postman_resource_1.default.getWorkspaceSpecifications(workspaceId, options.postmanApiKey, options.accessToken, (err, specifications) => {
                                if (err) {
                                    workspaceResolve(null);
                                    return;
                                }
                                const found = specifications.some((spec) => {
                                    if (typeof spec === 'string') {
                                        return spec === identifiedId;
                                    }
                                    const specId = spec.id || spec.specificationId;
                                    return specId === identifiedId;
                                });
                                workspaceResolve(found ? workspaceId : null);
                            });
                        }
                    });
                });
                return Promise.all(searchPromises);
            })
                .then((results) => {
                const workspaceId = results.find((result) => {
                    return result !== null;
                });
                if (!workspaceId) {
                    const entityType = idType === 'collection' ? 'Collection' : 'Specification';
                    reject(new Error(`${entityType} ${id} not found in any of your workspaces. ` +
                        'Please provide the workspace ID manually using the --workspace option.'));
                    return;
                }
                resolve({ workspaceId, idType });
            })
                .catch((err) => {
                reject(err);
            });
        });
    });
}
function registerConnectCommand(sdk) {
    sdk.command('connect <provider> <id>')
        .description('Connect a collection or specification to a Git repository for automated SDK PRs')
        .usage('<provider> <id> [options]')
        .option('-l, --language <language>', 'SDK language (required) [choices: typescript, python, java]', (value) => {
        if (!(0, sdk_utils_1.validateLanguage)(value)) {
            throw new Error(`Invalid language: "${value}". ` +
                `Must be one of: ${sdk_utils_1.SUPPORTED_LANGUAGES.join(', ')}`);
        }
        return value;
    })
        .option('-w, --workspace <workspaceId>', 'Workspace ID (optional)')
        .addHelpText('after', '\nProviders:\n' +
        '  github - Connect to GitHub repository\n\n' +
        'Note: Currently only collections are supported for GitHub integrations. ' +
        'Specification support is being worked on and may be available in a future release.\n\n' +
        'Examples:\n' +
        '  Connect a collection for TypeScript SDK output:\n' +
        '    postman sdk connect github 12345-abcde-67890 --language typescript\n\n' +
        '  Connect with specific workspace:\n' +
        '    postman sdk connect github 12345-abcde-67890 -l python -w my-workspace-id\n')
        .action((provider, id, command) => {
        let exitCode = 0;
        if (provider !== 'github') {
            print_1.default.lf(safe_1.default.red(`Error: Unsupported provider "${provider}". Currently only "github" is supported.`));
            process.exit(1);
        }
        if (!command.language) {
            print_1.default.lf(safe_1.default.red('Error: Language is required. Use --language or -l flag.'));
            print_1.default.lf(`Supported languages: ${sdk_utils_1.SUPPORTED_LANGUAGES.join(', ')}`);
            process.exit(1);
        }
        (0, waterfall_1.default)([
            (next) => {
                util_1.default.addApiKeyToOptions(command, {}, (err, options) => {
                    if (err) {
                        return next(err);
                    }
                    if (!options?.postmanApiKey && !options?.accessToken) {
                        return next(new Error(constants_1.ERROR_MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE));
                    }
                    next(null, options);
                });
            },
            (options, next) => {
                (0, sdk_utils_1.identifyIdType)(id, options, (err, idType, identifiedId) => {
                    if (err) {
                        return next(new Error(`ID identification failed: ${err.message}`));
                    }
                    if (!idType || idType === 'unknown' || !identifiedId) {
                        return next(new Error('ID identification failed. ' +
                            'Please check the ID and your access permissions.'));
                    }
                    if (idType === 'specification') {
                        return next(new Error('Specifications are not currently supported for GitHub integrations. ' +
                            'Please use a collection ID instead. ' +
                            'Support for specifications is being worked on and may be available in a future release.'));
                    }
                    next(null, { ...options, idType, identifiedId });
                });
            },
            (options, next) => {
                const integrationsService = (0, integrations_service_1.createIntegrationsService)({
                    accessToken: options.accessToken,
                    postmanApiKey: options.postmanApiKey
                });
                const pollForToken = (sessionId) => {
                    return new Promise((resolve, reject) => {
                        let attempts = 0;
                        const maxAttempts = 60;
                        const poll = () => {
                            if (attempts >= maxAttempts) {
                                reject(new Error('Authentication timeout. Please try again.'));
                                return;
                            }
                            sleep(5000).then(() => {
                                attempts++;
                                integrationsService.getOAuthSession(sessionId)
                                    .then((sessionStatus) => {
                                    if (sessionStatus.token && sessionStatus.tokenId && sessionStatus.token.accessToken) {
                                        print_1.default.lf(safe_1.default.green('✓ Authentication successful!'));
                                        resolve({
                                            tokenId: sessionStatus.tokenId,
                                            accessToken: sessionStatus.token.accessToken
                                        });
                                    }
                                    else {
                                        if (command.verbose && attempts % 6 === 0) {
                                            print_1.default.lf(safe_1.default.gray(`Still waiting... (${attempts * 5}s)`));
                                        }
                                        poll();
                                    }
                                })
                                    .catch((err) => {
                                    reject(err);
                                });
                            });
                        };
                        poll();
                    });
                };
                if (command.verbose) {
                    print_1.default.lf('Creating OAuth session...');
                }
                const INTEGRATION_TYPE = 'github';
                integrationsService.createOAuthSession(INTEGRATION_TYPE, 'integration_oauth')
                    .then((session) => {
                    print_1.default.lf(safe_1.default.cyan('\nOpening browser for GitHub authentication...'));
                    const integrationsServiceBaseUrl = new URL(session.redirectUrl).origin;
                    const oauthUrl = `${integrationsServiceBaseUrl}/oauth/authorize/${INTEGRATION_TYPE}?sessionId=${session.sessionId}&client_id=${session.clientId}&state=${session.state}&scope=${session.scope}&redirect_uri=${encodeURIComponent(session.redirectUrl)}`;
                    print_1.default.lf(safe_1.default.gray(`If the browser doesn't open, visit: ${oauthUrl}`));
                    return (0, open_1.default)(oauthUrl)
                        .then(() => {
                        print_1.default.lf(safe_1.default.yellow('Waiting for authentication...'));
                        return pollForToken(session.sessionId);
                    })
                        .then(({ tokenId, accessToken }) => {
                        return Promise.all([
                            integrationsService.getGitUser(tokenId),
                            integrationsService.getGitOrgs(tokenId)
                        ]).then(([user, orgs]) => {
                            return { tokenId, accessToken, user, orgs, session };
                        });
                    });
                })
                    .then(({ tokenId, accessToken, user, orgs, session }) => {
                    const ownerChoices = [
                        { title: `${user.name} (${user.username})`, value: user.username }
                    ].concat(orgs.map((org) => {
                        return {
                            title: org.name,
                            value: org.name
                        };
                    }));
                    return (0, prompts_1.default)({
                        type: 'autocomplete',
                        name: 'owner',
                        message: 'Select GitHub Organization/User (type to search)',
                        choices: ownerChoices,
                        suggest: createFuzzySearcher(ownerChoices)
                    }).then((ownerResponse) => {
                        if (!ownerResponse.owner) {
                            throw new Error('USER_CANCELLED');
                        }
                        return {
                            tokenId: tokenId,
                            accessToken: accessToken,
                            owner: ownerResponse.owner,
                            session: session
                        };
                    });
                })
                    .then(({ tokenId, accessToken, owner, session }) => {
                    print_1.default.lf(safe_1.default.cyan('Fetching repositories...'));
                    return integrationsService.getGitRepositories(tokenId, owner)
                        .then((repositories) => {
                        if (repositories.length === 0) {
                            throw new Error(`No repositories found for ${owner}`);
                        }
                        const repoChoices = repositories.map((repo) => {
                            return {
                                title: repo.name,
                                value: repo.name,
                                description: repo.full_name
                            };
                        });
                        return (0, prompts_1.default)({
                            type: 'autocomplete',
                            name: 'repository',
                            message: 'Select Repository (type to search)',
                            choices: repoChoices,
                            suggest: createFuzzySearcher(repoChoices)
                        }).then((repoResponse) => {
                            if (!repoResponse.repository) {
                                throw new Error('USER_CANCELLED');
                            }
                            return {
                                tokenId: tokenId,
                                accessToken: accessToken,
                                owner: owner,
                                repository: repoResponse.repository,
                                session: session
                            };
                        });
                    });
                })
                    .then(({ tokenId, accessToken, owner, repository, session }) => {
                    print_1.default.lf(safe_1.default.cyan('Fetching branches...'));
                    return integrationsService.getGitBranches(tokenId, owner, repository)
                        .then((branches) => {
                        if (branches.length === 0) {
                            throw new Error(`No branches found for ${owner}/${repository}`);
                        }
                        const branchChoices = branches.map((branch) => {
                            return {
                                title: branch.name,
                                value: branch.name
                            };
                        });
                        return (0, prompts_1.default)({
                            type: 'autocomplete',
                            name: 'branch',
                            message: 'Select Branch (type to search)',
                            choices: branchChoices,
                            suggest: createFuzzySearcher(branchChoices)
                        }).then((branchResponse) => {
                            if (!branchResponse.branch) {
                                throw new Error('USER_CANCELLED');
                            }
                            return {
                                tokenId: tokenId,
                                accessToken: accessToken,
                                owner: owner,
                                repository: repository,
                                branch: branchResponse.branch,
                                session: session
                            };
                        });
                    });
                })
                    .then(({ tokenId, accessToken, owner, repository, branch, session }) => {
                    return (0, prompts_1.default)({
                        type: 'text',
                        name: 'directory',
                        message: 'SDK directory path (ignore if not using a monorepo):',
                        initial: './',
                        validate: (value) => {
                            if (!value || value.trim().length === 0) {
                                return 'Directory path is required';
                            }
                            return true;
                        }
                    }).then((directoryResponse) => {
                        if (!directoryResponse.directory) {
                            throw new Error('USER_CANCELLED');
                        }
                        const directory = directoryResponse.directory.trim().endsWith('/') ?
                            directoryResponse.directory.trim() :
                            `${directoryResponse.directory.trim()}/`;
                        return {
                            tokenId: tokenId,
                            accessToken: accessToken,
                            owner: owner,
                            repository: repository,
                            branch: branch,
                            directory: directory,
                            session: session
                        };
                    });
                })
                    .then(({ tokenId, accessToken, owner, repository, branch, directory, session }) => {
                    let workspaceId = command.workspace;
                    if (!workspaceId) {
                        print_1.default.lf(safe_1.default.cyan('Fetching workspace ID from collection or specification...'));
                        return getWorkspaceIdFromId(id, {
                            postmanApiKey: options.postmanApiKey,
                            accessToken: options.accessToken
                        })
                            .then(({ workspaceId: fetchedWorkspaceId, idType }) => {
                            const entityType = idType === 'collection' ? 'collection' : 'specification';
                            print_1.default.lf(safe_1.default.green(`✓ Found workspace ID: ${fetchedWorkspaceId} (${entityType})`));
                            return {
                                tokenId: tokenId,
                                accessToken: accessToken,
                                owner: owner,
                                repository: repository,
                                branch: branch,
                                directory: directory,
                                workspaceId: fetchedWorkspaceId,
                                idType: idType,
                                session: session
                            };
                        })
                            .catch((err) => {
                            print_1.default.lf(safe_1.default.yellow(`⚠ ${err.message}`));
                            print_1.default.lf(safe_1.default.cyan('Please enter the workspace ID manually:'));
                            return (0, prompts_1.default)({
                                type: 'text',
                                name: 'workspaceId',
                                message: 'Enter Workspace ID (required)',
                                validate: (value) => {
                                    if (!value || value.trim().length === 0) {
                                        return 'Workspace ID is required';
                                    }
                                    return true;
                                }
                            }).then((workspaceResponse) => {
                                if (!workspaceResponse.workspaceId) {
                                    throw new Error('USER_CANCELLED');
                                }
                                return {
                                    tokenId: tokenId,
                                    accessToken: accessToken,
                                    owner: owner,
                                    repository: repository,
                                    branch: branch,
                                    directory: directory,
                                    workspaceId: workspaceResponse.workspaceId,
                                    session: session
                                };
                            });
                        });
                    }
                    return Promise.resolve({
                        tokenId: tokenId,
                        accessToken: accessToken,
                        owner: owner,
                        repository: repository,
                        branch: branch,
                        directory: directory,
                        workspaceId: workspaceId,
                        idType: 'collection',
                        session: session
                    });
                })
                    .then(({ accessToken, owner, repository, branch, directory, workspaceId, idType }) => {
                    print_1.default.lf(safe_1.default.cyan('Checking for existing integration...'));
                    if (!accessToken) {
                        throw new Error('OAuth token not available. Please try again.');
                    }
                    return integrationsService.getIntegrations(id, command.language, idType || 'collection')
                        .then((existingIntegrations) => {
                        if (existingIntegrations.length > 0) {
                            print_1.default.lf(safe_1.default.yellow(`Found ${existingIntegrations.length} existing integration(s) with the same collection and language. Deleting...`));
                            const deletePromises = existingIntegrations.map((integration) => {
                                return integrationsService.deleteIntegration(String(integration.id))
                                    .then(() => {
                                    print_1.default.lf(safe_1.default.gray(`  Deleted integration: ${integration.id}`));
                                })
                                    .catch((err) => {
                                    print_1.default.lf(safe_1.default.yellow(`  Warning: Failed to delete integration ${integration.id}: ${err.message}`));
                                });
                            });
                            return Promise.all(deletePromises).then(() => {
                                return Promise.resolve();
                            });
                        }
                        return Promise.resolve();
                    })
                        .then(() => {
                        print_1.default.lf(safe_1.default.cyan('Creating integration...'));
                        const integrationName = `My collection ${id} ${command.language} sdk integration`;
                        return integrationsService.createIntegration(workspaceId, {
                            name: integrationName,
                            rule: integrations_service_1.INTEGRATION_RULE,
                            entity: 'collection',
                            executor: id,
                            variables: {
                                access_token: accessToken,
                                repository: repository,
                                branch: branch,
                                owner: owner,
                                directory: directory,
                                language: command.language
                            }
                        }).then((integration) => {
                            print_1.default.lf(safe_1.default.green('\n✓ Integration created successfully!'));
                            print_1.default.lf(safe_1.default.gray(`Integration ID: ${integration.id}`));
                            print_1.default.lf(safe_1.default.gray(`Repository: ${owner}/${repository}`));
                            print_1.default.lf(safe_1.default.gray(`Branch: ${branch}`));
                            print_1.default.lf(safe_1.default.gray(`Language: ${command.language}`));
                            next(null, { integration });
                        });
                    });
                })
                    .catch((err) => {
                    next(err);
                });
            }
        ], (err) => {
            if (err) {
                if (err.message === 'USER_CANCELLED') {
                    process.exit(0);
                }
                if (err.message && err.message.includes('No authorization data')) {
                    print_1.default.lf(safe_1.default.red(`\nError: ${err.message}`));
                    print_1.default.lf(safe_1.default.yellow('\nPlease login first using: postman login'));
                }
                else {
                    print_1.default.lf(safe_1.default.red(`\nError: ${err.message || err}`));
                }
                exitCode = 1;
            }
            process.exit(exitCode);
        });
    });
}
