// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/generate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatBuildError = formatBuildError;
exports.registerGenerateCommand = registerGenerateCommand;
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const waterfall_1 = __importDefault(require("async/waterfall"));
const util_1 = __importDefault(require("../../postman-entities/util"));
const constants_1 = require("../../constants");
const prompts_1 = __importDefault(require("prompts"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const semver_1 = __importDefault(require("semver"));
const sdk_utils_1 = require("./sdk-utils");
const v3_filesystem_collection_1 = require("./v3-filesystem-collection");
const sdk_generator_service_1 = require("../../api/sdk-generator-service");
const pr_workflow_1 = require("./pr-workflow");
const config_service_1 = require("./config-service");
const integrations_service_1 = require("../../api/integrations-service");
const DEFAULT_SDK_VERSION = '1.0.0';
function sanitizeSdkName(name) {
    const sanitized = name.replace(/[^a-zA-Z0-9 ]/g, ' ').replace(/\s+/g, ' ').trim();
    return sanitized || 'SDK';
}
function deriveNameFromPath(filePath) {
    const resolved = path_1.default.resolve(filePath);
    try {
        if (fs_1.default.statSync(resolved).isDirectory()) {
            return path_1.default.basename(resolved);
        }
    }
    catch {
    }
    return path_1.default.basename(resolved, path_1.default.extname(resolved));
}
function validateSDKVersion(version) {
    if (!version) {
        return;
    }
    if (semver_1.default.valid(version)) {
        return;
    }
    const coerced = semver_1.default.coerce(version);
    if (coerced) {
        const suggestedVersion = coerced.version;
        if (suggestedVersion !== version) {
            throw new Error(`Invalid SDK version format: "${version}". ` +
                'SDK versions must follow semantic versioning (MAJOR.MINOR.PATCH format). ' +
                `Did you mean "${suggestedVersion}"?`);
        }
    }
    else {
        throw new Error(`Invalid SDK version format: "${version}". ` +
            'SDK versions must be valid semantic versions (MAJOR.MINOR.PATCH, e.g. 1.0.0).');
    }
}
function formatBuildError(apiErr) {
    const statusCode = apiErr.statusCode, originalMessage = apiErr.message || '', prefixPattern = /^(Bad Request|Unauthorized|Forbidden|Not Found|Internal Server Error):\s*/i;
    let errorMessage = originalMessage, cleanedMessage = errorMessage.replace(prefixPattern, '');
    if (statusCode === 429) {
        let message = safe_1.default.red('✗ Rate limit exceeded');
        message += '\n\nPlease try again later.';
        if (apiErr.rateLimitInfo?.reset) {
            const resetDate = new Date(apiErr.rateLimitInfo.reset * 1000);
            message += `\nRate limit resets at: ${resetDate.toLocaleTimeString()}`;
        }
        return message;
    }
    if (statusCode === 403) {
        return safe_1.default.red('✗ Permission denied') +
            '\n\nYou do not have permission to generate SDKs. ' +
            'Please check your account permissions or contact your administrator.';
    }
    if (statusCode === 400) {
        const lowerMessage = cleanedMessage.toLowerCase(), isLanguageError = lowerMessage.includes('language') &&
            (lowerMessage.includes('invalid') || lowerMessage.includes('not supported'));
        if (lowerMessage.includes('sdkversion') && lowerMessage.includes('semantic versioning')) {
            return safe_1.default.red('✗ Invalid SDK version format') +
                '\n\nThe SDK version must follow Semantic Versioning (SemVer) format: MAJOR.MINOR.PATCH' +
                '\n\nExamples:' +
                '\n  ✓ 2.1.0  (valid)' +
                '\n  ✓ 1.0.0  (valid)' +
                '\n  ✗ 2.1    (invalid - missing patch version)' +
                '\n  ✗ 2      (invalid - missing minor and patch versions)' +
                '\n\nPlease update your --sdk-version flag to use a valid SemVer format.' +
                '\nFor example: --sdk-version 2.1.0';
        }
        if (lowerMessage.includes('collection') || lowerMessage.includes('specification')) {
            if (lowerMessage.includes('not found') || lowerMessage.includes('invalid')) {
                return safe_1.default.red('✗ Invalid collection or specification') +
                    '\n\nThe provided ID is invalid or you do not have access to it.' +
                    '\n\nPlease verify:' +
                    '\n  • The ID is correct' +
                    '\n  • You have access to the collection/specification' +
                    '\n  • The ID is in UUID format (e.g., cc6d4ba8-8d30-4e88-85c2-cefa822621aa)';
            }
        }
        if (isLanguageError) {
            return safe_1.default.red('✗ Invalid language specified') +
                '\n\nOne or more of the specified languages is not supported.' +
                '\n\nSupported languages: typescript, python, java' +
                '\n\nPlease check your --language flag and try again.';
        }
        return safe_1.default.red('✗ Validation error') +
            '\n\n' + originalMessage +
            '\n\nPlease check your command parameters and try again.';
    }
    if (statusCode === 404) {
        return safe_1.default.red('✗ Resource not found') +
            '\n\nThe requested resource could not be found.' +
            '\n\nPlease verify:' +
            '\n  • The collection or specification ID is correct' +
            '\n  • You have access to the resource' +
            '\n  • The resource exists in your workspace';
    }
    if (statusCode === 500 || statusCode === 502 || statusCode === 503 || statusCode === 504) {
        return safe_1.default.red('✗ Server error') +
            '\n\nAn error occurred on the server while processing your request.' +
            '\n\nPlease try again in a few moments. If the problem persists, contact support.';
    }
    let defaultMessage = safe_1.default.red('✗ SDK generation failed');
    if (originalMessage) {
        defaultMessage += '\n\n' + originalMessage;
    }
    if (statusCode) {
        defaultMessage += `\n\nStatus code: ${statusCode}`;
    }
    return defaultMessage;
}
function registerGenerateCommand(sdk) {
    const languageChoices = sdk_utils_1.SUPPORTED_LANGUAGES.map((l) => { return `"${l}"`; }).join(', ');
    sdk.command('generate <idOrPath>')
        .alias('build')
        .description('Generate a new SDK from a Postman collection or specification.')
        .usage('<idOrPath> [options]')
        .configureOutput({
        outputError: (str, write) => {
            if (str.includes('idOrPath')) {
                write('error: Missing collection/specification ID, file path, or URL.\n' +
                    'Run `postman sdk generate --help` for more information.\n');
            }
            else {
                write(str);
            }
        }
    })
        .addHelpText('after', '\nInput:\n' +
        '  <idOrPath> can be either:\n' +
        '    - A Postman Collection ID or Specification ID (UUID format)\n' +
        '    - A file path to a collection (.json) or specification (.json, .yaml, .yml) file\n' +
        '    - A URL to a publicly accessible collection or specification file\n' +
        '  \n' +
        '  When using an ID, the SDK generation updates the version in Postman Cloud.\n' +
        '  When using a file path or URL, a one-off build is created that does not update ' +
        'the Postman Cloud.\n' +
        '  \n' +
        '  The command will automatically detect whether the input is a file, URL, or ID,\n' +
        '  and whether it represents a collection or specification.\n' +
        '\nExamples:\n' +
        '  Generate TypeScript SDK from collection ID:\n' +
        '    postman sdk generate 12345-abcde-67890 --language typescript --output-dir ./my-sdk\n\n' +
        '  Generate TypeScript SDK from specification ID:\n' +
        '    postman sdk generate 67890-fghij-12345 --language typescript --output-dir ./my-sdk\n\n' +
        '  Generate SDK from collection file (one-off build):\n' +
        '    postman sdk generate ./my-collection.json -l typescript -o ./my-sdk\n\n' +
        '  Generate SDK from OpenAPI specification file (one-off build):\n' +
        '    postman sdk generate ./openapi.yaml -l typescript -o ./my-sdk\n\n' +
        '  Generate SDK from publicly accessible URL (one-off build):\n' +
        '    postman sdk generate https://petstore.swagger.io/v2/swagger.json -l typescript -o ./my-sdk\n\n' +
        '  Generate multiple SDKs (TypeScript and Python):\n' +
        '    postman sdk generate 12345-abcde-67890 -l typescript python -o ./sdks\n\n' +
        '  Generate all supported languages:\n' +
        '    postman sdk generate 12345-abcde-67890 --all -o ./sdks\n\n' +
        '  Generate and auto-overwrite existing SDKs:\n' +
        '    postman sdk generate 12345-abcde-67890 -l typescript --yes -o ./sdks\n\n' +
        '  Verbose generation:\n' +
        '    postman sdk generate 12345-abcde-67890 -l java --verbose\n')
        .option('-l, --language <languages...>', `Target language(s) for SDK generation. Specify one or more languages. [choices: ${languageChoices}]`)
        .option('--all', 'Generate SDKs for all supported languages')
        .option('-o, --output-dir <path>', 'Directory where the generated SDK will be saved', './sdks')
        .option('--name <name>', 'Name for the generated SDK package')
        .option('--sdk-version <version>', 'Version number for the generated SDK')
        .option('--config <path>', 'Path to SDK generation config file (JSON). ' +
        'If not specified, will look for .postman/config.json (sdk key).')
        .option('-y, --yes', 'Skip all confirmation prompts and overwrite existing files')
        .option('--overwrite', 'Overwrite existing files without confirmation')
        .option('--verbose', 'Show detailed generation progress and logs')
        .option('--log-level <level>', 'Set build log detail level. "warn" (default) shows errors/warnings inline and collapses ' +
        'info notes; "info" expands all log entries', 'warn')
        .option('--pr', 'Create pull requests to configured GitHub repositories after SDK generation')
        .action((idOrPath, command) => {
        let exitCode = 0;
        (0, waterfall_1.default)([
            (next) => {
                util_1.default.addApiKeyToOptions(command, {}, (err, options) => {
                    if (err) {
                        return next(err);
                    }
                    if (!options.postmanApiKey && !options.accessToken) {
                        return next(new Error(constants_1.ERROR_MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE));
                    }
                    next(null, options);
                });
            },
            (options, next) => {
                if ((0, sdk_utils_1.isUrl)(idOrPath)) {
                    if (command.verbose) {
                        print_1.default.lf(`Detected URL: ${idOrPath}`);
                        print_1.default.lf('Fetching content from URL...');
                    }
                    (0, sdk_utils_1.fetchUrlToTempFile)(idOrPath)
                        .then((tempFilePath) => {
                        try {
                            const entityType = (0, sdk_utils_1.getEntityTypeFromFile)(tempFilePath);
                            if (command.verbose) {
                                print_1.default.lf(`Fetched content saved to temporary file: ${tempFilePath}`);
                                print_1.default.lf(`Entity type: ${entityType}`);
                            }
                            return next(null, {
                                type: 'file',
                                value: tempFilePath,
                                entityType: entityType,
                                tempFile: true
                            }, options);
                        }
                        catch (err) {
                            try {
                                if (fs_1.default.existsSync(tempFilePath)) {
                                    fs_1.default.unlinkSync(tempFilePath);
                                }
                            }
                            catch {
                            }
                            return next(new Error(`Failed to process fetched content: ${err.message}`));
                        }
                    })
                        .catch((err) => {
                        return next(new Error(`Failed to fetch URL: ${err.message}`));
                    });
                    return;
                }
                if ((0, sdk_utils_1.isFilePath)(idOrPath)) {
                    try {
                        let resolvedPath = idOrPath;
                        let entityType;
                        const v3Root = (0, v3_filesystem_collection_1.findV3FilesystemCollectionRootFromFile)(idOrPath);
                        if (v3Root) {
                            resolvedPath = v3Root;
                            entityType = 'collection';
                            if (command.verbose) {
                                print_1.default.lf(`Detected v3 request file; using collection root: ${v3Root}`);
                            }
                        }
                        else {
                            entityType = (0, sdk_utils_1.getEntityTypeFromFile)(idOrPath);
                        }
                        if (command.verbose) {
                            print_1.default.lf(`Detected file path: ${resolvedPath}`);
                            print_1.default.lf(`Entity type: ${entityType}`);
                        }
                        return next(null, { type: 'file', value: resolvedPath, entityType: entityType }, options);
                    }
                    catch (err) {
                        return next(new Error(`File validation failed: ${err.message}`));
                    }
                }
                (0, sdk_utils_1.extractCollectionId)(idOrPath, (err, id) => {
                    if (err) {
                        return next(err);
                    }
                    next(null, { type: 'id', value: id || idOrPath }, options);
                });
            },
            (input, options, next) => {
                if (command.pr && input.type === 'file') {
                    return next(new Error('PR creation is not supported for file-based one-off builds. ' +
                        'Please use a collection ID instead, or omit the --pr flag.'));
                }
                if (input.type === 'file' && input.entityType) {
                    if (command.pr && input.entityType === 'specification') {
                        return next(new Error('PR creation is not supported for specifications. ' +
                            'The integrations API does not currently support specifications. ' +
                            'Please use a collection ID instead, or omit the --pr flag. ' +
                            'Support for specifications is being worked on and may be available in a future release.'));
                    }
                    return next(null, input, options);
                }
                if (command.verbose) {
                    print_1.default.lf('Identifying ID type (collection or specification)...');
                }
                (0, sdk_utils_1.identifyIdType)(input.value, options, (err, idType, identifiedId) => {
                    if (err) {
                        return next(new Error(`ID identification failed: ${err.message}`));
                    }
                    if (!idType || idType === 'unknown' || !identifiedId) {
                        return next(new Error('ID identification failed. ' +
                            'Please check the ID and your access permissions.'));
                    }
                    if (command.verbose) {
                        print_1.default.lf(`Identified as: ${idType}`);
                    }
                    if (command.pr && idType === 'specification') {
                        return next(new Error('PR creation is not supported for specifications. ' +
                            'The integrations API does not currently support specifications. ' +
                            'Please use a collection ID instead, or omit the --pr flag. ' +
                            'Support for specifications is being worked on and may be available in a future release.'));
                    }
                    if (command.pr) {
                        const integrationsService = (0, integrations_service_1.createIntegrationsService)({
                            postmanApiKey: options.postmanApiKey,
                            accessToken: options.accessToken
                        });
                        return integrationsService.getIntegrations(identifiedId, undefined, idType)
                            .then((integrations) => {
                            if (!integrations || integrations.length === 0) {
                                const entityTypeName = idType === 'collection' ? 'collection' : 'specification';
                                return next(new Error(`No integrations configured for this ${entityTypeName}. ` +
                                    'Please run `postman sdk connect` to configure a GitHub integration ' +
                                    'before using the --pr flag.'));
                            }
                            next(null, { type: 'id', value: identifiedId, entityType: idType }, options);
                        })
                            .catch((integrationErr) => {
                            return next(new Error(`Failed to verify integrations for PR creation: ${integrationErr.message}`));
                        });
                    }
                    next(null, { type: 'id', value: identifiedId, entityType: idType }, options);
                });
            },
            (input, options, next) => {
                (async () => {
                    try {
                        let configData = null, configPath = null;
                        try {
                            configData = await (0, config_service_1.getConfigFile)(command.config);
                            configPath = (0, config_service_1.getConfigFilePath)(command.config);
                            if (configData && command.verbose) {
                                print_1.default.lf(`Loaded config from: ${configPath}`);
                            }
                            if (configData) {
                                if (configData.baseUrl && !command.baseUrl) {
                                    command.baseUrl = configData.baseUrl;
                                }
                            }
                        }
                        catch (err) {
                            return next(new Error(`Failed to load config file: ${err.message}`));
                        }
                        if (command.sdkVersion) {
                            try {
                                validateSDKVersion(command.sdkVersion);
                            }
                            catch (err) {
                                return next(err);
                            }
                        }
                        const selectLanguages = async () => {
                            if (command.all) {
                                if (command.language) {
                                    throw new Error('Cannot use both --all and --language flags. ' +
                                        'Please use either --all to generate all languages or ' +
                                        '--language to specify specific languages.');
                                }
                                const languages = [...sdk_utils_1.SUPPORTED_LANGUAGES];
                                if (command.verbose) {
                                    const langList = languages.join(', ');
                                    print_1.default.lf(`Generating SDKs for all supported languages: ${langList}`);
                                }
                                return languages;
                            }
                            if (command.language) {
                                return (0, sdk_utils_1.parseLanguages)(command.language);
                            }
                            if (configData && configData.languages && Array.isArray(configData.languages)) {
                                const configLanguages = configData.languages
                                    .map((lang) => { return lang.toLowerCase(); })
                                    .filter((lang) => {
                                    return sdk_utils_1.SUPPORTED_LANGUAGES.includes(lang);
                                });
                                if (configLanguages.length > 0) {
                                    if (command.verbose) {
                                        print_1.default.lf(`Using languages from config: ${configLanguages.join(', ')}`);
                                    }
                                    return configLanguages;
                                }
                            }
                            print_1.default.lf('No language specified. Please select the languages you want to generate:\n');
                            const response = await (0, prompts_1.default)({
                                type: 'multiselect',
                                name: 'languages',
                                message: 'Select target language(s) for SDK generation',
                                choices: sdk_utils_1.SUPPORTED_LANGUAGES.map((lang) => {
                                    return {
                                        title: lang,
                                        value: lang,
                                        selected: false
                                    };
                                }),
                                min: 1,
                                hint: '- Space to select. Return to submit',
                                instructions: false
                            });
                            if (!response || !response.languages || response.languages.length === 0) {
                                throw new Error('No languages selected. SDK generation cancelled.');
                            }
                            if (command.verbose) {
                                print_1.default.lf(`Selected languages: ${response.languages.join(', ')}`);
                            }
                            return response.languages;
                        };
                        const cleanSdkConfig = (config, selectedLanguages) => {
                            if (!config) {
                                return null;
                            }
                            const cleaned = { ...config }, customizations = cleaned.customizations;
                            delete cleaned.specFilePath;
                            delete cleaned.languages;
                            delete cleaned.sdkVersion;
                            delete cleaned.sdkName;
                            if (customizations &&
                                typeof customizations === 'object' &&
                                !Array.isArray(customizations)) {
                                const existingLanguageOptions = cleaned.languageOptions, normalizedLanguageOptions = (existingLanguageOptions &&
                                    typeof existingLanguageOptions === 'object' &&
                                    !Array.isArray(existingLanguageOptions)) ?
                                    existingLanguageOptions :
                                    {}, nextLanguageOptions = { ...normalizedLanguageOptions };
                                (selectedLanguages || []).forEach((lang) => {
                                    const existingPerLang = normalizedLanguageOptions[lang], perLang = (existingPerLang &&
                                        typeof existingPerLang === 'object' &&
                                        !Array.isArray(existingPerLang)) ?
                                        existingPerLang :
                                        {};
                                    nextLanguageOptions[lang] = { ...customizations, ...perLang };
                                });
                                cleaned.languageOptions = nextLanguageOptions;
                            }
                            return cleaned;
                        };
                        const languages = await selectLanguages(), cleanedConfig = cleanSdkConfig(configData, languages);
                        let resolvedSdkName = command.name, resolvedSdkVersion = command.sdkVersion;
                        if (configData?.languageOptions) {
                            for (const lang of languages) {
                                const langOpts = configData.languageOptions[lang];
                                if (!resolvedSdkName && langOpts?.sdkName) {
                                    resolvedSdkName = langOpts.sdkName;
                                }
                                else if (resolvedSdkName && langOpts?.sdkName &&
                                    langOpts.sdkName !== resolvedSdkName) {
                                    print_1.default.lf(safe_1.default.yellow(`Warning: languageOptions.${lang}.sdkName ` +
                                        `("${langOpts.sdkName}") differs from the ` +
                                        `resolved name ("${resolvedSdkName}"). ` +
                                        'Using the first resolved value.'));
                                }
                                if (!resolvedSdkVersion && langOpts?.sdkVersion) {
                                    resolvedSdkVersion = langOpts.sdkVersion;
                                }
                                else if (resolvedSdkVersion && langOpts?.sdkVersion &&
                                    langOpts.sdkVersion !== resolvedSdkVersion) {
                                    print_1.default.lf(safe_1.default.yellow(`Warning: languageOptions.${lang}.sdkVersion ` +
                                        `("${langOpts.sdkVersion}") differs from the ` +
                                        `resolved version ("${resolvedSdkVersion}"). ` +
                                        'Using the first resolved value.'));
                                }
                                if (resolvedSdkName && resolvedSdkVersion) {
                                    break;
                                }
                            }
                        }
                        if (!resolvedSdkName && configData?.sdkName) {
                            resolvedSdkName = configData.sdkName;
                        }
                        if (!resolvedSdkVersion && configData?.sdkVersion) {
                            resolvedSdkVersion = configData.sdkVersion;
                        }
                        if (!resolvedSdkVersion) {
                            resolvedSdkVersion = DEFAULT_SDK_VERSION;
                        }
                        const sdkOptions = {
                            languages: languages,
                            outputDir: command.outputDir,
                            name: resolvedSdkName,
                            sdkVersion: resolvedSdkVersion,
                            config: configPath || command.config,
                            sdkConfig: cleanedConfig,
                            yes: command.yes,
                            overwrite: command.overwrite,
                            verbose: command.verbose,
                            logLevel: command.logLevel === 'info' ? 'info' : 'warn',
                            postmanApiKey: options.postmanApiKey,
                            accessToken: options.accessToken,
                            baseUrl: command.baseUrl,
                            webhookUrl: command.webhookUrl,
                            watch: command.watch
                        };
                        return next(null, input, sdkOptions);
                    }
                    catch (err) {
                        return next(err);
                    }
                })();
            },
            (input, sdkOptions, next) => {
                if (sdkOptions.verbose) {
                    print_1.default.lf('Starting SDK generation...');
                    if (input.type === 'file') {
                        print_1.default.lf(`File: ${input.value}`);
                        print_1.default.lf(`Entity type: ${input.entityType}`);
                        print_1.default.lf('Build type: One-off (temporary, not stored in Postman Cloud)');
                    }
                    else {
                        const entityTypeName = input.entityType === 'collection' ? 'Collection' : 'Specification';
                        print_1.default.lf(`${entityTypeName} ID: ${input.value}`);
                        print_1.default.lf('Build type: Postman Cloud (updates stored SDK)');
                    }
                    print_1.default.lf(`Language(s): ${sdkOptions.languages.join(', ')}`);
                    print_1.default.lf(`Output: ${sdkOptions.outputDir}`);
                }
                (async () => {
                    try {
                        const service = (0, sdk_generator_service_1.createSDKGeneratorService)({
                            accessToken: sdkOptions.accessToken,
                            postmanApiKey: sdkOptions.postmanApiKey
                        });
                        print_1.default.wait(safe_1.default.cyan);
                        print_1.default.lf('Creating SDK build...');
                        if (input.type === 'file') {
                            const preparedFile = await (0, sdk_utils_1.prepareCollectionFile)(input.value, input.entityType, Boolean(sdkOptions.verbose));
                            const filePath = preparedFile.filePath;
                            const v3MergedFilePath = preparedFile.tempFilePath;
                            const rawName = sdkOptions.name || deriveNameFromPath(input.value);
                            const oneOffBuildOptions = {
                                filePath: filePath,
                                entityType: input.entityType,
                                sdkName: sanitizeSdkName(rawName),
                                languages: sdkOptions.languages,
                                sdkConfig: sdkOptions.sdkConfig
                            };
                            if (sdkOptions.sdkVersion) {
                                oneOffBuildOptions.sdkVersion = sdkOptions.sdkVersion;
                            }
                            await service.createOneOffBuild(oneOffBuildOptions)
                                .then(({ build, rateLimitInfo }) => {
                                print_1.default.unwait();
                                if (sdkOptions.verbose && rateLimitInfo.remaining !== undefined) {
                                    print_1.default.lf(safe_1.default.gray('Rate limit: ' +
                                        `${rateLimitInfo.remaining}/${rateLimitInfo.limit} remaining`));
                                }
                                print_1.default.lf(safe_1.default.green('✓ One-off build created successfully'));
                                print_1.default.lf(`Build ID: ${build.id}`);
                                print_1.default.lf(`Status: ${build.status}`);
                                const displayLangs = build.languages?.join(', ') ||
                                    sdkOptions.languages.join(', ');
                                print_1.default.lf(`Languages: ${displayLangs}`);
                                print_1.default.lf('\nWatching build progress...');
                                print_1.default.wait(safe_1.default.cyan);
                                return service.pollBuildStatus(build.id, {
                                    interval: 3000,
                                    timeout: 600000,
                                    onProgress: (updatedBuild) => {
                                        if (sdkOptions.verbose) {
                                            print_1.default.lf(`Status: ${updatedBuild.status}`);
                                        }
                                    }
                                }).then(async (completedBuild) => {
                                    print_1.default.unwait();
                                    completedBuild.collectionId = input.entityType === 'collection' ?
                                        input.value :
                                        undefined;
                                    completedBuild.apiSpecId = input.entityType === 'specification' ?
                                        input.value :
                                        undefined;
                                    await (0, sdk_utils_1.handleBuildCompletion)(completedBuild, sdkOptions.outputDir, Boolean(sdkOptions.yes || sdkOptions.overwrite), (err) => {
                                        (0, sdk_utils_1.cleanupTempFiles)([
                                            input.tempFile ? input.value : undefined,
                                            v3MergedFilePath
                                        ], Boolean(sdkOptions.verbose));
                                        if (err) {
                                            return next(err);
                                        }
                                        next(null);
                                    }, service, true);
                                }).catch((pollErr) => {
                                    print_1.default.unwait();
                                    (0, sdk_utils_1.cleanupTempFiles)(v3MergedFilePath);
                                    next(new Error(`Build polling failed: ${pollErr.message}`));
                                });
                            })
                                .catch((apiErr) => {
                                print_1.default.unwait();
                                (0, sdk_utils_1.cleanupTempFiles)(v3MergedFilePath);
                                let errorMessage = `SDK generation failed: ${apiErr.message}`;
                                if (apiErr.statusCode === 429) {
                                    errorMessage += '\nRate limit exceeded. Please try again later.';
                                    if (apiErr.rateLimitInfo?.reset) {
                                        const resetTimestamp = apiErr.rateLimitInfo.reset * 1000;
                                        const resetDate = new Date(resetTimestamp);
                                        const resetTime = resetDate.toLocaleTimeString();
                                        errorMessage += `\nRate limit resets at: ${resetTime}`;
                                    }
                                }
                                else if (apiErr.statusCode === 403) {
                                    errorMessage += '\nYou do not have permission to generate SDKs. ' +
                                        'Please check your account permissions.';
                                }
                                return next(new Error(errorMessage));
                            });
                            return;
                        }
                        const buildOptions = {
                            languages: sdkOptions.languages,
                            sdkConfig: sdkOptions.sdkConfig,
                            sdkName: sdkOptions.name ? sanitizeSdkName(sdkOptions.name) : undefined
                        };
                        if (sdkOptions.sdkVersion) {
                            buildOptions.sdkVersion = sdkOptions.sdkVersion;
                        }
                        if (input.entityType === 'collection') {
                            buildOptions.collectionId = input.value;
                        }
                        else if (input.entityType === 'specification') {
                            buildOptions.specificationId = input.value;
                        }
                        service.createBuild(buildOptions)
                            .then(({ build, rateLimitInfo }) => {
                            print_1.default.unwait();
                            if (sdkOptions.verbose && rateLimitInfo.remaining !== undefined) {
                                print_1.default.lf(safe_1.default.gray(`Rate limit: ${rateLimitInfo.remaining}/${rateLimitInfo.limit} remaining`));
                            }
                            print_1.default.lf(safe_1.default.green('✓ Build created successfully'));
                            print_1.default.lf(`Build ID: ${build.id}`);
                            print_1.default.lf(`Status: ${build.status}`);
                            const displayLangs = build.languages?.join(', ') ||
                                sdkOptions.languages.join(', ');
                            print_1.default.lf(`Languages: ${displayLangs}`);
                            print_1.default.lf('\nWatching build progress...');
                            print_1.default.wait(safe_1.default.cyan);
                            return service.pollBuildStatus(build.id, {
                                interval: 3000,
                                timeout: 600000,
                                onProgress: (updatedBuild) => {
                                    if (sdkOptions.verbose) {
                                        print_1.default.lf(`Status: ${updatedBuild.status}`);
                                    }
                                }
                            }).then(async (completedBuild) => {
                                print_1.default.unwait();
                                completedBuild.collectionId = input.entityType === 'collection' ?
                                    input.value :
                                    undefined;
                                completedBuild.apiSpecId = input.entityType === 'specification' ?
                                    input.value :
                                    undefined;
                                await (0, sdk_utils_1.handleBuildCompletion)(completedBuild, sdkOptions.outputDir, Boolean(sdkOptions.yes || sdkOptions.overwrite), (err) => {
                                    (0, sdk_utils_1.cleanupTempFiles)(input.tempFile ? input.value : undefined, Boolean(sdkOptions.verbose));
                                    if (err) {
                                        return next(err);
                                    }
                                    if (command.pr && input.entityType === 'collection') {
                                        return (0, pr_workflow_1.createPRsForIntegrations)(input.value, sdkOptions.languages, sdkOptions.sdkVersion, sdkOptions.outputDir, {
                                            postmanApiKey: sdkOptions.postmanApiKey,
                                            accessToken: sdkOptions.accessToken
                                        }, input.entityType).then(() => {
                                            return next(null);
                                        }).catch((prErr) => {
                                            print_1.default.lf(safe_1.default.yellow(`\n⚠ PR creation failed: ${prErr.message}`));
                                            return next(null);
                                        });
                                    }
                                    return next(null);
                                }, service, true, sdkOptions.logLevel || 'warn');
                            }).catch((pollErr) => {
                                print_1.default.unwait();
                                next(new Error(`Build polling failed: ${pollErr.message}`));
                            });
                        })
                            .catch((apiErr) => {
                            print_1.default.unwait();
                            const errorMessage = formatBuildError(apiErr);
                            return next(new Error(errorMessage));
                        });
                    }
                    catch (err) {
                        print_1.default.unwait();
                        return next(new Error(`Failed to initialize SDK service: ${err.message}`));
                    }
                })().catch((asyncErr) => {
                    return next(asyncErr);
                });
            }
        ], (err) => {
            if (err) {
                console.error(`Error: ${err.message || err}\n`);
                exitCode = 1;
            }
            process.exit(exitCode);
        });
    });
}
