// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/get.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerFetchCommand = registerFetchCommand;
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const waterfall_1 = __importDefault(require("async/waterfall"));
const util_1 = __importDefault(require("../../postman-entities/util"));
const constants_1 = require("../../constants");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const sdk_generator_service_1 = require("../../api/sdk-generator-service");
const sdk_utils_1 = require("./sdk-utils");
function registerFetchCommand(sdk) {
    sdk.command('fetch <buildId>')
        .alias('get')
        .alias('download')
        .description('Download SDKs from a build by its ID.')
        .usage('<buildId> [options]')
        .addHelpText('after', '\nExamples:\n' +
        '  Download SDKs from build:\n' +
        '    postman sdk get 12345\n\n' +
        '  Download to specific directory:\n' +
        '    postman sdk fetch 12345 --output-dir ./my-sdks\n\n' +
        '  Overwrite existing files:\n' +
        '    postman sdk download 12345 --overwrite\n\n' +
        '  Download specific language only:\n' +
        '    postman sdk fetch 12345 --language typescript\n')
        .option('-o, --output-dir <path>', 'Directory to save the downloaded SDK(s)', './sdks')
        .option('-l, --language <language>', 'Download only the specified language (e.g., typescript, python)')
        .option('--overwrite', 'Overwrite existing files without confirmation')
        .option('--verbose', 'Show detailed download progress')
        .option('--log-level <level>', 'Set build log detail level. "warn" (default) shows errors/warnings inline and collapses ' +
        'info notes; "info" expands all log entries', 'warn')
        .action((buildId, command) => {
        let exitCode = 0;
        (0, waterfall_1.default)([
            (next) => {
                util_1.default.addApiKeyToOptions(command, {}, (err, options) => {
                    if (err) {
                        return next(err);
                    }
                    if (!options.postmanApiKey && !options.accessToken) {
                        return next(new Error(constants_1.ERROR_MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE));
                    }
                    next(null, options);
                });
            },
            (options, next) => {
                const buildIdNum = parseInt(buildId, 10);
                if (isNaN(buildIdNum) || buildIdNum <= 0) {
                    return next(new Error('Invalid build ID provided. Build ID must be a positive number.'));
                }
                next(null, options);
            },
            (options, next) => {
                const downloadOptions = {
                    outputDir: command.outputDir,
                    language: command.language ? command.language.toUpperCase() : null,
                    overwrite: command.overwrite,
                    verbose: command.verbose,
                    logLevel: (command.logLevel === 'info' ? 'info' : 'warn'),
                    postmanApiKey: options.postmanApiKey,
                    accessToken: options.accessToken
                };
                if (downloadOptions.verbose) {
                    print_1.default.lf('Fetching build from Postman...');
                    print_1.default.lf(`Build ID: ${buildId}`);
                }
                try {
                    const service = (0, sdk_generator_service_1.createSDKGeneratorService)({
                        accessToken: downloadOptions.accessToken,
                        postmanApiKey: downloadOptions.postmanApiKey
                    }), buildIdNum = parseInt(buildId, 10);
                    print_1.default.wait(safe_1.default.cyan);
                    print_1.default.lf('Fetching build details...');
                    service.getBuild(buildIdNum)
                        .then(async ({ build, rateLimitInfo }) => {
                        if (downloadOptions.verbose && rateLimitInfo.remaining !== undefined) {
                            print_1.default.lf(safe_1.default.gray(`Rate limit: ${rateLimitInfo.remaining}/${rateLimitInfo.limit} remaining`));
                        }
                        print_1.default.lf(`Build Status: ${build.status}`);
                        const successfulArtifacts = (build.artifacts || []).filter((artifact) => {
                            return artifact.status === 'SUCCESS';
                        });
                        if (successfulArtifacts.length === 0) {
                            print_1.default.unwait();
                            await (0, sdk_utils_1.fetchAndDisplayBuildSummary)(service, buildIdNum, undefined, undefined, downloadOptions.logLevel);
                            throw new Error('No SDKs found in build');
                        }
                        const artifactsToDownload = [];
                        for (const artifact of successfulArtifacts) {
                            let language;
                            try {
                                if (artifact.sdk) {
                                    language = artifact.sdk.language.toLowerCase();
                                }
                                else if (artifact.bucketKey) {
                                    language = (0, sdk_utils_1.extractLanguageFromBucketKey)(artifact.bucketKey);
                                }
                                else {
                                    continue;
                                }
                                if (downloadOptions.language) {
                                    const requestedLanguage = downloadOptions.language.toUpperCase();
                                    if (language.toUpperCase() !== requestedLanguage) {
                                        continue;
                                    }
                                }
                                artifactsToDownload.push({ artifact, language });
                            }
                            catch {
                                continue;
                            }
                        }
                        if (artifactsToDownload.length === 0) {
                            print_1.default.unwait();
                            const availableLanguages = successfulArtifacts
                                .map((a) => {
                                if (a.sdk) {
                                    return a.sdk.language;
                                }
                                if (a.bucketKey) {
                                    try {
                                        return (0, sdk_utils_1.extractLanguageFromBucketKey)(a.bucketKey).toUpperCase();
                                    }
                                    catch {
                                        return null;
                                    }
                                }
                                return null;
                            })
                                .filter((lang) => {
                                return lang !== null;
                            });
                            if (downloadOptions.language) {
                                throw new Error(`No SDK found for language: ${downloadOptions.language}\n` +
                                    `Available: ${[...new Set(availableLanguages)].join(', ')}`);
                            }
                            throw new Error('No SDK artifacts found in build');
                        }
                        print_1.default.lf(`Found ${artifactsToDownload.length} SDK(s) to download`);
                        await (0, sdk_utils_1.fetchAndDisplayBuildSummary)(service, buildIdNum, undefined, undefined, downloadOptions.logLevel);
                        let downloadedCount = 0;
                        for (const { artifact, language } of artifactsToDownload) {
                            const displayLanguage = artifact.sdk?.language || language.toUpperCase();
                            print_1.default.lf(`\nDownloading ${displayLanguage} SDK...`);
                            const extractDir = path_1.default.join(downloadOptions.outputDir, language);
                            if (fs_1.default.existsSync(extractDir) && !downloadOptions.overwrite) {
                                print_1.default.lf(safe_1.default.yellow('  Skipped: SDK directory already exists (use --overwrite to replace)'));
                                continue;
                            }
                            if (!fs_1.default.existsSync(downloadOptions.outputDir)) {
                                fs_1.default.mkdirSync(downloadOptions.outputDir, { recursive: true });
                            }
                            try {
                                let idForDownload;
                                let idType;
                                if (build.collectionId) {
                                    idForDownload = build.collectionId;
                                    idType = 'collection';
                                }
                                else if (build.apiSpecId) {
                                    idForDownload = build.apiSpecId;
                                    idType = 'specification';
                                }
                                else {
                                    idForDownload = build.id;
                                    idType = 'build';
                                }
                                await (0, sdk_utils_1.downloadAndExtractSDK)(artifact, downloadOptions.outputDir, service, idForDownload, idType);
                                downloadedCount++;
                            }
                            catch (err) {
                                print_1.default.lf(safe_1.default.red(`  ✗ Failed: ${err.message}`));
                            }
                        }
                        print_1.default.unwait();
                        if (downloadedCount > 0) {
                            print_1.default.lf(safe_1.default.green(`\n✓ Downloaded and extracted ${downloadedCount} SDK(s) successfully!`));
                            print_1.default.lf('\nNext steps:');
                            print_1.default.lf('  Follow the README in the extracted SDK folder for integration instructions');
                        }
                        else {
                            print_1.default.lf(safe_1.default.yellow('\nNo SDKs were downloaded.'));
                        }
                        const failedArtifacts = (build.artifacts || []).filter((artifact) => {
                            return artifact.status === 'FAIL';
                        });
                        if (failedArtifacts.length > 0 && downloadedCount > 0) {
                            print_1.default.lf(safe_1.default.yellow(`\n⚠ ${failedArtifacts.length} artifact(s) in this build failed:`));
                        }
                        next(null);
                    })
                        .catch((apiErr) => {
                        print_1.default.unwait();
                        let errorMessage = `Build fetch failed: ${apiErr.message}`;
                        if (apiErr.statusCode === 404) {
                            errorMessage = `Build not found: ${buildId}\n` +
                                'Please verify the build ID is correct.';
                        }
                        else if (apiErr.statusCode === 429) {
                            errorMessage += '\nRate limit exceeded. Please try again later.';
                            if (apiErr.rateLimitInfo?.reset) {
                                const resetDate = new Date(apiErr.rateLimitInfo.reset * 1000);
                                errorMessage += `\nRate limit resets at: ${resetDate.toLocaleTimeString()}`;
                            }
                        }
                        return next(new Error(errorMessage));
                    });
                }
                catch (err) {
                    print_1.default.unwait();
                    return next(new Error(`Failed to initialize SDK service: ${err.message}`));
                }
            }
        ], (err) => {
            if (err) {
                console.error(`Error: ${err.message || err}\n`);
                exitCode = 1;
            }
            process.exit(exitCode);
        });
    });
}
