// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/git-operations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitOperations = void 0;
const isomorphic_git_1 = __importDefault(require("isomorphic-git"));
const node_1 = __importDefault(require("isomorphic-git/http/node"));
const node_fs_1 = __importDefault(require("node:fs"));
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const rest_1 = require("@octokit/rest");
const rimraf_1 = require("rimraf");
class GitOperations {
    prData;
    octokit;
    localRepoPath;
    constructor(prData) {
        this.prData = prData;
        this.octokit = new rest_1.Octokit({ auth: prData.accessToken });
        const tempPrefix = node_path_1.default.join(node_os_1.default.tmpdir(), 'postman-sdk-pr-');
        this.localRepoPath = node_fs_1.default.mkdtempSync(tempPrefix);
    }
    async clean() {
        try {
            await (0, rimraf_1.rimraf)(this.localRepoPath);
        }
        catch {
        }
    }
    async clone() {
        await isomorphic_git_1.default.clone({
            fs: node_fs_1.default,
            http: node_1.default,
            dir: this.localRepoPath,
            url: this.prData.repoUrl,
            onAuth: () => {
                return {
                    username: this.prData.accessToken,
                    password: 'x-oauth-basic'
                };
            }
        });
    }
    async checkout() {
        try {
            await isomorphic_git_1.default.fetch({
                fs: node_fs_1.default,
                http: node_1.default,
                dir: this.localRepoPath,
                remote: 'origin',
                ref: this.prData.targetBranch,
                depth: 1,
                onAuth: () => {
                    return {
                        username: this.prData.accessToken,
                        password: 'x-oauth-basic'
                    };
                }
            });
            await isomorphic_git_1.default.checkout({
                fs: node_fs_1.default,
                dir: this.localRepoPath,
                ref: `origin/${this.prData.targetBranch}`,
                track: true
            });
            await isomorphic_git_1.default.branch({
                fs: node_fs_1.default,
                dir: this.localRepoPath,
                ref: this.prData.branchName,
                checkout: true
            });
        }
        catch (error) {
            if (error instanceof isomorphic_git_1.default.Errors.NotFoundError) {
                throw new Error(`Target branch '${this.prData.targetBranch}' could not be found: ${error.message}`);
            }
            throw new Error('Unexpected error occurred while trying to fetch target branch ' +
                `${this.prData.targetBranch}: ${error.message}`);
        }
    }
    async stage(filePaths) {
        try {
            await Promise.all(filePaths.map((filePath) => {
                return isomorphic_git_1.default.add({ fs: node_fs_1.default, dir: this.localRepoPath, filepath: filePath });
            }));
        }
        catch (error) {
            throw new Error(`Unexpected error occurred while trying to stage changes: ${error.message}`);
        }
    }
    async commit() {
        try {
            await isomorphic_git_1.default.commit({
                fs: node_fs_1.default,
                dir: this.localRepoPath,
                message: this.prData.commitMessage,
                author: {
                    name: this.prData.authorName,
                    email: this.prData.authorEmail
                }
            });
        }
        catch (error) {
            throw new Error(`Unexpected error occurred while trying to commit changes: ${error.message}`);
        }
    }
    async push() {
        try {
            await isomorphic_git_1.default.push({
                fs: node_fs_1.default,
                http: node_1.default,
                dir: this.localRepoPath,
                remote: 'origin',
                ref: this.prData.branchName,
                onAuth: () => {
                    return {
                        username: this.prData.accessToken,
                        password: 'x-oauth-basic'
                    };
                }
            });
        }
        catch (error) {
            if (error instanceof isomorphic_git_1.default.Errors.HttpError) {
                const statusCode = error.data?.statusCode;
                const responseBody = error.data?.response;
                switch (statusCode) {
                    case 401: {
                        throw new Error(`Unauthorized access to push changes to branch ${this.prData.branchName}. ` +
                            `Response: ${responseBody || 'No additional details'}`);
                    }
                    case 403: {
                        const message = responseBody || 'No additional details';
                        throw new Error(`Access to push changes to branch ${this.prData.branchName} is forbidden. ` +
                            `GitHub response: ${message}\n\n` +
                            'This usually means:\n' +
                            '  1. The GitHub App doesn\'t have \'contents: write\' permission\n' +
                            '  2. The user hasn\'t authorized the app with sufficient permissions\n' +
                            '  3. The repository settings don\'t allow the app to write\n\n' +
                            'To fix this:\n' +
                            '  - Check the GitHub App\'s permissions in GitHub Settings > ' +
                            'Developer settings > GitHub Apps\n' +
                            '  - Ensure \'Repository permissions > Contents\' is set to ' +
                            '\'Read and write\'\n' +
                            '  - Re-authorize the app or update the integration if permissions were changed');
                    }
                    default: {
                        throw new Error('Unexpected error occurred while trying to push changes ' +
                            `(Status: ${statusCode}): ${error.message}. ` +
                            `Response: ${responseBody || 'No additional details'}`);
                    }
                }
            }
            throw new Error(`Unexpected error occurred while trying to push changes: ${error.message}`);
        }
    }
    async createPullRequest() {
        const { data } = await this.octokit.pulls.create({
            base: this.prData.targetBranch,
            body: this.prData.body,
            head: this.prData.branchName,
            owner: this.prData.owner,
            repo: this.prData.repo,
            title: this.prData.title
        });
        return data;
    }
    getLocalRepoPath() {
        return this.localRepoPath;
    }
}
exports.GitOperations = GitOperations;
