// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerInitCommand = registerInitCommand;
const promises_1 = __importDefault(require("node:fs/promises"));
const node_path_1 = __importDefault(require("node:path"));
const safe_1 = __importDefault(require("colors/safe"));
const console_1 = __importDefault(require("../../framework/console"));
const workspace_utils_1 = require("../workspace/workspace-utils");
const liblab_config_schema_1 = __importDefault(require("./schemas/liblab-config.schema"));
function registerInitCommand(sdk) {
    sdk.command('init')
        .description('Create or update .postman/config.json with an SDK configuration stub.')
        .option('-y, --yes', 'Skip prompts and overwrite existing sdk config')
        .action(async (command) => {
        const cwd = process.cwd(), postmanDir = node_path_1.default.join(cwd, '.postman'), configPath = node_path_1.default.join(postmanDir, 'config.json');
        await promises_1.default.mkdir(postmanDir, { recursive: true });
        let existing = null;
        try {
            existing = await (0, workspace_utils_1.readJSONFile)(configPath);
        }
        catch {
            existing = null;
        }
        const hasSdk = Boolean(existing && typeof existing === 'object' && existing.sdk);
        if (hasSdk && !command.yes) {
            console_1.default.log(safe_1.default.yellow(`SDK config already exists at ${configPath}. Re-run with --yes to overwrite.`));
            return;
        }
        const base = (existing && typeof existing === 'object') ? existing : {}, schemaProps = liblab_config_schema_1.default?.properties || {}, schemaExampleOrDefault = (key, fallback) => {
            const p = schemaProps?.[key];
            if (!p) {
                return fallback;
            }
            if (p.default !== undefined) {
                return p.default;
            }
            if (Array.isArray(p.examples) && p.examples.length > 0) {
                return p.examples[0];
            }
            return fallback;
        }, defaultsFromObjectSchema = (objSchema) => {
            if (!objSchema || typeof objSchema !== 'object') {
                return {};
            }
            const props = objSchema.properties;
            if (!props || typeof props !== 'object') {
                return {};
            }
            const out = {};
            Object.keys(props).forEach((k) => {
                const p = props[k];
                if (!p || typeof p !== 'object') {
                    return;
                }
                if (p.default !== undefined) {
                    out[k] = p.default;
                    return;
                }
                if (p.type === 'object') {
                    const nested = defaultsFromObjectSchema(p);
                    if (Object.keys(nested).length) {
                        out[k] = nested;
                    }
                }
            });
            return out;
        }, customizationsSchema = schemaProps?.customizations, customizationsDefaults = defaultsFromObjectSchema(customizationsSchema), customizations = (() => {
            const allowed = [
                'environments',
                'license',
                'retry'
            ], out = {};
            allowed.forEach((k) => {
                if (customizationsDefaults[k] !== undefined) {
                    out[k] = customizationsDefaults[k];
                    return;
                }
                const propSchema = customizationsSchema?.properties?.[k];
                const type = propSchema?.type;
                if (type === 'array') {
                    out[k] = [];
                }
                else if (type === 'object') {
                    out[k] = {};
                }
            });
            return out;
        })(), languageOptionsSchema = schemaProps?.languageOptions?.properties || {}, languageOptions = {
            typescript: (() => {
                const tsDefaults = defaultsFromObjectSchema(languageOptionsSchema?.typescript);
                delete tsDefaults.httpClient;
                delete tsDefaults.generateEnumAs;
                return tsDefaults;
            })(),
            java: (() => {
                const javaDefaults = defaultsFromObjectSchema(languageOptionsSchema?.java);
                delete javaDefaults.includeKotlinSnippets;
                return javaDefaults;
            })(),
            python: (() => {
                const pyDefaults = defaultsFromObjectSchema(languageOptionsSchema?.python);
                delete pyDefaults.alwaysInitializeOptionals;
                delete pyDefaults.enforceRequestValidation;
                delete pyDefaults.strictVersion;
                return pyDefaults;
            })()
        }, sdkTemplate = {
            sdkName: '',
            languages: schemaExampleOrDefault('languages', ['typescript']),
            customizations,
            languageOptions,
            analytics: {
                endpoint: '',
                exporter: 'otlp-http'
            }
        }, nextConfig = {
            ...base,
            schemaVersion: base.schemaVersion || '1',
            workspace: base.workspace || {},
            entities: base.entities || {},
            sdk: command.yes ? sdkTemplate : (base.sdk || sdkTemplate)
        }, content = `${JSON.stringify(nextConfig, null, 2)}\n`;
        await promises_1.default.writeFile(configPath, content, 'utf8');
        console_1.default.log(safe_1.default.green(`Successfully created/updated ${configPath}`));
    });
}
