// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerListCommand = registerListCommand;
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const waterfall_1 = __importDefault(require("async/waterfall"));
const util_1 = __importDefault(require("../../postman-entities/util"));
const constants_1 = require("../../constants");
const sdk_utils_1 = require("./sdk-utils");
const sdk_generator_service_1 = require("../../api/sdk-generator-service");
const prompts_1 = __importDefault(require("prompts"));
function padRight(str, width) {
    return str + ' '.repeat(Math.max(0, width - str.length));
}
function truncate(str, maxLength) {
    if (str.length <= maxLength) {
        return str;
    }
    return str.substring(0, maxLength - 3) + '...';
}
function registerListCommand(sdk) {
    sdk.command('list <id>')
        .alias('ls')
        .description('List SDK builds for a collection, extensible collection, or specification.')
        .usage('<id> [options]')
        .addHelpText('after', '\nInput:\n' +
        '  <id> can be a Postman Collection ID, Extensible Collection ID, or Specification ID\n' +
        '  The command will automatically detect the ID type\n' +
        '  Supported formats:\n' +
        '    Collection UUID:            12345678-abcdef12-1234-abcd-1234-abcdef123456\n' +
        '    Specification UUID:          12345678-1234-1234-1234-123456789012\n' +
        '    Extensible Collection ID:    698e05f4ecf468ef2465e43d (24-char hex)\n' +
        '\nExamples:\n' +
        '  List all builds for a collection:\n' +
        '    postman sdk list 12345-abcde-67890\n\n' +
        '  List all builds for an extensible collection:\n' +
        '    postman sdk list 698e05f4ecf468ef2465e43d\n\n' +
        '  List all builds for a specification:\n' +
        '    postman sdk list 67890-fghij-12345\n\n' +
        '  List only successful builds:\n' +
        '    postman sdk list 12345-abcde-67890 --status success\n\n' +
        '  List in-progress builds:\n' +
        '    postman sdk list 12345-abcde-67890 --status in-progress\n\n' +
        '  List failed builds:\n' +
        '    postman sdk list 12345-abcde-67890 --status failure\n\n' +
        '  Output as JSON:\n' +
        '    postman sdk list 12345-abcde-67890 --output json\n')
        .option('-s, --status <status>', 'Filter builds by status (case-insensitive) [choices: in-progress, success, failure]')
        .option('-o, --output <format>', 'Output format [choices: "json", "table"]', sdk_utils_1.validateOutputFormat, 'table')
        .option('--log-level <level>', 'Set build log detail level. "warn" (default) shows errors/warnings inline and collapses ' +
        'info notes; "info" expands all log entries', 'warn')
        .action((idOrPath, command) => {
        let exitCode = 0;
        (0, waterfall_1.default)([
            (next) => {
                util_1.default.addApiKeyToOptions(command, {}, (err, options) => {
                    if (err) {
                        return next(err);
                    }
                    if (!options.postmanApiKey && !options.accessToken) {
                        return next(new Error(constants_1.ERROR_MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE));
                    }
                    next(null, options);
                });
            },
            (options, next) => {
                (0, sdk_utils_1.extractCollectionId)(idOrPath, (err, id) => {
                    if (err) {
                        return next(err);
                    }
                    next(null, id, options);
                });
            },
            (id, options, next) => {
                if (command.verbose) {
                    print_1.default.lf('Identifying ID type (collection or specification)...');
                }
                (0, sdk_utils_1.identifyIdType)(id, options, (err, idType, identifiedId) => {
                    if (err) {
                        return next(new Error(`ID identification failed: ${err.message}`));
                    }
                    if (!idType || idType === 'unknown' || !identifiedId) {
                        return next(new Error('ID identification failed. ' +
                            'Please check the ID and your access permissions.'));
                    }
                    if (command.verbose) {
                        print_1.default.lf(`Identified as: ${idType}`);
                    }
                    next(null, identifiedId, idType, options);
                });
            },
            (id, idType, options, next) => {
                let normalizedStatus;
                if (command.status) {
                    const statusUpper = command.status.toUpperCase().replace(/-/g, '_'), validStatuses = ['IN_PROGRESS', 'SUCCESS', 'FAILURE'], statusMap = {
                        INPROGRESS: 'IN_PROGRESS',
                        'IN PROGRESS': 'IN_PROGRESS',
                        FAIL: 'FAILURE',
                        FAILED: 'FAILURE'
                    };
                    normalizedStatus = statusMap[statusUpper] || statusUpper;
                    if (!validStatuses.includes(normalizedStatus)) {
                        return next(new Error(`Invalid status: "${command.status}". ` +
                            'Must be one of: IN_PROGRESS, SUCCESS, FAILURE (case-insensitive)'));
                    }
                }
                const service = (0, sdk_generator_service_1.createSDKGeneratorService)({
                    accessToken: options.accessToken,
                    postmanApiKey: options.postmanApiKey
                }), listOptions = {};
                if (idType === 'collection') {
                    listOptions.collectionId = id;
                }
                else if (idType === 'specification') {
                    listOptions.specificationId = id;
                }
                if (command.verbose) {
                    print_1.default.lf('Fetching builds from Postman...');
                    if (idType === 'collection' && (0, sdk_utils_1.isExtensibleCollectionId)(id)) {
                        print_1.default.lf(`Extensible Collection ID: ${id}`);
                    }
                    else if (idType === 'collection') {
                        print_1.default.lf(`Collection ID: ${id}`);
                    }
                    else if (idType === 'specification') {
                        print_1.default.lf(`Specification ID: ${id}`);
                    }
                    if (normalizedStatus) {
                        print_1.default.lf(`Status filter: ${normalizedStatus}`);
                    }
                }
                print_1.default.wait(safe_1.default.cyan);
                if (normalizedStatus) {
                    listOptions.status = normalizedStatus;
                }
                service.listBuilds(listOptions)
                    .then(async (result) => {
                    const builds = result.data || [], rateLimitInfo = result.rateLimitInfo;
                    print_1.default.unwait();
                    if (command.verbose) {
                        if (rateLimitInfo.remaining !== undefined) {
                            print_1.default.lf(safe_1.default.gray(`Rate limit: ${rateLimitInfo.remaining}/${rateLimitInfo.limit} remaining`));
                        }
                        print_1.default.lf(safe_1.default.gray(`Raw response: ${JSON.stringify(result.data, null, 2)}`));
                    }
                    if (builds.length === 0) {
                        print_1.default.lf('No builds found.');
                        print_1.default.lf('\nGenerate SDKs with:');
                        print_1.default.lf('  postman sdk generate <COLLECTION_ID_OR_SPEC_ID> ' +
                            '--language typescript python');
                        return next(null);
                    }
                    if (command.verbose) {
                        print_1.default.lf(safe_1.default.gray(`Fetching details for ${builds.length} build(s)...`));
                    }
                    const buildDetailsPromises = builds.map((build) => {
                        return service.getBuild(build.id)
                            .then(({ build: fullBuild }) => {
                            return { build, fullBuild };
                        })
                            .catch((err) => {
                            if (command.verbose) {
                                const warningMsg = 'Warning: Could not fetch details for build ' +
                                    `${build.id}: ${err.message}`;
                                print_1.default.lf(safe_1.default.yellow(warningMsg));
                            }
                            return { build: build, fullBuild: null };
                        });
                    }), buildDetailsResults = await Promise.allSettled(buildDetailsPromises), buildsWithDetails = buildDetailsResults.map((result, index) => {
                        if (result.status === 'fulfilled') {
                            return result.value;
                        }
                        return { build: builds[index], fullBuild: null };
                    });
                    if (command.output === 'json') {
                        const jsonBuilds = buildsWithDetails.map(({ build, fullBuild }) => {
                            return fullBuild || build;
                        });
                        if (service.getBuildSummary) {
                            const failedJsonBuilds = jsonBuilds.filter((b) => {
                                return b.status === 'FAILURE';
                            });
                            const summaryPromises = failedJsonBuilds.map((b) => {
                                return service.getBuildSummary(undefined, undefined, b.id)
                                    .then(({ buildSummary }) => {
                                    if (buildSummary) {
                                        b.summary = buildSummary;
                                    }
                                })
                                    .catch(() => {
                                });
                            });
                            await Promise.allSettled(summaryPromises);
                        }
                        console.log(JSON.stringify({
                            builds: jsonBuilds,
                            total: jsonBuilds.length
                        }, null, 2));
                        return next(null);
                    }
                    print_1.default.lf(`\nFound ${builds.length} build${builds.length === 1 ? '' : 's'}:\n`);
                    const idWidth = 10, statusWidth = 15, languagesWidth = 25, dateWidth = 25, header = [
                        padRight('Build ID', idWidth),
                        padRight('Status', statusWidth),
                        padRight('Languages', languagesWidth),
                        padRight('Created', dateWidth)
                    ].join(' │ ');
                    print_1.default.lf(safe_1.default.bold(header));
                    print_1.default.lf('─'.repeat(header.length));
                    buildsWithDetails.forEach(({ build, fullBuild }) => {
                        if (!build) {
                            return;
                        }
                        const statusColor = build.status === 'SUCCESS' ? safe_1.default.green :
                            build.status === 'IN_PROGRESS' ? safe_1.default.yellow : safe_1.default.red, createdDate = build.createdAt ?
                            new Date(build.createdAt).toLocaleString('en-US', {
                                month: 'short',
                                day: 'numeric',
                                year: 'numeric',
                                hour: '2-digit',
                                minute: '2-digit'
                            }) : '-';
                        let languages;
                        if (fullBuild) {
                            if (fullBuild.languages && fullBuild.languages.length > 0) {
                                languages = fullBuild.languages.join(', ');
                            }
                            else if (fullBuild.artifacts && fullBuild.artifacts.length > 0) {
                                const languageSet = new Set();
                                for (const artifact of fullBuild.artifacts) {
                                    const hasSdkLanguage = artifact.sdk && artifact.sdk.language, hasBucketKey = artifact.bucketKey;
                                    if (hasSdkLanguage) {
                                        languageSet.add(artifact.sdk.language.toLowerCase());
                                        continue;
                                    }
                                    if (!hasBucketKey) {
                                        continue;
                                    }
                                    const lang = (0, sdk_utils_1.extractLanguageFromBucketKey)(artifact.bucketKey);
                                    if (!lang) {
                                        continue;
                                    }
                                    languageSet.add(lang.toLowerCase());
                                }
                                languages = languageSet.size > 0 ? Array.from(languageSet).join(', ') : '-';
                            }
                            else {
                                languages = '-';
                            }
                        }
                        else {
                            languages = '-';
                        }
                        const row = [
                            padRight(String(build.id || '-'), idWidth),
                            statusColor(padRight(String(build.status || '-'), statusWidth)),
                            padRight(truncate(languages, languagesWidth), languagesWidth),
                            padRight(createdDate, dateWidth)
                        ].join(' │ ');
                        print_1.default.lf(row);
                    });
                    const failedBuilds = buildsWithDetails.filter(({ build }) => {
                        return build.status === 'FAILURE';
                    });
                    if (failedBuilds.length > 0 && service.getBuildSummary) {
                        for (const { build: failedBuild } of failedBuilds) {
                            try {
                                const { buildSummary } = await service.getBuildSummary(undefined, undefined, failedBuild.id);
                                if (buildSummary) {
                                    print_1.default.lf(safe_1.default.gray(`\nBuild ${failedBuild.id} details:`));
                                    (0, sdk_utils_1.displayBuildSummaryDetails)(buildSummary, command.logLevel === 'info' ? 'info' : 'warn');
                                }
                            }
                            catch {
                            }
                        }
                    }
                    const successfulBuilds = buildsWithDetails.filter(({ build }) => {
                        return build.status === 'SUCCESS';
                    }), mostRecentSuccessful = successfulBuilds.length > 0 ?
                        successfulBuilds.reduce((latest, current) => {
                            return new Date(current.build.createdAt || 0) >
                                new Date(latest.build.createdAt || 0) ? current : latest;
                        }) : null;
                    print_1.default.lf('\nTo download SDKs from a build:');
                    print_1.default.lf('  postman sdk get <BUILD_ID> --output-dir ./sdks');
                    if (mostRecentSuccessful) {
                        print_1.default.lf('');
                        try {
                            const response = await (0, prompts_1.default)({
                                type: 'confirm',
                                name: 'download',
                                message: 'Download SDKs from the most recent successful build ' +
                                    `(${mostRecentSuccessful.build.id})?`,
                                initial: true
                            });
                            if (response.download) {
                                print_1.default.lf(safe_1.default.cyan('Downloading SDKs...'));
                                try {
                                    let fullBuild = mostRecentSuccessful.fullBuild;
                                    if (!fullBuild) {
                                        const buildId = mostRecentSuccessful.build.id, { build: fetchedBuild } = await service.getBuild(buildId);
                                        fullBuild = fetchedBuild;
                                    }
                                    await (0, sdk_utils_1.downloadBuildArtifacts)(fullBuild, './sdks', service, true);
                                    print_1.default.lf(safe_1.default.green('✓ SDKs downloaded successfully to ./sdks'));
                                }
                                catch (fetchErr) {
                                    const errorMsg = `Error fetching build details: ${fetchErr.message}`;
                                    print_1.default.lf(safe_1.default.red(errorMsg));
                                    throw fetchErr;
                                }
                            }
                        }
                        catch (downloadErr) {
                            print_1.default.lf(safe_1.default.red(`Error downloading SDKs: ${downloadErr.message}`));
                            return next(downloadErr);
                        }
                    }
                    return next(null);
                })
                    .catch((err) => {
                    print_1.default.unwait();
                    let errorMessage = err.message || 'Failed to list builds';
                    if (err.statusCode === 429) {
                        errorMessage = 'Failed to list builds: Rate limit exceeded. Please try again later.';
                    }
                    else if (err.statusCode === 404) {
                        errorMessage = 'No builds found for this collection.';
                    }
                    else if (err.statusCode >= 500) {
                        errorMessage = 'Server error while fetching builds. Please try again later.';
                    }
                    if (command.verbose && err.statusCode) {
                        errorMessage += `\nHTTP Status: ${err.statusCode}`;
                    }
                    return next(new Error(errorMessage));
                });
            }
        ], (err) => {
            if (err) {
                print_1.default.lf(safe_1.default.red(`Error: ${err.message || err}`));
                exitCode = 1;
            }
            process.exit(exitCode);
        });
    });
}
