// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/pr-workflow.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPRsForIntegrations = createPRsForIntegrations;
const node_path_1 = __importDefault(require("node:path"));
const node_fs_1 = __importDefault(require("node:fs"));
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const rest_1 = require("@octokit/rest");
const git_operations_1 = require("./git-operations");
const integrations_service_1 = require("../../api/integrations-service");
function createPRDataFromIntegration(integration, language, sdkVersion, _outputDir) {
    const variables = integration.variables || {};
    const sensitiveInputs = integration.sensitiveInputs || {};
    const accessToken = sensitiveInputs.access_token || variables.access_token;
    if (!accessToken) {
        return null;
    }
    const owner = variables.owner;
    const repo = variables.repository;
    const targetBranch = variables.branch || 'main';
    if (!owner || !repo) {
        return null;
    }
    const currentDate = new Date();
    const branchName = sdkVersion ?
        `sdk-update-v${sdkVersion}-${currentDate.getTime()}` :
        `sdk-update-${currentDate.getTime()}`;
    const commitMessage = sdkVersion ?
        `SDK update for ${language} v${sdkVersion}` :
        `SDK update for ${language}`;
    const title = sdkVersion ?
        `SDK update: ${language} v${sdkVersion}` :
        `SDK update: ${language}`;
    const body = '## Automatically generated PR with SDK updates\n\n' +
        `- Language: **${language}**\n` +
        (sdkVersion ? `- SDK version: **v${sdkVersion}**\n` : '') +
        '\nThis PR was automatically created by Postman CLI.';
    const authorName = owner;
    const authorEmail = `${owner}@users.noreply.github.com`;
    const repoUrl = `https://github.com/${owner}/${repo}.git`;
    return {
        owner: owner,
        repo: repo,
        branch: branchName,
        targetBranch: targetBranch,
        branchName: branchName,
        title: title,
        body: body,
        commitMessage: commitMessage,
        accessToken: accessToken,
        authorName: authorName,
        authorEmail: authorEmail,
        repoUrl: repoUrl
    };
}
function copySDKFiles(sourceDir, repoRoot, language, targetDirectory) {
    const languageDir = node_path_1.default.join(sourceDir, language);
    if (!node_fs_1.default.existsSync(languageDir)) {
        throw new Error(`SDK directory not found: ${languageDir}`);
    }
    let normalizedTargetDir = targetDirectory.trim();
    if (normalizedTargetDir === '.' || normalizedTargetDir === './' || normalizedTargetDir === '') {
        normalizedTargetDir = '';
    }
    else {
        if (normalizedTargetDir.startsWith('./')) {
            normalizedTargetDir = normalizedTargetDir.substring(2);
        }
        if (!normalizedTargetDir.endsWith('/')) {
            normalizedTargetDir = `${normalizedTargetDir}/`;
        }
    }
    const targetPath = normalizedTargetDir ?
        node_path_1.default.join(repoRoot, normalizedTargetDir) :
        repoRoot;
    const copiedFiles = [];
    function copyRecursive(src, dest) {
        const stats = node_fs_1.default.statSync(src);
        if (stats.isDirectory()) {
            if (!node_fs_1.default.existsSync(dest)) {
                node_fs_1.default.mkdirSync(dest, { recursive: true });
            }
            const files = node_fs_1.default.readdirSync(src);
            for (const file of files) {
                copyRecursive(node_path_1.default.join(src, file), node_path_1.default.join(dest, file));
            }
        }
        else {
            const destDirPath = node_path_1.default.dirname(dest);
            if (!node_fs_1.default.existsSync(destDirPath)) {
                node_fs_1.default.mkdirSync(destDirPath, { recursive: true });
            }
            node_fs_1.default.copyFileSync(src, dest);
            const relativePath = node_path_1.default.relative(repoRoot, dest);
            copiedFiles.push(relativePath);
        }
    }
    const filesInLanguageDir = node_fs_1.default.readdirSync(languageDir);
    let effectiveSourceDir = languageDir;
    if (filesInLanguageDir.length === 1) {
        const potentialSubDir = node_path_1.default.join(languageDir, filesInLanguageDir[0]);
        if (node_fs_1.default.statSync(potentialSubDir).isDirectory()) {
            effectiveSourceDir = potentialSubDir;
        }
    }
    const filesToCopy = node_fs_1.default.readdirSync(effectiveSourceDir);
    for (const file of filesToCopy) {
        const srcPath = node_path_1.default.join(effectiveSourceDir, file);
        const destPath = node_path_1.default.join(targetPath, file);
        copyRecursive(srcPath, destPath);
    }
    return copiedFiles;
}
async function createPRForIntegration(integration, language, sdkVersion, outputDir, _authOptions) {
    const prData = createPRDataFromIntegration(integration, language, sdkVersion, outputDir);
    if (!prData) {
        const integrationId = typeof integration.id === 'number' ? integration.id : integration.id;
        throw new Error(`Integration ${integrationId} is missing required configuration (owner, repo, or access token)`);
    }
    try {
        const octokit = new rest_1.Octokit({ auth: prData.accessToken });
        const { data: user } = await octokit.users.getAuthenticated();
        prData.authorName = user.login;
        prData.authorEmail = `${user.id}+${user.login}@users.noreply.github.com`;
    }
    catch (error) {
        print_1.default.lf(safe_1.default.yellow(`Warning: Could not fetch GitHub user info: ${error.message}`));
    }
    const gitOps = new git_operations_1.GitOperations(prData);
    try {
        await gitOps.clean();
        print_1.default.lf(safe_1.default.cyan(`Cloning repository ${prData.owner}/${prData.repo}...`));
        await gitOps.clone();
        print_1.default.lf(safe_1.default.cyan(`Creating branch ${prData.branchName}...`));
        await gitOps.checkout();
        const repoPath = gitOps.getLocalRepoPath();
        const integrationVariables = integration.variables || {};
        const targetDirectory = integrationVariables.directory || 'sdk-directory/';
        const copiedFiles = copySDKFiles(outputDir, repoPath, language, targetDirectory);
        if (copiedFiles.length === 0) {
            throw new Error('No files were copied to the repository');
        }
        print_1.default.lf(safe_1.default.green(`✓ Copied ${copiedFiles.length} file(s) to repository`));
        print_1.default.lf(safe_1.default.cyan('Staging changes...'));
        await gitOps.stage(copiedFiles);
        print_1.default.lf(safe_1.default.cyan('Committing changes...'));
        await gitOps.commit();
        print_1.default.lf(safe_1.default.cyan('Pushing branch...'));
        await gitOps.push();
        print_1.default.lf(safe_1.default.cyan('Creating pull request...'));
        const pr = await gitOps.createPullRequest();
        print_1.default.lf(safe_1.default.green(`✓ Pull request created: ${pr.html_url}`));
        await gitOps.clean();
        return {
            htmlUrl: pr.html_url,
            integration: integration
        };
    }
    catch (error) {
        await gitOps.clean();
        throw error;
    }
}
async function createPRsForIntegrations(id, languages, sdkVersion, outputDir, authOptions, entityType = 'collection') {
    const integrationsService = (0, integrations_service_1.createIntegrationsService)(authOptions);
    print_1.default.lf(safe_1.default.cyan('\nLooking up integrations...'));
    const allIntegrations = await integrationsService.getIntegrations(id, undefined, entityType);
    if (allIntegrations.length === 0) {
        const entityTypeName = entityType === 'collection' ? 'collection' : 'specification';
        print_1.default.lf(safe_1.default.yellow(`No integrations found for this ${entityTypeName}. Skipping PR creation.`));
        return;
    }
    const prPromises = [];
    for (const language of languages) {
        let languageIntegrations = allIntegrations.filter((integration) => {
            return integration.variables?.language === language;
        });
        if (languageIntegrations.length === 0) {
            print_1.default.lf(safe_1.default.yellow(`No integrations found for language ${language}. Skipping.`));
            continue;
        }
        const seen = new Map();
        const duplicates = [];
        for (const integration of languageIntegrations) {
            const owner = integration.variables?.owner || '';
            const repo = integration.variables?.repository || '';
            const branch = integration.variables?.branch || 'main';
            const key = `${owner}/${repo}:${branch}:${language}`;
            if (seen.has(key)) {
                const existing = seen.get(key);
                if (!existing) {
                    continue;
                }
                const currentId = typeof integration.id === 'number' ?
                    integration.id :
                    parseInt(String(integration.id), 10);
                const existingId = typeof existing.id === 'number' ?
                    existing.id :
                    parseInt(String(existing.id), 10);
                if (currentId > existingId) {
                    duplicates.push(existing);
                    seen.set(key, integration);
                }
                else {
                    duplicates.push(integration);
                }
            }
            else {
                seen.set(key, integration);
            }
        }
        languageIntegrations = Array.from(seen.values());
        if (duplicates.length > 0) {
            print_1.default.lf(safe_1.default.yellow(`Removed ${duplicates.length} duplicate integration(s) for ${language}:`));
            for (const duplicate of duplicates) {
                const owner = duplicate.variables?.owner || 'unknown';
                const repo = duplicate.variables?.repository || 'unknown';
                const branch = duplicate.variables?.branch || 'main';
                const duplicateId = typeof duplicate.id === 'number' ? duplicate.id : duplicate.id;
                print_1.default.lf(safe_1.default.gray(`  - Integration ${duplicateId}: ${owner}/${repo} (${branch})`));
            }
        }
        print_1.default.lf(safe_1.default.gray(`Found ${languageIntegrations.length} integration(s) for ${language}:`));
        for (const integration of languageIntegrations) {
            const owner = integration.variables?.owner || 'unknown';
            const repo = integration.variables?.repository || 'unknown';
            const branch = integration.variables?.branch || 'main';
            const integrationId = typeof integration.id === 'number' ? integration.id : integration.id;
            print_1.default.lf(safe_1.default.gray(`  - Integration ${integrationId}: ${owner}/${repo} (${branch})`));
        }
        for (const integration of languageIntegrations) {
            prPromises.push(integrationsService.getIntegrationById(String(integration.id))
                .then((fullIntegration) => {
                return createPRForIntegration(fullIntegration, language, sdkVersion, outputDir, authOptions);
            })
                .catch((error) => {
                const integrationId = typeof integration.id === 'number' ? integration.id : integration.id;
                print_1.default.lf(safe_1.default.red(`✗ Failed to create PR for integration ${integrationId} ` +
                    `(${language}): ${error.message}`));
                throw error;
            }));
        }
    }
    if (prPromises.length === 0) {
        print_1.default.lf(safe_1.default.yellow('No matching integrations found. Skipping PR creation.'));
        return;
    }
    print_1.default.lf(safe_1.default.cyan(`\nCreating ${prPromises.length} pull request(s)...`));
    const results = await Promise.allSettled(prPromises);
    let successCount = 0;
    let failureCount = 0;
    for (const result of results) {
        if (result.status === 'fulfilled') {
            successCount++;
        }
        else {
            failureCount++;
        }
    }
    if (successCount > 0) {
        print_1.default.lf(safe_1.default.green(`\n✓ Successfully created ${successCount} pull request(s)`));
    }
    if (failureCount > 0) {
        print_1.default.lf(safe_1.default.red(`\n✗ Failed to create ${failureCount} pull request(s)`));
    }
}
