// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/sdk-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUPPORTED_LANGUAGES = void 0;
exports.isExtensibleCollectionId = isExtensibleCollectionId;
exports.validateLanguage = validateLanguage;
exports.parseLanguages = parseLanguages;
exports.isUrl = isUrl;
exports.validateUrlProtocol = validateUrlProtocol;
exports.fetchUrlToTempFile = fetchUrlToTempFile;
exports.isFilePath = isFilePath;
exports.getEntityTypeFromFile = getEntityTypeFromFile;
exports.extractCollectionId = extractCollectionId;
exports.identifyIdType = identifyIdType;
exports.validateCollection = validateCollection;
exports.sdkExistsAtPath = sdkExistsAtPath;
exports.validateOutputFormat = validateOutputFormat;
exports.downloadFromArtifactUrl = downloadFromArtifactUrl;
exports.extractLanguageFromBucketKey = extractLanguageFromBucketKey;
exports.downloadAndExtractSDK = downloadAndExtractSDK;
exports.downloadBuildArtifacts = downloadBuildArtifacts;
exports.displayBuildSummaryDetails = displayBuildSummaryDetails;
exports.fetchAndDisplayBuildSummary = fetchAndDisplayBuildSummary;
exports.detectV3Collection = detectV3Collection;
exports.mergeV3CollectionToTempFile = mergeV3CollectionToTempFile;
exports.prepareCollectionFile = prepareCollectionFile;
exports.cleanupTempFiles = cleanupTempFiles;
exports.handleBuildCompletion = handleBuildCompletion;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const cmd_util_1 = __importDefault(require("../../commands/cmd-util"));
const postman_resource_1 = __importStar(require("../../api/postman-resource"));
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const adm_zip_1 = __importDefault(require("adm-zip"));
const request_1 = __importDefault(require("@postman/platform-req/lib/request"));
const sdk_generator_service_1 = require("../../api/sdk-generator-service");
const util_1 = __importDefault(require("../../util"));
const v3_collection_merge_1 = require("./v3-collection-merge");
const v3_to_v2_transform_1 = require("./v3-to-v2-transform");
const v3_filesystem_collection_1 = require("./v3-filesystem-collection");
function checkSpecificationExists(specId, options, callback) {
    const headers = {
        'User-Agent': util_1.default.userAgent,
        'x-pstmn-req-service': 'api-specification'
    };
    if (options.postmanApiKey) {
        headers['x-api-key'] = options.postmanApiKey;
    }
    else if (options.accessToken) {
        headers['x-access-token'] = options.accessToken;
    }
    const client = request_1.default.defaults({
        timeout: 30000,
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: { enabled: true }
            }
        }
    });
    const url = `${util_1.default.POSTMAN_GATEWAY_BASE_URL()}/specifications/${specId}`;
    client.get({
        url,
        headers
    }, (err, response) => {
        if (err) {
            return callback(err, false);
        }
        const exists = (/2../).test(response.statusCode);
        callback(null, exists);
    });
}
const EXTENSIBLE_COLLECTION_OBJECT_ID_REGEX = /^[0-9a-f]{24}$/i;
function checkExtensibleCollectionExists(collectionId, options, callback) {
    const headers = {
        'User-Agent': util_1.default.userAgent,
        'x-pstmn-req-service': 'collection-api'
    };
    if (options.postmanApiKey) {
        headers['X-Api-Key'] = options.postmanApiKey;
    }
    else if (options.accessToken) {
        headers['x-access-token'] = options.accessToken;
    }
    const client = request_1.default.defaults({
        timeout: 30000,
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: { enabled: true }
            }
        }
    });
    const url = `${util_1.default.POSTMAN_GATEWAY_BASE_URL()}/collection-exports/${collectionId}`;
    client.get({
        url,
        headers
    }, (err, response) => {
        if (err) {
            return callback(err, false);
        }
        const exists = (/^2\d{2}$/).test(String(response.statusCode));
        callback(null, exists);
    });
}
function isExtensibleCollectionId(id) {
    return (typeof id === 'string' && EXTENSIBLE_COLLECTION_OBJECT_ID_REGEX.test(id));
}
exports.SUPPORTED_LANGUAGES = ['typescript', 'python', 'java'];
function validateLanguage(language) {
    return exports.SUPPORTED_LANGUAGES.includes(language);
}
function parseLanguages(languageInput) {
    const languageArray = Array.isArray(languageInput) ?
        languageInput :
        [languageInput], invalidLanguages = [], validLanguages = [];
    for (const lang of languageArray) {
        if (validateLanguage(lang)) {
            if (!validLanguages.includes(lang)) {
                validLanguages.push(lang);
            }
        }
        else {
            invalidLanguages.push(lang);
        }
    }
    if (invalidLanguages.length > 0) {
        throw new Error(`Unsupported language(s): ${invalidLanguages.join(', ')}. ` +
            `Supported languages: ${exports.SUPPORTED_LANGUAGES.join(', ')}`);
    }
    if (validLanguages.length === 0) {
        throw new Error('At least one language must be specified.');
    }
    return validLanguages;
}
function detectUuidType(id) {
    const standardUuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    const prefixedUuidRegex = /^[0-9a-f]{8}-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    if (standardUuidRegex.test(id)) {
        return 'specification';
    }
    if (prefixedUuidRegex.test(id)) {
        return 'collection';
    }
    if (EXTENSIBLE_COLLECTION_OBJECT_ID_REGEX.test(id)) {
        return 'collection';
    }
    return 'unknown';
}
function isUrl(input) {
    if (!input || typeof input !== 'string') {
        return false;
    }
    const trimmed = input.trim();
    return trimmed.startsWith('http://') || trimmed.startsWith('https://');
}
const LOOPBACK_HOSTNAMES = new Set(['localhost', '127.0.0.1', '::1', '[::1]']);
function validateUrlProtocol(url) {
    let parsed;
    try {
        parsed = new URL(url);
    }
    catch {
        throw new Error(`Invalid URL: ${url}`);
    }
    if (parsed.protocol === 'https:') {
        return;
    }
    if (parsed.protocol === 'http:') {
        if (!LOOPBACK_HOSTNAMES.has(parsed.hostname)) {
            throw new Error('HTTP URLs are restricted to loopback addresses (localhost, 127.0.0.1, ::1) for security. ' +
                'Use HTTPS for remote URLs, or download the file first and pass a local path.');
        }
        print_1.default.lf(safe_1.default.yellow('Warning: Fetching over unencrypted HTTP. Consider using HTTPS.'));
        return;
    }
    throw new Error(`Unsupported URL protocol "${parsed.protocol}". Only HTTPS and HTTP (loopback only) are supported.`);
}
function fetchUrlToTempFile(url) {
    return new Promise((resolve, reject) => {
        try {
            validateUrlProtocol(url);
        }
        catch (err) {
            reject(err);
            return;
        }
        let urlPath;
        let urlExt;
        try {
            const parsedUrl = new URL(url);
            urlPath = parsedUrl.pathname;
            urlExt = path_1.default.extname(urlPath) || '.json';
        }
        catch (urlErr) {
            reject(new Error(`Invalid URL: ${urlErr.message}`));
            return;
        }
        const client = request_1.default.defaults({
            timeout: 30000,
            retry: {
                enabled: true,
                delay: 50,
                backOffNoise: 100,
                maxAttempts: 2,
                maxDelay: 1000,
                statusCodes: [500, 502, 503, 504],
                methods: {
                    GET: { enabled: true }
                }
            }
        }), tempDir = os_1.default.tmpdir(), tempFileName = `postman-sdk-${Date.now()}-${Math.random().toString(36).substring(7)}${urlExt}`, tempFilePath = path_1.default.join(tempDir, tempFileName);
        const options = {
            method: 'GET',
            url: url,
            followRedirect: true,
            encoding: 'utf8'
        };
        client(options, (err, response, body) => {
            if (err) {
                return reject(new Error(`Failed to fetch URL: ${err.message}`));
            }
            if (response.statusCode >= 400) {
                return reject(new Error(`HTTP ${response.statusCode}: Failed to fetch URL. ` +
                    'The URL may not be publicly accessible.'));
            }
            try {
                if (!fs_1.default.existsSync(tempDir)) {
                    fs_1.default.mkdirSync(tempDir, { recursive: true });
                }
                fs_1.default.writeFileSync(tempFilePath, body, 'utf8');
                resolve(tempFilePath);
            }
            catch (writeErr) {
                reject(new Error(`Failed to save fetched content to temporary file: ${writeErr.message}`));
            }
        });
    });
}
function isFilePath(input) {
    if (!input || typeof input !== 'string') {
        return false;
    }
    if (isUrl(input)) {
        return false;
    }
    if (input.endsWith('.json') ||
        input.endsWith('.yaml') ||
        input.endsWith('.yml')) {
        return true;
    }
    if (input.includes('/') || input.includes('\\')) {
        return true;
    }
    if (fs_1.default.existsSync(input)) {
        try {
            const stats = fs_1.default.statSync(input);
            return stats.isFile();
        }
        catch {
            return false;
        }
    }
    return false;
}
function getEntityTypeFromFile(filePath) {
    if (!fs_1.default.existsSync(filePath)) {
        throw new Error(`File not found: ${filePath}`);
    }
    if (path_1.default.basename(filePath).match(/^collection\.ya?ml$/i)) {
        if ((0, v3_collection_merge_1.isV3Collection)(filePath)) {
            return 'collection';
        }
    }
    if (fs_1.default.statSync(filePath).isDirectory()) {
        const v3CollectionPath = (0, v3_collection_merge_1.findV3CollectionInDirectory)(filePath);
        if (v3CollectionPath) {
            return 'collection';
        }
        if ((0, v3_filesystem_collection_1.isV3FilesystemCollectionRoot)(filePath)) {
            return 'collection';
        }
        throw new Error(`Unable to determine entity type from directory: ${filePath}. ` +
            'No collection.yaml or v3 filesystem layout (*.request.yaml / .resources/) found.');
    }
    try {
        const fileContent = fs_1.default.readFileSync(filePath, 'utf-8');
        let parsed;
        if (filePath.endsWith('.json')) {
            parsed = JSON.parse(fileContent);
        }
        else if (filePath.endsWith('.yaml') || filePath.endsWith('.yml')) {
            return 'specification';
        }
        else {
            try {
                parsed = JSON.parse(fileContent);
            }
            catch {
                throw new Error(`Unable to parse file: ${filePath}. ` +
                    'Expected a JSON collection or specification file.');
            }
        }
        if (parsed.info && parsed.item) {
            return 'collection';
        }
        if (parsed.openapi) {
            return 'specification';
        }
        if (parsed.swagger) {
            return 'specification';
        }
        if (parsed.info && parsed.paths) {
            return 'specification';
        }
        if (parsed.info) {
            return 'collection';
        }
        throw new Error(`Unable to determine entity type from file: ${filePath}. ` +
            'File does not appear to be a valid Postman Collection or OpenAPI specification.');
    }
    catch (err) {
        if (err.message.includes('Unable to determine') ||
            err.message.includes('File not found')) {
            throw err;
        }
        throw new Error(`Failed to read or parse file: ${filePath}. ${err.message}`);
    }
}
function extractCollectionId(id, callback) {
    if (!id || typeof id !== 'string') {
        return callback(new Error('Invalid ID provided. Expected a Postman Collection ID or Specification ID.'));
    }
    if (id.startsWith('http://') || id.startsWith('https://')) {
        return callback(new Error('URLs should be handled separately. Please use fetchUrlToTempFile() first.'));
    }
    const extractedUid = cmd_util_1.default.extractCollectionUIDFromCLIArg(id);
    if (extractedUid) {
        return callback(null, extractedUid);
    }
    const detectedType = detectUuidType(id);
    if (detectedType === 'specification' || detectedType === 'collection') {
        return callback(null, id);
    }
    return callback(new Error(`Invalid ID: "${id}". ` +
        'Expected a Postman Collection ID or Specification ID (UUID or ObjectId format).'));
}
function identifyIdType(id, options, callback) {
    if (!id) {
        return callback(new Error('ID is required for identification.'));
    }
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const likelyType = detectUuidType(id);
    if (isExtensibleCollectionId(id)) {
        checkExtensibleCollectionExists(id, options, (ecErr, ecExists) => {
            if (ecExists) {
                return callback(null, 'collection', id);
            }
            return callback(new Error(`ID "${id}" not found as a collection or specification. ` +
                'Please verify the ID and your access permissions.'), 'unknown');
        });
        return;
    }
    const checkCollectionFirst = likelyType === 'collection' || likelyType === 'unknown';
    if (checkCollectionFirst) {
        const collectionResource = new postman_resource_1.default(postman_resource_1.default.COLLECTION, id, options.postmanApiKey, options.accessToken);
        collectionResource.exists((collectionErr, isCollection) => {
            if (isCollection) {
                return callback(null, 'collection', id);
            }
            checkSpecificationExists(id, options, (specErr, isSpecification) => {
                if (isSpecification) {
                    return callback(null, 'specification', id);
                }
                return callback(new Error(`ID "${id}" not found as a collection or specification. ` +
                    'Please verify the ID and your access permissions.'), 'unknown');
            });
        });
    }
    else {
        checkSpecificationExists(id, options, (specErr, isSpecification) => {
            if (isSpecification) {
                return callback(null, 'specification', id);
            }
            const collectionResource = new postman_resource_1.default(postman_resource_1.default.COLLECTION, id, options.postmanApiKey, options.accessToken);
            collectionResource.exists((collectionErr, isCollection) => {
                if (isCollection) {
                    return callback(null, 'collection', id);
                }
                return callback(new Error(`ID "${id}" not found as a collection or specification. ` +
                    'Please verify the ID and your access permissions.'), 'unknown');
            });
        });
    }
}
function validateCollection(collectionId, options, callback) {
    if (!collectionId) {
        return callback(new Error('Collection ID is required for validation.'));
    }
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new postman_resource_1.default(postman_resource_1.COLLECTION, collectionId, options.postmanApiKey, options.accessToken);
    resource.exists((err, exists) => {
        if (err) {
            return callback(err, false);
        }
        if (!exists) {
            return callback(new Error(`Collection "${collectionId}" not found. ` +
                'Please verify the collection ID and your access permissions.'), false);
        }
        callback(null, true);
    });
}
function sdkExistsAtPath(outputPath) {
    if (!outputPath || typeof outputPath !== 'string') {
        return false;
    }
    try {
        if (!fs_1.default.existsSync(outputPath)) {
            return false;
        }
        const stats = fs_1.default.statSync(outputPath);
        if (!stats.isDirectory()) {
            return false;
        }
        const contents = fs_1.default.readdirSync(outputPath);
        return contents.length > 0;
    }
    catch {
        return false;
    }
}
function validateOutputFormat(format) {
    const validFormats = ['json', 'table'];
    if (!validFormats.includes(format)) {
        throw new Error(`Invalid output format: ${format}. Valid formats: ${validFormats.join(', ')}`);
    }
    return format;
}
function downloadFromArtifactUrl(artifactUrl, outputPath) {
    return new Promise((resolve, reject) => {
        const client = request_1.default.defaults({
            retry: {
                enabled: true,
                delay: 50,
                backOffNoise: 100,
                maxAttempts: 2,
                maxDelay: 1000,
                statusCodes: [500, 502, 503, 504],
                methods: {
                    GET: { enabled: true }
                }
            }
        }), options = {
            method: 'GET',
            url: artifactUrl,
            followRedirect: true,
            encoding: null
        };
        client(options, (err, response, body) => {
            if (err) {
                return reject(new Error(`Download failed: ${err.message}`));
            }
            if (response.statusCode >= 400) {
                return reject(new Error(`HTTP ${response.statusCode}: Failed to download artifact`));
            }
            try {
                const outputDir = path_1.default.dirname(outputPath);
                if (!fs_1.default.existsSync(outputDir)) {
                    fs_1.default.mkdirSync(outputDir, { recursive: true });
                }
                fs_1.default.writeFileSync(outputPath, body);
                resolve();
            }
            catch (writeErr) {
                reject(new Error(`Failed to write file: ${writeErr.message}`));
            }
        });
    });
}
function extractLanguageFromBucketKey(bucketKey) {
    const filename = bucketKey.split('/').pop() || '';
    const knownLanguages = ['typescript', 'python', 'java'];
    const match = filename.match(/-(\w+)-[\d.]+\.zip$/);
    if (match && match[1]) {
        const extractedLang = match[1].toLowerCase();
        if (knownLanguages.includes(extractedLang)) {
            return extractedLang;
        }
    }
    const lowerFilename = filename.toLowerCase();
    for (const lang of knownLanguages) {
        if (lowerFilename.includes(lang)) {
            return lang;
        }
    }
    throw new Error(`Could not extract language from bucket key: ${bucketKey}`);
}
async function downloadAndExtractSDK(artifact, outputDir, service, id, idType = 'collection') {
    let language;
    if (artifact.sdk) {
        language = artifact.sdk.language.toLowerCase();
    }
    else if (artifact.bucketKey) {
        language = extractLanguageFromBucketKey(artifact.bucketKey);
    }
    else {
        throw new Error('Artifact does not contain SDK metadata or bucket key');
    }
    const artifactUrl = await service.getSDKDownloadUrl(id, language, idType);
    const zipPath = path_1.default.join(outputDir, `${language}.zip`), extractDir = path_1.default.join(outputDir, language);
    if (fs_1.default.existsSync(extractDir)) {
        fs_1.default.rmSync(extractDir, { recursive: true, force: true });
    }
    await downloadFromArtifactUrl(artifactUrl, zipPath);
    const zip = new adm_zip_1.default(zipPath);
    zip.extractAllTo(extractDir, true);
    fs_1.default.unlinkSync(zipPath);
    print_1.default.lf(safe_1.default.green(`✓ ${language}: ${extractDir}`));
}
async function downloadBuildArtifacts(completedBuild, outputDir, service, skipConfirmation = false, forceUseBuildId = false) {
    if (!completedBuild.artifacts || completedBuild.artifacts.length === 0) {
        return;
    }
    const successfulArtifacts = completedBuild.artifacts.filter((artifact) => {
        return artifact.status === 'SUCCESS';
    });
    if (successfulArtifacts.length === 0) {
        return;
    }
    const existingDirs = [];
    for (const artifact of successfulArtifacts) {
        let language;
        try {
            if (artifact.sdk) {
                language = artifact.sdk.language.toLowerCase();
            }
            else if (artifact.bucketKey) {
                language = extractLanguageFromBucketKey(artifact.bucketKey);
            }
            else {
                continue;
            }
            const extractDir = path_1.default.join(outputDir, language);
            if (sdkExistsAtPath(extractDir)) {
                existingDirs.push(language);
            }
        }
        catch {
            continue;
        }
    }
    if (existingDirs.length > 0 && !skipConfirmation) {
        print_1.default.lf(safe_1.default.yellow(`\nThe following SDK directories already exist: ${existingDirs.join(', ')}`));
        const prompts = require('prompts');
        const response = await prompts({
            type: 'confirm',
            name: 'overwrite',
            message: 'Do you want to overwrite them?',
            initial: false
        });
        if (!response || !response.overwrite) {
            throw new Error('SDK generation cancelled. Existing directories were not overwritten.');
        }
    }
    print_1.default.lf('\nDownloading SDKs...');
    let downloadId;
    let downloadIdType;
    if (forceUseBuildId) {
        downloadId = completedBuild.id;
        downloadIdType = 'build';
    }
    else if (completedBuild.collectionId) {
        downloadId = completedBuild.collectionId;
        downloadIdType = 'collection';
    }
    else if (completedBuild.apiSpecId) {
        downloadId = completedBuild.apiSpecId;
        downloadIdType = 'specification';
    }
    else {
        downloadId = completedBuild.id;
        downloadIdType = 'build';
    }
    await Promise.all(successfulArtifacts.map(async (artifact) => {
        try {
            await downloadAndExtractSDK(artifact, outputDir, service, downloadId, downloadIdType);
        }
        catch (err) {
            let language = 'Unknown';
            try {
                if (artifact.sdk) {
                    language = artifact.sdk.language;
                }
                else if (artifact.bucketKey) {
                    language = extractLanguageFromBucketKey(artifact.bucketKey);
                }
            }
            catch {
            }
            print_1.default.lf(safe_1.default.red(`✗ ${language}: ${err.message}`));
        }
    }));
    print_1.default.lf(safe_1.default.green('\n✓ All SDKs downloaded successfully!'));
}
function wrapText(text, maxWidth, continuationIndent) {
    const words = text.split(/\s+/);
    if (words.length === 0) {
        return text;
    }
    const lines = [];
    let currentLine = words[0];
    for (let i = 1; i < words.length; i++) {
        const testLine = `${currentLine} ${words[i]}`;
        const effectiveWidth = lines.length === 0 ?
            maxWidth :
            maxWidth - continuationIndent.length;
        if (testLine.length <= effectiveWidth) {
            currentLine = testLine;
        }
        else {
            lines.push(currentLine);
            currentLine = words[i];
        }
    }
    lines.push(currentLine);
    return lines
        .map((line, idx) => {
        return idx === 0 ? line : `${continuationIndent}${line}`;
    })
        .join('\n');
}
function displayBuildSummaryDetails(summary, logLevel = 'warn') {
    if (summary.statusDetails?.message) {
        print_1.default.lf(safe_1.default.yellow(`\nBuild message: ${summary.statusDetails.message}`));
    }
    if (!summary.downloads || summary.downloads.length === 0) {
        return;
    }
    for (const download of summary.downloads) {
        if (download.status === 'failure' && download.statusDetails?.message) {
            print_1.default.lf(safe_1.default.red(`\n  \u2717 ${download.language}`));
            print_1.default.lf(safe_1.default.gray(`    ${download.statusDetails.message}`));
        }
    }
    const downloadWithLogs = summary.downloads.find((d) => {
        return d.logs && d.logs.length > 0;
    });
    if (!downloadWithLogs || !downloadWithLogs.logs) {
        return;
    }
    const seen = new Set();
    const uniqueLogs = [];
    for (const log of downloadWithLogs.logs) {
        const key = `${log.message}::${log.path || ''}`;
        if (!seen.has(key)) {
            seen.add(key);
            uniqueLogs.push(log);
        }
    }
    const isErrorOrWarning = (severity) => {
        return (severity === 'ERROR' ||
            severity === 'WARNING' ||
            severity === 'WARN');
    };
    const errorLogs = uniqueLogs.filter((l) => {
        return isErrorOrWarning(l.severity);
    });
    const infoLogs = uniqueLogs.filter((l) => {
        return !isErrorOrWarning(l.severity);
    });
    if (errorLogs.length === 0 && infoLogs.length === 0) {
        return;
    }
    print_1.default.lf('');
    print_1.default.lf(safe_1.default.bold('  Build Log Summary'));
    if (errorLogs.length > 0) {
        const endpointGroups = new Map();
        for (const log of errorLogs) {
            const endpoint = log.path || '(general)';
            if (!endpointGroups.has(endpoint)) {
                endpointGroups.set(endpoint, []);
            }
            endpointGroups.get(endpoint).push(log);
        }
        const MAX_ENDPOINTS = 10;
        const endpointEntries = Array.from(endpointGroups.entries());
        const displayCount = Math.min(endpointEntries.length, MAX_ENDPOINTS);
        const totalIssues = errorLogs.length;
        print_1.default.lf(safe_1.default.yellow(`\n  ${totalIssues} issue(s) across ${endpointEntries.length} endpoint(s):`));
        for (let i = 0; i < displayCount; i++) {
            const [endpoint, logs] = endpointEntries[i];
            print_1.default.lf('');
            print_1.default.lf(safe_1.default.yellow(`  ${i + 1}) "${endpoint}"`));
            for (const log of logs) {
                const severityColor = log.severity === 'ERROR' ? safe_1.default.red : safe_1.default.yellow;
                const severityLabel = log.severity === 'WARN' ? 'WARNING' : log.severity;
                print_1.default.lf(severityColor(`    ${severityLabel}: ${log.message}`));
            }
        }
        if (endpointEntries.length > MAX_ENDPOINTS) {
            print_1.default.lf('');
            print_1.default.lf(safe_1.default.gray(`  ... and ${endpointEntries.length - MAX_ENDPOINTS} more endpoint(s) with issues`));
        }
    }
    if (infoLogs.length > 0) {
        const uniqueMessages = [];
        const seenMessages = new Set();
        for (const log of infoLogs) {
            if (!seenMessages.has(log.message)) {
                seenMessages.add(log.message);
                uniqueMessages.push(log.message);
            }
        }
        const noteCount = uniqueMessages.length;
        if (logLevel !== 'info') {
            print_1.default.lf(safe_1.default.cyan(`\n  Notes: ${noteCount} note(s) (use --log-level info to expand)`));
        }
        else {
            const MAX_NOTES = 10;
            const displayCount = Math.min(uniqueMessages.length, MAX_NOTES);
            const bulletPrefix = '    \u2022 ';
            const wrapIndent = '      ';
            const maxWidth = 80;
            print_1.default.lf(safe_1.default.cyan(`\n  Notes (${noteCount}):`));
            for (let i = 0; i < displayCount; i++) {
                const wrapped = wrapText(uniqueMessages[i], maxWidth - bulletPrefix.length, wrapIndent);
                print_1.default.lf('');
                print_1.default.lf(`${bulletPrefix}${wrapped}`);
            }
            if (uniqueMessages.length > MAX_NOTES) {
                print_1.default.lf('');
                print_1.default.lf(safe_1.default.gray(`    ... and ${uniqueMessages.length - MAX_NOTES} more note(s)`));
            }
        }
    }
    print_1.default.lf('');
}
async function fetchAndDisplayBuildSummary(service, buildId, collectionId, specificationId, logLevel = 'warn') {
    if (!service || !service.getBuildSummary) {
        return;
    }
    try {
        const { buildSummary } = await service.getBuildSummary(collectionId, specificationId, buildId);
        if (buildSummary) {
            displayBuildSummaryDetails(buildSummary, logLevel);
        }
    }
    catch {
    }
}
function detectV3Collection(filePath) {
    try {
        if (fs_1.default.statSync(filePath).isDirectory()) {
            const refBased = (0, v3_collection_merge_1.findV3CollectionInDirectory)(filePath);
            if (refBased) {
                return refBased;
            }
            if ((0, v3_filesystem_collection_1.isV3FilesystemCollectionRoot)(filePath)) {
                return filePath;
            }
            return null;
        }
        if ((0, v3_collection_merge_1.isV3Collection)(filePath)) {
            return filePath;
        }
        return null;
    }
    catch {
        return null;
    }
}
async function mergeV3CollectionToTempFile(v3CollectionPath, verbose) {
    const isDir = fs_1.default.statSync(v3CollectionPath).isDirectory();
    if (verbose) {
        const colors = require('colors/safe');
        const detectMsg = isDir ?
            'Detected v3 filesystem collection - assembling...' :
            'Detected v3 collection - merging references...';
        print_1.default.lf(colors.cyan(detectMsg));
    }
    const tempDir = os_1.default.tmpdir();
    const v2TempFileName = `postman-v2-transformed-${Date.now()}-${Math.random().toString(36).substring(7)}.json`;
    const v2TransformedFilePath = path_1.default.join(tempDir, v2TempFileName);
    const v3TempFileName = `postman-v3-merged-${Date.now()}-${Math.random().toString(36).substring(7)}.json`;
    const v3MergedFilePath = path_1.default.join(tempDir, v3TempFileName);
    try {
        let v3Collection;
        if (isDir) {
            v3Collection = (0, v3_filesystem_collection_1.assembleV3CollectionFromFilesystem)(v3CollectionPath);
        }
        else {
            await (0, v3_collection_merge_1.mergeV3CollectionToFile)(v3CollectionPath, v3MergedFilePath);
            const v3Content = fs_1.default.readFileSync(v3MergedFilePath, 'utf8');
            v3Collection = JSON.parse(v3Content);
        }
        if (verbose) {
            const colors = require('colors/safe');
            const assembledMsg = isDir ?
                '✓ V3 filesystem collection assembled' :
                '✓ V3 collection merged successfully';
            print_1.default.lf(colors.green(assembledMsg));
            print_1.default.lf(colors.cyan('Transforming v3 to v2 format...'));
            const debugPath = path_1.default.join(tempDir, `postman-v3-debug-${Date.now()}.json`);
            fs_1.default.writeFileSync(debugPath, JSON.stringify(v3Collection, null, 2), 'utf8');
            print_1.default.lf(colors.gray(`Debug: assembled v3 written to ${debugPath}`));
        }
        const v2Collection = (0, v3_to_v2_transform_1.transformV3ToV2)(v3Collection);
        fs_1.default.writeFileSync(v2TransformedFilePath, JSON.stringify(v2Collection, null, 2), 'utf8');
        if (verbose) {
            const colors = require('colors/safe');
            print_1.default.lf(colors.green('✓ V3 collection transformed to V2 successfully'));
            print_1.default.lf(colors.gray(`Transformed file: ${v2TransformedFilePath}`));
        }
        try {
            if (fs_1.default.existsSync(v3MergedFilePath)) {
                fs_1.default.unlinkSync(v3MergedFilePath);
            }
        }
        catch {
        }
        return v2TransformedFilePath;
    }
    catch (err) {
        try {
            if (fs_1.default.existsSync(v3MergedFilePath)) {
                fs_1.default.unlinkSync(v3MergedFilePath);
            }
            if (fs_1.default.existsSync(v2TransformedFilePath)) {
                fs_1.default.unlinkSync(v2TransformedFilePath);
            }
        }
        catch {
        }
        const errorMessage = err.message || String(err);
        throw new Error(`Failed to process v3 collection: ${errorMessage}`);
    }
}
async function prepareCollectionFile(filePath, entityType, verbose) {
    if (entityType !== 'collection') {
        return { filePath };
    }
    const v3CollectionPath = detectV3Collection(filePath);
    if (!v3CollectionPath) {
        return { filePath };
    }
    const mergedFilePath = await mergeV3CollectionToTempFile(v3CollectionPath, verbose);
    return {
        filePath: mergedFilePath,
        tempFilePath: mergedFilePath
    };
}
function cleanupTempFiles(tempFilePaths, verbose = false) {
    const paths = Array.isArray(tempFilePaths) ?
        tempFilePaths :
        [tempFilePaths];
    for (const filePath of paths) {
        if (!filePath) {
            continue;
        }
        try {
            if (fs_1.default.existsSync(filePath)) {
                fs_1.default.unlinkSync(filePath);
                if (verbose) {
                    print_1.default.lf(safe_1.default.gray(`Cleaned up temporary file: ${filePath}`));
                }
            }
        }
        catch (cleanupErr) {
            if (verbose) {
                print_1.default.lf(safe_1.default.yellow(`Warning: Failed to clean up temporary file: ${cleanupErr.message}`));
            }
        }
    }
}
async function handleBuildCompletion(completedBuild, outputDir, skipConfirmation, next, service, forceUseBuildId = false, logLevel = 'warn') {
    try {
        const successfulArtifacts = completedBuild.artifacts?.filter((artifact) => {
            return artifact.status === 'SUCCESS';
        }) || [];
        const failedArtifacts = completedBuild.artifacts?.filter((artifact) => {
            return artifact.status === 'FAIL';
        }) || [];
        const hasSettledArtifacts = successfulArtifacts.length > 0 || failedArtifacts.length > 0;
        if (completedBuild.status === sdk_generator_service_1.BuildStatus.SUCCESS ||
            (hasSettledArtifacts &&
                completedBuild.status === sdk_generator_service_1.BuildStatus.IN_PROGRESS)) {
            if (successfulArtifacts.length > 0) {
                print_1.default.lf(safe_1.default.green('\n✓ Build completed successfully!'));
                if (completedBuild.duration) {
                    const durationSeconds = Math.round(completedBuild.duration / 1000);
                    print_1.default.lf(`Duration: ${durationSeconds}s`);
                }
                await fetchAndDisplayBuildSummary(service, completedBuild.id, completedBuild.collectionId, completedBuild.apiSpecId, logLevel);
                await downloadBuildArtifacts(completedBuild, outputDir, service, skipConfirmation, forceUseBuildId);
                return next(null);
            }
        }
        else if (completedBuild.status === sdk_generator_service_1.BuildStatus.FAILURE) {
            if (successfulArtifacts.length > 0) {
                print_1.default.lf(safe_1.default.yellow('\n⚠ Build completed with partial success'));
                if (completedBuild.duration) {
                    const durationSeconds = Math.round(completedBuild.duration / 1000);
                    print_1.default.lf(`Duration: ${durationSeconds}s`);
                }
                if (failedArtifacts.length > 0) {
                    print_1.default.lf(safe_1.default.yellow('\nFailed to generate some SDKs:'));
                    failedArtifacts.forEach((artifact) => {
                        const language = artifact.sdk?.language || artifact.artifactType;
                        print_1.default.lf(safe_1.default.red(`  ✗ ${language}`));
                        if (artifact.error) {
                            print_1.default.lf(safe_1.default.gray(`    ${artifact.error}`));
                        }
                    });
                }
                if (completedBuild.warnings &&
                    completedBuild.warnings.length > 0) {
                    print_1.default.lf('\nWarnings:');
                    completedBuild.warnings.forEach((warning) => {
                        print_1.default.lf(safe_1.default.yellow(`  ⚠ ${warning}`));
                    });
                }
                await fetchAndDisplayBuildSummary(service, completedBuild.id, completedBuild.collectionId, completedBuild.apiSpecId, logLevel);
                await downloadBuildArtifacts(completedBuild, outputDir, service, skipConfirmation, forceUseBuildId);
                print_1.default.lf(safe_1.default.green(`\n✓ Successfully generated ${successfulArtifacts.length} SDK(s), ` +
                    `${failedArtifacts.length} failed`));
                return next(null);
            }
            print_1.default.lf(safe_1.default.red('\n✗ Build failed'));
            if (failedArtifacts.length > 0) {
                print_1.default.lf('\nFailed artifacts:');
                failedArtifacts.forEach((artifact) => {
                    const language = artifact.sdk?.language || artifact.artifactType;
                    print_1.default.lf(safe_1.default.red(`  ✗ ${language}`));
                    if (artifact.error) {
                        print_1.default.lf(safe_1.default.gray(`    ${artifact.error}`));
                    }
                });
            }
            if (completedBuild.warnings && completedBuild.warnings.length > 0) {
                print_1.default.lf('\nWarnings:');
                completedBuild.warnings.forEach((warning) => {
                    print_1.default.lf(safe_1.default.yellow(`  ⚠ ${warning}`));
                });
            }
            await fetchAndDisplayBuildSummary(service, completedBuild.id, completedBuild.collectionId, completedBuild.apiSpecId, logLevel);
            return next(new Error('SDK generation failed. ' +
                'Check the error details above for more information.'));
        }
        return next(null);
    }
    catch (err) {
        return next(err);
    }
}
