// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/v3-collection-merge.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isV3Collection = isV3Collection;
exports.findV3CollectionInDirectory = findV3CollectionInDirectory;
exports.mergeV3Collection = mergeV3Collection;
exports.mergeV3CollectionToFile = mergeV3CollectionToFile;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
function isV3Collection(filePath) {
    try {
        if (!fs_1.default.existsSync(filePath)) {
            return false;
        }
        const basename = path_1.default.basename(filePath);
        if (!basename.match(/^collection\.ya?ml$/i)) {
            return false;
        }
        const content = fs_1.default.readFileSync(filePath, 'utf8');
        const parsed = js_yaml_1.default.load(content);
        return (parsed &&
            typeof parsed === 'object' &&
            parsed.$kind === 'collection' &&
            (parsed.$schema?.includes('/v3.') ||
                (Array.isArray(parsed.items) && parsed.items.length > 0 && parsed.items[0].ref)));
    }
    catch {
        return false;
    }
}
function findV3CollectionInDirectory(dirPath) {
    try {
        if (!fs_1.default.existsSync(dirPath) || !fs_1.default.statSync(dirPath).isDirectory()) {
            return null;
        }
        const yamlPath = path_1.default.join(dirPath, 'collection.yaml');
        const ymlPath = path_1.default.join(dirPath, 'collection.yml');
        if (fs_1.default.existsSync(yamlPath) && isV3Collection(yamlPath)) {
            return yamlPath;
        }
        else if (fs_1.default.existsSync(ymlPath) && isV3Collection(ymlPath)) {
            return ymlPath;
        }
        return null;
    }
    catch {
        return null;
    }
}
function readYamlOrJsonFile(filePath) {
    const content = fs_1.default.readFileSync(filePath, 'utf8');
    if (filePath.endsWith('.json')) {
        return JSON.parse(content);
    }
    return js_yaml_1.default.load(content);
}
function resolveRefPath(baseDir, ref) {
    const cleanRef = ref.startsWith('./') ? ref.slice(2) : ref;
    return path_1.default.resolve(baseDir, cleanRef);
}
function assertPathWithinRoot(collectionRoot, resolvedPath, originalRef) {
    const rootWithSep = collectionRoot.toLowerCase() + path_1.default.sep, targetLower = resolvedPath.toLowerCase();
    if (!targetLower.startsWith(rootWithSep)) {
        throw new Error(`Reference path escapes the collection directory: "${originalRef}" ` +
            `resolves to "${resolvedPath}" which is outside "${collectionRoot}".`);
    }
}
async function resolveRefsRecursively(obj, baseDir, collectionRoot) {
    if (!obj || typeof obj !== 'object') {
        return obj;
    }
    if (Array.isArray(obj)) {
        return Promise.all(obj.map((item) => {
            return resolveRefsRecursively(item, baseDir, collectionRoot);
        }));
    }
    const resolved = { ...obj };
    delete resolved.ref;
    if (resolved.items && Array.isArray(resolved.items)) {
        resolved.items = await resolveItemsArray(resolved.items, baseDir, collectionRoot);
    }
    for (const key in resolved) {
        if (resolved[key] && typeof resolved[key] === 'object') {
            resolved[key] = await resolveRefsRecursively(resolved[key], baseDir, collectionRoot);
        }
    }
    return resolved;
}
async function resolveItemsArray(items, baseDir, collectionRoot) {
    if (!items || !Array.isArray(items)) {
        return [];
    }
    const resolvedItems = [];
    for (const item of items) {
        if (item.ref) {
            const refPath = resolveRefPath(baseDir, item.ref);
            assertPathWithinRoot(collectionRoot, refPath, item.ref);
            if (!fs_1.default.existsSync(refPath)) {
                throw new Error(`Referenced file not found: ${refPath}`);
            }
            const refContent = readYamlOrJsonFile(refPath);
            const refDir = path_1.default.dirname(refPath);
            const resolved = await resolveRefsRecursively(refContent, refDir, collectionRoot);
            resolvedItems.push(resolved);
        }
        else {
            const resolved = await resolveRefsRecursively(item, baseDir, collectionRoot);
            resolvedItems.push(resolved);
        }
    }
    return resolvedItems;
}
async function mergeV3Collection(collectionPath) {
    if (!isV3Collection(collectionPath)) {
        throw new Error(`Not a valid v3 collection: ${collectionPath}`);
    }
    const collectionData = readYamlOrJsonFile(collectionPath);
    const collectionDir = path_1.default.resolve(path_1.default.dirname(collectionPath));
    const mergedCollection = await resolveRefsRecursively(collectionData, collectionDir, collectionDir);
    if (!mergedCollection.$schema) {
        mergedCollection.$schema = 'https://schema.postman.com/json/draft-2020-12/collection/v3.0.0/';
    }
    if (mergedCollection.$kind !== 'collection') {
        throw new Error('Merged collection missing required "$kind: collection" field');
    }
    if (!mergedCollection.name) {
        throw new Error('Merged collection missing required "name" field');
    }
    return mergedCollection;
}
async function mergeV3CollectionToFile(collectionPath, outputPath) {
    const mergedCollection = await mergeV3Collection(collectionPath);
    const outputDir = path_1.default.dirname(outputPath);
    if (!fs_1.default.existsSync(outputDir)) {
        fs_1.default.mkdirSync(outputDir, { recursive: true });
    }
    fs_1.default.writeFileSync(outputPath, JSON.stringify(mergedCollection, null, 2), 'utf8');
    return outputPath;
}
