// @postman/pm-bin-linux-x64/bin/lib/commands/sdk/v3-filesystem-collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isV3RequestFilePath = isV3RequestFilePath;
exports.isV3FilesystemCollectionRoot = isV3FilesystemCollectionRoot;
exports.findV3FilesystemCollectionRootFromFile = findV3FilesystemCollectionRootFromFile;
exports.assembleV3CollectionFromFilesystem = assembleV3CollectionFromFilesystem;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const REQUEST_FILE_SUFFIX = '.request.yaml';
const REQUEST_FILE_SUFFIX_YL = '.request.yml';
const RESOURCES_DIR = '.resources';
const DEFINITION_FILE = 'definition.yaml';
const DEFINITION_FILE_YL = 'definition.yml';
function readYamlFile(filePath) {
    const content = fs_1.default.readFileSync(filePath, 'utf8');
    return js_yaml_1.default.load(content);
}
function hasRequestFiles(dirPath) {
    if (!fs_1.default.existsSync(dirPath) || !fs_1.default.statSync(dirPath).isDirectory()) {
        return false;
    }
    const entries = fs_1.default.readdirSync(dirPath, { withFileTypes: true });
    const found = entries.some((e) => {
        if (!e.isFile()) {
            return false;
        }
        const name = e.name.toLowerCase();
        return name.endsWith(REQUEST_FILE_SUFFIX) || name.endsWith(REQUEST_FILE_SUFFIX_YL);
    });
    if (found) {
        return true;
    }
    for (const e of entries) {
        if (e.isDirectory() && !e.name.startsWith('.') && e.name !== RESOURCES_DIR) {
            if (hasRequestFiles(path_1.default.join(dirPath, e.name))) {
                return true;
            }
        }
    }
    return false;
}
function hasResourcesDir(dirPath) {
    const resPath = path_1.default.join(dirPath, RESOURCES_DIR);
    return fs_1.default.existsSync(resPath) && fs_1.default.statSync(resPath).isDirectory();
}
function isV3RequestFilePath(filePath) {
    const name = path_1.default.basename(filePath).toLowerCase();
    return name.endsWith(REQUEST_FILE_SUFFIX) || name.endsWith(REQUEST_FILE_SUFFIX_YL);
}
function isV3FilesystemCollectionRoot(dirPath) {
    try {
        if (!fs_1.default.existsSync(dirPath) || !fs_1.default.statSync(dirPath).isDirectory()) {
            return false;
        }
        if (fs_1.default.existsSync(path_1.default.join(dirPath, 'collection.yaml')) ||
            fs_1.default.existsSync(path_1.default.join(dirPath, 'collection.yml'))) {
            return false;
        }
        return hasRequestFiles(dirPath) || hasResourcesDir(dirPath);
    }
    catch {
        return false;
    }
}
function loadScopeDefinition(dirPath) {
    const defYaml = path_1.default.join(dirPath, RESOURCES_DIR, DEFINITION_FILE);
    const defYml = path_1.default.join(dirPath, RESOURCES_DIR, DEFINITION_FILE_YL);
    if (fs_1.default.existsSync(defYaml)) {
        if (fs_1.default.existsSync(defYml)) {
            const resDir = path_1.default.join(dirPath, RESOURCES_DIR);
            throw new Error(`Both ${DEFINITION_FILE} and ${DEFINITION_FILE_YL}` +
                ` exist in ${resDir}; only one allowed.`);
        }
        return readYamlFile(defYaml);
    }
    if (fs_1.default.existsSync(defYml)) {
        return readYamlFile(defYml);
    }
    return null;
}
function requestFilenameToRequestName(filename) {
    const lower = filename.toLowerCase();
    if (lower.endsWith(REQUEST_FILE_SUFFIX_YL)) {
        return filename.slice(0, -REQUEST_FILE_SUFFIX_YL.length);
    }
    return filename.slice(0, -REQUEST_FILE_SUFFIX.length);
}
const FILESYSTEM_ONLY_FIELDS = ['order', 'examples', 'scripts', '$schema'];
function stripFilesystemFields(obj) {
    if (!obj || typeof obj !== 'object') {
        return;
    }
    for (const field of FILESYSTEM_ONLY_FIELDS) {
        if (field in obj && typeof obj[field] === 'string') {
            delete obj[field];
        }
        if (field === 'order' && typeof obj[field] === 'number') {
            delete obj[field];
        }
    }
}
function compareOrderThenName(a, b) {
    const aOrder = typeof a.order === 'number' ? a.order : Infinity;
    const bOrder = typeof b.order === 'number' ? b.order : Infinity;
    if (aOrder !== bOrder) {
        return aOrder - bOrder;
    }
    return a.name.localeCompare(b.name, undefined, { sensitivity: 'base' });
}
function discoverItems(dirPath) {
    const entries = fs_1.default.readdirSync(dirPath, { withFileTypes: true });
    const requestFiles = [];
    const subdirs = [];
    for (const e of entries) {
        if (e.name.startsWith('.')) {
            continue;
        }
        const fullPath = path_1.default.join(dirPath, e.name);
        if (e.isFile()) {
            const lower = e.name.toLowerCase();
            if (lower.endsWith(REQUEST_FILE_SUFFIX) || lower.endsWith(REQUEST_FILE_SUFFIX_YL)) {
                const requestName = requestFilenameToRequestName(e.name);
                let order;
                try {
                    const content = readYamlFile(fullPath);
                    order = typeof content?.order === 'number' ? content.order : undefined;
                }
                catch {
                }
                requestFiles.push({ order, name: requestName, path: fullPath });
            }
        }
        else if (e.isDirectory() && e.name !== RESOURCES_DIR) {
            let order;
            try {
                const subDef = loadScopeDefinition(fullPath);
                order = typeof subDef?.order === 'number' ? subDef.order : undefined;
            }
            catch {
            }
            subdirs.push({ order, name: e.name, path: fullPath });
        }
    }
    const items = [];
    for (const r of requestFiles.sort(compareOrderThenName)) {
        const content = readYamlFile(r.path);
        if (!content.name) {
            content.name = r.name;
        }
        stripFilesystemFields(content);
        items.push(content);
    }
    for (const sub of [...subdirs].sort(compareOrderThenName)) {
        const subDef = loadScopeDefinition(sub.path);
        const childItems = discoverItems(sub.path);
        const folder = {
            $kind: 'collection',
            name: subDef?.name ?? sub.name,
            items: childItems
        };
        if (subDef?.description) {
            folder.description = subDef.description;
        }
        items.push(folder);
    }
    return items;
}
function findV3FilesystemCollectionRootFromFile(filePath) {
    try {
        if (!fs_1.default.existsSync(filePath) || !fs_1.default.statSync(filePath).isFile()) {
            return null;
        }
        if (!isV3RequestFilePath(filePath)) {
            return null;
        }
        let dir = path_1.default.dirname(path_1.default.resolve(filePath));
        const fsRoot = path_1.default.parse(dir).root;
        let topmostV3Root = null;
        while (dir !== path_1.default.dirname(dir)) {
            if (isV3FilesystemCollectionRoot(dir)) {
                topmostV3Root = dir;
                if (loadScopeDefinition(dir) !== null) {
                    return dir;
                }
            }
            if (dir === fsRoot) {
                break;
            }
            dir = path_1.default.dirname(dir);
        }
        return topmostV3Root;
    }
    catch {
        return null;
    }
}
function assembleV3CollectionFromFilesystem(rootDir, collectionName) {
    const definition = loadScopeDefinition(rootDir);
    const name = collectionName ?? definition?.name ?? path_1.default.basename(path_1.default.resolve(rootDir));
    const items = discoverItems(rootDir);
    const collection = {
        $schema: 'https://schema.postman.com/json/draft-2020-12/collection/v3.0.0/',
        $kind: 'collection',
        name,
        items
    };
    if (definition?.variables) {
        collection.variables = definition.variables;
    }
    if (definition?.description) {
        collection.description = definition.description;
    }
    return collection;
}
