// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/fetchers/fetchViaApiId.js
const get = require('lodash/get'), find = require('lodash/find'), includes = require('lodash/includes'), yml = require('yaml-js'), axios = require('axios'), util = require('../../../../lib/util'), { SCHEMA_TYPES, ASYNC_API_VERSION_TO_TYPE_MAP } = require('../constants');
async function fetchViaApiId(apiId, options, next) {
    const apiServiceOptions = {
        method: 'get',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-dev-platform',
            'x-postman-service': 'cloud-api'
        }
    };
    if (options.postmanApiKey) {
        apiServiceOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        apiServiceOptions.headers['x-access-token'] = options.accessToken;
    }
    let resBody;
    try {
        const response = await axios.request({
            url: `${util.POSTMAN_GATEWAY_BASE_URL()}/apis/${apiId}?populate=schemas&populate=versions&populate=relations`,
            ...apiServiceOptions
        });
        resBody = response.data;
        if (typeof resBody !== 'object') {
            return next('Something went wrong when fetching the API. Try again.');
        }
        if (response.status < 200 || response.status >= 300) {
            const errorMessage = resBody.detail || resBody.title ||
                'Something went wrong when fetching the API. Try again.';
            return next(errorMessage);
        }
        const api = resBody.data, apiName = api.name, relation = get(api, 'versions[0].relations', null);
        let apiDefinition, apiVersionId, schemaId, definitionResponseBody, schema = get(api, 'versions[0].schemas[0]', null);
        if (!schema && relation && relation.length) {
            apiVersionId = get(api, 'versions[0].id', null);
            apiDefinition = find(relation, { type: 'apiDefinition' });
            if (!apiDefinition) {
                return next('No schema definition found.');
            }
            schemaId = apiDefinition.id;
            const definitionResponse = await axios.request({
                url: `${util.POSTMAN_GATEWAY_BASE_URL()}/apis/${apiId}/versions/${apiVersionId}/schemas/${schemaId}`,
                ...apiServiceOptions
            });
            definitionResponseBody = definitionResponse.data;
            if (typeof definitionResponseBody !== 'object') {
                return next('Something went wrong when fetching the API. Try again.');
            }
            if (definitionResponse.status < 200 || definitionResponse.status >= 300) {
                const errorMessage = definitionResponseBody.detail || definitionResponseBody.title ||
                    'Something went wrong when fetching the API. Try again.';
                return next(errorMessage);
            }
            schema = get(definitionResponseBody, 'schema', null);
        }
        try {
            if (!schema) {
                throw new Error('No schema found.');
            }
            let { type, language } = schema;
            if (!schema.content && !schema.schema) {
                throw new Error('No schema found.');
            }
            if (!['json', 'yaml', 'yml'].includes(language)) {
                throw new Error(`Unsupported schema language: ${language}.`);
            }
            if (!includes(Object.keys(SCHEMA_TYPES), type)) {
                throw new Error('The schema provided is not an OpenAPI/AsyncAPI schema.');
            }
            if (language === 'yml') {
                language = 'yaml';
            }
            if (type === 'asyncapi2') {
                if (language === 'yaml') {
                    const schemaContent = yml.load(schema.content);
                    type = ASYNC_API_VERSION_TO_TYPE_MAP[schemaContent.asyncapi];
                }
                else if (language === 'json') {
                    const schemaContent = JSON.parse(schema.content);
                    type = ASYNC_API_VERSION_TO_TYPE_MAP[schemaContent.asyncapi];
                }
                if (!type) {
                    throw new Error('The provided AsyncAPI schema lacks a valid schema version.');
                }
            }
            return next(null, {
                id: apiId,
                name: apiName,
                content: schema.content || schema.schema,
                language: language,
                type: type
            });
        }
        catch (err) {
            err.message = `Couldn't parse the API. ${err.message}`;
            return next(err);
        }
    }
    catch (err) {
        return next(err);
    }
}
module.exports = {
    fetchViaApiId
};
