// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/fetchers/fetchViaConfig.js
require('../../../logger').init();
const _ = require('lodash'), Converter = require('openapi-to-postmanv2'), fs = require('fs').promises, path = require('path'), json = require('json-to-ast'), yaml = require('yaml-js'), configInterface = require('../apiConfig/configInterface');
const { SCHEMA_VERSIONS, LEARNING_CENTER_LINK, ASYNC_API_VERSION_TO_TYPE_MAP } = require('../constants');
function getLanguage(pathName) {
    let extension = path.extname(pathName).replace(/^\./, '').toLowerCase();
    if (extension === 'yml') {
        extension = 'yaml';
    }
    return extension;
}
async function fetchViaConfig(apiId, next) {
    try {
        const allApiIds = apiId ? [apiId] : [];
        let hasErrorOccurred = false, errorMessages = [];
        if (!apiId) {
            const postmanConfigApiIds = await configInterface.getAPIIds();
            allApiIds.push(...postmanConfigApiIds.result);
        }
        if (allApiIds.length === 0) {
            hasErrorOccurred = true;
            errorMessages.push('No apiIds not found in the config.');
            throw new Error('No apiIds not found in the config.');
        }
        const apiObjects = await Promise.all(allApiIds.map(async (apiId) => {
            try {
                let fetchSchemaFilePaths;
                try {
                    fetchSchemaFilePaths = await configInterface.getAPIDefinitionFilePaths(apiId);
                }
                catch (err) {
                    if (_.get(err, 'error.name') === 'failedToFetchData') {
                        fetchSchemaFilePaths = {
                            result: {
                                rootDirectory: '',
                                filePaths: []
                            }
                        };
                    }
                    else if (!_.isNil(_.get(err, 'error.message'))) {
                        throw new Error(err.error.message);
                    }
                }
                const rootDirectory = fetchSchemaFilePaths.result.rootDirectory, schemaFilePaths = fetchSchemaFilePaths.result.filePaths;
                if (schemaFilePaths.length === 0) {
                    throw new Error('No schema files found in the config.');
                }
                const allParsedFiles = await Promise.all(schemaFilePaths.map(async (filePath) => {
                    let constructedFilePath, ast = {};
                    if (fetchSchemaFilePaths.status === 'warning' && fetchSchemaFilePaths.reason === 'configNotSupported') {
                        constructedFilePath = filePath;
                    }
                    else {
                        constructedFilePath = path.posix.join('.', rootDirectory, filePath);
                    }
                    const language = getLanguage(constructedFilePath), content = await fs.readFile(constructedFilePath, 'utf-8');
                    if (language === 'json') {
                        ast = json(content);
                    }
                    else if (language === 'yaml') {
                        ast = yaml.compose(content);
                    }
                    return {
                        path: filePath,
                        fileName: filePath,
                        content: content,
                        ast: ast,
                        language: language
                    };
                }));
                let id = apiId, name, referenceMap, rootFile, allFiles, content, language, type;
                if (allParsedFiles.length > 1) {
                    let converterInput = {
                        type: 'multiFile',
                        options: { includeReferenceMap: true },
                        data: allParsedFiles
                    };
                    const response = await Converter.bundle(converterInput);
                    if (!response.result) {
                        throw new Error('Unable to bundle the schema files.');
                    }
                    let data = _.get(response, 'output.data[0]', {}), specification = response.output.specification;
                    if (!_.get(specification, 'type', '').toLowerCase().includes('openapi')) {
                        throw new Error('The schema provided is not an OpenAPI schema.');
                    }
                    if (!data) {
                        throw new Error('No schema found.');
                    }
                    language = getLanguage(data.rootFile.path);
                    name = data.rootFile.path;
                    referenceMap = data.referenceMap;
                    rootFile = data.rootFile;
                    allFiles = allParsedFiles;
                    content = data.bundledContent;
                    type = SCHEMA_VERSIONS[_.get(specification, 'version').replace(/\./g, '_')] || SCHEMA_VERSIONS['3_0'];
                }
                else if (allParsedFiles.length === 1) {
                    let file = allParsedFiles[0], parsed;
                    language = getLanguage(file.path);
                    name = file.path;
                    content = file.content;
                    if (language === 'json') {
                        parsed = JSON.parse(content);
                    }
                    else if (language === 'yaml') {
                        parsed = yaml.load(content);
                    }
                    if (!parsed) {
                        throw new Error('No schema found.');
                    }
                    if (!parsed.openapi && !parsed.swagger && !parsed.asyncapi) {
                        throw new Error('The provided schema is not an OpenAPI/AsyncAPI schema.');
                    }
                    if (parsed.swagger) {
                        type = SCHEMA_VERSIONS[parsed.swagger.replace(/\./g, '_').substring(0, 3)];
                    }
                    else if (parsed.asyncapi) {
                        type = ASYNC_API_VERSION_TO_TYPE_MAP[parsed.asyncapi];
                        if (!type) {
                            throw new Error('The provided AsyncAPI schema lacks a valid schema version.');
                        }
                    }
                    else {
                        type = SCHEMA_VERSIONS[parsed.openapi.replace(/\./g, '_').substring(0, 3)];
                    }
                }
                if (!['yaml', 'json'].includes(language)) {
                    throw new Error(`Unsupported schema language '${language}.`);
                }
                return ({
                    id,
                    name,
                    referenceMap,
                    rootFile,
                    allFiles,
                    content,
                    language,
                    type
                });
            }
            catch (err) {
                hasErrorOccurred = true;
                const errorMessage = err.message || 'Something went wrong';
                if (allApiIds.length === 1) {
                    throw err;
                }
                else {
                    logger.error(`Failed to lint ${apiId}: ${errorMessage}`);
                    errorMessages.push(`Failed to lint ${apiId}: ${errorMessage}`);
                    return {};
                }
            }
        }));
        let finalConfigs = [];
        _.forEach(apiObjects, (apiConfig) => {
            if (!_.isEmpty(apiConfig)) {
                finalConfigs.push(apiConfig);
            }
        });
        return next(null, {
            apiObjects: finalConfigs,
            countOfAPIs: allApiIds.length,
            hasErrorOccurred: hasErrorOccurred,
            errorMessages: errorMessages
        });
    }
    catch (err) {
        if (_.get(err, 'error.name') === 'configFileNotFound') {
            const message = `Couldn't find the Postman config file in the current directory.\nTry re-integrating this folder with your Postman API. ${LEARNING_CENTER_LINK}`;
            err.message = message;
        }
        else if (_.get(err, 'code') === 'ENOENT') {
            err.message = `Could not find the schema file: ${err.path}.`;
        }
        else {
            err.message = `Couldn't parse the API. ${err.message}`;
        }
        return next(err);
    }
}
async function checkIfLocalGitLinkedAPI(apiId) {
    try {
        const arrayOfAPIIds = await configInterface.getAPIIds(), doesAPIIdExist = _.includes(arrayOfAPIIds.result, apiId);
        return doesAPIIdExist;
    }
    catch (err) {
        return false;
    }
}
module.exports = {
    fetchViaConfig,
    checkIfLocalGitLinkedAPI
};
