// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/fetchers/fetchViaFile.js
const path = require('path'), fs = require('fs').promises, YAML = require('yaml-js'), { SCHEMA_VERSIONS, ASYNC_API_VERSION_TO_TYPE_MAP } = require('../constants');
function getSchemaType(schema, schemaVersion) {
    if (schema.asyncapi) {
        const type = ASYNC_API_VERSION_TO_TYPE_MAP[schemaVersion];
        if (!type) {
            throw new Error('The provided AsyncAPI schema lacks a valid schema version.');
        }
        return type;
    }
    let version = schemaVersion.substring(0, 3).replace(/\./g, '_'), type = SCHEMA_VERSIONS[version] || SCHEMA_VERSIONS['3_0'];
    return type;
}
async function fetchViaFile(config, next) {
    try {
        const content = await fs.readFile(config, 'utf8', 'r');
        let extension = path.extname(config).replace(/^\./, ''), schema, type;
        if (extension === 'json') {
            schema = JSON.parse(content);
        }
        else if (['yaml', 'yml'].includes(extension)) {
            schema = YAML.load(content);
            extension = 'yaml';
        }
        else {
            throw new Error(`Unsupported schema language: ${extension}.`);
        }
        if (!schema) {
            throw new Error('No schema found.');
        }
        if (!schema.openapi && !schema.swagger && !schema.asyncapi) {
            throw new Error('The schema provided is not an OpenAPI/AsyncAPI schema.');
        }
        const schemaVersion = schema && (schema.openapi || schema.swagger || schema.asyncapi);
        type = schemaVersion && getSchemaType(schema, schemaVersion);
        return next(null, {
            name: config,
            content: content,
            language: extension,
            type: type
        });
    }
    catch (err) {
        if (err.code && err.code === 'ENOENT') {
            err.message = 'Couldn\'t find the schema file at the given path.';
        }
        else {
            err.message = `Couldn't parse the API. ${err.message}`;
        }
        return next(err);
    }
}
module.exports = {
    fetchViaFile
};
