// @postman/pm-bin-linux-x64/bin/lib/commands/mock/controller/manager.js
const path = require('path'), fs = require('fs'), { isNil } = require('lodash'), colors = require('colors/safe'), print = require('../../../framework/print'), { parseManifest, isManifestFile } = require('../util/manifest'), { variableScopeToArray } = require('../util/variables');
let logger;
try {
    logger = require('../../../logger');
    if (!logger || !logger.info) {
        logger = console;
    }
}
catch (e) {
    logger = console;
}
function loadVariables(filePath) {
    var resolvedPath, json, values;
    if (!filePath || typeof filePath !== 'string') {
        return [];
    }
    resolvedPath = path.resolve(filePath);
    if (!fs.existsSync(resolvedPath)) {
        return [];
    }
    try {
        json = JSON.parse(fs.readFileSync(resolvedPath, 'utf8'));
        json = json.environment || json.globals || json;
        values = json.values || [];
        return values
            .filter(function (v) { return v && v.key; })
            .map(function (v) {
            return { key: String(v.key), value: !isNil(v.value) ? String(v.value) : '' };
        });
    }
    catch (_e) {
        return [];
    }
}
class MockServerManager {
    constructor(config = {}) {
        this.config = {
            port: 4500,
            autoRestart: false,
            maxRestarts: 0,
            name: 'Mock Server',
            ...config
        };
        this.config.logger = config.silent ? function () { } : logger.info.bind(logger);
        this.config.environment = typeof config.environment === 'string' ?
            loadVariables(config.environment) : variableScopeToArray(config.environment);
        this.config.globals = typeof config.globals === 'string' ?
            loadVariables(config.globals) : variableScopeToArray(config.globals);
        (config.envVar || []).forEach(function (v) {
            v && this.config.environment.push({ key: v.key, value: v.value });
        }, this);
        (config.globalVar || []).forEach(function (v) {
            v && this.config.globals.push({ key: v.key, value: v.value });
        }, this);
        this.mockServer = null;
        this.isRunning = false;
        this.manifest = null;
    }
    _initializeMockServer() {
        if (this.mockServer) {
            return;
        }
        try {
            const { SandboxedMockServer } = require('../runtime/sandbox.js');
            if (process.pkg && !this.config.workerPath) {
                const snapshotRoot = path.dirname(path.dirname(process.pkg.entrypoint));
                this.config.workerPath = path.join(snapshotRoot, 'lib', 'commands', 'mock', 'runtime', 'http-server-worker.js');
            }
            this.mockServer = new SandboxedMockServer(this.config);
            this.mockServer.on('worker-crash', (info) => {
                logger.error('Mock server worker crashed', {
                    operation: 'mockServer',
                    code: info.code,
                    signal: info.signal
                });
            });
            this.mockServer.on('worker-error', (error) => {
                logger.error('Mock server worker error', {
                    operation: 'mockServer',
                    error: error.error
                });
            });
        }
        catch (error) {
            throw new Error(`Failed to initialize mock server: ${error.message}`);
        }
    }
    async start(mockPathOrManifest) {
        try {
            let mockScriptPath, serverPort = this.config.port, serverName = this.config.name;
            if (isManifestFile(mockPathOrManifest)) {
                this.manifest = parseManifest(mockPathOrManifest);
                mockScriptPath = this.manifest.mockSrc;
                serverPort = this.manifest.port;
                serverName = this.manifest.name;
            }
            else {
                mockScriptPath = path.resolve(mockPathOrManifest);
                if (!fs.existsSync(mockScriptPath)) {
                    throw new Error(`Mock script not found: ${mockScriptPath}`);
                }
            }
            this.config.port = serverPort;
            this.config.name = serverName;
            this._log(colors.grey('Starting mock server...'));
            await this._initializeMockServer();
            await this.mockServer.loadHandler(mockScriptPath);
            await this.mockServer.start();
            this.isRunning = true;
            this._log(colors.green(`Mock server started: ${this.getDisplayName()}`));
            return {
                port: this.config.port,
                name: serverName,
                manifest: this.manifest
            };
        }
        catch (error) {
            logger.error('Failed to start mock server', {
                operation: 'mockServer',
                error: error.message
            });
            throw error;
        }
    }
    async stop() {
        if (!this.mockServer || !this.isRunning) {
            return;
        }
        await this.mockServer.stop();
        this.isRunning = false;
        this._log(colors.grey('Mock server stopped'));
    }
    getInfo() {
        if (!this.mockServer) {
            return { isRunning: false };
        }
        return this.mockServer.getInfo();
    }
    _log(message) {
        !this.config.silent && print.lf(message);
    }
    getDisplayName() {
        var info = this.getInfo();
        return this.manifest ?
            `${colors.cyan(this.manifest.name)} at ${colors.bold(info.url)}` :
            colors.bold(info.url);
    }
    getLogs() {
        if (!this.mockServer) {
            return [];
        }
        return this.mockServer.getLogs();
    }
}
module.exports = {
    MockServerManager
};
