// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/Service/RuleService.js
const axios = require('axios'), util = require('../../../util'), { SPECIFICATION_INPUT_TYPES } = require('../constants');
async function fetchGovernanceCustomFunctionsBundle(workspaceId, options) {
    try {
        let functionBundleServiceUri = `${util.POSTMAN_GATEWAY_BASE_URL()}/security/configure/function-bundle?` +
            `target=node&workspaceId=${workspaceId}`;
        const headers = {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'ruleset',
            'x-app-version': '10.15.2',
            Accept: 'application/vnd.postman.v2+json'
        };
        if (options.postmanApiKey) {
            headers['x-api-key'] = options.postmanApiKey;
        }
        else {
            headers['x-access-token'] = options.accessToken;
        }
        const response = await axios.request({
            url: functionBundleServiceUri,
            headers: headers,
            method: 'get'
        });
        if (response.status !== 200) {
            throw new Error('Failed to fetch custom functions bundle: received status code ' + response.status);
        }
        return response.data;
    }
    catch (err) {
        logger.error('Unable to retrieve governance custom functions bundle, errorMessage: ' +
            (err.message || 'Unknown error occurred.'));
        throw new Error('Unable to retrieve governance custom functions bundle');
    }
}
async function fetchGovernanceRulesetsByWorkspace(workspaceId, options) {
    try {
        const rulesetServiceUri = `${util.POSTMAN_GATEWAY_BASE_URL()}/checks/workspace/${workspaceId}?` +
            'serviceSlug=falcon_service', headers = {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'ruleset',
            'x-app-version': '10.15.2',
            Accept: 'application/vnd.postman.v2+json'
        };
        if (options.postmanApiKey) {
            headers['x-api-key'] = options.postmanApiKey;
        }
        else {
            headers['x-access-token'] = options.accessToken;
        }
        const response = await axios.request({
            url: rulesetServiceUri,
            headers: headers,
            method: 'get'
        });
        if (response.status !== 200) {
            throw new Error(`Failed to fetch workspace ruleset: received status code ${response.status}` +
                `data ${JSON.stringify(response.data)}`);
        }
        const customFunctionsBundle = await fetchGovernanceCustomFunctionsBundle(workspaceId, options);
        return {
            customFunctionsBundle: {
                value: customFunctionsBundle.bundle ?? '',
                runtime: 'node'
            },
            categories: [response.data]
        };
    }
    catch (err) {
        logger.error('Unable to retrieve workspace governance rulesets, errorMessage: ' +
            (err.message || 'Unknown error occurred.'));
        throw new Error('Unable to retrieve workspace governance rulesets');
    }
}
async function fetchGovernanceRulesets(specificationId, specificationInputType, { workspaceId }, options) {
    try {
        let rulesetServiceUri;
        if (workspaceId && specificationInputType === SPECIFICATION_INPUT_TYPES.FILE_PATH) {
            return fetchGovernanceRulesetsByWorkspace(workspaceId, options);
        }
        rulesetServiceUri = `${util.POSTMAN_GATEWAY_BASE_URL()}/security/configure/checks?` +
            'serviceSlug=falcon_service&include=customFunctionsBundle&customFunctionsBundle.runtime=node';
        if (specificationId && specificationInputType === SPECIFICATION_INPUT_TYPES.ID) {
            rulesetServiceUri += `&elementType=specification&elementId=${specificationId}`;
        }
        const headers = {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'ruleset',
            'x-app-version': '10.15.2',
            Accept: 'application/vnd.postman.v2+json'
        };
        if (options.postmanApiKey) {
            headers['x-api-key'] = options.postmanApiKey;
        }
        else {
            headers['x-access-token'] = options.accessToken;
        }
        const response = await axios.request({
            url: rulesetServiceUri,
            headers: headers,
            method: 'get'
        });
        if (response.status !== 200) {
            throw new Error('Failed to fetch ruleset: received status code ' + response.status);
        }
        return response.data;
    }
    catch (err) {
        logger.error('Unable to retrieve governance rulesets, errorMessage: ' +
            (err.message || 'Unknown error occurred.'));
        throw new Error('Unable to retrieve governance rulesets');
    }
}
module.exports = {
    fetchGovernanceRulesets
};
