// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/Service/SpecificationService.js
const axios = require('axios'), fs = require('fs'), path = require('path'), { v4: uuidv4 } = require('uuid'), util = require('../../../util'), { CustomError } = require('../Errors/CustomError'), { SPECIFICATION_INPUT_TYPES, FILE_TYPES } = require('../constants'), _ = require('lodash'), yaml = require('js-yaml'), { resolveRelativePath, convertToPosixPath } = require('../Util/FilePath'), { SPECIFICATION_TYPES, getSupportedSpecificationTypes } = require('../SpecificationTypeConfig'), ComponentLibraryService = require('./ComponentLibraryService'), { resolve } = require('@apidevtools/json-schema-ref-parser');
function safeParseFile(content) {
    try {
        let parsed = yaml.load(content);
        return parsed;
    }
    catch (err) {
        logger.error('Unable to parse file content:', err);
        return {};
    }
}
function getSpecNameFromRootFile(rootFile) {
    try {
        if (rootFile.path.endsWith('.json')) {
            const jsonRootFile = JSON.parse(rootFile.content);
            return jsonRootFile.info.title;
        }
        const yamlContent = yaml.load(rootFile.content);
        return yamlContent.info.title;
    }
    catch (error) {
        logger.error(`Error occurred while parsing YAML from root file: ${error.message || error}`);
    }
    throw new Error(`Could not infer the spec name from root file: ${rootFile.path}`);
}
async function getSpecDetails(specId, options) {
    const requestOptions = {
        method: 'get',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api'
        }
    };
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    try {
        const response = await axios.request({
            url: `${util.POSTMAN_GATEWAY_BASE_URL()}/specifications/${specId}`,
            ...requestOptions
        });
        if (response.status !== 200) {
            throw new Error('Failed to fetch specification details (ID: ' +
                specId +
                '). Server responded with status ' +
                response.status +
                ': ' +
                (response.data?.detail ||
                    response.data?.title ||
                    'No additional details provided.'));
        }
        return response.data.data;
    }
    catch (err) {
        logger.error(`Error occurred in getSpecDetails specId: ${specId}, errorDetails: ${err.message || err}`);
        if (err.response &&
            err.response.data &&
            err.response.data.error &&
            err.response.data.error.detail) {
            throw new CustomError('An error occurred while retrieving specification details, Details: ' +
                (err.response.data.error.detail ||
                    'No additional details provided.'));
        }
        throw new Error('An error occurred while retrieving specification details (ID: ' +
            specId +
            '): ' +
            (err.message || err));
    }
}
async function fetchSpecificationFiles(specificationId, options) {
    const requestOptions = {
        method: 'get',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api'
        }
    };
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    let allSpecificationFiles = [], specificationFiles = [], paginationCursor = null, hasMoreFilesToFetch = true;
    while (hasMoreFilesToFetch) {
        const baseUrl = util.POSTMAN_GATEWAY_BASE_URL(), params = 'fields=id,name,content,type,path', url = paginationCursor ?
            `${baseUrl}/specifications/${specificationId}/files?${params}&cursor=${paginationCursor}` :
            `${baseUrl}/specifications/${specificationId}/files?${params}`;
        let response;
        try {
            response = await axios.request({
                url,
                ...requestOptions
            });
        }
        catch (err) {
            if (err.response &&
                err.response.data &&
                err.response.data.error &&
                err.response.data.error.detail) {
                throw new CustomError('An error occurred while retrieving specification files, Details: ' +
                    (err.response.data.error.detail ||
                        'No additional details provided.'));
            }
            logger.error('Error occurred in fetchSpecificationFiles specId: ' +
                specificationId +
                ', errorDetails: ' +
                (err.message || err));
            throw new Error('An error occurred while retrieving specification files content (ID: ' +
                specificationId +
                '): ' +
                (err.message || err));
        }
        if (!response || response.status !== 200) {
            throw new Error('Failed to fetch specification files specId: ' +
                specificationId +
                ', Server responded with status ' +
                response.status);
        }
        specificationFiles = _.get(response, 'data.data', []);
        allSpecificationFiles.push(...specificationFiles);
        paginationCursor = _.get(response, 'data.meta.cursor.next', null);
        hasMoreFilesToFetch = paginationCursor;
    }
    if (!allSpecificationFiles.length) {
        throw new CustomError(`No files found for specification ID: ${specificationId}`);
    }
    return allSpecificationFiles;
}
async function fetchSpecificationDetailsById(specId, options) {
    try {
        const specDetails = await getSpecDetails(specId, options), type = SPECIFICATION_TYPES[specDetails.type];
        if (!type) {
            throw new CustomError(`Unsupported specification type: ${specDetails.type}`);
        }
        return {
            id: specId,
            name: specDetails.name,
            type: type.label,
            files: await fetchSpecificationFiles(specId, options)
        };
    }
    catch (err) {
        logger.error('An error occurred while retrieving specification (ID: ' +
            specId +
            '): errorDetails ' +
            (err.message || err));
        if (err instanceof CustomError) {
            throw err;
        }
        throw new Error(`An error occurred while retrieving specification (ID: ${specId})`);
    }
}
async function detectRootFiles(files) {
    for (const specificationVersionKey of Object.keys(SPECIFICATION_TYPES)) {
        const { detectRootFiles } = SPECIFICATION_TYPES[specificationVersionKey];
        if (!detectRootFiles) {
            return;
        }
        try {
            const result = await detectRootFiles(files);
            if (result) {
                return result;
            }
        }
        catch {
        }
    }
}
async function getAllRelatedFiles(rootFilePath) {
    for (const specificationVersionKey of Object.keys(SPECIFICATION_TYPES)) {
        const { getAllRelatedFiles } = SPECIFICATION_TYPES[specificationVersionKey];
        if (!getAllRelatedFiles) {
            return;
        }
        try {
            const result = await getAllRelatedFiles(rootFilePath);
            return result;
        }
        catch {
        }
    }
}
async function dereferenceComponentsInSpecification(specification, options) {
    const dereferencedFiles = [...specification.files], componentUrlToFile = new Map();
    for (const specificationFile of specification.files) {
        await resolve(safeParseFile(specificationFile.content), {
            resolve: {
                file: {
                    read: (file) => {
                        const matchedFile = specification.files.find((f) => {
                            return path.resolve(f.path) === file.url;
                        });
                        return matchedFile?.content || {};
                    },
                    canRead: true,
                    order: 2
                },
                http: {
                    order: 1,
                    canRead: (file) => {
                        return file.url.startsWith(util.COMPONENT_LIBRARY_BASE_URL());
                    },
                    read: (file) => {
                        const path = file.url.split(util.COMPONENT_LIBRARY_BASE_URL() + '/')[1], fileId = uuidv4(), filePath = `CommonComponents_${path}.yaml`;
                        specificationFile.content =
                            specificationFile.content.replaceAll(file.url, filePath);
                        componentUrlToFile.set(file.url, {
                            id: fileId,
                            path: filePath,
                            componentPath: path
                        });
                        return {};
                    }
                }
            }
        });
    }
    await Promise.all([...componentUrlToFile.entries()].map(async ([, componentFile]) => {
        try {
            const response = await ComponentLibraryService.resolveComponentByPath(componentFile.componentPath, {
                consumerId: specification.id || specification.path
            }, options);
            dereferencedFiles.push({
                id: componentFile.id,
                content: response,
                path: componentFile.path
            });
            return response;
        }
        catch (err) {
            logger.error('Failed to resolve component at path: ' +
                componentFile.componentPath +
                ', errorDetails: ' +
                (err.message || err));
            return {};
        }
    }));
    return dereferencedFiles;
}
async function fetchSpecificationDetailsByPath(rootFilePath) {
    try {
        const resolvedRootFilePath = resolveRelativePath(rootFilePath), rootFile = {
            path: convertToPosixPath(resolvedRootFilePath),
            content: fs.readFileSync(resolvedRootFilePath, 'utf-8', 'r'),
            type: FILE_TYPES.ROOT
        }, result = await detectRootFiles([rootFile]);
        if (!result) {
            throw new CustomError('Could not detect a valid root file in the provided path. ' +
                'Supported specification types are: ' +
                getSupportedSpecificationTypes().join(', '));
        }
        const { type, version } = result.output.specification;
        const specification = {
            path: rootFilePath,
            type: `${type} ${version}`,
            name: getSpecNameFromRootFile(rootFile)
        };
        const files = await getAllRelatedFiles(rootFilePath);
        files.some((file) => {
            if (file.path === rootFile.path) {
                file.type = FILE_TYPES.ROOT;
                return true;
            }
        });
        specification.files = files;
        return specification;
    }
    catch (err) {
        logger.error('An error occurred while retrieving specification by file path (filePath: ' +
            rootFilePath +
            '): errorDetails ' +
            (err.message || err));
        if (err instanceof CustomError) {
            throw err;
        }
        throw new Error(`An error occurred while retrieving specification (Path: ${rootFilePath})`);
    }
}
async function getWorkspaceSpecifications(workspaceId, options) {
    if (!options.postmanApiKey && !options.accessToken) {
        throw new Error('No authorization data found. Please use the `postman login` command.');
    }
    const requestOptions = {
        method: 'get',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'workspaces',
            'Content-Type': 'application/json'
        }
    }, url = `${util.POSTMAN_GATEWAY_BASE_URL()}/workspaces/${workspaceId}?include=elements`;
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    try {
        const response = await axios.request({
            url,
            ...requestOptions
        }), specIds = _.get(response, 'data.data.elements.specifications', []), specifications = [];
        if (response.status !== 200) {
            throw new Error(`Failed to get workspace specifications. Server responded with status ${response.status}: ${response.data?.detail ||
                response.data?.title ||
                'No additional details provided.'}`);
        }
        if (!Array.isArray(specIds) || specIds.length === 0) {
            return [];
        }
        for (const specId of specIds) {
            try {
                const specDetails = await getSpecDetails(specId, options), files = await fetchSpecificationFiles(specId, options), rootFile = files.find((file) => {
                    return file.type === 'ROOT';
                }), rootPath = rootFile ? rootFile.path : null;
                specifications.push({
                    id: specId,
                    name: specDetails.name,
                    type: specDetails.type,
                    rootPath: rootPath
                });
            }
            catch (err) {
                logger.error(`Failed to fetch spec details for ${specId}: ${err.message}`);
            }
        }
        return specifications;
    }
    catch (err) {
        logger.error(`Error in getWorkspaceSpecifications workspaceId: ${workspaceId}, error: ${err.message || err}`);
        if (err.response &&
            err.response.data &&
            err.response.data.error &&
            err.response.data.error.detail) {
            throw new CustomError('An error while retrieving workspace specifications,: ' +
                (err.response.data.error.detail ||
                    'No additional details provided.'));
        }
        throw new Error('An error occurred while retrieving workspace specifications: ' +
            (err.message || err));
    }
}
async function createSpecification(specData, workspaceId, options) {
    if (!options.postmanApiKey && !options.accessToken) {
        throw new Error('No authorization data found. Please use the `postman login` command.');
    }
    const requestOptions = {
        method: 'post',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api',
            'Content-Type': 'application/json'
        }
    }, payload = {
        name: specData.name,
        type: specData.type,
        files: [
            {
                path: specData.rootPath,
                content: specData.content,
                type: 'ROOT'
            }
        ]
    }, url = `${util.POSTMAN_GATEWAY_BASE_URL()}/specifications?containerType=workspace&containerId=${workspaceId}`;
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    try {
        const response = await axios.request({
            url: url,
            ...requestOptions,
            data: payload
        });
        if (!(/2../).test(response.status)) {
            const errorMessage = _.get(response, 'data.error.message') ||
                _.get(response, 'data.message') ||
                'Unknown error';
            throw new Error(`Failed to create specification: ${errorMessage}`);
        }
        return _.get(response, 'data.data') || response.data;
    }
    catch (err) {
        logger.error(`Error occurred in createSpecification, errorDetails: ${err.message || err}`);
        if (err.response &&
            err.response.data &&
            err.response.data.error &&
            err.response.data.error.detail) {
            throw new CustomError('An error occurred while creating specification, Details: ' +
                (err.response.data.error.detail ||
                    'No additional details provided.'));
        }
        throw new Error('An error occurred while creating specification: ' +
            (err.message || err));
    }
}
async function updateSpecificationFile(specId, filePath, content, options) {
    if (!options.postmanApiKey && !options.accessToken) {
        throw new Error('No authorization data found. Please use the `postman login` command.');
    }
    const requestOptions = {
        method: 'patch',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api',
            'Content-Type': 'application/json'
        }
    }, payload = [
        {
            op: 'replace',
            path: '/content',
            value: content
        }
    ], url = `${util.POSTMAN_GATEWAY_BASE_URL()}/specs/${specId}/files/${filePath}`;
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    try {
        const response = await axios.request({
            url: url,
            ...requestOptions,
            data: payload
        });
        if (!(/2../).test(response.status)) {
            const errorMessage = _.get(response, 'data.error.message') ||
                _.get(response, 'data.message') ||
                'Unknown error';
            throw new Error(`Failed to update specification file: ${errorMessage}`);
        }
        return _.get(response, 'data.data') || response.data;
    }
    catch (err) {
        logger.error(`Error in updateSpecificationFile specId: ${specId}, filePath: ${filePath}, err: ${err.message || err}`);
        if (err.response &&
            err.response.data &&
            err.response.data.error &&
            err.response.data.error.detail) {
            throw new CustomError('An error occurred while updating specification file, Details: ' +
                (err.response.data.error.detail ||
                    'No additional details provided.'));
        }
        throw new Error('An error occurred while updating specification file: ' +
            (err.message || err));
    }
}
async function fetchSpecificationDetails(specificationInput, specificationInputType, options) {
    let specification;
    if (specificationInputType === SPECIFICATION_INPUT_TYPES.ID) {
        specification = await fetchSpecificationDetailsById(specificationInput, options);
    }
    else if (specificationInputType === SPECIFICATION_INPUT_TYPES.FILE_PATH) {
        specification = await fetchSpecificationDetailsByPath(specificationInput);
    }
    if (!specification) {
        throw new Error('Invalid specification input type provided: ' +
            specificationInputType);
    }
    const dereferencedFiles = await dereferenceComponentsInSpecification(specification, options);
    return { ...specification, files: dereferencedFiles };
}
async function bulkModifyFiles(specId, payload, options) {
    if (!options.postmanApiKey && !options.accessToken) {
        throw new Error('No authorization data found. Please use the `postman login` command.');
    }
    const requestOptions = {
        method: 'post',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api',
            'Content-Type': 'application/json'
        }
    }, url = `${util.POSTMAN_GATEWAY_BASE_URL()}/specifications/${specId}/bulk-files`;
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    try {
        const response = await axios.request({
            url: url,
            ...requestOptions,
            data: payload
        });
        if (!(/2../).test(response.status)) {
            const errorMessage = _.get(response, 'data.error.message') ||
                _.get(response, 'data.message') ||
                'Unknown error';
            throw new Error(`Failed to bulk modify specification files: ${errorMessage}`);
        }
        return _.get(response, 'data.data') || response.data;
    }
    catch (err) {
        logger.error(`Error in bulkModifyFiles specId: ${specId}, url: ${url}, err: ${err.message || err}`);
        const wrappedError = new Error('An error occurred while bulk modifying specification files: ' +
            (err.message || err));
        wrappedError.response = err.response;
        throw wrappedError;
    }
}
async function upsertSpecCollectionLinks(specId, collectionLinks, options) {
    if (!options.postmanApiKey && !options.accessToken) {
        throw new Error('No authorization data found. Please use the `postman login` command.');
    }
    const sanitizedLinks = collectionLinks.map((link) => {
        return {
            collectionId: link.collectionId,
            ...(link.options && { options: link.options }),
            ...(link.syncOptions && { syncOptions: link.syncOptions })
        };
    }), requestOptions = {
        method: 'put',
        headers: {
            'User-Agent': util.userAgent,
            'x-pstmn-req-service': 'api-specification',
            'x-postman-service': 'postman-api',
            'Content-Type': 'application/json'
        }
    }, url = `${util.POSTMAN_GATEWAY_BASE_URL()}/specifications/${specId}/collections`;
    if (options.postmanApiKey) {
        requestOptions.headers['x-api-key'] = options.postmanApiKey;
    }
    else {
        requestOptions.headers['x-access-token'] = options.accessToken;
    }
    try {
        const response = await axios.request({
            url: url,
            ...requestOptions,
            data: sanitizedLinks
        });
        if (!(/2../).test(response.status)) {
            const errorMessage = _.get(response, 'data.error.message') ||
                _.get(response, 'data.message') ||
                'Unknown error';
            throw new Error(`Failed to push spec-collection links: ${errorMessage}`);
        }
        return response.data || {};
    }
    catch (err) {
        logger.error(`Error in upsertSpecCollectionLinks specId: ${specId}, url: ${url}, err: ${err.message || err}`);
        if (_.has(err, 'response.data.error')) {
            const detail = err.response.data.error.detail ||
                err.response.data.error.title ||
                'No additional details provided.';
            throw new Error(`An error occurred while pushing spec-collection links: ${detail}`);
        }
        throw new Error(`An error occurred while pushing spec-collection links: ${err.message || err}`);
    }
}
module.exports = {
    fetchSpecificationDetails,
    fetchSpecificationFiles,
    getWorkspaceSpecifications,
    createSpecification,
    updateSpecificationFile,
    bulkModifyFiles,
    upsertSpecCollectionLinks
};
