// @postman/pm-bin-linux-x64/bin/lib/postman-entities/spec/Service/SyntaxParserService.js
const { FILE_TYPES, SEVERITY_MAP, VALIDATION_TYPES } = require('../constants');
function mapDiagnosticSeverity(severity) {
    return SEVERITY_MAP[severity] || 'ERROR';
}
function transformDiagnostics(diagnostics) {
    if (!Array.isArray(diagnostics)) {
        return [];
    }
    return diagnostics.map((diagnostic) => {
        return {
            message: diagnostic.message || '',
            path: diagnostic.path || [],
            range: diagnostic.range || null,
            severity: mapDiagnosticSeverity(diagnostic.severity),
            file: diagnostic.source || '',
            code: diagnostic.code || 'syntax-error',
            type: VALIDATION_TYPES.SYNTAX
        };
    });
}
async function parseSyntax({ specificationFileDetails }) {
    try {
        const OpenAPIParser = require('@postman/openapi-parser').default ||
            require('@postman/openapi-parser'), rootFile = specificationFileDetails.find((file) => {
            return file.type === FILE_TYPES.ROOT;
        });
        if (!rootFile) {
            logger.warn('No root file found for syntax parsing');
            return [];
        }
        const nonRootFiles = specificationFileDetails.filter((file) => {
            return file.type !== FILE_TYPES.ROOT;
        }), parser = new OpenAPIParser({
            __unstable: {
                resolver: {
                    resolvers: [
                        {
                            schema: 'file',
                            read: (uri) => {
                                const filePath = uri.path ? uri.path() : uri, file = nonRootFiles.find((f) => {
                                    return f.path === filePath;
                                });
                                return file?.content || '';
                            },
                            canRead: true,
                            order: 1
                        },
                        {
                            schema: 'http',
                            canRead: false,
                            read: () => {
                                return undefined;
                            }
                        },
                        {
                            schema: 'https',
                            canRead: false,
                            read: () => {
                                return undefined;
                            }
                        }
                    ]
                }
            },
            ruleset: {
                useRecommended: false
            }
        }), output = await parser.parse(rootFile.content, {
            source: rootFile.path,
            validateOptions: {
                allowedSeverity: {
                    error: true,
                    warning: true,
                    hint: true,
                    info: true
                }
            }
        }), formattedDiagnostics = transformDiagnostics(output.diagnostics || []);
        logger.info(`Syntax parsing complete. Found ${formattedDiagnostics.length} diagnostic(s).`);
        return formattedDiagnostics;
    }
    catch (error) {
        logger.error(`Error during syntax parsing: ${error.message || error}`);
        return [];
    }
}
module.exports = {
    parseSyntax
};
