// @postman/pm-bin-linux-x64/bin/lib/commands/monitor/run/monitor-runner.js
const _ = require('lodash'), colors = require('colors/safe'), print = require('../../../../lib/framework/print'), cliUtils = require('../../../reporters/cli/cli-utils'), monitorService = require('../monitor-service'), monitorOutput = require('../monitor-output'), config = require('../../../config'), util = require('../../../util'), { ERROR_MESSAGES: COMMON_ERROR_MESSAGES } = require('../../../constants');
const { getRunStatus } = require('../monitor-utils');
const ERROR_MESSAGES = {
    MONITOR_NOT_FOUND: 'Monitor not found. Please verify the monitor ID and try again.',
    UNAUTHORIZED_MONITOR_FETCH: 'Invalid API key. Please ensure you are logged in with a valid API key.',
    FORBIDDEN_MONITOR_FETCH: 'Access denied. You do not have permission to access this monitor.',
    UNAUTHORIZED_MONITOR_RUN: 'You are not authorized to run this monitor. ' +
        'Please ensure you are logged in with a valid API key.',
    FORBIDDEN_MONITOR_RUN: 'Access denied. You do not have permission to run this monitor.',
    NO_AUTH: COMMON_ERROR_MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE
};
const MONITOR_FETCH_ERROR_MAP = {
    401: ERROR_MESSAGES.UNAUTHORIZED_MONITOR_FETCH,
    403: ERROR_MESSAGES.FORBIDDEN_MONITOR_FETCH,
    404: ERROR_MESSAGES.MONITOR_NOT_FOUND
};
const MONITOR_RUN_ERROR_MAP = {
    401: ERROR_MESSAGES.UNAUTHORIZED_MONITOR_RUN,
    403: ERROR_MESSAGES.FORBIDDEN_MONITOR_RUN
};
const JOB_POLLING_ERROR_MAP = {
    401: 'Invalid API key. Please ensure you are logged in with a valid API key.',
    403: 'Access denied. You do not have permission to access this job status.',
    404: 'Job not found. The monitor run may have expired or been deleted.'
};
function getMonitorRunUrl(monitorId, jobId) {
    return `${util.POSTMAN_ARTEMIS_BASE_URL()}/monitor/${monitorId}?job=${jobId}`;
}
function getAuth() {
    return new Promise((resolve, reject) => {
        config.get({}, { command: 'default' }, (err, options) => {
            if (err) {
                return reject(err);
            }
            util.extractPostmanApiKey(options.postmanApiKey, (extractErr, apiKey) => {
                if (extractErr) {
                    return reject(extractErr);
                }
                const authResult = {
                    postmanApiKey: apiKey,
                    accessToken: options.accessToken
                };
                return resolve(authResult);
            });
        });
    });
}
function getErrorMessage(error) {
    const errorMsg = _.get(error, 'response.data.error.message', error.message);
    return errorMsg.replace('JobTemplate', 'Monitor');
}
async function enrichMonitorInfo({ monitor, postmanApiKey, accessToken }) {
    const enrichedMonitor = _.cloneDeep(monitor);
    const monitorData = enrichedMonitor.monitor;
    if (!monitorData) {
        return enrichedMonitor;
    }
    try {
        if (monitorData.collectionUid) {
            const collection = await monitorService.getCollectionById({
                collectionId: monitorData.collectionUid,
                postmanApiKey: postmanApiKey,
                accessToken: accessToken
            });
            monitorData.collectionName = _.get(collection, 'info.name');
        }
        if (monitorData.environmentUid) {
            const environment = await monitorService.getEnvironmentById({
                environmentId: monitorData.environmentUid,
                postmanApiKey: postmanApiKey,
                accessToken: accessToken
            });
            monitorData.environmentName = _.get(environment, 'name');
        }
    }
    catch (error) {
        logger.error(`Error enriching monitor info: ${error.message}`);
    }
    return enrichedMonitor;
}
async function fetchMonitor({ monitorId, postmanApiKey, accessToken }) {
    const symbols = cliUtils.symbols();
    print('Fetching monitor details... ');
    print().wait();
    try {
        const monitor = await monitorService.getMonitor({
            monitorId,
            postmanApiKey,
            accessToken
        });
        const enrichedMonitor = await enrichMonitorInfo({ monitor, postmanApiKey, accessToken });
        print.lf(colors.green(symbols.ok));
        return enrichedMonitor;
    }
    catch (error) {
        print.lf(colors.red(symbols.error));
        const msg = (error.response && MONITOR_FETCH_ERROR_MAP[error.response.status]) || getErrorMessage(error);
        throw new Error(msg);
    }
}
const POLLING_THRESHOLDS = {
    VERY_LONG: 30 * 60 * 1000,
    LONG: 10 * 60 * 1000,
    MEDIUM: 5 * 60 * 1000,
    SHORT: 60 * 1000,
    VERY_SHORT: 30 * 1000
};
const POLLING_INTERVALS = {
    VERY_LONG: 5 * 60 * 1000,
    LONG: 2 * 60 * 1000,
    MEDIUM: 60 * 1000,
    SHORT: 30 * 1000,
    VERY_SHORT: 10 * 1000,
    DEFAULT: 5 * 1000
};
function calculateSmartPollingConfig({ monitor, userTimeoutMs }) {
    const lastRun = _.get(monitor, 'monitor.lastRun');
    let pollingInterval = POLLING_INTERVALS.DEFAULT;
    if (lastRun && lastRun.startedAt && lastRun.finishedAt) {
        const lastRunDuration = new Date(lastRun.finishedAt) - new Date(lastRun.startedAt);
        if (lastRunDuration > POLLING_THRESHOLDS.VERY_LONG) {
            pollingInterval = POLLING_INTERVALS.VERY_LONG;
        }
        else if (lastRunDuration > POLLING_THRESHOLDS.LONG) {
            pollingInterval = POLLING_INTERVALS.LONG;
        }
        else if (lastRunDuration > POLLING_THRESHOLDS.MEDIUM) {
            pollingInterval = POLLING_INTERVALS.MEDIUM;
        }
        else if (lastRunDuration > POLLING_THRESHOLDS.SHORT) {
            pollingInterval = POLLING_INTERVALS.SHORT;
        }
        else if (lastRunDuration > POLLING_THRESHOLDS.VERY_SHORT) {
            pollingInterval = POLLING_INTERVALS.VERY_SHORT;
        }
        else {
            pollingInterval = POLLING_INTERVALS.DEFAULT;
        }
    }
    if (userTimeoutMs !== null && userTimeoutMs > 0 && pollingInterval > userTimeoutMs / 3) {
        pollingInterval = Math.max(5000, Math.floor(userTimeoutMs / 4));
    }
    return {
        INTERVAL_MS: pollingInterval,
        MAX_ATTEMPTS: userTimeoutMs !== null ? Math.ceil(userTimeoutMs / pollingInterval) : Infinity,
        HAS_TIMEOUT: userTimeoutMs !== null,
        TIMEOUT_MS: userTimeoutMs,
        LAST_RUN_DURATION: (lastRun && lastRun.startedAt && lastRun.finishedAt) ?
            (new Date(lastRun.finishedAt) - new Date(lastRun.startedAt)) : null
    };
}
function calculateSleepInterval(pollingConfig, startTime) {
    let sleepInterval = pollingConfig.INTERVAL_MS;
    if (pollingConfig.HAS_TIMEOUT) {
        const elapsedTime = Date.now() - startTime;
        const remainingTimeout = pollingConfig.TIMEOUT_MS - elapsedTime;
        if (remainingTimeout <= pollingConfig.INTERVAL_MS) {
            sleepInterval = Math.max(1000, remainingTimeout - 1000);
        }
    }
    return sleepInterval;
}
function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
async function pollForRunCompletion({ monitorId, jobId, postmanApiKey, accessToken, pollingConfig }) {
    const symbols = cliUtils.symbols();
    let attempts = 0;
    const startTime = Date.now();
    while (pollingConfig.HAS_TIMEOUT ? attempts < pollingConfig.MAX_ATTEMPTS : true) {
        try {
            const jobDetails = await monitorService.getJobDetails({
                jobId,
                postmanApiKey,
                accessToken
            });
            if (jobDetails.finishedAt) {
                print.lf(colors.green(symbols.ok));
                return jobDetails;
            }
            const sleepInterval = calculateSleepInterval(pollingConfig, startTime);
            if (sleepInterval <= 0) {
                break;
            }
            await sleep(sleepInterval);
            attempts++;
        }
        catch (error) {
            print.lf(colors.red(symbols.error));
            const msg = (error.response && JOB_POLLING_ERROR_MAP[error.response.status]) || getErrorMessage(error);
            throw new Error(msg);
        }
    }
    throw new Error(`Monitor run polling timed out after ${util.prettyms(pollingConfig.TIMEOUT_MS)}. ` +
        'The run may still be in progress.');
}
async function executeMonitorRun({ monitorName, monitorId, postmanApiKey, accessToken, pollingConfig }) {
    const symbols = cliUtils.symbols();
    print(`Submitting run request for monitor "${monitorName}"... `);
    print().wait();
    try {
        const runResponse = await monitorService.runMonitor({
            monitorId,
            postmanApiKey,
            accessToken
        });
        const jobId = _.get(runResponse, 'data.data.id');
        if (!jobId) {
            print.lf(colors.red(symbols.error));
            throw new Error('Monitor run started but no job ID was returned for polling.');
        }
        print.lf(colors.green(symbols.ok));
        print('Waiting for monitor run to complete ... ');
        print().wait();
        const completedRun = await pollForRunCompletion({
            monitorId,
            jobId,
            postmanApiKey,
            accessToken,
            pollingConfig
        });
        return { data: { run: completedRun }, statusCode: 200 };
    }
    catch (error) {
        print.lf(colors.red(symbols.error));
        const msg = (error.response && MONITOR_RUN_ERROR_MAP[error.response.status]) || getErrorMessage(error);
        throw new Error(msg);
    }
}
async function runMonitor(monitorId, options = {}) {
    if (!monitorId || !monitorId.trim()) {
        throw new Error('Monitor ID must not be empty. Please provide a valid monitor ID.');
    }
    const auth = await getAuth();
    if (!auth.postmanApiKey && !auth.accessToken) {
        throw new Error(ERROR_MESSAGES.NO_AUTH);
    }
    try {
        const monitor = await fetchMonitor({
            monitorId: monitorId,
            postmanApiKey: auth.postmanApiKey,
            accessToken: auth.accessToken
        });
        const userTimeoutMs = !_.isNil(options.timeout) ?
            parseInt(options.timeout, 10) : null;
        const pollingConfig = calculateSmartPollingConfig({ monitor, userTimeoutMs });
        monitorOutput.displayMonitorDetails(monitor);
        const runResponse = await executeMonitorRun({
            monitorName: _.get(monitor, 'monitor.name'),
            monitorId: monitorId,
            postmanApiKey: auth.postmanApiKey,
            accessToken: auth.accessToken,
            pollingConfig: pollingConfig
        });
        monitorOutput.displayMonitorRunResults(_.get(runResponse, 'data.run'));
        const result = {
            success: getRunStatus(_.get(runResponse, 'data.run', {})) === 'success',
            monitorRunUrl: getMonitorRunUrl(_.get(monitor, 'monitor.id'), _.get(runResponse, 'data.run.jobId'))
        };
        if (result.monitorRunUrl) {
            print.lf(colors.gray('View the detailed run report in the Postman app \n') + result.monitorRunUrl + '\n');
        }
        return result;
    }
    catch (error) {
        print.unwait();
        throw error;
    }
}
module.exports = {
    runMonitor
};
