// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/formatters/utils/getRangeAndFileForConfigMode.js
const { getLineNumberForSchema } = require('./lineService');
function getRangeAndFileForConfigMode(result, api) {
    const pathsArr = [...result.path], referenceMap = api.referenceMap;
    if (!referenceMap.root) {
        referenceMap.root = api.rootFile;
    }
    let generatedPath = '#', relevantReferenceMap, index;
    for (let i = 0; i < pathsArr.length; i++) {
        generatedPath += `/${pathsArr[i]}`;
        if (referenceMap[generatedPath]) {
            relevantReferenceMap = referenceMap[generatedPath];
            index = i;
        }
    }
    if (!relevantReferenceMap) {
        relevantReferenceMap = referenceMap.root;
    }
    let mapSplit = relevantReferenceMap.path.split('#/'), modifiedPath = pathsArr.splice(index);
    if (mapSplit[1]) {
        modifiedPath[0] = mapSplit[1];
    }
    const fileDetails = api.allFiles.filter((x) => { return x.path === mapSplit[0]; })[0], rangeDetails = getLineNumberForSchema(fileDetails.ast, modifiedPath, fileDetails.language), range = {
        start: {
            line: rangeDetails.startLineNumber,
            character: rangeDetails.startColumn
        },
        end: {
            line: rangeDetails.endLineNumber,
            character: rangeDetails.endColumn
        }
    };
    return {
        range: range,
        file: fileDetails.path
    };
}
module.exports = { getRangeAndFileForConfigMode };
