// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/formatters/utils/lineService.js
const _ = require('lodash'), MAP_TAG = 'tag:yaml.org,2002:map', SEQ_TAG = 'tag:yaml.org,2002:seq';
function computeLineNumberForYAMLSchema(current, path, last) {
    if (!current) {
        if (last && last.start_mark) {
            const range = {};
            range.startLineNumber = _.get(last, 'start_mark.line', 0);
            range.startColumn = _.get(last, 'start_mark.column', 0);
            if (last.end_mark) {
                range.endLineNumber = _.get(last, 'end_mark.line', 0);
                range.endColumn = _.get(last, 'end_mark.column', 0);
            }
            return range;
        }
        return {
            startLineNumber: 0,
            startColumn: 0,
            endLineNumber: 0,
            endColumn: 0
        };
    }
    if (path.length && current.tag === MAP_TAG) {
        for (let i = 0; i < _.get(current, 'value.length', 0); i++) {
            const pair = current.value[i], key = pair[0], value = pair[1];
            if (key.value === path[0]) {
                return computeLineNumberForYAMLSchema(value, path.slice(1), current);
            }
        }
    }
    if (path.length && current.tag === SEQ_TAG) {
        var item = current.value[path[0]];
        if (item && item.tag) {
            return computeLineNumberForYAMLSchema(item, path.slice(1), current.value);
        }
    }
    return {
        startLineNumber: current.start_mark.line,
        startColumn: current.start_mark.column,
        endLineNumber: current.end_mark.line,
        endColumn: current.end_mark.column
    };
}
function computeLineNumberForJSONSchema(current, path, last) {
    if (!current) {
        if (last && last.loc.start.line) {
            const range = {};
            range.startLineNumber = _.get(last, 'loc.start.line', 1) - 1;
            range.startColumn = _.get(last, 'loc.start.column', 1);
            if (last.loc.end) {
                range.endLineNumber = _.get(last, 'loc.end.line', 1) - 1;
                range.endColumn = _.get(last, 'loc.end.column', 1);
            }
            return range;
        }
        return {
            startLineNumber: 0,
            startColumn: 0,
            endLineNumber: 0,
            endColumn: 0
        };
    }
    if (path.length && current.type === 'Object') {
        for (let i = 0; i < current.children.length; i++) {
            if (current.children[i].key.value === path[0]) {
                return computeLineNumberForJSONSchema(current.children[i].value, path.slice(1), current);
            }
        }
    }
    if (path.length && current.type === 'Array') {
        return computeLineNumberForJSONSchema(current.children[path[0]], path.slice(1), current);
    }
    if (path.length && current.type === 'Property') {
        if (current.key.value === path[0]) {
            return computeLineNumberForJSONSchema(current.value, path.slice(1), current);
        }
    }
    return {
        startLineNumber: current.loc.start.line - 1,
        startColumn: current.loc.start.column,
        endLineNumber: current.loc.end.line - 1,
        endColumn: current.loc.end.column
    };
}
function getLineNumberForSchema(ast, path, type) {
    if (!_.isArray(path)) {
        throw new TypeError('path should be an array of strings');
    }
    if (type === 'json') {
        return computeLineNumberForJSONSchema(ast, path);
    }
    return computeLineNumberForYAMLSchema(ast, path);
}
module.exports = {
    getLineNumberForSchema
};
