// @postman/pm-bin-linux-x64/bin/lib/login/browser-auth-success-static.html.js
module.exports = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=320, initial-scale=1, maximum-scale=1, user-scalable=no">
  <title>Postman CLI - Login</title>
  <link rel="shortcut icon" href="/favicon.ico">
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;600&display=swap" rel="stylesheet">

  <style>
    /* CSS Variables - Only used ones */
    :root {
      --content-color-primary: #212121;
      --content-color-secondary: #6B6B6B;
      --content-color-link: #0265D2;
      --background-color-primary: #FFFFFF;
      --background-color-secondary: #F9F9F9;
      --text-family-default: Inter,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans','Droid Sans',Helvetica,Arial,sans-serif;
      --text-size-s: 11px;
      --text-size-l: 14px;
      --text-size-xxxl: 21px;
      --text-weight-bold: 600;
      --line-height-s: 16px;
      --line-height-l: 24px;
      --line-height-m: 20px;
      --line-height-xl: 28px;
      --spacing-s: 8px;
      --spacing-l: 16px;
      --spacing-xl: 24px;
      --spacing-xxl: 32px;
      --spacing-xxxl: 48px;
      --border-radius-l: 8px;
      --shadow-card: 0 0 0 1px rgba(0,0,0,.04),0 4px 16px -4px rgba(0,0,0,.16);
    }

    /* Reset and Base Styles */
    * {
      box-sizing: border-box;
    }

    body {
      font-family: var(--text-family-default);
      margin: 0;
      padding: 0;
      background-color: var(--background-color-secondary);
      color: var(--content-color-primary);
      min-height: 100vh;
    }

    /* Layout */
    .container {
      align-items: center;
      display: flex;
      flex-direction: column;
      min-height: 100vh;
      justify-content: space-between;
      margin: 0;
      padding: var(--spacing-l);
      position: relative;
      width: 100%;
    }

    .main {
      padding: var(--spacing-l);
      width: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    @media screen and (min-width: 720px) {
      .main {
        padding: var(--spacing-xxl) var(--spacing-l);
      }
    }

    /* Card Styles */
    .card {
      background-color: var(--background-color-primary);
      box-shadow: var(--shadow-card);
      border-radius: var(--border-radius-l);
      margin: var(--spacing-xxl) auto var(--spacing-xl);
      padding: var(--spacing-xxl);
      position: relative;
      width: 100%;
      max-width: 360px;
    }

    @media screen and (min-width: 720px) {
      .card {
        margin: var(--spacing-xxxl) auto var(--spacing-xxl);
        padding: var(--spacing-xxl);
        width: 360px;
      }
    }

    /* Logo */
    .logo {
      display: block;
      margin: 0 auto var(--spacing-xl) auto;
      text-align: center;
      width: 120px;
      height: auto;
    }

    /* Typography */
    .heading-1 {
      font-size: var(--text-size-xxxl);
      font-weight: var(--text-weight-bold);
      letter-spacing: -0.32px;
      line-height: var(--line-height-xl);
      margin: 0;
    }

    .text-large {
      font-size: var(--text-size-l);
      line-height: var(--line-height-l);
    }

    .text-center {
      text-align: center;
    }

    .spacing-top-l {
      margin-top: var(--spacing-l);
    }

    /* Footer */
    .footer {
      font-family: var(--text-family-default);
      line-height: var(--line-height-m);
      font-size: var(--text-size-s);
      width: 100%;
      text-align: center;
      color: var(--content-color-secondary);
      padding: var(--spacing-l);
    }

    .text-subdued {
      color: var(--content-color-secondary);
    }

    .link-muted {
      color: var(--content-color-secondary);
      text-decoration: none;
    }

    .link-muted:hover {
      text-decoration: underline;
    }

    .separator {
      margin: 0 var(--spacing-s);
      font-style: normal;
    }
  </style>
</head>
<body>
  <div class="container">
    <div class="main">
      <div class="card">
        <img src="https://static.getpostman.com/assets/pm-logo-1.svg" alt="Postman Logo" class="logo">
        <h1 class="heading-1 text-center">Authentication Successful!</h1>
        <p class="text-large text-center spacing-top-l">You can now return to your terminal to continue.</p>
      </div>
    </div>

    <footer class="footer">
      <span translate="no" class="text-subdued">©2025 Postman, Inc. All rights reserved.</span>
      <i class="separator">·</i>
      <a class="link-muted" href="https://www.postman.com/legal/eula" target="_blank">Terms of use</a>
      <i class="separator">·</i>
      <a class="link-muted" href="https://www.postman.com/legal/privacy-policy" target="_blank">Privacy policy</a>
    </footer>
  </div>
</body>
</html>`;
