// @postman/pm-bin-linux-x64/bin/lib/login/index.js
const _ = require('lodash'), waterfall = require('async/waterfall'), colors = require('colors/safe'), prompts = require('prompts'), rcfile = require('../config/rc-file'), print = require('../framework/print'), util = require('../util'), console = require('../framework/console'), regionUtil = require('../region-util'), PKCEAuth = require('./pkce-auth'), { REGIONS } = require('../constants'), DEFAULT = 'default', USER_ACTION = 'userAction', ALIAS = 'alias', REGION = 'region', OVERRIDE_PERMISSION = 'overridePermission', POSTMAN_API_KEY = 'postmanApiKey', INPUT_PROMPTS = {
    userAction: 'Default authentication details already exist.\nChoose an option to continue',
    userActionChoices: [
        { title: 'Override it', value: 'override' },
        { title: 'Store as an alias', value: 'new_alias' },
        { title: 'Abort', value: 'abort' }
    ],
    alias: 'Alias',
    overridePermission: 'The alias already exists.\nDo you want to override it?',
    postmanApiKey: 'Postman API Key'
}, V1_API_KEY_REGEX = /^[a-f0-9]{32}$/, V2_API_KEY_REGEX = /^(PMAK-[a-f0-9]{24}-[a-f0-9]{34})$/, INVALID_API_KEY = 'The provided Postman API key is invalid. Please check your key and region, then try again.' +
    '\n\n  Refer: https://learning.postman.com/docs/postman-cli/postman-cli-options/#postman-login', ABORT_MESSAGE = 'Login aborted.', INVALID_INPUT = 'Invalid input.', KEY_VALIDATION_FALLBACK_ERROR_MSG = 'Error occurred in validating API key. Please try again.';
function validateRegion(region) {
    if (!region) {
        return REGIONS.US;
    }
    const normalizedRegion = region.toLowerCase();
    if (normalizedRegion === REGIONS.EU) {
        return REGIONS.EU;
    }
    throw new Error('Invalid region provided. Use "eu" for EU region, or omit --region for default region.');
}
module.exports = (options, callback) => {
    if (options.alias) {
        console.warn(colors.yellow('[warning] --alias is discontinued and will be removed in a future release.\n'));
    }
    let targetRegion, alias, fileData;
    try {
        targetRegion = validateRegion(options.region);
    }
    catch (err) {
        return callback(err);
    }
    const getData = async (field, opts, cb) => {
        let promptOptions = {
            type: opts.type,
            name: 'response',
            hint: ' ',
            message: (INPUT_PROMPTS[field])
        };
        if (opts.boolean) {
            promptOptions = {
                ...promptOptions,
                active: 'Yes',
                inactive: 'No'
            };
        }
        if (opts.choices) {
            promptOptions = {
                ...promptOptions,
                choices: opts.choices
            };
        }
        !_.isUndefined(opts.initial) && (promptOptions.initial = opts.initial);
        util.signal(field);
        var { response } = await prompts.prompt(promptOptions);
        if (_.isString(response)) {
            response = response.trim();
        }
        _.set(options, field, response);
        if (_.isUndefined(options[field])) {
            process.exit(1);
        }
        if (options[field] === '') {
            return cb(INVALID_INPUT);
        }
        return cb(null);
    }, validateKey = (key, callback) => {
        if (!V1_API_KEY_REGEX.test(options.postmanApiKey) &&
            !V2_API_KEY_REGEX.test(options.postmanApiKey)) {
            return callback(null, { isValidKey: false, region: null });
        }
        regionUtil.validateApiKeyForRegion(key, targetRegion, ['user.username', 'user.id'], (err, validRegionInfo) => {
            if (err) {
                return callback(err);
            }
            if (validRegionInfo && validRegionInfo.isValidKey) {
                const username = _.get(validRegionInfo, 'authInfo.user.username'), userId = _.get(validRegionInfo, 'authInfo.user.id');
                print.lf(`Logged in using api key of user: ${username} `);
                if (validRegionInfo.region !== REGIONS.US) {
                    print.lf(colors.yellow(`\nPostman CLI is logging in to the ${validRegionInfo.region} region. ` +
                        '\nPlease note that some features may not be available in this region.'));
                }
                return callback(null, {
                    isValidKey: true,
                    region: validRegionInfo.region,
                    username: username,
                    userId: userId
                });
            }
            return callback(null, {
                isValidKey: false,
                region: targetRegion,
                error: INVALID_API_KEY
            });
        });
    };
    alias = options.alias || DEFAULT;
    if (options.withApiKey) {
        waterfall([
            rcfile.load,
            (data, next) => {
                options.postmanApiKey = options.withApiKey.trim();
                validateKey(options.postmanApiKey, (err, result) => {
                    if (err) {
                        if (err && err.cause === regionUtil.BAD_API_KEY_ERROR) {
                            return next(util.augmentError(new Error(INVALID_API_KEY), err.message));
                        }
                        return next(util.augmentError(new Error(KEY_VALIDATION_FALLBACK_ERROR_MSG), err.message));
                    }
                    const { isValidKey, region, username, userId } = result;
                    if (!isValidKey) {
                        let profile = _.find(_.get(data, 'login._profiles'), _.matchesProperty('alias', alias));
                        if (profile) {
                            let userError = profile.username ? `for user - ${profile.username}` : '', errMessage = INVALID_API_KEY +
                                ((alias || options.alias) !== 'default' ?
                                    `\nDefaulting to already stored API key with alias ${options.alias} ${userError}` :
                                    `\nDefaulting to already stored API key ${userError}`);
                            return next(new Error(errMessage));
                        }
                        return next(new Error(INVALID_API_KEY));
                    }
                    return next(null, data, region, username, userId);
                });
            },
            (data, region, username, userId, next) => {
                fileData = data;
                options.alias = options.alias || DEFAULT;
                !fileData && (fileData = {});
                !fileData.login && (fileData.login = {});
                !fileData.login._profiles && (fileData.login._profiles = []);
                if (_.has(fileData, 'login._profiles')) {
                    fileData.login._profiles = _.reject(fileData.login._profiles, [ALIAS, options.alias]);
                }
                regionUtil.setCurrentRegionForProcess(region);
                fileData.login._profiles.push({
                    [ALIAS]: options[ALIAS],
                    [POSTMAN_API_KEY]: options[POSTMAN_API_KEY],
                    [REGION]: region,
                    username: username,
                    userId: userId
                });
                return next(null, fileData);
            },
            rcfile.store
        ], callback);
    }
    else {
        waterfall([
            rcfile.load,
            (data, next) => {
                fileData = data;
                let defaultAliasDetails = _.has(fileData, 'login._profiles') &&
                    _.filter(fileData.login._profiles, [ALIAS, alias]);
                if (_.isEmpty(defaultAliasDetails)) {
                    options.alias = alias;
                    return next(null);
                }
                return getData(USER_ACTION, { type: 'select', choices: INPUT_PROMPTS.userActionChoices }, () => {
                    if (options.userAction === 'abort') {
                        process.exit(1);
                    }
                    print('\n');
                    if (options.userAction === 'override') {
                        options.alias = alias;
                    }
                    return next(null);
                });
            },
            (next) => {
                if (options.alias) {
                    return next(null);
                }
                return getData(ALIAS, { type: 'text' }, next);
            },
            (next) => {
                let previousData = _.has(fileData, 'login._profiles') &&
                    _.filter(fileData.login._profiles, [ALIAS, options.alias]);
                if (_.isEmpty(previousData)) {
                    return next(null);
                }
                if (options.alias === 'default') {
                    return next(null);
                }
                return getData(OVERRIDE_PERMISSION, { type: 'toggle', boolean: true, initial: true }, () => {
                    if (!options.overridePermission) {
                        console.error(ABORT_MESSAGE);
                        process.exit(1);
                    }
                    return next(null);
                });
            },
            (next) => {
                const pkceAuth = new PKCEAuth(options.region);
                pkceAuth.authenticate()
                    .then((tokenInfo) => {
                    const username = (tokenInfo.user &&
                        (tokenInfo.user.email || tokenInfo.user.username)) || 'Unknown User';
                    if (_.has(fileData, 'login._profiles')) {
                        fileData.login._profiles = _.reject(fileData.login._profiles, [ALIAS, options.alias]);
                    }
                    next(null, {
                        accessToken: tokenInfo.accessToken,
                        user: tokenInfo.user,
                        username: username
                    });
                }, () => {
                    next(new Error('Authentication failed: Unable to connect to Postman authentication services. Please check your internet connection and try again. If the issue persists, contact Postman support.'));
                });
            },
            (authInfo, next) => {
                try {
                    const region = regionUtil.getCurrentRegion(), iapubBaseUrl = util.POSTMAN_IAPUB_BASE_URL(region), requestFactory = require('@postman/platform-req/lib/request'), client = requestFactory.defaults({ json: true, timeout: 10000 });
                    client.get({
                        url: `${iapubBaseUrl}/api/sessions/current`,
                        headers: { 'x-access-token': authInfo.accessToken, 'User-Agent': util.userAgent }
                    }, (err, response) => {
                        if (!err && response && response.statusCode === 200 &&
                            response.body && response.body.session && response.body.session.identity) {
                            authInfo.userId = response.body.session.identity.user;
                        }
                        return next(null, authInfo);
                    });
                }
                catch (e) {
                    return next(null, authInfo);
                }
            },
            (authInfo, next) => {
                !fileData && (fileData = {});
                !fileData.login && (fileData.login = {});
                !fileData.login._profiles && (fileData.login._profiles = []);
                fileData.login._profiles.push({
                    [ALIAS]: options[ALIAS],
                    accessToken: authInfo.accessToken,
                    username: authInfo.username,
                    userId: authInfo.userId,
                    ...(options.region && { [REGION]: options.region })
                });
                return next(null, fileData);
            },
            rcfile.store
        ], callback);
    }
};
