// @postman/pm-bin-linux-x64/bin/lib/login/pkce-auth.js
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
const crypto = require('crypto');
const http = require('http');
const { URL } = require('url');
const axios = require('axios');
const colors = require('colors/safe');
const open = require('open');
const isCI = require('is-ci');
const util = require('../util');
const print = require('../framework/print');
const cliUtils = require('../reporters/cli/cli-utils');
const browserAuthSuccessStatic = require('./browser-auth-success-static.html');
const { getBanner } = require('./banner');
const REDIRECT_URI_BASE = 'http://localhost';
const AUTH_TIMEOUT_MS = 5 * 60 * 1000;
class PKCEAuth {
    constructor(region = 'us') {
        this.server = null;
        this.codeVerifier = null;
        this.codeChallenge = null;
        this.redirectUri = null;
        this.port = 64203;
        this.baseURL = util.POSTMAN_IAPUB_BASE_URL(region);
        this.maxPortAttempts = 100;
    }
    generatePKCEKeyPair() {
        this.codeVerifier = crypto.randomBytes(32).toString('hex');
        this.codeChallenge = crypto.createHash('sha256').update(this.codeVerifier).digest('hex');
        return {
            codeVerifier: this.codeVerifier,
            codeChallenge: this.codeChallenge
        };
    }
    startLocalServer() {
        return this.tryStartServer(this.port, 0);
    }
    tryStartServer(port, attempt) {
        if (attempt > this.maxPortAttempts) {
            return Promise.reject(new Error(`Failed to start local server. Tried ports: ${port - this.maxPortAttempts} to ${port - 1}.`));
        }
        const server = http.createServer();
        return new Promise((resolve, reject) => {
            server.listen(port, 'localhost', () => {
                this.port = port;
                this.redirectUri = `${REDIRECT_URI_BASE}:${this.port}/auth/callback`;
                this.server = server;
                if (attempt > 0) {
                    print.lf(colors.yellow('⚠️  Default port was occupied, using port %d instead'), port);
                }
                resolve();
            });
            server.on('error', (error) => {
                if (error.code === 'EADDRINUSE' && attempt < this.maxPortAttempts) {
                    const nextPort = port + 1;
                    resolve(this.tryStartServer(nextPort, attempt + 1));
                }
                else if (error.code === 'EADDRINUSE') {
                    const errorMsg = `Unable to find an available port after ${this.maxPortAttempts + 1} attempts. ` +
                        'Please try again later.';
                    reject(new Error(errorMsg));
                }
                else {
                    reject(error);
                }
            });
        });
    }
    waitForAuthorizationCode() {
        return new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                this.cleanup();
                reject(new Error('Authentication timeout. Please try again.'));
            }, AUTH_TIMEOUT_MS);
            this.server.on('request', (req, res) => {
                if (!req.url || !req.url.startsWith('/auth/callback')) {
                    res.writeHead(404, { 'Content-Type': 'text/plain' });
                    res.end('Not Found');
                    return;
                }
                const url = new URL(req.url, this.redirectUri);
                const authCode = url.searchParams.get('code');
                const error = url.searchParams.get('error');
                const errorDescription = url.searchParams.get('error_description');
                if (error) {
                    res.writeHead(400, { 'Content-Type': 'text/html' });
                    res.end(`
                        <html>
                            <body>
                                <h2>Authentication Failed</h2>
                                <p>Error: ${error}</p>
                                <p>Description: ${errorDescription || 'Unknown error occurred'}</p>
                                <p>You can close this window and try again.</p>
                            </body>
                        </html>
                    `);
                    clearTimeout(timeout);
                    this.cleanup();
                    reject(new Error(`Authentication failed: ${error} - ${errorDescription}`));
                }
                else if (authCode) {
                    res.writeHead(200, { 'Content-Type': 'text/html' });
                    res.end(browserAuthSuccessStatic);
                    clearTimeout(timeout);
                    resolve(authCode);
                }
                else {
                    res.writeHead(400, { 'Content-Type': 'text/html' });
                    res.end(`
                        <html>
                            <body>
                                <h2>Invalid Request</h2>
                                <p>No authorization code received. Please try again.</p>
                            </body>
                        </html>
                    `);
                }
            });
        });
    }
    async initiateAuthFlow() {
        try {
            const requestPayload = {
                challenge: this.codeChallenge,
                redirect_uri: this.redirectUri,
                target: 'login'
            };
            const userAgent = 'postman-cli-' + (util.version.match(/^(\d+\.\d+\.\d)/)[1] || util.version);
            const response = await axios.post(`${this.baseURL}/api/public/pkce/exchange/init`, requestPayload, {
                headers: {
                    'User-Agent': userAgent
                }
            });
            if (response.status !== 200) {
                throw new Error(`Failed to initiate auth flow: ${response.status} ${response.data}`);
            }
            const authUrl = response.data.redirectUrl;
            if (!authUrl) {
                throw new Error('Invalid response: Missing authorization URL');
            }
            let finalUrl = authUrl;
            if (!(/^https?:\/\//).test(finalUrl)) {
                finalUrl = `https://${finalUrl}`;
            }
            return finalUrl;
        }
        catch (error) {
            if (error.response) {
                throw new Error(`Failed to initiate auth flow: ${JSON.stringify(error.response.data)}`);
            }
            throw error;
        }
    }
    async exchangeCodeForToken(authorizationCode) {
        try {
            const endpoint = `${this.baseURL}/api/handover/pkce/${this.codeVerifier}/consume`;
            const requestPayload = { authorizationCode };
            const response = await axios.post(endpoint, requestPayload);
            if (response.status !== 200) {
                throw new Error(`Failed to exchange code for token: ${response.status} ${response.statusText}`);
            }
            const data = response.data;
            if (!data ||
                !data.refreshToken ||
                !data.refreshToken.access_token ||
                !data.sessionData ||
                !data.sessionData.data ||
                !data.sessionData.data.user) {
                throw new Error('Invalid response: Missing required authentication data');
            }
            const tokenInfo = {
                accessToken: data.refreshToken.access_token,
                user: data.sessionData.data.user,
                expiresIn: data.refreshToken.expires_in
            };
            return tokenInfo;
        }
        catch (error) {
            if (error.response) {
                throw new Error(`Failed to exchange code for token: ${error.response.status} ${error}`);
            }
            throw error;
        }
    }
    static async openBrowser(authUrl) {
        try {
            let urlObj;
            try {
                urlObj = new URL(authUrl);
            }
            catch (e) {
                return false;
            }
            if (urlObj.protocol !== 'http:' && urlObj.protocol !== 'https:') {
                return false;
            }
            await open(authUrl, { wait: false });
            return true;
        }
        catch (error) {
            return false;
        }
    }
    async authenticate() {
        try {
            print.lf('Logging in using Postman Cloud.');
            this.generatePKCEKeyPair();
            await this.startLocalServer();
            const authUrl = await this.initiateAuthFlow();
            const browserOpened = await PKCEAuth.openBrowser(authUrl);
            if (browserOpened) {
                print.lf(colors.gray('→ Please complete the authentication in the browser'));
            }
            else {
                print.lf(colors.gray('→ Please open the following URL in your browser to continue:'));
                print.lf('   %s', authUrl);
            }
            print.lf('');
            print.print('Waiting for authentication to complete ');
            print.wait();
            const authCode = await this.waitForAuthorizationCode();
            print.unwait();
            print.lf(colors.green('✓'));
            const tokenInfo = await this.exchangeCodeForToken(authCode);
            const user = tokenInfo.user || {};
            const username = user.email || user.username || 'Unknown User';
            const name = user.name || username;
            const email = user.email;
            const teamName = user.teamName;
            if (!isCI) {
                const isWindows = (/^win/).test(process.platform), asciiArt = getBanner(1);
                let colorSupport = null;
                try {
                    const supportsColor = await Promise.resolve().then(() => __importStar(require('supports-color')));
                    colorSupport = supportsColor.default.stdout;
                }
                catch (error) {
                    colorSupport = null;
                }
                if (isWindows) {
                    const { exec } = require('child_process');
                    exec('cls', () => {
                    });
                }
                else {
                    print.print('\x1B[2J\x1B[0f');
                }
                if (cliUtils.noTTY() || !colors.enabled) {
                    print.lf(asciiArt);
                }
                else if (colorSupport && colorSupport.has16m) {
                    print.lf('\x1b[38;2;255;108;55m' + asciiArt + '\x1b[0m');
                }
                else if (colorSupport && colorSupport.has256) {
                    print.lf('\x1b[38;5;214m' + asciiArt + '\x1b[0m');
                }
                else if (colorSupport && colorSupport.hasBasic) {
                    print.lf(colors.yellow(asciiArt));
                }
                else {
                    print.lf(asciiArt);
                }
            }
            if (email) {
                print.lf('Welcome %s <%s>', colors.cyan(name), colors.cyan(email));
            }
            else {
                print.lf('Welcome %s!', colors.cyan(name));
            }
            if (teamName) {
                print.lf('  Team: %s', colors.cyan(teamName));
            }
            print.lf('');
            return tokenInfo;
        }
        catch (error) {
            print.lf(colors.red('\nAuthentication failed: %s'), colors.red(error.message));
            this.cleanup();
            throw error;
        }
    }
    cleanup() {
        if (this.server) {
            this.server.close();
            this.server = null;
        }
        this.codeVerifier = null;
        this.codeChallenge = null;
        this.redirectUri = null;
        this.port = null;
    }
}
module.exports = PKCEAuth;
