// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/apiPublishContentFetchers.js
const _ = require('lodash'), path = require('path'), fs = require('fs'), os = require('os'), colors = require('colors/safe'), print = require('../../framework/print'), configInterface = require('./apiConfig/configInterface'), { getAllRelatedFiles } = require('./apiConfig/openAPIReader'), { DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR, DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE, DEFAULT_PUBLISH_ERROR_MESSAGE } = require('./constants'), { ROOT_FILE_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS, SCHEMA_FOLDER_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS } = require('./apiConfig/constants');
function fetchIds(apiId, options, callback) {
    const collections = options.apiDetails.collections, schemas = options.apiDetails.schemas, versionName = options.name, releaseNotes = options.releaseNotes, publishPayload = {
        name: versionName,
        releaseNotes: releaseNotes
    };
    if (!_.isEmpty(collections)) {
        publishPayload.collections = collections;
    }
    if (!_.isEmpty(schemas)) {
        publishPayload.schemas = schemas;
    }
    options.publishPayload = publishPayload;
    return callback(null, options);
}
async function fetchPaths(apiId, options, callback) {
    try {
        const versionName = options.name, releaseNotes = options.releaseNotes, publishPayload = {
            name: versionName,
            releaseNotes: releaseNotes,
            branch: options.currentBranchName
        }, fetchCollections = await configInterface.getCollectionFilePaths(apiId), fetchSchemas = await configInterface.getAPIDefinitionFilePaths(apiId), collectionDir = _.get(fetchCollections, 'result.rootDirectory'), collections = _.map(_.get(fetchCollections, 'result.filePaths'), (filePath) => {
            return ({ filePath: path.posix.join('.', collectionDir, filePath) });
        }), schemaDir = _.get(fetchSchemas, 'result.rootDirectory', ''), schemaRootFile = _.get(fetchSchemas, 'result.rootFile', '');
        let schemas = [];
        if (!_.isEmpty(schemaDir)) {
            schemas = [{ directoryPath: schemaDir }];
        }
        if (!_.isEmpty(schemaRootFile)) {
            schemas = [{ filePath: schemaRootFile }];
        }
        if (!_.isEmpty(collections)) {
            publishPayload.collections = collections;
        }
        if (!_.isEmpty(schemas)) {
            publishPayload.schemas = schemas;
        }
        options.publishPayload = publishPayload;
        return callback(null, options);
    }
    catch (err) {
        const errorCode = _.get(err, 'error.name', ''), errorMessage = _.get(err, 'error.message', DEFAULT_PUBLISH_ERROR_MESSAGE);
        if (errorCode === 'configFileNotFound' || errorCode === 'configNotReadable') {
            return callback(DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR, null);
        }
        return callback(errorMessage, null);
    }
}
async function fetchContents(apiId, options, callback) {
    try {
        const versionName = options.name, releaseNotes = options.releaseNotes, publishPayload = {
            name: versionName,
            releaseNotes: releaseNotes,
            branch: options.currentBranchName
        }, fetchCollections = await configInterface.getCollectionFilePaths(apiId), fetchSchemas = await configInterface.getAPIDefinitionFilePaths(apiId), collectionDir = _.get(fetchCollections, 'result.rootDirectory'), collectionPaths = _.map(_.get(fetchCollections, 'result.filePaths'), (filePath) => {
            return path.posix.join('.', collectionDir, filePath);
        }), schemaDir = _.get(fetchSchemas, 'result.rootDirectory'), schemasPaths = _.map(_.get(fetchSchemas, 'result.filePaths'), (filePath) => {
            return {
                relativePath: filePath,
                absolutePath: path.posix.join('.', schemaDir, filePath)
            };
        }), fetchConfigVersion = await configInterface.getConfigCacheVersion(apiId), configVersion = _.get(fetchConfigVersion, 'result', ''), isRootFileBasedIntegration = _.includes(ROOT_FILE_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS, configVersion), schemaRootFile = _.get(fetchSchemas, 'result.rootFile', ''), fetchAPIDefinitionMeta = await configInterface.getAPIDefinitionMeta(apiId), rootFiles = _.get(fetchAPIDefinitionMeta, 'result.rootFiles', []), type = _.get(fetchAPIDefinitionMeta, 'result.type', '');
        let collections = [], schemaFiles = [], schemas = {};
        if (!_.isEmpty(collectionPaths)) {
            _.forEach(collectionPaths, (collectionPath) => {
                let collectionContent = '';
                try {
                    collectionContent = fs.readFileSync(collectionPath, 'utf8', 'r');
                }
                catch (err) {
                    if (err.code && err.code === 'ENOENT') {
                        const error = {
                            error: {
                                message: `Could not find the collection file at ${collectionPath}. Your local ` +
                                    'repository or your API in Postman is not in sync with remote.'
                            }
                        };
                        throw error;
                    }
                    throw err;
                }
                collections.push({
                    filePath: collectionPath,
                    contents: collectionContent
                });
            });
            publishPayload.collections = collections;
        }
        if (isRootFileBasedIntegration && _.includes(['openapi:3', 'openapi:2', 'openapi:1'], type)) {
            const schemaFilesToPublish = await getAllRelatedFiles(type, schemaRootFile), formattedSchemaFilesToPublish = [];
            let isConfigOutOfSync = false, sortedFormattedSchemaFilesToPublish = [], sortedSchemasPaths = [];
            for (const schemaFile of schemaFilesToPublish) {
                formattedSchemaFilesToPublish.push({
                    relativePath: schemaFile.path,
                    absolutePath: schemaFile.path
                });
                schemaFiles.push({
                    filePath: schemaFile.path,
                    contents: schemaFile.content
                });
            }
            sortedFormattedSchemaFilesToPublish = _.sortBy(formattedSchemaFilesToPublish, 'relativePath');
            sortedSchemasPaths = _.sortBy(schemasPaths, 'relativePath');
            isConfigOutOfSync = !_.isEqual(sortedFormattedSchemaFilesToPublish, sortedSchemasPaths);
            if (isConfigOutOfSync) {
                print(`${colors.yellow('\n\nWARNING:\nPublished files may differ from those in the Postman app.')}`);
                print(`${colors.yellow(`${DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE}\n`)}`);
            }
        }
        else if (!_.isEmpty(schemasPaths)) {
            _.forEach(schemasPaths, (schemaPath) => {
                const absoluteSchemaPath = schemaPath.absolutePath, relativeSchemaPath = schemaPath.relativePath;
                let schemaContent = '';
                try {
                    schemaContent = fs.readFileSync(absoluteSchemaPath, 'utf8', 'r');
                }
                catch (err) {
                    if (err.code && err.code === 'ENOENT') {
                        const errObj = {
                            error: {
                                message: `Could not find the schema file at ${absoluteSchemaPath}. Your local ` +
                                    'repository or your API in Postman is not in sync with remote.'
                            }
                        };
                        throw errObj;
                    }
                    throw err;
                }
                schemaFiles.push({
                    filePath: relativeSchemaPath,
                    contents: schemaContent
                });
            });
        }
        if (!_.isEmpty(schemaFiles)) {
            schemas = {
                rootFiles: rootFiles,
                type: type,
                Files: schemaFiles
            };
            publishPayload.schemas = schemas;
        }
        options.publishPayload = publishPayload;
        return callback(null, options);
    }
    catch (err) {
        const errorCode = _.get(err, 'error.name', ''), errorMessage = _.get(err, 'error.message', DEFAULT_PUBLISH_ERROR_MESSAGE);
        if (errorCode === 'configFileNotFound' || errorCode === 'configNotReadable') {
            return callback(DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR, null);
        }
        return callback(errorMessage, null);
    }
}
function validateIds(apiId, options, callback) {
    try {
        const versionName = options.name, releaseNotes = options.releaseNotes, publishPayload = {
            name: versionName,
            releaseNotes: releaseNotes
        }, collectionsFromCommand = options.collections, apiDefinitionFromCommand = options.apiDefinition, allCollections = _.map(options.apiDetails.collections, ('id')), apiDefinitionId = _.map(options.apiDetails.schemas, ('id'))[0];
        if (!_.isEmpty(collectionsFromCommand)) {
            _.forEach(collectionsFromCommand, (collectionFromCommand) => {
                if (!_.includes(allCollections, collectionFromCommand)) {
                    const error = {
                        message: `Collection ${collectionFromCommand} is not part of API ${apiId}`
                    };
                    throw error;
                }
            });
            const collections = _.map(collectionsFromCommand, (collectionFromCommand) => {
                return { id: collectionFromCommand };
            });
            publishPayload.collections = collections;
        }
        if (!_.isEmpty(apiDefinitionFromCommand)) {
            if (!_.isEqual(apiDefinitionId, apiDefinitionFromCommand)) {
                const error = {
                    message: `API Definition ${apiDefinitionFromCommand} is not part of API ${apiId}`
                };
                throw error;
            }
            const schemas = [{ id: apiDefinitionFromCommand }];
            publishPayload.schemas = schemas;
        }
        options.publishPayload = publishPayload;
        return callback(null, options);
    }
    catch (err) {
        const errorMessage = _.get(err, 'message', DEFAULT_PUBLISH_ERROR_MESSAGE);
        return callback(errorMessage, null);
    }
}
async function validatePaths(apiId, options, callback) {
    try {
        const versionName = options.name, releaseNotes = options.releaseNotes, publishPayload = {
            name: versionName,
            releaseNotes: releaseNotes,
            branch: options.currentBranchName
        }, fetchCollections = await configInterface.getCollectionFilePaths(apiId), fetchSchemas = await configInterface.getAPIDefinitionFilePaths(apiId), fetchConfigVersion = await configInterface.getConfigCacheVersion(apiId), configVersion = _.get(fetchConfigVersion, 'result', ''), isRootFileBasedIntegration = _.includes(ROOT_FILE_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS, configVersion), isFolderBasedIntegration = _.includes(SCHEMA_FOLDER_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS, configVersion), collectionDir = _.get(fetchCollections, 'result.rootDirectory'), collectionsFromConfig = _.map(_.get(fetchCollections, 'result.filePaths'), (filePath) => {
            return path.posix.join('.', collectionDir, filePath);
        }), schemaDir = _.get(fetchSchemas, 'result.rootDirectory'), schemaRootFile = _.get(fetchSchemas, 'result.rootFile', '');
        let collectionsFromCommand = options.collections.map((collPath) => {
            return path.relative(process.cwd(), path.resolve(collPath));
        }), apiDefinitionFromCommand = path.relative(process.cwd(), path.resolve(options.apiDefinition)), schemas;
        if (os.platform() === 'win32') {
            collectionsFromCommand = collectionsFromCommand.map((collPath) => {
                return collPath.replaceAll(path.win32.sep.toString(), path.posix.sep.toString());
            });
            apiDefinitionFromCommand = apiDefinitionFromCommand.replaceAll(path.win32.sep.toString(), path.posix.sep.toString());
        }
        if (!_.isEmpty(collectionsFromCommand)) {
            _.forEach(collectionsFromCommand, (collectionFromCommand) => {
                if (!_.includes(collectionsFromConfig, collectionFromCommand)) {
                    const error = {
                        error: {
                            message: `Collection is not part of API ${apiId} or it doesn't exist ` +
                                `at path ${collectionFromCommand}.` + DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                        }
                    };
                    throw error;
                }
            });
            const collections = _.map(collectionsFromCommand, (collectionFromCommand) => {
                return { filePath: collectionFromCommand };
            });
            publishPayload.collections = collections;
        }
        if (!_.isEmpty(apiDefinitionFromCommand)) {
            if (!fs.existsSync(apiDefinitionFromCommand)) {
                const error = {
                    error: {
                        message: `${apiDefinitionFromCommand} does not exist.` +
                            DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                    }
                };
                throw error;
            }
            else if (isFolderBasedIntegration) {
                if (!fs.lstatSync(apiDefinitionFromCommand).isDirectory()) {
                    const error = {
                        error: {
                            message: `${apiDefinitionFromCommand} is a file. ` +
                                'Please provide the schema directory path.'
                        }
                    };
                    throw error;
                }
                else if (_.isEmpty(schemaDir) || !_.isEqual(schemaDir, apiDefinitionFromCommand)) {
                    const error = {
                        error: {
                            message: `API ${apiId} doesn't use ${apiDefinitionFromCommand} directory.` +
                                DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                        }
                    };
                    throw error;
                }
                schemas = [{ directoryPath: apiDefinitionFromCommand }];
            }
            else if (isRootFileBasedIntegration) {
                if (!fs.lstatSync(apiDefinitionFromCommand).isFile()) {
                    const error = {
                        error: {
                            message: `${apiDefinitionFromCommand} is a folder. ` +
                                'Please provide the schema root file path.'
                        }
                    };
                    throw error;
                }
                else if (_.isEmpty(schemaRootFile) || !_.isEqual(schemaRootFile, apiDefinitionFromCommand)) {
                    const error = {
                        error: {
                            message: `${apiDefinitionFromCommand} is not the schema root file for API ${apiId}.` +
                                DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                        }
                    };
                    throw error;
                }
                schemas = [{ filePath: apiDefinitionFromCommand }];
            }
            else {
                const error = {
                    error: {
                        message: 'API config version is not supported. Try updating CLI to latest version.'
                    }
                };
                throw error;
            }
            publishPayload.schemas = schemas;
        }
        options.publishPayload = publishPayload;
        return callback(null, options);
    }
    catch (err) {
        const errorCode = _.get(err, 'error.name', ''), errorMessage = _.get(err, 'error.message', DEFAULT_PUBLISH_ERROR_MESSAGE);
        if (errorCode === 'configFileNotFound' || errorCode === 'configNotReadable') {
            return callback(DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR, null);
        }
        return callback(errorMessage, null);
    }
}
async function validateContents(apiId, options, callback) {
    try {
        const versionName = options.name, releaseNotes = options.releaseNotes, publishPayload = {
            name: versionName,
            releaseNotes: releaseNotes,
            branch: options.currentBranchName
        }, fetchCollections = await configInterface.getCollectionFilePaths(apiId), fetchSchemas = await configInterface.getAPIDefinitionFilePaths(apiId), fetchConfigVersion = await configInterface.getConfigCacheVersion(apiId), configVersion = _.get(fetchConfigVersion, 'result', ''), isRootFileBasedIntegration = _.includes(ROOT_FILE_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS, configVersion), isFolderBasedIntegration = _.includes(SCHEMA_FOLDER_TYPE_INTEGRATION_API_ENTITY_CONFIG_VERSIONS, configVersion), collectionDir = _.get(fetchCollections, 'result.rootDirectory'), collectionsFromConfig = _.map(_.get(fetchCollections, 'result.filePaths'), (filePath) => {
            return path.posix.join('.', collectionDir, filePath);
        }), schemaDir = _.get(fetchSchemas, 'result.rootDirectory'), schemaRootFile = _.get(fetchSchemas, 'result.rootFile', ''), schemasPaths = _.map(_.get(fetchSchemas, 'result.filePaths'), (filePath) => {
            return {
                relativePath: filePath,
                absolutePath: path.posix.join('.', schemaDir, filePath)
            };
        });
        let collectionsFromCommand = options.collections.map((collPath) => {
            return path.relative(process.cwd(), path.resolve(collPath));
        }), apiDefinitionFromCommand = path.relative(process.cwd(), path.resolve(options.apiDefinition)), collections = [], schemaFiles = [], schemas = {};
        if (os.platform() === 'win32') {
            collectionsFromCommand = collectionsFromCommand.map((collPath) => {
                return collPath.replaceAll(path.win32.sep.toString(), path.posix.sep.toString());
            });
            apiDefinitionFromCommand = apiDefinitionFromCommand.replaceAll(path.win32.sep.toString(), path.posix.sep.toString());
        }
        if (!_.isEmpty(collectionsFromCommand)) {
            _.forEach(collectionsFromCommand, (collectionFromCommand) => {
                if (!_.includes(collectionsFromConfig, collectionFromCommand) ||
                    !fs.existsSync(collectionFromCommand)) {
                    const error = {
                        error: {
                            message: `Collection is not part of API ${apiId} or it doesn't exist ` +
                                `at path ${collectionFromCommand}.` + DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                        }
                    };
                    throw error;
                }
            });
        }
        if (!_.isEmpty(apiDefinitionFromCommand)) {
            if (!fs.existsSync(apiDefinitionFromCommand)) {
                const error = {
                    error: {
                        message: `${apiDefinitionFromCommand} does not exist.` +
                            DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                    }
                };
                throw error;
            }
            else if (isFolderBasedIntegration) {
                if (!fs.lstatSync(apiDefinitionFromCommand).isDirectory()) {
                    const error = {
                        error: {
                            message: `${apiDefinitionFromCommand} is a file. ` +
                                'Please provide the schema directory path.'
                        }
                    };
                    throw error;
                }
                else if (_.isEmpty(schemaDir) || !_.isEqual(schemaDir, apiDefinitionFromCommand)) {
                    const error = {
                        error: {
                            message: `API ${apiId} doesn't use ${apiDefinitionFromCommand} directory.` +
                                DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                        }
                    };
                    throw error;
                }
            }
            else if (isRootFileBasedIntegration) {
                if (!fs.lstatSync(apiDefinitionFromCommand).isFile()) {
                    const error = {
                        error: {
                            message: `${apiDefinitionFromCommand} is a folder. ` +
                                'Please provide the schema root file path.'
                        }
                    };
                    throw error;
                }
                else if (_.isEmpty(schemaRootFile) || !_.isEqual(schemaRootFile, apiDefinitionFromCommand)) {
                    const error = {
                        error: {
                            message: `${apiDefinitionFromCommand} is not the schema root file for API ${apiId}.` +
                                DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                        }
                    };
                    throw error;
                }
            }
            else {
                const error = {
                    error: {
                        message: 'API config version is not supported. Try updating CLI to latest version.'
                    }
                };
                throw error;
            }
        }
        if (!_.isEmpty(collectionsFromCommand)) {
            _.forEach(collectionsFromCommand, (collectionPath) => {
                let collectionContent = '';
                try {
                    collectionContent = fs.readFileSync(collectionPath, 'utf8', 'r');
                }
                catch (err) {
                    if (err.code && err.code === 'ENOENT') {
                        const error = {
                            error: {
                                message: `Could not find the collection file at ${collectionPath}.` +
                                    DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                            }
                        };
                        throw error;
                    }
                    throw err;
                }
                collections.push({
                    filePath: collectionPath,
                    contents: collectionContent
                });
            });
            publishPayload.collections = collections;
        }
        if (!_.isEmpty(apiDefinitionFromCommand)) {
            const fetchAPIDefinitionMeta = await configInterface.getAPIDefinitionMeta(apiId), rootFiles = _.get(fetchAPIDefinitionMeta, 'result.rootFiles', []), type = _.get(fetchAPIDefinitionMeta, 'result.type', '');
            if (isRootFileBasedIntegration && _.includes(['openapi:3', 'openapi:2', 'openapi:1'], type)) {
                const schemaFilesToPublish = await getAllRelatedFiles(type, schemaRootFile), formattedSchemaFilesToPublish = [];
                let isConfigOutOfSync = false, sortedFormattedSchemaFilesToPublish = [], sortedSchemasPaths = [];
                for (const schemaFile of schemaFilesToPublish) {
                    formattedSchemaFilesToPublish.push({
                        relativePath: schemaFile.path,
                        absolutePath: schemaFile.path
                    });
                    schemaFiles.push({
                        filePath: schemaFile.path,
                        contents: schemaFile.content
                    });
                }
                sortedFormattedSchemaFilesToPublish = _.sortBy(formattedSchemaFilesToPublish, 'relativePath');
                sortedSchemasPaths = _.sortBy(schemasPaths, 'relativePath');
                isConfigOutOfSync = !_.isEqual(sortedFormattedSchemaFilesToPublish, sortedSchemasPaths);
                if (isConfigOutOfSync) {
                    print(`${colors.yellow('\n\nWARNING:\nCould not find some files from config.')}`);
                    print(`${colors.yellow(`${DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE}\n`)}`);
                }
            }
            else {
                _.forEach(schemasPaths, (schemaPath) => {
                    const absoluteSchemaPath = schemaPath.absolutePath, relativeSchemaPath = schemaPath.relativePath;
                    let schemaContent = '';
                    try {
                        schemaContent = fs.readFileSync(absoluteSchemaPath, 'utf8', 'r');
                    }
                    catch (err) {
                        if (err.code && err.code === 'ENOENT') {
                            const errObj = {
                                error: {
                                    message: `Could not find the schema file at ${absoluteSchemaPath}.` +
                                        DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE
                                }
                            };
                            throw errObj;
                        }
                        throw err;
                    }
                    schemaFiles.push({
                        filePath: relativeSchemaPath,
                        contents: schemaContent
                    });
                });
            }
            schemas = {
                rootFiles: rootFiles,
                type: type,
                Files: schemaFiles
            };
            publishPayload.schemas = schemas;
        }
        options.publishPayload = publishPayload;
        return callback(null, options);
    }
    catch (err) {
        const errorCode = _.get(err, 'error.name', ''), errorMessage = _.get(err, 'error.message', DEFAULT_PUBLISH_ERROR_MESSAGE);
        if (errorCode === 'configFileNotFound' || errorCode === 'configNotReadable') {
            return callback(DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR, null);
        }
        return callback(errorMessage, null);
    }
}
module.exports = {
    fetchIds,
    fetchPaths,
    fetchContents,
    validateIds,
    validatePaths,
    validateContents
};
