// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/constants.js
const util = require('../../util'), REPORT_API_URL = `${util.POSTMAN_GATEWAY_BASE_URL()}/governance/ci/builds`, LF = '\n', POSTMAN_CONFIG_FILE_NAME = '.postman/api', LEARNING_CENTER_LINK = 'https://go.pstmn.io/v10-git-local-docs', SCHEMA_SPECTRAL_TYPES = {
    openapi3: 'OpenAPI 3.0',
    openapi3_1: 'OpenAPI 3.1',
    openapi2: 'OpenAPI 2.0',
    swagger2: 'OpenAPI 2.0',
    asyncapi2: 'AsyncAPI 2',
    asyncapi2_0: 'AsyncAPI 2.0',
    asyncapi2_1: 'AsyncAPI 2.1',
    asyncapi2_2: 'AsyncAPI 2.2',
    asyncapi2_3: 'AsyncAPI 2.3',
    asyncapi2_4: 'AsyncAPI 2.4',
    asyncapi2_5: 'AsyncAPI 2.5',
    asyncapi2_6: 'AsyncAPI 2.6'
}, SCHEMA_TYPES = {
    openapi3: 'OpenAPI 3.0',
    openapi3_1: 'OpenAPI 3.1',
    openapi2: 'OpenAPI 2.0',
    swagger2: 'OpenAPI 2.0',
    asyncapi2: 'AsyncAPI 2',
    asyncapi2_0: 'AsyncAPI 2',
    asyncapi2_1: 'AsyncAPI 2',
    asyncapi2_2: 'AsyncAPI 2',
    asyncapi2_3: 'AsyncAPI 2',
    asyncapi2_4: 'AsyncAPI 2',
    asyncapi2_5: 'AsyncAPI 2',
    asyncapi2_6: 'AsyncAPI 2'
}, SCHEMA_VERSIONS = {
    '3_0': 'openapi3',
    '3_1': 'openapi3_1',
    '2_0': 'openapi2'
}, ASYNC_API_VERSION_TO_TYPE_MAP = {
    '2.0.0': 'asyncapi2_0',
    '2.1.0': 'asyncapi2_1',
    '2.2.0': 'asyncapi2_2',
    '2.3.0': 'asyncapi2_3',
    '2.4.0': 'asyncapi2_4',
    '2.5.0': 'asyncapi2_5',
    '2.6.0': 'asyncapi2_6'
}, SEVERITY_COLOR_MAP = {
    ERROR: 'red',
    WARN: 'yellow',
    INFO: 'cyan',
    HINT: 'white'
}, COMMAND_MODE = {
    CONFIG: 0,
    LOCAL_FILE: 1,
    API_ID: 2
}, EXIT_CODE = {
    OK: 0,
    ERROR: 1
}, SEVERITY_FAIL_ORDER = {
    ERROR: 4,
    WARN: 3,
    INFO: 2,
    HINT: 1
}, SPECTRAL_VERSION_6 = 'SPECTRAL 6.0', NO_AUTHORIZATION_DATA = 'No authorization data found. Please use the `postman login` command.', DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR = 'Failed to fetch contents to publish ~ ' +
    'Config file is missing or cannot be read.' +
    '\nPlease ensure .postman folder is present. ' +
    'Your local repository or your API in Postman is not in sync with remote.', DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE = '\nPostman config might be in an inconsistent state. ' +
    'Please keep your local repository and your API in Postman in sync with remote.', DEFAULT_PUBLISH_ERROR_MESSAGE = 'Something went wrong while fetching contents to publish';
module.exports = {
    REPORT_API_URL,
    LEARNING_CENTER_LINK,
    LF,
    POSTMAN_CONFIG_FILE_NAME,
    SCHEMA_SPECTRAL_TYPES,
    SCHEMA_TYPES,
    SCHEMA_VERSIONS,
    SEVERITY_COLOR_MAP,
    COMMAND_MODE,
    EXIT_CODE,
    SEVERITY_FAIL_ORDER,
    SPECTRAL_VERSION_6,
    NO_AUTHORIZATION_DATA,
    DEFAULT_PUBLISH_MISSING_CONFIG_FILE_ERROR,
    DEFAULT_PUBLISH_INCONSISTENT_RESOLUTION_MESSAGE,
    DEFAULT_PUBLISH_ERROR_MESSAGE,
    ASYNC_API_VERSION_TO_TYPE_MAP
};
