// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/index.js
require('../../../lib/logger').init();
const path = require('path');
const waterfall = require('async/waterfall');
const parallel = require('async/parallel');
const mapSeries = require('async/mapSeries');
const _ = require('lodash');
const Spinner = require('cli-spinner').Spinner;
const colors = require('colors/safe');
const request = require('postman-request');
const GovernanceSandbox = require('@postman/governance-sandbox');
const { RuleUtils } = require('@postman/api-governance-utils');
const APIIdFetcher = require('./fetchers/fetchViaApiId');
const ConfigFetcher = require('./fetchers/fetchViaConfig');
const FileFetcher = require('./fetchers/fetchViaFile');
const tableFormatter = require('./formatters/tableFormatter');
const print = require('../../framework/print');
const util = require('../util');
const libUtil = require('../../util');
const getUserAgent = require('../../../lib/util').userAgent;
const { SCHEMA_SPECTRAL_TYPES, SPECTRAL_VERSION_6, COMMAND_MODE, EXIT_CODE, LF, SEVERITY_FAIL_ORDER, DEFAULT_PUBLISH_ERROR_MESSAGE, NO_AUTHORIZATION_DATA } = require('./constants');
const Reporting = require('./reporting');
const PublishPayload = require('./apiPublishContentFetchers');
let fileFetchingMode;
function getCommandMode(args) {
    if (!args) {
        return COMMAND_MODE.CONFIG;
    }
    let fileExtension = path.extname(args).replace('.', '');
    return ['json', 'yaml', 'yml'].includes(fileExtension) ? COMMAND_MODE.LOCAL_FILE : COMMAND_MODE.API_ID;
}
function fetchTaskStatus(postmanApiKey, accessToken, checkTaskStatusURI) {
    return new Promise((resolve) => {
        const headers = {
            'x-pstmn-req-service': 'api-dev-platform',
            Accept: 'application/vnd.api.v10+json',
            'x-postman-service': 'postman-api'
        };
        if (postmanApiKey) {
            headers['x-api-key'] = postmanApiKey;
        }
        else {
            headers['x-access-token'] = accessToken;
        }
        request.get(checkTaskStatusURI, {
            headers: headers,
            resolveWithFullResponse: true
        }, (err, response, body) => {
            if (err) {
                return resolve({ error: err });
            }
            if (response.statusCode === 403) {
                return resolve({
                    error: `Status code:${response.statusCode} Forbidden. ` +
                        'Please use `postman login` command'
                });
            }
            let taskDetails;
            try {
                taskDetails = JSON.parse(body);
            }
            catch (err) {
                logger.error(err);
                return resolve({ error: 'Something went wrong while publishing your API. Please try again...' });
            }
            if (response.statusCode < 200 || response.statusCode >= 300) {
                const errorMessage = taskDetails.detail || taskDetails.title;
                return resolve({ error: `Status code:${response.statusCode} ${errorMessage}.` });
            }
            return resolve(taskDetails);
        });
    });
}
function fetchRulesetServiceUri(commandMode, args) {
    const defaultURI = `${libUtil.POSTMAN_GATEWAY_BASE_URL()}/security/configure/checks?` +
        'serviceSlug=falcon_service&include=customFunctionsBundle&customFunctionsBundle.runtime=node';
    if (commandMode === COMMAND_MODE.API_ID) {
        return defaultURI + `&elementType=api&elementId=${args}`;
    }
    return defaultURI;
}
module.exports = {
    validate: (args, command, callback) => {
        let spinner = new Spinner();
        spinner.setSpinnerString(19);
        spinner.startSpinner = function (title) {
            this.setSpinnerTitle(title);
            this.start();
        };
        spinner.stopSpinner = function () {
            this.stop(true);
        };
        const commandMode = getCommandMode(args);
        return waterfall([
            (next) => {
                util.addApiKeyToOptions(command, {}, next);
            },
            (option, next) => {
                option.integrationId = command.integrationId;
                next(null, option);
            },
            (options, next) => {
                parallel([
                    (parallelNext) => {
                        const rulesetServiceUri = fetchRulesetServiceUri(commandMode, args), headers = {
                            'User-Agent': getUserAgent,
                            'x-pstmn-req-service': 'ruleset',
                            'x-app-version': '10.15.2'
                        };
                        if (options.postmanApiKey) {
                            headers['x-api-key'] = options.postmanApiKey;
                        }
                        else {
                            headers['x-access-token'] = options.accessToken;
                        }
                        request.get(rulesetServiceUri, {
                            headers
                        }, (err, response, body) => {
                            if (err) {
                                logger.error(`Error while fetching the ruleset - ${JSON.stringify(err)}`);
                                return parallelNext(err);
                            }
                            if (response.statusCode < 200 || response.statusCode >= 300) {
                                logger.error(`Error in response while fetching the ruleset -
                                        ${JSON.stringify(response)}`);
                                return parallelNext(new Error(`Status code:${response.statusCode} Something went wrong when fetching the ruleset. Try again.`));
                            }
                            let rulesets;
                            try {
                                rulesets = JSON.parse(body);
                            }
                            catch (err) {
                                err.message = `Error encountered when parsing the ruleset: ${err.message}`;
                                return parallelNext(err);
                            }
                            return parallelNext(null, rulesets);
                        });
                    },
                    (parallelNext) => {
                        switch (commandMode) {
                            case COMMAND_MODE.LOCAL_FILE:
                                spinner.startSpinner('Fetching the API from the given file path');
                                fileFetchingMode = 'FILE';
                                FileFetcher.fetchViaFile(args, parallelNext);
                                break;
                            case COMMAND_MODE.API_ID:
                                ConfigFetcher.checkIfLocalGitLinkedAPI(args)
                                    .then((isGitLinked) => {
                                    if (!isGitLinked) {
                                        spinner.startSpinner(`Fetching the API with ID ${args}`);
                                        fileFetchingMode = 'API_ID';
                                        APIIdFetcher.fetchViaApiId(args, options, parallelNext);
                                    }
                                    else {
                                        print(`${colors.yellow('Since the provided API ID matches with the git linked API ID, the local schema/definition file will be used for linting.')}${LF}`);
                                        spinner.startSpinner('Fetching the api from the postman config');
                                        fileFetchingMode = 'CONFIG';
                                        ConfigFetcher.fetchViaConfig(args, parallelNext);
                                    }
                                })
                                    .catch((err) => {
                                    parallelNext(err);
                                });
                                break;
                            case COMMAND_MODE.CONFIG:
                            default:
                                spinner.startSpinner('Fetching the API from Postman config');
                                fileFetchingMode = 'CONFIG';
                                ConfigFetcher.fetchViaConfig(args, parallelNext);
                                break;
                        }
                    }
                ], (err, results) => {
                    spinner.stopSpinner();
                    if (err) {
                        logger.error(`Error while fetching the rule and content - ${JSON.stringify(err)}`);
                        return next(err, results[1]);
                    }
                    if (['API_ID', 'FILE'].includes(fileFetchingMode)) {
                        const resultObject = {
                            apiObjects: [results[1]],
                            countOfAPIs: 1,
                            hasErrorOccurred: false,
                            errorMessages: []
                        };
                        return next(null, results[0], resultObject, options);
                    }
                    return next(null, results[0], results[1], options);
                });
            },
            (rulesets, { apiObjects, countOfAPIs, hasErrorOccurred, errorMessages }, options, next) => {
                return mapSeries(apiObjects, (api, mapCallback) => {
                    const { content, type } = api;
                    let language = api.language;
                    if (language === 'yml') {
                        language = 'yaml';
                    }
                    let spectralRules = RuleUtils.getSpectralRules(rulesets, {
                        isEnabled: true,
                        engine: SPECTRAL_VERSION_6,
                        entityConfig: { type: SCHEMA_SPECTRAL_TYPES[type] || SCHEMA_SPECTRAL_TYPES.openapi3 }
                    });
                    if (!spectralRules || !spectralRules.rules || !spectralRules.spectralConfig) {
                        print(`${colors.yellow('No rules found for linting.')}${LF}`);
                        return callback(null, { code: EXIT_CODE.OK });
                    }
                    return waterfall([
                        async () => {
                            const customFunctionsBundle = _.get(rulesets, 'customFunctionsBundle.value', '');
                            const validateOptions = JSON.stringify({
                                file: {
                                    language: language && language.toLowerCase(),
                                    contents: content
                                },
                                ruleset: spectralRules.spectralConfig
                            });
                            const code = `
                                        (async () => {
                                            let global = {};

                                            // custom functions bundle
                                            ${customFunctionsBundle};

                                            let validateOptions = ${validateOptions};

                                            validateOptions['customFunctions'] = global;

                                            try {
                                                const warnings = await apiValidator.validate(validateOptions);

                                                if (warnings) {
                                                  apiValidator.result({ type: 'warnings', warnings: warnings });
                                                }
                                              }
                                              catch(error) {
                                                apiValidator.result({ type: 'error', error: error });
                                              }
                                        })()
                                        `;
                            const warnings = await new Promise((resolve, reject) => {
                                try {
                                    GovernanceSandbox.createContext({}, (err, ctx) => {
                                        if (err) {
                                            logger.error(`Error while creating context for governance sandbox -
                                                ${JSON.stringify(err)}`);
                                            return reject(err);
                                        }
                                        ctx.on('apiValidator.result', (message) => {
                                            switch (message.type) {
                                                case 'error':
                                                    logger.error(`Error while validating schema -
                                                        ${JSON.stringify(message)}`);
                                                    return reject(message.error);
                                                case 'warnings':
                                                    return resolve(message.warnings);
                                                default:
                                                    logger.error(`Unhandled Error occurred while validating schema -
                                                        ${JSON.stringify(message)}`);
                                                    return reject(new Error(`Sandbox: unhandled result type:${message.type} received`));
                                            }
                                        });
                                        ctx.on('error', (err) => {
                                            err && logger.error(`Error thrown by sandbox - ${JSON.stringify(err)}`);
                                            return reject(err);
                                        });
                                        ctx.execute(code, (err) => {
                                            if (err) {
                                                logger.error(`Error thrown by sandbox while executing -
                                                    ${JSON.stringify(err)}`);
                                                return reject(err);
                                            }
                                            return false;
                                        });
                                    });
                                }
                                catch (err) {
                                    reject(err);
                                }
                            });
                            return warnings;
                        },
                        (results, nestedNext) => {
                            let categorizedResults = {}, validResults = [], isFailSeverityFound = false, { failSeverity } = libUtil.commanderToObject(command);
                            results.forEach((result) => {
                                const ruleName = result.code, rule = spectralRules.rules[ruleName], tag = _.get(spectralRules.rules[ruleName], 'config.tags[0]', 'others');
                                if (!rule) {
                                    return;
                                }
                                result.tag = tag;
                                if (!categorizedResults[tag]) {
                                    categorizedResults[tag] = [];
                                }
                                result.severity = rule.severity;
                                result.possibleFixUrl = _.get(rule, 'mitigation.possibleFixUrl', '');
                                if (failSeverity &&
                                    SEVERITY_FAIL_ORDER[failSeverity] <= SEVERITY_FAIL_ORDER[rule.severity.toUpperCase()]) {
                                    isFailSeverityFound = true;
                                }
                                validResults.push(result);
                                categorizedResults[tag].push(result);
                            });
                            const output = tableFormatter(api, commandMode, categorizedResults);
                            print(output);
                            return nestedNext(null, { api, validResults, isFailSeverityFound });
                        }
                    ], mapCallback);
                }, (err, mapSeriesResults) => {
                    if (err) {
                        logger.error(`Error while fetching results - ${JSON.stringify(err)}`);
                        return next(err);
                    }
                    const isFailSeverityFoundCombined = mapSeriesResults.reduce((accumulator, apiResult) => {
                        return accumulator || apiResult.isFailSeverityFound;
                    }, false);
                    return next(null, mapSeriesResults, isFailSeverityFoundCombined, countOfAPIs, hasErrorOccurred, errorMessages, options);
                });
            },
            (allResults, isFailSeverityFound, countOfAPIs, hasErrorOccurred, errorMessages, options, next) => {
                let output;
                if (commandMode !== COMMAND_MODE.API_ID) {
                    print(`${colors.yellow('\nWARNING:\nShowing linting results for rules in "All Workspaces" group across APIs in your repo. To get results from other workspace groups, add an API ID and consider reconfiguring your Postman CLI integration. To learn more about Postman CLI, visit: https://go.pstmn.io/postman-cli-faq.')}${LF}`);
                }
                if (Reporting.shouldSendReport(commandMode) && countOfAPIs === 1) {
                    const { api, validResults: results } = allResults[0];
                    spinner.startSpinner('Sending report to Postman');
                    return Reporting.sendReport(api, options, results, (err, res) => {
                        spinner.stopSpinner();
                        if (err) {
                            logger.error(`Error while sending the report - ${JSON.stringify(err)}`);
                            output = `${colors.yellow(err)}${LF}`;
                        }
                        else {
                            output = `${colors.green('Report sent.')}${LF}`;
                            let reportUrl = res.reportUrl;
                            if (reportUrl) {
                                output += `You can view your schema warnings along with other build information here ${reportUrl}${LF}`;
                            }
                            output += `${colors.yellow('\nWARNING:\nCurrent repo contains only one API and linting results work. To continue receiving linting results for more than one integration or API, consider reconfiguring your Postman CLI integration. For more information visit: https://go.pstmn.io/postman-cli-faq')}${LF}`;
                        }
                        print(output);
                        return next(null, {
                            code: (isFailSeverityFound || hasErrorOccurred) ?
                                EXIT_CODE.ERROR : EXIT_CODE.OK
                        });
                    });
                }
                if (countOfAPIs > 1) {
                    if (hasErrorOccurred) {
                        _.forEach(errorMessages, (errorMessage) => {
                            print(`${colors.red(`\n${errorMessage}`)}${LF}`);
                        });
                    }
                    if (allResults.length >= 1) {
                        print(`${colors.yellow('\nWARNING:\nShowing linting results for all APIs in your repo. This will not be sent to Postman. To get accurate linting results, add an API ID and consider reconfiguring your Postman CLI integration. For more information visit: https://go.pstmn.io/postman-cli-faq')}${LF}`);
                    }
                }
                const hasError = isFailSeverityFound || hasErrorOccurred;
                hasError && logger.error(`Is error found - ${JSON.stringify({
                    isFailSeverityFound,
                    hasErrorOccurred
                })}`);
                return next(null, { code: hasError ? EXIT_CODE.ERROR : EXIT_CODE.OK });
            }
        ], callback);
    },
    publish: (apiId, command, callback) => {
        let spinner = new Spinner();
        spinner.setSpinnerString(19);
        spinner.startSpinner = function (title) {
            this.setSpinnerTitle(title);
            this.start();
        };
        spinner.stopSpinner = function () {
            this.stop(true);
        };
        const hasUserSpecifiedEntitiesToPublish = !_.isEmpty(_.get(command, 'collections')) ||
            !_.isEmpty(_.get(command, 'apiDefinition'));
        let isGitLinked = false, isIntegrationTypeOnPrem = false;
        return waterfall([
            (next) => {
                util.addApiKeyToOptions(command, {}, next);
            },
            (option, next) => {
                option.name = _.get(command, 'name', '');
                option.releaseNotes = _.get(command, 'releaseNotes', '');
                option.collections = _.get(command, 'collections', []);
                option.apiDefinition = _.get(command, 'apiDefinition', '');
                option.doNotPoll = _.get(command, 'doNotPoll', false);
                next(null, option);
            },
            (options, next) => {
                spinner.startSpinner('Fetching contents to publish');
                const postmanApiKey = options.postmanApiKey;
                if (!postmanApiKey && !options.accessToken) {
                    return next(NO_AUTHORIZATION_DATA, null);
                }
                const getAPIUri = `${libUtil.POSTMAN_GATEWAY_BASE_URL()}/apis/${apiId}?include=schemas,collections,gitInfo`;
                const headers = {
                    'x-pstmn-req-service': 'api-dev-platform',
                    Accept: 'application/vnd.api.v10+json',
                    'x-postman-service': 'postman-api'
                };
                if (postmanApiKey) {
                    headers['x-api-key'] = postmanApiKey;
                }
                else {
                    headers['x-access-token'] = options.accessToken;
                }
                request.get(getAPIUri, {
                    headers: headers,
                    resolveWithFullResponse: true
                }, (err, response, body) => {
                    if (err) {
                        return next(err, null);
                    }
                    let apiDetails;
                    try {
                        apiDetails = JSON.parse(body);
                    }
                    catch (err) {
                        logger.error(err);
                        return next(DEFAULT_PUBLISH_ERROR_MESSAGE, null);
                    }
                    if (response.statusCode === 403 && apiDetails?.error?.name === 'notLoggedInError') {
                        return next(`Status code: ${response.statusCode} Forbidden. ` +
                            'Please use `postman login` command', null);
                    }
                    if (response.statusCode < 200 || response.statusCode >= 300) {
                        const errorMessage = apiDetails.detail || apiDetails.title ||
                            'Something went wrong while fetching data from our servers';
                        return next(`Status code: ${response.statusCode} ${errorMessage}.`, null);
                    }
                    options.apiDetails = apiDetails;
                    isGitLinked = !_.isEmpty(_.get(apiDetails, 'gitInfo'));
                    isIntegrationTypeOnPrem = !_.isEmpty(_.get(apiDetails, 'gitInfo.domain'));
                    if (isGitLinked) {
                        spinner.stopSpinner();
                        try {
                            options.currentBranchName = util.getCurrentBranchName();
                            print(`Branch found: ${options.currentBranchName}\n`);
                        }
                        catch (err) {
                            return next(err, null);
                        }
                    }
                    return next(null, options);
                });
            },
            (options, next) => {
                spinner.startSpinner('Fetching contents to publish');
                if (!hasUserSpecifiedEntitiesToPublish && !isGitLinked && !isIntegrationTypeOnPrem) {
                    return PublishPayload.fetchIds(apiId, options, next);
                }
                if (!hasUserSpecifiedEntitiesToPublish && isGitLinked && !isIntegrationTypeOnPrem) {
                    return PublishPayload.fetchPaths(apiId, options, next);
                }
                if (!hasUserSpecifiedEntitiesToPublish && isGitLinked && isIntegrationTypeOnPrem) {
                    return PublishPayload.fetchContents(apiId, options, next);
                }
                if (hasUserSpecifiedEntitiesToPublish && !isGitLinked && !isIntegrationTypeOnPrem) {
                    return PublishPayload.validateIds(apiId, options, next);
                }
                if (hasUserSpecifiedEntitiesToPublish && isGitLinked && !isIntegrationTypeOnPrem) {
                    return PublishPayload.validatePaths(apiId, options, next);
                }
                if (hasUserSpecifiedEntitiesToPublish && isGitLinked && isIntegrationTypeOnPrem) {
                    return PublishPayload.validateContents(apiId, options, next);
                }
                return next('Failed to determine contents to publish', null);
            },
            (options, next) => {
                spinner.startSpinner('Publishing API to Postman');
                const postmanApiKey = options.postmanApiKey, publishAPIUri = `${libUtil.POSTMAN_GATEWAY_BASE_URL()}/v2/apis/${apiId}/versions`;
                const headers = {
                    'x-pstmn-req-service': 'api-dev-platform',
                    Accept: 'application/vnd.api.v10+json',
                    'x-postman-service': 'postman-api'
                };
                if (postmanApiKey) {
                    headers['x-api-key'] = postmanApiKey;
                }
                else {
                    headers['x-access-token'] = options.accessToken;
                }
                request.post(publishAPIUri, {
                    headers: headers,
                    body: JSON.stringify(options.publishPayload),
                    resolveWithFullResponse: true
                }, (err, response, body) => {
                    if (err) {
                        return next(err, null);
                    }
                    let publishedAPIDetails;
                    try {
                        publishedAPIDetails = JSON.parse(body);
                    }
                    catch (err) {
                        logger.error(err);
                        return next(DEFAULT_PUBLISH_ERROR_MESSAGE, null);
                    }
                    if (response.statusCode === 403 && publishedAPIDetails.title === 'Unauthorized') {
                        return next(`Status code: ${response.statusCode} Forbidden. ` +
                            'Please use `postman login` command', null);
                    }
                    if (response.statusCode < 200 || response.statusCode >= 300) {
                        const errorMessage = publishedAPIDetails.detail || publishedAPIDetails.title ||
                            'Something went wrong while publishing the API';
                        return next(`Status code: ${response.statusCode} ${errorMessage}.`, null);
                    }
                    const parsedRawHeaders = _.chain(response.rawHeaders)
                        .chunk(2)
                        .map(([key, value]) => {
                        return {
                            key: _.toLower(key),
                            value: value
                        };
                    })
                        .keyBy('key')
                        .mapValues('value')
                        .value(), checkTaskStatusURI = _.get(parsedRawHeaders, 'location');
                    options.releaseId = publishedAPIDetails.id;
                    options.checkTaskStatusURI = checkTaskStatusURI;
                    spinner.stopSpinner();
                    print.lf(`\nSuccessfully triggered publish for API: ${apiId}`);
                    return next(null, options);
                });
            },
            async (options) => {
                if (options.doNotPoll) {
                    print.lf('You can view your published version here ' +
                        `https://go.postman.co/apis/${apiId}?version=${options.releaseId}` +
                        '\n\nFor more information visit: https://go.pstmn.io/postman-cli-faq');
                    return { code: EXIT_CODE.OK };
                }
                const postmanApiKey = options.postmanApiKey, checkTaskStatusURI = `${libUtil.POSTMAN_GATEWAY_BASE_URL()}${options.checkTaskStatusURI}`, sleepUntil = (time) => {
                    return new Promise((resolve) => {
                        setTimeout(resolve, time);
                    });
                };
                let shouldStopPolling = false, waitTime = 0;
                do {
                    spinner.startSpinner('Checking status of the published version');
                    await sleepUntil(waitTime * 1000);
                    const response = await fetchTaskStatus(postmanApiKey, options.accessToken, checkTaskStatusURI);
                    if (!_.isEmpty(response.error)) {
                        throw response.error;
                    }
                    if (response.status === 'pending') {
                        waitTime += 5;
                        spinner.stopSpinner();
                        print.lf(`\nPublish in progress...\nChecking again in ${waitTime}sec...`);
                    }
                    if (response.status === 'failed') {
                        shouldStopPolling = true;
                        spinner.stopSpinner();
                        print.lf(`\nFailed to publish your API: ${response.details.error.message}` +
                            '\n\nFor more information visit: https://go.pstmn.io/postman-cli-faq');
                    }
                    if (response.status === 'completed') {
                        shouldStopPolling = true;
                        spinner.stopSpinner();
                        print.lf('\nSuccessfully published your API. You can view your published version ' +
                            `here https://go.postman.co/apis/${apiId}?version=${options.releaseId}` +
                            '\n\nFor more information visit: https://go.pstmn.io/postman-cli-faq');
                    }
                } while (!shouldStopPolling);
                return { code: EXIT_CODE.OK };
            }
        ], callback);
    }
};
