// @postman/pm-bin-linux-x64/bin/lib/postman-entities/api/reporting.js
const _ = require('lodash');
const request = require('postman-request');
const CIInfo = require('../../run-upload/helpers/ci-info');
const util = require('../../../lib/util');
const { REPORT_API_URL, COMMAND_MODE } = require('./constants');
function shouldSendReport(commandMode) {
    if (![COMMAND_MODE.CONFIG, COMMAND_MODE.API_ID].includes(commandMode)) {
        return false;
    }
    const isCI = require('is-ci');
    if (!isCI) {
        return false;
    }
    return true;
}
function sendReport(api, options, results, callback) {
    const env = process.env, integrationId = options && options.integrationId || env && env.INTEGRATION_ID, apiId = _.get(api, 'id');
    if (!apiId) {
        return callback('Cannot send the report as API ID is missing from the config.');
    }
    if (!integrationId) {
        return callback('Couldn\'t send the report. Please set the Integration ID and try again');
    }
    let ciInfo = CIInfo.getCiInfo(env), mapped = results.map((result) => {
        return {
            type: result.tag,
            severity: result.severity,
            message: result.message,
            location: {
                start: {
                    line: result.range.start.line,
                    column: result.range.start.character
                },
                end: {
                    line: result.range.end.line,
                    column: result.range.end.character
                }
            },
            dataPath: result.path,
            possibleFixUrl: result.possibleFixUrl
        };
    }), body = {
        integrationId: integrationId,
        branch: {
            name: ciInfo.branch,
            source: ciInfo.branchSource
        },
        build: {
            id: ciInfo.pipelineMeta.id,
            url: ciInfo.pipelineMeta.url,
            source: ciInfo.pipelineMeta.source
        },
        apis: [{ id: apiId, warnings: mapped }]
    }, headers = {
        'content-type': 'application/json',
        'x-pstmn-req-service': 'ruleset',
        'x-app-version': '10.15.2',
        Accept: 'application/vnd.postman.v2+json',
        'User-Agent': util.userAgent
    };
    if (options.postmanApiKey) {
        headers['x-api-key'] = options.postmanApiKey;
    }
    else if (options.accessToken) {
        headers['x-access-token'] = options.accessToken;
    }
    return request.post(REPORT_API_URL, {
        headers: headers,
        body: JSON.stringify(body)
    }, (err, res, body) => {
        if (err) {
            const message = err && err.message || err.toString();
            return callback(`Couldn't send the report. ${message}`);
        }
        else if (res.statusCode < 200 || res.statusCode >= 300) {
            return callback(`Status Code: [${res.statusCode}] Couldn't send the report.`);
        }
        try {
            return callback(null, JSON.parse(body));
        }
        catch (err) {
            return callback(null);
        }
    });
}
module.exports = {
    shouldSendReport,
    sendReport
};
