// @postman/pm-bin-linux-x64/bin/lib/commands/workspace/push.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const commander_1 = require("commander");
const waterfall_1 = __importDefault(require("async/waterfall"));
const eachSeries_1 = __importDefault(require("async/eachSeries"));
const readline_1 = __importDefault(require("readline"));
const util_1 = __importDefault(require("../../postman-entities/util"));
const console_1 = __importDefault(require("../../framework/console"));
const workspace_utils_1 = require("./workspace-utils");
const v3_collection_1 = require("../../postman-entities/collection/v3-collection");
const SpecificationService_1 = require("../../postman-entities/spec/Service/SpecificationService");
const resources_1 = require("./resources");
const workflows_1 = require("./workflows");
const uuid_1 = require("uuid");
const safe_1 = __importDefault(require("colors/safe"));
const analytics_1 = require("../workspace/analytics");
function matchesId(workspaceId, localId) {
    return workspaceId === localId || workspaceId.endsWith(`-${localId}`);
}
function createDefaultStats() {
    return {
        collectionsUpdated: 0,
        environmentsUpdated: 0,
        specificationsUpdated: 0,
        sdksUploaded: 0,
        collectionsSkipped: 0,
        environmentsSkipped: 0,
        specificationsSkipped: 0,
        collectionsCreated: 0,
        environmentsCreated: 0,
        specificationsCreated: 0,
        collectionErrors: 0,
        environmentErrors: 0,
        specificationErrors: 0,
        collectionCreateErrors: 0,
        environmentCreateErrors: 0,
        specificationCreateErrors: 0,
        sdkErrors: 0
    };
}
async function processJsonCollectionFile(filePath, options) {
    const collection = await (0, workspace_utils_1.readJSONFile)(filePath);
    const name = collection.info?.name || filePath;
    if (options.configSource === 'resources') {
        let cloudId = null;
        try {
            const lookup = await (0, resources_1.getCloudIdForPath)(filePath, 'collections');
            cloudId = lookup.cloudId;
        }
        catch {
        }
        if (cloudId) {
            options.localCollections?.set(cloudId, {
                filePath: filePath,
                data: collection,
                name
            });
        }
        else if (options.yes) {
            console_1.default.log(safe_1.default.yellow(`Collection "${name}": not found on cloud, skipping (--yes)`));
            options.stats.collectionsSkipped++;
        }
        else {
            console_1.default.log(safe_1.default.cyan(`Collection "${name}": no cloud mapping in resources.yaml, will be created`));
            options.collectionsToCreate?.push({
                localId: '',
                data: collection,
                filePath: filePath,
                name
            });
        }
    }
    else {
        const collectionId = collection.info?._postman_id;
        if (collectionId) {
            options.localCollections?.set(collectionId, {
                filePath: filePath,
                data: collection,
                name
            });
        }
    }
}
async function processEnvironmentFile(filePath, options) {
    const environment = await (0, workspace_utils_1.readEnvironmentFile)(filePath);
    const name = environment.name || filePath;
    if (options.configSource === 'resources') {
        let cloudId = null;
        try {
            const lookup = await (0, resources_1.getCloudIdForPath)(filePath, 'environments');
            cloudId = lookup.cloudId;
        }
        catch {
        }
        if (cloudId) {
            options.localEnvironments?.set(cloudId, {
                filePath: filePath,
                data: environment,
                name
            });
        }
        else if (options.yes) {
            console_1.default.log(safe_1.default.yellow(`Environment "${name}": not found on cloud, skipping (--yes)`));
            options.stats.environmentsSkipped++;
        }
        else {
            console_1.default.log(safe_1.default.cyan(`Environment "${name}": no cloud mapping in resources.yaml, will be created`));
            options.environmentsToCreate?.push({
                localId: '',
                data: environment,
                filePath: filePath,
                name
            });
        }
    }
    else {
        const environmentId = environment.id;
        if (environmentId) {
            options.localEnvironments?.set(environmentId, {
                filePath: filePath,
                data: environment,
                name
            });
        }
    }
}
function determineSyncCollections(options) {
    options.collectionsToUpdate = [];
    options.v3CollectionsToUpdate = [];
    if (!options.collectionsToCreate) {
        options.collectionsToCreate = [];
    }
    if (!options.v3CollectionsToCreate) {
        options.v3CollectionsToCreate = [];
    }
    const wsCollectionIds = options.workspaceCollections ?? [];
    options.localCollections?.forEach((localCollection, localCollectionId) => {
        const matchingWorkspaceId = wsCollectionIds.find((wsId) => {
            return matchesId(wsId, localCollectionId);
        });
        if (matchingWorkspaceId) {
            options.collectionsToUpdate?.push({
                localId: localCollectionId,
                workspaceId: matchingWorkspaceId
            });
        }
        else if (options.yes) {
            console_1.default.log(safe_1.default.yellow(`Collection "${localCollection.name}": not found on cloud, skipping (--yes)`));
            options.stats.collectionsSkipped++;
        }
        else {
            console_1.default.log(safe_1.default.cyan(`Collection "${localCollection.name}": not found in workspace, will be created`));
            options.collectionsToCreate?.push({
                localId: localCollectionId,
                data: localCollection.data,
                filePath: localCollection.filePath,
                name: localCollection.name
            });
        }
    });
    for (const v3Entry of (options.v3Collections || [])) {
        if (v3Entry.cloudId) {
            options.v3CollectionsToUpdate?.push({
                entry: v3Entry,
                cloudCollectionId: v3Entry.cloudId
            });
        }
        else if (options.yes) {
            console_1.default.log(safe_1.default.yellow(`Collection "${v3Entry.name}": not found on cloud, skipping (--yes)`));
            options.stats.collectionsSkipped++;
        }
        else {
            console_1.default.log(safe_1.default.cyan(`V3 collection "${v3Entry.name}": no cloud ID, will be created`));
            options.v3CollectionsToCreate?.push({ entry: v3Entry });
        }
    }
}
function determineSyncEnvironments(options) {
    options.environmentsToUpdate = [];
    if (!options.environmentsToCreate) {
        options.environmentsToCreate = [];
    }
    const wsEnvironmentIds = options.workspaceEnvironments ?? [];
    options.localEnvironments?.forEach((localEnvironment, localEnvironmentId) => {
        const matchingWorkspaceId = wsEnvironmentIds.find((wsId) => {
            return matchesId(wsId, localEnvironmentId);
        });
        if (matchingWorkspaceId) {
            options.environmentsToUpdate?.push({
                localId: localEnvironmentId,
                workspaceId: matchingWorkspaceId
            });
        }
        else if (options.yes) {
            console_1.default.log(safe_1.default.yellow(`Environment "${localEnvironment.name}": not found on cloud, skipping (--yes)`));
            options.stats.environmentsSkipped++;
        }
        else {
            console_1.default.log(safe_1.default.cyan(`Environment "${localEnvironment.name}": not found in workspace, will be created`));
            options.environmentsToCreate?.push({
                localId: localEnvironmentId,
                data: localEnvironment.data,
                filePath: localEnvironment.filePath,
                name: localEnvironment.name
            });
        }
    });
}
async function determineSyncSpecifications(options) {
    options.specificationsToUpdate = [];
    if (!options.specificationsToCreate) {
        options.specificationsToCreate = [];
    }
    for (const [normalizedPath, localSpec] of (options.localSpecifications ?? new Map())) {
        let specCloudId = null;
        try {
            const lookup = await (0, resources_1.getCloudIdForPath)(localSpec.filePath, 'specs');
            specCloudId = lookup.cloudId;
        }
        catch {
        }
        if (specCloudId) {
            options.specificationsToUpdate?.push({
                localRootPath: normalizedPath,
                specId: specCloudId,
                filePath: normalizedPath
            });
            continue;
        }
        const match = options.cloudSpecificationMap?.get(normalizedPath);
        if (match) {
            options.specificationsToUpdate?.push({
                localRootPath: normalizedPath,
                specId: match.specId,
                filePath: match.rootPath
            });
        }
        else if (options.yes) {
            console_1.default.log(safe_1.default.yellow(`Specification "${localSpec.name}": not found on cloud, skipping (--yes)`));
            options.stats.specificationsSkipped++;
        }
        else {
            console_1.default.log(safe_1.default.cyan(`Specification "${localSpec.name}": no cloud mapping, will be created`));
            options.specificationsToCreate?.push({
                normalizedPath,
                localSpec
            });
        }
    }
}
async function determineSyncSpecCollectionLinks(options) {
    options.specCollectionLinksToPush = [];
    try {
        const workflowsConfig = await (0, workflows_1.readWorkflowsConfig)();
        if (!workflowsConfig || !workflowsConfig.workflows?.syncSpecToCollection?.length) {
            return;
        }
        const resolvedLinks = await (0, workflows_1.resolveWorkflowLinks)(workflowsConfig, undefined, options.localSpecifications);
        options.specCollectionLinksToPush = resolvedLinks;
    }
    catch (e) {
        console_1.default.error(safe_1.default.red(`Failed to resolve workflow links: ${e.message}`));
    }
}
const pushCommand = function (workspace) {
    workspace
        .command('push')
        .description('Push local collection, environment, and specification changes to Postman workspace.')
        .option('--collections-dir <path>', 'Path to the collections directory', 'postman/collections')
        .option('--environments-dir <path>', 'Path to the environments directory', 'postman/environments')
        .option('--no-prepare', 'Skip the prepare step before pushing')
        .option('-y, --yes', 'Skip all confirmation prompts')
        .addOption(new commander_1.Option('--report-events', 'Report run events').hideHelp())
        .action((command) => {
        let options, exitCode = 0;
        (0, waterfall_1.default)([
            (next) => {
                (0, workspace_utils_1.readPostmanConfig)()
                    .then(async (config) => {
                    options = { ...command, workspaceId: config.workspaceId, stats: createDefaultStats() };
                    const resourcesConfig = await (0, resources_1.readResourcesConfig)();
                    options.configSource = resourcesConfig?.workspace?.id ? 'resources' : 'configJson';
                    options.collectionFiles = await (0, workspace_utils_1.discoverEntities)('collections', process.cwd());
                    options.environmentFiles = await (0, workspace_utils_1.discoverEntities)('environments', process.cwd());
                    options.specificationFiles = await (0, workspace_utils_1.discoverEntities)('specs', process.cwd());
                    options.discoveredSDKs = await (0, workspace_utils_1.discoverSDKs)(process.cwd());
                    next(null, options);
                })
                    .catch((err) => {
                    next(err);
                });
            },
            (opts, next) => {
                util_1.default.addApiKeyToOptions(opts, {}, (err, updatedOpts) => {
                    if (err) {
                        return next(err);
                    }
                    options = updatedOpts ?? options;
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                if (!options.workspaceId) {
                    return next(new Error('No workspace ID found. Provide workspace.id in .postman/resources.yaml or .postman/config.json.'));
                }
                console_1.default.log(safe_1.default.grey(`Using workspace: ${options.workspaceId}\n`));
                return next(null, options);
            },
            (opts, next) => {
                options = opts;
                if (options.workspaceCollections) {
                    return next(null, options);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                (0, workspace_utils_1.getWorkspaceCollections)(workspaceId, options, (err, collections) => {
                    if (err) {
                        return next(err);
                    }
                    options.workspaceCollections = collections;
                    console_1.default.log(safe_1.default.grey(`Found ${(collections ?? []).length} collection(s) in workspace`));
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                if (options.workspaceEnvironments) {
                    return next(null, options);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                (0, workspace_utils_1.getWorkspaceEnvironments)(workspaceId, options, (err, environments) => {
                    if (err) {
                        return next(err);
                    }
                    options.workspaceEnvironments = environments;
                    console_1.default.log(safe_1.default.grey(`Found ${(environments ?? []).length} environment(s) in workspace`));
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                if (options.workspaceSpecifications) {
                    return next(null, options);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                (0, workspace_utils_1.getWorkspaceSpecifications)(workspaceId, options, (err, specifications) => {
                    if (err) {
                        return next(err);
                    }
                    options.workspaceSpecifications = specifications;
                    console_1.default.log(safe_1.default.grey(`Found ${(specifications ?? []).length} specification(s) in workspace`));
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                console_1.default.log(safe_1.default.grey(`Found ${options.discoveredSDKs?.length || 0} SDK(s) in workspace`));
                next(null, options);
            },
            (opts, next) => {
                options = opts;
                if (!command.reportEvents) {
                    return next(null, options);
                }
                if (!options.workspaceId) {
                    return next(null, options);
                }
                (0, analytics_1.sendWorkspacePushEvents)({ workspaceId: options.workspaceId })
                    .then(() => { return next(null, options); })
                    .catch((e) => { return next(e); });
            },
            (opts, next) => {
                options = opts;
                Promise.resolve()
                    .then(async () => {
                    const jsonCollectionFiles = [];
                    for (const filePath of (options.collectionFiles || [])) {
                        if (!(await (0, v3_collection_1.isV3CollectionDirectory)(filePath))) {
                            jsonCollectionFiles.push(filePath);
                        }
                    }
                    const jsonEnvironmentFiles = (options.environmentFiles || []).filter(workspace_utils_1.isJsonFile);
                    const shouldSkipPrepare = options.prepare === false, collectionStatuses = await (0, workspace_utils_1.validateCollections)(jsonCollectionFiles, options, options.workspaceCollections || []), environmentStatuses = await (0, workspace_utils_1.validateEnvironments)(jsonEnvironmentFiles, options, options.workspaceEnvironments || []), specValidationResult = await (0, workspace_utils_1.validateSpecifications)(options.specificationFiles || [], options, options.workspaceSpecifications || []), hasPrepareChanges = collectionStatuses.some((s) => { return s.needsNewId || (s.nestedIdsFixed || 0) > 0; }) ||
                        environmentStatuses.some((s) => { return s.needsNewId; });
                    options.localSpecifications = specValidationResult.localSpecs;
                    options.cloudSpecificationMap = specValidationResult.cloudSpecMap;
                    if (!hasPrepareChanges) {
                        return next(null, options);
                    }
                    if (shouldSkipPrepare) {
                        return next(new Error('Prepare required: local files have invalid/missing IDs. Remove --no-prepare, run prepare, or use --yes to auto-fix.'));
                    }
                    const shouldAutoPrepare = Boolean(options?.yes);
                    if (shouldAutoPrepare) {
                        await (0, workspace_utils_1.regenerateCollections)(collectionStatuses);
                        await (0, workspace_utils_1.regenerateEnvironments)(environmentStatuses);
                        (0, workspace_utils_1.printPrepareSummary)(collectionStatuses, environmentStatuses);
                        return next(null, options);
                    }
                    if (!process.stdin.isTTY) {
                        return next(new Error('Prepare required but cannot prompt in non-interactive mode. Re-run with --yes or run prepare first.'));
                    }
                    const rl = readline_1.default.createInterface({ input: process.stdin, output: process.stdout });
                    rl.question(safe_1.default.cyan('Prepare detected changes to local IDs. Run prepare now? (y/N): '), async (answer) => {
                        rl.close();
                        const a = answer.trim().toLowerCase(), doPrepare = a === 'y' || a === 'yes';
                        if (!doPrepare) {
                            return next(new Error('Push aborted: prepare changes were required and not approved.'));
                        }
                        try {
                            await (0, workspace_utils_1.regenerateCollections)(collectionStatuses);
                            await (0, workspace_utils_1.regenerateEnvironments)(environmentStatuses);
                            (0, workspace_utils_1.printPrepareSummary)(collectionStatuses, environmentStatuses);
                            return next(null, options);
                        }
                        catch (e) {
                            return next(e);
                        }
                    });
                })
                    .catch((e) => { return next(e); });
            },
            (opts, next) => {
                options = opts;
                options.localCollections = new Map();
                options.v3Collections = [];
                options.collectionsToCreate = [];
                const collectionFiles = options.collectionFiles ?? [];
                if (collectionFiles.length === 0) {
                    return next(null, options);
                }
                Promise.resolve()
                    .then(async () => {
                    for (const filePath of collectionFiles) {
                        if (await (0, v3_collection_1.isV3CollectionDirectory)(filePath)) {
                            const name = await (0, v3_collection_1.getV3CollectionName)(filePath);
                            let cloudId = null;
                            let source = 'ec';
                            try {
                                const lookup = await (0, resources_1.getCloudIdForPath)(filePath, 'collections');
                                cloudId = lookup.cloudId;
                                source = lookup.source;
                            }
                            catch {
                            }
                            options.v3Collections?.push({
                                dirPath: filePath,
                                name,
                                cloudId,
                                source
                            });
                        }
                        else {
                            try {
                                await processJsonCollectionFile(filePath, options);
                            }
                            catch {
                            }
                        }
                    }
                    const localCollSize = options.localCollections?.size ?? 0;
                    const v3CollLength = options.v3Collections?.length ?? 0;
                    const createCollLength = options.collectionsToCreate?.length ?? 0;
                    const totalCollections = localCollSize + v3CollLength + createCollLength;
                    console_1.default.log(safe_1.default.grey(`Found ${totalCollections} local collection(s) (${localCollSize} JSON, ${v3CollLength} v3 YAML, ${createCollLength} new)`));
                    next(null, options);
                })
                    .catch((e) => { return next(e); });
            },
            (opts, next) => {
                options = opts;
                options.localEnvironments = new Map();
                options.environmentsToCreate = [];
                const environmentFiles = options.environmentFiles ?? [];
                if (environmentFiles.length === 0) {
                    return next(null, options);
                }
                Promise.resolve()
                    .then(async () => {
                    for (const filePath of environmentFiles) {
                        try {
                            await processEnvironmentFile(filePath, options);
                        }
                        catch {
                        }
                    }
                    const localEnvSize = options.localEnvironments?.size ?? 0;
                    const createEnvLength = options.environmentsToCreate?.length ?? 0;
                    const totalEnvs = localEnvSize + createEnvLength;
                    console_1.default.log(safe_1.default.grey(`Found ${totalEnvs} local environment(s) (${localEnvSize} existing, ${createEnvLength} new)\n`));
                    next(null, options);
                })
                    .catch((e) => { return next(e); });
            },
            (opts, next) => {
                options = opts;
                if (options.workspaceCollections && options.workspaceCollections.length >= 0) {
                    return next(null, options);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                (0, workspace_utils_1.getWorkspaceCollections)(workspaceId, options, (err, collections) => {
                    if (err) {
                        return next(err);
                    }
                    options.workspaceCollections = collections;
                    console_1.default.log(safe_1.default.grey(`Found ${(collections ?? []).length} collection(s) in workspace`));
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                if (options.workspaceEnvironments && options.workspaceEnvironments.length >= 0) {
                    return next(null, options);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                (0, workspace_utils_1.getWorkspaceEnvironments)(workspaceId, options, (err, environments) => {
                    if (err) {
                        return next(err);
                    }
                    options.workspaceEnvironments = environments;
                    console_1.default.log(safe_1.default.grey(`Found ${(environments ?? []).length} environment(s) in workspace`));
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                determineSyncCollections(options);
                next(null, options);
            },
            (opts, next) => {
                options = opts;
                determineSyncEnvironments(options);
                next(null, options);
            },
            (opts, next) => {
                options = opts;
                determineSyncSpecifications(options)
                    .then(() => { next(null, options); })
                    .catch((e) => { return next(e); });
            },
            (opts, next) => {
                options = opts;
                determineSyncSpecCollectionLinks(options)
                    .then(() => { next(null, options); })
                    .catch((e) => { return next(e); });
            },
            (opts, next) => {
                options = opts;
                options.stats = {
                    ...options.stats,
                    collectionsUpdated: 0,
                    environmentsUpdated: 0,
                    specificationsUpdated: 0,
                    sdksUploaded: 0,
                    collectionsCreated: 0,
                    environmentsCreated: 0,
                    specificationsCreated: 0,
                    collectionErrors: 0,
                    environmentErrors: 0,
                    specificationErrors: 0,
                    collectionCreateErrors: 0,
                    environmentCreateErrors: 0,
                    specificationCreateErrors: 0,
                    sdkErrors: 0
                };
                const totalCollectionsToUpdate = (options.collectionsToUpdate?.length ?? 0) + (options.v3CollectionsToUpdate?.length ?? 0), totalCollectionsToCreate = (options.collectionsToCreate?.length ?? 0) + (options.v3CollectionsToCreate?.length ?? 0), totalEnvironmentsToCreate = options.environmentsToCreate?.length ?? 0, totalSpecsToCreate = options.specificationsToCreate?.length ?? 0;
                console_1.default.log(safe_1.default.cyan('\nSync Plan:'));
                console_1.default.log(safe_1.default.grey(`  Collections: ${totalCollectionsToUpdate} to update${totalCollectionsToCreate ? `, ${totalCollectionsToCreate} to create` : ''}`));
                console_1.default.log(safe_1.default.grey(`  Environments: ${options.environmentsToUpdate?.length ?? 0} to update${totalEnvironmentsToCreate ? `, ${totalEnvironmentsToCreate} to create` : ''}`));
                console_1.default.log(safe_1.default.grey(`  Specifications: ${options.specificationsToUpdate?.length ?? 0} to update${totalSpecsToCreate ? `, ${totalSpecsToCreate} to create` : ''}`));
                console_1.default.log(safe_1.default.grey(`  SDKs: ${options.discoveredSDKs?.length ?? 0} to sync\n`));
                if (options.yes) {
                    return next(null, options);
                }
                if (!process.stdin.isTTY) {
                    console_1.default.log(safe_1.default.grey('Non-interactive terminal detected. Use --yes to proceed without confirmation.'));
                    options.collectionsToUpdate = [];
                    options.v3CollectionsToUpdate = [];
                    options.collectionsToCreate = [];
                    options.v3CollectionsToCreate = [];
                    options.environmentsToUpdate = [];
                    options.environmentsToCreate = [];
                    options.specificationsToUpdate = [];
                    options.specificationsToCreate = [];
                    options.specCollectionLinksToPush = [];
                    return next(null, options);
                }
                const rl = readline_1.default.createInterface({ input: process.stdin, output: process.stdout });
                rl.question(safe_1.default.cyan('Proceed with pushing changes to the workspace? (y/N): '), (answer) => {
                    rl.close();
                    const a = answer.trim().toLowerCase(), proceed = a === 'y' || a === 'yes';
                    if (!proceed) {
                        console_1.default.log(safe_1.default.yellow('Push aborted by user. No changes applied.'));
                        options.collectionsToUpdate = [];
                        options.v3CollectionsToUpdate = [];
                        options.collectionsToCreate = [];
                        options.v3CollectionsToCreate = [];
                        options.environmentsToUpdate = [];
                        options.environmentsToCreate = [];
                        options.specificationsToUpdate = [];
                        options.specificationsToCreate = [];
                        options.specCollectionLinksToPush = [];
                    }
                    next(null, options);
                });
            },
            (opts, next) => {
                options = opts;
                const collectionsToUpdate = options.collectionsToUpdate ?? [];
                if (collectionsToUpdate.length === 0) {
                    return next(null);
                }
                console_1.default.log(safe_1.default.cyan('\nSyncing collections...\n'));
                (0, eachSeries_1.default)(collectionsToUpdate, (updateInfo, callback) => {
                    const localCollection = options.localCollections?.get(updateInfo.localId);
                    if (!localCollection) {
                        return callback();
                    }
                    (0, workspace_utils_1.updateCollection)(updateInfo.workspaceId, localCollection.data, options, (err) => {
                        if (err) {
                            console_1.default.error(safe_1.default.red(`${localCollection.name}: Failed to sync - ${err.message}`));
                            options.stats.collectionErrors++;
                        }
                        else {
                            console_1.default.log(safe_1.default.green(`${localCollection.name}: Synced successfully`));
                            options.stats.collectionsUpdated++;
                        }
                        callback();
                    });
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                if (!options.v3CollectionsToUpdate || options.v3CollectionsToUpdate.length === 0) {
                    return next(null);
                }
                console_1.default.log(safe_1.default.cyan('\nSyncing v3 collections...\n'));
                (0, eachSeries_1.default)(options.v3CollectionsToUpdate, (updateInfo, callback) => {
                    const { entry, cloudCollectionId } = updateInfo;
                    (async () => {
                        try {
                            if (entry.source === 'ec') {
                                await (0, workspace_utils_1.updateV3CollectionEC)(cloudCollectionId, entry.dirPath, entry.name, options);
                            }
                            else {
                                const v2Collection = await (0, v3_collection_1.readV3CollectionAsV2)(entry.dirPath);
                                await new Promise((resolve, reject) => {
                                    (0, workspace_utils_1.updateV3CollectionV2)(cloudCollectionId, v2Collection, options, (err) => {
                                        if (err) {
                                            return reject(err);
                                        }
                                        resolve();
                                    });
                                });
                            }
                            console_1.default.log(safe_1.default.green(`${entry.name}: Synced successfully`));
                            options.stats.collectionsUpdated++;
                        }
                        catch (e) {
                            console_1.default.error(safe_1.default.red(`${entry.name}: Failed to sync - ${e.message}`));
                            options.stats.collectionErrors++;
                        }
                        callback();
                    })();
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                const environmentsToUpdate = options.environmentsToUpdate ?? [];
                if (environmentsToUpdate.length === 0) {
                    return next(null);
                }
                console_1.default.log(safe_1.default.cyan('\nSyncing environments...\n'));
                (0, eachSeries_1.default)(environmentsToUpdate, (updateInfo, callback) => {
                    const localEnvironment = options.localEnvironments?.get(updateInfo.localId);
                    if (!localEnvironment) {
                        return callback();
                    }
                    (0, workspace_utils_1.updateEnvironment)(updateInfo.workspaceId, localEnvironment.data, options, (err) => {
                        if (err) {
                            console_1.default.error(safe_1.default.red(`${localEnvironment.name}: Failed to sync - ${err.message}`));
                            options.stats.environmentErrors++;
                        }
                        else {
                            console_1.default.log(safe_1.default.green(`${localEnvironment.name}: Synced successfully`));
                            options.stats.environmentsUpdated++;
                        }
                        callback();
                    });
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                const specificationsToUpdate = options.specificationsToUpdate ?? [];
                if (specificationsToUpdate.length === 0) {
                    return next(null);
                }
                console_1.default.log(safe_1.default.cyan('\nSyncing specifications...\n'));
                const normalizePath = (p) => { return p.replace(/\\/g, '/').replace(/^\/+/, ''); };
                (0, eachSeries_1.default)(specificationsToUpdate, (updateInfo, callback) => {
                    const localSpec = options.localSpecifications?.get(updateInfo.localRootPath);
                    if (!localSpec) {
                        return callback();
                    }
                    (async () => {
                        try {
                            const cloudFiles = await (0, workspace_utils_1.fetchSpecificationFiles)(updateInfo.specId, options);
                            const cloudPathMap = new Map();
                            for (const f of cloudFiles) {
                                cloudPathMap.set(normalizePath(f.path), f);
                            }
                            const localRootPath = normalizePath(updateInfo.localRootPath);
                            const localPathMap = new Map();
                            localPathMap.set(localRootPath, {
                                content: localSpec.content,
                                type: 'ROOT'
                            });
                            const bulkPayload = {};
                            const filesToUpdate = [];
                            for (const [localPath, localFile] of localPathMap) {
                                const cloudFile = cloudPathMap.get(localPath);
                                if (cloudFile) {
                                    filesToUpdate.push({
                                        id: cloudFile.id,
                                        content: localFile.content
                                    });
                                }
                            }
                            if (filesToUpdate.length > 0) {
                                bulkPayload.update = filesToUpdate;
                            }
                            const filesToCreate = [];
                            for (const [localPath, localFile] of localPathMap) {
                                if (!cloudPathMap.has(localPath)) {
                                    filesToCreate.push({
                                        path: localPath,
                                        content: localFile.content,
                                        type: localFile.type
                                    });
                                }
                            }
                            if (filesToCreate.length > 0) {
                                bulkPayload.create = filesToCreate;
                            }
                            if (Object.keys(bulkPayload).length > 0) {
                                await (0, workspace_utils_1.bulkModifySpecificationFiles)(updateInfo.specId, bulkPayload, options);
                            }
                            console_1.default.log(safe_1.default.green(`${localSpec.name}: Synced successfully`));
                            options.stats.specificationsUpdated++;
                        }
                        catch (e) {
                            const detail = e.response ? JSON.stringify({
                                status: e.response.status,
                                statusText: e.response.statusText,
                                url: e.response.config?.url,
                                data: e.response.data
                            }, null, 2) : 'No response object';
                            console_1.default.error(safe_1.default.red(`${localSpec.name}: Failed to sync - ${e.message}`));
                            console_1.default.error(safe_1.default.red(`  Error detail: ${detail}`));
                            options.stats.specificationErrors++;
                        }
                        callback();
                    })();
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                if (!options.collectionsToCreate || options.collectionsToCreate.length === 0) {
                    return next(null);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                console_1.default.log(safe_1.default.cyan('\nCreating new collections...\n'));
                (0, eachSeries_1.default)(options.collectionsToCreate, (createInfo, callback) => {
                    (0, workspace_utils_1.createCollection)(createInfo.data, workspaceId, options, (err, body) => {
                        if (err) {
                            console_1.default.error(safe_1.default.red(`${createInfo.name}: Failed to create - ${err.message}`));
                            options.stats.collectionCreateErrors++;
                            return callback();
                        }
                        const collectionId = body?.info?._postman_id || body?.collection?.id || body?.id;
                        if (!collectionId) {
                            console_1.default.error(safe_1.default.red(`${createInfo.name}: Created but failed to determine ID from response`));
                            options.stats.collectionCreateErrors++;
                            return callback();
                        }
                        (0, workspace_utils_1.getWorkspaceCollections)(workspaceId, options, (wsErr, wsCollections) => {
                            if (wsErr || !wsCollections) {
                                console_1.default.error(safe_1.default.red(`${createInfo.name}: Created but failed to resolve workspace ID`));
                                options.stats.collectionCreateErrors++;
                                return callback();
                            }
                            const uid = wsCollections.find((wsId) => { return matchesId(wsId, collectionId); });
                            if (!uid) {
                                console_1.default.error(safe_1.default.red(`${createInfo.name}: Created but failed to resolve workspace ID`));
                                options.stats.collectionCreateErrors++;
                                return callback();
                            }
                            (async () => {
                                try {
                                    await (0, resources_1.setCloudIdForPath)(createInfo.filePath, 'collections', uid);
                                    await (0, workspace_utils_1.writeCollectionFileFromServer)(createInfo.filePath, body);
                                    console_1.default.log(safe_1.default.green(`${createInfo.name}: Created successfully`));
                                    options.stats.collectionsCreated++;
                                }
                                catch (e) {
                                    console_1.default.error(safe_1.default.red(`${createInfo.name}: Created on server but failed to persist locally - ${e.message}`));
                                    options.stats.collectionCreateErrors++;
                                }
                                callback();
                            })();
                        });
                    });
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                if (!options.v3CollectionsToCreate || options.v3CollectionsToCreate.length === 0) {
                    return next(null);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                console_1.default.log(safe_1.default.cyan('\nCreating new v3 collections...\n'));
                (0, eachSeries_1.default)(options.v3CollectionsToCreate, (createInfo, callback) => {
                    const { entry } = createInfo;
                    (async () => {
                        try {
                            const v2Collection = await (0, v3_collection_1.readV3CollectionAsV2)(entry.dirPath);
                            const body = await new Promise((resolve, reject) => {
                                (0, workspace_utils_1.createCollection)(v2Collection, workspaceId, options, (err, result) => {
                                    if (err) {
                                        return reject(err);
                                    }
                                    resolve(result);
                                });
                            });
                            const collectionId = body?.info?._postman_id || body?.collection?.id || body?.id;
                            if (!collectionId) {
                                console_1.default.error(safe_1.default.red(`${entry.name}: Created but failed to determine ID from response`));
                                options.stats.collectionCreateErrors++;
                                return callback();
                            }
                            const wsCollections = await new Promise((resolve, reject) => {
                                (0, workspace_utils_1.getWorkspaceCollections)(workspaceId, options, (wsErr, collections) => {
                                    if (wsErr) {
                                        return reject(wsErr);
                                    }
                                    resolve(collections || []);
                                });
                            });
                            const uid = wsCollections.find((wsId) => { return matchesId(wsId, collectionId); });
                            if (!uid) {
                                console_1.default.error(safe_1.default.red(`${entry.name}: Created but failed to resolve workspace ID`));
                                options.stats.collectionCreateErrors++;
                                return callback();
                            }
                            await (0, resources_1.setCloudIdForPath)(entry.dirPath, 'collections', uid);
                            console_1.default.log(safe_1.default.green(`${entry.name}: Created successfully`));
                            options.stats.collectionsCreated++;
                        }
                        catch (e) {
                            console_1.default.error(safe_1.default.red(`${entry.name}: Failed to create - ${e.message}`));
                            options.stats.collectionCreateErrors++;
                        }
                        callback();
                    })();
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                if (!options.environmentsToCreate || options.environmentsToCreate.length === 0) {
                    return next(null);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                console_1.default.log(safe_1.default.cyan('\nCreating new environments...\n'));
                (0, eachSeries_1.default)(options.environmentsToCreate, (createInfo, callback) => {
                    const envData = { ...createInfo.data };
                    if (!envData.id) {
                        envData.id = (0, uuid_1.v4)();
                    }
                    (0, workspace_utils_1.createEnvironment)(envData, workspaceId, options, (err, body) => {
                        if (err) {
                            console_1.default.error(safe_1.default.red(`${createInfo.name}: Failed to create - ${err.message}`));
                            options.stats.environmentCreateErrors++;
                            return callback();
                        }
                        const environmentId = body?.id || envData.id;
                        (0, workspace_utils_1.getWorkspaceEnvironments)(workspaceId, options, (wsErr, wsEnvironments) => {
                            if (wsErr || !wsEnvironments) {
                                console_1.default.error(safe_1.default.red(`${createInfo.name}: Created but failed to resolve workspace ID`));
                                options.stats.environmentCreateErrors++;
                                return callback();
                            }
                            const uid = wsEnvironments.find((wsId) => { return matchesId(wsId, environmentId); });
                            if (!uid) {
                                console_1.default.error(safe_1.default.red(`${createInfo.name}: Created but failed to resolve workspace ID`));
                                options.stats.environmentCreateErrors++;
                                return callback();
                            }
                            (async () => {
                                try {
                                    await (0, resources_1.setCloudIdForPath)(createInfo.filePath, 'environments', uid);
                                    console_1.default.log(safe_1.default.green(`${createInfo.name}: Created successfully`));
                                    options.stats.environmentsCreated++;
                                }
                                catch (e) {
                                    console_1.default.error(safe_1.default.red(`${createInfo.name}: Created on server but failed to persist locally - ${e.message}`));
                                    options.stats.environmentCreateErrors++;
                                }
                                callback();
                            })();
                        });
                    });
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                if (!options.specificationsToCreate || options.specificationsToCreate.length === 0) {
                    return next(null);
                }
                const workspaceId = options.workspaceId;
                if (!workspaceId) {
                    return next(new Error('Workspace ID not set'));
                }
                console_1.default.log(safe_1.default.cyan('\nCreating new specifications...\n'));
                (0, eachSeries_1.default)(options.specificationsToCreate, (createInfo, callback) => {
                    const { normalizedPath, localSpec } = createInfo;
                    const specData = {
                        name: localSpec.name,
                        type: localSpec.type,
                        content: localSpec.content,
                        rootPath: normalizedPath
                    };
                    (0, workspace_utils_1.createSpecification)(specData, workspaceId, options, (err, body) => {
                        if (err) {
                            console_1.default.error(safe_1.default.red(`${localSpec.name}: Failed to create - ${err.message}`));
                            options.stats.specificationCreateErrors++;
                            return callback();
                        }
                        const specId = body?.id;
                        if (!specId) {
                            console_1.default.error(safe_1.default.red(`${localSpec.name}: Created but failed to determine ID from response`));
                            options.stats.specificationCreateErrors++;
                            return callback();
                        }
                        (async () => {
                            try {
                                await (0, resources_1.setCloudIdForPath)(localSpec.filePath, 'specs', specId);
                                console_1.default.log(safe_1.default.green(`${localSpec.name}: Created successfully`));
                                options.stats.specificationsCreated++;
                            }
                            catch (e) {
                                console_1.default.error(safe_1.default.red(`${localSpec.name}: Created on server but failed to persist locally - ${e.message}`));
                                options.stats.specificationCreateErrors++;
                            }
                            callback();
                        })();
                    });
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                const specCollectionLinks = options.specCollectionLinksToPush ?? [];
                if (specCollectionLinks.length === 0) {
                    return next(null);
                }
                console_1.default.log(safe_1.default.cyan('\nPushing spec-collection links...\n'));
                const linksBySpec = new Map();
                for (const link of specCollectionLinks) {
                    if (!linksBySpec.has(link.specId)) {
                        linksBySpec.set(link.specId, []);
                    }
                    const { specId, ...collectionLink } = link;
                    linksBySpec.get(specId)?.push(collectionLink);
                }
                const specIds = Array.from(linksBySpec.keys());
                (0, eachSeries_1.default)(specIds, (specId, callback) => {
                    const collectionLinks = linksBySpec.get(specId) || [];
                    const specName = collectionLinks[0]?.specName || specId;
                    (async () => {
                        try {
                            const response = await (0, SpecificationService_1.upsertSpecCollectionLinks)(specId, collectionLinks, options);
                            const createdCount = response.data?.created?.length || 0;
                            const updatedCount = response.data?.updated?.length || 0;
                            if (createdCount > 0 || updatedCount > 0) {
                                console_1.default.log(safe_1.default.green(`${specName}: ${createdCount} link(s) created, ${updatedCount} link(s) updated`));
                            }
                        }
                        catch (e) {
                            const detail = e.response ? JSON.stringify({
                                status: e.response.status,
                                statusText: e.response.statusText,
                                url: e.response.config?.url,
                                data: e.response.data
                            }, null, 2) : 'No response object';
                            console_1.default.error(safe_1.default.red(`${specName}: Failed to push links - ${e.message}`));
                            console_1.default.error(safe_1.default.red(`  Error detail: ${detail}`));
                        }
                        callback();
                    })();
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                if (!options.discoveredSDKs || options.discoveredSDKs.length === 0) {
                    return next(null);
                }
                (0, workspace_utils_1.populateSDKCloudHashes)(options.discoveredSDKs, options)
                    .then(() => {
                    return next(null);
                })
                    .catch(() => {
                    console_1.default.log(safe_1.default.dim('  ℹ Could not fetch build summaries, will upload all SDKs'));
                    return next(null);
                });
            },
            (next) => {
                if (!options.discoveredSDKs || options.discoveredSDKs.length === 0) {
                    return next(null);
                }
                console_1.default.log(safe_1.default.cyan('\nSyncing SDKs...\n'));
                (0, eachSeries_1.default)(options.discoveredSDKs, (sdk, callback) => {
                    const sdkDisplayName = `${sdk.entityName} (${sdk.language})`;
                    (0, workspace_utils_1.syncSDK)(sdk, options, (err, result) => {
                        if (err) {
                            console_1.default.error(safe_1.default.red(`${sdkDisplayName}: Failed - ${err.message}`));
                            options.stats.sdkErrors++;
                        }
                        else if (result?.uploaded) {
                            console_1.default.log(safe_1.default.green(`${sdkDisplayName}: Uploaded successfully`));
                            options.stats.sdksUploaded++;
                        }
                        else {
                            console_1.default.log(safe_1.default.dim(`${sdkDisplayName}: No changes detected, skipped`));
                            options.stats.sdksUploaded++;
                        }
                        callback();
                    });
                }, (err) => {
                    next(err);
                });
            },
            (next) => {
                const totalSuccess = options.stats.collectionsUpdated +
                    options.stats.collectionsCreated +
                    options.stats.environmentsUpdated +
                    options.stats.environmentsCreated +
                    options.stats.specificationsUpdated +
                    options.stats.specificationsCreated +
                    options.stats.sdksUploaded, totalSkipped = options.stats.collectionsSkipped +
                    options.stats.environmentsSkipped +
                    options.stats.specificationsSkipped, totalErrors = options.stats.collectionErrors + options.stats.collectionCreateErrors +
                    options.stats.environmentErrors + options.stats.environmentCreateErrors +
                    options.stats.specificationErrors + options.stats.specificationCreateErrors +
                    options.stats.sdkErrors;
                console_1.default.log(safe_1.default.cyan('\n═══════════════════════════════════════════════════════════'));
                console_1.default.log(safe_1.default.cyan('                      PUSH SUMMARY'));
                console_1.default.log(safe_1.default.cyan('═══════════════════════════════════════════════════════════\n'));
                console_1.default.log(safe_1.default.white('Collections:'));
                console_1.default.log(safe_1.default.grey(`  Updated: ${options.stats.collectionsUpdated}`));
                if (options.stats.collectionsCreated > 0) {
                    console_1.default.log(safe_1.default.green(`  Created: ${options.stats.collectionsCreated}`));
                }
                if (options.stats.collectionsSkipped > 0) {
                    console_1.default.log(safe_1.default.yellow(`  Skipped: ${options.stats.collectionsSkipped}`));
                }
                console_1.default.log(safe_1.default.white('\nEnvironments:'));
                console_1.default.log(safe_1.default.grey(`  Updated: ${options.stats.environmentsUpdated}`));
                if (options.stats.environmentsCreated > 0) {
                    console_1.default.log(safe_1.default.green(`  Created: ${options.stats.environmentsCreated}`));
                }
                if (options.stats.environmentsSkipped > 0) {
                    console_1.default.log(safe_1.default.yellow(`  Skipped: ${options.stats.environmentsSkipped}`));
                }
                console_1.default.log(safe_1.default.white('\nSpecifications:'));
                console_1.default.log(safe_1.default.grey(`  Updated: ${options.stats.specificationsUpdated}`));
                if (options.stats.specificationsCreated > 0) {
                    console_1.default.log(safe_1.default.green(`  Created: ${options.stats.specificationsCreated}`));
                }
                if (options.stats.specificationsSkipped > 0) {
                    console_1.default.log(safe_1.default.yellow(`  Skipped: ${options.stats.specificationsSkipped}`));
                }
                console_1.default.log(safe_1.default.white('\nSDKs:'));
                console_1.default.log(safe_1.default.grey(`  Uploaded: ${options.stats.sdksUploaded}`));
                console_1.default.log(safe_1.default.white('\nTotal:'));
                console_1.default.log(safe_1.default.green(`  Success: ${totalSuccess}`));
                if (totalSkipped > 0) {
                    console_1.default.log(safe_1.default.yellow(`  Skipped: ${totalSkipped}`));
                }
                if (totalErrors > 0) {
                    console_1.default.log(safe_1.default.red(`  Errors: ${totalErrors}`));
                    exitCode = 1;
                }
                console_1.default.log(safe_1.default.cyan('\n═══════════════════════════════════════════════════════════\n'));
                next(null);
            }
        ], (err) => {
            if (err) {
                console_1.default.error(safe_1.default.red(`\nError: ${err.message || err}`));
                logger.error(`Workspace push command failed: ${err.message || err}`);
                exitCode = 1;
            }
            process.exit(exitCode);
        });
    });
};
pushCommand._testing = {
    processJsonCollectionFile,
    processEnvironmentFile,
    determineSyncCollections,
    determineSyncEnvironments,
    determineSyncSpecifications,
    determineSyncSpecCollectionLinks,
    matchesId,
    readWorkflowsConfig: workflows_1.readWorkflowsConfig,
    resolveWorkflowLinks: workflows_1.resolveWorkflowLinks
};
module.exports = pushCommand;
