// @postman/pm-bin-linux-x64/bin/lib/commands/workspace/resources.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.classifyCloudCollectionIdSource = classifyCloudCollectionIdSource;
exports.readResourcesConfig = readResourcesConfig;
exports.writeResourcesConfig = writeResourcesConfig;
exports.getCloudIdForPath = getCloudIdForPath;
exports.setCloudIdForPath = setCloudIdForPath;
exports.setWorkspaceId = setWorkspaceId;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const POSTMAN_DIR_NAME = '.postman';
const RESOURCES_FILE_NAME = 'resources.yaml';
const MONGO_OBJECT_ID_REGEX = /^[a-f0-9]{24}$/i;
const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const UID_REGEX = /^[^-]+-[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
function classifyCloudCollectionIdSource(cloudId) {
    if (!cloudId) {
        throw new Error(`Cloud collection ID is required but not found in ${RESOURCES_FILE_NAME}`);
    }
    if (MONGO_OBJECT_ID_REGEX.test(cloudId)) {
        return 'ec';
    }
    if (UUID_REGEX.test(cloudId) || UID_REGEX.test(cloudId)) {
        return 'v2';
    }
    throw new Error(`Unrecognized cloud collection ID format: ${cloudId}`);
}
function getFirstCloudId(cloudIds) {
    if (typeof cloudIds === 'string' && cloudIds.length > 0) {
        return cloudIds;
    }
    if (!Array.isArray(cloudIds) || cloudIds.length === 0) {
        return null;
    }
    const firstId = cloudIds[0];
    if (typeof firstId !== 'string' || firstId.length === 0) {
        return null;
    }
    return firstId;
}
function getResourcesFilePath(workingDir) {
    const baseDir = workingDir || process.cwd();
    return path_1.default.join(baseDir, POSTMAN_DIR_NAME, RESOURCES_FILE_NAME);
}
function normalizeResourcePath(inputPath, workingDir) {
    const baseDir = workingDir || process.cwd();
    const configRoot = path_1.default.join(baseDir, POSTMAN_DIR_NAME);
    if (path_1.default.isAbsolute(inputPath)) {
        return path_1.default.relative(configRoot, inputPath).split(path_1.default.sep).join('/');
    }
    return inputPath.split(path_1.default.sep).join('/');
}
function resolveResourcePath(pathRef, workingDir) {
    if (path_1.default.isAbsolute(pathRef)) {
        return path_1.default.normalize(pathRef);
    }
    const baseDir = workingDir || process.cwd();
    const configRoot = path_1.default.join(baseDir, POSTMAN_DIR_NAME);
    return path_1.default.normalize(path_1.default.resolve(configRoot, pathRef));
}
async function readResourcesConfig(workingDir) {
    const filePath = getResourcesFilePath(workingDir);
    try {
        const content = await promises_1.default.readFile(filePath, 'utf8');
        const parsed = js_yaml_1.default.load(content);
        if (!parsed || typeof parsed !== 'object') {
            return {};
        }
        return parsed;
    }
    catch (err) {
        if (err.code === 'ENOENT') {
            return {};
        }
        throw new Error(`Failed to read ${POSTMAN_DIR_NAME}/${RESOURCES_FILE_NAME}: ${err.message}`);
    }
}
async function writeResourcesConfig(config, workingDir) {
    const filePath = getResourcesFilePath(workingDir);
    const dir = path_1.default.dirname(filePath);
    await promises_1.default.mkdir(dir, { recursive: true });
    const content = js_yaml_1.default.dump(config, { lineWidth: -1, noRefs: true });
    await promises_1.default.writeFile(filePath, content, 'utf8');
}
async function getCloudIdForPath(dirPath, entityType, workingDir) {
    const config = await readResourcesConfig(workingDir);
    const cloudResources = config.cloudResources?.[entityType];
    if (!cloudResources) {
        throw new Error(`No cloud ${entityType} mappings found in ${RESOURCES_FILE_NAME} for path: ${dirPath}`);
    }
    const resolvedDirPath = path_1.default.normalize(path_1.default.resolve(dirPath));
    for (const [pathRef, cloudIds] of Object.entries(cloudResources)) {
        const resolvedConfigPath = resolveResourcePath(pathRef, workingDir);
        if (path_1.default.normalize(resolvedConfigPath) === resolvedDirPath) {
            const cloudId = getFirstCloudId(cloudIds);
            return {
                cloudId: cloudId,
                source: classifyCloudCollectionIdSource(cloudId)
            };
        }
    }
    throw new Error(`No cloud ID mapping found in ${RESOURCES_FILE_NAME} for ${entityType} path: ${dirPath}`);
}
async function setCloudIdForPath(dirPath, entityType, cloudId, workingDir) {
    const config = await readResourcesConfig(workingDir);
    if (!config.cloudResources) {
        config.cloudResources = {};
    }
    if (!config.cloudResources[entityType]) {
        config.cloudResources[entityType] = {};
    }
    const relativePath = normalizeResourcePath(dirPath, workingDir);
    config.cloudResources[entityType][relativePath] = cloudId;
    await writeResourcesConfig(config, workingDir);
}
async function setWorkspaceId(workspaceId, workingDir) {
    const config = await readResourcesConfig(workingDir);
    if (!config.workspace) {
        config.workspace = {};
    }
    config.workspace.id = workspaceId;
    await writeResourcesConfig(config, workingDir);
}
