// @postman/pm-bin-linux-x64/bin/lib/commands/workspace/workspace-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readEnvironmentFiles = exports.readEnvironmentFile = exports.isJsonFile = void 0;
exports.readCollectionFiles = readCollectionFiles;
exports.readSpecificationFiles = readSpecificationFiles;
exports.readJSONFile = readJSONFile;
exports.writeJSONFile = writeJSONFile;
exports.writeCollectionFileFromServer = writeCollectionFileFromServer;
exports.readPostmanConfig = readPostmanConfig;
exports.discoverEntities = discoverEntities;
exports.updateCollection = updateCollection;
exports.updateV3CollectionEC = updateV3CollectionEC;
exports.updateV3CollectionV2 = updateV3CollectionV2;
exports.updateEnvironment = updateEnvironment;
exports.checkCollectionExists = checkCollectionExists;
exports.checkEnvironmentExists = checkEnvironmentExists;
exports.getWorkspaceCollections = getWorkspaceCollections;
exports.getWorkspaceEnvironments = getWorkspaceEnvironments;
exports.createCollection = createCollection;
exports.createEnvironment = createEnvironment;
exports.deleteCollection = deleteCollection;
exports.deleteEnvironment = deleteEnvironment;
exports.regenerateCollectionIds = regenerateCollectionIds;
exports.regenerateEnvironmentId = regenerateEnvironmentId;
exports.validateCollections = validateCollections;
exports.validateEnvironments = validateEnvironments;
exports.regenerateCollections = regenerateCollections;
exports.regenerateEnvironments = regenerateEnvironments;
exports.getWorkspaceSpecifications = getWorkspaceSpecifications;
exports.createSpecification = createSpecification;
exports.getSpecRootFilePath = getSpecRootFilePath;
exports.updateSpecification = updateSpecification;
exports.bulkModifySpecificationFiles = bulkModifySpecificationFiles;
exports.fetchSpecificationFiles = fetchSpecificationFiles;
exports.updateSpecificationFileAsync = updateSpecificationFileAsync;
exports.detectSpecType = detectSpecType;
exports.readSpecFile = readSpecFile;
exports.normalizeSpecPath = normalizeSpecPath;
exports.validateSpecifications = validateSpecifications;
exports.printPrepareSummary = printPrepareSummary;
exports.discoverSDKs = discoverSDKs;
exports.zipDirectory = zipDirectory;
exports.computeSdkHash = computeSdkHash;
exports.populateSDKCloudHashes = populateSDKCloudHashes;
exports.syncSDK = syncSDK;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const uuid_1 = require("uuid");
const liquid_json_1 = __importDefault(require("liquid-json"));
const safe_1 = __importDefault(require("colors/safe"));
const console_1 = __importDefault(require("../../framework/console"));
const print_1 = __importDefault(require("../../framework/print"));
const v3_collection_1 = require("../../postman-entities/collection/v3-collection");
const PostmanResource = require('../../api/postman-resource'), { COLLECTION, ENVIRONMENT } = PostmanResource;
const SpecificationService = require('../../postman-entities/spec/Service/SpecificationService');
const file_utils_1 = require("../../postman-entities/file-utils");
Object.defineProperty(exports, "isJsonFile", { enumerable: true, get: function () { return file_utils_1.isJsonFile; } });
const environment_1 = require("../../postman-entities/environment/environment");
Object.defineProperty(exports, "readEnvironmentFile", { enumerable: true, get: function () { return environment_1.readEnvironmentFile; } });
Object.defineProperty(exports, "readEnvironmentFiles", { enumerable: true, get: function () { return environment_1.readEnvironmentFiles; } });
const resources_1 = require("./resources");
function readCollectionFiles(directory) {
    return (0, file_utils_1.scanDirectoryForFiles)(directory, [file_utils_1.JSON_EXTENSION]);
}
function readSpecificationFiles(directory) {
    return (0, file_utils_1.scanDirectoryForFiles)(directory, file_utils_1.JSON_AND_YAML_EXTENSIONS);
}
async function readJSONFile(filePath) {
    try {
        const content = await promises_1.default.readFile(filePath, 'utf8');
        return liquid_json_1.default.parse(content);
    }
    catch (err) {
        throw new Error(`Failed to read/parse ${filePath}: ${err.message}`);
    }
}
async function writeJSONFile(filePath, data) {
    try {
        const content = JSON.stringify(data, null, 2);
        await promises_1.default.writeFile(filePath, content, 'utf8');
    }
    catch (err) {
        throw new Error(`Failed to write ${filePath}: ${err.message}`);
    }
}
async function writeCollectionFileFromServer(filePath, serverCollection) {
    await writeJSONFile(filePath, serverCollection);
}
async function tryReadFromResources(baseDir) {
    const { readResourcesConfig } = require('./resources');
    const resourcesConfig = await readResourcesConfig(baseDir), workspaceId = resourcesConfig?.workspace?.id;
    if (!workspaceId) {
        return null;
    }
    const configDir = path_1.default.join(baseDir, '.postman'), localResources = resourcesConfig.localResources || {}, resolveResourcePaths = (paths) => {
        if (!Array.isArray(paths)) {
            return undefined;
        }
        return paths
            .filter((p) => { return typeof p === 'string' && p.trim().length > 0; })
            .map((p) => { return path_1.default.resolve(configDir, p.trim()); });
    };
    return {
        workspaceId,
        entities: {
            collections: resolveResourcePaths(localResources.collections),
            environments: resolveResourcePaths(localResources.environments),
            specs: resolveResourcePaths(localResources.specs),
            flows: resolveResourcePaths(localResources.flows)
        }
    };
}
async function tryReadFromConfigJson(baseDir) {
    const configDir = path_1.default.join(baseDir, '.postman'), jsonPath = path_1.default.join(configDir, 'config.json');
    try {
        const jsonRaw = await promises_1.default.readFile(jsonPath, 'utf8'), config = JSON.parse(jsonRaw), workspaceId = config?.workspace?.id;
        if (!workspaceId) {
            return null;
        }
        const resolveList = (arr, label, exts, allowDirectories = false) => {
            if (!Array.isArray(arr)) {
                return undefined;
            }
            const results = [];
            arr.forEach((p) => {
                if (typeof p !== 'string') {
                    console_1.default.log(safe_1.default.yellow(`Invalid ${label} path (not a string): ${String(p)} - skipping`));
                    return;
                }
                const raw = p.trim();
                if (!raw) {
                    console_1.default.log(safe_1.default.yellow(`Invalid ${label} path (empty) - skipping`));
                    return;
                }
                const lower = raw.toLowerCase();
                const allowed = exts.some((ext) => { return lower.endsWith(ext); });
                if (!allowed && !allowDirectories) {
                    console_1.default.log(safe_1.default.yellow(`Invalid ${label} path (unsupported extension): ${p} - skipping`));
                    return;
                }
                results.push(path_1.default.resolve(configDir, raw));
            });
            return results;
        }, entities = {
            collections: resolveList(config?.entities?.collections, 'collection', [file_utils_1.JSON_EXTENSION], true),
            environments: resolveList(config?.entities?.environments, 'environment', file_utils_1.JSON_AND_YAML_EXTENSIONS),
            specs: resolveList(config?.entities?.specs, 'spec', file_utils_1.JSON_AND_YAML_EXTENSIONS),
            flows: resolveList(config?.entities?.flows, 'flow', [file_utils_1.JSON_EXTENSION])
        };
        return { workspaceId, entities };
    }
    catch (err) {
        if (err instanceof SyntaxError) {
            throw new Error(`Invalid JSON in .postman/config.json: ${err.message}`);
        }
        if (err.code !== 'ENOENT') {
            throw err;
        }
        return null;
    }
}
async function readPostmanConfig(workingDir) {
    const baseDir = workingDir || process.cwd();
    const fromResources = await tryReadFromResources(baseDir);
    if (fromResources) {
        return fromResources;
    }
    const fromConfigJson = await tryReadFromConfigJson(baseDir);
    if (fromConfigJson) {
        return fromConfigJson;
    }
    return { workspaceId: undefined, entities: {} };
}
async function discoverEntities(entityType, workingDir) {
    const baseDir = workingDir || process.cwd(), entitiesFromConfig = [], allFilePaths = new Set(), entityDir = path_1.default.join(baseDir, 'postman', entityType);
    try {
        const config = await readPostmanConfig(workingDir), configEntities = config?.entities?.[entityType] || [];
        for (const filePath of configEntities) {
            if (filePath) {
                entitiesFromConfig.push(filePath);
                allFilePaths.add(filePath);
            }
        }
    }
    catch (err) {
        print_1.default.lf(safe_1.default.yellow(`Could not read config.json: ${err.message}`));
    }
    try {
        let readFunction;
        if (entityType === 'collections') {
            readFunction = readCollectionFiles;
        }
        else if (entityType === 'environments') {
            readFunction = environment_1.readEnvironmentFiles;
        }
        else if (entityType === 'specs') {
            readFunction = readSpecificationFiles;
        }
        const scannedFiles = await readFunction(entityDir);
        if (!scannedFiles) {
            return;
        }
        for (const filePath of scannedFiles) {
            const resolvedPath = path_1.default.resolve(filePath);
            allFilePaths.add(resolvedPath);
        }
    }
    catch (err) {
        print_1.default.lf(safe_1.default.grey(`Could not scan ${entityDir}: ${err.message}`));
    }
    if (entityType === 'collections') {
        try {
            const v3Dirs = await (0, v3_collection_1.discoverV3Collections)(baseDir);
            for (const dirPath of v3Dirs) {
                allFilePaths.add(path_1.default.resolve(dirPath));
            }
        }
        catch (err) {
            print_1.default.lf(safe_1.default.grey(`Could not scan for v3 collections: ${err.message}`));
        }
    }
    return Array.from(allFilePaths);
}
function updateCollection(collectionId, collectionData, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(COLLECTION, collectionId, options.postmanApiKey, options.accessToken);
    return resource.deepUpdate(collectionData, options.userId, collectionId, callback);
}
async function updateV3CollectionEC(cloudCollectionId, dirPath, collectionName, options) {
    const { readV3CollectionAsExtensible } = require('../../postman-entities/collection/v3-collection');
    const { updateExtensibleCollection } = require('../../api/ec-resource');
    const extensible = await readV3CollectionAsExtensible(dirPath);
    await updateExtensibleCollection(cloudCollectionId, extensible, collectionName, {
        postmanApiKey: options.postmanApiKey,
        accessToken: options.accessToken
    });
}
function updateV3CollectionV2(cloudCollectionId, v2CollectionData, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    if (v2CollectionData?.info) {
        v2CollectionData.info._postman_id = cloudCollectionId;
    }
    const resource = new PostmanResource(COLLECTION, cloudCollectionId, options.postmanApiKey, options.accessToken);
    return resource.deepUpdate(v2CollectionData, options.userId, cloudCollectionId, callback);
}
function updateEnvironment(environmentId, environmentData, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(ENVIRONMENT, environmentId, options.postmanApiKey, options.accessToken), rawValues = (environmentData && Array.isArray(environmentData.values)) ? environmentData.values : [], values = rawValues.map((v) => {
        return {
            key: v.key,
            value: v.value,
            enabled: v.enabled !== false,
            type: v.type || 'default'
        };
    });
    return resource.updateRaw({ values }, callback);
}
function checkCollectionExists(collectionId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(COLLECTION, collectionId, options.postmanApiKey, options.accessToken);
    return resource.exists(callback);
}
function checkEnvironmentExists(environmentId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(ENVIRONMENT, environmentId, options.postmanApiKey, options.accessToken);
    return resource.exists(callback);
}
function getWorkspaceCollections(workspaceId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    return PostmanResource.getWorkspaceCollections(workspaceId, options.postmanApiKey, options.accessToken, callback);
}
function getWorkspaceEnvironments(workspaceId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    return PostmanResource.getWorkspaceEnvironments(workspaceId, options.postmanApiKey, options.accessToken, callback);
}
function createCollection(collectionData, workspaceId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(COLLECTION, '', options.postmanApiKey, options.accessToken);
    return resource.createInWorkspace(collectionData, workspaceId, callback);
}
function createEnvironment(environmentData, workspaceId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(ENVIRONMENT, '', options.postmanApiKey, options.accessToken);
    return resource.createInWorkspace(environmentData, workspaceId, callback);
}
function deleteCollection(collectionId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(COLLECTION, collectionId, options.postmanApiKey, options.accessToken);
    return resource.delete(callback);
}
function deleteEnvironment(environmentId, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const resource = new PostmanResource(ENVIRONMENT, environmentId, options.postmanApiKey, options.accessToken);
    return resource.delete(callback);
}
function regenerateCollectionIds(collection) {
    const newCollection = JSON.parse(JSON.stringify(collection));
    newCollection.info._postman_id = (0, uuid_1.v4)();
    function regenerateItemIds(items) {
        if (!items || !Array.isArray(items)) {
            return;
        }
        items.forEach((item) => {
            if (item.id) {
                item.id = (0, uuid_1.v4)();
            }
            if (item.item && Array.isArray(item.item)) {
                regenerateItemIds(item.item);
            }
        });
    }
    if (newCollection.item && Array.isArray(newCollection.item)) {
        regenerateItemIds(newCollection.item);
    }
    return newCollection;
}
function regenerateEnvironmentId(environment) {
    const newEnvironment = JSON.parse(JSON.stringify(environment));
    newEnvironment.id = (0, uuid_1.v4)();
    return newEnvironment;
}
function isValidUuidV4(value) {
    if (typeof value !== 'string') {
        return false;
    }
    return (/^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i).test(value);
}
function ensureItemAndResponseIds(collection, forceAll = false) {
    const updated = JSON.parse(JSON.stringify(collection));
    let changes = 0;
    function fixItem(item) {
        if (forceAll || !isValidUuidV4(item.id)) {
            item.id = (0, uuid_1.v4)();
            changes++;
        }
        if (Array.isArray(item.response)) {
            item.response.forEach((resp) => {
                if (forceAll || !isValidUuidV4(resp.id)) {
                    resp.id = (0, uuid_1.v4)();
                    changes++;
                }
            });
        }
        if (Array.isArray(item.item)) {
            item.item.forEach((child) => { fixItem(child); });
        }
    }
    if (Array.isArray(updated.item)) {
        updated.item.forEach((it) => { fixItem(it); });
    }
    return { updated, changes };
}
async function validateCollections(collectionFiles, options, workspaceCollectionIds) {
    const statuses = [];
    console_1.default.log(safe_1.default.cyan('\nValidating collections...\n'));
    if (!collectionFiles || collectionFiles.length === 0) {
        console_1.default.log(safe_1.default.cyan('\nNo environments found...\n'));
        return statuses;
    }
    for (const filePath of collectionFiles) {
        const fileName = path_1.default.basename(filePath);
        try {
            let collection = await readJSONFile(filePath);
            const collectionId = collection.info?._postman_id, collectionName = collection.info?.name, resolvedName = collectionName || fileName, { changes } = ensureItemAndResponseIds(collection);
            if (changes > 0) {
                console_1.default.log(safe_1.default.yellow(`${fileName}: Detected ${changes} invalid/missing item/response ID(s) (will be fixed during prepare).`));
            }
            if (!collectionId || !isValidUuidV4(collectionId)) {
                console_1.default.log(safe_1.default.yellow(`${fileName}: No _postman_id found in collection, will generate new IDs.`));
                statuses.push({ filePath: filePath, name: resolvedName, needsNewId: true, nestedIdsFixed: changes });
                continue;
            }
            let existsGlobally = false;
            await new Promise((resolve) => {
                checkCollectionExists(collectionId, options, (err, exists) => {
                    if (err) {
                        console_1.default.error(safe_1.default.red(`${fileName}: Error checking collection: ${err.message}`));
                        return resolve(undefined);
                    }
                    existsGlobally = Boolean(exists);
                    resolve(undefined);
                });
            });
            if (!existsGlobally) {
                const msg = options.yes ?
                    `${fileName}: Collection not found on cloud,` +
                        ' skipping (--yes)' :
                    `${fileName}: Collection not found in Postman` +
                        ' (any workspace), will create new.';
                console_1.default.log(safe_1.default.yellow(msg));
                statuses.push({ filePath: filePath, name: resolvedName, needsNewId: false,
                    originalId: collectionId, nestedIdsFixed: changes });
                continue;
            }
            if (Array.isArray(workspaceCollectionIds)) {
                const existsInWorkspace = workspaceCollectionIds.some((wsId) => {
                    return wsId === collectionId || wsId.endsWith(`-${collectionId}`);
                });
                if (!existsInWorkspace) {
                    const msg = options.yes ?
                        `${fileName}: Collection not found on cloud,` +
                            ' skipping (--yes)' :
                        `${fileName}: Collection exists in another` +
                            ' workspace. Will create independent copy' +
                            ' with new IDs.';
                    console_1.default.log(safe_1.default.yellow(msg));
                    statuses.push({ filePath: filePath, name: resolvedName, needsNewId: true,
                        originalId: collectionId, nestedIdsFixed: changes });
                    continue;
                }
            }
            console_1.default.log(safe_1.default.green(`${fileName}: Collection exists in workspace, ready to push`));
            statuses.push({ filePath: filePath, name: resolvedName, needsNewId: false,
                originalId: collectionId, nestedIdsFixed: changes });
        }
        catch (err) {
            console_1.default.error(safe_1.default.red(`✗ ${fileName}: ${err.message}`));
            throw err;
        }
    }
    return statuses;
}
async function validateEnvironments(environmentFiles, options, workspaceEnvironmentIds) {
    const statuses = [];
    console_1.default.log(safe_1.default.cyan('\nValidating environments...\n'));
    if (!environmentFiles || environmentFiles.length === 0) {
        console_1.default.log(safe_1.default.cyan('\nNo environments found...\n'));
        return statuses;
    }
    for (const filePath of environmentFiles) {
        const fileName = path_1.default.basename(filePath);
        try {
            const environment = await (0, environment_1.readEnvironmentFile)(filePath), environmentId = environment.id, environmentName = environment.name, resolvedName = environmentName || fileName;
            if (!environmentId) {
                console_1.default.log(safe_1.default.yellow(`${fileName}: No id found in environment, will generate new IDs.`));
                statuses.push({ filePath: filePath, name: resolvedName, needsNewId: true });
                continue;
            }
            let existsGlobally = false;
            await new Promise((resolve) => {
                checkEnvironmentExists(environmentId, options, (err, exists) => {
                    if (err) {
                        console_1.default.error(safe_1.default.red(`✗ ${fileName}: Error checking environment: ${err.message}`));
                        return resolve(undefined);
                    }
                    existsGlobally = Boolean(exists);
                    resolve(undefined);
                });
            });
            if (!existsGlobally) {
                const msg = options.yes ?
                    `${fileName}: Environment not found on cloud,` +
                        ' skipping (--yes)' :
                    `${fileName}: Environment not found in Postman` +
                        ' (any workspace), will create new.';
                console_1.default.log(safe_1.default.yellow(msg));
                statuses.push({ filePath: filePath, name: resolvedName, needsNewId: false, originalId: environmentId });
                continue;
            }
            if (Array.isArray(workspaceEnvironmentIds)) {
                const existsInWorkspace = workspaceEnvironmentIds.some((wsId) => {
                    return wsId === environmentId || wsId.endsWith(`-${environmentId}`);
                });
                if (!existsInWorkspace) {
                    const msg = options.yes ?
                        `${fileName}: Environment not found on cloud,` +
                            ' skipping (--yes)' :
                        `${fileName}: Environment exists in another` +
                            ' workspace. Will create independent copy' +
                            ' with new ID.';
                    console_1.default.log(safe_1.default.yellow(msg));
                    statuses.push({ filePath: filePath, name: resolvedName, needsNewId: true, originalId: environmentId });
                    continue;
                }
            }
            console_1.default.log(safe_1.default.green(`${fileName}: Environment exists in workspace, ready to push.`));
            statuses.push({ filePath: filePath, name: resolvedName, needsNewId: false, originalId: environmentId });
        }
        catch (err) {
            console_1.default.error(safe_1.default.red(`✗ ${fileName}: ${err.message}`));
            throw err;
        }
    }
    return statuses;
}
async function regenerateCollections(statuses) {
    if (!statuses || statuses.length === 0) {
        return;
    }
    const toProcess = statuses.filter((s) => { return s.needsNewId || (s.nestedIdsFixed || 0) > 0; });
    if (toProcess.length === 0) {
        return;
    }
    console_1.default.log(safe_1.default.cyan(`\nRegenerating IDs for ${toProcess.length} collection(s)...`));
    function logCollectionRegeneration(fileName, status, nestedChanges) {
        if (status.needsNewId) {
            console_1.default.log(safe_1.default.green(`${fileName}: Updated collection ID.`));
            if (status.originalId) {
                console_1.default.log(safe_1.default.grey(`  Old ID: ${status.originalId} → New ID: ${status.newId}`));
            }
            else {
                console_1.default.log(safe_1.default.grey(`  New ID: ${status.newId}`));
            }
        }
        if (nestedChanges > 0) {
            console_1.default.log(safe_1.default.green(`${fileName}: Fixed ${nestedChanges} invalid/missing item/response ID(s).`));
        }
    }
    for (const status of toProcess) {
        const fileName = path_1.default.basename(status.filePath);
        try {
            const original = await readJSONFile(status.filePath), updated = JSON.parse(JSON.stringify(original));
            if (status.needsNewId) {
                const newId = (0, uuid_1.v4)();
                updated.info = updated.info || {};
                updated.info._postman_id = newId;
                status.newId = newId;
            }
            const nestedResult = ensureItemAndResponseIds(updated, Boolean(status.needsNewId)), wroteChanges = status.needsNewId || nestedResult.changes > 0;
            if (!wroteChanges) {
                continue;
            }
            await writeJSONFile(status.filePath, nestedResult.updated);
            logCollectionRegeneration(fileName, status, nestedResult.changes);
        }
        catch (err) {
            console_1.default.error(safe_1.default.red(`✗ ${fileName}: Failed to regenerate IDs: ${err.message}`));
            throw err;
        }
    }
}
async function regenerateEnvironments(statuses) {
    if (!statuses || statuses.length === 0) {
        return;
    }
    const toProcess = statuses.filter((s) => { return s.needsNewId; });
    if (toProcess.length === 0) {
        return;
    }
    console_1.default.log(safe_1.default.cyan(`\nRegenerating IDs for ${toProcess.length} environment(s)...\n`));
    for (const status of toProcess) {
        const fileName = path_1.default.basename(status.filePath);
        try {
            const environment = await readJSONFile(status.filePath), updated = regenerateEnvironmentId(environment), newId = updated.id;
            status.newId = newId;
            await writeJSONFile(status.filePath, updated);
            console_1.default.log(safe_1.default.green(`✓ ${fileName}: Generated new ID`));
            if (status.originalId) {
                console_1.default.log(safe_1.default.grey(`  Old ID: ${status.originalId} → New ID: ${newId}`));
            }
            else {
                console_1.default.log(safe_1.default.grey(`  New ID: ${newId}`));
            }
        }
        catch (err) {
            console_1.default.error(safe_1.default.red(`✗ ${fileName}: Failed to regenerate ID: ${err.message}`));
            throw err;
        }
    }
}
function getWorkspaceSpecifications(workspaceId, options, callback) {
    SpecificationService.getWorkspaceSpecifications(workspaceId, options)
        .then((specs) => { callback(null, specs); })
        .catch((err) => { callback(err); });
}
function createSpecification(specData, workspaceId, options, callback) {
    SpecificationService.createSpecification(specData, workspaceId, options)
        .then((result) => { callback(null, result); })
        .catch((err) => { callback(err); });
}
async function getSpecRootFilePath(specId, options) {
    const files = await SpecificationService.fetchSpecificationFiles(specId, options);
    const rootFile = files.find((f) => { return f.type === 'ROOT'; });
    return rootFile ? rootFile.path : null;
}
function updateSpecification(specId, filePath, content, options, callback) {
    SpecificationService.updateSpecificationFile(specId, filePath, content, options)
        .then((result) => { callback(null, result); })
        .catch((err) => { callback(err); });
}
function bulkModifySpecificationFiles(specId, payload, options) {
    return SpecificationService.bulkModifyFiles(specId, payload, options);
}
function fetchSpecificationFiles(specId, options) {
    return SpecificationService.fetchSpecificationFiles(specId, options);
}
function updateSpecificationFileAsync(specId, filePath, content, options) {
    return SpecificationService.updateSpecificationFile(specId, filePath, content, options);
}
function detectSpecType(content) {
    if (content.includes('openapi: 3') || content.includes('"openapi": "3') || content.includes("'openapi': '3")) {
        return 'OPENAPI:3.0';
    }
    if (content.includes('swagger: "2') || content.includes('"swagger": "2') || content.includes("'swagger': '2")) {
        return 'OPENAPI:2.0';
    }
    return 'OPENAPI:3.0';
}
async function readSpecFile(filePath) {
    try {
        const content = await promises_1.default.readFile(filePath, 'utf8'), type = detectSpecType(content);
        return { content, type };
    }
    catch (err) {
        throw new Error(`Failed to read specification file ${filePath}: ${err.message}`);
    }
}
function normalizeSpecPath(specPath) {
    return specPath.replace(/^\//, '').replace(/\\/g, '/');
}
async function validateSpecifications(specificationFiles, options, workspaceSpecs = []) {
    const localSpecs = new Map(), cloudSpecMap = new Map(), projectRoot = process.cwd();
    console_1.default.log(safe_1.default.cyan('\nValidating specifications...\n'));
    if (!specificationFiles || specificationFiles.length === 0) {
        console_1.default.log(safe_1.default.grey('No specifications found\n'));
        return { localSpecs, cloudSpecMap };
    }
    for (const filePath of specificationFiles) {
        const fileName = path_1.default.basename(filePath);
        try {
            const { content, type } = await readSpecFile(filePath), relativeToProjectRoot = path_1.default.relative(projectRoot, filePath), normalizedPath = normalizeSpecPath(relativeToProjectRoot);
            let specName = path_1.default.basename(filePath, path_1.default.extname(filePath));
            try {
                const yaml = require('js-yaml'), parsed = yaml.load(content);
                if (parsed && parsed.info && parsed.info.title) {
                    specName = parsed.info.title;
                }
            }
            catch (err) {
                specName = path_1.default.basename(filePath, path_1.default.extname(filePath));
            }
            localSpecs.set(normalizedPath, {
                name: specName,
                type: type,
                content: content,
                filePath: filePath
            });
            console_1.default.log(safe_1.default.green(`${fileName}: Valid specification file`));
        }
        catch (err) {
            console_1.default.error(safe_1.default.red(`${fileName}: ${err.message}`));
            throw err;
        }
    }
    if (workspaceSpecs.length > 0) {
        for (const cloudSpec of workspaceSpecs) {
            if (!cloudSpec.name) {
                console_1.default.log(safe_1.default.yellow(`Skipping spec with ID ${cloudSpec.id || 'unknown'}: missing name`));
                continue;
            }
            if (cloudSpec.rootPath) {
                const normalizedRootPath = normalizeSpecPath(cloudSpec.rootPath);
                cloudSpecMap.set(normalizedRootPath, {
                    specId: cloudSpec.id,
                    name: cloudSpec.name,
                    type: cloudSpec.type,
                    rootPath: cloudSpec.rootPath
                });
            }
            else {
                console_1.default.log(safe_1.default.yellow(`  Cloud spec: "${cloudSpec.name}" has no root path, skipping`));
            }
        }
    }
    return { localSpecs, cloudSpecMap };
}
function printPrepareSummary(collectionStatuses, environmentStatuses) {
    const collectionsReady = collectionStatuses.filter((s) => { return !s.needsNewId || s.newId; }).length, environmentsReady = environmentStatuses.filter((s) => { return !s.needsNewId || s.newId; }).length, totalReady = collectionsReady + environmentsReady, collectionsRegenerated = collectionStatuses.filter((s) => { return s.needsNewId && s.newId; }).length, environmentsRegenerated = environmentStatuses.filter((s) => { return s.needsNewId && s.newId; }).length, totalRegenerated = collectionsRegenerated + environmentsRegenerated;
    console_1.default.log(safe_1.default.cyan('\n═══════════════════════════════════════════════════════════'));
    console_1.default.log(safe_1.default.cyan('                     PREPARE SUMMARY'));
    console_1.default.log(safe_1.default.cyan('═══════════════════════════════════════════════════════════\n'));
    if (totalRegenerated > 0) {
        console_1.default.log(safe_1.default.yellow(`Regenerated IDs: ${totalRegenerated} (${collectionsRegenerated} collections, ${environmentsRegenerated} environments)`));
    }
    console_1.default.log(safe_1.default.green(`Ready to push: ${totalReady} entities (${collectionsReady} collections, ${environmentsReady} environments)`));
    console_1.default.log(safe_1.default.cyan('\n───────────────────────────────────────────────────────────\n'));
}
async function discoverSDKs(workingDir) {
    const baseDir = workingDir || process.cwd(), sdksDir = path_1.default.join(baseDir, 'postman', 'sdks'), discoveredSDKs = [];
    try {
        await promises_1.default.access(sdksDir);
    }
    catch {
        return [];
    }
    function mapEntityTypeToInternal(metadataEntityType) {
        if (metadataEntityType === 'collections') {
            return 'collection';
        }
        return 'specification';
    }
    async function processCategoryDirectory(categoryDir, sdkList) {
        const sdkNameDirs = await promises_1.default.readdir(categoryDir, { withFileTypes: true });
        for (const sdkNameDir of sdkNameDirs) {
            if (!sdkNameDir.isDirectory()) {
                continue;
            }
            const sdkNamePath = path_1.default.join(categoryDir, sdkNameDir.name);
            await processSDKNameDirectory(sdkNamePath, sdkNameDir.name, sdkList);
        }
    }
    async function resolveCloudEntityId(metadata, sdkNameDirName) {
        const resourcesConfig = await (0, resources_1.readResourcesConfig)(baseDir);
        const cloudResources = resourcesConfig.cloudResources?.[metadata.entityType];
        if (!cloudResources) {
            throw new Error(`No cloudResources.${metadata.entityType} found in .postman/resources.yaml`);
        }
        if (metadata.entityType === 'collections') {
            for (const [pathKey, cloudId] of Object.entries(cloudResources)) {
                if (path_1.default.basename(pathKey) === sdkNameDirName) {
                    return cloudId;
                }
            }
            throw new Error(`No cloud ID found in .postman/resources.yaml for collection "${sdkNameDirName}"`);
        }
        else if (metadata.entityType === 'specs') {
            let specAbsPath;
            try {
                specAbsPath = Buffer.from(metadata.entityId, 'base64').toString('utf8');
            }
            catch {
                throw new Error(`Failed to decode spec entityId for "${sdkNameDirName}"`);
            }
            const configRoot = path_1.default.join(baseDir, '.postman');
            const relSpecPath = path_1.default.relative(configRoot, specAbsPath).split(path_1.default.sep).join('/');
            for (const [pathKey, cloudId] of Object.entries(cloudResources)) {
                const normalizedKey = pathKey.split(path_1.default.sep).join('/');
                if (normalizedKey === relSpecPath) {
                    return cloudId;
                }
            }
            throw new Error(`No cloud ID found in .postman/resources.yaml for spec "${sdkNameDirName}" (path: ${relSpecPath})`);
        }
        else {
            throw new Error(`Unsupported entity type "${metadata.entityType}"`);
        }
    }
    async function processSDKNameDirectory(sdkNamePath, sdkNameDirName, sdkList) {
        const metadataPath = path_1.default.join(sdkNamePath, 'metadata.json');
        let metadata;
        try {
            const metadataContent = await promises_1.default.readFile(metadataPath, 'utf8');
            metadata = JSON.parse(metadataContent);
            if (!metadata.entityType || (metadata.entityType === 'specs' && !metadata.entityId)) {
                console_1.default.log(safe_1.default.yellow(`Skipping ${sdkNameDirName}: Invalid metadata.json structure`));
                return;
            }
            if (metadata.entityType !== 'collections' && metadata.entityType !== 'specs') {
                console_1.default.log(safe_1.default.yellow(`Skipping ${sdkNameDirName}: Unsupported entityType "${metadata.entityType}" in metadata.json. Expected "collections" or "specs".`));
                return;
            }
        }
        catch (err) {
            console_1.default.log(safe_1.default.yellow(`Skipping ${sdkNameDirName}: Cannot read metadata.json - ${err.message}`));
            return;
        }
        let cloudEntityId;
        try {
            cloudEntityId = await resolveCloudEntityId(metadata, sdkNameDirName);
        }
        catch (err) {
            console_1.default.log(safe_1.default.yellow(`Skipping ${sdkNameDirName}: ${err.message}`));
            return;
        }
        const languageDirs = await promises_1.default.readdir(sdkNamePath, { withFileTypes: true });
        for (const languageDir of languageDirs) {
            if (!languageDir.isDirectory() || languageDir.name === 'metadata.json') {
                continue;
            }
            const language = languageDir.name;
            const sdkPath = path_1.default.join(sdkNamePath, language);
            const entityName = sdkNameDirName;
            const internalEntityType = mapEntityTypeToInternal(metadata.entityType);
            sdkList.push({
                sdkDir: sdkPath,
                language: language,
                entityType: internalEntityType,
                entityId: cloudEntityId,
                entityName: entityName
            });
        }
    }
    try {
        const entries = await promises_1.default.readdir(sdksDir, { withFileTypes: true });
        for (const entry of entries) {
            if (!entry.isDirectory()) {
                continue;
            }
            if (entry.name !== 'Collections SDKs' && entry.name !== 'Specs SDKs') {
                continue;
            }
            const categoryDir = path_1.default.join(sdksDir, entry.name);
            await processCategoryDirectory(categoryDir, discoveredSDKs);
        }
    }
    catch (err) {
        throw new Error(`Failed to discover SDKs: ${err.message}`);
    }
    return discoveredSDKs;
}
function loadGitignorePatterns(directoryPath) {
    const ignore = require('ignore');
    const fsSync = require('fs');
    const gitignorePath = path_1.default.join(directoryPath, '.gitignore');
    const ig = ignore();
    if (fsSync.existsSync(gitignorePath)) {
        const gitignoreContent = fsSync.readFileSync(gitignorePath, 'utf-8');
        ig.add(gitignoreContent);
    }
    return ig;
}
function shouldIgnorePath(relativePath, igInstance, isDirectory) {
    let normalizedPath = relativePath.split(path_1.default.sep).join('/');
    if (isDirectory && !normalizedPath.endsWith('/')) {
        normalizedPath += '/';
    }
    if (isDirectory) {
        return igInstance.ignores(normalizedPath) || igInstance.ignores(normalizedPath.replace(/\/$/, ''));
    }
    return igInstance.ignores(normalizedPath);
}
async function zipDirectory(sourceDir, outputPath) {
    const AdmZip = require('adm-zip');
    const fsSync = require('fs');
    const zip = new AdmZip();
    const ig = loadGitignorePatterns(sourceDir);
    async function addDirectoryToZip(currentPath, zipPath = '') {
        const entries = await promises_1.default.readdir(currentPath, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path_1.default.join(currentPath, entry.name);
            const relativePath = path_1.default.relative(sourceDir, fullPath);
            const zipEntryPath = zipPath ? path_1.default.join(zipPath, entry.name) : entry.name;
            if (shouldIgnorePath(relativePath, ig, entry.isDirectory())) {
                continue;
            }
            if (entry.isDirectory()) {
                await addDirectoryToZip(fullPath, zipEntryPath);
            }
            else {
                const fileContent = fsSync.readFileSync(fullPath);
                zip.addFile(zipEntryPath, fileContent);
            }
        }
    }
    await addDirectoryToZip(sourceDir);
    zip.writeZip(outputPath);
}
async function computeSdkHash(directoryPath) {
    const { hashElement } = require('folder-hash');
    const ig = loadGitignorePatterns(directoryPath);
    const options = {
        folders: {
            exclude: (folderPath) => {
                let normalizedPath = folderPath.split(path_1.default.sep).join('/');
                if (!normalizedPath.endsWith('/')) {
                    normalizedPath += '/';
                }
                return ig.ignores(normalizedPath) || ig.ignores(folderPath.split(path_1.default.sep).join('/'));
            },
            ignoreRootName: true
        },
        files: {
            exclude: (fileRelativePath) => {
                const normalizedPath = fileRelativePath.split(path_1.default.sep).join('/');
                return ig.ignores(normalizedPath);
            }
        },
        algo: 'sha256',
        encoding: 'hex'
    };
    const result = await hashElement(directoryPath, options);
    return result.hash;
}
async function populateSDKCloudHashes(sdks, options) {
    if (!options.postmanApiKey && !options.accessToken) {
        return;
    }
    const { createSDKGeneratorService } = require('../../api/sdk-generator-service');
    const sdkService = createSDKGeneratorService({
        postmanApiKey: options.postmanApiKey,
        accessToken: options.accessToken
    });
    const entitiesMap = new Map();
    for (const sdk of sdks) {
        const entityKey = `${sdk.entityType}:${sdk.entityId}`;
        const entitySDKs = entitiesMap.get(entityKey);
        if (!entitySDKs) {
            entitiesMap.set(entityKey, [sdk]);
        }
        else {
            entitySDKs.push(sdk);
        }
    }
    for (const entitySDKs of entitiesMap.values()) {
        const firstSDK = entitySDKs[0];
        const collectionId = firstSDK.entityType === 'collection' ? firstSDK.entityId : undefined;
        const specificationId = firstSDK.entityType === 'specification' ? firstSDK.entityId : undefined;
        if (!collectionId && !specificationId) {
            continue;
        }
        try {
            const { buildSummary } = await sdkService.getBuildSummary(collectionId, specificationId);
            if (!buildSummary || !buildSummary.downloads) {
                continue;
            }
            for (const sdk of entitySDKs) {
                const downloadInfo = buildSummary.downloads.find((download) => {
                    return download.language === sdk.language;
                });
                if (downloadInfo?.contentHash) {
                    sdk.cloudContentHash = downloadInfo.contentHash;
                }
            }
        }
        catch {
            console_1.default.log(safe_1.default.yellow(`  ⚠ Could not check cloud SDK(s) for ${firstSDK.entityName}, ` +
                'will upload all languages for consistency.'));
        }
    }
}
async function uploadSDK(sdk, sdkService, collectionId, specificationId) {
    const osModule = require('os'), pathModule = require('path'), fsModule = require('fs');
    const tmpDir = osModule.tmpdir();
    const zipFileName = `${sdk.entityName}-${sdk.language}-${Date.now()}.zip`;
    const zipFilePath = pathModule.join(tmpDir, zipFileName);
    try {
        await zipDirectory(sdk.sdkDir, zipFilePath);
        await sdkService.uploadSDK(zipFilePath, sdk.language, collectionId, specificationId);
        try {
            fsModule.unlinkSync(zipFilePath);
        }
        catch {
        }
    }
    catch (err) {
        try {
            fsModule.unlinkSync(zipFilePath);
        }
        catch {
        }
        throw err;
    }
}
function syncSDK(sdk, options, callback) {
    if (!options.postmanApiKey && !options.accessToken) {
        return callback(new Error('No authorization data found. Please use the `postman login` command.'));
    }
    const { createSDKGeneratorService } = require('../../api/sdk-generator-service');
    const sdkService = createSDKGeneratorService({
        postmanApiKey: options.postmanApiKey,
        accessToken: options.accessToken
    });
    const collectionId = sdk.entityType === 'collection' ? sdk.entityId : undefined;
    const specificationId = sdk.entityType === 'specification' ? sdk.entityId : undefined;
    (async () => {
        try {
            if (sdk.cloudContentHash) {
                const localHash = await computeSdkHash(sdk.sdkDir);
                if (localHash === sdk.cloudContentHash) {
                    return callback(null, { uploaded: false });
                }
            }
            await uploadSDK(sdk, sdkService, collectionId, specificationId);
            return callback(null, { uploaded: true });
        }
        catch (err) {
            const isHashError = err.message && err.message.includes('hash');
            if (isHashError) {
                console_1.default.log(safe_1.default.yellow(`  ⚠ Could not verify SDK hash for ${sdk.entityName} (${sdk.language}), uploading anyway: ${err.message}`));
                try {
                    await uploadSDK(sdk, sdkService, collectionId, specificationId);
                    return callback(null, { uploaded: true });
                }
                catch (uploadErr) {
                    return callback(uploadErr);
                }
            }
            return callback(err);
        }
    })();
}
