// @postman/pm-bin-linux-x64/bin/lib/framework/feature-flags/index.js
const _ = require('lodash'), requestFactory = require('@postman/platform-req/lib/request'), cliEnvironment = require('../../config/cli-environment'), rcfile = require('../../config/rc-file'), console = require('../console');
let flagsCache = null, initPromise = null, authenticated = false;
function getClient() {
    return requestFactory.defaults({
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                POST: { enabled: true }
            }
        }
    });
}
function getAuthHeaders(profile) {
    const headers = {
        'Content-Type': 'application/json',
        'x-pstmn-req-service': 'features'
    };
    if (profile.accessToken) {
        headers['x-access-token'] = profile.accessToken;
    }
    else if (profile.postmanApiKey) {
        headers['X-Api-Key'] = profile.postmanApiKey;
    }
    return headers;
}
function getDefaultProfile() {
    try {
        const data = rcfile.loadSync(['home']), profiles = _.get(data, 'login._profiles', []), profile = _.find(profiles, ['alias', 'default']);
        return profile || null;
    }
    catch (e) {
        return null;
    }
}
function init(callback) {
    if (!callback) {
        return new Promise((resolve, reject) => {
            init((err) => { return err ? reject(err) : resolve(); });
        });
    }
    if (flagsCache) {
        return callback(null);
    }
    if (initPromise) {
        return initPromise.then(() => { return callback(null); });
    }
    const settings = cliEnvironment.getSettings(), featureNames = settings.enableFeatureFlags, profile = getDefaultProfile();
    if (!featureNames || !featureNames.length) {
        flagsCache = {};
        return callback(null);
    }
    if (!profile || !profile.userId || (!profile.accessToken && !profile.postmanApiKey)) {
        flagsCache = {};
        authenticated = false;
        return callback(null);
    }
    authenticated = true;
    const util = require('../../util'), gatewayBaseUrl = util.POSTMAN_GATEWAY_BASE_URL(), url = `${gatewayBaseUrl}/features/list?entityType=user&entityValue=${profile.userId}`, client = getClient(), requestOptions = {
        url: url,
        method: 'POST',
        headers: getAuthHeaders(profile),
        body: JSON.stringify({ features: featureNames }),
        timeout: 10000
    };
    initPromise = new Promise((resolve) => {
        function done(parsed) {
            flagsCache = parsed;
            initPromise = null;
            resolve();
            callback(null);
        }
        client(requestOptions, (err, response, body) => {
            if (err) {
                console.warn(`Feature flags: failed to fetch — ${err.message || err}`);
                return done({});
            }
            try {
                if (typeof body === 'string') {
                    body = JSON.parse(body);
                }
            }
            catch (e) {
                console.warn('Feature flags: invalid JSON response');
                return done({});
            }
            if (!response || response.statusCode >= 400) {
                console.warn(`Feature flags: HTTP ${response && response.statusCode}`);
                return done({});
            }
            const features = body && body.data && body.data.features, parsed = {};
            if (!features || typeof features !== 'object') {
                return done({});
            }
            Object.keys(features).forEach((name) => {
                parsed[name] = features[name].value;
            });
            return done(parsed);
        });
    });
}
function get(featureName) {
    if (!flagsCache) {
        return undefined;
    }
    return flagsCache[featureName];
}
function getAll() {
    return flagsCache;
}
function isEnabled(featureName) {
    return get(featureName) === true;
}
function isLoggedIn() {
    return authenticated;
}
function clearCache() {
    flagsCache = null;
    initPromise = null;
    authenticated = false;
}
module.exports = {
    init: init,
    get: get,
    getAll: getAll,
    isEnabled: isEnabled,
    isLoggedIn: isLoggedIn,
    clearCache: clearCache
};
