// @postman/pm-bin-linux-x64/bin/lib/commands/flows/flows-trigger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print = __importStar(require("../../framework/print"));
const utils_1 = require("./cmd/utils");
const handler_1 = require("./cmd/handler");
const zod_1 = require("zod");
const client_1 = __importDefault(require("./api/client"));
const axios_1 = __importDefault(require("axios"));
const input_functions_1 = require("./run/input-functions");
const chalk_1 = __importDefault(require("chalk"));
function redactToken(token, showSecrets) {
    return showSecrets ? token : (token.length >= 3 ? `${token.slice(0, 3)}*****` : '*****');
}
const TriggerFlowCommandSchema = zod_1.z.object({
    verbose: zod_1.z.boolean().default(false),
    debug: zod_1.z.boolean().default(false),
    json: zod_1.z.boolean().default(false),
    flow: zod_1.z.string().min(1, 'Flow ID is required'),
    input: zod_1.z.array(zod_1.z.object({ key: zod_1.z.string(), value: zod_1.z.string() })).default([]),
    inputFile: zod_1.z.array(zod_1.z.string()).default([]),
    headers: zod_1.z.array(zod_1.z.object({ key: zod_1.z.string(), value: zod_1.z.string() })).default([]),
    query: zod_1.z.array(zod_1.z.object({ key: zod_1.z.string(), value: zod_1.z.string() })).default([]),
    dryRun: zod_1.z.boolean().default(false),
    showSecrets: zod_1.z.boolean().default(false),
    result: zod_1.z.boolean().default(false)
});
async function getFlowTarget(command, auth) {
    try {
        const targetResponse = await client_1.default.getTarget({
            accessToken: auth.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            flowId: command.flow
        });
        return targetResponse;
    }
    catch (error) {
        if (error.message.includes('404') || error.message.includes('not found')) {
            throw new Error(`Flow ${command.flow} is not deployed. ` +
                'Please deploy this flow first before triggering it.');
        }
        throw error;
    }
}
function buildInputPayload(command) {
    const payload = {};
    if (command.inputFile.length > 0) {
        const fileInputs = (0, input_functions_1.processInputFiles)(command.inputFile);
        Object.assign(payload, fileInputs);
    }
    command.input.forEach(({ key, value }) => {
        payload[key] = value;
    });
    return payload;
}
async function triggerFlowRequest(url, payload, authToken, command) {
    const headers = {
        'Content-Type': 'application/json'
    };
    if (authToken) {
        headers.Authorization = `Bearer ${authToken}`;
    }
    if (command?.headers) {
        command.headers.forEach(({ key, value }) => {
            headers[key] = value;
        });
    }
    let requestUrl = url;
    if (command?.query && command.query.length > 0) {
        const queryParams = new URLSearchParams();
        command.query.forEach(({ key, value }) => {
            queryParams.append(key, value);
        });
        requestUrl = `${url}?${queryParams.toString()}`;
    }
    if (command?.verbose) {
        print.lf(`Making request to: ${requestUrl}`);
        print.lf(`Payload: ${JSON.stringify(payload, null, 2)}`);
        if (authToken) {
            print.lf('Using authentication token');
        }
    }
    try {
        const response = await axios_1.default.post(requestUrl, payload, { headers });
        return response;
    }
    catch (error) {
        if (error.response) {
            return error.response;
        }
        else if (error.request) {
            throw new Error(`Flow trigger failed: No response received from ${requestUrl}`);
        }
        else {
            throw new Error(`Flow trigger failed: ${error.message}`);
        }
    }
}
async function triggerFlow(command) {
    try {
        const auth = await (0, utils_1.getAuthForCommand)(command, 'flows');
        if (command.verbose) {
            print.lf(`Getting deployment information for flow ${command.flow}...`);
        }
        const targetResponse = await getFlowTarget(command, auth);
        const target = targetResponse.target;
        if (!target.deployment) {
            throw new Error(`Flow ${command.flow} has a target but is not deployed. ` +
                'Please deploy this flow first before triggering it.');
        }
        const url = target.url;
        if (!url || !url.basePath || !url.namespace) {
            throw new Error(`Flow ${command.flow} does not have a valid URL`);
        }
        const webhookUrl = `${url.basePath}${url.namespace}/${url.path}`;
        if (command.verbose) {
            print.lf(`Target URL: ${webhookUrl}`);
        }
        if (target.enabled === false) {
            throw new Error('Flow target is disabled. Please enable the target before triggering the flow.');
        }
        const payload = buildInputPayload(command);
        let authToken;
        const enabledAuthConfig = target.auth &&
            Object.values(target.auth).find((auth) => {
                return auth.enabled && auth.encryptedAuthToken;
            });
        if (enabledAuthConfig) {
            if (command.verbose) {
                print.lf('Flow has authentication enabled, decrypting auth token...');
            }
            try {
                const decryptedAuthResponse = await client_1.default.decryptAuth({
                    accessToken: auth.accessToken,
                    service: 'flows',
                    verbose: command.verbose,
                    debug: command.debug,
                    targetId: target.id,
                    encryptedAuthToken: enabledAuthConfig.encryptedAuthToken
                });
                authToken = decryptedAuthResponse.token;
                if (command.verbose) {
                    print.lf('Auth token retrieved successfully');
                }
            }
            catch (error) {
                throw new Error(`Failed to decrypt auth token for flow: ${error.message}`);
            }
        }
        else if (command.verbose) {
            print.lf('Flow does not have authentication enabled');
        }
        if (command.dryRun) {
            let finalUrl = webhookUrl;
            if (command.query && command.query.length > 0) {
                const queryParams = new URLSearchParams();
                command.query.forEach(({ key, value }) => {
                    queryParams.append(key, value);
                });
                finalUrl = `${webhookUrl}?${queryParams.toString()}`;
            }
            print.lf('Dry run triggering the flow (no request will be made)');
            print.lf(`Target URL: ${finalUrl}`);
            if (authToken) {
                print.lf('Authentication: Bearer token (enabled)');
            }
            else {
                print.lf('Authentication: None');
            }
            print.lf('Headers:');
            print.lf('  Content-Type: application/json');
            if (authToken) {
                const displayToken = redactToken(authToken, command.showSecrets);
                print.lf(`  Authorization: Bearer ${displayToken}`);
            }
            if (command.headers) {
                command.headers.forEach(({ key, value }) => {
                    print.lf(`  ${key}: ${value}`);
                });
            }
            if (Object.keys(payload).length > 0) {
                print.lf('Payload:');
                const payloadLines = JSON.stringify(payload, null, 2).split('\n');
                payloadLines.forEach((line) => {
                    print.lf(line);
                });
            }
            else {
                print.lf('Payload: (empty)');
            }
            return {
                flowId: command.flow,
                webhookUrl: finalUrl,
                payload,
                dryRun: true
            };
        }
        const response = await triggerFlowRequest(webhookUrl, payload, authToken, command);
        return {
            flowId: command.flow,
            webhookUrl,
            payload,
            response: {
                status: response.status,
                statusText: response.statusText,
                headers: response.headers,
                data: response.data
            }
        };
    }
    catch (error) {
        throw new Error(`Failed to trigger flow: ${error.message}`);
    }
}
async function runCommand(command) {
    try {
        const result = await triggerFlow(command);
        if (command.json) {
            print.lf(JSON.stringify(result, null, 4));
            return;
        }
        if (result.dryRun) {
            return;
        }
        const isSuccess = result.response.status === 200;
        const runId = result.response.headers?.['x-run-id'];
        const statusLine = `${result.response.status} ${result.response.statusText}`;
        const coloredStatus = isSuccess ? chalk_1.default.green(statusLine) : chalk_1.default.red(statusLine);
        if (command.result) {
            if (result.response.data) {
                const responseStr = typeof result.response.data === 'string' ?
                    result.response.data :
                    JSON.stringify(result.response.data, null, 2);
                print.lf(responseStr);
            }
        }
        else {
            if (runId) {
                print.lf(`Run ID: ${runId} (${coloredStatus})`);
            }
            else {
                print.lf(`Status: ${coloredStatus}`);
            }
            if (command.verbose) {
                print.lf(`  Flow ID: ${result.flowId}`);
                print.lf(`  Target URL: ${result.webhookUrl}`);
                if (Object.keys(result.payload).length > 0) {
                    print.lf(`  Input Payload: ${JSON.stringify(result.payload, null, 2)}`);
                }
                print.lf('');
                print.lf('Headers:');
                if (result.response.headers) {
                    Object.entries(result.response.headers).forEach(([key, value]) => {
                        print.lf(`  ${key}: ${value}`);
                    });
                }
                else {
                    print.lf('  (no headers)');
                }
                print.lf('');
                print.lf('Response:');
            }
            if (result.response.data) {
                const responseStr = typeof result.response.data === 'string' ?
                    result.response.data :
                    JSON.stringify(result.response.data, null, 2);
                print.lf(command.verbose ? `  ${responseStr}` : responseStr);
            }
            else if (command.verbose) {
                print.lf('  (empty response)');
            }
        }
        if (!isSuccess) {
            process.exit(1);
        }
    }
    catch (error) {
        print.lf(error.message);
        process.exit(1);
    }
}
exports.default = {
    validateCommand: (input) => { return (0, handler_1.validateCommand)(input, TriggerFlowCommandSchema); },
    runCommand: runCommand
};
