// @postman/pm-bin-linux-x64/bin/lib/commands/flows/flows-list.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print = __importStar(require("../../framework/print"));
const cli_table3_1 = __importDefault(require("cli-table3"));
const client_1 = __importDefault(require("./api/client"));
const utils_1 = require("./cmd/utils");
const handler_1 = require("./cmd/handler");
const zod_1 = require("zod");
const config = __importStar(require("../../config"));
const cmdUtil = __importStar(require("../cmd-util"));
const libUtil = __importStar(require("../../util"));
const constants_1 = require("../../constants");
const safe_1 = __importDefault(require("colors/safe"));
const ListFlowsCommandSchema = zod_1.z.object({
    verbose: zod_1.z.boolean().default(false),
    debug: zod_1.z.boolean().default(false),
    json: zod_1.z.boolean().default(false),
    workspace: zod_1.z.string().min(1, 'Workspace ID is required'),
    paginate: zod_1.z.boolean().default(false),
    filter: zod_1.z.string().optional(),
    sort: zod_1.z.enum(['name', 'updated']).default('updated')
});
function formatFlowsTable(flows) {
    const table = new cli_table3_1.default({
        chars: {
            top: '', 'top-mid': '', 'top-left': '', 'top-right': '',
            bottom: '', 'bottom-mid': '', 'bottom-left': '', 'bottom-right': '',
            left: '', 'left-mid': '', mid: '', 'mid-mid': '',
            right: '', 'right-mid': '', middle: ' '
        },
        colAligns: ['left', 'left', 'left'],
        colWidths: [20, 25, 80],
        style: {
            'padding-left': 0,
            'padding-right': 0,
            head: [],
            border: []
        }
    });
    table.push([
        'Last Updated',
        'ID',
        'Name'
    ]);
    flows.forEach((flow) => {
        const lastUpdated = flow.updatedAt ? (0, utils_1.formatDateTime)(new Date(flow.updatedAt)) : 'N/A';
        const displayName = flow.isDeployed ?
            `${flow.name} ${safe_1.default.green('●')}` :
            flow.name;
        table.push([
            lastUpdated,
            flow.id,
            displayName
        ]);
    });
    return table;
}
function sortFlows(flows, sortBy = 'updated') {
    return [...flows].sort((a, b) => {
        if (sortBy === 'name') {
            return a.name.localeCompare(b.name, undefined, { sensitivity: 'base' });
        }
        const dateA = a.updatedAt ? new Date(a.updatedAt).getTime() : 0;
        const dateB = b.updatedAt ? new Date(b.updatedAt).getTime() : 0;
        return dateB - dateA;
    });
}
async function displayFlowsPaginated(flows, pageSize = 25) {
    let currentIndex = 0;
    while (currentIndex < flows.length) {
        const endIndex = Math.min(currentIndex + pageSize, flows.length), currentPage = flows.slice(currentIndex, endIndex);
        print.lf(formatFlowsTable(currentPage).toString());
        currentIndex = endIndex;
        if (currentIndex < flows.length) {
            print.lf(`\nShowing ${Math.min(currentIndex, flows.length)} of ${flows.length} flows`);
            const continueViewing = await (0, utils_1.promptForMore)('flows');
            if (!continueViewing) {
                break;
            }
            print.lf('');
        }
    }
    if (currentIndex >= flows.length) {
        print.lf(`\nShowing all ${flows.length} flows`);
    }
}
async function getFlowsList(command) {
    try {
        const options = await new Promise((resolve, reject) => {
            config.get(command, { command: 'flows' }, (err, opts) => {
                if (err) {
                    return reject(new Error(`Error getting configuration: ${err.message || err}`));
                }
                const mergedOptions = { ...opts, ...cmdUtil.commanderToObject(command) };
                libUtil.extractPostmanApiKey(mergedOptions.postmanApiKey, (extractErr, postmanApiKey) => {
                    if (!postmanApiKey && !mergedOptions.accessToken) {
                        return reject(new Error(constants_1.MESSAGES.NO_AUTHORIZATION_DATA_WITH_REFERENCE));
                    }
                    mergedOptions.postmanApiKey = postmanApiKey;
                    mergedOptions.accessToken = postmanApiKey || mergedOptions.accessToken;
                    resolve(mergedOptions);
                });
            });
        });
        const flowsOptions = {
            accessToken: options.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            workspace: command.workspace
        };
        const flows = await client_1.default.listFlows(flowsOptions);
        return flows;
    }
    catch (error) {
        throw new Error(`Failed to list flows: ${error.message}`);
    }
}
async function runCommand(command) {
    try {
        let flows = await getFlowsList(command);
        flows = sortFlows(flows, command.sort);
        if (command.filter) {
            try {
                flows = (0, utils_1.filterByName)(flows, command.filter);
                if (flows.length === 0 && command.verbose) {
                    print.lf('No flows found.');
                    return;
                }
            }
            catch (filterError) {
                print.lf(filterError.message);
                process.exit(1);
            }
        }
        if (command.paginate) {
            await displayFlowsPaginated(flows);
            return;
        }
        if (command.json) {
            print.lf(JSON.stringify(flows, null, 4));
            return;
        }
        print.lf(formatFlowsTable(flows).toString());
        if (command.verbose) {
            print.lf(`\nShowing all ${flows.length} flows`);
        }
    }
    catch (error) {
        print.lf(error.message);
        process.exit(1);
    }
}
exports.default = {
    validateCommand: (input) => { return (0, handler_1.validateCommand)(input, ListFlowsCommandSchema); },
    runCommand: runCommand
};
