// @postman/pm-bin-linux-x64/bin/lib/commands/flows/flows-list-runs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTimeRange = parseTimeRange;
const print = __importStar(require("../../framework/print"));
const cli_table3_1 = __importDefault(require("cli-table3"));
const client_1 = __importDefault(require("./api/client"));
const utils_1 = require("./cmd/utils");
const handler_1 = require("./cmd/handler");
const zod_1 = require("zod");
const ListRunsCommandSchema = zod_1.z.object({
    verbose: zod_1.z.boolean().default(false),
    debug: zod_1.z.boolean().default(false),
    json: zod_1.z.boolean().default(false),
    workspace: zod_1.z.string().min(1, 'Workspace ID is required'),
    flow: zod_1.z.string().optional(),
    range: zod_1.z.object({
        startMs: zod_1.z.number(),
        endMs: zod_1.z.number()
    })
});
function formatRunsTable(response) {
    const table = new cli_table3_1.default({
        chars: {
            top: '', 'top-mid': '', 'top-left': '', 'top-right': '',
            bottom: '', 'bottom-mid': '', 'bottom-left': '', 'bottom-right': '',
            left: '', 'left-mid': '', mid: '', 'mid-mid': '',
            right: '', 'right-mid': '', middle: ' '
        },
        colAligns: ['left', 'left', 'left', 'left', 'left'],
        colWidths: [25, 34, 10, 10, 50],
        style: {
            'padding-left': 0,
            'padding-right': 0,
            head: [],
            border: []
        }
    });
    table.push([
        'Timestamp',
        'Run ID',
        'Duration',
        'Status',
        'Flow'
    ]);
    const runs = [...response.history];
    runs.sort((a, b) => {
        return (b.start_time_ms || 0) - (a.start_time_ms || 0);
    });
    runs.forEach((run) => {
        const timestamp = new Date(run.start_time_ms).toISOString();
        const runId = run.session_id || 'N/A';
        const flowName = response.meta.actions[run.action_id]?.name || 'Unknown';
        const duration = run.duration_ms ? `${run.duration_ms}ms` : 'N/A';
        const status = run.status || 'Unknown';
        table.push([
            timestamp,
            runId,
            duration,
            status,
            flowName
        ]);
    });
    return table;
}
function parseTimeRange(range) {
    const now = new Date();
    const endMs = now.getTime();
    const match = range.match(/^(\d+)([mhd])$/);
    if (!match) {
        throw new Error('Invalid range format. Use formats like 30m, 2h, or 3d.');
    }
    const value = parseInt(match[1], 10);
    const unit = match[2];
    let startMs;
    switch (unit) {
        case 'm':
            startMs = endMs - value * 60 * 1000;
            break;
        case 'h':
            startMs = endMs - value * 3600 * 1000;
            break;
        case 'd':
            startMs = endMs - value * 86400 * 1000;
            break;
        default:
            throw new Error('Invalid time unit. Use m for minutes, h for hours, or d for days.');
    }
    if (endMs - startMs > 3 * 24 * 3600 * 1000) {
        throw new Error('Time range cannot exceed 3 days');
    }
    return { startMs, endMs };
}
async function getFlowsRunList(command) {
    try {
        const auth = await (0, utils_1.getAuthForCommand)(command, 'flows');
        const options = {
            accessToken: auth.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            workspaceId: command.workspace,
            flowId: command.flow,
            startMs: command.range.startMs,
            endMs: command.range.endMs
        };
        const runs = await client_1.default.listRuns(options);
        return runs;
    }
    catch (error) {
        throw new Error(`Failed to list runs: ${error.message}`);
    }
}
async function runCommand(command) {
    try {
        const response = await getFlowsRunList(command);
        if (command.json) {
            print.lf(JSON.stringify(response, null, 4));
            return;
        }
        if (response.history && response.history.length > 0) {
            const total = response.meta.total;
            print.lf(`${total} run(s)\n`);
            print.lf(formatRunsTable(response).toString());
        }
        else {
            print.lf('No runs found');
        }
    }
    catch (error) {
        print.lf(error.message);
        process.exit(1);
    }
}
exports.default = {
    validateCommand: (input) => { return (0, handler_1.validateCommand)(input, ListRunsCommandSchema); },
    runCommand: runCommand
};
