// @postman/pm-bin-linux-x64/bin/lib/commands/flows/flows-deploy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print = __importStar(require("../../framework/print"));
const utils_1 = require("./cmd/utils");
const handler_1 = require("./cmd/handler");
const zod_1 = require("zod");
const client_1 = __importDefault(require("./api/client"));
const either_1 = require("./types/either");
const DeployFlowCommandSchema = zod_1.z.object({
    verbose: zod_1.z.boolean().default(false),
    debug: zod_1.z.boolean().default(false),
    json: zod_1.z.boolean().default(false),
    flow: zod_1.z.string().regex(/^[a-f0-9]{24}$/, 'Flow ID must be a 24-character hex string (e.g. 69a70d07fc0dbce42fe9c202)'),
    path: zod_1.z.string()
        .regex(/^\/[a-zA-Z0-9_-]+$/, 'URL path must start with / and contain only letters, numbers, hyphens, and underscores (e.g. /my-trigger)')
        .transform((p) => { return p.slice(1); }),
    timeout: zod_1.z.string().optional(),
    auth: zod_1.z.boolean().default(false)
}).refine((data) => {
    if (data.timeout) {
        const match = data.timeout.match(/^(\d+)ms$/);
        if (!match) {
            return false;
        }
        const timeoutValue = parseInt(match[1], 10);
        if (timeoutValue < 5000 || timeoutValue > 60000) {
            return false;
        }
    }
    return true;
}, {
    message: 'Timeout must be between 5000ms and 60000ms (5s to 60s)',
    path: ['timeout']
});
async function deployFlow(command) {
    try {
        const auth = await (0, utils_1.getAuthForCommand)(command, 'flows');
        let targetId;
        let targetResponse;
        if (command.verbose) {
            print.lf('Creating deployment target...');
        }
        const createResult = await client_1.default.createTarget({
            accessToken: auth.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            flowId: command.flow,
            path: command.path
        });
        if ((0, either_1.isLeft)(createResult)) {
            print.lf('Target already exists, checking existing path...');
            targetResponse = await client_1.default.getTarget({
                accessToken: auth.accessToken,
                service: 'flows',
                verbose: command.verbose,
                debug: command.debug,
                flowId: command.flow
            });
            targetId = targetResponse.target.id;
            const existingPath = targetResponse.target.url?.path;
            if (existingPath !== command.path) {
                print.lf(`Updating path from "${existingPath}" to "${command.path}"...`);
                await client_1.default.updateTargetPath({
                    accessToken: auth.accessToken,
                    service: 'flows',
                    verbose: command.verbose,
                    debug: command.debug,
                    targetId: targetId,
                    path: command.path
                });
                if (command.verbose) {
                    print.lf('Path updated');
                }
            }
            else if (command.verbose) {
                print.lf('Path matches, using existing target');
            }
        }
        else {
            targetResponse = createResult.right;
            targetId = targetResponse.target.id;
            if (command.debug) {
                print.lf(`Target created: ${targetId}`);
            }
        }
        if (command.verbose) {
            print.lf('Deploying target...');
        }
        let timeoutValue;
        if (command.timeout) {
            const match = command.timeout.match(/^(\d+)ms$/);
            if (match) {
                timeoutValue = parseInt(match[1], 10);
            }
        }
        const deployResponse = await client_1.default.deployTarget({
            accessToken: auth.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            flowId: command.flow,
            targetId: targetId,
            timeout: timeoutValue
        });
        if (command.verbose) {
            print.lf(`Deployment created: ${deployResponse.id}`);
        }
        let authResponse;
        let decryptedAuthResponse;
        if (command.auth) {
            if (command.verbose) {
                print.lf('Enabling authentication...');
            }
            authResponse = await client_1.default.enableAuth({
                accessToken: auth.accessToken,
                service: 'flows',
                verbose: command.verbose,
                debug: command.debug,
                flowId: command.flow,
                targetId: targetId
            });
            if (command.verbose) {
                print.lf('Authentication enabled');
                print.lf('Decrypting auth token...');
            }
            decryptedAuthResponse = await client_1.default.decryptAuth({
                accessToken: auth.accessToken,
                service: 'flows',
                verbose: command.verbose,
                debug: command.debug,
                targetId: targetId,
                encryptedAuthToken: authResponse.encryptedAuthToken
            });
            if (command.verbose) {
                print.lf('Auth token decrypted');
            }
        }
        return {
            target: targetResponse,
            deployment: deployResponse,
            auth: authResponse,
            decryptedAuth: decryptedAuthResponse?.token
        };
    }
    catch (error) {
        throw new Error(`Failed to deploy flow: ${error.message}`);
    }
}
async function runCommand(command) {
    try {
        const response = await deployFlow(command);
        if (command.json) {
            print.lf(JSON.stringify(response, null, 4));
            return;
        }
        const url = response.target.target.url;
        const webhookUrl = url ? `${url.basePath}${url.namespace}/${url.path}` : null;
        print.lf('Flow deployed successfully');
        print.lf(`  Flow ID: ${command.flow}`);
        print.lf(`  Deployment ID: ${response.deployment.id}`);
        if (command.debug) {
            print.lf(`  Target ID: ${response.target.target.id}`);
        }
        if (webhookUrl) {
            print.lf(`  Trigger URL: ${webhookUrl}`);
        }
        if (command.timeout) {
            print.lf(`  Timeout: ${command.timeout}`);
        }
        if (command.auth && response.decryptedAuth) {
            print.lf('  Authentication: Enabled');
            print.lf(`  Auth Token: ${response.decryptedAuth}`);
        }
        if (response.target.target.enabled === false) {
            print.lf('');
            print.lf('Note: trigger is currently off. Use "postman flows update --trigger on" to activate.');
        }
    }
    catch (error) {
        print.lf(error.message);
        process.exit(1);
    }
}
exports.default = {
    validateCommand: (input) => { return (0, handler_1.validateCommand)(input, DeployFlowCommandSchema); },
    runCommand: runCommand
};
