// @postman/pm-bin-linux-x64/bin/lib/commands/flows/flows-analyze-runs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print = __importStar(require("../../framework/print"));
const client_1 = __importDefault(require("./api/client"));
const utils_1 = require("./cmd/utils");
const handler_1 = require("./cmd/handler");
const zod_1 = require("zod");
const AnalyzeRunCommandSchema = zod_1.z.object({
    verbose: zod_1.z.boolean().default(false),
    debug: zod_1.z.boolean().default(false),
    json: zod_1.z.boolean().default(false),
    flow: zod_1.z.string().optional(),
    run: zod_1.z.string().min(1, 'Run ID is required'),
    logs: zod_1.z.boolean().default(false),
    filter: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional().transform((val) => {
        if (!val) {
            return undefined;
        }
        return Array.isArray(val) ? val : [val];
    })
});
async function getFlowsRun(command) {
    try {
        const auth = await (0, utils_1.getAuthForCommand)(command, 'flows');
        const options = {
            accessToken: auth.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            flowId: command.flow || '',
            runId: command.run
        };
        const run = await client_1.default.getRun(options);
        return run;
    }
    catch (error) {
        throw new Error(`Failed to get run: ${error.message}`);
    }
}
function getBlockName(processId) {
    return processId;
}
function analyzeIncompleteProcesses(logs) {
    const processStates = {};
    logs.forEach((log) => {
        if (!processStates[log.process_id]) {
            processStates[log.process_id] = {
                process_id: log.process_id,
                hasExecuting: false,
                hasExecuted: false,
                executingTime: null,
                executedTime: null
            };
        }
        if (log.event === 'process:executing') {
            processStates[log.process_id].hasExecuting = true;
            processStates[log.process_id].executingTime = log.timestamp;
        }
        else if (log.event === 'process:executed') {
            processStates[log.process_id].hasExecuted = true;
            processStates[log.process_id].executedTime = log.timestamp;
        }
    });
    const incomplete = Object.values(processStates)
        .filter((state) => { return state.hasExecuting && !state.hasExecuted; })
        .map((state) => {
        return {
            process_id: state.process_id,
            started_at: new Date(state.executingTime).toISOString(),
            timestamp: state.executingTime
        };
    });
    return incomplete;
}
async function runCommand(command) {
    try {
        const runData = await getFlowsRun(command);
        if (command.json) {
            print.lf(JSON.stringify(runData, null, 4));
            return;
        }
        if (command.verbose) {
            print.lf(`Analyzing run: ${command.run}`);
        }
        const events = runData?.data?.events || runData?.events;
        if (events && events.length > 0) {
            const logs = events;
            const sortedLogs = logs
                .map((entry) => {
                return {
                    timestamp: entry.timestamp,
                    event: entry.event,
                    process_id: entry.process_id || 'unknown',
                    order: entry.order,
                    session_id: entry.session_id,
                    inputs: entry.inputs,
                    outputs: entry.outputs
                };
            })
                .sort((a, b) => { return a.timestamp - b.timestamp; });
            const timestamps = sortedLogs.map((log) => { return log.timestamp; });
            const earliestTime = Math.min(...timestamps);
            const latestTime = Math.max(...timestamps);
            const totalDuration = latestTime - earliestTime;
            const durationSeconds = (totalDuration / 1000).toFixed(2);
            const durationMinutes = (totalDuration / 60000).toFixed(2);
            const eventCounts = {};
            sortedLogs.forEach((log) => {
                eventCounts[log.event] = (eventCounts[log.event] || 0) + 1;
            });
            const incompleteProcesses = analyzeIncompleteProcesses(sortedLogs);
            print.lf('\n=== Execution Summary ===');
            print.lf(`Started:  ${new Date(earliestTime).toISOString()}`);
            print.lf(`Ended:    ${new Date(latestTime).toISOString()}`);
            print.lf(`Duration: ${durationSeconds}s (${durationMinutes} minutes)`);
            print.lf('\nEvent Summary:');
            Object.entries(eventCounts).forEach(([event, count]) => {
                print.lf(`  ${event}: ${count}`);
            });
            if (incompleteProcesses.length > 0) {
                print.lf('\n⚠️  WARNING: Incomplete Processes Found!');
                print.lf(`${incompleteProcesses.length} process(es) started but never completed:\n`);
                incompleteProcesses.forEach((proc) => {
                    const blockName = getBlockName(proc.process_id);
                    print.lf(`  ❌ ${proc.process_id} (${blockName})`);
                    print.lf(`     Started at: ${proc.started_at}`);
                });
            }
            else {
                print.lf('\n✓ All processes completed successfully');
            }
            if (command.logs) {
                print.lf('\n=== Detailed Event Log ===');
                const filteredLogs = command.filter && command.filter.length > 0 ?
                    sortedLogs.filter((log) => {
                        if (!log.process_id) {
                            return false;
                        }
                        return command.filter.some((prefix) => {
                            return log.process_id.startsWith(prefix);
                        });
                    }) :
                    sortedLogs;
                if (command.filter && command.filter.length > 0) {
                    print.lf(`Filtering by block ID prefix: ${command.filter.join(', ')}`);
                    print.lf(`Matching entries: ${filteredLogs.length} of ${sortedLogs.length}\n`);
                }
                let separator = '';
                let startTime = null;
                let displayedCount = 0;
                const waitForKeyPress = () => {
                    return new Promise((resolve) => {
                        print.lf('\n--- Press any key to continue ---');
                        process.stdin.setRawMode(true);
                        process.stdin.resume();
                        process.stdin.once('data', () => {
                            process.stdin.setRawMode(false);
                            process.stdin.pause();
                            resolve();
                        });
                    });
                };
                const processLogs = async () => {
                    for (let index = 0; index < filteredLogs.length; index++) {
                        const log = filteredLogs[index];
                        if (index === 0) {
                            startTime = log.timestamp;
                        }
                        let timeDisplay;
                        const eventName = log.event.includes(':') ? log.event.split(':')[1] : log.event;
                        if (eventName === 'events') {
                            continue;
                        }
                        print.print(separator);
                        separator = '\n';
                        if (eventName === 'start' || eventName === 'end') {
                            timeDisplay = new Date(log.timestamp).toISOString();
                            print.lf(`${timeDisplay} | ${eventName.padEnd(10)}`);
                            displayedCount++;
                            if (displayedCount % 20 === 0 && index < filteredLogs.length - 1) {
                                await waitForKeyPress();
                            }
                            continue;
                        }
                        else if (startTime !== null) {
                            const delta = log.timestamp - startTime;
                            timeDisplay = `+${delta}ms`;
                        }
                        else {
                            timeDisplay = new Date(log.timestamp).toISOString();
                        }
                        const blockName = getBlockName(log.process_id);
                        const entry = eventName === 'executing' ? 'start' : 'end';
                        print.lf(`${timeDisplay} | ${blockName} | ${entry.padEnd(8)}`);
                        if (log.inputs && log.inputs.length > 0) {
                            log.inputs.forEach((input) => {
                                const valueStr = JSON.stringify(input.value);
                                const displayValue = valueStr.length > 128 ?
                                    `${valueStr.substring(0, 128)}...` : valueStr;
                                print.lf(`  └─ input: ${input.name} = ${displayValue}`);
                            });
                        }
                        if (log.outputs && log.outputs.length > 0) {
                            log.outputs.forEach((output) => {
                                const valueStr = JSON.stringify(output.value);
                                const displayValue = valueStr.length > 128 ?
                                    `${valueStr.substring(0, 128)}...` : valueStr;
                                print.lf(`  └─ output: ${output.name} = ${displayValue}`);
                            });
                        }
                        displayedCount++;
                        if (displayedCount % 20 === 0 && index < filteredLogs.length - 1) {
                            await waitForKeyPress();
                        }
                    }
                };
                await processLogs();
            }
            print.lf(`\nTotal entries: ${sortedLogs.length}`);
        }
        else {
            print.lf(`No events found for run ${command.run}`);
        }
    }
    catch (error) {
        print.lf(error.message);
        process.exit(1);
    }
}
exports.default = {
    validateCommand: (input) => { return (0, handler_1.validateCommand)(input, AnalyzeRunCommandSchema); },
    runCommand: runCommand
};
