// @postman/pm-bin-linux-x64/bin/lib/commands/flows/flows-update.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print = __importStar(require("../../framework/print"));
const utils_1 = require("./cmd/utils");
const handler_1 = require("./cmd/handler");
const zod_1 = require("zod");
const client_1 = __importDefault(require("./api/client"));
const UpdateFlowCommandSchema = zod_1.z.object({
    verbose: zod_1.z.boolean().default(false),
    debug: zod_1.z.boolean().default(false),
    json: zod_1.z.boolean().default(false),
    flow: zod_1.z.string().min(1, 'Flow ID is required'),
    trigger: zod_1.z.enum(['on', 'off']).optional(),
    auth: zod_1.z.enum(['on', 'off']).optional()
}).refine((data) => {
    return data.trigger !== undefined || data.auth !== undefined;
}, {
    message: 'At least one option is required: --trigger or --auth',
    path: ['trigger']
});
async function updateFlow(command) {
    try {
        const auth = await (0, utils_1.getAuthForCommand)(command, 'flows');
        if (command.verbose) {
            print.lf(`Getting deployment information for flow ${command.flow}...`);
        }
        const targetResponse = await client_1.default.getTarget({
            accessToken: auth.accessToken,
            service: 'flows',
            verbose: command.verbose,
            debug: command.debug,
            flowId: command.flow
        });
        const target = targetResponse.target;
        if (!target.deployment) {
            throw new Error(`Flow ${command.flow} is not deployed. ` +
                'Please deploy this flow first before updating it.');
        }
        const result = { flowId: command.flow, targetId: target.id };
        if (command.trigger !== undefined) {
            const enabled = command.trigger === 'on';
            if (command.verbose) {
                print.lf(`${enabled ? 'Enabling' : 'Disabling'} trigger...`);
            }
            await client_1.default.maintenance({
                accessToken: auth.accessToken,
                service: 'flows',
                verbose: command.verbose,
                debug: command.debug,
                targetId: target.id,
                enabled
            });
            result.trigger = command.trigger;
            if (command.verbose) {
                print.lf(`Trigger ${enabled ? 'enabled' : 'disabled'}`);
            }
        }
        if (command.auth !== undefined) {
            const enableAuth = command.auth === 'on';
            if (command.verbose) {
                print.lf(`${enableAuth ? 'Enabling' : 'Disabling'} authentication...`);
            }
            if (enableAuth) {
                const authResponse = await client_1.default.enableAuth({
                    accessToken: auth.accessToken,
                    service: 'flows',
                    verbose: command.verbose,
                    debug: command.debug,
                    flowId: command.flow,
                    targetId: target.id
                });
                const decryptedAuth = await client_1.default.decryptAuth({
                    accessToken: auth.accessToken,
                    service: 'flows',
                    verbose: command.verbose,
                    debug: command.debug,
                    targetId: target.id,
                    encryptedAuthToken: authResponse.encryptedAuthToken
                });
                result.auth = command.auth;
                result.authToken = decryptedAuth.token;
            }
            else {
                await client_1.default.disableAuth({
                    accessToken: auth.accessToken,
                    service: 'flows',
                    verbose: command.verbose,
                    debug: command.debug,
                    targetId: target.id
                });
                result.auth = command.auth;
            }
            if (command.verbose) {
                print.lf(`Authentication ${enableAuth ? 'enabled' : 'disabled'}`);
            }
        }
        return result;
    }
    catch (error) {
        throw new Error(`Failed to update flow: ${error.message}`);
    }
}
async function runCommand(command) {
    try {
        const result = await updateFlow(command);
        if (command.json) {
            print.lf(JSON.stringify(result, null, 4));
            return;
        }
        print.lf('Flow updated successfully');
        print.lf(`  Flow ID: ${result.flowId}`);
        if (command.debug) {
            print.lf(`  Target ID: ${result.targetId}`);
        }
        if (result.trigger !== undefined) {
            print.lf(`  Trigger: ${result.trigger}`);
        }
        if (result.auth !== undefined) {
            print.lf(`  Authentication: ${result.auth}`);
            if (result.authToken) {
                print.lf(`  Auth Token: ${result.authToken}`);
            }
        }
    }
    catch (error) {
        print.lf(error.message);
        process.exit(1);
    }
}
exports.default = {
    validateCommand: (input) => { return (0, handler_1.validateCommand)(input, UpdateFlowCommandSchema); },
    runCommand: runCommand
};
