// @postman/pm-bin-linux-x64/bin/lib/commands/flows/run-request-from-file.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runRequestFromFile = runRequestFromFile;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const v3_filesystem_collection_1 = require("../sdk/v3-filesystem-collection");
const v3_loader_1 = require("../../run/v3-loader");
const REQUEST_FILE_SUFFIX = '.request.yaml';
const REQUEST_FILE_SUFFIX_YL = '.request.yml';
function getRequestNameFromFile(requestFilePath) {
    try {
        const content = fs_1.default.readFileSync(requestFilePath, 'utf8');
        const parsed = js_yaml_1.default.load(content);
        if (parsed && typeof parsed.name === 'string' && parsed.name.trim()) {
            return parsed.name.trim();
        }
    }
    catch {
    }
    const basename = path_1.default.basename(requestFilePath);
    if (basename.toLowerCase().endsWith(REQUEST_FILE_SUFFIX_YL)) {
        return basename.slice(0, -REQUEST_FILE_SUFFIX_YL.length);
    }
    return basename.slice(0, -REQUEST_FILE_SUFFIX.length);
}
function responseBodyToString(body) {
    if (typeof body === 'string') {
        return body;
    }
    if (Buffer.isBuffer(body)) {
        return body.toString('utf8');
    }
    const bufLike = body;
    if (bufLike &&
        typeof bufLike === 'object' &&
        bufLike.type === 'Buffer' &&
        Array.isArray(bufLike.data)) {
        return Buffer.from(bufLike.data).toString('utf8');
    }
    return '{}';
}
async function runRequestFromFile(requestFilePath, runOptions) {
    const normalizedPath = path_1.default.resolve(requestFilePath);
    if (!fs_1.default.existsSync(normalizedPath)) {
        throw new Error(`Request file not found: ${normalizedPath}`);
    }
    if (!fs_1.default.statSync(normalizedPath).isFile()) {
        throw new Error(`Path is not a file: ${normalizedPath}`);
    }
    if (!(0, v3_filesystem_collection_1.isV3RequestFilePath)(normalizedPath)) {
        throw new Error(`Path is not a v3 request file (*.request.yaml or *.request.yml): ${normalizedPath}`);
    }
    const collectionRoot = (0, v3_filesystem_collection_1.findV3FilesystemCollectionRootFromFile)(normalizedPath);
    if (!collectionRoot) {
        throw new Error(`Request file is not part of a v3 filesystem collection (no collection root found): ${normalizedPath}`);
    }
    const collection = await (0, v3_loader_1.loadV3AsExtensibleCollection)(collectionRoot);
    if (!collection) {
        throw new Error(`Failed to load v3 collection from root: ${collectionRoot}`);
    }
    const requestName = getRequestNameFromFile(normalizedPath);
    const globalVar = Object.entries(runOptions?.variables || {}).map(([key, value]) => {
        return { key, value };
    });
    const options = {
        collection,
        i: [requestName],
        workingDir: collectionRoot,
        silent: true,
        globalVar,
        envVar: runOptions?.envVar
    };
    return new Promise((resolve, reject) => {
        const run = require('../../run');
        run(options, (err, summary) => {
            if (err) {
                reject(err);
            }
            else {
                const [runSummary] = summary.run.executions;
                const response = runSummary.protocol.response ?? {};
                const body = responseBodyToString(response.body ?? '{}');
                const statusCode = response.statusCode ?? 0;
                const responseHeaders = response.headers ?? [];
                const responseTime = runSummary.protocol.responseTime ?? 0;
                const contentTypeHeader = responseHeaders.find((h) => {
                    return h.key.toLowerCase() === 'content-type';
                });
                const contentType = contentTypeHeader?.value ?? 'application/octet-stream';
                const [mimeType, ...charsetParts] = contentType.split(';').map((s) => {
                    return s.trim();
                });
                const charsetPart = charsetParts.find((s) => {
                    return s.toLowerCase().startsWith('charset=');
                });
                const charset = charsetPart?.split('=')[1]?.trim() ?? 'utf8';
                const mimeFormat = mimeType.includes('json') ? 'json' : mimeType.includes('xml') ? 'xml' : 'text';
                const fileExtension = mimeFormat === 'json' ? 'json' : mimeFormat === 'xml' ? 'xml' : 'txt';
                const contentInfo = response.contentInfo ?? {
                    contentType,
                    mimeType: mimeType.split('/')[0] ?? 'application',
                    mimeFormat,
                    charset,
                    fileExtension,
                    fileName: `response.${fileExtension}`
                };
                resolve({
                    response: {
                        responseTime,
                        code: statusCode,
                        headers: responseHeaders,
                        status: response.status ?? (statusCode === 200 ? 'OK' : String(statusCode)),
                        body,
                        contentInfo,
                        isMedia: response.isMedia ?? false
                    },
                    assertions: runSummary.assertions.map((assertion, index) => {
                        return {
                            index: index + 1,
                            name: assertion.assertion,
                            passed: !assertion.error,
                            skipped: assertion.skipped,
                            error: assertion.error
                        };
                    })
                });
            }
        });
    });
}
