// @postman/pm-bin-linux-x64/bin/lib/commands/monitor/monitor-service.js
const request = require('postman-request'), util = require('../../util'), PostmanResource = require('../../api/postman-resource');
class MonitorService {
    constructor() {
        this.commonHeaders = {
            ...util.defaultHeaders,
            'x-original-user-agent': util.userAgent
        };
    }
    static _buildApiUrl(endpoint) {
        return `${util.POSTMAN_GATEWAY_BASE_URL()}${endpoint}`;
    }
    _buildAuthHeaders(postmanApiKey, accessToken, service) {
        const headers = { ...this.commonHeaders };
        if (postmanApiKey) {
            headers['x-api-key'] = postmanApiKey;
        }
        else if (accessToken) {
            headers['x-access-token'] = accessToken;
        }
        if (service) {
            headers['x-pstmn-req-service'] = service;
        }
        return headers;
    }
    static _transformJobToLastRun(job) {
        return {
            jobId: job.id,
            startedAt: job.startedAt,
            finishedAt: job.finishedAt,
            status: job.status?.state,
            stats: job.results || MonitorService._buildStatsFromJobDetails(job)
        };
    }
    static _buildStatsFromJobDetails(jobDetails) {
        return {
            runCount: jobDetails.results?.runCount || 0,
            abortedCount: jobDetails.results?.abortedCount || 0,
            errorCount: jobDetails.results?.errorCount || 0,
            warningCount: jobDetails.results?.warningCount || 0,
            requests: {
                total: jobDetails.results?.requestCount || 0
            },
            assertions: {
                total: (jobDetails.results?.passedTestCount || 0) +
                    (jobDetails.results?.failedTestCount || 0),
                failed: jobDetails.results?.failedTestCount || 0
            },
            responseLatency: jobDetails.results?.responseLatency || 0,
            responseSize: jobDetails.results?.responseSize || 0
        };
    }
    _makeRequest({ method, url, postmanApiKey, accessToken, service }) {
        const headers = this._buildAuthHeaders(postmanApiKey, accessToken, service);
        const requestConfig = { url, headers };
        return new Promise((resolve, reject) => {
            request[method](requestConfig, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                let data;
                try {
                    data = body && JSON.parse(body);
                }
                catch (parseError) {
                    return reject(new Error(`Failed to parse response: ${parseError.message}`));
                }
                if (response.statusCode < 200 || response.statusCode >= 300) {
                    const errorMessage = `HTTP ${response.statusCode}: ${(data && data.message) || 'Request failed'}`, error = new Error(errorMessage);
                    error.response = { status: response.statusCode, data: data };
                    return reject(error);
                }
                resolve({ data: data, statusCode: response.statusCode });
            });
        });
    }
    async _getMonitorJobs({ monitorId, postmanApiKey, accessToken }) {
        const url = MonitorService._buildApiUrl(`/jobtemplates/${monitorId}/jobs`);
        const response = await this._makeRequest({
            method: 'get',
            url: url,
            postmanApiKey: postmanApiKey,
            accessToken: accessToken,
            service: 'newman-remote-api'
        });
        return response.data;
    }
    async _getLastSuccessfulJob({ monitorId, postmanApiKey, accessToken }) {
        const jobs = await this._getMonitorJobs({ monitorId, postmanApiKey, accessToken });
        const completedJob = jobs.data.find((job) => { return job.startedAt !== null && job.finishedAt !== null; });
        return completedJob || null;
    }
    async getMonitor({ monitorId, postmanApiKey, accessToken }) {
        const url = MonitorService._buildApiUrl(`/jobtemplates/${monitorId}`);
        const response = await this._makeRequest({
            method: 'get',
            url: url,
            postmanApiKey: postmanApiKey,
            accessToken: accessToken,
            service: 'newman-remote-api'
        });
        const rawData = response.data?.data || {};
        let lastRun = null;
        try {
            const lastSuccessfulJob = await this._getLastSuccessfulJob({
                monitorId,
                postmanApiKey,
                accessToken
            });
            if (lastSuccessfulJob) {
                lastRun = MonitorService._transformJobToLastRun(lastSuccessfulJob);
            }
        }
        catch (error) {
        }
        const transformedData = {
            monitor: {
                id: rawData.id,
                name: rawData.name,
                active: rawData.active,
                collection: rawData.collection,
                collectionUid: rawData.collection,
                environmentUid: rawData.environment || null,
                options: rawData.options || {},
                distribution: rawData.distribution || [],
                notifications: rawData.notifications || {},
                notificationLimit: rawData.notificationLimit || 3,
                files: rawData.files || { data: [] },
                retry: rawData.retry || { attempts: null },
                lastRun: lastRun,
                createdBy: rawData.createdBy,
                createdAt: rawData.createdAt,
                updatedAt: rawData.updatedAt
            }
        };
        return transformedData;
    }
    async runMonitor({ monitorId, postmanApiKey, accessToken }) {
        const url = MonitorService._buildApiUrl(`/jobtemplates/${monitorId}/jobs`);
        const response = await this._makeRequest({
            method: 'post',
            url: url,
            postmanApiKey: postmanApiKey,
            accessToken: accessToken,
            service: 'newman-remote-api'
        });
        return { data: response.data, statusCode: response.statusCode };
    }
    getCollectionById({ collectionId, postmanApiKey, accessToken }) {
        return new Promise((resolve, reject) => {
            const collectionResource = new PostmanResource('collection', collectionId, postmanApiKey, accessToken);
            collectionResource.get((err, data) => {
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
    async getJobDetails({ jobId, postmanApiKey, accessToken }) {
        const url = MonitorService._buildApiUrl(`/jobs/${jobId}`);
        const response = await this._makeRequest({
            method: 'get',
            url: url,
            postmanApiKey: postmanApiKey,
            accessToken: accessToken,
            service: 'newman-remote-api'
        });
        const jobDetails = response.data.data;
        return MonitorService._transformJobToLastRun(jobDetails);
    }
    getEnvironmentById({ environmentId, postmanApiKey, accessToken }) {
        return new Promise((resolve, reject) => {
            const environmentResource = new PostmanResource('environment', environmentId, postmanApiKey, accessToken);
            environmentResource.get((err, data) => {
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
}
module.exports = new MonitorService();
