// @postman/pm-bin-linux-x64/bin/lib/commands/monitor/monitor-output.js
const _ = require('lodash'), colors = require('colors/safe'), Table = require('cli-table3'), util = require('../../util'), print = require('../../framework/print'), { getRunStatus } = require('./monitor-utils');
function displayMonitorDetails(monitor) {
    const monitorName = _.get(monitor, 'monitor.name', 'Unknown Monitor');
    const active = _.get(monitor, 'monitor.active');
    const collectionName = _.get(monitor, 'monitor.collectionName');
    const environmentName = _.get(monitor, 'monitor.environmentName');
    const retryAttempts = _.get(monitor, 'monitor.retry.attempts');
    const distribution = _.get(monitor, 'monitor.distribution');
    const dataFiles = _.get(monitor, 'monitor.files.data');
    const strictSSL = _.get(monitor, 'monitor.options.strictSSL');
    const followRedirects = _.get(monitor, 'monitor.options.followRedirects');
    const requestTimeout = _.get(monitor, 'monitor.options.requestTimeout');
    const requestDelay = _.get(monitor, 'monitor.options.requestDelay');
    let output = '\n';
    output += `~ ${colors.cyan('Monitor:')} ` +
        `${monitorName}\n`;
    output += '─ Active: ' +
        `${active ? colors.green('Yes') : colors.red('No')}\n`;
    output += '─ Collection: ' +
        `${collectionName || 'Not specified'}\n`;
    output += '─ Environment: ' +
        `${environmentName || 'Not specified'}\n`;
    output += '─ Retry: ' +
        `${retryAttempts ? `${retryAttempts} attempts` : 'Default'}\n`;
    output += '─ Strict SSL: ' +
        `${strictSSL ? 'Enabled' : 'Disabled'}\n`;
    output += '─ Follow Redirects: ' +
        `${followRedirects ? 'Enabled' : 'Disabled'}\n`;
    output += '─ Request Timeout: ' +
        `${requestTimeout ? `${requestTimeout}ms` : 'Default'}\n`;
    output += '─ Request Delay: ' +
        `${requestDelay ? `${requestDelay}ms` : 'None'}\n`;
    if (distribution && distribution.length > 0) {
        const runners = distribution.map((d) => { return d.region; }).join(', ');
        output += '─ Runners: ' +
            `${runners}\n`;
    }
    else {
        output += '─ Runners: ' +
            'Select Automatically\n';
    }
    if (dataFiles && dataFiles.length > 0) {
        dataFiles.forEach((file) => {
            const fileSize = file.size ? `${(file.size / 1024).toFixed(2)} KB` : 'Unknown size';
            output += '─ Data File: ' +
                `${file.name} (${file.format}, ${fileSize})\n`;
        });
    }
    print.lf(output);
}
function displayMonitorRunResults(run) {
    let output = '\n';
    const statisticsTable = new Table({
        chars: undefined,
        style: { head: [] },
        head: ['', 'executed', '  failed'],
        colAligns: ['right', 'right', 'right'],
        colWidths: [20, 20, 20]
    });
    const runCount = {
        total: _.get(run, 'stats.runCount', 1),
        failed: _.get(run, 'stats.abortedCount', 0)
    };
    const runCountLabel = runCount.failed ? colors.red('runs') : colors.green('runs');
    statisticsTable.push([
        runCountLabel,
        runCount.total,
        runCount.failed ? colors.red(runCount.failed) : runCount.failed
    ]);
    const requestsTotal = _.get(run, 'stats.requestCount', 0);
    const requestsFailed = _.get(run, 'stats.errorCount', 0);
    const requestsLabel = requestsFailed ? colors.red('requests') : colors.green('requests');
    statisticsTable.push([
        requestsLabel,
        requestsTotal,
        requestsFailed ? colors.red(requestsFailed) : requestsFailed
    ]);
    const assertionsTotal = _.get(run, 'stats.passedTestCount', 0) + _.get(run, 'stats.failedTestCount', 0);
    const assertionsFailed = _.get(run, 'stats.failedTestCount', 0);
    const assertionsLabel = assertionsFailed ? colors.red('assertions') : colors.green('assertions');
    statisticsTable.push([
        assertionsLabel,
        assertionsTotal,
        assertionsFailed ? colors.red(assertionsFailed) : assertionsFailed
    ]);
    const startedAt = _.get(run, 'startedAt');
    const finishedAt = _.get(run, 'finishedAt');
    if (startedAt && finishedAt) {
        const duration = new Date(finishedAt) - new Date(startedAt);
        statisticsTable.push([{
                colSpan: 3,
                content: `total run duration: ${util.prettyms(duration)}`,
                hAlign: 'left'
            }]);
    }
    const responseSize = _.get(run, 'stats.responseSize');
    if (responseSize) {
        const dataSizeKB = (responseSize / 1024).toFixed(2);
        statisticsTable.push([{
                colSpan: 3,
                content: `total data received: ${dataSizeKB} KB (approx)`,
                hAlign: 'left'
            }]);
    }
    const responseLatency = _.get(run, 'stats.responseLatency');
    if (responseLatency && requestsTotal > 0) {
        const avgTime = Math.round(responseLatency / requestsTotal);
        statisticsTable.push([{
                colSpan: 3,
                content: `average response time: ${avgTime} ms`,
                hAlign: 'left'
            }]);
    }
    output += statisticsTable.toString() + '\n';
    const derivedStatus = getRunStatus(run);
    const runStatus = derivedStatus.toUpperCase();
    let resultText = '';
    if (runStatus === 'SUCCESS') {
        resultText = `Result: ${colors.green('SUCCESS')}`;
    }
    else if (runStatus === 'FAILED') {
        resultText = `Result: ${colors.red('FAILED')}`;
    }
    else if (runStatus === 'ERROR') {
        resultText = `Result: ${colors.red('ERROR')}`;
    }
    else if (runStatus === 'ABORTED') {
        resultText = `Result: ${colors.red('ABORTED')}`;
    }
    else if (runStatus === 'RUNNING') {
        resultText = `Result: ${colors.yellow('RUNNING')}`;
    }
    else {
        resultText = `Result: ${colors.gray('UNKNOWN')}`;
    }
    output += `${resultText}\n`;
    print.lf(output);
}
module.exports = {
    displayMonitorDetails,
    displayMonitorRunResults
};
