// @postman/pm-bin-linux-x64/bin/lib/commands/collection/migrate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const v3_export_1 = require("@postman/v3.export");
const v2_1 = require("@postman/runtime.models/v2");
const transforms_1 = require("@postman/runtime.models/transforms");
const console = require('../../framework/console');
const SCHEMA_VERSION_RE = /\/v(\d+\.\d+\.\d+)\//;
function extractSchemaVersion(schema) {
    const match = SCHEMA_VERSION_RE.exec(schema);
    return match ? match[1] : null;
}
function assertV21Schema(schema) {
    const version = extractSchemaVersion(schema);
    if (version !== '2.1.0') {
        throw new Error(`Unsupported collection schema version "${version || schema}". Only v2.1.0 collections are supported. ` +
            'Please export your collection in v2.1 format from Postman and try again.');
    }
}
function assertOutputDirEmpty(dirPath) {
    if (!fs_1.default.existsSync(dirPath)) {
        return;
    }
    const entries = fs_1.default.readdirSync(dirPath);
    if (entries.length > 0) {
        throw new Error(`Output directory "${dirPath}" already exists and is not empty.`);
    }
}
function migrateCollection(collectionPath, options) {
    const resolvedPath = path_1.default.resolve(collectionPath);
    if (!fs_1.default.existsSync(resolvedPath)) {
        throw new Error(`Collection file not found: ${resolvedPath}`);
    }
    const raw = fs_1.default.readFileSync(resolvedPath, 'utf-8');
    let collectionJson;
    try {
        collectionJson = JSON.parse(raw);
    }
    catch {
        throw new Error(`Failed to parse JSON from: ${resolvedPath}`);
    }
    const schema = collectionJson?.info?.schema;
    if (!schema) {
        throw new Error('Invalid collection: missing "info.schema" field.');
    }
    assertV21Schema(schema);
    let parsed;
    try {
        parsed = v2_1.Collection.parse(collectionJson);
    }
    catch (err) {
        if (err.issues) {
            console.error('Collection validation issues:');
            for (const issue of err.issues) {
                const loc = issue.path
                    ?.map((p) => {
                    return p.key ?? p;
                })
                    .join('.') || '(root)';
                console.error(`  - ${loc}: ${issue.message}`);
            }
        }
        throw new Error('Failed to parse collection: the file does not conform to the v2.1 schema.');
    }
    const v3Collection = (0, transforms_1.transform)(v2_1.Collection, transforms_1.FormatVersion.V3, parsed);
    const tmpDir = fs_1.default.mkdtempDisposableSync(path_1.default.join(os_1.default.tmpdir(), 'postman-migrate-'));
    const { files, rootPath } = (0, v3_export_1.splitCollection)(v3Collection, {
        collectionDirName: options.output,
        rootPrefix: tmpDir.path.replaceAll(path_1.default.sep, path_1.default.posix.sep)
    });
    const collectionDirName = path_1.default.posix.basename(rootPath);
    const outputDir = options.output ?
        path_1.default.resolve(options.output) :
        path_1.default.resolve(collectionDirName);
    assertOutputDirEmpty(outputDir);
    try {
        for (const file of files) {
            fs_1.default.mkdirSync(path_1.default.dirname(file.path), { recursive: true });
            fs_1.default.writeFileSync(file.path, file.content, 'utf-8');
        }
        fs_1.default.renameSync(rootPath, outputDir);
        return outputDir;
    }
    finally {
        tmpDir.remove();
    }
}
function migrateCommand(program) {
    program
        .command('migrate <path>')
        .description('Migrate a v2.1 collection to the v3 format')
        .option('-o, --output <dir>', 'Output directory (default: ./<collection-name>)')
        .action((collectionPath, options) => {
        try {
            const outputDir = migrateCollection(collectionPath, options);
            console.info(`Collection migrated to: ${outputDir}`);
        }
        catch (err) {
            console.error(`Error: ${err.message || err}`);
            process.exit(1);
        }
    });
}
migrateCommand._testing = {
    migrateCollection,
    assertV21Schema,
    extractSchemaVersion,
    assertOutputDirEmpty
};
module.exports = migrateCommand;
