// @postman/pm-bin-linux-x64/bin/lib/run/unified/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRunContext = buildRunContext;
const _ = require("lodash");
const variables_1 = require("./variables");
const PackageResolver = require('../package-resolver');
const RequestResolver = require('../request-resolver');
const console = require('../../framework/console');
function promisifyCallback(fn, ...args) {
    return new Promise((resolve, reject) => {
        fn(...args, (err, result) => {
            if (err) {
                return reject(err);
            }
            return resolve(result);
        });
    });
}
function findItemByNameOrId(node, nameOrId) {
    if (!node) {
        return null;
    }
    if (node.id === nameOrId) {
        return node;
    }
    if (node.title && node.title.toLowerCase() === nameOrId.toLowerCase()) {
        return node;
    }
    if (Array.isArray(node.children)) {
        for (const child of node.children) {
            const found = findItemByNameOrId(child, nameOrId);
            if (found) {
                return found;
            }
        }
    }
    return null;
}
function collectLeafItemIds(node, result) {
    if (!node) {
        return result;
    }
    if (node.type !== 'folder' && node.type !== 'collection' && node.id) {
        result.push(node.id);
    }
    if (Array.isArray(node.children)) {
        node.children.forEach((child) => { return collectLeafItemIds(child, result); });
    }
    return result;
}
function resolveSelections(collection, selections) {
    const resolved = [];
    const notFound = [];
    selections.forEach((nameOrId) => {
        const item = findItemByNameOrId(collection, nameOrId);
        if (!item) {
            notFound.push(nameOrId);
            return;
        }
        if (item.type === 'folder') {
            collectLeafItemIds(item, resolved);
        }
        else {
            resolved.push(item.id);
        }
    });
    if (notFound.length) {
        throw new Error('The following items were not found in the collection: ' + notFound.join(', '));
    }
    return resolved;
}
function buildVariableScopes(options, collection) {
    const variables = {};
    const envScope = (0, variables_1.adaptVariableScope)(options.environment);
    const globalsScope = (0, variables_1.adaptVariableScope)(options.globals);
    if (envScope) {
        variables.environment = envScope;
    }
    if (globalsScope) {
        variables.workspace = globalsScope;
    }
    variables.local = (0, variables_1.createVariableStore)([]);
    const collectionVars = _.get(collection, 'payload.variables', []);
    if (_.isArray(collectionVars)) {
        const store = collectionVars.map((v) => {
            return {
                key: v.key, value: v.value, disabled: Boolean(v.disabled)
            };
        });
        variables.collection = (0, variables_1.createVariableStore)(store, { disabled: false });
    }
    return variables;
}
function buildCollectionOptions(options, collection) {
    const opts = {};
    if (options.iterationCount && options.iterationCount > 1) {
        opts.iterations = options.iterationCount;
    }
    if (options.iterationData?.length) {
        opts.data = options.iterationData;
    }
    if (options.delayRequest) {
        opts.itemDelayMs = options.delayRequest;
    }
    if (options.i) {
        const raw = _.isArray(options.i) ? options.i : [options.i];
        const ids = resolveSelections(collection, raw);
        opts.selection = ids.map((id) => { return { id }; });
    }
    return Object.keys(opts).length ? opts : undefined;
}
function buildHttpOptions(options) {
    const http = {};
    if (options.timeoutRequest) {
        http.requestTimeoutMs = options.timeoutRequest;
    }
    if (_.has(options, 'ignoreRedirects')) {
        http.followRedirects = !options.ignoreRedirects;
    }
    return Object.keys(http).length ? http : undefined;
}
function normalizeCertField(val) {
    if (!val) {
        return undefined;
    }
    return typeof val === 'string' ? { src: val } : val;
}
function buildCertificateContext(options) {
    const definitions = [];
    if (_.isArray(options.sslClientCertList) && options.sslClientCertList.length) {
        options.sslClientCertList.forEach((entry) => {
            definitions.push({
                match: entry.matches || entry.match || '*',
                cert: normalizeCertField(entry.cert),
                key: normalizeCertField(entry.key),
                pfx: normalizeCertField(entry.pfx),
                passphrase: entry.passphrase || undefined
            });
        });
    }
    if (options.sslClientCert) {
        definitions.push({
            match: ['*'],
            cert: { src: options.sslClientCert },
            key: options.sslClientKey ? { src: options.sslClientKey } : undefined,
            passphrase: options.sslClientPassphrase || undefined
        });
    }
    if (!definitions.length && !options.sslExtraCaCerts) {
        return undefined;
    }
    const certs = {};
    if (definitions.length) {
        certs.definitions = definitions;
    }
    if (options.sslExtraCaCerts) {
        certs.ca = { src: options.sslExtraCaCerts };
    }
    return certs;
}
function buildCookieContext(options) {
    if (!options.cookieJar) {
        return undefined;
    }
    const jar = options.cookieJar;
    return {
        store: jar.store,
        getCookieString(url) {
            return new Promise((resolve, reject) => {
                jar.getCookieString(url, (err, str) => {
                    err ? reject(err) : resolve(str);
                });
            });
        },
        setCookie(cookie, url) {
            return new Promise((resolve, reject) => {
                jar.setCookie(cookie, url, (err) => {
                    err ? reject(err) : resolve();
                });
            });
        }
    };
}
function buildScriptingOptions(options) {
    const scripting = {};
    const packageResolver = new PackageResolver({
        postmanApiKey: options.postmanApiKey,
        accessToken: options.accessToken,
        logger: {
            info() { },
            error(msg) { console.error(msg); }
        }
    });
    scripting.packageResolver = (input) => {
        return promisifyCallback(packageResolver.resolvePackages.bind(packageResolver), input)
            .then((result) => { return result || {}; });
    };
    const requestResolver = new RequestResolver({
        postmanApiKey: options.postmanApiKey,
        accessToken: options.accessToken
    });
    scripting.requestResolver = (requestId, resolverContext) => {
        return promisifyCallback(requestResolver.resolveRequest.bind(requestResolver), requestId, resolverContext)
            .then((result) => { return result || {}; });
    };
    return scripting;
}
function buildRunContext(options, collection) {
    const context = {};
    context.variables = buildVariableScopes(options, collection);
    context.options = buildCollectionOptions(options, collection);
    context.http = buildHttpOptions(options);
    context.tls = { strict: _.has(options, 'insecure') ? !options.insecure : true };
    context.certificates = buildCertificateContext(options);
    context.proxy = { useEnvironment: true };
    context.cookies = buildCookieContext(options);
    context.scripting = buildScriptingOptions(options);
    return context;
}
