// @postman/pm-bin-linux-x64/bin/lib/run/unified/run.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeUnifiedRun = executeUnifiedRun;
const node_http_1 = __importDefault(require("node:http"));
const node_https_1 = __importDefault(require("node:https"));
const context_1 = require("./context");
const normalizer_1 = require("./proto/normalizer");
const resolver_1 = require("./proto/resolver");
const events_1 = require("./events");
const bail_1 = require("./bail");
const runtime_1 = require("@postman/runtime");
const print = require('../../framework/print');
const colors = require('colors/safe');
const getUserFacingError = require('../get-user-facing-error');
const cliSymbols = require('../../reporters/cli/cli-utils-symbols');
const cliUtils = require('../../reporters/cli/cli-utils');
const featureFlags = require('../../framework/feature-flags');
const SUPPORTED_ITEM_TYPES = new Set([
    'collection', 'folder',
    'http-request', 'graphql-request', 'grpc-request',
    'http-example', 'graphql-example', 'grpc-example'
]);
const FLAG_GATED_TYPES = {
    'grpc-request': { flag: 'grpc_protocol_execution_allowed', label: 'gRPC' },
    'graphql-request': { flag: 'graphql_v2_protocol_execution_allowed', label: 'GraphQL' }
};
function getBlockedProtocols(node) {
    const blocked = new Set();
    function walk(n) {
        const gated = FLAG_GATED_TYPES[n.type];
        if (gated && !featureFlags.isEnabled(gated.flag)) {
            blocked.add(gated.label);
        }
        if (Array.isArray(n.children)) {
            n.children.forEach(walk);
        }
    }
    walk(node);
    return Array.from(blocked);
}
function filterUnsupportedItems(node, selectedIds) {
    const skipped = [];
    function walk(parent) {
        if (!Array.isArray(parent.children)) {
            return;
        }
        parent.children = parent.children.filter((child) => {
            if (!SUPPORTED_ITEM_TYPES.has(child.type)) {
                if (!selectedIds || selectedIds.has(child.id)) {
                    skipped.push({ name: child.title || child.id, type: child.type });
                }
                return false;
            }
            walk(child);
            if (child.type === 'folder' && Array.isArray(child.children) && child.children.length === 0) {
                return false;
            }
            return true;
        });
    }
    walk(node);
    return skipped;
}
function destroyKeepAliveAgents() {
    node_http_1.default.globalAgent.destroy();
    node_https_1.default.globalAgent.destroy();
}
async function executeUnifiedRun(options, emitter, runContext, callback) {
    try {
        const collection = options.collection;
        const runtimeContext = (0, context_1.buildRunContext)(options, collection);
        const auth = {
            postmanApiKey: options.postmanApiKey,
            accessToken: options.accessToken
        };
        const blockedProtocols = getBlockedProtocols(collection);
        if (blockedProtocols.length) {
            const protocolList = blockedProtocols.join(' and ');
            let message;
            if (!featureFlags.isLoggedIn()) {
                message = `This collection contains ${protocolList} requests. ` +
                    'Please log in to run this collection.\n' +
                    'Run `postman login` to authenticate.';
            }
            else {
                message = `This collection contains ${protocolList} requests. ` +
                    `Please upgrade your plan to run collections with ${protocolList} protocol support.\n` +
                    'Visit https://www.postman.com/pricing/ for more details.';
            }
            return callback(new Error(message), null);
        }
        if (!options.silent) {
            print.lf(colors.grey('Running your collection...'));
        }
        if (options.timeoutScript && !options.silent) {
            const symbols = cliSymbols(cliUtils.usePlainText());
            print.lf(colors.yellow('\n  %s --timeout-script is not supported for this collection and will be ignored.\n'), symbols.warning);
        }
        let selectedIds;
        const ctxSelection = runtimeContext.options?.selection;
        if (ctxSelection) {
            selectedIds = new Set(ctxSelection.map((s) => { return s.id; }));
        }
        const skipped = filterUnsupportedItems(collection, selectedIds);
        if (skipped.length && !options.silent) {
            const symbols = cliSymbols(cliUtils.usePlainText());
            print.lf(colors.yellow('\n  %s Skipping %d unsupported item(s):\n'), symbols.warning, skipped.length);
            skipped.forEach((s) => {
                print.lf('  %s %s %s', symbols.sub, s.name, colors.grey('(' + s.type + ')'));
            });
            print.lf('');
        }
        (0, normalizer_1.normalizeGrpcSettings)(collection);
        const runtime = (0, runtime_1.createRuntime)();
        const reflectionMethod = (request) => {
            const reflection = runtime?.next?.methods?.grpc?.reflection;
            if (!reflection) {
                throw new Error('gRPC reflection is not exposed by runtime');
            }
            return reflection(request);
        };
        await (0, resolver_1.resolveCollectionSchemas)(collection, auth, reflectionMethod);
        const run = runtime.execute(collection, runtimeContext);
        const bail = (0, bail_1.createBailController)(options);
        let abortedEarly = false;
        let abortError = null;
        const globalTimeout = options.timeout ?
            setTimeout(() => {
                abortedEarly = true;
                abortError = { message: 'callback timed out' };
                run.cancel();
                run.close?.();
            }, options.timeout) :
            undefined;
        for await (const event of run) {
            (0, events_1.processEvent)(event, runContext);
            if (bail.beforeEmit(event)) {
                abortedEarly = true;
                run.cancel();
                run.close?.();
                break;
            }
            emitter.emit(event.type, event);
            if (bail.afterEmit(event)) {
                abortedEarly = true;
                run.cancel();
                run.close?.();
                break;
            }
        }
        if (globalTimeout) {
            clearTimeout(globalTimeout);
        }
        if (abortedEarly && run.abort) {
            try {
                await run.abort();
            }
            catch { }
        }
        emitter.emit('up:beforeDone', null, { summary: emitter.summary });
        emitter.emit('up:done', abortError, emitter.summary);
        destroyKeepAliveAgents();
        return callback(null, emitter.summary);
    }
    catch (err) {
        emitter.emit('up:done', getUserFacingError(err), emitter.summary);
        destroyKeepAliveAgents();
        return callback(getUserFacingError(err), emitter.summary);
    }
}
