// @postman/pm-bin-linux-x64/bin/lib/run/unified/summary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attachUnifiedTrackers = attachUnifiedTrackers;
const _ = require("lodash");
const ITEM_TYPE_TO_PROTOCOL = {
    'http-request': 'http',
    'graphql-request': 'graphql',
    'grpc-request': 'grpc',
    'ws-raw-request': 'websocket',
    'ws-socketio-request': 'socketio',
    'mqtt-request': 'mqtt',
    'llm-request': 'llm',
    'mcp-request': 'mcp'
};
const SCRIPT_STAT_KEYS = {
    beforeRequest: ['prerequestScripts', 'prerequests'],
    beforeInvoke: ['prerequestScripts', 'prerequests'],
    afterResponse: ['testScripts', 'tests']
};
function getScriptErrorSource(scriptType) {
    if (scriptType === 'beforeRequest' || scriptType === 'beforeInvoke') {
        return 'prerequest-script';
    }
    if (scriptType === 'afterResponse') {
        return 'test-script';
    }
    return 'script';
}
function updateTimingPhase(timings, phase, currentValue, requestCount) {
    if (!currentValue) {
        return;
    }
    const previousAverage = timings[phase + 'Average'];
    const previousVariance = timings[phase + 'Sd'] ** 2;
    const delta1 = currentValue - previousAverage;
    timings[phase + 'Average'] = (previousAverage * (requestCount - 1) + currentValue) / requestCount;
    timings[phase + 'Min'] = timings[phase + 'Min'] ?
        Math.min(timings[phase + 'Min'], currentValue) : currentValue;
    timings[phase + 'Max'] = Math.max(timings[phase + 'Max'] || 0, currentValue);
    const delta2 = currentValue - timings[phase + 'Average'];
    const currentVariance = (previousVariance * (requestCount - 1) + (delta1 * delta2)) / requestCount;
    timings[phase + 'Sd'] = Math.sqrt(currentVariance);
}
function handleProtocolError(event, executionCache, scriptErrorsHandled, summary, scriptKeys) {
    const correlationId = _.get(event, 'context.correlationId');
    const execution = correlationId ? executionCache[correlationId] : undefined;
    const error = _.get(event, 'payload.error', {});
    const errorMessage = (error.message || error.code || String(error));
    if (correlationId) {
        const isScriptCaused = scriptKeys.some((key) => { return scriptErrorsHandled[correlationId + ':' + key]; });
        if (isScriptCaused) {
            return;
        }
    }
    if (execution) {
        execution.requestError = error;
        execution.requestFailed = true;
        execution.errors.push({ error, source: 'request' });
    }
    summary.run.failures.push({
        error: { name: error.name || 'Error', message: errorMessage },
        at: 'request',
        source: execution ? execution.item : undefined,
        cursor: {}
    });
}
function accumulateStreamTransfer(event, executionCache) {
    const correlationId = _.get(event, 'context.correlationId');
    const execution = correlationId ? executionCache[correlationId] : undefined;
    const payload = _.get(event, 'payload', {});
    if (execution && payload.data) {
        let dataText;
        try {
            dataText = JSON.stringify(payload.data);
        }
        catch {
            dataText = String(payload.data);
        }
        execution.streamTransferTotal = (execution.streamTransferTotal || 0) + Buffer.byteLength(dataText);
    }
}
function updateScriptStatsPending(summary, scriptType) {
    const keys = SCRIPT_STAT_KEYS[scriptType];
    if (keys) {
        keys.forEach((key) => { summary.run.stats[key].pending += 1; });
    }
}
function updateScriptStatsComplete(summary, scriptType) {
    const keys = SCRIPT_STAT_KEYS[scriptType];
    if (keys) {
        keys.forEach((key) => {
            if (summary.run.stats[key].pending) {
                summary.run.stats[key].pending -= 1;
            }
            summary.run.stats[key].total += 1;
        });
    }
}
function attachUnifiedTrackers(summary, emitter, options) {
    const executionCache = {};
    const verbose = Boolean(options?.verbose);
    const scriptErrorsHandled = {};
    const graphqlSendTimestamps = {};
    let successfulResponseCount = 0;
    emitter.on('run:execute', () => {
        if (!summary.run.timings.started) {
            summary.run.timings.started = Date.now();
        }
    });
    emitter.on('up:beforeDone', () => {
        summary.run.timings.completed = Date.now();
        _.forEach(summary.run.stats, (tracker) => {
            if (tracker.pending > 0) {
                tracker.total += tracker.pending;
                tracker.pending = 0;
            }
        });
    });
    emitter.on('up:done', (err) => {
        if (err) {
            summary.run.error = err;
        }
    });
    emitter.on('collection:iteration', () => { summary.run.stats.iterations.pending += 1; });
    emitter.on('collection:iteration-end', () => {
        if (summary.run.stats.iterations.pending) {
            summary.run.stats.iterations.pending -= 1;
        }
        summary.run.stats.iterations.total += 1;
    });
    emitter.on('collection:item', () => {
        summary.run.stats.items.pending += 1;
        summary.run.stats.requests.pending += 1;
    });
    emitter.on('collection:item-end', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        if (summary.run.stats.items.pending) {
            summary.run.stats.items.pending -= 1;
        }
        summary.run.stats.items.total += 1;
        if (summary.run.stats.requests.pending) {
            summary.run.stats.requests.pending -= 1;
        }
        if (!execution) {
            return;
        }
        summary.run.stats.requests.total += 1;
        if (execution.requestFailed || !execution.protocol.request) {
            execution.requestFailed = true;
            summary.run.stats.requests.failed += 1;
            if (!execution.errors.length) {
                const errorMessage = 'Request could not be completed';
                execution.errors.push({ error: { name: 'Error', message: errorMessage }, source: 'request' });
                summary.run.failures.push({
                    error: { name: 'Error', message: errorMessage },
                    at: 'request',
                    source: execution.item,
                    cursor: {}
                });
            }
            return;
        }
        const hasResponse = Boolean(execution.protocol.response) || execution.responseReceived;
        if (!hasResponse) {
            return;
        }
        summary.run.stats.responses.total += 1;
        const response = execution.protocol.response;
        if (response?.sizes) {
            const sizes = response.sizes;
            summary.run.transfers.responseTotal += (sizes.body || 0);
        }
        if (execution.streamTransferTotal) {
            summary.run.transfers.responseTotal += execution.streamTransferTotal;
        }
        const timings = response?.timings;
        const responseTime = execution.protocol.responseTime ||
            (timings?.total) || 0;
        if (responseTime) {
            summary.run.stats.responses.totalResponseTime =
                (summary.run.stats.responses.totalResponseTime || 0) + responseTime;
        }
        successfulResponseCount += 1;
        if (responseTime) {
            const phases = { response: responseTime };
            if (verbose && timings) {
                phases.dns = timings.dns;
                phases.firstByte = timings.firstByte;
            }
            _.forEach(phases, (value, phase) => {
                updateTimingPhase(summary.run.timings, phase, value, successfulResponseCount);
            });
        }
    });
    emitter.on('scripting:execute', (event) => {
        const scriptType = _.get(event, 'payload.source.eventName');
        summary.run.stats.scripts.pending += 1;
        updateScriptStatsPending(summary, scriptType);
    });
    emitter.on('scripting:complete', (event) => {
        const scriptType = _.get(event, 'payload.source.eventName');
        if (summary.run.stats.scripts.pending) {
            summary.run.stats.scripts.pending -= 1;
        }
        summary.run.stats.scripts.total += 1;
        updateScriptStatsComplete(summary, scriptType);
    });
    emitter.on('scripting:error', (event) => {
        const scriptType = _.get(event, 'payload.source.eventName');
        const error = _.get(event, 'payload.error', {});
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        const errorSource = getScriptErrorSource(scriptType);
        if (scriptType === 'beforeRequest' || scriptType === 'beforeInvoke') {
            summary.run.stats.prerequestScripts.failed += 1;
        }
        else if (scriptType === 'afterResponse') {
            summary.run.stats.testScripts.failed += 1;
        }
        summary.run.stats.scripts.failed += 1;
        if (execution) {
            execution.errors.push({ error, source: errorSource });
        }
        summary.run.failures.push({
            error: { name: error.name || 'Error', message: error.message || String(error) },
            at: errorSource,
            source: execution ? execution.item : undefined,
            cursor: {}
        });
        if (correlationId && scriptType) {
            scriptErrorsHandled[correlationId + ':' + scriptType] = true;
        }
    });
    emitter.on('scripting:assertion', (event) => {
        const assertions = _.get(event, 'payload.assertions', []);
        assertions.forEach((a) => {
            summary.run.stats.assertions.total += 1;
            if (!a.passed && !a.skipped) {
                summary.run.stats.assertions.failed += 1;
            }
        });
    });
    emitter.on('http:error', (event) => {
        handleProtocolError(event, executionCache, scriptErrorsHandled, summary, ['beforeRequest', 'afterResponse']);
    });
    emitter.on('grpc:error', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        if (execution && !execution.protocol.type) {
            execution.protocol.type = 'grpc';
        }
        handleProtocolError(event, executionCache, scriptErrorsHandled, summary, ['beforeInvoke', 'beforeRequest', 'afterResponse']);
    });
    emitter.on('graphql:received-response-stream-start', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        const sendTimestamp = correlationId ? graphqlSendTimestamps[correlationId] : undefined;
        if (execution) {
            execution.responseReceived = true;
            execution.protocol.responseTime = sendTimestamp ? Date.now() - sendTimestamp : 0;
        }
        if (correlationId) {
            delete graphqlSendTimestamps[correlationId];
        }
    });
    emitter.on('graphql:received-response-stream-data', (event) => {
        accumulateStreamTransfer(event, executionCache);
    });
    emitter.on('grpc:received-response-data', (event) => {
        accumulateStreamTransfer(event, executionCache);
    });
    emitter.on('grpc:invoke', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        if (!execution) {
            return;
        }
        execution.protocol.type = 'grpc';
        execution.protocol.request = {
            url: _.get(event, 'payload.url'),
            methodPath: _.get(event, 'payload.methodPath'),
            message: _.get(event, 'payload.message'),
            metadata: _.get(event, 'payload.metadata', [])
        };
    });
    emitter.on('grpc:status', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        const payload = _.get(event, 'payload', {});
        if (!execution) {
            return;
        }
        execution.responseReceived = true;
        const timings = payload.timings;
        execution.protocol.responseTime = (timings?.total) || 0;
        execution.protocol.response = {
            statusCode: payload.statusCode,
            statusMessage: payload.statusMessage,
            metadata: payload.metadata,
            timings: payload.timings
        };
    });
    emitter.on('collection:item', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const itemRef = _.get(event, 'payload.item');
        const item = emitter.runContext?.resolveItemRef?.(itemRef);
        if (!correlationId) {
            return;
        }
        const itemName = (item && emitter.runContext?.getFullName?.(item.id)) ||
            (item && (item.title || item.id)) || itemRef;
        const itemTypeKey = item?.type;
        const protocolType = (itemTypeKey && ITEM_TYPE_TO_PROTOCOL[itemTypeKey]) || undefined;
        const execution = {
            item: item ? { id: item.id, name: itemName } : { id: itemRef, name: itemRef },
            iteration: _.get(event, 'context.execution.iteration', 1) - 1,
            correlationId,
            protocol: protocolType ? { type: protocolType } : {},
            assertions: [],
            errors: []
        };
        summary.run.executions.push(execution);
        executionCache[correlationId] = execution;
    });
    emitter.on('graphql:send-request', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        if (correlationId) {
            graphqlSendTimestamps[correlationId] = Date.now();
        }
        if (!execution) {
            return;
        }
        execution.protocol.type = 'graphql';
        execution.protocol.request = {
            url: _.get(event, 'payload.url'),
            operationType: _.get(event, 'payload.operationType'),
            query: _.get(event, 'payload.query'),
            headers: _.get(event, 'payload.headers', [])
        };
    });
    emitter.on('http:send-request', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        if (!execution) {
            return;
        }
        if (!execution.protocol.type) {
            execution.protocol.type = 'http';
        }
        execution.protocol.request = {
            url: _.get(event, 'payload.url'),
            method: _.get(event, 'payload.method'),
            headers: _.get(event, 'payload.headers', []),
            body: _.get(event, 'payload.body')
        };
    });
    emitter.on('http:received-response-end', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        const res = event._assembledResponse;
        if (!execution || !res) {
            return;
        }
        execution.protocol.response = res;
    });
    emitter.on('scripting:assertion', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        const assertions = _.get(event, 'payload.assertions', []);
        if (!execution) {
            return;
        }
        assertions.forEach((a) => {
            execution.assertions.push({
                assertion: a.name,
                skipped: a.skipped,
                error: (a.passed || a.skipped) ? undefined : (a.error || { message: a.name })
            });
            if (!a.passed && !a.skipped) {
                summary.run.failures.push({
                    error: {
                        name: _.get(a, 'error.name', 'AssertionError'),
                        test: a.name,
                        message: _.get(a, 'error.message', a.name || ''),
                        index: a.index
                    },
                    at: 'assertion:' + a.index + ' in test-script',
                    source: execution.item || undefined,
                    cursor: {}
                });
            }
        });
    });
    emitter.on('scripting:complete', (event) => {
        const correlationId = _.get(event, 'context.correlationId');
        const execution = correlationId ? executionCache[correlationId] : undefined;
        const scriptType = _.get(event, 'payload.source.eventName');
        const scriptError = _.get(event, 'payload.result.error');
        if (!execution) {
            return;
        }
        if (scriptError && !scriptErrorsHandled[correlationId + ':' + scriptType]) {
            const errorSource = getScriptErrorSource(scriptType);
            execution.errors.push({ error: scriptError, source: errorSource });
            summary.run.failures.push({
                error: scriptError,
                at: errorSource,
                source: execution.item || undefined,
                cursor: {}
            });
        }
    });
}
