// @postman/pm-bin-linux-x64/bin/lib/perftest/duckdb-service.js
const path = require('path');
const os = require('os');
const fs = require('fs');
const { isNil } = require('lodash');
let _DuckDBInstance = null;
function getDuckDBInstance() {
    if (!_DuckDBInstance) {
        try {
            if (typeof process.pkg !== 'undefined') {
                const externalPath = path.join(path.dirname(process.execPath), 'lib', 'node_modules', '@duckdb', 'node-api');
                if (!fs.existsSync(externalPath)) {
                    throw new Error(`DuckDB external path not found: ${externalPath}`);
                }
                _DuckDBInstance = require(externalPath).DuckDBInstance;
            }
            else {
                _DuckDBInstance = require('@duckdb/node-api').DuckDBInstance;
            }
        }
        catch (error) {
            console.error('Failed to load DuckDB:', error.message);
            throw error;
        }
    }
    return _DuckDBInstance;
}
const TARGET_TIME_SERIES_BUCKETS = 30;
class DuckDBService {
    constructor() {
        this.dbPath = path.join(os.homedir(), '.postman', 'perftest', 'perftest.db');
        this.dbInstance = null;
        this.writeConn = null;
        this.initialized = false;
    }
    async init() {
        if (this.initialized && this.writeConn && this.dbInstance) {
            return;
        }
        try {
            const dbDir = path.dirname(this.dbPath);
            if (!fs.existsSync(dbDir)) {
                fs.mkdirSync(dbDir, { recursive: true });
            }
            const DuckDBInstance = getDuckDBInstance();
            this.dbInstance = await DuckDBInstance.create(this.dbPath);
            this.writeConn = await this.dbInstance.connect();
            await this.createTables();
            this.initialized = true;
        }
        catch (error) {
            console.error('✗ Failed to initialize DuckDB:', error.message);
            throw error;
        }
    }
    async createTables() {
        try {
            await this.writeConn.run('DROP TABLE IF EXISTS request_metrics');
            await this.writeConn.run('DROP TABLE IF EXISTS response_details');
            await this.writeConn.run('DROP TABLE IF EXISTS assertion_details');
            await this.writeConn.run('DROP TABLE IF EXISTS script_error_details');
        }
        catch (error) {
        }
        const createRequestMetricsSQL = `
            CREATE TABLE request_metrics (
                id UUID NOT NULL PRIMARY KEY,
                run_id VARCHAR NOT NULL,
                timestamp TIMESTAMP NOT NULL,
                metric_name VARCHAR NOT NULL,
                value VARCHAR NOT NULL,
                unit VARCHAR,
                aggregation VARCHAR NOT NULL,
                sensor_name VARCHAR NOT NULL,
                sensor_type VARCHAR NOT NULL,
                sensor_id VARCHAR NOT NULL,
                method VARCHAR NOT NULL,
                vu_id VARCHAR NOT NULL,
                vu_count INTEGER,
                region VARCHAR NOT NULL,
                error_data JSON,
                has_error BOOLEAN,
                error_name VARCHAR,
                execution_context VARCHAR,
                http_status_code INTEGER,
                is_client_error BOOLEAN
            )
        `, createResponseDetailsSQL = `
                CREATE TABLE response_details (
                    id UUID NOT NULL PRIMARY KEY,
                    run_id VARCHAR NOT NULL,
                    request_id VARCHAR NOT NULL,
                    error_metric_name VARCHAR NOT NULL,
                    response_hash VARCHAR NOT NULL,
                    request_data JSON,
                    response_data JSON,
                    item_data JSON,
                    virtual_user_data JSON,
                    error_data JSON,
                    response_time DOUBLE,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            `, createAssertionDetailsSQL = `
                CREATE TABLE assertion_details (
                    id UUID NOT NULL PRIMARY KEY,
                    run_id VARCHAR NOT NULL,
                    request_id VARCHAR NOT NULL,
                    timestamp TIMESTAMP NOT NULL,
                    test_name VARCHAR NOT NULL,
                    test_status VARCHAR NOT NULL,
                    error_name VARCHAR,
                    error_message TEXT,
                    vu_id VARCHAR,
                    vu_count INTEGER,
                    execution_context VARCHAR,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            `, createScriptErrorDetailsSQL = `
                CREATE TABLE script_error_details (
                    id UUID NOT NULL PRIMARY KEY,
                    run_id VARCHAR NOT NULL,
                    request_id VARCHAR NOT NULL,
                    timestamp TIMESTAMP NOT NULL,
                    script_phase VARCHAR NOT NULL,
                    error_type VARCHAR NOT NULL,
                    error_name VARCHAR NOT NULL,
                    error_message TEXT,
                    vu_id VARCHAR,
                    vu_count INTEGER,
                    execution_context VARCHAR,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            `;
        await this.writeConn.run(createRequestMetricsSQL);
        await this.writeConn.run(createResponseDetailsSQL);
        await this.writeConn.run(createAssertionDetailsSQL);
        await this.writeConn.run(createScriptErrorDetailsSQL);
    }
    static processRawMetrics(rawDataArray) {
        const transformedMetrics = [], responseDetails = [];
        for (const innerArray of rawDataArray) {
            for (const rawMetric of innerArray) {
                const hasErrorData = Boolean(rawMetric.errorData), hasError = hasErrorData || Boolean(rawMetric.error), errorName = hasErrorData ?
                    rawMetric.errorData.name || rawMetric.error : rawMetric.error, errorDetails = hasErrorData ? JSON.stringify(rawMetric.errorData) : null;
                let httpStatusCode = null, isClientError = false;
                if (hasErrorData && rawMetric.errorData.response?.code) {
                    const code = rawMetric.errorData.response.code;
                    if (typeof code === 'number' || (typeof code === 'string' && (/^\d+$/).test(code))) {
                        httpStatusCode = typeof code === 'number' ? code : parseInt(code, 10);
                        isClientError = httpStatusCode >= 400 && httpStatusCode < 500;
                    }
                }
                transformedMetrics.push({
                    run_id: rawMetric.runId,
                    timestamp: rawMetric.timestamp,
                    metric_name: 'http.response_time',
                    value: rawMetric.responseTime,
                    unit: 'ms',
                    aggregation: 'raw',
                    sensor_name: rawMetric.request?.name || 'unknown',
                    sensor_type: 'request',
                    sensor_id: rawMetric.requestId,
                    method: rawMetric.request?.method || 'GET',
                    vu_id: rawMetric.vuId || 'unknown',
                    vu_count: rawMetric.vuCount || null,
                    region: 'local',
                    error_data: errorDetails,
                    has_error: hasError,
                    error_name: errorName,
                    execution_context: rawMetric.executionContext || null,
                    http_status_code: httpStatusCode,
                    is_client_error: isClientError
                });
                if (hasErrorData) {
                    const { response: responseData = {}, request: requestData = {}, virtualUser: virtualUserData = {} } = rawMetric.errorData;
                    responseDetails.push({
                        run_id: rawMetric.runId,
                        request_id: rawMetric.requestId,
                        error_metric_name: errorName,
                        response_hash: DuckDBService._createResponseHash(responseData, requestData),
                        request_data: JSON.stringify(requestData),
                        response_data: JSON.stringify(responseData),
                        item_data: '{}',
                        virtual_user_data: JSON.stringify(virtualUserData),
                        error_data: errorDetails,
                        response_time: rawMetric.responseTime || null
                    });
                }
            }
        }
        return { transformedMetrics, responseDetails };
    }
    static _createResponseHash(responseData, requestData) {
        const hashData = `${requestData.method || 'GET'}-` +
            `${requestData.url || 'unknown'}-` +
            `${responseData.status || 'unknown'}`, len = hashData.length;
        let hash = 0;
        for (let i = 0; i < len; i++) {
            hash = ((hash << 5) - hash) + hashData.charCodeAt(i);
            hash |= 0;
        }
        return hash.toString();
    }
    static processAssertionMetrics(assertionArray) {
        const assertionDetails = [];
        for (const assertion of assertionArray[0]) {
            let testStatus = 'pass';
            if (assertion.skipped) {
                testStatus = 'skip';
            }
            else if (!assertion.passed) {
                testStatus = 'fail';
            }
            assertionDetails.push({
                run_id: assertion.runId,
                request_id: assertion.requestId,
                timestamp: assertion.timestamp,
                test_name: assertion.name,
                test_status: testStatus,
                error_name: assertion.errorName || null,
                error_message: assertion.errorMessage || null,
                vu_id: assertion.vuId || null,
                vu_count: assertion.vuCount || null,
                execution_context: assertion.executionContext || null
            });
        }
        return assertionDetails;
    }
    static processScriptErrorMetrics(scriptErrorArray) {
        const scriptErrorDetails = [];
        for (const scriptError of scriptErrorArray[0]) {
            scriptErrorDetails.push({
                run_id: scriptError.runId,
                request_id: scriptError.requestId,
                timestamp: scriptError.timestamp,
                script_phase: scriptError.scriptPhase || 'test',
                error_type: scriptError.errorType || 'Error',
                error_name: scriptError.errorName || 'ScriptError',
                error_message: scriptError.errorMessage || null,
                vu_id: scriptError.vuId || null,
                vu_count: scriptError.vuCount || null,
                execution_context: scriptError.executionContext || null
            });
        }
        return scriptErrorDetails;
    }
    async insertRequestMetrics(metrics) {
        if (!metrics || metrics.length === 0) {
            return;
        }
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const escapeString = (val) => {
                if (isNil(val)) {
                    return '';
                }
                return String(val).replace(/'/g, '\'\'');
            }, formatTimestamp = (val) => {
                if (!val) {
                    return '';
                }
                try {
                    const d = new Date(val);
                    if (!isNaN(d.getTime())) {
                        return d.toISOString();
                    }
                }
                catch (_) {
                }
                return escapeString(val);
            }, values = metrics.map((metric) => {
                const ts = formatTimestamp(metric.timestamp), timestampVal = ts ? `'${ts}'` : 'NULL', unitVal = metric.unit ? `'${escapeString(metric.unit)}'` : 'NULL', errorDataVal = metric.error_data ?
                    `'${escapeString(metric.error_data)}'` : 'NULL', errorNameVal = metric.error_name ?
                    `'${escapeString(metric.error_name)}'` : 'NULL', execContextVal = metric.execution_context ?
                    `'${escapeString(metric.execution_context)}'` : 'NULL', vuCountVal = !isNil(metric.vu_count) ?
                    metric.vu_count : 'NULL', statusCodeVal = !isNil(metric.http_status_code) ?
                    metric.http_status_code : 'NULL', hasErrorVal = metric.has_error ? 'TRUE' : 'FALSE', isClientErrorVal = metric.is_client_error ? 'TRUE' : 'FALSE';
                return `(
                        gen_random_uuid(),
                        '${escapeString(metric.run_id)}',
                        ${timestampVal},
                        '${escapeString(metric.metric_name)}',
                        '${escapeString(String(metric.value))}',
                        ${unitVal},
                        '${escapeString(metric.aggregation)}',
                        '${escapeString(metric.sensor_name)}',
                        '${escapeString(metric.sensor_type)}',
                        '${escapeString(metric.sensor_id)}',
                        '${escapeString(metric.method)}',
                        '${escapeString(metric.vu_id)}',
                        ${vuCountVal},
                        '${escapeString(metric.region)}',
                        ${errorDataVal},
                        ${hasErrorVal},
                        ${errorNameVal},
                        ${execContextVal},
                        ${statusCodeVal},
                        ${isClientErrorVal}
                    )`;
            }).join(', '), bulkInsertSQL = `
                    INSERT INTO request_metrics (
                        id, run_id, timestamp, metric_name, value, unit, aggregation,
                        sensor_name, sensor_type, sensor_id, method, vu_id, vu_count, region, error_data,
                        has_error, error_name, execution_context, http_status_code, is_client_error
                    ) VALUES ${values}
                `;
            await this.writeConn.run(bulkInsertSQL);
        }
        catch (error) {
            console.error('✗ Failed to insert request metrics:', error.message);
            throw error;
        }
    }
    async insertResponseDetails(responseDetails) {
        if (!responseDetails || responseDetails.length === 0) {
            return;
        }
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const queries = responseDetails.map((detail) => {
                const requestData = detail.request_data ?
                    `'${detail.request_data.replace(/'/g, '\'\'')}'` : 'NULL', responseData = detail.response_data ?
                    `'${detail.response_data.replace(/'/g, '\'\'')}'` : 'NULL', itemData = detail.item_data ?
                    `'${detail.item_data.replace(/'/g, '\'\'')}'` : 'NULL', virtualUserData = detail.virtual_user_data ?
                    `'${detail.virtual_user_data.replace(/'/g, '\'\'')}'` : 'NULL', errorData = detail.error_data ?
                    `'${detail.error_data.replace(/'/g, '\'\'')}'` : 'NULL';
                return `
                    INSERT INTO response_details (
                        id, run_id, request_id, error_metric_name, response_hash,
                        request_data, response_data, item_data, virtual_user_data,
                        error_data, response_time
                    ) VALUES (
                        gen_random_uuid(),
                        '${detail.run_id}',
                        '${detail.request_id}',
                        '${detail.error_metric_name}',
                        '${detail.response_hash}',
                        ${requestData},
                        ${responseData},
                        ${itemData},
                        ${virtualUserData},
                        ${errorData},
                        ${detail.response_time || 'NULL'}
                    )
                `;
            });
            await Promise.all(queries.map((sql) => {
                return this.writeConn.run(sql);
            }));
        }
        catch (error) {
            console.error('✗ Failed to insert response details:', error.message);
            throw error;
        }
    }
    async insertAssertionDetails(assertionDetails) {
        if (!assertionDetails || assertionDetails.length === 0) {
            return;
        }
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const escapeString = (val) => {
                if (isNil(val)) {
                    return '';
                }
                return String(val).replace(/'/g, '\'\'');
            }, formatTimestamp = (val) => {
                if (!val) {
                    return '';
                }
                try {
                    const d = new Date(val);
                    if (!isNaN(d.getTime())) {
                        return d.toISOString();
                    }
                }
                catch (_) {
                }
                return escapeString(val);
            }, values = assertionDetails.map((detail) => {
                const ts = formatTimestamp(detail.timestamp), timestampVal = ts ? `'${ts}'` : 'NULL', errorNameVal = detail.error_name ?
                    `'${escapeString(detail.error_name)}'` : 'NULL', errorMessageVal = detail.error_message ?
                    `'${escapeString(detail.error_message)}'` : 'NULL', vuIdVal = detail.vu_id ?
                    `'${escapeString(detail.vu_id)}'` : 'NULL', vuCountVal = !isNil(detail.vu_count) ?
                    detail.vu_count : 'NULL', execContextVal = detail.execution_context ?
                    `'${escapeString(detail.execution_context)}'` : 'NULL';
                return `(
                        gen_random_uuid(),
                        '${escapeString(detail.run_id)}',
                        '${escapeString(detail.request_id)}',
                        ${timestampVal},
                        '${escapeString(detail.test_name)}',
                        '${escapeString(detail.test_status)}',
                        ${errorNameVal},
                        ${errorMessageVal},
                        ${vuIdVal},
                        ${vuCountVal},
                        ${execContextVal}
                    )`;
            }).join(', '), bulkInsertSQL = `
                    INSERT INTO assertion_details (
                        id, run_id, request_id, timestamp, test_name, test_status,
                        error_name, error_message, vu_id, vu_count, execution_context
                    ) VALUES ${values}
                `;
            await this.writeConn.run(bulkInsertSQL);
        }
        catch (error) {
            console.error('✗ Failed to insert assertion details:', error.message);
            throw error;
        }
    }
    async insertScriptErrorDetails(scriptErrorDetails) {
        if (!scriptErrorDetails || scriptErrorDetails.length === 0) {
            return;
        }
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const escapeString = (val) => {
                if (isNil(val)) {
                    return '';
                }
                return String(val).replace(/'/g, '\'\'');
            }, formatTimestamp = (val) => {
                if (!val) {
                    return '';
                }
                try {
                    const d = new Date(val);
                    if (!isNaN(d.getTime())) {
                        return d.toISOString();
                    }
                }
                catch (_) {
                }
                return escapeString(val);
            }, values = scriptErrorDetails.map((detail) => {
                const ts = formatTimestamp(detail.timestamp), timestampVal = ts ? `'${ts}'` : 'NULL', errorMessageVal = detail.error_message ?
                    `'${escapeString(detail.error_message)}'` : 'NULL', vuIdVal = detail.vu_id ?
                    `'${escapeString(detail.vu_id)}'` : 'NULL', vuCountVal = !isNil(detail.vu_count) ?
                    detail.vu_count : 'NULL', execContextVal = detail.execution_context ?
                    `'${escapeString(detail.execution_context)}'` : 'NULL';
                return `(
                        gen_random_uuid(),
                        '${escapeString(detail.run_id)}',
                        '${escapeString(detail.request_id)}',
                        ${timestampVal},
                        '${escapeString(detail.script_phase)}',
                        '${escapeString(detail.error_type)}',
                        '${escapeString(detail.error_name)}',
                        ${errorMessageVal},
                        ${vuIdVal},
                        ${vuCountVal},
                        ${execContextVal}
                    )`;
            }).join(', '), bulkInsertSQL = `
                    INSERT INTO script_error_details (
                        id, run_id, request_id, timestamp, script_phase, error_type,
                        error_name, error_message, vu_id, vu_count, execution_context
                    ) VALUES ${values}
                `;
            await this.writeConn.run(bulkInsertSQL);
        }
        catch (error) {
            console.error('✗ Failed to insert script error details:', error.message);
            throw error;
        }
    }
    async getSummaryAggregations(runId, startedAt, finishedAt) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const responseTimeSql = `
                    SELECT
                        AVG(CAST(value AS DOUBLE)) AS avg,
                        MIN(CAST(value AS DOUBLE)) AS min,
                        MAX(CAST(value AS DOUBLE)) AS max,
                        quantile(CAST(value AS DOUBLE), 0.90) AS percentile90,
                        quantile(CAST(value AS DOUBLE), 0.95) AS percentile95,
                        quantile(CAST(value AS DOUBLE), 0.99) AS percentile99
                    FROM request_metrics
                    WHERE run_id = '${runId}' AND metric_name = 'http.response_time'
                `, errorStatsSql = `
                    WITH all_executions AS (
                        SELECT DISTINCT
                            sensor_id as request_id,
                            execution_context,
                            sensor_id || '||' || execution_context as execution_key
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                            AND metric_name = 'http.response_time'
                    ),
                    transport_error_executions AS (
                        SELECT DISTINCT sensor_id || '||' || execution_context as execution_key
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND has_error = TRUE AND execution_context IS NOT NULL
                    ),
                    script_error_executions AS (
                        SELECT DISTINCT ae.execution_key
                        FROM all_executions ae
                        INNER JOIN script_error_details sec
                            ON ae.request_id = sec.request_id
                            AND ae.execution_context = sec.execution_context
                        WHERE sec.run_id = '${runId}'
                    ),
                    executions_with_errors AS (
                        SELECT execution_key FROM transport_error_executions
                        UNION
                        SELECT execution_key FROM script_error_executions
                    )
                    SELECT
                        COUNT(DISTINCT ae.execution_key) as total_executions,
                        COUNT(DISTINCT ee.execution_key) as executions_with_errors,
                        COUNT(DISTINCT te.execution_key) as transport_error_executions,
                        COUNT(DISTINCT se.execution_key) as script_error_executions
                    FROM all_executions ae
                    LEFT JOIN executions_with_errors ee ON ae.execution_key = ee.execution_key
                    LEFT JOIN transport_error_executions te ON ae.execution_key = te.execution_key
                    LEFT JOIN script_error_executions se ON ae.execution_key = se.execution_key
                `, errorTypesSql = `
                    WITH valid_execution_contexts AS (
                        SELECT DISTINCT execution_context
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                    ),
                    transport_errors AS (
                        SELECT
                            COALESCE(error_name, value) AS errorName,
                            'transport' AS errorType,
                            COUNT(DISTINCT execution_context) AS count
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND has_error = TRUE AND execution_context IS NOT NULL
                        GROUP BY 1, 2
                    ),
                    script_errors AS (
                        SELECT
                            error_name AS errorName,
                            'script' AS errorType,
                            COUNT(DISTINCT execution_context) AS count
                        FROM script_error_details
                        WHERE run_id = '${runId}'
                            AND execution_context IN (SELECT execution_context FROM valid_execution_contexts)
                        GROUP BY 1, 2
                    )
                    SELECT * FROM transport_errors
                    UNION ALL
                    SELECT * FROM script_errors
                    ORDER BY count DESC
                `, throughputSql = `
                    SELECT COUNT(*) AS total_requests
                    FROM request_metrics
                    WHERE run_id = '${runId}' AND metric_name = 'http.response_time'
                `, vuSql = `
                    SELECT AVG(CAST(vu_count AS DOUBLE)) AS avg_vu
                    FROM request_metrics
                    WHERE run_id = '${runId}' AND vu_count IS NOT NULL
                `, [responseTimeResult, errorStatsResult, errorTypesResult, throughputResult, vuResult] = await Promise.all([
                this.writeConn.runAndReadAll(responseTimeSql),
                this.writeConn.runAndReadAll(errorStatsSql),
                this.writeConn.runAndReadAll(errorTypesSql),
                this.writeConn.runAndReadAll(throughputSql),
                this.writeConn.runAndReadAll(vuSql)
            ]), responseTime = responseTimeResult.getRowObjects()[0] || {}, errorStats = errorStatsResult.getRowObjects()[0] ||
                { total_executions: 0, executions_with_errors: 0, transport_error_executions: 0, script_error_executions: 0 }, errorTypes = errorTypesResult.getRowObjects(), throughput = throughputResult.getRowObjects()[0] || { total_requests: 0 }, vu = vuResult.getRowObjects()[0] || { avg_vu: 0 }, errorTypesObj = {}, totalExecutions = Number(errorStats.total_executions) || 0, executionsWithErrors = Number(errorStats.executions_with_errors) || 0, transportErrorExecutions = Number(errorStats.transport_error_executions) || 0, scriptErrorExecutions = Number(errorStats.script_error_executions) || 0, errorPercentage = totalExecutions > 0 ?
                (executionsWithErrors * 100.0 / totalExecutions) :
                0, startMs = new Date(startedAt).getTime(), endMs = new Date(finishedAt).getTime(), durationSeconds = Math.max(1, (endMs - startMs) / 1000), rate = Number((Number(throughput.total_requests) /
                durationSeconds).toFixed(2));
            errorTypes.forEach((row) => {
                errorTypesObj[row.errorName] = {
                    count: Number(row.count) || 0,
                    category: row.errorType
                };
            });
            return {
                responseTime: {
                    avg: responseTime.avg !== null ?
                        Math.round(Number(responseTime.avg)) : 0,
                    min: responseTime.min !== null ?
                        Math.round(Number(responseTime.min)) : 0,
                    max: responseTime.max !== null ?
                        Math.round(Number(responseTime.max)) : 0,
                    percentile90: responseTime.percentile90 !== null ?
                        Math.round(Number(responseTime.percentile90)) : 0,
                    percentile95: responseTime.percentile95 !== null ?
                        Math.round(Number(responseTime.percentile95)) : 0,
                    percentile99: responseTime.percentile99 !== null ?
                        Math.round(Number(responseTime.percentile99)) : 0
                },
                errors: {
                    percentage: Number(errorPercentage.toFixed(2)),
                    segmentation: {
                        errorTypes: errorTypesObj
                    },
                    totalExecutions: totalExecutions,
                    transportErrorExecutions: transportErrorExecutions,
                    scriptErrorExecutions: scriptErrorExecutions
                },
                requestThroughput: {
                    sum: Number(throughput.total_requests) || 0,
                    rate: rate
                },
                virtualUsers: {
                    avg: Math.round(Number(vu.avg_vu)) || 0
                }
            };
        }
        catch (error) {
            console.error('✗ Failed to get summary aggregations:', error.message);
            throw error;
        }
    }
    async getAggregationsByRequest(runId, startedAt, finishedAt) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                WITH request_executions AS (
                    SELECT
                        sensor_id as request_id,
                        sensor_name as request_name,
                        method,
                        execution_context,
                        CAST(value AS DOUBLE) as response_time,
                        vu_count,
                        has_error as has_transport_error
                    FROM request_metrics
                    WHERE run_id = '${runId}' AND metric_name = 'http.response_time'
                        AND value IS NOT NULL AND execution_context IS NOT NULL
                ),
                script_error_contexts AS (
                    SELECT DISTINCT request_id, execution_context
                    FROM script_error_details
                    WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                ),
                request_stats AS (
                    SELECT
                        re.request_id as requestId,
                        re.request_name as requestName,
                        re.method,
                        AVG(re.response_time) AS avg,
                        MIN(re.response_time) AS min,
                        MAX(re.response_time) AS max,
                        quantile(re.response_time, 0.90) AS p90,
                        quantile(re.response_time, 0.95) AS p95,
                        quantile(re.response_time, 0.99) AS p99,
                        MAX(re.vu_count) AS virtual_users_count,
                        COUNT(*)::DOUBLE AS throughput,
                        COUNT(DISTINCT re.execution_context) AS total_executions,
                        COUNT(DISTINCT CASE WHEN re.has_transport_error THEN re.execution_context END) AS transport_error_executions,
                        COUNT(DISTINCT CASE WHEN sec.execution_context IS NOT NULL THEN re.execution_context END) AS script_error_executions,
                        COUNT(DISTINCT CASE WHEN re.has_transport_error OR sec.execution_context IS NOT NULL
                              THEN re.execution_context END) AS executions_with_errors
                    FROM request_executions re
                    LEFT JOIN script_error_contexts sec ON re.request_id = sec.request_id AND re.execution_context = sec.execution_context
                    GROUP BY re.request_id, re.request_name, re.method
                )
                SELECT *, (executions_with_errors * 100.0 / NULLIF(total_executions, 0)) AS error_rate
                FROM request_stats
                ORDER BY throughput DESC
            `, errorTypesByRequestSql = `
                    WITH valid_executions AS (
                        SELECT DISTINCT sensor_id as request_id, execution_context
                        FROM request_metrics WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                    ),
                    transport_errors AS (
                        SELECT
                            sensor_id as requestId,
                            COALESCE(error_name, value) as errorName,
                            'transport' as errorType,
                            COUNT(DISTINCT execution_context) as errorCount
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND has_error = TRUE AND execution_context IS NOT NULL
                        GROUP BY 1, 2, 3
                    ),
                    script_errors AS (
                        SELECT
                            se.request_id as requestId,
                            se.error_name as errorName,
                            'script' as errorType,
                            COUNT(DISTINCT se.execution_context) as errorCount
                        FROM script_error_details se
                        INNER JOIN valid_executions ve ON se.request_id = ve.request_id
                            AND se.execution_context = ve.execution_context
                        WHERE se.run_id = '${runId}'
                        GROUP BY 1, 2, 3
                    )
                    SELECT * FROM transport_errors
                    UNION ALL
                    SELECT * FROM script_errors
                    ORDER BY requestId, errorCount DESC
                `, [result, errorTypesResult] = await Promise.all([
                this.writeConn.runAndReadAll(sql),
                this.writeConn.runAndReadAll(errorTypesByRequestSql)
            ]), rows = result.getRowObjects(), errorTypesRows = errorTypesResult.getRowObjects(), errorTypesByRequest = {};
            errorTypesRows.forEach((row) => {
                if (!errorTypesByRequest[row.requestId]) {
                    errorTypesByRequest[row.requestId] = {};
                }
                errorTypesByRequest[row.requestId][row.errorName] = {
                    count: Number(row.errorCount) || 0,
                    category: row.errorType
                };
            });
            const startMs = new Date(startedAt).getTime(), endMs = new Date(finishedAt).getTime(), durationSeconds = Math.max(1, (endMs - startMs) / 1000), groupedByRequest = {};
            for (const row of rows) {
                const totalExecutions = Number(row.total_executions) || 0, transportErrorExecutions = Number(row.transport_error_executions) || 0, scriptErrorExecutions = Number(row.script_error_executions) || 0, errorPercentage = Number(row.error_rate) || 0;
                groupedByRequest[row.requestId] = {
                    aggregatedMetrics: {
                        responseTime: {
                            avg: row.avg !== null ? Math.round(Number(row.avg)) : 0,
                            min: row.min !== null ? Math.round(Number(row.min)) : 0,
                            max: row.max !== null ? Math.round(Number(row.max)) : 0,
                            percentile90: row.p90 !== null ?
                                Math.round(Number(row.p90)) : 0,
                            percentile95: row.p95 !== null ?
                                Math.round(Number(row.p95)) : 0,
                            percentile99: row.p99 !== null ?
                                Math.round(Number(row.p99)) : 0
                        },
                        errors: {
                            percentage: Number(errorPercentage.toFixed(2)),
                            segmentation: {
                                errorTypes: errorTypesByRequest[row.requestId] || {}
                            },
                            totalExecutions: totalExecutions,
                            transportErrorExecutions: transportErrorExecutions,
                            scriptErrorExecutions: scriptErrorExecutions
                        },
                        requestThroughput: {
                            sum: Number(row.throughput) || 0
                        },
                        virtualUsers: {
                            avg: Math.round(Number(row.virtual_users_count)) || 0
                        }
                    }
                };
            }
            return groupedByRequest;
        }
        catch (error) {
            console.error('✗ Failed to get aggregations by request:', error.message);
            throw error;
        }
    }
    async getAggregationsByVu(runId) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                SELECT
                    vu_id as vuId,
                    AVG(CAST(value AS DOUBLE)) AS avg,
                    MIN(CAST(value AS DOUBLE)) AS min,
                    MAX(CAST(value AS DOUBLE)) AS max,
                    quantile(CAST(value AS DOUBLE), 0.90) AS percentile90,
                    quantile(CAST(value AS DOUBLE), 0.95) AS percentile95,
                    quantile(CAST(value AS DOUBLE), 0.99) AS percentile99,
                    COUNT(*) AS total_count,
                    SUM(CASE WHEN has_error = TRUE THEN 1 ELSE 0 END) AS error_count,
                    AVG(CAST(vu_count AS DOUBLE)) AS avg_vu
                FROM request_metrics
                WHERE run_id = '${runId}' AND metric_name = 'http.response_time'
                GROUP BY vu_id
            `, result = await this.writeConn.runAndReadAll(sql), rows = result.getRowObjects(), groupedByVuId = {};
            for (const row of rows) {
                const errorPercentage = row.total_count > 0 ?
                    (Number(row.error_count) * 100.0 / Number(row.total_count)) :
                    0;
                groupedByVuId[row.vuId] = {
                    aggregatedMetrics: {
                        responseTime: {
                            avg: row.avg !== null ? Math.round(Number(row.avg)) : 0,
                            min: row.min !== null ? Math.round(Number(row.min)) : 0,
                            max: row.max !== null ? Math.round(Number(row.max)) : 0,
                            percentile90: row.percentile90 !== null ?
                                Math.round(Number(row.percentile90)) : 0,
                            percentile95: row.percentile95 !== null ?
                                Math.round(Number(row.percentile95)) : 0,
                            percentile99: row.percentile99 !== null ?
                                Math.round(Number(row.percentile99)) : 0
                        },
                        errors: {
                            percentage: Number(errorPercentage.toFixed(2)),
                            segmentation: { errorTypes: {} }
                        },
                        requestThroughput: {
                            sum: Number(row.total_count) || 0
                        },
                        virtualUsers: {
                            avg: Math.round(Number(row.avg_vu)) || 0
                        }
                    }
                };
            }
            return groupedByVuId;
        }
        catch (error) {
            console.error('✗ Failed to get aggregations by VU:', error.message);
            throw error;
        }
    }
    async getTimeSeriesAggregations(runId, bucketSize) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const overallSql = `
                WITH all_executions AS (
                    SELECT
                        to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                        sensor_id as request_id,
                        execution_context,
                        CAST(value AS DOUBLE) as response_time,
                        vu_count,
                        has_error as has_transport_error
                    FROM request_metrics
                    WHERE run_id = '${runId}' AND metric_name = 'http.response_time'
                        AND execution_context IS NOT NULL
                ),
                script_error_contexts AS (
                    SELECT
                        to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                        request_id,
                        execution_context
                    FROM script_error_details
                    WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                )
                SELECT
                    ae.time_window,
                    AVG(ae.response_time) AS avg,
                    MIN(ae.response_time) AS min,
                    MAX(ae.response_time) AS max,
                    quantile(ae.response_time, 0.90) AS percentile90,
                    quantile(ae.response_time, 0.95) AS percentile95,
                    quantile(ae.response_time, 0.99) AS percentile99,
                    COUNT(*) AS total_count,
                    MAX(ae.vu_count) AS max_vu,
                    COUNT(DISTINCT CASE WHEN ae.has_transport_error OR sec.execution_context IS NOT NULL
                          THEN ae.request_id || '||' || ae.execution_context END) AS executions_with_errors,
                    COUNT(DISTINCT ae.request_id || '||' || ae.execution_context) AS total_executions
                FROM all_executions ae
                LEFT JOIN script_error_contexts sec
                    ON ae.time_window = sec.time_window
                    AND ae.request_id = sec.request_id
                    AND ae.execution_context = sec.execution_context
                GROUP BY ae.time_window
                ORDER BY ae.time_window
            `, perRequestSql = `
                    WITH request_executions AS (
                        SELECT
                            to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                            sensor_id as request_id,
                            execution_context,
                            CAST(value AS DOUBLE) as response_time,
                            vu_count,
                            has_error as has_transport_error
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND metric_name = 'http.response_time'
                            AND value IS NOT NULL AND execution_context IS NOT NULL
                    ),
                    script_error_contexts AS (
                        SELECT
                            to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                            request_id,
                            execution_context
                        FROM script_error_details
                        WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                    )
                    SELECT
                        re.time_window,
                        re.request_id as requestId,
                        AVG(re.response_time) AS avg,
                        MIN(re.response_time) AS min,
                        MAX(re.response_time) AS max,
                        quantile(re.response_time, 0.90) AS percentile90,
                        quantile(re.response_time, 0.95) AS percentile95,
                        quantile(re.response_time, 0.99) AS percentile99,
                        COUNT(*) AS total_count,
                        MAX(re.vu_count) AS max_vu,
                        COUNT(DISTINCT re.execution_context) AS total_executions,
                        COUNT(DISTINCT CASE WHEN re.has_transport_error THEN re.execution_context END) AS transport_error_executions,
                        COUNT(DISTINCT CASE WHEN sec.execution_context IS NOT NULL THEN re.execution_context END) AS script_error_executions,
                        COUNT(DISTINCT CASE WHEN re.has_transport_error OR sec.execution_context IS NOT NULL
                              THEN re.execution_context END) AS executions_with_errors
                    FROM request_executions re
                    LEFT JOIN script_error_contexts sec
                        ON re.time_window = sec.time_window
                        AND re.request_id = sec.request_id
                        AND re.execution_context = sec.execution_context
                    GROUP BY re.time_window, re.request_id
                    ORDER BY re.time_window, re.request_id
                `, errorTypesByWindowSql = `
                    WITH valid_executions AS (
                        SELECT DISTINCT sensor_id as request_id, execution_context
                        FROM request_metrics WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                    ),
                    transport_errors AS (
                        SELECT
                            to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                            COALESCE(error_name, value) as errorName,
                            'transport' as errorType,
                            COUNT(DISTINCT execution_context) as errorCount
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND has_error = TRUE AND execution_context IS NOT NULL
                        GROUP BY 1, 2, 3
                    ),
                    script_errors AS (
                        SELECT
                            to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                            se.error_name as errorName,
                            'script' as errorType,
                            COUNT(DISTINCT se.execution_context) as errorCount
                        FROM script_error_details se
                        INNER JOIN valid_executions ve ON se.request_id = ve.request_id
                            AND se.execution_context = ve.execution_context
                        WHERE se.run_id = '${runId}'
                        GROUP BY 1, 2, 3
                    )
                    SELECT * FROM transport_errors
                    UNION ALL
                    SELECT * FROM script_errors
                    ORDER BY 1, errorCount DESC
                `, errorTypesByRequestAndWindowSql = `
                    WITH valid_executions AS (
                        SELECT DISTINCT sensor_id as request_id, execution_context
                        FROM request_metrics WHERE run_id = '${runId}' AND execution_context IS NOT NULL
                    ),
                    transport_errors AS (
                        SELECT
                            to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                            sensor_id as requestId,
                            COALESCE(error_name, value) as errorName,
                            'transport' as errorType,
                            COUNT(DISTINCT execution_context) as errorCount
                        FROM request_metrics
                        WHERE run_id = '${runId}' AND has_error = TRUE AND execution_context IS NOT NULL
                        GROUP BY 1, 2, 3, 4
                    ),
                    script_errors AS (
                        SELECT
                            to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                            se.request_id as requestId,
                            se.error_name as errorName,
                            'script' as errorType,
                            COUNT(DISTINCT se.execution_context) as errorCount
                        FROM script_error_details se
                        INNER JOIN valid_executions ve ON se.request_id = ve.request_id
                            AND se.execution_context = ve.execution_context
                        WHERE se.run_id = '${runId}'
                        GROUP BY 1, 2, 3, 4
                    )
                    SELECT * FROM transport_errors
                    UNION ALL
                    SELECT * FROM script_errors
                    ORDER BY 1, requestId, errorCount DESC
                `, [overallResult, perRequestResult, errorTypesByWindowResult, errorTypesByRequestAndWindowResult] = await Promise.all([
                this.writeConn.runAndReadAll(overallSql),
                this.writeConn.runAndReadAll(perRequestSql),
                this.writeConn.runAndReadAll(errorTypesByWindowSql),
                this.writeConn.runAndReadAll(errorTypesByRequestAndWindowSql)
            ]), overallRows = overallResult.getRowObjects(), perRequestRows = perRequestResult.getRowObjects(), errorTypesByWindowRows = errorTypesByWindowResult.getRowObjects(), errorTypesByRequestAndWindowRows = errorTypesByRequestAndWindowResult.getRowObjects(), errorTypesByWindow = {};
            errorTypesByWindowRows.forEach((row) => {
                const window = new Date(String(row.time_window)).toISOString();
                if (!errorTypesByWindow[window]) {
                    errorTypesByWindow[window] = {};
                }
                errorTypesByWindow[window][row.errorName] = {
                    count: Number(row.errorCount) || 0,
                    category: row.errorType
                };
            });
            const errorTypesByRequestAndWindow = {};
            errorTypesByRequestAndWindowRows.forEach((row) => {
                const window = new Date(String(row.time_window)).toISOString();
                if (!errorTypesByRequestAndWindow[window]) {
                    errorTypesByRequestAndWindow[window] = {};
                }
                if (!errorTypesByRequestAndWindow[window][row.requestId]) {
                    errorTypesByRequestAndWindow[window][row.requestId] = {};
                }
                errorTypesByRequestAndWindow[window][row.requestId][row.errorName] = {
                    count: Number(row.errorCount) || 0,
                    category: row.errorType
                };
            });
            const perRequestByWindow = {};
            perRequestRows.forEach((row) => {
                const window = new Date(String(row.time_window)).toISOString();
                if (!perRequestByWindow[window]) {
                    perRequestByWindow[window] = {};
                }
                const totalExecutions = Number(row.total_executions) || 0, executionsWithErrors = Number(row.executions_with_errors) || 0, transportErrorExecutions = Number(row.transport_error_executions) || 0, scriptErrorExecutions = Number(row.script_error_executions) || 0, errorPercentage = totalExecutions > 0 ?
                    (executionsWithErrors * 100.0 / totalExecutions) :
                    0;
                perRequestByWindow[window][row.requestId] = {
                    aggregatedMetrics: {
                        responseTime: {
                            avg: row.avg !== null ? Math.round(Number(row.avg)) : 0,
                            min: row.min !== null ? Math.round(Number(row.min)) : 0,
                            max: row.max !== null ? Math.round(Number(row.max)) : 0,
                            percentile90: row.percentile90 !== null ?
                                Math.round(Number(row.percentile90)) : 0,
                            percentile95: row.percentile95 !== null ?
                                Math.round(Number(row.percentile95)) : 0,
                            percentile99: row.percentile99 !== null ?
                                Math.round(Number(row.percentile99)) : 0
                        },
                        errors: {
                            percentage: Number(errorPercentage.toFixed(2)),
                            segmentation: {
                                errorTypes: (errorTypesByRequestAndWindow[window] &&
                                    errorTypesByRequestAndWindow[window][row.requestId]) ||
                                    {}
                            },
                            totalExecutions: totalExecutions,
                            transportErrorExecutions: transportErrorExecutions,
                            scriptErrorExecutions: scriptErrorExecutions
                        },
                        requestThroughput: {
                            sum: Number(row.total_count) || 0
                        },
                        virtualUsers: {
                            avg: Math.round(Number(row.max_vu)) || 0
                        }
                    }
                };
            });
            const timeSeries = overallRows.map((row) => {
                const window = new Date(String(row.time_window)).toISOString(), totalExecutions = Number(row.total_executions) || null, executionsWithErrors = Number(row.executions_with_errors) || null, errorPercentage = totalExecutions > 0 ?
                    (executionsWithErrors * 100.0 / totalExecutions) :
                    0;
                return {
                    window,
                    aggregatedMetrics: {
                        responseTime: {
                            avg: row.avg !== null ? Math.round(Number(row.avg)) : 0,
                            min: row.min !== null ? Math.round(Number(row.min)) : 0,
                            max: row.max !== null ? Math.round(Number(row.max)) : 0,
                            percentile90: row.percentile90 !== null ?
                                Math.round(Number(row.percentile90)) : 0,
                            percentile95: row.percentile95 !== null ?
                                Math.round(Number(row.percentile95)) : 0,
                            percentile99: row.percentile99 !== null ?
                                Math.round(Number(row.percentile99)) : 0
                        },
                        errors: {
                            percentage: Number(errorPercentage.toFixed(2)),
                            segmentation: {
                                errorTypes: errorTypesByWindow[window] || {}
                            },
                            totalExecutions: totalExecutions,
                            transportErrorExecutions: null,
                            scriptErrorExecutions: null
                        },
                        requestThroughput: {
                            sum: Number(row.total_count) || 0
                        },
                        virtualUsers: {
                            avg: Math.round(Number(row.max_vu)) || 0
                        }
                    },
                    groupedByRequest: perRequestByWindow[window] || {}
                };
            });
            return timeSeries;
        }
        catch (error) {
            console.error('✗ Failed to get time series aggregations:', error.message);
            throw error;
        }
    }
    async calculateBucketSize(runId) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                SELECT
                    MIN(timestamp) as start_time,
                    MAX(timestamp) as end_time
                FROM request_metrics
                WHERE run_id = '${runId}'
            `;
            const result = await this.writeConn.runAndReadAll(sql), row = result.getRowObjects()[0];
            if (!row || !row.start_time || !row.end_time) {
                return 1;
            }
            const startMs = new Date(String(row.start_time)).getTime(), endMs = new Date(String(row.end_time)).getTime();
            if (isNaN(startMs) || isNaN(endMs) || endMs <= startMs) {
                return 1;
            }
            const durationSeconds = Math.max(1, Math.floor((endMs - startMs) / 1000)), bucketSize = Math.max(1, Math.ceil(durationSeconds / TARGET_TIME_SERIES_BUCKETS));
            return bucketSize;
        }
        catch (error) {
            console.error('✗ Failed to calculate bucket size:', error.message);
            return 1;
        }
    }
    async getRequestFailureRate(runId) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                WITH all_request_executions AS (
                    SELECT DISTINCT
                        sensor_id as request_id,
                        execution_context
                    FROM request_metrics
                    WHERE run_id = '${runId}'
                        AND execution_context IS NOT NULL
                ),
                executions_with_failures AS (
                    SELECT DISTINCT
                        request_id,
                        execution_context
                    FROM assertion_details
                    WHERE run_id = '${runId}'
                        AND test_status = 'fail'
                        AND execution_context IS NOT NULL
                )
                SELECT
                    COUNT(DISTINCT are.request_id || '||' || are.execution_context) as total_requests,
                    COUNT(DISTINCT CASE
                        WHEN ewf.execution_context IS NOT NULL
                        THEN are.request_id || '||' || are.execution_context
                    END) as requests_with_failed_assertions
                FROM all_request_executions are
                LEFT JOIN executions_with_failures ewf
                    ON are.request_id = ewf.request_id
                    AND are.execution_context = ewf.execution_context
            `;
            const result = await this.writeConn.runAndReadAll(sql);
            const row = result.getRowObjects()[0];
            if (!row) {
                return {
                    totalRequests: 0,
                    requestsWithFailedAssertions: 0,
                    failurePercentage: 0
                };
            }
            const totalRequests = Number(row.total_requests) || 0;
            const requestsWithFailedAssertions = Number(row.requests_with_failed_assertions) || 0;
            const failurePercentage = totalRequests > 0 ?
                (requestsWithFailedAssertions / totalRequests) * 100 : 0;
            return {
                totalRequests,
                requestsWithFailedAssertions,
                failurePercentage: Number(failurePercentage.toFixed(2))
            };
        }
        catch (error) {
            console.error('✗ Error getting request failure rate:', error.message);
            return {
                totalRequests: 0,
                requestsWithFailedAssertions: 0,
                failurePercentage: 0
            };
        }
    }
    async getRequestFailureRateByRequest(runId) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                WITH all_request_executions AS (
                    SELECT
                        rm.sensor_id as request_id,
                        rm.sensor_name as request_name,
                        rm.method as request_method,
                        rm.execution_context
                    FROM request_metrics rm
                    WHERE rm.run_id = '${runId}'
                        AND rm.execution_context IS NOT NULL
                ),
                executions_with_failures AS (
                    SELECT DISTINCT
                        ad.request_id,
                        ad.execution_context
                    FROM assertion_details ad
                    WHERE ad.run_id = '${runId}'
                        AND ad.test_status = 'fail'
                        AND ad.execution_context IS NOT NULL
                ),
                assertion_counts AS (
                    SELECT
                        request_id,
                        COUNT(CASE WHEN test_status = 'pass' THEN 1 END) as pass_count,
                        COUNT(CASE WHEN test_status = 'fail' THEN 1 END) as fail_count,
                        COUNT(CASE WHEN test_status = 'skip' THEN 1 END) as skip_count,
                        COUNT(*) as total_count
                    FROM assertion_details
                    WHERE run_id = '${runId}'
                    GROUP BY request_id
                )
                SELECT
                    are.request_id as requestId,
                    are.request_name as requestName,
                    are.request_method as requestMethod,
                    COUNT(DISTINCT are.execution_context) as totalExecutions,
                    COUNT(DISTINCT CASE
                        WHEN ewf.execution_context IS NOT NULL
                        THEN are.execution_context
                    END) as executionsWithFailedAssertions,
                    COALESCE(ac.pass_count, 0) as passCount,
                    COALESCE(ac.fail_count, 0) as failCount,
                    COALESCE(ac.skip_count, 0) as skipCount,
                    COALESCE(ac.total_count, 0) as totalCount
                FROM all_request_executions are
                LEFT JOIN executions_with_failures ewf
                    ON are.request_id = ewf.request_id
                    AND are.execution_context = ewf.execution_context
                LEFT JOIN assertion_counts ac
                    ON are.request_id = ac.request_id
                GROUP BY are.request_id, are.request_name, are.request_method,
                         ac.pass_count, ac.fail_count, ac.skip_count, ac.total_count
                ORDER BY are.request_name
            `;
            const result = await this.writeConn.runAndReadAll(sql);
            const rows = result.getRowObjects();
            return rows.map((row) => {
                const totalExecutions = Number(row.totalExecutions) || 0;
                const executionsWithFailedAssertions = Number(row.executionsWithFailedAssertions) || 0;
                const failurePercentage = totalExecutions > 0 ?
                    (executionsWithFailedAssertions / totalExecutions) * 100 : 0;
                return {
                    requestId: row.requestId,
                    requestName: row.requestName,
                    requestMethod: row.requestMethod,
                    totalExecutions,
                    executionsWithFailedAssertions,
                    failurePercentage: Number(failurePercentage.toFixed(2)),
                    passCount: Number(row.passCount) || 0,
                    failCount: Number(row.failCount) || 0,
                    skipCount: Number(row.skipCount) || 0,
                    totalCount: Number(row.totalCount) || 0
                };
            });
        }
        catch (error) {
            console.error('✗ Error getting request failure rate by request:', error.message);
            return [];
        }
    }
    async getTestPassRateSummary(runId) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                SELECT
                    COUNT(*) as totalTests,
                    COUNT(CASE WHEN test_status = 'pass' THEN 1 END) as passedTests,
                    COUNT(CASE WHEN test_status = 'fail' THEN 1 END) as failedTests,
                    COUNT(CASE WHEN test_status = 'skip' THEN 1 END) as skippedTests
                FROM assertion_details
                WHERE run_id = '${runId}'
            `;
            const result = await this.writeConn.runAndReadAll(sql);
            const row = result.getRowObjects()[0];
            if (!row) {
                return {
                    totalTests: 0,
                    passedTests: 0,
                    failedTests: 0,
                    skippedTests: 0,
                    passRate: 0
                };
            }
            const totalTests = Number(row.totalTests) || 0;
            const passedTests = Number(row.passedTests) || 0;
            const failedTests = Number(row.failedTests) || 0;
            const skippedTests = Number(row.skippedTests) || 0;
            const totalExecutedTests = passedTests + failedTests;
            const passRate = totalExecutedTests > 0 ? (passedTests / totalExecutedTests) * 100 : 0;
            return {
                totalTests,
                passedTests,
                failedTests,
                skippedTests,
                passRate: Number(passRate.toFixed(2))
            };
        }
        catch (error) {
            console.error('✗ Error getting test pass rate summary:', error.message);
            return {
                totalTests: 0,
                passedTests: 0,
                failedTests: 0,
                skippedTests: 0,
                passRate: 0
            };
        }
    }
    async getTestFailureTimeSeries(runId, bucketSize) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                WITH all_executions_by_window AS (
                    SELECT
                        to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                        sensor_id as request_id,
                        execution_context,
                        vu_id
                    FROM request_metrics
                    WHERE run_id = '${runId}'
                        AND execution_context IS NOT NULL
                ),
                failed_executions_by_window AS (
                    SELECT DISTINCT
                        to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                        request_id,
                        execution_context
                    FROM assertion_details
                    WHERE run_id = '${runId}'
                        AND test_status = 'fail'
                        AND execution_context IS NOT NULL
                )
                SELECT
                    ae.time_window,
                    COUNT(DISTINCT CASE
                        WHEN fe.execution_context IS NOT NULL
                        THEN ae.request_id || '||' || ae.execution_context
                    END) as failed_request_count,
                    COUNT(DISTINCT ae.vu_id) as vu_count
                FROM all_executions_by_window ae
                LEFT JOIN failed_executions_by_window fe
                    ON ae.time_window = fe.time_window
                    AND ae.request_id = fe.request_id
                    AND ae.execution_context = fe.execution_context
                GROUP BY ae.time_window
                ORDER BY ae.time_window
            `;
            const result = await this.writeConn.runAndReadAll(sql);
            const rows = result.getRowObjects();
            return rows.map((row) => {
                const timeWindow = new Date(String(row.time_window));
                return {
                    window: timeWindow.getTime(),
                    failed_request_count: Number(row.failed_request_count) || 0,
                    vu_count: Number(row.vu_count) || 0
                };
            });
        }
        catch (error) {
            console.error('✗ Error getting test failure time series:', error.message);
            return [];
        }
    }
    async getTestFailureByRequestTimeSeries(runId, bucketSize) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                WITH window_data AS (
                    SELECT
                        to_timestamp(floor(epoch(timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                        sensor_id as request_id,
                        execution_context,
                        vu_id
                    FROM request_metrics
                    WHERE run_id = '${runId}'
                        AND execution_context IS NOT NULL
                ),
                failed_assertions AS (
                    SELECT
                        to_timestamp(floor(epoch(ad.timestamp) / ${bucketSize}) * ${bucketSize}) AS time_window,
                        ad.request_id,
                        COUNT(*) as assertion_fail_count
                    FROM assertion_details ad
                    WHERE ad.run_id = '${runId}'
                        AND ad.test_status = 'fail'
                        AND ad.execution_context IS NOT NULL
                    GROUP BY time_window, ad.request_id
                )
                SELECT
                    wd.request_id as requestId,
                    wd.time_window,
                    COALESCE(fa.assertion_fail_count, 0) as failed_assertion_count,
                    COUNT(DISTINCT wd.vu_id) as vu_count
                FROM window_data wd
                LEFT JOIN failed_assertions fa
                    ON wd.time_window = fa.time_window
                    AND wd.request_id = fa.request_id
                GROUP BY wd.request_id, wd.time_window, fa.assertion_fail_count
                ORDER BY wd.request_id, wd.time_window
            `;
            const result = await this.writeConn.runAndReadAll(sql);
            const rows = result.getRowObjects();
            const byRequest = {};
            rows.forEach((row) => {
                const requestId = row.requestId;
                if (!byRequest[requestId]) {
                    byRequest[requestId] = [];
                }
                const timeWindow = new Date(String(row.time_window));
                byRequest[requestId].push({
                    window: timeWindow.getTime(),
                    failed_assertion_count: Number(row.failed_assertion_count) || 0,
                    vu_count: Number(row.vu_count) || 0
                });
            });
            return byRequest;
        }
        catch (error) {
            console.error('✗ Error getting test failure by request time series:', error.message);
            return {};
        }
    }
    async getTestDetailsByRequest(runId, requestId) {
        if (!this.initialized || !this.writeConn) {
            await this.init();
        }
        try {
            const sql = `
                SELECT
                    ad.request_id as requestId,
                    rm.sensor_name as requestName,
                    rm.method as requestMethod,
                    ad.test_name as testName,
                    MAX(CASE WHEN ad.test_status = 'fail' THEN ad.error_name END) as errorName,
                    MAX(CASE WHEN ad.test_status = 'fail' THEN ad.error_message END) as errorMessage,
                    COUNT(CASE WHEN ad.test_status = 'fail' THEN 1 END) as failCount,
                    COUNT(CASE WHEN ad.test_status = 'pass' THEN 1 END) as passCount,
                    COUNT(*) as totalCount
                FROM assertion_details ad
                INNER JOIN (
                    SELECT DISTINCT run_id, sensor_id, sensor_name, method
                    FROM request_metrics
                ) rm ON ad.run_id = rm.run_id AND ad.request_id = rm.sensor_id
                WHERE ad.run_id = '${runId}'
                    AND ad.request_id = '${requestId}'
                GROUP BY ad.request_id, rm.sensor_name, rm.method, ad.test_name
                HAVING COUNT(CASE WHEN ad.test_status = 'fail' THEN 1 END) > 0
                ORDER BY failCount DESC, ad.test_name
            `;
            const result = await this.writeConn.runAndReadAll(sql);
            const rows = result.getRowObjects();
            return rows.map((row) => ({
                requestId: row.requestId,
                requestName: row.requestName,
                requestMethod: row.requestMethod,
                testName: row.testName,
                errorName: row.errorName,
                errorMessage: row.errorMessage,
                failCount: Number(row.failCount) || 0,
                passCount: Number(row.passCount) || 0,
                totalCount: Number(row.totalCount) || 0
            }));
        }
        catch (error) {
            console.error('✗ Error getting test details by request:', error.message);
            return [];
        }
    }
    close() {
        try {
            if (this.writeConn) {
                this.writeConn.disconnectSync();
                this.writeConn = null;
            }
            if (this.dbInstance) {
                this.dbInstance.closeSync();
                this.dbInstance = null;
            }
            this.initialized = false;
        }
        catch (error) {
            console.error('✗ Error closing DuckDB:', error.message);
        }
    }
}
module.exports = DuckDBService;
