// @postman/pm-bin-linux-x64/bin/lib/perftest/index.js
const _ = require('lodash'), colors = require('colors'), Table = require('cli-table3'), Perftest = require('@postman/performance-test'), print = require('../framework/print'), { getMachineSpecs, flattenCollection, getCollectionStructure, mergeRunConfig } = require('./utils'), getOptions = require('../run/options'), { createPerformanceTest, savePerformanceTestResults } = require('./perftest-service'), { parsePassIfCondition, evaluatePassIfCondition, buildChecksPayload } = require('../commands/performance/pass-if'), DuckDBService = require('./duckdb-service'), POSTMAN_UID_REGEX = /^([0-9A-Z]+)-([0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12})$/i;
function formatTime(seconds) {
    const mins = Math.floor(seconds / 60);
    const secs = Math.floor(seconds % 60);
    return `${mins.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
}
function formatNumber(num) {
    return num.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}
function displayCompletionSummary(runNumber, runId, metrics, durationSeconds, totalRequests, collectionId) {
    const finalRps = durationSeconds > 0 ? (totalRequests / durationSeconds).toFixed(2) : '0.00';
    const actualDuration = formatTime(durationSeconds || 0);
    const avgResponseTime = metrics?.responseTime?.avg?.toFixed(2) || 'N/A';
    const p90 = metrics?.responseTime?.percentile90?.toFixed(2) || 'N/A';
    const p95 = metrics?.responseTime?.percentile95?.toFixed(2) || 'N/A';
    const p99 = metrics?.responseTime?.percentile99?.toFixed(2) || 'N/A';
    const errorRate = metrics?.errors?.percentage?.toFixed(2) || '0.00';
    const testFailureRate = metrics?.testFailures?.percentage?.toFixed(2) || '0.00';
    const errorColor = parseFloat(errorRate) > 5 ? 'red' : 'green';
    const errorText = errorColor === 'red' ?
        colors.bold(colors.red(errorRate + '%')) :
        colors.bold(colors.green(errorRate + '%'));
    const testFailureColor = parseFloat(testFailureRate) > 5 ? 'red' : 'green';
    const testFailureText = testFailureColor === 'red' ?
        colors.bold(colors.red(testFailureRate + '%')) :
        colors.bold(colors.green(testFailureRate + '%'));
    const summaryTable = new Table({
        style: { head: [] },
        colAligns: ['right', 'right'],
        colWidths: [35, 35]
    });
    summaryTable.push([{
            colSpan: 2,
            content: colors.bold(colors.reset(`Test Run: #${runNumber}`)) + ` | Run ID: ${runId || 'N/A'}`,
            hAlign: 'left'
        }]);
    summaryTable.push([{ colSpan: 2, content: colors.reset('📊 Performance Summary'), hAlign: 'left' }], ['  Total Requests', colors.bold(colors.reset(formatNumber(totalRequests)))], ['  Average Request Per Second', colors.bold(colors.reset(finalRps + ' req/s'))], ['  Test Duration', colors.bold(colors.reset(actualDuration))]);
    summaryTable.push([{ colSpan: 2, content: colors.reset('⏱  Response Time'), hAlign: 'left' }], ['  99th Percentile (p99)', colors.bold(colors.reset(p99 + ' ms'))], ['  95th Percentile (p95)', colors.bold(colors.reset(p95 + ' ms'))], ['  90th Percentile (p90)', colors.bold(colors.reset(p90 + ' ms'))], ['  Average:', colors.bold(colors.reset(avgResponseTime + ' ms'))]);
    summaryTable.push([{ colSpan: 2, content: colors.reset('🎯 Quality'), hAlign: 'left' }], ['  Error Percentage', errorText], ['  Test Failure Rate', testFailureText]);
    print.lf();
    print(summaryTable.toString());
    print.lf();
    const regionUtil = require('../region-util');
    const artemisBaseUrl = regionUtil.getARTEMISBaseURL();
    const baseUrl = artemisBaseUrl.replace(/\/\/go\.eu\./g, '//go.');
    const performanceRunsUrl = `${baseUrl}/run/performance/${runId}?static=true`;
    print.lf('View on Postman for detailed reports and analytics:', colors.grey(performanceRunsUrl));
}
function enrichAggregatedMetrics(aggregatedMetrics, startedAt, finishedAt) {
    if (aggregatedMetrics?.summary?.aggregatedMetrics) {
        const startTime = new Date(startedAt).getTime();
        const endTime = new Date(finishedAt).getTime();
        const durationSeconds = Math.floor((endTime - startTime) / 1000);
        const totalRequests = aggregatedMetrics.summary.aggregatedMetrics.requestThroughput?.sum || 0;
        const rps = durationSeconds > 0 ? (totalRequests / durationSeconds) : 0;
        aggregatedMetrics.summary.aggregatedMetrics.requestThroughput.rate = Number(rps.toFixed(2));
    }
    return aggregatedMetrics;
}
async function aggregateMetricsFromDuckDB(duckDBService, runId, startedAt, finishedAt, testError) {
    try {
        const bucketSize = await duckDBService.calculateBucketSize(runId);
        const [summaryMetrics, groupedByRequest, groupedByVuId, timeSeries, requestFailureRate, requestFailureRateByRequest, testFailureTimeSeries, testFailureByRequestTimeSeries] = await Promise.all([
            duckDBService.getSummaryAggregations(runId, startedAt, finishedAt),
            duckDBService.getAggregationsByRequest(runId, startedAt, finishedAt),
            duckDBService.getAggregationsByVu(runId),
            duckDBService.getTimeSeriesAggregations(runId, bucketSize),
            duckDBService.getRequestFailureRate(runId),
            duckDBService.getRequestFailureRateByRequest(runId),
            duckDBService.getTestFailureTimeSeries(runId, bucketSize),
            duckDBService.getTestFailureByRequestTimeSeries(runId, bucketSize)
        ]);
        summaryMetrics.testFailures = {
            percentage: requestFailureRate.failurePercentage,
            totalRequests: requestFailureRate.totalRequests,
            requestsWithFailedAssertions: requestFailureRate.requestsWithFailedAssertions
        };
        requestFailureRateByRequest.forEach((reqData) => {
            if (groupedByRequest[reqData.requestId]) {
                groupedByRequest[reqData.requestId].aggregatedMetrics.testFailures = {
                    percentage: reqData.failurePercentage,
                    totalExecutions: reqData.totalExecutions,
                    executionsWithFailedAssertions: reqData.executionsWithFailedAssertions,
                    passCount: reqData.passCount,
                    failCount: reqData.failCount,
                    skipCount: reqData.skipCount,
                    totalCount: reqData.totalCount,
                    requestName: reqData.requestName,
                    requestMethod: reqData.requestMethod
                };
            }
        });
        const testDetailsByRequest = {};
        for (const reqData of requestFailureRateByRequest) {
            if (reqData.failCount > 0) {
                const testDetails = await duckDBService.getTestDetailsByRequest(runId, reqData.requestId);
                testDetailsByRequest[reqData.requestId] = testDetails.map((detail) => ({
                    testName: detail.testName,
                    errorName: detail.errorName,
                    errorMessage: detail.errorMessage,
                    failCount: detail.failCount,
                    passCount: detail.passCount,
                    totalCount: detail.totalCount
                }));
            }
        }
        const state = testError ? 'finished:error' : 'finished', aggregatedResult = {
            summary: {
                aggregatedMetrics: summaryMetrics,
                groupedByRequest: groupedByRequest,
                groupedByVuId: groupedByVuId,
                testMetrics: {
                    overallSummary: null,
                    testFailureTimeSeries: testFailureTimeSeries,
                    testFailureByRequestTimeSeries: testFailureByRequestTimeSeries,
                    testDetailsByRequest: testDetailsByRequest
                }
            },
            timeSeries: timeSeries,
            state: state,
            startedAt: startedAt,
            finishedAt: finishedAt
        };
        return aggregatedResult;
    }
    catch (error) {
        console.error('✗ Failed to aggregate metrics from DuckDB:', error.message);
        return {
            summary: {
                aggregatedMetrics: {
                    responseTime: { avg: 0, min: 0, max: 0, percentile90: 0, percentile95: 0, percentile99: 0 },
                    errors: { percentage: 0, segmentation: { errorTypes: {} } },
                    requestThroughput: { sum: 0, rate: 0 },
                    virtualUsers: { avg: 0 },
                    testFailures: { percentage: 0, totalRequests: 0, requestsWithFailedAssertions: 0 }
                },
                groupedByRequest: {},
                groupedByVuId: {},
                testMetrics: {
                    overallSummary: null,
                    testFailureTimeSeries: [],
                    testFailureByRequestTimeSeries: {},
                    testDetailsByRequest: {}
                }
            },
            timeSeries: [],
            state: testError ? 'finished:error' : 'finished',
            startedAt: startedAt,
            finishedAt: finishedAt
        };
    }
}
async function executeRun(runConfig, options, outputHandler, callback) {
    const { name, environment, collectionId, globals, loadProfile } = runConfig;
    const runOptions = {
        collection: collectionId,
        collectionUID: collectionId,
        environment: environment,
        ...options
    };
    getOptions(runOptions, async function (err, config) {
        if (err) {
            return callback(err);
        }
        if (!_.isObject(config.collection)) {
            return callback(new Error('expecting a collection to run'));
        }
        try {
            await runPerftest(config, loadProfile, runConfig, options, outputHandler, callback);
        }
        catch (error) {
            callback(new Error(`Error running performance test: ${error.message}`));
        }
    });
}
async function runPerftest(config, loadProfile, runConfig, options, outputHandler, callback) {
    const { collectionUID, collection, environment, globals } = config, duckDBService = new DuckDBService();
    outputHandler.printLog(`collectionUID: ${collectionUID}`, 'info');
    try {
        await duckDBService.init();
    }
    catch (error) {
        outputHandler.printLog(`Failed to initialize DuckDB: ${error.message}`, 'error');
        return callback(new Error(`Cannot initialize DuckDB for metrics storage: ${error.message}`));
    }
    let optionsPerf = {
        collection: collection.toJSON(),
        environment: environment ? JSON.parse(JSON.stringify(environment)) : {},
        globals: globals ? JSON.parse(JSON.stringify(globals)) : {},
        startedAt: new Date().toISOString(),
        loadProfile: loadProfile
    }, originalExecArgv;
    if (runConfig.dataFileConfig) {
        optionsPerf.dataFileConfig = runConfig.dataFileConfig;
    }
    if (typeof process.pkg !== 'undefined') {
        originalExecArgv = process.execArgv.slice();
        outputHandler.printLog(`[pkg] Original process.execArgv: ${JSON.stringify(originalExecArgv)}`, 'info');
    }
    const match = POSTMAN_UID_REGEX.exec(collectionUID || ''), createdBy = match ? match[1] : undefined;
    if (createdBy) {
        _.set(config, 'createdBy', createdBy);
    }
    else {
        return callback(new Error('error while parsing createbBy'));
    }
    const peakVUs = runConfig.loadProfile.peakVUs ||
        Math.max(...loadProfile.controlPoints.map(cp => cp.load.count));
    const createReqObj = {
        name: runConfig.name,
        collection: {
            id: collectionUID
        },
        runSource: 'cli',
        machineSpecs: getMachineSpecs(),
        options: {
            virtualUsers: peakVUs,
            profile: loadProfile,
            requestSelection: {
                selectedItems: flattenCollection(optionsPerf.collection.item, config.createdBy),
                meta: {
                    collectionStructure: getCollectionStructure(optionsPerf.collection.item, config.createdBy)
                }
            }
        }
    };
    if (options.passIfCondition) {
        createReqObj.checks = buildChecksPayload(options.passIfCondition);
    }
    outputHandler.printLog('Creating performance test run...', 'info');
    let data, runNumber;
    try {
        data = await createPerformanceTest({
            data: createReqObj,
            postmanApiKey: config.postmanApiKey,
            accessToken: config.accessToken
        });
        optionsPerf.runId = data.id;
        runNumber = data.runNumber;
    }
    catch (error) {
        outputHandler.printLog(`Failed to create performance test: ${JSON.stringify(error, null, 2)}`, 'error');
        return callback(new Error('Cannot create performance test run'));
    }
    let test = await new Perftest().create(optionsPerf);
    async function cleanup() {
        if (progressInterval) {
            clearInterval(progressInterval);
        }
        if (test) {
            test.removeAllListeners('init');
            test.removeAllListeners('start');
            test.removeAllListeners('systemMetricsBreached');
            test.removeAllListeners('log');
            test.removeAllListeners('receiveOutput');
            test.removeAllListeners('receiveAssertionMetrics');
            test = null;
        }
        if (duckDBService) {
            await duckDBService.close();
        }
        if (typeof process.pkg !== 'undefined' && originalExecArgv) {
            process.execArgv = originalExecArgv;
            outputHandler.printLog('[pkg] Restored process.execArgv', 'info');
        }
        optionsPerf = null;
    }
    if (!test) {
        callback(new Error('cannot create peformance test run'));
        return;
    }
    outputHandler.printLog(`Performance test created - id: #${runNumber} - ${optionsPerf.runId}`, 'info');
    test.on('init', (err) => {
        if (err) {
            outputHandler.printLog(`Initialization error: ${err.message}`, 'error');
            return;
        }
        outputHandler.printInit();
    });
    test.on('start', (err) => {
        if (err) {
            outputHandler.printLog(err.message || err, 'error');
            return;
        }
        outputHandler.printStart();
    });
    test.on('systemMetricsBreached', (data) => {
        const { metric, value, threshold, type, consecutiveBreaches } = data;
        outputHandler.printSystemMetricsBreached(data);
        if (type === 'critical') {
            test.stop();
            cleanup();
            callback(new Error(`Performance test stopped due to critical system metrics breach: ${metric} = ${value}`));
        }
    });
    test.on('log', (log) => {
        if (log && log.message) {
            const level = log.level || 'info';
            outputHandler.printLog(log.message, level);
        }
    });
    let totalRequests = 0;
    let totalErrors = 0;
    let totalResponseTime = 0;
    const testStartTime = Date.now();
    const testDurationMs = loadProfile.options.totalDuration.millis;
    const progressInterval = setInterval(() => {
        const elapsed = Date.now() - testStartTime;
        const progress = Math.min(100, (elapsed / testDurationMs) * 100);
        const avgResponseTime = totalRequests > 0 ? totalResponseTime / totalRequests : 0;
        outputHandler.updateProgress({
            totalRequests,
            errorCount: totalErrors,
            avgResponseTime,
            currentVUs: test.actualVUCount,
            progress
        });
        if (progress >= 100) {
            clearInterval(progressInterval);
        }
    }, 500);
    test.recieveOutput(async (data) => {
        try {
            const { transformedMetrics, responseDetails } = DuckDBService.processRawMetrics(data);
            if (transformedMetrics.length > 0) {
                transformedMetrics.forEach(metric => {
                    outputHandler.printLog(JSON.stringify(metric), 'debug');
                    totalRequests++;
                    totalResponseTime += metric.value || 0;
                    if (metric.has_error) {
                        outputHandler.printLog(`Error making request: ${metric.error_name}`, 'error');
                        totalErrors++;
                    }
                });
                await duckDBService.insertRequestMetrics(transformedMetrics);
            }
            if (responseDetails.length > 0) {
                await duckDBService.insertResponseDetails(responseDetails);
            }
            const elapsed = Date.now() - testStartTime;
            const progress = Math.min(100, (elapsed / testDurationMs) * 100);
            const avgResponseTime = totalRequests > 0 ? totalResponseTime / totalRequests : 0;
            outputHandler.updateProgress({
                totalRequests,
                errorCount: totalErrors,
                avgResponseTime,
                currentVUs: test.actualVUCount,
                progress
            });
        }
        catch (error) {
            outputHandler.printLog(`Failed to store metrics in DuckDB: ${error.message}`, 'error');
        }
    });
    test.receiveAssertionMetrics(async (data) => {
        try {
            const assertionDetails = DuckDBService.processAssertionMetrics(data);
            if (assertionDetails.length > 0) {
                await duckDBService.insertAssertionDetails(assertionDetails);
            }
        }
        catch (error) {
            outputHandler.printLog(`Failed to store assertions in DuckDB: ${error.message}`, 'error');
        }
    });
    test.receiveScriptErrorMetrics(async (data) => {
        try {
            const scriptErrorDetails = DuckDBService.processScriptErrorMetrics(data);
            if (scriptErrorDetails.length > 0) {
                await duckDBService.insertScriptErrorDetails(scriptErrorDetails);
            }
        }
        catch (error) {
            outputHandler.printLog(`Failed to store script errors in DuckDB: ${error.message}`, 'error');
        }
    });
    test.start(async (error) => {
        const finishedAt = new Date().toISOString();
        if (error) {
            outputHandler.printLog('Aggregating metrics from DuckDB...', 'info');
            let aggregatedMetrics = null;
            try {
                aggregatedMetrics = await aggregateMetricsFromDuckDB(duckDBService, optionsPerf.runId, optionsPerf.startedAt, finishedAt, error);
                enrichAggregatedMetrics(aggregatedMetrics, optionsPerf.startedAt, finishedAt);
                outputHandler.printLog(`Performance test saving results with error state id: #${runNumber}`, 'info');
                try {
                    await savePerformanceTestResults({
                        perfRunId: optionsPerf.runId,
                        data: aggregatedMetrics,
                        postmanApiKey: config.postmanApiKey,
                        accessToken: config.accessToken
                    });
                    outputHandler.printLog(`Performance test results saved (with error) id: #${runNumber}`, 'info');
                }
                catch (saveError) {
                    outputHandler.printLog(`Failed to save performance test results: ${saveError.message}`, 'error');
                    if (saveError.response) {
                        outputHandler.printLog(`API Response Status: ${saveError.response.status}`, 'error');
                        outputHandler.printLog(`API Response Body: ${JSON.stringify(saveError.response.data, null, 2)}`, 'error');
                    }
                    outputHandler.printLog(`Request Data: ${JSON.stringify(aggregatedMetrics, null, 2)}`, 'error');
                }
            }
            catch (aggError) {
                outputHandler.printLog(`Failed to aggregate metrics: ${aggError.message}`, 'error');
            }
            outputHandler.destroy();
            await cleanup();
            return callback(error);
        }
        outputHandler.printLog('Aggregating metrics from DuckDB...', 'info');
        let aggregatedMetrics = null;
        try {
            aggregatedMetrics = await aggregateMetricsFromDuckDB(duckDBService, optionsPerf.runId, optionsPerf.startedAt, finishedAt, null);
        }
        catch (aggError) {
            return callback(new Error(`Failed to aggregate metrics: ${aggError.message}`));
        }
        enrichAggregatedMetrics(aggregatedMetrics, optionsPerf.startedAt, finishedAt);
        outputHandler.printLog(`Performance test saving results id: #${runNumber}`, 'info');
        try {
            await savePerformanceTestResults({
                perfRunId: optionsPerf.runId,
                data: aggregatedMetrics,
                postmanApiKey: config.postmanApiKey,
                accessToken: config.accessToken
            });
            outputHandler.printLog(`Performance test results saved id: #${runNumber}`, 'info');
        }
        catch (saveError) {
            outputHandler.printLog(`Failed to save performance test results: ${saveError.message}`, 'error');
            if (saveError.response) {
                outputHandler.printLog(`API Response Status: ${saveError.response.status}`, 'error');
                outputHandler.printLog(`API Response Body: ${JSON.stringify(saveError.response.data, null, 2)}`, 'error');
            }
            outputHandler.printLog(`Request Data: ${JSON.stringify(aggregatedMetrics, null, 2)}`, 'error');
        }
        outputHandler.destroy();
        if (aggregatedMetrics?.summary?.aggregatedMetrics) {
            const durationSeconds = Math.floor((new Date(finishedAt).getTime() - new Date(optionsPerf.startedAt).getTime()) / 1000);
            const totalRequests = aggregatedMetrics.summary.aggregatedMetrics.requestThroughput?.sum || 0;
            displayCompletionSummary(runNumber, optionsPerf.runId, aggregatedMetrics.summary.aggregatedMetrics, durationSeconds, totalRequests, collectionUID);
            if (options.passIf) {
                let parsedCondition = null;
                try {
                    parsedCondition = parsePassIfCondition(options.passIf);
                }
                catch (parseError) {
                    await cleanup();
                    return callback(new Error(`Invalid pass-if condition: ${parseError.message}`), aggregatedMetrics);
                }
                const evaluation = evaluatePassIfCondition(parsedCondition, aggregatedMetrics.summary.aggregatedMetrics);
                if (evaluation.passed) {
                    console.log('');
                    console.log(colors.green('✓ Pass-if condition met: ') + colors.white(evaluation.message));
                    console.log('');
                }
                if (!evaluation.passed) {
                    await cleanup();
                    return callback(new Error(`Pass-if condition failed: ${evaluation.message}`), aggregatedMetrics);
                }
            }
        }
        await cleanup();
        return callback(null, aggregatedMetrics);
    });
    return test;
}
module.exports = function (configs, options, outputHandler, callback) {
    if (!_.isFunction(callback)) {
        callback = _.noop;
    }
    executeRun(configs, options, outputHandler, callback);
};
