// @postman/pm-bin-linux-x64/bin/lib/perftest/perftest-service.js
const request = require('postman-request'), util = require('../util');
function _buildAuthHeaders(postmanApiKey, accessToken) {
    const headers = {
        ...util.defaultHeaders,
        'x-pstmn-req-service': 'performance-test-api',
        'Content-Type': 'application/json'
    };
    if (postmanApiKey) {
        headers['x-api-key'] = postmanApiKey;
    }
    else if (accessToken) {
        headers['x-access-token'] = accessToken;
    }
    return headers;
}
function _makeRequest({ method, url, body, postmanApiKey, accessToken }) {
    const headers = _buildAuthHeaders(postmanApiKey, accessToken), requestConfig = { url, headers };
    if (body) {
        requestConfig.body = JSON.stringify(body);
    }
    return new Promise((resolve, reject) => {
        request[method](requestConfig, (err, response, responseBody) => {
            if (err) {
                return reject(new Error(`Network error: ${err.message}`));
            }
            let data;
            try {
                data = responseBody && JSON.parse(responseBody);
            }
            catch (parseError) {
                return reject(new Error(`Failed to parse response: ${parseError.message}`));
            }
            if (response.statusCode < 200 || response.statusCode >= 300) {
                const errorMessage = `HTTP ${response.statusCode}: ${(data && data.message) || 'Request failed'}`, error = new Error(errorMessage);
                error.response = {
                    status: response.statusCode,
                    statusText: response.statusMessage,
                    headers: response.headers,
                    data: data
                };
                error.request = {
                    method: method.toUpperCase(),
                    url: url,
                    headers: headers,
                    body: body,
                    bodySize: body ? JSON.stringify(body).length : 0
                };
                return reject(error);
            }
            resolve({ data: data, statusCode: response.statusCode });
        });
    });
}
async function createPerformanceTest({ data, postmanApiKey, accessToken }) {
    const url = `${util.POSTMAN_GATEWAY_BASE_URL()}/performance-runs`, response = await _makeRequest({
        method: 'post',
        url: url,
        body: data,
        postmanApiKey: postmanApiKey,
        accessToken: accessToken
    });
    return response.data?.data?.data || response.data?.data;
}
async function savePerformanceTestResults({ perfRunId, data, postmanApiKey, accessToken }) {
    const url = `${util.POSTMAN_GATEWAY_BASE_URL()}/performance-runs/${perfRunId}/results`;
    await _makeRequest({
        method: 'post',
        url: url,
        body: data,
        postmanApiKey: postmanApiKey,
        accessToken: accessToken
    });
    return true;
}
module.exports = {
    createPerformanceTest,
    savePerformanceTestResults
};
