// @postman/pm-bin-linux-x64/bin/lib/perftest/utils.js
const os = require('os'), yaml = require('js-yaml'), Ajv = require('ajv'), betterAjvErrors = require('better-ajv-errors').default || require('better-ajv-errors'), performanceRunParametersSchema = require('./schemas/performance-run-parameters');
function getMachineSpecs() {
    const cpus = os.cpus().map((cpu) => {
        return { model: cpu.model, speed: cpu.speed };
    }), totalMemory = os.totalmem(), freeMemory = os.freemem(), platform = os.platform(), arch = os.arch();
    return {
        cpu: {
            count: cpus.length,
            details: cpus
        },
        memory: {
            total: totalMemory,
            free: freeMemory,
            used: totalMemory - freeMemory
        },
        os: {
            platform,
            arch
        }
    };
}
function getFixedLoadProfile(vuCount = 20, duration = 1) {
    return {
        type: 'fixed',
        shorthand: 'Fixed',
        description: `${vuCount} virtual users run for ${duration} ${duration > 1 ? 'minutes' : 'minute'}, each executing all requests sequentially.`,
        options: {
            totalDuration: {
                millis: duration * 60000
            }
        },
        controlPoints: [
            {
                id: 'initial',
                load: {
                    count: vuCount
                },
                time: {
                    millis: 0
                }
            },
            {
                id: 'final',
                load: {
                    count: vuCount
                },
                time: {
                    millis: duration * 60000
                }
            }
        ]
    };
}
function getRampUpLoadProfile(vuCount = 20, duration = 1) {
    const durationMillis = duration * 60000;
    const startVUs = Math.max(1, Math.floor(vuCount * 0.25));
    const rampUpDuration = Math.floor(durationMillis * 0.25);
    return {
        type: 'ramp-up',
        shorthand: 'Ramp-Up',
        description: `${startVUs} virtual users ramp up to ${vuCount} over ${Math.floor(rampUpDuration / 1000)}s, then maintain for ${duration} ${duration > 1 ? 'minutes' : 'minute'}.`,
        options: {
            totalDuration: { millis: durationMillis },
            rampUpDuration: { millis: rampUpDuration }
        },
        controlPoints: [
            { id: 'initial', load: { count: startVUs }, time: { millis: 0 } },
            { id: 'rampLeadStart', load: { count: startVUs }, time: { millis: Math.floor(durationMillis * 0.25) } },
            { id: 'rampLeadEnd', load: { count: vuCount }, time: { millis: Math.floor(durationMillis * 0.5) } },
            { id: 'final', load: { count: vuCount }, time: { millis: durationMillis } }
        ]
    };
}
function getSpikeLoadProfile(vuCount = 20, duration = 1) {
    const durationMillis = duration * 60000;
    const baselineVUs = Math.max(1, Math.floor(vuCount * 0.1));
    return {
        type: 'spike',
        shorthand: 'Spike',
        description: `${baselineVUs} virtual users spike to ${vuCount}, then drop back over ${duration} ${duration > 1 ? 'minutes' : 'minute'}.`,
        options: {
            totalDuration: { millis: durationMillis }
        },
        controlPoints: [
            { id: 'initial', load: { count: baselineVUs }, time: { millis: 0 } },
            { id: 'spikeStart', load: { count: baselineVUs }, time: { millis: Math.floor(durationMillis * 0.4) } },
            { id: 'spikePeak', load: { count: vuCount }, time: { millis: Math.floor(durationMillis * 0.5) } },
            { id: 'spikeEnd', load: { count: baselineVUs }, time: { millis: Math.floor(durationMillis * 0.6) } },
            { id: 'final', load: { count: baselineVUs }, time: { millis: durationMillis } }
        ]
    };
}
function getPeakLoadProfile(vuCount = 20, duration = 1) {
    const durationMillis = duration * 60000;
    const baselineVUs = Math.max(1, Math.floor(vuCount * 0.2));
    return {
        type: 'peak',
        shorthand: 'Peak',
        description: `${baselineVUs} virtual users ramp up to ${vuCount}, maintain peak, then ramp down over ${duration} ${duration > 1 ? 'minutes' : 'minute'}.`,
        options: {
            totalDuration: { millis: durationMillis }
        },
        controlPoints: [
            { id: 'initial', load: { count: baselineVUs }, time: { millis: 0 } },
            { id: 'rampLeadStart', load: { count: baselineVUs }, time: { millis: Math.floor(durationMillis * 0.2) } },
            { id: 'rampLeadEnd', load: { count: vuCount }, time: { millis: Math.floor(durationMillis * 0.4) } },
            { id: 'rampDownStart', load: { count: vuCount }, time: { millis: Math.floor(durationMillis * 0.6) } },
            { id: 'rampDownEnd', load: { count: baselineVUs }, time: { millis: Math.floor(durationMillis * 0.8) } },
            { id: 'final', load: { count: baselineVUs }, time: { millis: durationMillis } }
        ]
    };
}
function getCollectionStructure(inputArray, createdBy) {
    return inputArray.map((obj) => {
        if (Array.isArray(obj.item)) {
            return {
                id: `${createdBy}-${obj.id}`,
                item: getCollectionStructure(obj.item, createdBy)
            };
        }
        return {
            id: `${createdBy}-${obj.id}`,
            name: obj.name,
            method: obj.request?.method || null
        };
    });
}
function flattenCollection(inputArray, createdBy) {
    const result = [];
    function recurse(items) {
        items.forEach((item) => {
            if (item.request) {
                result.push({
                    id: `${createdBy}-${item.id}`,
                    type: 'request'
                });
            }
            if (Array.isArray(item.item)) {
                recurse(item.item);
            }
        });
    }
    recurse(inputArray);
    return result;
}
function validateLoadProfile(loadProfile, context) {
    const errors = [], { duration, peakVUs, load } = loadProfile;
    if (!loadProfile) {
        return errors;
    }
    if (load && load.ramp) {
        const { up, down } = load.ramp;
        const totalRampDuration = up + down;
        if (totalRampDuration > duration) {
            errors.push({
                message: `Ramp up (${up}s) + ramp down (${down}s) = ${totalRampDuration}s exceeds total duration (${duration}s)`,
                instancePath: `/${context}/loadProfile/load/ramp`,
                type: 'ramp-duration-exceeded'
            });
        }
    }
    if (load && load.stages) {
        const totalStageDuration = load.stages.reduce((sum, stage) => {
            return sum + stage.duration;
        }, 0);
        if (totalStageDuration !== duration) {
            errors.push({
                message: `Sum of stage durations (${totalStageDuration}s) does not equal total duration (${duration}s)`,
                instancePath: `/${context}/loadProfile/load/stages`,
                type: 'stages-duration-mismatch'
            });
        }
        load.stages.forEach((stage, index) => {
            if (stage.vus > peakVUs) {
                errors.push({
                    message: `Stage ${index + 1}${stage.name ? ` "${stage.name}"` : ''} VUs (${stage.vus}) exceeds peakVUs (${peakVUs})`,
                    instancePath: `/${context}/loadProfile/load/stages/${index}/vus`,
                    type: 'stage-vus-exceeded'
                });
            }
        });
    }
    return errors;
}
function validatePerformanceRunParameters(yamlString) {
    try {
        const parsedData = yaml.load(yamlString);
        const ajv = new Ajv({ allErrors: true, verbose: true });
        const validate = ajv.compile(performanceRunParametersSchema);
        const hasDefaultLoadProfile = parsedData.defaults && parsedData.defaults.loadProfile;
        const schemaValid = validate(parsedData);
        const customErrors = [];
        if (!schemaValid) {
            const betterErrors = betterAjvErrors(performanceRunParametersSchema, parsedData, validate.errors, { format: 'cli', indent: 2 });
            return {
                valid: false,
                data: null,
                errors: validate.errors,
                betterErrors: betterErrors
            };
        }
        if (parsedData.runs) {
            parsedData.runs.forEach((run, index) => {
                if (!hasDefaultLoadProfile && !run.loadProfile) {
                    customErrors.push({
                        message: 'Load profile is required either in defaults or in each run',
                        instancePath: `/runs[${index}]`,
                        type: 'missing-load-profile'
                    });
                }
            });
        }
        if (parsedData.defaults && parsedData.defaults.loadProfile) {
            const errors = validateLoadProfile(parsedData.defaults.loadProfile, 'defaults');
            customErrors.push(...errors);
        }
        if (parsedData.runs) {
            parsedData.runs.forEach((run, index) => {
                if (run.loadProfile) {
                    const errors = validateLoadProfile(run.loadProfile, `runs[${index}]`);
                    customErrors.push(...errors);
                }
            });
        }
        if (customErrors.length > 0) {
            return {
                valid: false,
                data: null,
                errors: customErrors
            };
        }
        return {
            valid: true,
            data: parsedData,
            errors: null
        };
    }
    catch (error) {
        return {
            valid: false,
            data: null,
            errors: [{
                    message: error.message,
                    type: 'yaml-parse-error'
                }]
        };
    }
}
function translateLoadProfile(loadProfile) {
    const { duration, peakVUs, load } = loadProfile, durationMillis = duration * 1000;
    let controlPoints = [];
    if (load.ramp) {
        const { up } = load.ramp, upMillis = up * 1000;
        controlPoints = [
            {
                id: 'initial',
                load: { count: 0 },
                time: { millis: 0 }
            },
            {
                id: 'rampLeadStart',
                load: { count: 0 },
                time: { millis: 1 }
            },
            {
                id: 'rampLeadEnd',
                load: { count: peakVUs },
                time: { millis: upMillis }
            },
            {
                id: 'final',
                load: { count: peakVUs },
                time: { millis: durationMillis }
            }
        ];
    }
    else if (load.stages) {
        let cumulativeTime = 0;
        controlPoints = load.stages.map((stage, index) => {
            const point = {
                id: stage.name || `stage-${index}`,
                load: { count: stage.vus },
                time: { millis: cumulativeTime + (stage.duration * 1000) }
            };
            cumulativeTime += stage.duration * 1000;
            return point;
        });
        if (controlPoints[0].time.millis > 0) {
            controlPoints.unshift({
                id: 'initial',
                load: { count: 0 },
                time: { millis: 0 }
            });
        }
    }
    return {
        type: load.ramp ? 'ramp-up' : 'stages',
        shorthand: load.ramp ? 'Ramp-Up' : 'Custom Stages',
        description: `Performance test with ${peakVUs} peak virtual users over ${duration} seconds`,
        options: {
            totalDuration: { millis: durationMillis },
            rampUpDuration: { millis: load.ramp.up * 1000 }
        },
        controlPoints: controlPoints
    };
}
function mergeRunConfig(defaults, run) {
    const merged = {
        ...run,
        loadProfile: run.loadProfile || defaults?.loadProfile,
        assertions: {
            ...(defaults?.assertions || {}),
            ...(run.assertions || {})
        }
    };
    if (defaults?.assertions?.responseTime && run.assertions?.responseTime) {
        merged.assertions.responseTime = {
            ...defaults.assertions.responseTime,
            ...run.assertions.responseTime
        };
    }
    return merged;
}
function csvToDataFileContent(csvContent) {
    const lines = csvContent.split('\n').filter(line => line.trim() !== '');
    const headers = lines[0].split(',').map(h => h.trim());
    const contents = [];
    for (let i = 1; i < lines.length; i++) {
        const values = lines[i].split(',');
        const row = {};
        headers.forEach((header, index) => {
            row[header] = values[index] || '';
        });
        contents.push(row);
    }
    return contents;
}
module.exports = {
    getMachineSpecs,
    getFixedLoadProfile,
    getRampUpLoadProfile,
    getSpikeLoadProfile,
    getPeakLoadProfile,
    getCollectionStructure,
    flattenCollection,
    validatePerformanceRunParameters,
    translateLoadProfile,
    mergeRunConfig,
    csvToDataFileContent
};
