// @postman/pm-bin-linux-x64/bin/lib/logout/index.js
const _ = require('lodash'), waterfall = require('async/waterfall'), prompts = require('prompts'), rcfile = require('../config/rc-file'), env = require('../config/process-env'), util = require('../util'), request = require('postman-request'), colors = require('colors/safe'), console = require('../framework/console'), ALIAS = 'alias', DEFAULT = 'default', USER_PERMISSION = 'userPermission', LOGOUT_CONFIRMATION = (alias) => {
    return alias === DEFAULT ? 'Are you sure you want to log out?' :
        `Are you sure you want to log out of the \`${alias}\` alias?`;
}, ALIAS_INPUT_PROMPT = 'Select the alias', ABORT_MESSAGE = 'Log out aborted.', NO_CREDENTIALS_AVAILABLE = 'No credentials available.' +
    '\n   Use the login command to store credentials.', ALIAS_NOT_PRESENT = 'Alias not present.';
module.exports = (options, callback) => {
    if (options.alias) {
        console.warn(colors.yellow('[warning] --alias is discontinued and will be removed in a future release.\n'));
    }
    let fileData, alias = options.alias || DEFAULT;
    waterfall([
        rcfile.load,
        (data, next) => {
            fileData = data;
            env.load((err, envOptions) => {
                if (err) {
                    return next(err);
                }
                return next(null, _.get(envOptions, 'logout.alias'));
            });
        },
        (sessionAlias, next) => {
            let profiles, initialProfile, promptOptions;
            fileData.login && (profiles = fileData.login._profiles);
            if (_.isEmpty(profiles)) {
                return next(new Error(NO_CREDENTIALS_AVAILABLE));
            }
            if (profiles.length === 1) {
                alias = _.head(profiles).alias;
                return next(null);
            }
            if (options.alias) {
                return next(null);
            }
            promptOptions = {
                type: 'autocomplete',
                name: ALIAS,
                message: ALIAS_INPUT_PROMPT,
                choices: _.map(profiles, (profile) => { return { title: profile.alias }; }),
                initial: sessionAlias
            };
            [initialProfile] = _.filter(profiles, [ALIAS, sessionAlias]);
            initialProfile && (promptOptions.initial = initialProfile.alias);
            util.signal(ALIAS);
            prompts.prompt(promptOptions)
                .then((response) => {
                if (!_.has(response, ALIAS)) {
                    process.exit(1);
                }
                alias = response.alias;
                if (!alias) {
                    return next(ALIAS_NOT_PRESENT);
                }
                return next(null);
            })
                .catch(next);
        },
        (next) => {
            let promptOptions = {
                type: 'toggle',
                name: USER_PERMISSION,
                message: LOGOUT_CONFIRMATION(alias),
                active: 'Yes',
                inactive: 'No',
                initial: true
            };
            util.signal(USER_PERMISSION);
            prompts.prompt(promptOptions)
                .then((response) => {
                if (!_.has(response, USER_PERMISSION)) {
                    process.exit(1);
                }
                if (!response.userPermission) {
                    return next(ABORT_MESSAGE);
                }
                return next(null);
            })
                .catch(next);
        },
        (next) => {
            const profile = _.find(_.get(fileData, 'login._profiles', []), ['alias', alias]), accessToken = profile ? profile.accessToken : null, gatewayBaseUrl = util.POSTMAN_GATEWAY_BASE_URL(), url = `${gatewayBaseUrl}/api/sessions/current`, options = {
                method: 'DELETE',
                url: url,
                headers: {
                    'x-access-token': accessToken,
                    'x-pstmn-req-service': 'identity'
                },
                timeout: 10000,
                json: true
            };
            if (!accessToken) {
                return next(null);
            }
            request(options, (error, response) => {
                if (error) {
                    console.warn('Warning: Failed to invalidate session on server:', error.message);
                }
                else if (response && (response.statusCode < 200 || response.statusCode >= 300)) {
                    console.warn('Warning: Session invalidation returned status:', response.statusCode);
                }
                return next(null);
            });
        },
        (next) => {
            fileData.login._profiles = _.reject(fileData.login._profiles, [ALIAS, alias]);
            return next(null, fileData);
        },
        rcfile.store
    ], callback);
};
