// @postman/pm-bin-linux-x64/bin/lib/commands/runner/logging/index.js
const winston = require('winston'), os = require('os'), { version } = require('../../../../package.json');
class RunnerLogger {
    constructor(config = {}) {
        this.config = {
            level: config.level || 'info',
            enabled: config.enabled !== false,
            service: config.service || 'postman-runner',
            version: version,
            prettyFormat: config.prettyFormat || false
        };
        this.instanceInfo = {
            hostname: os.hostname(),
            pid: process.pid,
            arch: os.arch(),
            platform: os.platform()
        };
        this.logger = winston.createLogger({
            level: this.config.level,
            silent: !this.config.enabled,
            format: winston.format.combine(winston.format.timestamp(), winston.format.printf(this._formatLogEntry.bind(this))),
            transports: [new winston.transports.Console()]
        });
    }
    info(message, meta = {}) { this.logger.info(message, { meta }); }
    debug(message, meta = {}) { this.logger.debug(message, { meta }); }
    error(message, meta = {}) { this.logger.error(message, { meta }); }
    warn(message, meta = {}) { this.logger.warn(message, { meta }); }
    _formatLogEntry(info) {
        const logEntry = {
            timestamp: info.timestamp,
            level: info.level,
            message: RunnerLogger._redactSensitiveData(info.message),
            'service.name': this.config.service,
            'service.version': this.config.version,
            'process.pid': process.pid,
            'host.name': this.instanceInfo.hostname,
            'host.arch': this.instanceInfo.arch,
            'host.platform': this.instanceInfo.platform,
            'host.type': this.instanceInfo.type,
            ...RunnerLogger._formatMeta(info.meta || {})
        };
        return this.config.prettyFormat ?
            JSON.stringify(logEntry, null, 2) :
            JSON.stringify(logEntry);
    }
    static _redactSensitiveData(value) {
        if (typeof value === 'string') {
            let redacted = value;
            redacted = redacted.replace(/PMAK-[a-f0-9]{24}-[a-f0-9]{34}/g, '[[REDACTED_API_KEY]]');
            redacted = redacted.replace(/\b[a-f0-9]{32}\b/g, '[[REDACTED_API_KEY]]');
            redacted = redacted.replace(/(Bearer\s+)[^\s"',}]+/gi, '$1[[REDACTED_TOKEN]]');
            redacted = redacted.replace(/(Authorization:\s*)[^\s"',}]+/gi, '$1[[REDACTED_AUTH]]');
            return redacted;
        }
        if (Array.isArray(value)) {
            return value.map((item) => { return RunnerLogger._redactSensitiveData(item); });
        }
        if (value && typeof value === 'object') {
            const redacted = {};
            Object.keys(value).forEach((key) => {
                const lowerKey = key.toLowerCase();
                if (lowerKey.includes('key') || lowerKey.includes('token') ||
                    lowerKey.includes('password') || lowerKey.includes('secret') ||
                    lowerKey.includes('auth')) {
                    redacted[key] = '[[REDACTED]]';
                }
                else {
                    redacted[key] = RunnerLogger._redactSensitiveData(value[key]);
                }
            });
            return redacted;
        }
        return value;
    }
    static _formatMeta(meta) {
        return RunnerLogger._redactSensitiveData(meta);
    }
}
module.exports = RunnerLogger;
