// @postman/pm-bin-linux-x64/bin/lib/commands/runner/logging/runtime-event-logger.js
const { getLogger } = require('./logger-factory');
function getLogLevelForEvent(eventName) {
    const errorEvents = ['error', 'exception'], warnEvents = ['pause'];
    if (errorEvents.includes(eventName)) {
        return 'error';
    }
    if (warnEvents.includes(eventName)) {
        return 'warn';
    }
    return 'info';
}
function extractEventMeta(eventData) {
    const meta = {};
    if (!eventData) {
        return meta;
    }
    if (eventData.error) {
        if (typeof eventData.error === 'object') {
            meta.errorMessage = eventData.error.message || eventData.error.toString();
            meta.errorType = eventData.error.constructor?.name || 'Unknown';
            meta.errorCode = eventData.error.code;
            meta.errorStack = eventData.error.stack;
        }
        else {
            meta.errorMessage = String(eventData.error);
        }
    }
    if (eventData.cursor) {
        meta.iteration = eventData.cursor.iteration;
        meta.position = eventData.cursor.position;
        meta.cursorLength = eventData.cursor.length;
        meta.cursorCycles = eventData.cursor.cycles;
        meta.httpRequestId = eventData.cursor.httpRequestId;
        meta.scriptId = eventData.cursor.scriptId;
        meta.executionId = eventData.cursor.execution;
    }
    if (eventData.item) {
        meta.itemName = eventData.item.name;
        meta.itemId = eventData.item.id;
    }
    if (eventData.request) {
        meta.method = eventData.request.method;
        meta.requestContentType = eventData.request.headers?.get('content-type');
    }
    if (eventData.response) {
        meta.statusCode = eventData.response.code;
        meta.statusText = eventData.response.status;
        meta.duration = eventData.response.responseTime;
        meta.responseSize = eventData.response.responseSize;
        meta.responseContentType = eventData.response.headers?.get('content-type');
    }
    if (eventData.assertion) {
        const assertions = Array.isArray(eventData.assertion) ? eventData.assertion : [eventData.assertion];
        meta.assertionCount = assertions.length;
        meta.assertionsPassed = assertions.filter((a) => {
            return a.passed;
        }).length;
        meta.assertionsFailed = assertions.filter((a) => {
            return !a.passed;
        }).length;
        meta.assertionsSkipped = assertions.filter((a) => {
            return a.skipped;
        }).length;
        const failedAssertions = assertions.filter((a) => {
            return !a.passed;
        }).map((a) => {
            return a.name;
        });
        if (failedAssertions.length > 0) {
            meta.failedAssertions = failedAssertions;
        }
    }
    if (eventData.scriptResults) {
        meta.scriptResultsCount = eventData.scriptResults.length;
    }
    if (eventData.trace) {
        meta.traceType = eventData.trace.type;
        meta.traceSource = eventData.trace.source;
    }
    return meta;
}
function logRuntimeEvent(eventName, eventData, context = {}) {
    try {
        const logger = getLogger();
        const safeEventData = eventData || {}, logLevel = getLogLevelForEvent(eventName), meta = {
            event: eventName,
            runnerId: context.runnerId,
            runId: context.runId,
            jobId: context.jobId,
            ...extractEventMeta(safeEventData)
        };
        logger[logLevel](`[Runtime Event: ${eventName}]`, meta);
    }
    catch (error) {
        const logger = getLogger();
        logger.error('Failed to log runtime event', {
            operation: 'logRuntimeEvent',
            errorMessage: error.message,
            errorStack: error.stack,
            eventName: eventName
        });
    }
}
module.exports = logRuntimeEvent;
