// @postman/pm-bin-linux-x64/bin/lib/commands/mock/runtime/console-capture.js
function initConsoleCapture(worker, options) {
    const { onLog, maxLogs = 100, logsBuffer } = options || {};
    if (!worker || !onLog) {
        return () => { };
    }
    function handleOutput(data, level) {
        const msg = data.toString().trim();
        if (!msg) {
            return;
        }
        const logEntry = {
            id: `console-${Date.now()}-${Math.random().toString(36).substring(2, 11)}`,
            type: 'console',
            level,
            args: [msg],
            timestamp: new Date().toISOString(),
            workerId: worker.pid
        };
        if (logsBuffer) {
            logsBuffer.push(logEntry);
            if (logsBuffer.length > maxLogs) {
                logsBuffer.shift();
            }
        }
        onLog(logEntry);
    }
    const stdoutHandler = (data) => handleOutput(data, 'log');
    const stderrHandler = (data) => handleOutput(data, data.toString().toLowerCase().includes('warn') ? 'warn' : 'error');
    if (worker.stdout) {
        worker.stdout.on('data', stdoutHandler);
    }
    if (worker.stderr) {
        worker.stderr.on('data', stderrHandler);
    }
    return () => {
        worker.stdout?.off('data', stdoutHandler);
        worker.stderr?.off('data', stderrHandler);
    };
}
module.exports = { initConsoleCapture };
