// @postman/pm-bin-linux-x64/bin/lib/commands/mock/runtime/http-server-worker.js
if (typeof process.send === 'undefined') {
    console.warn('[HTTPServerWorker] ⚠️  Not running as forked process - worker code will not execute');
    console.warn('[HTTPServerWorker] This is normal if this file is being scanned during build/import resolution');
    if (typeof module !== 'undefined') {
        module.exports = {};
    }
}
else {
    (function () {
        const { existsSync, readFileSync } = require('fs');
        const { resolve } = require('path');
        const net = require('net');
        const { createRequire } = require('module');
        const { installHttpLogging } = require('./http-logger');
        const { installPmGlobal } = require('./pm');
        const installScenarios = function () { };
        const updateScenariosConfig = function () { };
        const nativeRequire = createRequire(__filename);
        let examplesData = [];
        let requestsData = [];
        let environmentData = [];
        let globalsData = [];
        let scenarios = null;
        let logs = [];
        const maxLogs = 100;
        function sendLog(log) {
            const logWithId = {
                ...log,
                id: Date.now() + Math.random(),
                workerId: process.pid
            };
            logs.push(logWithId);
            if (logs.length > maxLogs) {
                logs.shift();
            }
            try {
                if (process.send) {
                    process.send({
                        type: 'log',
                        payload: logWithId
                    });
                }
            }
            catch (err) {
                console.error('[HTTPServerWorker] Error sending log:', err.message);
            }
        }
        function initializeHttpLogging() {
            installHttpLogging({
                onLog: sendLog,
                maxBodySize: 10000,
                captureRequestBody: true,
                captureResponseBody: true
            });
        }
        function validateMockFile(filePath) {
            try {
                if (!existsSync(filePath)) {
                    throw new Error(`Mock file not found: ${filePath}`);
                }
                const content = readFileSync(filePath, 'utf8');
                try {
                    new Function(content);
                }
                catch (syntaxError) {
                    throw new Error(`Syntax error in mock file: ${syntaxError.message}`);
                }
                return {
                    success: true,
                    message: 'Mock file validated successfully'
                };
            }
            catch (error) {
                console.error('[HTTPServerWorker] Validation error:', error.message);
                return {
                    success: false,
                    error: error.message,
                    stack: error.stack
                };
            }
        }
        function isPortAvailable(port) {
            return new Promise((resolveCheck) => {
                const socket = new net.Socket();
                socket.setTimeout(1000);
                socket.on('connect', () => {
                    socket.destroy();
                    resolveCheck(false);
                });
                socket.on('error', (err) => {
                    socket.destroy();
                    if (err && err.code === 'ECONNREFUSED') {
                        resolveCheck(true);
                    }
                    else {
                        resolveCheck(false);
                    }
                });
                socket.on('timeout', () => {
                    socket.destroy();
                    resolveCheck(true);
                });
                socket.connect(port, '127.0.0.1');
            });
        }
        function waitForPort(port, timeout = 10000) {
            return new Promise((resolvePort) => {
                const startTime = Date.now();
                const check = () => {
                    const socket = new net.Socket();
                    socket.setTimeout(200);
                    socket.on('connect', () => {
                        socket.destroy();
                        resolvePort(true);
                    });
                    socket.on('error', () => {
                        socket.destroy();
                        if (Date.now() - startTime < timeout) {
                            setTimeout(check, 100);
                        }
                        else {
                            resolvePort(false);
                        }
                    });
                    socket.on('timeout', () => {
                        socket.destroy();
                        if (Date.now() - startTime < timeout) {
                            setTimeout(check, 100);
                        }
                        else {
                            resolvePort(false);
                        }
                    });
                    socket.connect(port, '127.0.0.1');
                };
                check();
            });
        }
        async function executeMockFile(filePath, port) {
            try {
                if (!existsSync(filePath)) {
                    throw new Error(`Mock file not found: ${filePath}`);
                }
                const portAvailable = await isPortAvailable(port);
                if (!portAvailable) {
                    throw new Error(`Port ${port} is already in use. Free it or use another.`);
                }
                initializeHttpLogging();
                installScenarios(scenarios || {});
                installPmGlobal({
                    examples: examplesData,
                    requests: requestsData,
                    environment: environmentData,
                    globals: globalsData
                });
                process.env.PORT = String(port);
                const resolvedPath = resolve(filePath);
                delete nativeRequire.cache[nativeRequire.resolve(resolvedPath)];
                nativeRequire(resolvedPath);
                const serverReady = await waitForPort(port, 10000);
                if (!serverReady) {
                    throw new Error(`Mock server did not start on port ${port} within timeout. ` +
                        'Make sure your mock file calls server.listen() with process.env.PORT');
                }
                return {
                    success: true,
                    port,
                    url: `http://localhost:${port}`
                };
            }
            catch (error) {
                console.error('Error executing mock file:', error.message);
                return {
                    success: false,
                    error: error.message,
                    stack: error.stack
                };
            }
        }
        process.on('message', async (message) => {
            const { id, type, payload } = message;
            try {
                switch (type) {
                    case 'validate': {
                        const { handlerPath } = payload;
                        const validationResult = validateMockFile(handlerPath);
                        if (process.send) {
                            process.send({
                                id,
                                type: validationResult.success ? 'success' : 'error',
                                payload: validationResult
                            });
                        }
                        break;
                    }
                    case 'start': {
                        const { handlerPath, port, examples, requests, scenarios: scenariosPayload, environment, globals } = payload;
                        if (examples && Array.isArray(examples)) {
                            examplesData = examples;
                        }
                        if (requests && Array.isArray(requests)) {
                            requestsData = requests;
                        }
                        if (environment && Array.isArray(environment)) {
                            environmentData = environment;
                        }
                        if (globals && Array.isArray(globals)) {
                            globalsData = globals;
                        }
                        if (scenariosPayload) {
                            scenarios = scenariosPayload;
                        }
                        const validationResult = validateMockFile(handlerPath);
                        if (!validationResult.success) {
                            if (process.send) {
                                process.send({
                                    id,
                                    type: 'error',
                                    payload: {
                                        error: validationResult.error,
                                        stack: validationResult.stack
                                    }
                                });
                            }
                            return;
                        }
                        const result = await executeMockFile(handlerPath, port || 3000);
                        if (process.send) {
                            process.send({
                                id,
                                type: result.success ? 'success' : 'error',
                                payload: result
                            });
                        }
                        break;
                    }
                    case 'update-scenarios': {
                        const { scenarios: newScenarios } = payload;
                        scenarios = newScenarios;
                        updateScenariosConfig(newScenarios);
                        if (process.send) {
                            process.send({
                                id,
                                type: 'success',
                                payload: { updated: true }
                            });
                        }
                        break;
                    }
                    case 'get-logs': {
                        if (process.send) {
                            process.send({
                                id,
                                type: 'success',
                                payload: { logs }
                            });
                        }
                        break;
                    }
                    case 'ping': {
                        if (process.send) {
                            process.send({
                                id,
                                type: 'pong',
                                payload: { status: 'alive', pid: process.pid }
                            });
                        }
                        break;
                    }
                    case 'shutdown': {
                        if (process.send) {
                            process.send({
                                id,
                                type: 'success',
                                payload: { shutdown: true }
                            });
                        }
                        setTimeout(() => {
                            process.exit(0);
                        }, 100);
                        break;
                    }
                    default:
                        if (process.send) {
                            process.send({
                                id,
                                type: 'error',
                                payload: {
                                    error: `Unknown message type: ${type}`
                                }
                            });
                        }
                }
            }
            catch (error) {
                console.error('[HTTPServerWorker] Error handling message:', error);
                if (process.send) {
                    process.send({
                        id,
                        type: 'error',
                        payload: {
                            error: error.message,
                            stack: error.stack
                        }
                    });
                }
            }
        });
        process.on('uncaughtException', (error) => {
            console.error('[HTTPServerWorker] Uncaught exception:', error);
            try {
                if (process.send) {
                    process.send({
                        type: 'worker-error',
                        payload: {
                            error: error.message,
                            stack: error.stack,
                            type: 'uncaughtException'
                        }
                    });
                }
            }
            catch (e) {
            }
            setTimeout(() => {
                process.exit(1);
            }, 100);
        });
        process.on('unhandledRejection', (reason) => {
            console.error('[HTTPServerWorker] Unhandled rejection:', reason);
            try {
                if (process.send) {
                    process.send({
                        type: 'worker-error',
                        payload: {
                            error: String(reason),
                            type: 'unhandledRejection'
                        }
                    });
                }
            }
            catch (e) {
            }
            setTimeout(() => {
                process.exit(1);
            }, 100);
        });
        process.on('disconnect', () => {
            process.exit(0);
        });
        try {
            if (process.send) {
                process.send({ type: 'ready', payload: { pid: process.pid } });
            }
            else {
                console.error('[HTTPServerWorker] process.send is not available');
                process.exit(1);
            }
        }
        catch (error) {
            console.error('[HTTPServerWorker] Failed to send ready signal:', error.message);
            process.exit(1);
        }
    }());
}
