// @postman/pm-bin-linux-x64/bin/lib/commands/mock/runtime/http-logger.js
const http = require('http');
const { parse: parseUrl } = require('url');
let logCallback = null;
let config = {
    maxBodySize: 10000,
    captureRequestBody: true,
    captureResponseBody: true
};
function installHttpLogging(options) {
    config = { ...config, ...options };
    logCallback = options.onLog || logCallback;
    if (!logCallback) {
        throw new Error('HttpLogger: onLog callback is required');
    }
    const originalCreateServer = http.createServer;
    http.createServer = function (requestListener) {
        const wrappedListener = function (req, res) {
            const startTime = process.hrtime.bigint();
            const parsedUrl = parseUrl(req.url, true);
            const log = {
                timestamp: new Date().toISOString(),
                method: req.method,
                path: parsedUrl.pathname,
                query: parsedUrl.query,
                headers: { ...req.headers },
                requestSize: 0,
                responseSize: 0
            };
            if (config.captureRequestBody) {
                const requestBodyChunks = [];
                req.on('data', (chunk) => {
                    requestBodyChunks.push(chunk);
                    log.requestSize += chunk.length;
                });
                req.on('end', () => {
                    if (requestBodyChunks.length > 0) {
                        const bodyBuffer = Buffer.concat(requestBodyChunks);
                        try {
                            const bodyString = bodyBuffer.toString('utf8');
                            const contentType = req.headers['content-type'] || '';
                            if (contentType.includes('application/json') && bodyString) {
                                log.body = JSON.parse(bodyString);
                            }
                            else if (bodyBuffer.length < config.maxBodySize) {
                                log.body = bodyString;
                            }
                            else {
                                log.body = `<truncated: ${bodyBuffer.length} bytes, type: ${contentType}>`;
                            }
                        }
                        catch (e) {
                            log.body = `<parse error: ${e.message}>`;
                        }
                    }
                });
            }
            const originalWrite = res.write.bind(res);
            const originalEnd = res.end.bind(res);
            const originalWriteHead = res.writeHead.bind(res);
            const responseChunks = [];
            let responseHeaders = {};
            let headersWritten = false;
            res.writeHead = function (statusCode, statusMessage, headers) {
                headersWritten = true;
                log.statusCode = statusCode;
                const normalizeHeaders = (hdrs) => {
                    const normalized = {};
                    for (const [key, value] of Object.entries(hdrs)) {
                        normalized[key.toLowerCase()] = value;
                    }
                    return normalized;
                };
                if (typeof statusMessage === 'object') {
                    responseHeaders = { ...normalizeHeaders(this.getHeaders()), ...normalizeHeaders(statusMessage) };
                }
                else if (headers) {
                    responseHeaders = { ...normalizeHeaders(this.getHeaders()), ...normalizeHeaders(headers) };
                }
                else {
                    responseHeaders = { ...normalizeHeaders(this.getHeaders()) };
                }
                return originalWriteHead.apply(this, arguments);
            };
            res.write = function (chunk, encoding, callback) {
                if (config.captureResponseBody && chunk) {
                    const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);
                    responseChunks.push(buffer);
                    log.responseSize += buffer.length;
                }
                return originalWrite.call(this, chunk, encoding, callback);
            };
            res.end = function (chunk, encoding, callback) {
                if (config.captureResponseBody && chunk) {
                    const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);
                    responseChunks.push(buffer);
                    log.responseSize += buffer.length;
                }
                const endTime = process.hrtime.bigint();
                log.duration = Number(endTime - startTime) / 1_000_000;
                log.statusCode = log.statusCode || res.statusCode || 200;
                if (!headersWritten) {
                    const hdrs = res.getHeaders();
                    responseHeaders = {};
                    for (const [key, value] of Object.entries(hdrs)) {
                        responseHeaders[key.toLowerCase()] = value;
                    }
                }
                log.responseHeaders = responseHeaders;
                if (config.captureResponseBody && responseChunks.length > 0) {
                    const responseBuffer = Buffer.concat(responseChunks);
                    try {
                        const contentType = responseHeaders['content-type'] || '';
                        if (responseBuffer.length < config.maxBodySize) {
                            const responseString = responseBuffer.toString('utf8');
                            if (contentType.includes('application/json') && responseString) {
                                log.response = JSON.parse(responseString);
                            }
                            else if (contentType.includes('text') || contentType.includes('application/json')) {
                                log.response = responseString;
                            }
                            else {
                                log.response = `<binary: ${responseBuffer.length} bytes, type: ${contentType}>`;
                            }
                        }
                        else {
                            log.response = `<large payload: ${responseBuffer.length} bytes, type: ${contentType}>`;
                        }
                    }
                    catch (e) {
                        log.response = `<parse error: ${e.message}>`;
                    }
                }
                try {
                    logCallback(log);
                }
                catch (callbackError) {
                    console.error('[HttpLogger] Error in log callback:', callbackError);
                }
                return originalEnd.call(this, chunk, encoding, callback);
            };
            try {
                if (requestListener) {
                    requestListener(req, res);
                }
            }
            catch (error) {
                console.error('[HttpLogger] Request handler error:', error);
                const endTime = process.hrtime.bigint();
                log.statusCode = 500;
                log.duration = Number(endTime - startTime) / 1_000_000;
                log.error = error.message;
                log.errorStack = error.stack;
                try {
                    logCallback(log);
                }
                catch (callbackError) {
                    console.error('[HttpLogger] Error in log callback:', callbackError);
                }
                if (!res.headersSent) {
                    res.writeHead(500, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({
                        error: 'Internal server error',
                        message: error.message
                    }));
                }
            }
        };
        return originalCreateServer.call(this, wrappedListener);
    };
}
function uninstallHttpLogging() {
    logCallback = null;
}
function setLogCallback(callback) {
    logCallback = callback;
}
function updateConfig(newConfig) {
    config = { ...config, ...newConfig };
}
module.exports = {
    installHttpLogging,
    uninstallHttpLogging,
    setLogCallback,
    updateConfig
};
