// @postman/pm-bin-linux-x64/bin/lib/commands/mock/runtime/sandbox.js
const { fork } = require('child_process');
const { EventEmitter } = require('events');
const { join } = require('path');
const { initConsoleCapture } = require('./console-capture');
class SandboxedMockServer extends EventEmitter {
    constructor(config = {}) {
        super();
        this.config = {
            port: 3000,
            logger: console.log,
            timeout: 5000,
            maxLogs: 100,
            autoRestart: true,
            maxRestarts: 3,
            examples: [],
            requests: [],
            environment: [],
            globals: [],
            scenarios: null,
            workerPath: null,
            ...config
        };
        this.worker = null;
        this.handlerPath = null;
        this.serverInfo = null;
        this.logs = [];
        this.messageIdCounter = 0;
        this.pendingRequests = new Map();
        this.restartCount = 0;
        this.isShuttingDown = false;
        this.examples = this.config.examples || [];
        this.requests = this.config.requests || [];
        this.environment = this.config.environment || [];
        this.globals = this.config.globals || [];
        this.scenarios = this.config.scenarios || null;
    }
    async _createWorker() {
        return new Promise((resolve, reject) => {
            const fs = require('fs');
            try {
                const workerPath = this.config.workerPath || join(__dirname, 'http-server-worker.js');
                if (!fs.existsSync(workerPath)) {
                    const error = `Worker file not found at: ${workerPath}`;
                    this.config.logger(error);
                    reject(new Error(error));
                    return;
                }
                let worker;
                try {
                    worker = fork(workerPath, [], {
                        stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
                        execArgv: []
                    });
                }
                catch (forkError) {
                    this.config.logger('Failed to fork worker:', forkError.message);
                    throw new Error(`Failed to fork worker process: ${forkError.message}`);
                }
                if (!worker) {
                    throw new Error('Worker process failed to start (fork returned null)');
                }
                if (!worker.send || typeof worker.send !== 'function') {
                    this.config.logger('Worker IPC channel not available');
                    try {
                        worker.kill();
                    }
                    catch (forkError) {
                        this.config.logger('Failed to kill worker:', forkError.message);
                    }
                    throw new Error('Worker IPC channel not available - fork may have failed');
                }
                let workerReady = false;
                let creationTimeout = null;
                let readyHandler, errorHandler, exitHandler;
                readyHandler = (message) => {
                    if (message.type === 'ready') {
                        workerReady = true;
                        clearTimeout(creationTimeout);
                        worker.off('message', readyHandler);
                        worker.off('error', errorHandler);
                        worker.off('exit', exitHandler);
                        resolve(worker);
                    }
                };
                errorHandler = (error) => {
                    if (!workerReady) {
                        clearTimeout(creationTimeout);
                        this.config.logger('Worker creation error:', error);
                        worker.off('message', readyHandler);
                        worker.off('exit', exitHandler);
                        try {
                            worker.kill();
                        }
                        catch (e) {
                        }
                        reject(new Error(`Worker creation failed: ${error.message}`));
                    }
                };
                exitHandler = (code, signal) => {
                    if (!workerReady) {
                        clearTimeout(creationTimeout);
                        const msg = `Worker exited during creation (code: ${code}, signal: ${signal})`;
                        this.config.logger(`${msg}`);
                        worker.off('message', readyHandler);
                        worker.off('error', errorHandler);
                        reject(new Error(msg));
                    }
                    else {
                        this._handleWorkerExit(code, signal);
                    }
                };
                worker.on('message', readyHandler);
                worker.on('error', errorHandler);
                worker.on('exit', exitHandler);
                initConsoleCapture(worker, {
                    onLog: (logEntry) => this.emit('log', logEntry),
                    maxLogs: this.config.maxLogs,
                    logsBuffer: this.logs
                });
                creationTimeout = setTimeout(() => {
                    if (!workerReady) {
                        this.config.logger('Worker creation timeout');
                        worker.off('message', readyHandler);
                        worker.off('error', errorHandler);
                        worker.off('exit', exitHandler);
                        try {
                            worker.kill('SIGTERM');
                        }
                        catch (killError) {
                        }
                        reject(new Error('Worker creation timeout'));
                    }
                }, 10000);
            }
            catch (error) {
                this.config.logger('Error creating worker:', error.message);
                reject(error);
            }
        });
    }
    _handleWorkerExit(code, signal) {
        const exitError = new Error(`Worker exited unexpectedly (code: ${code}, signal: ${signal})`);
        for (const [, request] of this.pendingRequests.entries()) {
            clearTimeout(request.timeoutId);
            request.reject(exitError);
        }
        this.pendingRequests.clear();
        this.worker = null;
        if (this.isShuttingDown) {
            return;
        }
        this.config.logger(`Worker crashed (code: ${code}, signal: ${signal})`);
        this.emit('worker-crash', { code, signal, restartCount: this.restartCount });
        if (this.config.autoRestart && this.restartCount < this.config.maxRestarts) {
            this.restartCount++;
            const msg = `Auto-restarting (attempt ${this.restartCount}/${this.config.maxRestarts})`;
            this.config.logger(`${msg}`);
            setTimeout(() => {
                this._restartWorker().catch((err) => {
                    this.config.logger('Restart failed:', err.message);
                    this.emit('restart-failed', err);
                });
            }, 1000);
        }
        else {
            this.emit('worker-died', { code, signal });
        }
    }
    async _restartWorker() {
        try {
            this.worker = await this._createWorker();
            this._setupWorkerListeners();
            const result = await this._sendMessage('start', {
                handlerPath: this.handlerPath,
                port: this.config.port,
                config: {},
                examples: this.examples,
                requests: this.requests,
                environment: this.environment,
                globals: this.globals,
                scenarios: this.scenarios
            });
            if (result.success) {
                this.serverInfo = result;
                this.config.logger(`Restarted: ${result.url}`);
                this.emit('worker-restarted', { url: result.url, attempt: this.restartCount });
            }
            else {
                throw new Error(result.error || 'Failed to start server in restarted worker');
            }
        }
        catch (error) {
            this.config.logger('Restart failed:', error.message);
            throw error;
        }
    }
    _setupWorkerListeners() {
        if (!this.worker) {
            return;
        }
        this.worker.on('message', (message) => {
            const { id, type, payload } = message;
            if (id && this.pendingRequests.has(id)) {
                const request = this.pendingRequests.get(id);
                clearTimeout(request.timeoutId);
                this.pendingRequests.delete(id);
                if (type === 'success') {
                    request.resolve(payload);
                }
                else if (type === 'error') {
                    request.reject(new Error(payload.error || 'Worker error'));
                }
                else {
                    request.resolve(payload);
                }
                return;
            }
            if (type === 'log') {
                this.logs.push(payload);
                if (this.logs.length > this.config.maxLogs) {
                    this.logs.shift();
                }
                this.emit('log', payload);
            }
            else if (type === 'pm-variable-mutated') {
                this.emit('pm-variable-mutated', payload);
            }
            else if (type === 'worker-error') {
                this.config.logger('Worker error:', payload.error);
                this.emit('worker-error', payload);
                const errorLog = {
                    id: `worker-error-${Date.now()}`,
                    type: 'server_error',
                    timestamp: new Date().toISOString(),
                    message: `Worker error: ${payload.error}`,
                    error: payload.error,
                    errorStack: payload.stack
                };
                this.logs.push(errorLog);
                this.emit('log', errorLog);
                const workerError = new Error(payload.error || 'Worker error');
                for (const [, request] of this.pendingRequests.entries()) {
                    clearTimeout(request.timeoutId);
                    request.reject(workerError);
                }
                this.pendingRequests.clear();
            }
        });
    }
    _sendMessage(type, payload, timeout = 30000) {
        return new Promise((resolve, reject) => {
            if (!this.worker) {
                reject(new Error('No worker available'));
                return undefined;
            }
            const id = ++this.messageIdCounter;
            const timeoutId = setTimeout(() => {
                this.pendingRequests.delete(id);
                reject(new Error('Worker message timeout'));
            }, timeout);
            this.pendingRequests.set(id, { resolve, reject, timeoutId });
            this.worker.send({ id, type, payload }, (error) => {
                if (error) {
                    const request = this.pendingRequests.get(id);
                    if (request) {
                        clearTimeout(request.timeoutId);
                        this.pendingRequests.delete(id);
                    }
                    reject(error);
                }
            });
        });
    }
    async loadHandler(scriptPath) {
        const fs = require('fs');
        const path = require('path');
        if (!fs.existsSync(scriptPath)) {
            throw new Error(`Mock file not found: ${scriptPath}`);
        }
        this.handlerPath = path.resolve(scriptPath);
        return {
            success: true,
            message: 'Handler path validated and stored. Full validation happens in worker.'
        };
    }
    async start() {
        if (!this.handlerPath) {
            throw new Error('No handler loaded. Call loadHandler() first.');
        }
        try {
            this.worker = await this._createWorker();
            this._setupWorkerListeners();
            const result = await this._sendMessage('start', {
                handlerPath: this.handlerPath,
                port: this.config.port,
                config: {},
                examples: this.examples,
                requests: this.requests,
                environment: this.environment,
                globals: this.globals,
                scenarios: this.scenarios
            });
            if (!result.success) {
                throw new Error(result.error || 'Failed to start mock server in worker');
            }
            this.serverInfo = result;
            this.restartCount = 0;
            return {
                port: result.port,
                url: result.url,
                workerId: this.worker.pid
            };
        }
        catch (error) {
            this.config.logger('Failed to start mock server:', error.message);
            if (this.worker) {
                try {
                    this.worker.kill();
                }
                catch (e) {
                }
                this.worker = null;
            }
            throw error;
        }
    }
    async stop() {
        if (!this.worker) {
            return;
        }
        this.isShuttingDown = true;
        try {
            await this._sendMessage('shutdown', {}, 5000);
        }
        catch (error) {
            try {
                this.worker.kill('SIGKILL');
            }
            catch (e) {
            }
        }
        this.worker = null;
        this.serverInfo = null;
        this.isShuttingDown = false;
    }
    getLogs() {
        return [...this.logs];
    }
    clearLogs() {
        this.logs = [];
    }
    getInfo() {
        return {
            port: this.serverInfo ? this.serverInfo.port : null,
            isRunning: Boolean(this.worker),
            workerId: this.worker ? this.worker.pid : null,
            url: this.serverInfo ? this.serverInfo.url : null,
            logCount: this.logs.length,
            restartCount: this.restartCount,
            handlerPath: this.handlerPath
        };
    }
    async updateScenarios(scenarios) {
        if (!this.worker) {
            throw new Error('Server is not running');
        }
        this.scenarios = scenarios;
        return this._sendMessage('update-scenarios', { scenarios });
    }
}
module.exports = {
    SandboxedMockServer
};
