// @postman/pm-bin-linux-x64/bin/lib/commands/flows/run/flows-runner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runFlow = runFlow;
const path_1 = __importDefault(require("path"));
const fs_1 = require("fs");
const _ = __importStar(require("lodash"));
const liquid_json_1 = __importDefault(require("liquid-json"));
const file_utils_1 = require("../../../postman-entities/file-utils");
const flowsCompiler = __importStar(require("./flows-compiler"));
const localExecutor = __importStar(require("./local-executor"));
async function loadScopeFileToVarArray(filePath, workingDir) {
    const resolved = path_1.default.isAbsolute(filePath) ?
        path_1.default.normalize(filePath) :
        path_1.default.resolve(workingDir, filePath);
    const raw = await fs_1.promises.readFile(resolved, 'utf8');
    let data;
    if ((0, file_utils_1.isYamlFile)(resolved)) {
        const yaml = require('js-yaml');
        data = yaml.load(raw);
    }
    else {
        data = liquid_json_1.default.parse(raw.trim());
    }
    const source = (data && data.values) || data;
    if (!source || !Array.isArray(source)) {
        return [];
    }
    return source
        .filter((v) => {
        return typeof v === 'object' && v !== null && !_.isNil(v.key);
    })
        .map((v) => {
        return { key: String(v.key), value: String(v.value) };
    });
}
async function resolveEnvironmentToEnvVar(environmentPath, workingDir) {
    let list = [];
    if (environmentPath && typeof environmentPath === 'string') {
        list = await loadScopeFileToVarArray(environmentPath, workingDir);
    }
    const byKey = new Map(list.map((v) => { return [v.key, v]; }));
    return [...byKey.values()];
}
async function runFlow(flowPath, options = {}) {
    const workingDir = options.workingDir || process.cwd(), looksLikePath = flowPath.includes(path_1.default.sep) ||
        flowPath.startsWith('.') ||
        (/\.(json|postman_flow\.json)$/i).test(flowPath);
    if (looksLikePath) {
        const resolvedPath = path_1.default.isAbsolute(flowPath) ?
            path_1.default.normalize(flowPath) :
            path_1.default.normalize(path_1.default.resolve(workingDir, flowPath)), workingDirResolved = path_1.default.resolve(workingDir), relativeToWorking = path_1.default.relative(workingDirResolved, resolvedPath);
        if (relativeToWorking.startsWith('..') || path_1.default.isAbsolute(relativeToWorking)) {
            const pathError = new Error(`Flow path must be under the working directory: ${flowPath}`);
            pathError.code = 'FLOW_PATH_OUTSIDE_WORKING_DIR';
            throw pathError;
        }
        const raw = await (async () => {
            try {
                await fs_1.promises.access(resolvedPath);
            }
            catch (err) {
                if (err.code === 'ENOENT') {
                    const fileNotFoundError = new Error(`Flow file not found: ${flowPath}`);
                    fileNotFoundError.code = 'FLOW_FILE_NOT_FOUND';
                    throw fileNotFoundError;
                }
                throw err;
            }
            return fs_1.promises.readFile(resolvedPath, 'utf8');
        })();
        let flowSource, compiled, optionsInput;
        try {
            flowSource = JSON.parse(raw);
        }
        catch (parseErr) {
            const error = new Error(`Invalid flow JSON: ${resolvedPath} — ${parseErr.message}`);
            error.code = 'FLOW_JSON_INVALID';
            throw error;
        }
        if (!flowSource) {
            const error = new Error(`Invalid flow JSON: ${resolvedPath}`);
            error.code = 'FLOW_JSON_INVALID';
            throw error;
        }
        const flowSchema = flowSource.flow, flowInputs = flowSchema.io || {}, scenarios = flowSchema.scenarios, requiredInputNames = Object.values(flowInputs)
            .filter((io) => {
            return io && io.type === 'input' && io.name;
        })
            .map((io) => {
            return io.name;
        });
        optionsInput = options.input || {};
        if (options.scenario) {
            if (!scenarios || typeof scenarios !== 'object') {
                const scenarioError = new Error('Flow has no scenarios defined.');
                scenarioError.code = 'FLOW_SCENARIO_NO_SCENARIOS';
                throw scenarioError;
            }
            const scenarioEntry = Object.values(scenarios).find((s) => {
                return s.name === options.scenario;
            });
            if (!scenarioEntry) {
                const names = Object.values(scenarios).map((s) => {
                    return s.name;
                }).filter(Boolean).join(', ') || '(none named)', scenarioNotFoundError = new Error(`Scenario "${options.scenario}" not found. Available: ${names}`);
                scenarioNotFoundError.code = 'FLOW_SCENARIO_NOT_FOUND';
                throw scenarioNotFoundError;
            }
            const scenarioKeyVal = (scenarioEntry.data && typeof scenarioEntry.data === 'object') ?
                Object.values(scenarioEntry.data).reduce((acc, item) => {
                    if (item && item.key !== undefined && item.key !== null) {
                        acc[item.key] = item.value;
                    }
                    return acc;
                }, {}) :
                {};
            optionsInput = { ...scenarioKeyVal, ...optionsInput };
        }
        if (requiredInputNames.length > 0) {
            const missing = requiredInputNames.filter((name) => {
                return optionsInput[name] === undefined;
            });
            if (missing.length > 0) {
                const passedEntries = requiredInputNames
                    .filter((name) => {
                    return optionsInput[name] !== undefined;
                })
                    .map((name) => {
                    return `${name}=${String(optionsInput[name])}`;
                }), inputsError = new Error('The flow requires inputs that were not provided.\n' +
                    `  Required inputs: ${requiredInputNames.join(', ')}\n` +
                    `  Passed/available: ${passedEntries.length ? passedEntries.join(', ') : '(none)'}\n` +
                    `  Missing required inputs: ${missing.join(', ')}`);
                inputsError.code = 'FLOW_INPUTS_MISSING';
                throw inputsError;
            }
        }
        const input = Object.keys(optionsInput).reduce((inputs, key) => {
            const inputId = Object.keys(flowInputs).find((id) => {
                return flowInputs[id].type === 'input' && flowInputs[id].name === key;
            });
            if (inputId) {
                inputs[inputId] = optionsInput[key];
            }
            return inputs;
        }, {}), globals = options.globals || {};
        try {
            compiled = await flowsCompiler.compileFlow(flowSource, {
                basePath: path_1.default.dirname(resolvedPath)
            });
        }
        catch (compileErr) {
            const compileError = new Error(`Flow compilation failed: ${compileErr.message}`);
            compileError.code = 'FLOW_COMPILE_FAILED';
            compileError.cause = compileErr;
            throw compileError;
        }
        let flowResources = null;
        try {
            const flowResourcesPath = path_1.default.join(path_1.default.dirname(resolvedPath), 'flow-resources.json');
            const rawResources = await fs_1.promises.readFile(flowResourcesPath, 'utf8');
            flowResources = JSON.parse(rawResources);
        }
        catch (err) {
            if (err.code !== 'ENOENT') {
                throw err;
            }
        }
        const flowId = flowSource.info._postman_id;
        const dependencies = flowResources?.flows?.[flowId]?.dependencies ?? [];
        const resolvedEnvVar = await resolveEnvironmentToEnvVar(options.environment, workingDir);
        const runResult = await localExecutor.runCompiled(compiled, input, globals, {
            verbose: options.verbose,
            collections: options.collections,
            postmanApiKey: options.postmanApiKey,
            accessToken: options.accessToken,
            flowSource: flowSource,
            flowDependencies: dependencies,
            envVar: resolvedEnvVar
        });
        if (!runResult.success) {
            const runError = new Error(runResult.error || 'Flow run failed');
            runError.code = 'FLOW_RUN_FAILED';
            runError.failedNodeId = runResult?.failedNodeId;
            runError.failedNodeSpan = runResult?.failedNodeSpan;
            runError.logs = runResult.logs;
            runError.outputs = runResult.outputs || {};
            runError.flow = flowSchema;
            runError.inputs = optionsInput;
            throw runError;
        }
        const outputs = Object.keys(runResult.outputs || {}).reduce((outputs, key) => {
            const entry = Object.entries(flowInputs).find(([id, io]) => {
                return io.type === 'output' && id === key;
            });
            const output = entry ? entry[1] : null;
            if (output?.name) {
                outputs[output.name] = runResult?.outputs?.[key];
            }
            return outputs;
        }, {});
        return {
            success: true,
            flow: flowSource,
            compiled: compiled,
            outputs: outputs,
            inputs: optionsInput,
            logs: runResult.logs
        };
    }
    const idError = new Error('Running flows by ID is not supported with `postman flows run` yet. ' +
        'Please provide a path to a local flow JSON file.');
    idError.code = 'FLOW_ID_NOT_SUPPORTED';
    throw idError;
}
exports.default = {
    runFlow
};
