// @postman/pm-bin-linux-x64/bin/lib/commands/flows/run/flows-run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runFlowsCommand = runFlowsCommand;
const asyncWaterfall = require('async/waterfall');
const colors = require('colors/safe');
const cmdUtil = require('../../cmd-util');
const libUtil = require('../../../util');
const config = require('../../../config');
const consoleFramework = require('../../../framework/console');
const print = require('../../../framework/print');
const flowsRunner = require('./flows-runner');
const printer = require('./log-printer');
const userType = require('../../../user-type');
const inputFunctions = require('./input-functions');
const { processInputFiles } = inputFunctions;
function runFlowsCommand(pathArg, command) {
    let options;
    command.alias = 'default';
    asyncWaterfall([
        (next) => {
            return config.get(command, { command: 'run' }, next);
        },
        (commandFromConfig, next) => {
            options = cmdUtil.commanderToObject(commandFromConfig);
            options.path = pathArg;
            try {
                let keyValueInputs = {}, fileInputs = {};
                if (Array.isArray(options.input)) {
                    keyValueInputs = options.input.reduce((acc, { key, value }) => {
                        acc[key] = value;
                        return acc;
                    }, {});
                }
                if (Array.isArray(options.inputFile)) {
                    fileInputs = processInputFiles(options.inputFile);
                }
                options.input = { ...fileInputs, ...keyValueInputs };
            }
            catch (inputError) {
                return next(inputError);
            }
            return next(null, options);
        },
        (opts, next) => {
            libUtil.extractPostmanApiKey(opts.postmanApiKey, (err, postmanApiKey) => {
                opts.postmanApiKey = postmanApiKey;
                return next(err, opts);
            });
        }
    ], async (err, opts) => {
        if (err) {
            consoleFramework.error(`Error: ${err.message || err}\n`);
            if (err.help) {
                consoleFramework.error(err.help);
            }
            process.exit(1);
        }
        if (!opts.postmanApiKey && !opts.accessToken) {
            consoleFramework.error('Error: Authentication required. Please run `postman login` to authenticate.\n');
            process.exit(1);
        }
        let resolvedUserType, exitCode = 0;
        try {
            resolvedUserType = await userType.getUserType(opts.postmanApiKey, opts.accessToken);
        }
        catch (userTypeErr) {
            consoleFramework.error(`Error: ${userTypeErr.message}\n`);
            process.exit(1);
        }
        if (resolvedUserType !== 'enterprise') {
            consoleFramework.error(colors.yellow('Flows Runner is available to Enterprise accounts for running automations on-premise with private APIs.'));
            consoleFramework.error('');
            consoleFramework.error(colors.cyan('  • Already on an Enterprise team?'), 'Log in to access it', colors.gray('(postman login)'));
            consoleFramework.error(colors.cyan('  • Want this feature?'), 'Upgrade to Enterprise', colors.gray('(https://www.postman.com/pricing/)'));
            consoleFramework.error(colors.cyan('  • Not on Enterprise?'), 'You can run flows automatically via API triggers and webhooks.');
            consoleFramework.error('    First deploy your flows, then use the Collection Runner or a direct API call.');
            consoleFramework.error('    Learn more:', colors.gray('https://learning.postman.com/docs/postman-flows/build-flows/structure/actions'));
            consoleFramework.error('');
            process.exit(1);
        }
        if (!opts.path) {
            consoleFramework.error('Error: Path is required. Please provide a path to the flow.\n');
            process.exit(1);
        }
        try {
            if (!opts.silent) {
                print('Running flow... ');
                print.wait(colors.cyan);
            }
            const result = await flowsRunner.runFlow(opts.path, opts);
            exitCode = result && result.success ? 0 : 1;
            if (opts.suppressExitCode) {
                exitCode = 0;
            }
            if (!opts.silent && result && result.success) {
                printer.displayExecutionResults(result, { verbose: opts.verbose });
            }
        }
        catch (runErr) {
            if (!opts.silent) {
                print.unwait(true);
            }
            consoleFramework.error(`\nError: ${runErr.message || runErr}\n`);
            if (opts.verbose && runErr.logs && runErr.logs.length > 0) {
                printer.displayExecutionResults({
                    success: false,
                    logs: runErr.logs,
                    flow: runErr.flow,
                    outputs: runErr.outputs || {},
                    inputs: runErr.inputs || {}
                }, { verbose: true });
            }
            exitCode = 1;
            if (opts.suppressExitCode) {
                exitCode = 0;
            }
        }
        process.exit(exitCode);
    });
}
