// @postman/pm-bin-linux-x64/bin/lib/commands/flows/run/flows-compiler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfigTransformer = void 0;
exports.genSourceId = genSourceId;
exports.extractModuleFlowIds = extractModuleFlowIds;
exports.loadModuleFlow = loadModuleFlow;
exports.buildModuleSource = buildModuleSource;
exports.buildBlockListFromLib = buildBlockListFromLib;
exports.compileFlow = compileFlow;
const path_1 = __importDefault(require("path"));
const fs_1 = require("fs");
const _ = __importStar(require("lodash"));
exports.defaultConfigTransformer = {
    typescript: (code) => {
        if (typeof code !== 'string') {
            return 'throw new Error(\'Typescript compiler unavailable\')';
        }
        return code;
    }
};
function genSourceId(flowId, versionId) {
    return `${flowId}@${versionId}`;
}
function extractModuleFlowIds(nodes) {
    if (!nodes || typeof nodes !== 'object') {
        return [];
    }
    return Object.values(nodes)
        .filter((node) => {
        return node && node.type === 'system/module@1' && node.config && node.config.flowId;
    })
        .map((node) => {
        return node.config.flowId;
    })
        .filter((flowId) => {
        return Boolean(flowId);
    });
}
async function loadModuleFlow(flowId, basePath) {
    const basePathResolved = path_1.default.resolve(basePath), resolvedPath = path_1.default.isAbsolute(flowId) ?
        path_1.default.normalize(flowId) :
        path_1.default.normalize(path_1.default.join(basePathResolved, flowId)), relativeToBase = path_1.default.relative(basePathResolved, path_1.default.resolve(resolvedPath));
    if (relativeToBase.startsWith('..') || path_1.default.isAbsolute(relativeToBase)) {
        throw new Error(`Module flow path escapes base directory: "${flowId}". ` +
            'Resolved path must be under the main flow directory.');
    }
    const raw = await fs_1.promises.readFile(resolvedPath, 'utf8'), parsed = JSON.parse(raw);
    return _.isNil(parsed.flow) ? parsed : parsed.flow;
}
function buildModuleSource(moduleId, flow, blockList, MSource, MTypeNode, MConnection) {
    const source = new MSource(moduleId), nodes = flow.nodes || {}, connections = flow.connections || {};
    if (flow.scenarios || flow.io) {
        source.setMeta({
            module: {
                scenarios: flow.scenarios || {},
                io: flow.io || {}
            }
        });
    }
    for (const [id, node] of Object.entries(nodes)) {
        const def = blockList[node.type];
        if (!def) {
            throw new Error(`Unknown block type: ${node.type} (node ${id}) in module ${moduleId}. ` +
                'Ensure @postman/flows-block-lib includes this block.');
        }
        source.setNode(new MTypeNode(id, node.type, def.block, node.config || {}));
    }
    for (const [id, connection] of Object.entries(connections)) {
        source.setConnection(new MConnection(id, [connection.source, connection.sourcePort], [connection.target, connection.targetPort]));
    }
    if (flow.io) {
        Object.entries(flow.io).forEach(([id, io]) => {
            if (io.type === 'input') {
                source.setInput(id);
            }
            else if (io.type === 'output') {
                source.setOutput(id);
            }
        });
    }
    return source;
}
function buildBlockListFromLib(Blocks) {
    if (!Blocks || typeof Blocks !== 'object') {
        return {};
    }
    const list = {}, entries = Object.values(Blocks).filter((b) => {
        return b && typeof b.id === 'string';
    });
    entries.forEach((blockDef) => {
        list[blockDef.id] = { block: blockDef };
    });
    return list;
}
async function compileFlow(flowSource, options = {}) {
    const flow = flowSource.flow, nodes = flow?.nodes || {}, connections = flow?.connections || {}, sourceId = 'main', basePath = options.basePath || '';
    if (!flow || typeof flow !== 'object') {
        throw new Error('Flow schema must be a non-null object');
    }
    let MSource;
    let MCompiler;
    let MConnection;
    let MTypeNode;
    let blockList;
    let allCode;
    try {
        const compiler = require('@postman/merlin-lang.compiler');
        MSource = compiler.MSource;
        MCompiler = compiler.MCompiler;
        MConnection = compiler.MConnection;
        MTypeNode = compiler.MTypeNode;
        if (!MSource || !MCompiler || !MConnection || !MTypeNode) {
            throw new Error('@postman/merlin-lang.compiler did not export MSource, MCompiler, MConnection, MTypeNode');
        }
    }
    catch (err) {
        if (err.code === 'MODULE_NOT_FOUND' || err.message?.includes('Cannot find module')) {
            throw new Error('Flow compilation requires @postman/merlin-lang.compiler (and optionally @postman/flows-block-lib).');
        }
        throw err;
    }
    try {
        const { Blocks } = require('@postman/flows-block-lib');
        blockList = buildBlockListFromLib(Blocks);
    }
    catch (err) {
        if (err.code === 'MODULE_NOT_FOUND' || err.message?.includes('Cannot find module')) {
            throw new Error('Flow compilation requires @postman/flows-block-lib for block definitions.');
        }
        throw err;
    }
    const configTransformer = options.configTransformer || exports.defaultConfigTransformer, source = new MSource(sourceId), modulesMap = new Map(), depQueue = extractModuleFlowIds(nodes), processedFlowIds = new Set();
    if (flow.scenarios || flow.io) {
        source.setMeta({
            module: {
                scenarios: flow.scenarios || {},
                io: flow.io || {}
            }
        });
    }
    for (const [id, node] of Object.entries(nodes)) {
        const def = blockList[node.type];
        if (!def) {
            throw new Error(`Unknown block type: ${node.type} (node ${id}). ` +
                'Ensure @postman/flows-block-lib includes this block.');
        }
        source.setNode(new MTypeNode(id, node.type, def.block, node.config || {}));
    }
    for (const [id, connection] of Object.entries(connections)) {
        source.setConnection(new MConnection(id, [connection.source, connection.sourcePort], [connection.target, connection.targetPort]));
    }
    if (flow.io) {
        Object.entries(flow.io).forEach(([id, io]) => {
            if (io.type === 'input') {
                source.setInput(id);
            }
            else if (io.type === 'output') {
                source.setOutput(id);
            }
        });
    }
    while (depQueue.length > 0) {
        const flowId = depQueue.shift(), versionId = 'local', moduleId = genSourceId(flowId, versionId);
        if (!flowId || processedFlowIds.has(flowId)) {
            continue;
        }
        processedFlowIds.add(flowId);
        if (modulesMap.has(moduleId)) {
            continue;
        }
        let moduleFlow;
        try {
            moduleFlow = await loadModuleFlow(flowId, basePath);
        }
        catch (err) {
            throw new Error(`Failed to load module flow "${flowId}" (${err.message}). ` +
                'Ensure basePath is set to the main flow directory when using sub-flows.');
        }
        const moduleSource = buildModuleSource(moduleId, moduleFlow, blockList, MSource, MTypeNode, MConnection), nestedFlowIds = extractModuleFlowIds(moduleFlow.nodes || {});
        modulesMap.set(moduleId, moduleSource);
        nestedFlowIds
            .filter((id) => {
            return !processedFlowIds.has(id);
        })
            .forEach((id) => {
            depQueue.push(id);
        });
    }
    const mergedSources = [source, ...modulesMap.values()], mainCompiler = new MCompiler(sourceId, mergedSources, configTransformer), code = mainCompiler.compile();
    allCode = { [sourceId]: code };
    if (modulesMap.size > 0) {
        const moduleSourcesList = [...modulesMap.values()];
        for (const [moduleId] of modulesMap) {
            const moduleCompiler = new MCompiler(moduleId, moduleSourcesList, configTransformer);
            allCode[moduleId] = moduleCompiler.compile();
        }
    }
    return Promise.resolve({ code, sourceId, allCode });
}
