// @postman/pm-bin-linux-x64/bin/lib/commands/flows/run/local-executor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildDispatches = buildDispatches;
exports.runCompiled = runCompiled;
const events_1 = __importDefault(require("events"));
const fs_1 = __importDefault(require("fs"));
const _ = __importStar(require("lodash"));
const run_request_from_file_1 = require("../run-request-from-file");
const ai_client_1 = __importDefault(require("../api/ai-client"));
function buildDispatches() {
    const { Dispatch: BlockDispatch } = require('@postman/flows-block-lib'), extensions = new Map();
    extensions.set('ev/endpoint@1', BlockDispatch.StartV2('local'));
    extensions.set('ev/endpoint@2', BlockDispatch.StartInput('local'));
    extensions.set('ev/endpoint@3', BlockDispatch.StartInputV3('local'));
    extensions.set('logic/manipulate@1', BlockDispatch.Evaluate);
    extensions.set('logic/if@2', BlockDispatch.If2);
    extensions.set('logic/condition@1', BlockDispatch.Condition);
    extensions.set('logic/validate@1', BlockDispatch.ValidateSchema);
    if (BlockDispatch.Or) {
        extensions.set('logic/or@1', BlockDispatch.Or);
    }
    extensions.set('task/http-request@1', BlockDispatch.Request);
    extensions.set('task/integration@1', BlockDispatch.IntegrationBlock);
    extensions.set('value/reference@2', BlockDispatch.Reference);
    extensions.set('value/static-reference@1', BlockDispatch.Reference);
    extensions.set('value/template@1', BlockDispatch.Template);
    extensions.set('portal/in@1', BlockDispatch.PortalValue);
    extensions.set('portal/out@2', BlockDispatch.PortalValue);
    extensions.set('portal/config@1', BlockDispatch.ConfigBlock);
    extensions.set('output/viewer@1', BlockDispatch.Viewer);
    extensions.set('LIST_TO_STREAM', BlockDispatch.For);
    extensions.set('STREAM_TO_LIST', BlockDispatch.Collect);
    extensions.set('LOOP_N_TIME', BlockDispatch.LoopNTimes);
    extensions.set('TERMINAL', BlockDispatch.Log);
    extensions.set('DELAY', BlockDispatch.Delay);
    extensions.set('TYPE_BOOLEAN', BlockDispatch.Value);
    extensions.set('TYPE_NULL', BlockDispatch.NullValue);
    extensions.set('TYPE_NUMBER', BlockDispatch.Value);
    extensions.set('TYPE_STRING', BlockDispatch.Value);
    extensions.set('TYPE_REGEX', BlockDispatch.RegexValue);
    extensions.set('TYPE_NOW', BlockDispatch.NowValue);
    extensions.set('TYPE_DATE', BlockDispatch.DateValue);
    extensions.set('TYPE_DATETIME', BlockDispatch.DateTimeValue);
    extensions.set('TYPE_LIST', BlockDispatch.Value);
    extensions.set('TYPE_RECORD', BlockDispatch.Value);
    extensions.set('port/output@1', BlockDispatch.PortOutput);
    extensions.set('ev/output@1', BlockDispatch.FlowOutput);
    extensions.set('ev/output@2', BlockDispatch.FlowOutput2);
    extensions.set('system/module@1', BlockDispatch.ModuleBlock);
    extensions.set('action/request@1', BlockDispatch.ActionRequest);
    extensions.set('action/response@1', BlockDispatch.ActionResponse('local'));
    extensions.set('action/schedule@1', BlockDispatch.Schedule);
    if (BlockDispatch.AIRequest) {
        extensions.set('task/ai-request@1', BlockDispatch.AIRequest);
    }
    if (BlockDispatch.AIGenerate) {
        extensions.set('ai/generative@1', BlockDispatch.AIGenerate('local'));
    }
    if (BlockDispatch.AIAgent) {
        extensions.set('ai/actor@1', BlockDispatch.AIAgent);
    }
    if (BlockDispatch.AIAgentV2) {
        extensions.set('ai/agent@2', BlockDispatch.AIAgentV2);
    }
    return [...extensions.entries()];
}
function getNodeLabel(flowSchema, processId) {
    if (!processId) {
        return processId || '';
    }
    const nodeId = processId.split('/').pop() || '', nodes = (flowSchema && flowSchema.flow && flowSchema.flow.nodes) || (flowSchema && flowSchema.nodes) || {}, node = nodes[nodeId];
    return (node && node.type) ? `${node.type} (${nodeId})` : processId;
}
async function runCompiled(compiled, input = {}, globals = {}, options = {}) {
    const { VM, Summarizer } = require('@postman/merlin-lang.vm');
    const { executeJavascript } = require('@postman/merlin-lang.quickjs');
    const TJ = require('teleport-javascript'), stdlibPath = require.resolve('@postman/merlin-lang.quickjs/dist/stdlib.bundle.js'), stdlib = fs_1.default.readFileSync(stdlibPath, 'utf8'), dispatches = buildDispatches(), files = compiled.allCode || { [compiled.sourceId]: compiled.code }, entryPoint = compiled.sourceId || 'main', summarizer = new Summarizer(files, {}), events = new events_1.default(), logs = [], sessionErrors = [], toMessageValue = (value) => {
        const hasToJS = typeof value === 'object' && value !== null && typeof value.toJS === 'function';
        return hasToJS ? value.toJS() : value;
    };
    events.on('terminal', (idSpan, value) => {
        logs.push({
            type: 'terminal',
            level: 'info',
            processId: idSpan && idSpan.id,
            span: idSpan && idSpan.span,
            message: toMessageValue(value)
        });
    });
    events.on('visualization', (idSpan, value) => {
        logs.push({
            type: 'visualization',
            level: 'info',
            processId: idSpan && idSpan.id,
            span: idSpan && idSpan.span,
            message: toMessageValue(value)
        });
    });
    events.on('detail', (idSpan, value) => {
        logs.push({
            type: 'detail',
            level: 'info',
            processId: idSpan && idSpan.id,
            span: idSpan && idSpan.span,
            message: toMessageValue(value)
        });
    });
    const observer = {
        onProcessExecuting(process) {
            const inputs = process.inputs.map((port) => {
                const logVal = port.log();
                return { name: port.name, value: TJ.stringify(logVal) };
            });
            logs.push({
                type: 'process:executing',
                event: 'process:executing',
                processId: process.id,
                span: process.span,
                timestamp: Date.now(),
                inputs: inputs
            });
        },
        onProcessExecuted(process) {
            const outputs = process.outputs.map((port) => {
                const logVal = port.log();
                return { name: port.name, value: TJ.stringify(logVal) };
            });
            logs.push({
                type: 'process:executed',
                event: 'process:executed',
                processId: process.id,
                span: process.span,
                timestamp: Date.now(),
                outputs: outputs
            });
        },
        onProcessError(process, error) {
            const errMsg = error.message || String(error);
            sessionErrors.push({
                type: 'process',
                error: errMsg,
                processId: process.id,
                span: process.span
            });
            logs.push({
                type: 'process:error',
                event: 'process:error',
                processId: process.id,
                span: process.span,
                timestamp: Date.now(),
                error: errMsg
            });
        },
        onProcessTerminated(process) {
            logs.push({
                type: 'process:terminated',
                event: 'process:terminated',
                processId: process.id,
                span: process.span,
                timestamp: Date.now()
            });
        },
        onProcessBreak() { },
        onExecutionStart() { },
        onExecutionEnd() { },
        onExecutionError() { },
        onInputArgv() { },
        onOutputArgv() { },
        onSchedulerError(error) {
            const errMsg = error.message || String(error);
            sessionErrors.push({ type: 'scheduler', error: errMsg });
        }
    }, callAsync = async (topic, ...data) => {
        if (topic === 'request') {
            const requestData = Array.isArray(data[0]) ? data[0][0] : data[0];
            if (!requestData) {
                throw new Error('Request block requires collection/request context. Use --collection for local runs.');
            }
            const collectionId = requestData.collection;
            const requestId = requestData.request;
            const [variables] = requestData.variables ?? [];
            if (!collectionId || !requestId) {
                throw new Error('Request block requires both collection and request identifiers.');
            }
            const flowDependencies = options.flowDependencies ?? [];
            const dependency = flowDependencies.find(({ local }) => { return local.collectionId === collectionId && local.requestId === requestId; });
            const requestFilePath = dependency?.local?.filePath;
            if (requestFilePath) {
                const runOpts = { variables };
                if (Array.isArray(options.envVar) && options.envVar.length > 0) {
                    runOpts.envVar = options.envVar;
                }
                return (0, run_request_from_file_1.runRequestFromFile)(requestFilePath, runOpts);
            }
            throw new Error('Request block requires a dependency entry for this collection/request. ' +
                'Use flow-resources.json to map the request to a local *.request.yaml file.');
        }
        if (topic === 'ai-actor') {
            const param1 = Array.isArray(data[0]) ? data[0][0] : data[0];
            if (!param1) {
                throw new Error('AI Actor block requires provider, model, messages, tools, and optional outputSchema.');
            }
            return ai_client_1.default.actorChat({
                accessToken: options.accessToken,
                debug: options.debug,
                ...param1
            });
        }
        if (topic === 'ai-agent-v2') {
            const param1 = Array.isArray(data[0]) ? data[0][0] : data[0];
            if (!param1) {
                throw new Error('AI Agent block requires model, systemPrompt, context, mcpServers, and optional outputSchema.');
            }
            return ai_client_1.default.agentExecute({
                accessToken: options.accessToken,
                debug: options.debug,
                ...param1
            });
        }
        if (topic === 'generateText') {
            const param1 = Array.isArray(data[0]) ? data[0][0] : data[0];
            if (!param1) {
                throw new Error('Generate Text block requires prompt, and optional predictability, creativity, model.');
            }
            const response = await ai_client_1.default.generateText({
                accessToken: options.accessToken,
                debug: options.debug,
                ...param1
            });
            return (response && response.text !== undefined && response.text !== null) ?
                String(response.text) :
                '';
        }
        if (topic === 'generateAuto') {
            const param1 = Array.isArray(data[0]) ? data[0][0] : data[0];
            if (!param1) {
                throw new Error('Generate Auto block requires prompt, and optional predictability, creativity, model.');
            }
            const responseBody = await ai_client_1.default.generateAuto({
                accessToken: options.accessToken,
                debug: options.debug,
                ...param1
            }), intent = responseBody && responseBody.intent, hasContent = responseBody && responseBody.content !== undefined &&
                responseBody.content !== null, content = hasContent ? responseBody.content : '';
            if (intent === 'generate_text' || intent === 'generate_json') {
                return { type: 'text', result: content };
            }
            if (intent === 'generate_image') {
                return { type: 'image', result: Buffer.from(content, 'base64') };
            }
            if (responseBody && responseBody.content !== undefined) {
                return { type: 'text', result: String(responseBody.content || '') };
            }
            throw new Error('An unexpected error occurred, please try again.');
        }
        if (topic === 'getConsentStatus') {
            return true;
        }
        throw new Error(`callAsync not implemented: ${topic}`);
    }, vm = new VM(files, globals || {}, {
        io: {
            dispatch: (language, script, inputPayload) => {
                if (language === 'javascript') {
                    const stdlibArr = [stdlib];
                    return executeJavascript(script, inputPayload, {
                        stdlib: stdlibArr,
                        print: (level, message) => {
                            const parsed = TJ.parse(message), levelVal = level, messageVal = parsed;
                            logs.push({
                                type: 'console',
                                level: levelVal,
                                message: messageVal
                            });
                        }
                    });
                }
                throw new Error(`Unsupported language: ${language}`);
            },
            callAsync: callAsync,
            callSync: () => {
                throw new Error('callSync not implemented');
            }
        },
        dispatches: dispatches,
        counter: summarizer,
        events: events,
        process: { autoLatch: false },
        observer: observer
    }), argv = Object.entries(input), runOptions = { debugger: null };
    try {
        const result = await vm.run(entryPoint, argv, runOptions), outputs = {};
        if (Array.isArray(result)) {
            result.forEach(([key, value]) => {
                outputs[key] = value;
            });
        }
        return { success: true, outputs: outputs, logs: logs };
    }
    catch (err) {
        const flowSchema = options.flowSource, sessionDetail = sessionErrors.length ?
            sessionErrors.map((e) => {
                return _.isNil(e.processId) ?
                    e.error :
                    `${getNodeLabel(flowSchema, e.processId)} failed: ${e.error}`;
            }).join('\n') :
            '', detail = sessionDetail ? `${err.message}\n${sessionDetail}` : err.message, firstProcessError = sessionErrors.find((e) => { return !_.isNil(e.processId); });
        return {
            success: false,
            error: detail,
            logs: logs,
            failedNodeId: firstProcessError && firstProcessError.processId,
            failedNodeSpan: firstProcessError && firstProcessError.span
        };
    }
}
