// @postman/pm-bin-linux-x64/bin/lib/commands/flows/run/request-handler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchRequest = fetchRequest;
exports.executeRequest = executeRequest;
const api_1 = require("../../../api");
const util = __importStar(require("../../../util"));
const postman_request_1 = __importDefault(require("postman-request"));
function applyVariables(value, variables) {
    if (!value || !variables || typeof value !== 'string') {
        return value;
    }
    let nextValue = value;
    Object.keys(variables).forEach((key) => {
        nextValue = nextValue.split(`{{${key}}}`).join(variables[key]);
    });
    return nextValue;
}
function buildUrl(req, variables) {
    if (!req || !req.url) {
        return '';
    }
    if (typeof req.url === 'string') {
        return applyVariables(req.url, variables);
    }
    if (req.url.raw) {
        let raw = applyVariables(req.url.raw, variables);
        if (!raw.startsWith('http://') && !raw.startsWith('https://')) {
            raw = `https://${raw}`;
        }
        return raw;
    }
    const protocol = req.url.protocol || 'https', host = Array.isArray(req.url.host) ? req.url.host.join('.') : (req.url.host || ''), pathPart = Array.isArray(req.url.path) ? `/${req.url.path.join('/')}` : (req.url.path || ''), query = Array.isArray(req.url.query) ?
        `?${req.url.query
            .filter((q) => { return !q.disabled; })
            .map((q) => {
            return `${encodeURIComponent(q.key)}=${encodeURIComponent(q.value)}`;
        })
            .join('&')}` :
        '';
    return applyVariables(`${protocol}://${host}${pathPart}${query}`, variables);
}
async function fetchRequest(collectionId, requestId, options = {}) {
    if (!collectionId) {
        throw new Error('collectionId is required to fetch a request');
    }
    if (!requestId) {
        throw new Error('requestId is required to fetch a request');
    }
    let postmanApiKey, accessToken, collectionResource, collection;
    postmanApiKey = util.getValidPostmanAPIKey({ postmanApiKey: options.postmanApiKey });
    accessToken = options.accessToken;
    if (!postmanApiKey && !accessToken) {
        throw new Error('No authorization data found. Please use the `postman login` command.');
    }
    const requestResource = new api_1.PostmanResource('request', requestId, postmanApiKey, accessToken);
    const request = await new Promise((resolve, reject) => {
        requestResource.get((err, data) => {
            if (err) {
                return reject(err);
            }
            return resolve(data);
        });
    });
    collectionResource = new api_1.PostmanResource('collection', collectionId, postmanApiKey, accessToken);
    collection = await new Promise((resolve, reject) => {
        collectionResource.get((err, data) => {
            if (err) {
                return reject(err);
            }
            return resolve(data);
        });
    });
    if (!request) {
        throw new Error(`Request not found: ${requestId} in collection ${collectionId}`);
    }
    return {
        collection: collection,
        request: request
    };
}
function executeRequest(input, options = {}) {
    if (!input) {
        return Promise.reject(new Error('Request payload is required to execute a request'));
    }
    let req, variables, url, headers, requestOptions;
    req = input.request || input;
    variables = input.variables || options.variables || {};
    url = buildUrl(req, variables);
    headers = {};
    (req.header || []).forEach((header) => {
        if (header && !header.disabled) {
            headers[header.key] = applyVariables(header.value || '', variables);
        }
    });
    requestOptions = {
        url: url,
        method: req.method || 'GET',
        headers: headers,
        timeout: options.timeoutRequest || options.timeout || 30000,
        followRedirects: typeof options.ignoreRedirects === 'boolean' ? !options.ignoreRedirects : undefined,
        strictSSL: typeof options.insecure === 'boolean' ? !options.insecure : undefined
    };
    if (req.body && req.body.mode === 'raw' && req.body.raw) {
        requestOptions.body = applyVariables(req.body.raw, variables);
    }
    return new Promise((resolve, reject) => {
        (0, postman_request_1.default)(requestOptions, (error, response, body) => {
            if (error) {
                return reject(error);
            }
            let parsedBody = body;
            try {
                if (typeof body === 'string') {
                    parsedBody = JSON.parse(body);
                }
            }
            catch {
                parsedBody = body;
            }
            const responseHeaders = Object.entries(response.headers || {}).map(([key, value]) => {
                return { key: key, value: value };
            });
            return resolve({
                response: {
                    code: response.statusCode,
                    status: response.statusMessage,
                    headers: responseHeaders,
                    body: typeof parsedBody === 'string' ? parsedBody : JSON.stringify(parsedBody),
                    responseTime: response.elapsedTime || 0,
                    contentInfo: { mimeFormat: 'json' }
                },
                assertions: []
            });
        });
    });
}
exports.default = {
    fetchRequest,
    executeRequest
};
