// @postman/pm-bin-linux-x64/bin/lib/commands/runner/metrics/index.js
const http = require('http'), { getLogger } = require('../logging/logger-factory'), logger = getLogger();
class MetricsServer {
    constructor({ port = 9090, host = '0.0.0.0', runnerId }) {
        this.port = port;
        this.host = host;
        this.runnerId = runnerId;
        this.server = null;
        this.isRunning = false;
        this.startTime = new Date();
        this.readyState = true;
        this.version = require('../../../../package.json').version;
    }
    _getHealthStatus() {
        return {
            status: 'ok',
            timestamp: new Date().toISOString(),
            uptime: Math.floor((Date.now() - this.startTime) / 1000),
            runnerId: this.runnerId,
            version: this.version
        };
    }
    _getReadinessStatus() {
        return {
            status: this.readyState ? 'ready' : 'not_ready',
            timestamp: new Date().toISOString(),
            runnerId: this.runnerId,
            version: this.version
        };
    }
    start() {
        return new Promise((resolve, reject) => {
            if (this.isRunning) {
                resolve();
                return;
            }
            this.server = http.createServer((req, res) => {
                if (req.method !== 'GET') {
                    MetricsServer._respond(res, 405, { error: 'Method Not Allowed' });
                    return;
                }
                if (req.url === '/health/live') {
                    MetricsServer._respond(res, 200, this._getHealthStatus());
                    return;
                }
                if (req.url === '/health/ready') {
                    const readiness = this._getReadinessStatus(), statusCode = readiness.status === 'ready' ? 200 : 503;
                    MetricsServer._respond(res, statusCode, readiness);
                    return;
                }
                MetricsServer._respond(res, 404, {
                    error: 'Not Found',
                    availableEndpoints: ['/health/live', '/health/ready']
                });
            });
            this.server.on('error', (err) => {
                const msg = err.code === 'EADDRINUSE' ?
                    'Port ' + this.port + ' is already in use' : err.message;
                logger.error('Metrics server error', {
                    errorMessage: msg,
                    errorCode: err.code,
                    port: this.port
                });
                reject(err);
            });
            this.server.listen(this.port, this.host, () => {
                this.isRunning = true;
                logger.info('Metrics server started', {
                    host: this.host,
                    port: this.port,
                    endpoints: ['/health/live', '/health/ready']
                });
                resolve();
            });
        });
    }
    shutdown() {
        return new Promise((resolve) => {
            if (!this.isRunning || !this.server) {
                resolve();
                return;
            }
            this.server.close(() => {
                this.isRunning = false;
                logger.info('Metrics server stopped');
                resolve();
            });
            setTimeout(() => {
                if (this.isRunning && this.server.closeAllConnections) {
                    this.server.closeAllConnections();
                    this.isRunning = false;
                }
                resolve();
            }, 5000);
        });
    }
    static _respond(res, statusCode, body) {
        res.writeHead(statusCode, {
            'Content-Type': 'application/json',
            'Cache-Control': 'no-cache, no-store, must-revalidate'
        });
        res.end(JSON.stringify(body, null, 2));
    }
}
module.exports = MetricsServer;
