// @postman/pm-bin-linux-x64/bin/lib/commands/runner/proxy/authorization/service.js
const request = require('postman-request');
class AuthorizationService {
    constructor(logger, baseUrl) {
        if (!baseUrl) {
            throw new Error('Authorization Service base URL is required');
        }
        this._logger = logger;
        this._baseUrl = baseUrl;
        this._timeout = 5000;
    }
    checkHealth() {
        const healthUrl = `${this._baseUrl}/health`;
        return new Promise((resolve) => {
            request.get({
                url: healthUrl,
                timeout: this._timeout,
                strictSSL: false
            }, (error, response) => {
                if (error) {
                    this._logger.error(`Authorization service is not reachable: ${error.message}`);
                    resolve(false);
                    return;
                }
                if (response.statusCode === 200 || response.statusCode === 404) {
                    this._logger.info('Authorization service is reachable');
                    resolve(true);
                    return;
                }
                this._logger.warn(`Authorization service returned status: ${response.statusCode}`);
                resolve(false);
            });
        });
    }
    validateRequest(validationData) {
        const authzUrl = `${this._baseUrl}/validate`, urlPath = validationData.url ? validationData.url.split('?')[0] : 'unknown';
        this._logger.info(`Making authorization request for: ${validationData.method || 'UNKNOWN'} ${urlPath}`);
        return new Promise((resolve) => {
            request.post({
                url: authzUrl,
                body: validationData,
                json: true,
                timeout: this._timeout,
                strictSSL: false,
                headers: {
                    'Content-Type': 'application/json'
                }
            }, (error, response) => {
                if (error) {
                    this._handleAuthorizationError(error);
                    resolve(false);
                    return;
                }
                const statusMsg = response.statusCode === 200 ?
                    'Authorization successful, allowing request to proceed' :
                    `Authorization denied by validation service with status: ${response.statusCode}`;
                if (response.statusCode === 200) {
                    this._logger.info(statusMsg);
                    resolve(true);
                    return;
                }
                this._logger.error(statusMsg);
                resolve(false);
            });
        });
    }
    _handleAuthorizationError(authzError) {
        if (authzError.code === 'ETIMEDOUT' || authzError.code === 'ESOCKETTIMEDOUT') {
            this._logger.error('Authorization service timeout');
            return;
        }
        if (authzError.code === 'ECONNREFUSED') {
            this._logger.error('Authorization service connection refused - service may be down');
            return;
        }
        const connectionErrMsg = `Authorization service error: ${authzError.message}`;
        this._logger.error(connectionErrMsg);
    }
}
module.exports = AuthorizationService;
