// @postman/pm-bin-linux-x64/bin/lib/commands/runner/proxy/authorization/validator.js
const { processRequestHeaders, headersToJson } = require('../utils/request-utils'), AuthorizationService = require('./service');
class AuthorizationValidator {
    constructor(logger, baseUrl) {
        if (!baseUrl) {
            throw new Error('Authorization Service base URL is required');
        }
        this._logger = logger;
        this._authService = new AuthorizationService(logger, baseUrl);
    }
    async validateRequest(requestContext) {
        try {
            const validationData = AuthorizationValidator._buildValidationData(requestContext);
            return await this._authService.validateRequest(validationData);
        }
        catch (authError) {
            this._logger.error(`Authorization check failed: ${authError.message}`);
            return false;
        }
    }
    static _buildValidationData(requestContext) {
        const { clientToProxyRequest, fullUrl, requestBody } = requestContext, method = clientToProxyRequest.method || 'GET', urlObj = new URL(fullUrl), url = fullUrl, headersMap = processRequestHeaders(clientToProxyRequest.headers), headersJson = headersToJson(headersMap);
        let queryParams = '';
        if (urlObj.search) {
            queryParams = urlObj.search.substring(1);
        }
        return {
            url: url,
            queryParams: queryParams,
            body: requestBody,
            headers: headersJson,
            method: method
        };
    }
}
module.exports = AuthorizationValidator;
