// @postman/pm-bin-linux-x64/bin/lib/reporters/util.js
const _ = require('lodash'), util = {
    extractScriptErrors: (scripts, source) => {
        return scripts
            .map((script) => { return { ...(_.omit(script.error, ['checksum', 'id'])), source }; });
    },
    buildAssertionData: (assertions) => {
        const updatedAssertions = _.map(assertions, (assert) => {
            const status = assert.skipped ? 'skipped' :
                (assert.error ? 'failed' :
                    'passed');
            return { name: assert.assertion, error: assert.error, status: status };
        });
        return updatedAssertions;
    },
    getCollectionRunMetadata: (summaryData) => {
        return {
            collectionId: summaryData.collection.id,
            collectionName: summaryData.collection.name,
            started: summaryData.run.timings.started,
            completed: summaryData.run.timings.completed,
            duration: summaryData.run.timings.completed - summaryData.run.timings.started
        };
    },
    getRunSummaryStats: (summaryData) => {
        return {
            iterations: {
                executed: summaryData.run.stats.iterations.total,
                errors: summaryData.run.stats.iterations.failed
            },
            executedRequests: {
                executed: summaryData.run.stats.requests.total,
                errors: summaryData.run.stats.requests.failed
            },
            prerequestScripts: {
                executed: summaryData.run.stats.prerequestScripts.total,
                errors: summaryData.run.stats.prerequestScripts.failed
            },
            postresponseScripts: {
                executed: summaryData.run.stats.testScripts.total,
                errors: summaryData.run.stats.testScripts.failed
            },
            tests: {
                executed: summaryData.run.stats.assertions.total,
                failed: summaryData.run.stats.assertions.failed
            },
            timeStats: {
                ...(_.pick(summaryData.run.timings, ['responseAverage', 'responseMin', 'responseMax'])),
                responseStandardDeviation: _.get(summaryData.run.timings, 'responseSd')
            }
        };
    },
    getTransformedAssertions: (assertions) => {
        return util.buildAssertionData(assertions);
    },
    extractExecutionErrors: (execution, runFailures) => {
        const errors = [];
        if (execution.requestError) {
            errors.push({
                ...(_.omit(execution.requestError, ['checksum', 'id'])),
                source: 'request'
            });
        }
        if (execution.prerequestScript && execution.prerequestScript.length) {
            errors.push(...util.extractScriptErrors(execution.prerequestScript, 'pre-request'));
        }
        if (execution.testScript && execution.testScript.length) {
            errors.push(...util.extractScriptErrors(execution.testScript, 'post-response'));
        }
        runFailures.length && runFailures.forEach((failureItem) => {
            if (failureItem.at === 'request' &&
                !failureItem.source.name &&
                failureItem.cursor.scriptId &&
                (failureItem.cursor && execution.cursor && failureItem.cursor.ref === execution.cursor.ref)) {
                errors.push({
                    ...(_.omit(failureItem.error, ['checksum', 'id'])),
                    source: 'script-level-request'
                });
            }
        });
        return errors;
    },
    transformRunSummaryData: (summaryData) => {
        const data = {
            run: {}
        };
        data.run.meta = util.getCollectionRunMetadata(summaryData);
        data.run.summary = util.getRunSummaryStats(summaryData);
        data.run.executions = summaryData.run.executions.map((execution) => {
            const filteredExecution = {
                iterationCount: _.get(execution.cursor, 'iteration'),
                requestExecuted: {
                    id: _.get(execution.item, 'id'),
                    name: _.get(execution.item, 'name'),
                    ...execution.request
                },
                ...(execution.response && {
                    response: { ...execution.response, headers: execution.response.headers }
                }),
                tests: util.getTransformedAssertions(execution.assertions),
                errors: []
            };
            filteredExecution.errors = util.extractExecutionErrors(execution, summaryData.run.failures);
            return filteredExecution;
        });
        data.run.runError = summaryData.run.error;
        return data;
    }
};
module.exports = util;
