// @postman/pm-bin-linux-x64/bin/lib/commands/cmd-util.js
require('../logger').init();
const _ = require('lodash'), fs = require('fs').promises, POSTMAN_UID_REGEX = /^([0-9A-Z]+)-([0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12})$/i, COLLECTION_UID_FROM_URL_REGEX = /https?:\/\/api\.(?:get)?postman.*\.com\/(?:collections)\/([A-Za-z0-9-]+)/, ENVIRONMENT_UID_FROM_URL_REGEX = /https?:\/\/api\.(?:get)?postman.*\.com\/(?:environments)\/([A-Za-z0-9-]+)/;
module.exports = {
    cast: {
        integer: (arg) => {
            const num = Number(arg);
            if (!_.isSafeInteger(num) || num <= 0) {
                logger.error(`Invalid integer argument: ${arg}`);
                throw new Error('The value must be a positive integer.');
            }
            return num.valueOf();
        },
        memoize: (val, memo = []) => {
            memo.push(val);
            return memo;
        },
        memoizeKeyVal: (val, memo = []) => {
            let arg, eqIndex = val.indexOf('=');
            arg = eqIndex !== -1 ? {
                key: val.slice(0, eqIndex),
                value: val.slice(eqIndex + 1)
            } : {
                key: val,
                value: undefined
            };
            memo.push(arg);
            return memo;
        },
        csvParse: (list) => {
            return _.split(list, ',');
        },
        colorOptions: (value) => {
            if (!(/^(auto|on|off)$/).test(value)) {
                throw new Error(`invalid value \`${value}\` for --color. Expected: auto|on|off`);
            }
            return value;
        },
        validateFailedSeverity: (value) => {
            if (!(/^(error|warn|info|hint)$/i).test(value)) {
                throw new Error(`invalid value \`${value}\` for  --fail-severity. Expected: error|warn|info|hint`);
            }
            return value.toUpperCase();
        },
        validateFailedSeverityForSpec: (value) => {
            if (!(/^(error|warning|info|hint)$/i).test(value)) {
                throw new Error(`invalid value \`${value}\` for  --fail-severity. Expected: error|warning|info|hint`);
            }
            return value.toUpperCase();
        }
    },
    commanderToObject: (command) => {
        return _.reduce(command, (result, value, key) => {
            const validProp = !_.startsWith(key, '_') && !_.includes(['commands', 'options', 'parent'], key);
            validProp && (result[key] = value);
            return result;
        }, {});
    },
    parseNestedOptions: (argv, optionPrefix, options) => {
        let args = [], parsed = { _generic: {} }, name, path, len, eqIndex, i;
        for (i = 0, len = argv.length; i < len; i++) {
            const arg = argv[i];
            if (!_.startsWith(arg, optionPrefix)) {
                continue;
            }
            eqIndex = arg.indexOf('=');
            if (eqIndex !== -1) {
                args.push(arg.slice(0, eqIndex), arg.slice(eqIndex + 1));
            }
            else if (argv[i + 1] && !_.startsWith(argv[i + 1], '-')) {
                args.push(arg, argv[++i]);
            }
            else {
                args.push(arg);
            }
        }
        _.forEach(options, (option) => { parsed[option] = {}; });
        for (i = 0, len = args.length; i < len; i++) {
            const arg = args[i].replace(optionPrefix, '');
            name = _.split(arg, '-', 1)[0];
            path = _.includes(options, name) ?
                [name, _.camelCase(arg.replace(name + '-', ''))].join('.') :
                ['_generic', _.camelCase(arg)].join('.');
            _.set(parsed, path, (!args[i + 1] || _.startsWith(args[i + 1], '-')) ? true : args[++i]);
        }
        return parsed;
    },
    extractApiId: async (collectionPath) => {
        try {
            const configInterface = require('../postman-entities/api/apiConfig/configInterface'), arrayOfAPIIds = await configInterface.findAPIIdForCollectionPath(collectionPath);
            if (_.isEmpty(arrayOfAPIIds.result) || !_.isArray(arrayOfAPIIds.result)) {
                return null;
            }
            if (arrayOfAPIIds.result.length > 1) {
                const collectionFile = await fs.readFile(collectionPath, 'utf8', 'r'), collectionObject = JSON.parse(collectionFile), collectionUID = _.get(collectionObject, 'info._uid', ''), fetchParentAPIId = await configInterface.findAPIIdForCollectionUID(collectionUID);
                return fetchParentAPIId.result[0];
            }
            return arrayOfAPIIds.result[0];
        }
        catch (err) {
            const errorCode = _.get(err, 'error.name'), errorMessage = _.get(err, 'error.message');
            if (errorCode === 'configFileNotFound') {
                return null;
            }
            if (errorCode === 'configNotReadable') {
                logger.error(`Unable to read API ID from ./postman/api.
                Please ensure that the file exists and is readable.`);
                return null;
            }
            if (!_.isEmpty(errorMessage)) {
                logger.error(errorMessage);
                return null;
            }
            logger.error(err);
            return null;
        }
    },
    extractCollectionUIDFromCLIArg: (collectionArg) => {
        if (!_.isString(collectionArg)) {
            return null;
        }
        let parsedValues;
        if ((parsedValues = POSTMAN_UID_REGEX.exec(collectionArg))) {
            return parsedValues[0];
        }
        if ((parsedValues = COLLECTION_UID_FROM_URL_REGEX.exec(collectionArg))) {
            return parsedValues[1];
        }
        return null;
    },
    extractEnvironmentUIDFromCLIArg: (environmentArg) => {
        if (!_.isString(environmentArg)) {
            return null;
        }
        let parsedValues;
        if ((parsedValues = POSTMAN_UID_REGEX.exec(environmentArg))) {
            return parsedValues[0];
        }
        if ((parsedValues = ENVIRONMENT_UID_FROM_URL_REGEX.exec(environmentArg))) {
            return parsedValues[1];
        }
        return null;
    },
    getCommandHierarchy: (actionCommand) => {
        let current = actionCommand, hierarchy = [];
        while (current && _.isFunction(current.name) && current.name()) {
            hierarchy.unshift(current.name());
            current = current.parent;
        }
        hierarchy.shift();
        return hierarchy;
    },
    validateOutputFormat: (value) => {
        if (!(/^(json|csv)$/i).test(value)) {
            throw new Error(`invalid value \`${value}\` for  --output. Expected: json|csv`);
        }
        return value.toUpperCase();
    }
};
