// @postman/pm-bin-linux-x64/bin/lib/commands/request/collection-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCollectionFromOptions = createCollectionFromOptions;
const fs = require('fs');
function processScriptOption(script) {
    if (!script) {
        return [];
    }
    if (script.startsWith('@')) {
        const filePath = script.substring(1);
        try {
            const fileContent = fs.readFileSync(filePath, 'utf8');
            return fileContent.split('\n');
        }
        catch (error) {
            throw new Error(`Failed to read script file '${filePath}': ${error.message}`);
        }
    }
    else {
        return [script];
    }
}
function createCollectionFromOptions(method, url, options) {
    var headers = [], collection, events = [], formDataFields = [], auth;
    if (options.header && options.header.length > 0) {
        options.header.forEach((header) => {
            const [key, value] = header.split(':').map((s) => {
                return s.trim();
            });
            if (key && value) {
                headers.push({
                    key: key,
                    value: value,
                    disabled: false,
                    description: `Header: ${key}`
                });
            }
        });
    }
    if (options.auth && Object.keys(options.auth).length > 0) {
        const authTypes = Object.keys(options.auth);
        if (authTypes.length > 0) {
            const authType = authTypes[0], authParams = options.auth[authType];
            auth = {
                type: authType,
                [authType]: authParams
            };
        }
    }
    if (options.form && options.form.length > 0) {
        options.form.forEach((field) => {
            const equalIndex = field.indexOf('=');
            if (equalIndex > 0) {
                const key = field.substring(0, equalIndex).trim();
                let value = field.substring(equalIndex + 1).trim();
                if (value.startsWith('@')) {
                    const filePath = value.substring(1);
                    formDataFields.push({
                        key: key,
                        src: filePath,
                        type: 'file',
                        enabled: true
                    });
                }
                else {
                    formDataFields.push({
                        key: key,
                        value: value,
                        type: 'text',
                        enabled: true
                    });
                }
            }
        });
        if (!headers.find((h) => {
            return h.key.toLowerCase() === 'content-type';
        })) {
            headers.push({
                key: 'Content-Type',
                value: 'multipart/form-data',
                disabled: false,
                description: 'Auto-set for form data'
            });
        }
    }
    if (options.body && options.body.startsWith('@')) {
        const filePath = options.body.substring(1);
        formDataFields.push({
            key: 'file',
            src: filePath,
            type: 'file',
            enabled: true
        });
    }
    if (options.body && !headers.find((h) => {
        return h.key.toLowerCase() === 'content-type';
    })) {
        if (options.body.trim().startsWith('{') || options.body.trim().startsWith('[')) {
            headers.push({
                key: 'Content-Type',
                value: 'application/json',
                disabled: false,
                description: 'Auto-detected JSON content type'
            });
        }
    }
    if (options.scriptPreRequest) {
        try {
            const preRequestScript = processScriptOption(options.scriptPreRequest);
            events.push({
                listen: 'prerequest',
                script: {
                    type: 'text/javascript',
                    exec: preRequestScript
                }
            });
        }
        catch (error) {
            throw new Error(`Pre-request script error: ${error.message}`);
        }
    }
    if (options.scriptPostRequest) {
        try {
            const postRequestScript = processScriptOption(options.scriptPostRequest);
            events.push({
                listen: 'test',
                script: {
                    type: 'text/javascript',
                    exec: postRequestScript
                }
            });
        }
        catch (error) {
            throw new Error(`Post-request script error: ${error.message}`);
        }
    }
    let body;
    if (formDataFields.length > 0) {
        body = {
            mode: 'formdata',
            formdata: formDataFields,
            disabled: false
        };
    }
    else {
        body = {
            mode: 'raw',
            raw: options.body || '',
            disabled: !options.body
        };
    }
    const request = {
        url: url,
        method: method,
        header: headers,
        body: body
    }, protocolProfileBehavior = {}, item = {
        name: `${method} ${url}`,
        request: request,
        event: events.length > 0 ? events : undefined
    };
    if (auth) {
        request.auth = auth;
    }
    if (options.redirectsIgnore !== undefined) {
        protocolProfileBehavior.followRedirects = !options.redirectsIgnore;
    }
    if (options.redirectsMax !== undefined && options.redirectsMax > 0) {
        protocolProfileBehavior.maxRedirects = options.redirectsMax;
    }
    if (options.redirectsFollowMethod) {
        protocolProfileBehavior.followOriginalHttpMethod = true;
    }
    if (options.redirectsRemoveReferrer) {
        protocolProfileBehavior.removeRefererHeaderOnRedirect = true;
    }
    if (Object.keys(protocolProfileBehavior).length > 0) {
        item.protocolProfileBehavior = protocolProfileBehavior;
    }
    collection = {
        info: {
            name: 'CLI Request Collection',
            description: 'Collection generated from postman request CLI command',
            schema: 'http://schema.getpostman.com/collection/json/v2.1.0/draft-04/collection.json'
        },
        item: [item]
    };
    return collection;
}
