// @postman/pm-bin-linux-x64/bin/lib/commands/request/output.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const print_1 = __importDefault(require("../../framework/print"));
const safe_1 = __importDefault(require("colors/safe"));
const cli_utils_symbols_1 = __importDefault(require("../../reporters/cli/cli-utils-symbols"));
const pretty_ms_1 = __importDefault(require("pretty-ms"));
const util_1 = require("util");
const cli_utils_1 = __importDefault(require("../../reporters/cli/cli-utils"));
const utils_1 = __importDefault(require("./utils"));
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
safe_1.default.setTheme({
    log: 'grey',
    info: 'cyan',
    warn: 'yellow',
    debug: 'blue',
    error: 'red'
});
util_1.inspect.styles.string = 'grey';
util_1.inspect.styles.special = 'white';
const printLog = (log, prefix) => {
    let levelIcon, messageColor = safe_1.default.gray;
    switch (log.level) {
        case 'error':
            levelIcon = safe_1.default.red('✗');
            messageColor = safe_1.default.red;
            break;
        case 'warn':
            levelIcon = safe_1.default.yellow('⚠');
            messageColor = safe_1.default.yellow;
            break;
        case 'info':
            levelIcon = safe_1.default.blue('ℹ');
            break;
        case 'debug':
            levelIcon = safe_1.default.gray('⋯');
            break;
        default:
            levelIcon = safe_1.default.gray('◦');
    }
    print_1.default.lf(`${prefix}   ${levelIcon} ${messageColor(log.message)}`);
};
function RequestOutput(options) {
    this.options = options || {};
    this.symbols = (0, cli_utils_symbols_1.default)(options.disableUnicode);
    this.wrapWidth = utils_1.default.getWrapWidth();
    this.wrap = cli_utils_1.default.wrapper();
}
RequestOutput.prototype.printHeaders = function (headers, prefix) {
    print_1.default.lf(`${prefix} Headers:`);
    headers.each((header) => {
        const headerKeyLength = header.key.length + 2, wrapWidth = this.wrapWidth - headerKeyLength;
        let headerValue = header.value.slice(0, wrapWidth);
        if (header.value.length > wrapWidth) {
            headerValue += utils_1.default.wrapTextForBox(header.value.slice(wrapWidth), `\n  ${safe_1.default.white(this.symbols.console.middle)}   ${' '.repeat(headerKeyLength)}`, wrapWidth);
        }
        print_1.default.lf(`${prefix}   ${safe_1.default.cyan(header.key)}: ${safe_1.default.gray(headerValue)}`);
    });
};
RequestOutput.prototype.printRequestLine = function (request) {
    if (this.options.responseOnly) {
        return;
    }
    const sanitizedUrl = request.url.toString()
        .replace(/\r?\n/g, ' ')
        .replace(/\t/g, ' ')
        .replace(/\s+/g, ' ')
        .trim(), requestDescription = `${safe_1.default.bold(request.method)} ${safe_1.default.gray(safe_1.default.bold(sanitizedUrl))}`;
    (0, print_1.default)('\n%s %s ', this.symbols.root, safe_1.default.reset(requestDescription));
};
RequestOutput.prototype.printResponseLine = function (response, request) {
    if (this.options.responseOnly) {
        return;
    }
    let reqSize = 0;
    if (request.body.mode === 'formdata' && request.body.formdata) {
        request.body.formdata.each((field) => {
            reqSize += Buffer.byteLength(field.key || '', 'utf8');
            if (typeof field.value === 'string') {
                reqSize += Buffer.byteLength(field.value, 'utf8');
            }
            else if (field.src) {
                reqSize += Buffer.byteLength(field.src, 'utf8');
            }
        });
    }
    else if (request.body[request.body.mode]) {
        reqSize = Buffer.byteLength(request.body[request.body.mode] || '', 'utf8');
    }
    const responseBody = response.body || (response.stream ? response.stream.toString() : ''), statusCode = response.code, statusText = response.status || '', resSize = Buffer.byteLength(responseBody || '', 'utf8'), reqSizeFormatted = utils_1.default.formatSize(reqSize), resSizeFormatted = utils_1.default.formatSize(resSize), durationFormatted = (0, pretty_ms_1.default)(response.responseTime), reqHeadersCount = request.headers.count(), resHeadersCount = response.headers.count(), statusLine = [
        `${statusCode} ${statusText}`,
        `${durationFormatted} ${safe_1.default.gray('time')}`,
        `${reqSizeFormatted}${safe_1.default.gray(this.symbols.up)} ${resSizeFormatted}${safe_1.default.gray(this.symbols.down)} ${safe_1.default.gray('size')}`,
        `${reqHeadersCount}${safe_1.default.gray(this.symbols.up)} ${resHeadersCount}${safe_1.default.gray(this.symbols.down)} ${safe_1.default.gray('headers')}`,
        `0 ${safe_1.default.gray('cookies')}`
    ].join(` ${safe_1.default.gray(this.symbols.star)} `);
    print_1.default.lf(`\n  ${statusLine}`);
};
RequestOutput.prototype.printRequest = function (request) {
    if (this.options.responseOnly) {
        return;
    }
    if (!request.body.raw && request.body.mode !== 'formdata' && !this.options.verbose) {
        return;
    }
    if (request.body.mode === 'formdata' && request.body.formdata) {
        if (!this.options.verbose) {
            return;
        }
        let formDataSize = 0;
        request.body.formdata.each((field) => {
            formDataSize += Buffer.byteLength(field.key || '', 'utf8');
            if (typeof field.value === 'string') {
                formDataSize += Buffer.byteLength(field.value, 'utf8');
            }
            else if (field.src) {
                formDataSize += Buffer.byteLength(field.src, 'utf8');
            }
        });
        const formDataSizeFormatted = utils_1.default.formatSize(formDataSize), headerCount = request.headers.count(), metadata = [
            'formdata',
            formDataSizeFormatted,
            this.options.verbose && headerCount > 0 ? `${headerCount} ${safe_1.default.gray('headers')}` : null
        ].filter(Boolean).join(` ${safe_1.default.gray(this.symbols.star)} `);
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.top)} ${safe_1.default.white(this.symbols.up)} ${metadata}`);
        if (this.options.verbose) {
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} URL: ${safe_1.default.gray(request.url.toString())}`);
            if (request.headers.count() > 0) {
                this.printHeaders(request.headers, `  ${safe_1.default.white(this.symbols.console.middle)}`);
            }
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} Body:`);
        }
        request.body.formdata.each((field) => {
            if (field.type === 'file') {
                print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} ${this.options.verbose ? '  ' : ''}${safe_1.default.cyan(field.key)}: ${safe_1.default.yellow('@' + field.src)}`);
            }
            else {
                print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} ${this.options.verbose ? '  ' : ''}${safe_1.default.cyan(field.key)}: ${safe_1.default.gray(field.value)}`);
            }
        });
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
        return;
    }
    if (request.body.raw || this.options.verbose) {
        const reqText = request.body.raw || '', reqSize = Buffer.byteLength(reqText, 'utf8'), reqSizeFormatted = utils_1.default.formatSize(reqSize), contentType = request.headers.count() > 0 ?
            request.headers.find((h) => { return h.key.toLowerCase() === 'content-type'; }, this)?.value : '', contentTypeInfo = utils_1.default.parseContentType(contentType), bodyMode = contentTypeInfo.isJson ? 'json' : 'raw', headerCount = request.headers.count(), metadata = [
            bodyMode,
            reqSizeFormatted,
            this.options.verbose && headerCount > 0 ? `${headerCount} ${safe_1.default.gray('headers')}` : null
        ].filter(Boolean).join(` ${safe_1.default.gray(this.symbols.star)} `);
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.top)} ${safe_1.default.white(this.symbols.up)} ${metadata}`);
        if (this.options.verbose) {
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} URL: ${safe_1.default.gray(request.url.toString())}`);
            if (request.headers.count() > 0) {
                this.printHeaders(request.headers, `  ${safe_1.default.white(this.symbols.console.middle)}`);
            }
            if (reqText) {
                print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} Body:`);
            }
        }
        if (reqText) {
            const wrappedText = utils_1.default.wrapTextForBox(reqText, `  ${safe_1.default.white(this.symbols.console.middle)} ${this.options.verbose ? '  ' : ''}`, this.wrapWidth);
            (0, print_1.default)(safe_1.default.gray(utils_1.default.cleanWrappedText(wrappedText) + '\n'));
        }
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
    }
};
RequestOutput.prototype.printResponse = function (response) {
    if (this.options.responseOnly) {
        const body = response.body || (response.stream ? response.stream.toString() : '');
        (0, print_1.default)(body, '\n');
        return;
    }
    const body = response.body || (response.stream ? response.stream.toString() : ''), contentTypeHeader = response.headers.find((header) => {
        return header.key.toLowerCase() === 'content-type';
    }, this), contentTypeValue = contentTypeHeader?.value || '', resSize = Buffer.byteLength(body, 'utf8'), resSizeFormatted = utils_1.default.formatSize(resSize), contentTypeInfo = utils_1.default.parseContentType(contentTypeValue), headerCount = response.headers.count(), resSummary = [
        contentTypeInfo.contentType,
        contentTypeInfo.mimeFormat,
        contentTypeInfo.isJson ? 'json' : 'text',
        contentTypeInfo.charset,
        resSizeFormatted,
        this.options.verbose && headerCount > 0 ? `${headerCount} ${safe_1.default.gray('headers')}` : null
    ].filter(Boolean).join(` ${safe_1.default.gray(this.symbols.star)} `);
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.top)} ${safe_1.default.white(this.symbols.down)} ${resSummary}`);
    if (this.options.verbose && response.headers.count() > 0) {
        this.printHeaders(response.headers, `  ${safe_1.default.white(this.symbols.console.middle)}`);
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} Body:`);
    }
    let formattedBody, wrappedText;
    if (!this.options.raw && contentTypeInfo.isJson) {
        try {
            const parsed = JSON.parse(body);
            const jsonColorizer = require('json-colorizer');
            formattedBody = jsonColorizer(JSON.stringify(parsed, null, 2));
        }
        catch {
            formattedBody = body;
        }
    }
    else {
        formattedBody = body;
    }
    wrappedText = utils_1.default.wrapTextForBox(formattedBody, `  ${safe_1.default.white(this.symbols.console.middle)} ${this.options.verbose ? '  ' : ''}`, this.wrapWidth);
    (0, print_1.default)(safe_1.default.gray(utils_1.default.cleanWrappedText(wrappedText) + '\n'));
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
};
RequestOutput.prototype.printError = function (error) {
    const message = error.message || error;
    print_1.default.lf(`${safe_1.default.red(this.symbols.error)} Error: ${message}`);
};
RequestOutput.prototype.printRetrying = function (retryCount, maxRetries) {
    print_1.default.lf(`\n  ⟳ Retrying request (${retryCount}/${maxRetries})...`);
};
RequestOutput.prototype.printRetrySuccess = function (retryCount) {
    print_1.default.lf(`✓ Request succeeded on retry ${retryCount}`);
};
RequestOutput.prototype.printDebugSummary = function (attempts, totalTime, retryDelay) {
    const totalRedirects = attempts.reduce((sum, att) => {
        return sum + (att.redirects?.length || 0);
    }, 0), metadata = [
        `${attempts.length} attempts`,
        `${totalTime}ms`,
        totalRedirects > 0 ? `${totalRedirects} redirects` : null
    ].filter(Boolean).join(' ★ ');
    print_1.default.lf(`\n  ${safe_1.default.white(this.symbols.console.top)} 🔍 debug ${safe_1.default.gray(this.symbols.star)} ${safe_1.default.gray(metadata)}`);
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} Summary:`);
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}   Total attempts: ${attempts.length}`);
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}   Total time: ${totalTime}ms`);
    if (attempts.length > 1) {
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}   Retry delay: ${retryDelay}ms`);
    }
    if (totalRedirects > 0) {
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}   Total redirects: ${totalRedirects}`);
    }
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}`);
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} Attempt Details:`);
    attempts.forEach((attempt) => {
        const status = attempt.success ? safe_1.default.green('✓') : safe_1.default.red('✗');
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}   ${status} Attempt ${attempt.attemptNumber}: ${attempt.statusCode || 'ERROR'} (${attempt.responseTime || 0}ms)`);
        if (attempt.redirects && attempt.redirects.length > 0) {
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}      Redirects: ${attempt.redirects.length}`);
            attempt.redirects.forEach((redirect, index) => {
                print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}        ${index + 1}. ${redirect.statusCode} → ${redirect.url} (${redirect.responseTime}ms)`);
            });
        }
    });
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
};
RequestOutput.prototype.printCompletion = function (duration) {
    if (this.options.responseOnly) {
        return;
    }
    print_1.default.lf(`\n${this.symbols.root} Completed in ${(0, pretty_ms_1.default)(duration)}`);
};
RequestOutput.prototype.saveResponse = function (response, request, filePath, testResults, scriptLogs) {
    try {
        const dir = path.dirname(filePath), responseJSON = response.toJSON(), requestJSON = request.toJSON(), responseData = {
            request: {
                ...requestJSON,
                url: request.url.toString(),
                body: request.body[request.body.mode]
            },
            response: {
                ...responseJSON,
                body: response.body || (response.stream ? response.stream.toString() : ''),
                stream: undefined
            },
            timestamp: new Date().toISOString()
        }, results = testResults || [], passedTests = results.filter((t) => { return t.passed; }), failedTests = results.filter((t) => { return !t.passed; });
        responseData.tests = {
            total: results.length,
            passed: passedTests.length,
            failed: failedTests.length,
            results: results.map((test) => {
                return {
                    name: test.name,
                    passed: test.passed,
                    consoleLogs: test.consoleLogs || [],
                    error: test.error ? {
                        message: test.error.message,
                        name: test.error.name
                    } : undefined
                };
            })
        };
        responseData.scriptLogs = scriptLogs || [];
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }
        fs.writeFileSync(filePath, JSON.stringify(responseData, null, 2), 'utf8');
        if (!this.options.responseOnly) {
            print_1.default.lf(`\n  ${safe_1.default.white(this.symbols.console.top)} ${safe_1.default.green(this.symbols.ok)} saved ${safe_1.default.gray(this.symbols.star)} ${safe_1.default.gray('full response')}`);
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} ${safe_1.default.gray(filePath)}`);
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
        }
    }
    catch (error) {
        print_1.default.lf(`\n${safe_1.default.red(this.symbols.error)} Failed to save response: ${error.message}`);
    }
};
RequestOutput.prototype.printTestResults = function (testResults, scriptLogs) {
    if (this.options.responseOnly) {
        return;
    }
    const hasTests = testResults && testResults.length > 0, hasScriptLogs = scriptLogs && scriptLogs.length > 0;
    if (!hasTests && !hasScriptLogs) {
        return;
    }
    print_1.default.lf('');
    if (hasScriptLogs) {
        const logIcon = safe_1.default.cyan('◈');
        print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.top)} ${logIcon} Script Logs ${safe_1.default.gray(this.symbols.star)} ${safe_1.default.gray(`${scriptLogs.length} messages`)}`);
        scriptLogs.forEach((log) => {
            printLog(log, `  ${safe_1.default.white(this.symbols.console.middle)}`);
        });
        if (!hasTests) {
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
            return;
        }
    }
    if (!hasTests) {
        return;
    }
    const failedTests = testResults.filter((t) => { return !t.passed; }), passedTests = testResults.filter((t) => { return t.passed; }), metadata = [
        passedTests.length > 0 ? `${passedTests.length} ${safe_1.default.green('passed')}` : null,
        failedTests.length > 0 ? `${failedTests.length} ${safe_1.default.red('failed')}` : null
    ].filter(Boolean).join(` ${safe_1.default.gray(this.symbols.star)} `), icon = failedTests.length > 0 ? safe_1.default.red(this.symbols.error) : safe_1.default.green(this.symbols.ok), topSymbol = hasScriptLogs ?
        `  ${safe_1.default.white(this.symbols.console.middle)}` :
        `  ${safe_1.default.white(this.symbols.console.top)}`;
    print_1.default.lf(`${topSymbol} ${icon} tests ${safe_1.default.gray(this.symbols.star)} ${metadata}`);
    testResults.forEach((test) => {
        if (test.passed) {
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} ${safe_1.default.green(this.symbols.ok)} ${safe_1.default.gray(test.name)}`);
        }
        else {
            print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)} ${safe_1.default.red(this.symbols.error)} ${test.name}`);
            if (test.error && test.error.message) {
                const errorMsg = test.error.message;
                print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.middle)}   ${safe_1.default.gray(errorMsg)}`);
            }
        }
        if (test.consoleLogs && test.consoleLogs.length > 0) {
            test.consoleLogs.forEach((log) => {
                printLog(log, `  ${safe_1.default.white(this.symbols.console.middle)}`);
            });
        }
    });
    print_1.default.lf(`  ${safe_1.default.white(this.symbols.console.bottom)}`);
    if (failedTests.length > 0) {
        process.exitCode = failedTests.length;
    }
};
exports.default = RequestOutput;
