// @postman/pm-bin-linux-x64/bin/lib/commands/request/request-runner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runRequest = runRequest;
const output_1 = __importDefault(require("./output"));
const print_1 = __importDefault(require("../../framework/print"));
const run_1 = __importDefault(require("../../run"));
const util_1 = require("../../postman-entities/util");
const cmd_util_1 = __importDefault(require("../cmd-util"));
const collection_builder_1 = require("./collection-builder");
const OVERRIDE_SPINNER_WITH_SPACE = true;
function extractAuthTypes(argv) {
    const authTypes = new Set();
    for (const arg of argv) {
        if (arg.startsWith('--auth-')) {
            const parts = arg.substring(7).split('-');
            if (parts.length > 0) {
                authTypes.add(parts[0]);
            }
        }
    }
    return Array.from(authTypes);
}
function handleCollectionRunComplete(err, context) {
    if (err) {
        context.requestError = err;
    }
    print_1.default.unwait(OVERRIDE_SPINNER_WITH_SPACE);
}
function handleBeforeRequest(error, data, context, output, showOutput) {
    const { request } = data;
    context.lastRequest = request;
    if (error) {
        context.requestError = error;
        return;
    }
    if (showOutput) {
        output.printRequestLine(request);
        if (!output.options.responseOnly) {
            print_1.default.wait();
        }
    }
}
function handleRequest(error, args, context, output, showOutput) {
    const { request, response, history } = args;
    context.lastResponse = response;
    context.lastRequest = request;
    context.responseTime = response?.responseTime || 0;
    if (history && history.execution && history.execution.data) {
        const historyData = history.execution.data;
        if (historyData.length > 1) {
            context.redirects = historyData.slice(0, -1).map((redirect) => {
                return {
                    url: redirect.request?.href || redirect.request?.url?.toString() || 'unknown',
                    statusCode: redirect.response?.statusCode || redirect.response?.code || 0,
                    responseTime: redirect.timings?.total || 0
                };
            });
        }
    }
    if (error) {
        context.requestError = error;
    }
    if (showOutput && response) {
        print_1.default.unwait(OVERRIDE_SPINNER_WITH_SPACE);
        output.printResponseLine(response, request);
        output.printRequest(request);
        output.printResponse(response);
    }
}
function handleConsole(error, data, testResults, scriptLogs) {
    if (error) {
        return;
    }
    const { level, messages } = data, formattedMessages = messages.map((msg) => {
        if (typeof msg === 'object' && msg !== null) {
            try {
                return JSON.stringify(msg, null, 2);
            }
            catch {
                return String(msg);
            }
        }
        return String(msg);
    }), message = formattedMessages.join(' '), logEntry = {
        level: level || 'log',
        message: message,
        timestamp: new Date().toISOString()
    };
    if (testResults.length > 0) {
        const currentTest = testResults[testResults.length - 1];
        if (!currentTest.consoleLogs) {
            currentTest.consoleLogs = [];
        }
        currentTest.consoleLogs.push(logEntry);
    }
    else {
        scriptLogs.push(logEntry);
    }
}
function handleAssertion(error, assertion, testResults) {
    testResults.push({
        name: assertion.assertion,
        passed: !error,
        error: error,
        consoleLogs: []
    });
}
function handleDoneWithTests(err, context, output, showOutput, testResults, scriptLogs, enrichedOptions, resolve) {
    if (err) {
        return resolve({
            success: false,
            response: context.lastResponse,
            error: err,
            request: context.lastRequest,
            responseTime: context.responseTime,
            redirects: context.redirects,
            testResults: testResults,
            scriptLogs: scriptLogs
        });
    }
    if (showOutput) {
        output.printTestResults(testResults, scriptLogs);
    }
    if (enrichedOptions.output && context.lastResponse && context.lastRequest) {
        output.saveResponse(context.lastResponse, context.lastRequest, enrichedOptions.output, testResults, scriptLogs);
    }
    const isSuccess = !context.requestError &&
        context.lastResponse &&
        context.lastResponse.code >= 200 &&
        context.lastResponse.code < 400;
    resolve({
        success: isSuccess,
        response: context.lastResponse,
        error: context.requestError,
        request: context.lastRequest,
        responseTime: context.responseTime,
        redirects: context.redirects,
        testResults: testResults,
        scriptLogs: scriptLogs
    });
}
function handleDone(err, context, output, enrichedOptions, resolve) {
    if (err) {
        return resolve({
            success: false,
            response: context.lastResponse,
            error: err,
            request: context.lastRequest,
            responseTime: context.responseTime,
            redirects: context.redirects
        });
    }
    if (enrichedOptions.output && context.lastResponse && context.lastRequest) {
        output.saveResponse(context.lastResponse, context.lastRequest, enrichedOptions.output);
    }
    const isSuccess = !context.requestError &&
        context.lastResponse &&
        context.lastResponse.code >= 200 &&
        context.lastResponse.code < 400;
    resolve({
        success: isSuccess,
        response: context.lastResponse,
        error: context.requestError,
        request: context.lastRequest,
        responseTime: context.responseTime,
        redirects: context.redirects
    });
}
async function runRequest(methodOrUrl, urlOrNothing, options, requestCommand) {
    if (options.quiet) {
        options.responseOnly = true;
    }
    const output = new output_1.default(options), argv = requestCommand.parent._originalArgs || process.argv, authTypes = extractAuthTypes(argv), parsedAuthOptions = cmd_util_1.default.parseNestedOptions(argv, '--auth-', authTypes), authOptions = Object.keys(parsedAuthOptions)
        .filter((key) => { return key !== '_generic'; })
        .reduce((acc, key) => {
        acc[key] = parsedAuthOptions[key];
        return acc;
    }, {}), httpMethods = ['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'HEAD', 'OPTIONS'];
    let method, url;
    if (urlOrNothing) {
        method = methodOrUrl.toUpperCase();
        url = urlOrNothing;
    }
    else if (methodOrUrl) {
        if (httpMethods.includes(methodOrUrl.toUpperCase())) {
            output.printError('URL is required when method is specified');
            process.exit(-1);
        }
        else {
            method = 'GET';
            url = methodOrUrl;
        }
    }
    else {
        output.printError('URL is required for request.\n' +
            'please run `postman request --help` for more information.');
        process.exit(-1);
    }
    const maxRetries = options.retry || 0, retryDelay = options.retryDelay || 1000, enrichedOptions = await new Promise((resolve, reject) => {
        (0, util_1.addApiKeyToOptions)(options, {}, (err, options) => {
            if (err) {
                output.printError(err.message || err);
                reject(err);
            }
            options.auth = authOptions;
            resolve(options);
        });
    }), collection = (0, collection_builder_1.createCollectionFromOptions)(method, url, enrichedOptions), environmentUID = cmd_util_1.default.extractEnvironmentUIDFromCLIArg(enrichedOptions.environment), runOptions = {
        ...enrichedOptions,
        collection: collection,
        reporters: [],
        silent: true,
        skipRunUpload: true,
        timeoutRequest: enrichedOptions.timeout,
        environmentUID: environmentUID
    }, executeRequest = (showOutput) => {
        return new Promise((resolve) => {
            let isRequestFinished = false;
            const context = {
                requestError: null,
                lastResponse: null,
                lastRequest: null,
                responseTime: 0,
                redirects: []
            }, testResults = [], scriptLogs = [], emitter = (0, run_1.default)(runOptions, (err) => {
                handleCollectionRunComplete(err, context);
            });
            emitter.on('beforeRequest', (error, data) => {
                handleBeforeRequest(error, data, context, output, showOutput && !isRequestFinished);
            });
            emitter.on('request', (error, args) => {
                handleRequest(error, args, context, output, showOutput);
            });
            emitter.on('console', (error, data) => {
                handleConsole(error, data, testResults, scriptLogs);
            });
            if (enrichedOptions.scriptPostRequest && enrichedOptions.scriptPostRequest.length > 0) {
                emitter.on('assertion', (error, assertion) => {
                    handleAssertion(error, assertion, testResults);
                });
                emitter.on('done', (err) => {
                    isRequestFinished = true;
                    handleDoneWithTests(err, context, output, showOutput, testResults, scriptLogs, enrichedOptions, resolve);
                });
            }
            else {
                emitter.on('done', (err) => {
                    isRequestFinished = true;
                    if (showOutput) {
                        output.printTestResults(testResults, scriptLogs);
                    }
                    handleDone(err, context, output, enrichedOptions, resolve);
                });
            }
        });
    }, isDebugMode = enrichedOptions.debug, attempts = [];
    let result = await executeRequest(isDebugMode || maxRetries === 0), totalTime = result.responseTime || 0;
    attempts.push({
        attemptNumber: 1,
        success: result.success,
        responseTime: result.responseTime,
        statusCode: result.response?.code,
        redirects: result.redirects || []
    });
    if (!result.success && result.error && !maxRetries) {
        output.printError(`Request failed: ${result.error.message || result.error}`);
    }
    if (!result.success && maxRetries > 0) {
        let retryCount = 0, currentResult = result;
        while (retryCount < maxRetries && !currentResult.success) {
            retryCount++;
            if (isDebugMode) {
                output.printRetrying(retryCount, maxRetries);
            }
            await new Promise((resolve) => {
                setTimeout(resolve, retryDelay);
            });
            currentResult = await executeRequest(isDebugMode);
            totalTime += currentResult.responseTime || 0;
            attempts.push({
                attemptNumber: retryCount + 1,
                success: currentResult.success,
                responseTime: currentResult.responseTime,
                statusCode: currentResult.response?.code,
                redirects: currentResult.redirects || []
            });
            if (currentResult.success) {
                if (isDebugMode) {
                    output.printRetrySuccess(retryCount);
                }
                result = currentResult;
                break;
            }
        }
        if (!isDebugMode && result.request && result.response) {
            output.printRequestLine(result.request);
            output.printResponseLine(result.response, result.request);
            output.printRequest(result.request);
            output.printResponse(result.response);
            output.printTestResults(result.testResults, result.scriptLogs);
            if (enrichedOptions.output) {
                output.saveResponse(result.response, result.request, enrichedOptions.output, result.testResults, result.scriptLogs);
            }
        }
    }
    if (isDebugMode && (attempts.length > 1 || (attempts[0] && attempts[0].redirects.length > 0))) {
        output.printDebugSummary(attempts, totalTime, retryDelay);
    }
}
