// @postman/pm-bin-linux-x64/bin/lib/commands/request/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const wrap = require('word-wrap'), isCI = require('is-ci'), cliUtils = require('../../reporters/cli/cli-utils'), { filesize } = require('filesize');
const utils = {
    getDimension: function () {
        var tty, width, height;
        try {
            tty = require('tty');
        }
        catch {
            tty = null;
        }
        if (tty && tty.isatty(1) && tty.isatty(2)) {
            if (process.stdout.getWindowSize) {
                width = process.stdout.getWindowSize()[0];
                height = process.stdout.getWindowSize()[1];
            }
            else if (tty.getWindowSize) {
                width = tty.getWindowSize()[1];
                height = tty.getWindowSize()[0];
            }
            else if (process.stdout.columns && process.stdout.rows) {
                height = process.stdout.rows;
                width = process.stdout.columns;
            }
        }
        return {
            exists: !(isCI || cliUtils.noTTY()),
            width: width,
            height: height
        };
    },
    getWrapWidth: function () {
        var dimension = this.getDimension();
        return ((dimension.exists && (dimension.width > 20)) ? dimension.width : 60) - 6;
    },
    wrapTextForBox: function (text, indent, width) {
        return wrap(text, {
            indent: indent,
            width: width || this.getWrapWidth(),
            cut: true
        });
    },
    parseContentType: function (contentType) {
        contentType = contentType || '';
        var mimeType = contentType.split(';')[0] || '', charset = contentType.includes('charset=') ?
            contentType.split('charset=')[1].trim() : 'utf8';
        return {
            contentType: contentType,
            mimeType: mimeType,
            mimeFormat: mimeType.split('/')[1] || 'text',
            charset: charset,
            isJson: contentType.includes('application/json')
        };
    },
    formatSize: function (bytes) {
        return bytes > 1024 ? filesize(bytes) : `${bytes}B`;
    },
    cleanWrappedText: function (text) {
        return text.replace(/\n\s*\n/g, '\n');
    },
    collectValues: function (value, previous = []) {
        return previous.concat([value]);
    },
    formatOptionDescription: function (lines, indent = 35) {
        const padding = ' '.repeat(indent);
        return lines.map((line, index) => {
            if (index === 0) {
                return line;
            }
            return padding + line;
        }).join('\n');
    }
};
exports.default = utils;
