// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/modules/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const colors = require("colors/safe");
const util = require("../../../util");
const http_1 = require("./http");
const E = '';
function extractOperationName(query, operationName) {
    if (operationName) {
        return operationName;
    }
    if (!query) {
        return '';
    }
    const match = query.match(/^\s*(query|mutation|subscription)\s+(\w+)/m);
    return match ? match[2] : '';
}
const register = function ({ emitter, state, output, options }) {
    const symbols = output.symbols;
    const pendingData = state.getStore('graphql');
    emitter.on('graphql:send-request', function (event) {
        const query = event.payload.query || '', operationType = event.payload.operationType || 'query', operationName = extractOperationName(query, event.payload.operationName), url = event.payload.url || '', correlationId = event.context.correlationId, label = operationType.toUpperCase();
        if (correlationId) {
            state.claimCorrelation(correlationId, 'graphql');
            pendingData.set(correlationId, {
                query,
                operationType,
                operationName,
                variables: (typeof event.payload.variables === 'object' ?
                    event.payload.variables : {}),
                url,
                headers: event.payload.headers || [],
                sendTimestamp: Date.now(),
                streamMessageCount: 0
            });
        }
        const requestLabel = operationName ? (label + ' ' + operationName) : label;
        output.printRequestLine(requestLabel, url);
    });
    emitter.on('http:received-response-end', function (event) {
        const correlationId = event.context.correlationId;
        if (!correlationId || state.getOwner(correlationId) !== 'graphql') {
            return;
        }
        const gqlData = pendingData.get(correlationId);
        if (gqlData) {
            gqlData.httpResponse = event._assembledResponse;
        }
    });
    emitter.on('http:error', function (event) {
        const correlationId = event.context.correlationId;
        if (!correlationId || state.getOwner(correlationId) !== 'graphql') {
            return;
        }
        const error = event.payload.error || {}, errorMessage = error.message || error.code || String(error), gqlData = pendingData.get(correlationId);
        if (gqlData) {
            gqlData.httpError = {
                message: errorMessage,
                name: error.name || 'Error'
            };
        }
    });
    emitter.on('graphql:received-response', function (event) {
        const correlationId = event.context.correlationId, gqlData = correlationId ? pendingData.get(correlationId) : undefined, httpRes = gqlData && gqlData.httpResponse, httpError = gqlData && gqlData.httpError, gqlPayload = event.payload, gqlErrors = gqlPayload?.errors;
        if (correlationId) {
            state.releaseCorrelation(correlationId);
            pendingData.delete(correlationId);
        }
        if (!httpRes) {
            output.printError(httpError ? httpError.message : 'Could not read response data.');
            return;
        }
        const totalSize = (httpRes.sizes.body || 0) + (httpRes.sizes.headers || 0), totalTime = httpRes.timings.total || 0;
        if (!options.verbose) {
            output.printStatusBracket(httpRes.statusCode, (0, http_1.getStatusText)(httpRes.statusCode, httpRes.statusText), totalSize, totalTime);
            output.printGraphqlErrors(gqlErrors);
            return;
        }
        const reqHeadersLen = gqlData ? gqlData.headers.length : 0, resHeadersLen = httpRes.headers ? httpRes.headers.length : 0, queryText = gqlData ? gqlData.query : '', queryLen = queryText ? Buffer.byteLength(queryText) : 0;
        let variablesText = E, resBodyText = E, resBodyLen = 0;
        if (gqlData && gqlData.variables && Object.keys(gqlData.variables).length) {
            try {
                variablesText = JSON.stringify(gqlData.variables, null, 2);
            }
            catch {
                variablesText = String(gqlData.variables);
            }
        }
        if (httpRes.body) {
            resBodyText = Buffer.isBuffer(httpRes.body) ? httpRes.body.toString() : String(httpRes.body);
            resBodyLen = httpRes.sizes.body || Buffer.byteLength(resBodyText);
        }
        const resSummary = output.parseResponseContentType(httpRes.headers);
        const cookieCount = output.countResponseCookies(httpRes.headers);
        output.printSummaryLine([
            httpRes.statusCode + ' ' + (0, http_1.getStatusText)(httpRes.statusCode, httpRes.statusText),
            util.prettyms(totalTime) + ' ' + colors.gray('time'),
            (queryLen ? util.filesize(queryLen) + colors.gray(symbols.up) + ' ' : '') +
                util.filesize(totalSize) + colors.gray(symbols.down) + ' ' + colors.gray('size'),
            reqHeadersLen + colors.gray(symbols.up) + ' ' +
                resHeadersLen + colors.gray(symbols.down) + ' ' + colors.gray('headers'),
            cookieCount + ' ' + colors.gray('cookies')
        ]);
        if (queryLen) {
            output.printBodyBox(queryText, queryLen, symbols.up, 'graphql');
            output.printSpacer();
        }
        if (variablesText) {
            output.printBodyBox(variablesText, Buffer.byteLength(variablesText), symbols.up, 'variables');
            output.printSpacer();
        }
        if (resBodyLen) {
            output.printBodyBox(resBodyText, resBodyLen, symbols.down, resSummary);
        }
        else if (resSummary) {
            output.printContentLabel(symbols.down, resSummary);
        }
        if (gqlErrors && gqlErrors.length) {
            output.printSpacer();
            output.printGraphqlErrors(gqlErrors);
        }
        output.printTimingTable(httpRes.timings);
    });
    emitter.on('graphql:received-response-stream-start', function (event) {
        const correlationId = event.context.correlationId, gqlData = correlationId ? pendingData.get(correlationId) : undefined, elapsed = gqlData && gqlData.sendTimestamp ? Date.now() - gqlData.sendTimestamp : 0;
        if (!options.verbose) {
            output.printStatusMessage('[connected, ' + util.prettyms(elapsed) + ']');
            return;
        }
        output.printSummaryLine([
            'connected',
            util.prettyms(elapsed) + ' ' + colors.gray('time')
        ]);
        if (gqlData && gqlData.query) {
            output.printBodyBox(gqlData.query, Buffer.byteLength(gqlData.query), symbols.up, 'graphql');
            output.printSpacer();
        }
        if (gqlData && gqlData.variables && Object.keys(gqlData.variables).length) {
            let variablesText;
            try {
                variablesText = JSON.stringify(gqlData.variables, null, 2);
            }
            catch {
                variablesText = String(gqlData.variables);
            }
            output.printBodyBox(variablesText, Buffer.byteLength(variablesText), symbols.up, 'variables');
            output.printSpacer();
        }
    });
    emitter.on('graphql:received-response-stream-data', function (event) {
        const correlationId = event.context.correlationId, gqlData = correlationId ? pendingData.get(correlationId) : undefined, gqlPayload = event.payload, gqlErrors = gqlPayload?.errors;
        let dataText = E, dataLen = 0, messageNum;
        if (gqlData) {
            gqlData.streamMessageCount++;
            messageNum = gqlData.streamMessageCount;
        }
        else {
            messageNum = 1;
        }
        if (gqlPayload?.data) {
            try {
                dataText = JSON.stringify(gqlPayload.data);
            }
            catch {
                dataText = String(gqlPayload.data);
            }
            dataLen = Buffer.byteLength(dataText);
        }
        if (!options.verbose) {
            output.printGraphqlErrors(gqlErrors);
            return;
        }
        if (dataLen) {
            let prettyData;
            try {
                prettyData = JSON.stringify(gqlPayload?.data, null, 2);
            }
            catch {
                prettyData = dataText;
            }
            output.printBodyBox(prettyData, Buffer.byteLength(prettyData), symbols.down, 'message ' + messageNum, dataLen);
            output.printSpacer();
        }
        output.printGraphqlErrors(gqlErrors);
    });
};
exports.register = register;
