// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/modules/grpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const colors = require("colors/safe");
const util = require("../../../util");
const print = require("../../../framework/print");
const SPC = ' ', E = '', GRPC_STATUS_CODES = {
    0: 'OK', 1: 'CANCELLED', 2: 'UNKNOWN', 3: 'INVALID_ARGUMENT',
    4: 'DEADLINE_EXCEEDED', 5: 'NOT_FOUND', 6: 'ALREADY_EXISTS',
    7: 'PERMISSION_DENIED', 8: 'RESOURCE_EXHAUSTED', 9: 'FAILED_PRECONDITION',
    10: 'ABORTED', 11: 'OUT_OF_RANGE', 12: 'UNIMPLEMENTED',
    13: 'INTERNAL', 14: 'UNAVAILABLE', 15: 'DATA_LOSS', 16: 'UNAUTHENTICATED'
};
const register = function ({ emitter, state, output, options }) {
    const symbols = output.symbols;
    const pendingData = state.getStore('grpc');
    let activeCorrelationId = null;
    const deferrer = {
        deferConsole(color, message) {
            if (!activeCorrelationId) {
                return;
            }
            const data = pendingData.get(activeCorrelationId);
            if (data) {
                data.deferredConsole.push({ color, message });
            }
        },
        flushDeferredConsole() {
            if (!activeCorrelationId) {
                return;
            }
            const data = pendingData.get(activeCorrelationId);
            flushDeferred(data);
        }
    };
    function flushDeferred(grpcData) {
        if (grpcData && grpcData.deferredConsole.length) {
            grpcData.deferredConsole.forEach(function (deferred) {
                output.printConsoleBox(deferred.color, deferred.message);
            });
            grpcData.deferredConsole = [];
        }
    }
    emitter.on('grpc:invoke', function (event) {
        const methodPath = event.payload.methodPath || '', url = event.payload.url || '', correlationId = event.context.correlationId;
        if (correlationId) {
            state.claimCorrelation(correlationId, 'grpc');
            pendingData.set(correlationId, {
                methodPath,
                url,
                message: event.payload.message,
                metadata: event.payload.metadata || [],
                totalResponseSize: 0,
                streamMessageCount: 0,
                isRequestStreamed: false,
                isResponseStreamed: false,
                deferredConsole: []
            });
            activeCorrelationId = correlationId;
            state.setConsoleDeferrer(deferrer);
        }
        output.printRequestLine(methodPath, url);
    });
    emitter.on('grpc:sent-request-header', function (event) {
        const correlationId = event.context.correlationId, grpcData = correlationId ? pendingData.get(correlationId) : undefined;
        if (grpcData) {
            grpcData.isRequestStreamed = event.payload.isRequestStreamed;
            grpcData.isResponseStreamed = event.payload.isResponseStreamed;
        }
    });
    emitter.on('grpc:received-response-data', function (event) {
        const correlationId = event.context.correlationId, grpcData = correlationId ? pendingData.get(correlationId) : undefined, payload = event.payload;
        let dataText = E, dataLen = 0, messageNum;
        if (grpcData && payload.data) {
            try {
                dataText = JSON.stringify(payload.data);
            }
            catch {
                dataText = String(payload.data);
            }
            dataLen = Buffer.byteLength(dataText);
            grpcData.totalResponseSize += dataLen;
            grpcData.streamMessageCount++;
            grpcData.lastResponseData = payload.data;
            messageNum = grpcData.streamMessageCount;
        }
        else {
            messageNum = 1;
        }
        if (!options.verbose || !(grpcData && grpcData.isResponseStreamed)) {
            return;
        }
        if (dataLen) {
            let prettyData;
            try {
                prettyData = JSON.stringify(payload.data, null, 2);
            }
            catch {
                prettyData = dataText;
            }
            output.printBodyBox(prettyData, Buffer.byteLength(prettyData), symbols.down, 'message ' + messageNum, dataLen);
            print.lf(SPC);
        }
    });
    emitter.on('grpc:status', function (event) {
        const payload = event.payload, correlationId = event.context.correlationId, grpcData = correlationId ? pendingData.get(correlationId) : undefined, statusCode = payload.statusCode, statusMessage = payload.statusMessage || GRPC_STATUS_CODES[statusCode] || '', totalTime = payload.timings.total || 0, totalSize = grpcData ? grpcData.totalResponseSize : 0;
        activeCorrelationId = null;
        state.clearConsoleDeferrer();
        if (!options.verbose) {
            output.printStatusBracket(statusCode, statusMessage, totalSize, totalTime);
            flushDeferred(grpcData);
            if (correlationId) {
                pendingData.delete(correlationId);
            }
            return;
        }
        const reqMetadataLen = grpcData ? grpcData.metadata.length : 0, resMetadataLen = payload.metadata ? payload.metadata.length : 0;
        let messageContent = E, messageLen = 0, lastResponseText = E, lastResponseLen = 0;
        if (grpcData && grpcData.message && grpcData.message.content) {
            messageContent = grpcData.message.content;
            messageLen = Buffer.byteLength(messageContent);
        }
        if (grpcData && grpcData.lastResponseData) {
            try {
                lastResponseText = JSON.stringify(grpcData.lastResponseData, null, 2);
            }
            catch {
                lastResponseText = String(grpcData.lastResponseData);
            }
            lastResponseLen = Buffer.byteLength(lastResponseText);
        }
        output.printSummaryLine([
            statusCode + ' ' + statusMessage,
            util.prettyms(totalTime) + ' ' + colors.gray('time'),
            (messageLen ? util.filesize(messageLen) + colors.gray(symbols.up) + ' ' : '') +
                util.filesize(totalSize) + colors.gray(symbols.down) + ' ' + colors.gray('size'),
            reqMetadataLen + colors.gray(symbols.up) + ' ' +
                resMetadataLen + colors.gray(symbols.down) + ' ' + colors.gray('metadata')
        ]);
        if (messageLen) {
            output.printBodyBox(messageContent, messageLen, symbols.up, 'message');
            print.lf(SPC);
        }
        if (lastResponseLen && grpcData && !grpcData.isResponseStreamed) {
            output.printBodyBox(lastResponseText, lastResponseLen, symbols.down, 'response');
        }
        flushDeferred(grpcData);
        if (correlationId) {
            pendingData.delete(correlationId);
        }
    });
    emitter.on('grpc:error', function (event) {
        const error = event.payload.error || {}, errorMessage = error.message || error.code || String(error), correlationId = event.context.correlationId;
        activeCorrelationId = null;
        state.clearConsoleDeferrer();
        if (correlationId && state.hasScriptError(correlationId)) {
            return;
        }
        if (correlationId && state.getOwner(correlationId) !== 'grpc') {
            const itemRef = event.context.item, runContext = emitter.runContext, item = runContext && runContext.resolveItemRef(itemRef), methodPath = item && (item?.payload?.methodPath || ''), url = item && (item?.payload?.url || '');
            if (methodPath || url) {
                print('  %s %s ', colors.gray.bold(methodPath), colors.gray.bold(url));
            }
        }
        print.lf(colors.red('[errored]'));
        print.lf(colors.red('     %s'), errorMessage);
        if (correlationId) {
            const grpcData = pendingData.get(correlationId);
            flushDeferred(grpcData);
            pendingData.delete(correlationId);
        }
    });
    emitter.on('grpc:transient-error', function (event) {
        const error = event.payload.error || {}, errorMessage = error.message || String(error);
        output.printWarning('gRPC Warning: ' + errorMessage);
    });
    emitter.on('collection:item-end', function (event) {
        const correlationId = event.context.correlationId;
        if (correlationId && state.getOwner(correlationId) === 'grpc') {
            flushDeferred(pendingData.get(correlationId));
            if (activeCorrelationId === correlationId) {
                activeCorrelationId = null;
            }
        }
    });
};
exports.register = register;
