// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/cli-utils.js
var inspect = require('util').inspect, wrap = require('word-wrap'), symbols = require('./cli-utils-symbols'), cliUtils;
inspect.styles.string = 'grey';
inspect.styles.name = 'white';
cliUtils = {
    symbols: symbols,
    cliTableTemplate_Blank: {
        top: '',
        'top-mid': '',
        'top-left': '',
        'top-right': '',
        bottom: '',
        'bottom-mid': '',
        'bottom-left': '',
        'bottom-right': '',
        middle: '',
        mid: ' ',
        'mid-mid': '',
        'mid-left': '',
        'mid-right': '',
        left: '',
        'left-mid': '',
        'left-left': '',
        'left-right': '',
        right: '',
        'right-mid': '',
        'right-left': '',
        'right-right': ''
    },
    cliTableTemplateFallback: {
        top: '-',
        'top-mid': '-',
        'top-left': '-',
        'top-right': '-',
        bottom: '-',
        'bottom-mid': '-',
        'bottom-left': '-',
        'bottom-right': '-',
        middle: '|',
        mid: '-',
        'mid-mid': '+',
        'mid-left': '-',
        'mid-right': '-',
        left: '|',
        'left-mid': '-',
        'left-left': '-',
        'left-right': '-',
        right: '|',
        'right-mid': '-',
        'right-left': '-',
        'right-right': '-'
    },
    padLeft: function (nr, n, str) {
        return Array(n - String(nr).length + 1).join(str || '0') + nr;
    },
    noTTY: function () {
        return !process.stdout.isTTY;
    },
    _plainTextEnabled: false,
    enablePlainText: function () {
        cliUtils._plainTextEnabled = true;
    },
    usePlainText: function () {
        return cliUtils._plainTextEnabled || !process.stdout.isTTY;
    },
    inspector: function () {
        var dimension = cliUtils.dimension(), options = {
            depth: 25,
            maxArrayLength: 100,
            colors: require('colors/safe').enabled,
            breakLength: ((dimension.exists && (dimension.width > 20)) ? dimension.width : 60) - 16
        };
        return function (item) {
            return inspect(item, options);
        };
    },
    wrapper: function () {
        var dimension = cliUtils.dimension(), width = ((dimension.exists && (dimension.width > 20)) ? dimension.width : 60) - 6;
        return function (text, indent) {
            return wrap(text, {
                indent: indent,
                width: width,
                cut: true
            });
        };
    },
    dimension: function () {
        var tty, width, height;
        try {
            tty = require('tty');
        }
        catch (e) {
            tty = null;
        }
        if (tty && tty.isatty(1) && tty.isatty(2)) {
            if (process.stdout.getWindowSize) {
                width = process.stdout.getWindowSize(1)[0];
                height = process.stdout.getWindowSize(1)[1];
            }
            else if (tty.getWindowSize) {
                width = tty.getWindowSize()[1];
                height = tty.getWindowSize()[0];
            }
            else if (process.stdout.columns && process.stdout.rows) {
                height = process.stdout.rows;
                width = process.stdout.columns;
            }
        }
        return {
            exists: !(Boolean(process.env.CI) || !process.stdout.isTTY),
            width: width,
            height: height
        };
    }
};
module.exports = cliUtils;
