// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/cli-utils-symbols.js
var IS_WINDOWS = (/^win/).test(process.platform), subsets, symbols;
subsets = {
    regular: {
        console: {
            top: '┌',
            middle: '│',
            bottom: '└'
        },
        dot: '.',
        folder: '❏',
        root: '→',
        sub: '↳',
        ok: '✓',
        error: '✖',
        warning: '⚠',
        star: '★',
        up: '↑',
        down: '↓'
    },
    encoded: {
        console: {
            top: '\u250C',
            middle: '\u2502',
            bottom: '\u2514'
        },
        dot: '.',
        folder: '\u25A1',
        root: '\u2192',
        sub: '\u2514',
        ok: '\u221A',
        error: '\u00D7',
        warning: '\u26A0',
        star: '\u2605',
        up: '\u2191',
        down: '\u2193'
    },
    plainText: {
        console: {
            top: '-',
            middle: '|',
            bottom: '-'
        },
        dot: '.',
        folder: 'Folder',
        root: 'Root',
        sub: 'Sub-folder',
        ok: 'Pass',
        error: 'Fail',
        warning: '!',
        star: '*',
        up: '^',
        down: 'v'
    }
};
symbols = function (disableUnicode) {
    if (disableUnicode) {
        return subsets.plainText;
    }
    if (IS_WINDOWS) {
        return subsets.encoded;
    }
    return subsets.regular;
};
module.exports = symbols;
