// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/index.js
var _ = require('lodash'), sdk = require('postman-collection'), colors = require('colors/safe'), Table = require('cli-table3'), format = require('util').format, util = require('../../util'), cliUtils = require('./cli-utils'), { registerUnifiedHandlers } = require('./unified-reporter'), print = require('../../framework/print'), pad = cliUtils.padLeft, LF = '\n', SPC = ' ', DOT = '.', E = '', CACHED_TIMING_PHASE = '(cache)', TIMING_TABLE_HEADERS = {
    prepare: 'prepare',
    wait: 'wait',
    dns: 'dns-lookup',
    tcp: 'tcp-handshake',
    secureHandshake: 'ssl-handshake',
    firstByte: 'transfer-start',
    download: 'download',
    process: 'process',
    total: 'total'
}, BODY_CLIP_SIZE = 2048, PostmanCLIReporter, timestamp, extractSNR;
colors.setTheme({
    log: 'grey',
    info: 'cyan',
    warn: 'yellow',
    debug: 'blue',
    error: 'red'
});
extractSNR = function (executions) {
    var snr;
    _.isArray(executions) && _.forEachRight(executions, function (execution) {
        var nextReq = _.get(execution, 'result.return.nextRequest');
        if (nextReq) {
            snr = nextReq;
            return false;
        }
    });
    return snr;
};
PostmanCLIReporter = function (emitter, reporterOptions, options) {
    options.disableUnicode && cliUtils.enablePlainText();
    var currentGroup = options.collection, inspect = cliUtils.inspector(), wrap = cliUtils.wrapper(), symbols = cliUtils.symbols(cliUtils.usePlainText());
    if (reporterOptions.silent || options.silent) {
        return;
    }
    emitter.on('done', function () {
        var run = this.summary.run;
        if (!reporterOptions.noSummary) {
            print(LF + PostmanCLIReporter.parseStatistics(run.stats, run.timings, run.transfers) + LF);
        }
        if (!reporterOptions.noFailures && run.failures && run.failures.length) {
            print(LF + PostmanCLIReporter.parseFailures(run.failures) + LF);
        }
    });
    emitter.on('start', function () {
        var collectionIdentifier = currentGroup && (currentGroup.name || currentGroup.id);
        if (!reporterOptions.noBanner) {
            print('%s\n\n', colors.reset('postman'));
        }
        collectionIdentifier && print.lf('%s', colors.reset(collectionIdentifier));
    });
    emitter.on('beforeIteration', function (err, o) {
        if (err || o.cursor.cycles <= 1) {
            return;
        }
        print.lf(LF + colors.gray.underline('Iteration %d/%d'), o.cursor.iteration + 1, o.cursor.cycles);
    });
    emitter.on('test', function (err, o) {
        if (err) {
            return;
        }
        var snr = extractSNR(o.executions);
        if (snr) {
            print.lf(LF + colors.gray('Attempting to set next request to', snr));
        }
    });
    emitter.on('beforeItem', function (err, o) {
        if (err) {
            return;
        }
        var itemGroup = o.item.parent(), root = !itemGroup || (itemGroup === options.collection);
        if (itemGroup && (currentGroup !== itemGroup)) {
            !root && print('\n%s %s', symbols.folder, colors.reset(util.getFullName(itemGroup)));
            currentGroup = itemGroup;
        }
        const symbolToPrint = root ? symbols.root : symbols.sub;
        o.item && print.lf('\n%s %s\n', symbolToPrint, colors.reset(o.item.name || E));
    });
    emitter.on('beforeRequest', function (err, o) {
        if (err || !o.request) {
            return;
        }
        if (reporterOptions.showTimestamps) {
            var currentTime = new Date();
            timestamp = '[' + currentTime.toLocaleString() + ']';
            print('  %s  %s %s ', colors.gray.bold(timestamp), colors.gray.bold(o.request.method), colors.gray.bold(o.request.url.toString()));
        }
        else {
            print('  %s %s ', colors.gray.bold(o.request.method), colors.gray.bold(o.request.url.toString()));
        }
        !cliUtils.usePlainText() && print().wait(colors.gray);
    });
    emitter.on('request', function (err, o) {
        if (err) {
            print.lf(colors.red('[errored]'));
            print.lf(colors.red('     %s'), err.message);
            return;
        }
        if (!(o.request && o.response)) {
            print.lf(colors.red('[errored]'));
            print.lf(colors.red('     %s'), 'Internal error! Could not read response data.');
            return;
        }
        if (!options.verbose) {
            print.lf(colors.gray('[%d %s, %s, %s]'), o.response.code, o.response.reason(), util.filesize(o.response.size().total), util.prettyms(o.response.responseTime));
            return;
        }
        let req = o.request, res = o.response, reqSize = util.filesize(req.size().total), resSize = util.filesize(res.size().total), code = res.code, reason = res.reason(), mime = res.contentInfo() || {}, timings = _.last(_.get(o, 'history.execution.data')), reqHeadersLen = _.get(req, 'headers.members.length'), resHeadersLen = _.get(res, 'headers.members.length'), resTime = util.prettyms(res.responseTime || 0), reqText = (options.verbose && req.body) ? req.body.toString() : E, reqTextLen = req.size().body || Buffer.byteLength(reqText), resText = options.verbose ? res.text() : E, resTextLen = res.size().body || Buffer.byteLength(resText), reqBodyMode = _.get(req, 'body.mode', ''), resSummary = [
            `${mime.contentType}`,
            `${mime.mimeType}`,
            `${mime.mimeFormat}`,
            `${mime.charset}`
        ].join(` ${colors.gray(symbols.star)} `);
        print.lf(SPC);
        print.lf(SPC + SPC + [
            `${code} ${reason}`,
            `${resTime} ${colors.gray('time')}`,
            `${reqSize}${colors.gray(symbols.up)} ${resSize}${colors.gray(symbols.down)} ${colors.gray('size')}`,
            `${reqHeadersLen}${colors.gray(symbols.up)} ` +
                `${resHeadersLen}${colors.gray(symbols.down)} ${colors.gray('headers')}`,
            `${_.get(res, 'cookies.members.length')} ${colors.gray('cookies')}`
        ].join(` ${colors.gray(symbols.star)} `));
        if (reqTextLen) {
            if (reqTextLen > BODY_CLIP_SIZE) {
                reqText = reqText.substr(0, BODY_CLIP_SIZE) +
                    colors.brightWhite(`\n(showing ${util.filesize(BODY_CLIP_SIZE)}/${util.filesize(reqTextLen)})`);
            }
            reqText = wrap(reqText, `  ${colors.white(symbols.console.middle)} `);
            print.buffer(`  ${colors.white(symbols.console.top)} ${colors.white(symbols.up)} ${reqBodyMode} ${colors.gray(symbols.star)} ${util.filesize(reqTextLen)}\n`, colors.white(`  ${symbols.console.bottom}`))
                .nobuffer(colors.gray(reqText.replace(/\n\s*\n/g, LF) + LF));
            print.lf(SPC);
        }
        if (resTextLen) {
            if (resTextLen > BODY_CLIP_SIZE) {
                resText = resText.substr(0, BODY_CLIP_SIZE) +
                    colors.brightWhite(`\n(showing ${util.filesize(BODY_CLIP_SIZE)}/${util.filesize(resTextLen)})`);
            }
            resText = wrap(resText, `  ${colors.white(symbols.console.middle)} `);
            print.buffer(`  ${colors.white(symbols.console.top)} ${colors.white(symbols.down)} ${resSummary} ${colors.gray(symbols.star)} ${util.filesize(resTextLen)}\n`, colors.white(`  ${symbols.console.bottom}`))
                .nobuffer(colors.gray(resText.replace(/\n\s*\n/g, LF) + LF));
        }
        else {
            print.lf(`  ${symbols.down} ${resSummary}`);
        }
        timings = timings && timings.timings;
        if (timings) {
            let timingPhases = util.beautifyTime(sdk.Response.timingPhases(timings)), timingTable = new Table({
                chars: _.defaults({ mid: '', middle: '' }, cliUtils.cliTableTemplate_Blank),
                colAligns: _.fill(Array(_.size(timingPhases)), 'left'),
                style: { 'padding-left': 2 }
            });
            timingPhases = _.transform(TIMING_TABLE_HEADERS, (result, header, key) => {
                if (_.has(timingPhases, key)) {
                    result.headers.push(colors.white(header));
                    result.values.push(colors.log(timingPhases[key] || CACHED_TIMING_PHASE));
                }
            }, { headers: [], values: [] });
            timingTable.push(timingPhases.headers);
            timingTable.push(timingPhases.values);
            print(LF + timingTable + LF + LF);
        }
    });
    emitter.on('script', function (err, o) {
        err && print.lf(colors.red.bold('%s⠄ %s in %s-script'), pad(this.summary.run.failures.length, 3, SPC), err.name, o.event && o.event.listen || 'unknown');
    });
    !reporterOptions.noAssertions && emitter.on('assertion', function (err, o) {
        var passed = !err;
        if (o.skipped && !reporterOptions.noSuccessAssertions) {
            print.lf('%s %s', colors.cyan('  - '), colors.cyan('[skipped] ' + o.assertion));
            return;
        }
        if (passed && reporterOptions.noSuccessAssertions) {
            return;
        }
        if (reporterOptions.showTimestamps) {
            timestamp = '[' + new Date().toLocaleTimeString() + ']';
            print.lf('  %s%s %s', colors.gray(timestamp), passed ? colors.green(`  ${symbols.ok} `) :
                colors.red.bold(pad(this.summary.run.failures.length, 2, SPC) + symbols.dot), passed ?
                colors.gray(o.assertion) : colors.red.bold(o.assertion));
        }
        else {
            print.lf('%s %s', passed ? colors.green(`  ${symbols.ok} `) :
                colors.red.bold(pad(this.summary.run.failures.length, 3, SPC) + symbols.dot), passed ?
                colors.gray(o.assertion) : colors.red.bold(o.assertion));
        }
    });
    !reporterOptions.noConsole && emitter.on('console', function (err, o) {
        if (err) {
            return;
        }
        var color = colors[o.level] || colors.gray, message;
        message = wrap(_.reduce(o.messages, function (log, message) {
            return (log += (log ? colors.white(', ') : '') + inspect(message));
        }, E), `  ${color(symbols.console.middle)} `);
        if (reporterOptions.showTimestamps) {
            print(LF + '  %s', colors.gray('[' + new Date().toLocaleTimeString() + ']' + LF));
        }
        print.buffer(color(`  ${symbols.console.top}\n`), color(`  ${symbols.console.bottom}\n`))
            .nobuffer(colors.gray(message.replace(/\n\s*\n/g, LF) + LF));
    });
    registerUnifiedHandlers(emitter, reporterOptions, options);
};
_.assignIn(PostmanCLIReporter, {
    parseStatistics(stats, timings, transfers) {
        var summaryTable;
        summaryTable = new Table({
            chars: cliUtils.usePlainText() && cliUtils.cliTableTemplateFallback,
            style: { head: [] },
            head: [E, 'executed', '  failed'],
            colAligns: ['right', 'right', 'right'],
            colWidths: [25]
        });
        stats && _.forEach([{
                source: 'iterations',
                label: 'iterations'
            }, {
                source: 'requests',
                label: 'requests'
            }, {
                source: 'testScripts',
                label: 'test-scripts'
            }, {
                source: 'prerequestScripts',
                label: 'prerequest-scripts'
            }, {
                source: 'assertions',
                label: 'assertions'
            }], function (row) {
            var metric = stats[row.source], label = row.label;
            label = metric.failed ? colors.red(label) : (metric.pending ? label : colors.green(label));
            summaryTable.push([
                label,
                metric.total,
                (metric.failed ? colors.red(metric.failed) : metric.failed)
            ]);
        });
        timings && summaryTable.push([{
                colSpan: 3,
                content: format('total run duration: %s', util.prettyms(timings.completed - timings.started)),
                hAlign: 'left'
            }]);
        transfers && summaryTable.push([{
                colSpan: 3,
                content: format('total data received: %s (approx)', util.filesize(transfers.responseTotal)),
                hAlign: 'left'
            }]);
        timings && _.forEach({
            response: 'average response time:',
            dns: 'average DNS lookup time:',
            firstByte: 'average first byte time:'
        }, (value, key) => {
            timings[`${key}Average`] && summaryTable.push([{
                    colSpan: 3,
                    content: format(`${value} %s [min: %s, max: %s, s.d.: %s]`, util.prettyms(timings[`${key}Average`]), util.prettyms(timings[`${key}Min`]), util.prettyms(timings[`${key}Max`]), util.prettyms(timings[`${key}Sd`])),
                    hAlign: 'left'
                }]);
        });
        return summaryTable;
    },
    parseFailures(failures) {
        var failureTable = new Table({
            head: [{
                    hAlign: 'right',
                    content: colors.red.underline('#')
                }, colors.red.underline('failure'),
                colors.red.underline('detail')],
            chars: cliUtils.cliTableTemplate_Blank,
            wordWrap: true,
            colAligns: ['right'],
            colWidths: cliUtils.noTTY() ? [] : (function (size, indexOrder) {
                var colWidths = [];
                if (size.exists && size.width && (size.width > 20)) {
                    colWidths[0] = indexOrder + 3;
                    colWidths[1] = parseInt((size.width - colWidths[0]) * 0.2, 10);
                    colWidths[2] = parseInt(size.width - (colWidths[0] + colWidths[1] + 5), 10);
                }
                return colWidths;
            }(cliUtils.dimension(), Number(failures.length.toString().length)))
        });
        _.forEach(failures, function (failure, index) {
            var name = failure.error && failure.error.name || E, message = failure.error && failure.error.test || E;
            failure.cursor && (failure.cursor.cycles > 1) &&
                (name += LF + colors.gray('iteration: ' + (failure.cursor.iteration + 1)));
            failure.error && (message += LF + colors.gray(failure.error.message || E));
            failure.at && (message += LF + colors.gray('at ' + failure.at));
            if (failure.source) {
                var sourceName = util.getFullName(failure.source) ||
                    failure.source.name || failure.source.title || failure.source.id;
                sourceName && (message += format(colors.gray('\ninside "%s"'), sourceName));
            }
            failureTable.push([pad(Number(index + 1), Number(failures.length.toString().length)).toString() +
                    DOT, name, message]);
        });
        return failureTable;
    }
});
PostmanCLIReporter.prototype.dominant = true;
module.exports = PostmanCLIReporter;
