// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/output.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Output = void 0;
const _ = require("lodash");
const colors = require("colors/safe");
const Table = require("cli-table3");
const util = require("../../util");
const cliUtils = require("./cli-utils");
const print = require("../../framework/print");
const LF = '\n', SPC = ' ', E = '', CACHED_TIMING_PHASE = '(cache)', TIMING_TABLE_HEADERS = {
    prepare: 'prepare',
    wait: 'wait',
    dns: 'dns-lookup',
    tcp: 'tcp-handshake',
    secureHandshake: 'ssl-handshake',
    firstByte: 'transfer-start',
    download: 'download',
    process: 'process',
    total: 'total'
}, BODY_CLIP_SIZE = 2048;
class Output {
    options;
    reporterOptions;
    wrap;
    symbols;
    inspect;
    usePlainText;
    constructor(options, reporterOptions) {
        this.options = options;
        this.reporterOptions = reporterOptions;
        options.disableUnicode && cliUtils.enablePlainText();
        this.usePlainText = cliUtils.usePlainText();
        this.wrap = cliUtils.wrapper();
        this.symbols = cliUtils.symbols(this.usePlainText);
        this.inspect = cliUtils.inspector();
    }
    printBanner() {
        if (!this.reporterOptions.noBanner) {
            print('%s\n\n', colors.reset('postman'));
        }
    }
    printCollectionName(name) {
        if (name) {
            print.lf('%s', colors.reset(name));
        }
    }
    printIteration(iteration, iterations) {
        if (iterations <= 1) {
            return;
        }
        print.lf(LF + colors.gray.underline('Iteration %d/%d'), iteration, iterations);
    }
    printConsoleBox(color, message) {
        if (this.reporterOptions.showTimestamps) {
            print(LF + '  %s', colors.gray('[' + new Date().toLocaleTimeString() + ']' + LF));
        }
        print.buffer(color('  ' + this.symbols.console.top + '\n'), color('  ' + this.symbols.console.bottom + '\n')).nobuffer(colors.gray(message.replace(/\n\s*\n/g, LF) + LF));
    }
    printRequestLine(label, url) {
        if (this.reporterOptions.showTimestamps) {
            const tsStr = '[' + new Date().toLocaleString() + ']';
            print('  %s  %s %s ', colors.gray.bold(tsStr), colors.gray.bold(label), colors.gray.bold(url));
        }
        else {
            print('  %s %s ', colors.gray.bold(label), colors.gray.bold(url));
        }
        !this.usePlainText && print().wait(colors.gray);
    }
    printStatusBracket(statusCode, statusText, totalSize, totalTime) {
        print.lf(colors.gray('[%d %s, %s, %s]'), statusCode, statusText, util.filesize(totalSize), util.prettyms(totalTime));
    }
    printSummaryLine(parts) {
        print.lf(SPC);
        print.lf(SPC + SPC + parts.join(' ' + colors.gray(this.symbols.star) + ' '));
    }
    printBodyBox(text, textLen, direction, modeLabel, labelSize) {
        if (textLen > BODY_CLIP_SIZE) {
            text = text.substr(0, BODY_CLIP_SIZE) +
                colors.brightWhite('\n(showing ' + util.filesize(BODY_CLIP_SIZE) +
                    '/' + util.filesize(textLen) + ')');
        }
        text = this.wrap(text, '  ' + colors.white(this.symbols.console.middle) + ' ');
        const boxLabel = '  ' + colors.white(this.symbols.console.top) + ' ' +
            colors.white(direction) + ' ' +
            (modeLabel ? modeLabel + ' ' + colors.gray(this.symbols.star) + ' ' : '') +
            util.filesize(labelSize !== undefined ? labelSize : textLen) + '\n';
        print.buffer(boxLabel, colors.white('  ' + this.symbols.console.bottom))
            .nobuffer(colors.gray(text.replace(/\n\s*\n/g, LF) + LF));
    }
    parseResponseContentType(headers) {
        if (!headers || !headers.length) {
            return E;
        }
        const contentTypeHeader = _.find(headers, function (h) {
            return h.key && h.key.toLowerCase() === 'content-type';
        });
        if (!contentTypeHeader || !contentTypeHeader.value) {
            return E;
        }
        const ctParts = contentTypeHeader.value.split(';'), mimeType = (ctParts[0] || '').trim(), charsetMatch = contentTypeHeader.value.match(/charset=([^\s;]+)/i), charset = charsetMatch ? charsetMatch[1] : '', mimeParts = mimeType.split('/'), mainType = mimeParts[0] || '', subType = mimeParts[1] || '';
        return [mimeType, mainType, subType, charset]
            .filter(Boolean)
            .join(' ' + colors.gray(this.symbols.star) + ' ');
    }
    countResponseCookies(headers) {
        if (!headers || !headers.length) {
            return 0;
        }
        return _.filter(headers, function (h) {
            return h.key && h.key.toLowerCase() === 'set-cookie';
        }).length;
    }
    printTimingTable(timings) {
        if (!timings) {
            return;
        }
        const timingTable = new Table({
            chars: _.defaults({ mid: '', middle: '' }, cliUtils.cliTableTemplate_Blank),
            colAligns: _.fill(Array(_.size(timings)), 'left'),
            style: { 'padding-left': 2 }
        }), timingHeaders = [], timingValues = [];
        _.forEach(TIMING_TABLE_HEADERS, function (header, key) {
            if (_.has(timings, key)) {
                timingHeaders.push(colors.white(header));
                timingValues.push(colors.log(timings[key] ? util.prettyms(timings[key]) : CACHED_TIMING_PHASE));
            }
        });
        if (timingHeaders.length) {
            timingTable.push(timingHeaders);
            timingTable.push(timingValues);
            print(LF + timingTable + LF + LF);
        }
    }
    printWarning(message) {
        print.lf(colors.yellow('  %s %s'), this.symbols.warning, message);
    }
    printError(message) {
        print.lf(colors.red('[errored]'));
        print.lf(colors.red('     %s'), message);
    }
    printSpacer() {
        print.lf(SPC);
    }
    printStatusMessage(message) {
        print.lf(colors.gray(message));
    }
    printContentLabel(direction, label) {
        print.lf('  ' + direction + ' ' + label);
    }
    printGraphqlErrors(errors) {
        if (!errors || !errors.length) {
            return;
        }
        _.forEach(errors, (gqlErr) => {
            this.printWarning('GraphQL Error: ' + (gqlErr.message || String(gqlErr)));
        });
    }
}
exports.Output = Output;
