// @postman/pm-bin-linux-x64/bin/lib/reporters/cli/state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunState = void 0;
class RunState {
    correlationOwners = new Map();
    scriptErrors = new Set();
    requestsSent = new Set();
    stores = new Map();
    _failureCount = 0;
    _consoleDeferrer = null;
    _currentGroup = null;
    getStore(namespace) {
        if (!this.stores.has(namespace)) {
            this.stores.set(namespace, new Map());
        }
        return this.stores.get(namespace);
    }
    getCurrentGroup() {
        return this._currentGroup;
    }
    setCurrentGroup(group) {
        this._currentGroup = group;
    }
    clearCurrentGroup() {
        this._currentGroup = null;
    }
    claimCorrelation(id, owner) {
        this.correlationOwners.set(id, owner);
    }
    getOwner(id) {
        return this.correlationOwners.get(id);
    }
    releaseCorrelation(id) {
        this.correlationOwners.delete(id);
    }
    clearCorrelation(id) {
        for (const store of this.stores.values()) {
            store.delete(id);
        }
        this.correlationOwners.delete(id);
        this.scriptErrors.delete(id);
        this.requestsSent.delete(id);
        if (this._consoleDeferrer) {
            this._consoleDeferrer = null;
        }
    }
    markRequestSent(id) {
        this.requestsSent.add(id);
    }
    wasRequestSent(id) {
        return this.requestsSent.has(id);
    }
    markScriptError(id) {
        this.scriptErrors.add(id);
    }
    hasScriptError(id) {
        return this.scriptErrors.has(id);
    }
    incrementFailures() {
        return ++this._failureCount;
    }
    getFailureCount() {
        return this._failureCount;
    }
    setConsoleDeferrer(deferrer) {
        this._consoleDeferrer = deferrer;
    }
    getConsoleDeferrer() {
        return this._consoleDeferrer;
    }
    clearConsoleDeferrer() {
        this._consoleDeferrer = null;
    }
}
exports.RunState = RunState;
