// @postman/pm-bin-linux-x64/bin/lib/commands/runner/analytics/api.js
const request = require('postman-request'), util = require('../../../util'), SERVICE_NAME = 'monitoring-orchestrator', TIMEOUT = 30000, MAX_RETRIES = 2, RETRY_DELAY_MS = 100, RETRYABLE_ERROR_CODES = ['ECONNREFUSED', 'ENOTFOUND', 'ECONNRESET', 'ETIMEDOUT', 'EAI_AGAIN'], RETRYABLE_STATUS_CODES = [502, 503];
function isRetryable(err, response) {
    if (err && RETRYABLE_ERROR_CODES.includes(err.code)) {
        return true;
    }
    if (response && RETRYABLE_STATUS_CODES.includes(response.statusCode)) {
        return true;
    }
    return false;
}
function delay(ms) {
    return new Promise((resolve) => { setTimeout(resolve, ms); });
}
function doRequest(url, headers, body) {
    return new Promise((resolve) => {
        request.post({
            url: url,
            headers: headers,
            body: body,
            timeout: TIMEOUT
        }, (err, response, responseBody) => {
            if (err) {
                resolve({
                    success: false,
                    error: err.message,
                    retryable: isRetryable(err, null)
                });
                return;
            }
            if (response.statusCode >= 200 && response.statusCode < 300) {
                resolve({ success: true });
                return;
            }
            let errorMsg = `HTTP ${response.statusCode}`;
            try {
                const parsed = responseBody && JSON.parse(responseBody);
                if (parsed?.message) {
                    errorMsg += `: ${parsed.message}`;
                }
            }
            catch { }
            resolve({
                success: false,
                error: errorMsg,
                retryable: isRetryable(null, response)
            });
        });
    });
}
class RunnerEventsApi {
    constructor(options) {
        if (!options?.runId) {
            throw new Error('runId is required');
        }
        if (!options?.eventingKey) {
            throw new Error('eventingKey is required');
        }
        this.runId = options.runId;
        this.eventingKey = options.eventingKey;
        this.baseUrl = process.env.POSTMAN_RUNNER_URL || util.POSTMAN_GATEWAY_BASE_URL();
    }
    async initAuth() {
        return true;
    }
    buildHeaders() {
        return {
            ...util.defaultHeaders,
            'x-original-user-agent': util.userAgent,
            'x-pstmn-req-service': SERVICE_NAME,
            'X-Eventing-Key': this.eventingKey,
            'Content-Type': 'application/json'
        };
    }
    async insert(table, rows) {
        if (!rows || rows.length === 0) {
            return { success: true };
        }
        const url = `${this.baseUrl}/run/${this.runId}/events`, headers = this.buildHeaders(), body = JSON.stringify({
            data: {
                table: table,
                data: rows
            }
        });
        let lastError = null;
        for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
            if (attempt > 0) {
                await delay(RETRY_DELAY_MS * attempt);
            }
            const result = await doRequest(url, headers, body);
            if (result.success) {
                return result;
            }
            if (!result.retryable || attempt === MAX_RETRIES) {
                return { success: false, error: result.error };
            }
            lastError = result.error;
        }
        return { success: false, error: lastError };
    }
}
function create(options) {
    return new RunnerEventsApi(options);
}
module.exports = {
    RunnerEventsApi: RunnerEventsApi,
    create: create
};
