// @postman/pm-bin-linux-x64/bin/lib/commands/runner/analytics/diff-tracker.js
const jsonpatch = require('fast-json-patch'), _ = require('lodash'), fs = require('fs'), path = require('path');
const DEBUG = Boolean(process.env.POSTMAN_DEBUG_ANALYTICS), DEBUG_DIR = path.join(__dirname, '../../../../runner');
function serialize(runLog) {
    if (!runLog) {
        return {};
    }
    if (typeof runLog.toJSON === 'function') {
        return runLog.toJSON();
    }
    return JSON.parse(JSON.stringify(runLog));
}
function reconstructFromPatches(initial, patchesList) {
    return patchesList.reduce((doc, patches) => {
        return patches?.length ? jsonpatch.applyPatch(doc, patches, true, false).newDocument : doc;
    }, _.cloneDeep(initial));
}
class DiffTracker {
    constructor() {
        this.previousSnapshot = null;
        this.isFirstEvent = true;
        this.debugLog = [];
    }
    computeDiff(runLog) {
        const currentSnapshot = serialize(runLog);
        if (this.isFirstEvent) {
            this.isFirstEvent = false;
            this.previousSnapshot = _.cloneDeep(currentSnapshot);
            if (DEBUG) {
                this.debugLog.push({ type: 'seed', data: currentSnapshot });
            }
            return { isSeed: true, runLog: currentSnapshot };
        }
        const patches = jsonpatch.compare(this.previousSnapshot, currentSnapshot);
        this.previousSnapshot = _.cloneDeep(currentSnapshot);
        if (DEBUG) {
            this.debugLog.push({ type: 'patch', data: patches });
        }
        return { isSeed: false, patches: patches };
    }
    reconstruct() {
        if (!this.debugLog.length) {
            return null;
        }
        const initial = this.debugLog[0]?.data;
        const patches = this.debugLog.slice(1).map((e) => { return e.data; });
        return reconstructFromPatches(initial, patches);
    }
    writeDebugLog(filename = 'analytics-debug.json') {
        if (!DEBUG || !this.debugLog.length) {
            return;
        }
        if (!fs.existsSync(DEBUG_DIR)) {
            fs.mkdirSync(DEBUG_DIR, { recursive: true });
        }
        fs.writeFileSync(path.join(DEBUG_DIR, filename), JSON.stringify(this.debugLog, null, 2));
    }
    writeReconstructed(filename = 'analytics-reconstructed.json') {
        if (!DEBUG || !this.debugLog.length) {
            return;
        }
        const reconstructed = this.reconstruct();
        if (reconstructed) {
            fs.writeFileSync(path.join(DEBUG_DIR, filename), JSON.stringify(reconstructed, null, 2));
        }
    }
    writeActual(runLog, filename = 'analytics-actual.json') {
        if (!DEBUG || !runLog || !this.debugLog.length) {
            return;
        }
        fs.writeFileSync(path.join(DEBUG_DIR, filename), JSON.stringify(serialize(runLog), null, 2));
    }
    reset() {
        this.previousSnapshot = null;
        this.isFirstEvent = true;
        this.debugLog = [];
    }
}
module.exports = DiffTracker;
module.exports.reconstructFromPatches = reconstructFromPatches;
