// @postman/pm-bin-linux-x64/bin/lib/commands/runner/analytics/event-streamer.js
const { v4: uuidv4 } = require('uuid'), { createClient } = require('../../../framework/clickhouse'), { create: createRunnerEventsApi } = require('./api'), DiffTracker = require('./diff-tracker'), { buildBaseEvent, extractServiceInfo, extractEndpointPattern, generateEndpointId } = require('./event-builder');
class EventStreamer {
    constructor(options) {
        const api = createRunnerEventsApi({
            runId: options.runId,
            eventingKey: options.eventIngestion?.eventingKey
        });
        this.client = createClient({ api });
        this.diffTracker = new DiffTracker();
        this.runLog = null;
        this.state = {
            eventSpan: uuidv4(),
            eventIndex: 0,
            runId: options.runId || '',
            jobId: options.jobId || '',
            runnerId: options.runnerId || '',
            collectionId: options.collectionId || '',
            collectionName: options.collectionName || '',
            workspaceId: options.workspaceId || '',
            environmentId: options.environmentId || '',
            environmentName: options.environmentName || '',
            orgId: options.orgId || '',
            monitorId: options.monitorId || '',
            executionEnv: options.executionEnv || '',
            trigger: options.trigger || ''
        };
        this.pendingResponses = new Map();
        this.requestAssertions = new Map();
        this.stats = {
            requests: { total: 0, failed: 0 },
            assertions: { total: 0, failed: 0 },
            scripts: { total: 0, failed: 0 }
        };
        this.timings = {
            started: null,
            completed: null,
            responseTimes: []
        };
        this.tableName = options.eventIngestion?.tableName || 'postman_event_log_v3';
        this.debug = Boolean(process.env.POSTMAN_DEBUG_ANALYTICS);
    }
    setRunLog(runLog) {
        this.runLog = runLog;
    }
    _computeRunLogSnapshot() {
        if (!this.runLog) {
            return null;
        }
        const diff = this.diffTracker.computeDiff(this.runLog);
        return {
            is_seed: diff.isSeed,
            ...(diff.isSeed ? { run_log: diff.runLog } : { patches: diff.patches })
        };
    }
    onStart(args) {
        try {
            this.timings.started = Date.now();
            const event = buildBaseEvent(this.state, 'run_started'), runLogDiff = this._computeRunLogSnapshot();
            event.event_data = {
                collection_name: this.state.collectionName || '',
                environment_id: this.state.environmentId || null,
                environment_name: this.state.environmentName || null,
                job_id: this.state.jobId || null,
                run_id: this.state.runId || null,
                monitor_id: this.state.monitorId || null,
                execution_environment: this.state.executionEnv || null,
                executor: {
                    arch: process.arch,
                    name: 'postman-runner',
                    platform: process.platform,
                    version: require('../../../../package.json').version
                },
                execution_plan: {
                    total_iterations: String(args?.cursor?.cycles || 0),
                    total_requests: String(args?.cursor?.length || 0)
                },
                run_log_snapshot: runLogDiff
            };
            this.client.buffer(this.tableName, event);
        }
        catch (error) {
            this._log('Error in onStart:', error.message);
        }
    }
    onResponse(args) {
        try {
            this.stats.requests.total++;
            if (args.error) {
                this.stats.requests.failed++;
            }
            if (args.response?.responseTime) {
                this.timings.responseTimes.push(args.response.responseTime);
            }
            const key = `${args.cursor?.iteration || 0}:${args.cursor?.position || 0}`, svc = extractServiceInfo(args.request?.url), endpointPattern = extractEndpointPattern(args.request?.url), method = args.request?.method || 'GET', endpointId = generateEndpointId(svc.serviceName || '', method, endpointPattern);
            this.pendingResponses.set(key, {
                timestamp: new Date().toISOString(),
                serviceName: svc.serviceName,
                baseUrl: svc.baseUrl,
                endpointPattern: endpointPattern || '',
                endpointMethod: method,
                endpointId: endpointId || '',
                item: { id: args.item?.id || '', name: args.item?.name || '' },
                request: {
                    method: method,
                    url: args.request?.url?.toString() || ''
                },
                response: {
                    code: args.response?.code || 0,
                    status: args.response?.status || '',
                    time_ms: args.response?.responseTime || 0,
                    size_bytes: args.response?.responseSize || 0
                },
                position: {
                    iteration: args.cursor?.iteration || 0,
                    request_index: args.cursor?.position || 0
                },
                _urlObject: args.request?.url
            });
        }
        catch (error) {
            this._log('Error in onResponse:', error.message);
        }
    }
    onAssertion(args) {
        try {
            const key = `${args.cursor?.iteration || 0}:${args.cursor?.position || 0}`;
            let data = this.requestAssertions.get(key);
            if (!data) {
                const resp = this.pendingResponses.get(key), urlObject = resp?._urlObject || args.item?.request?.url, pattern = extractEndpointPattern(urlObject), method = resp?.request?.method || args.item?.request?.method || '', endpointId = generateEndpointId(resp?.serviceName || '', method, pattern);
                data = {
                    endpointPattern: pattern || '',
                    endpointMethod: method,
                    endpointId: endpointId || '',
                    total: 0,
                    passed: 0,
                    failed: 0,
                    skipped: 0
                };
                this.requestAssertions.set(key, data);
            }
            const assertions = args.assertion;
            if (!assertions) {
                return;
            }
            const assertionList = Array.isArray(assertions) ? assertions : [assertions];
            for (const assertion of assertionList) {
                if (!assertion) {
                    continue;
                }
                data.total++;
                this.stats.assertions.total++;
                if (assertion.skipped) {
                    data.skipped++;
                }
                else if (!assertion.passed) {
                    data.failed++;
                    this.stats.assertions.failed++;
                }
                else {
                    data.passed++;
                }
            }
        }
        catch (error) {
            this._log('Error in onAssertion:', error.message);
        }
    }
    onItem(args) {
        try {
            const key = `${args.cursor?.iteration || 0}:${args.cursor?.position || 0}`, resp = this.pendingResponses.get(key), asserts = this.requestAssertions.get(key);
            if (!resp) {
                return;
            }
            this.pendingResponses.delete(key);
            this.requestAssertions.delete(key);
            const event = buildBaseEvent(this.state, 'request_executed'), runLogDiff = this._computeRunLogSnapshot();
            event.timestamp = resp.timestamp;
            event.service_name = resp.serviceName || '';
            event.service_base_url = resp.baseUrl || '';
            event.endpoint_pattern = asserts?.endpointPattern || resp.endpointPattern || '';
            event.endpoint_method = asserts?.endpointMethod || resp.endpointMethod || '';
            event.endpoint_id = asserts?.endpointId || resp.endpointId || '';
            event.event_data = {
                item: resp.item,
                request: resp.request,
                response: resp.response,
                position: resp.position,
                job_id: this.state.jobId || null,
                run_id: this.state.runId || null,
                monitor_id: this.state.monitorId || null,
                execution_environment: this.state.executionEnv || null,
                tests: asserts ? {
                    total_assertions: asserts.total,
                    passed_assertions: asserts.passed,
                    failed_assertions: asserts.failed,
                    skipped_assertions: asserts.skipped,
                    has_failures: asserts.failed > 0,
                    has_exceptions: false
                } : undefined,
                run_log_snapshot: runLogDiff
            };
            this.client.buffer(this.tableName, event);
        }
        catch (error) {
            this._log('Error in onItem:', error.message);
        }
    }
    onDone() {
        try {
            this.timings.completed = Date.now();
            for (const [key, resp] of this.pendingResponses) {
                const asserts = this.requestAssertions.get(key), event = buildBaseEvent(this.state, 'request_executed');
                event.timestamp = resp.timestamp;
                event.service_name = resp.serviceName || '';
                event.service_base_url = resp.baseUrl || '';
                event.endpoint_pattern = asserts?.endpointPattern || resp.endpointPattern || '';
                event.endpoint_method = asserts?.endpointMethod || resp.endpointMethod || '';
                event.endpoint_id = asserts?.endpointId || resp.endpointId || '';
                event.event_data = {
                    item: resp.item,
                    request: resp.request,
                    response: resp.response,
                    position: resp.position,
                    job_id: this.state.jobId || null,
                    run_id: this.state.runId || null,
                    monitor_id: this.state.monitorId || null,
                    execution_environment: this.state.executionEnv || null,
                    tests: asserts ? {
                        total_assertions: asserts.total,
                        passed_assertions: asserts.passed,
                        failed_assertions: asserts.failed,
                        skipped_assertions: asserts.skipped,
                        has_failures: asserts.failed > 0,
                        has_exceptions: false
                    } : undefined,
                    run_log_snapshot: null
                };
                this.client.buffer(this.tableName, event);
            }
            this.pendingResponses.clear();
            this.requestAssertions.clear();
            const responseTimes = this.timings.responseTimes;
            let responseAverage = 0, responseMin = 0, responseMax = 0;
            if (responseTimes.length > 0) {
                responseAverage = responseTimes.reduce(function (a, b) { return a + b; }, 0) / responseTimes.length;
                responseMin = Math.min.apply(null, responseTimes);
                responseMax = Math.max.apply(null, responseTimes);
            }
            const computedTimings = {
                responseAverage: responseAverage,
                responseMin: responseMin,
                responseMax: responseMax,
                started: this.timings.started,
                completed: this.timings.completed
            }, event = buildBaseEvent(this.state, 'run_completed'), runLogDiff = this._computeRunLogSnapshot();
            event.event_data = {
                summary: {
                    run_passed: this.stats.assertions.failed === 0 &&
                        this.stats.scripts.failed === 0 &&
                        this.stats.requests.failed === 0,
                    total_duration_ms: (this.timings.completed || 0) - (this.timings.started || 0),
                    total_requests: this.stats.requests.total,
                    failed_requests: this.stats.requests.failed,
                    total_assertions: this.stats.assertions.total,
                    failed_assertions: this.stats.assertions.failed,
                    total_scripts: this.stats.scripts.total,
                    failed_scripts: this.stats.scripts.failed
                },
                job_id: this.state.jobId || null,
                run_id: this.state.runId || null,
                monitor_id: this.state.monitorId || null,
                execution_environment: this.state.executionEnv || null,
                stats: this.stats,
                timings: computedTimings,
                run_log_snapshot: runLogDiff
            };
            this.client.buffer(this.tableName, event);
            this.diffTracker.writeDebugLog();
            this.diffTracker.writeReconstructed();
            this.diffTracker.writeActual(this.runLog);
        }
        catch (error) {
            this._log('Error in onDone:', error.message);
        }
    }
    onScript(args) {
        try {
            this.stats.scripts.total++;
            if (args.error) {
                this.stats.scripts.failed++;
            }
        }
        catch (error) {
            this._log('Error in onScript:', error.message);
        }
    }
    onError(args) {
        try {
            const event = buildBaseEvent(this.state, 'run_error'), runLogDiff = this._computeRunLogSnapshot();
            event.event_data = {
                error: {
                    message: args.error?.message || 'Unknown error',
                    code: args.error?.code || null,
                    stack: args.error?.stack || null
                },
                job_id: this.state.jobId || null,
                run_id: this.state.runId || null,
                monitor_id: this.state.monitorId || null,
                execution_environment: this.state.executionEnv || null,
                run_log_snapshot: runLogDiff
            };
            this.client.buffer(this.tableName, event);
        }
        catch (error) {
            this._log('Error in onError:', error.message);
        }
    }
    onEvent(eventName, eventArgs) {
        switch (eventName) {
            case 'start':
                this.onStart(eventArgs);
                break;
            case 'response':
                this.onResponse(eventArgs);
                break;
            case 'assertion':
                this.onAssertion(eventArgs);
                break;
            case 'script':
                this.onScript(eventArgs);
                break;
            case 'item':
                this.onItem(eventArgs);
                break;
            case 'done':
                this.onDone(eventArgs);
                break;
            case 'error':
                this.onError(eventArgs);
                break;
            default:
                break;
        }
    }
    async flush() {
        try {
            await this.client.flush(this.tableName);
        }
        catch (error) {
            this._log('Flush error:', error.message);
        }
    }
    destroy() {
        this.diffTracker.reset();
        this.pendingResponses.clear();
        this.requestAssertions.clear();
        this.timings.responseTimes = [];
        this.client.destroy();
    }
    _log(...args) {
        if (this.debug) {
            console.log('[Runner Analytics]', ...args);
        }
    }
}
module.exports = EventStreamer;
