// @postman/pm-bin-linux-x64/bin/lib/commands/runner/analytics/event-builder.js
const { extractServiceInfo, extractEndpointPattern, generateEndpointId } = require('../../../framework/clickhouse/event-helpers');
function buildBaseEvent(state, eventType) {
    const trigger = state.trigger || 'unknown';
    return {
        event_type: eventType,
        event_span: state.eventSpan,
        event_index: state.eventIndex++,
        timestamp: new Date().toISOString(),
        collection_id: state.collectionId || '',
        workspace_id: state.workspaceId || '',
        environment_id: state.environmentId || '',
        org_id: String(state.orgId || ''),
        system_environment: state.executionEnv || 'default',
        executor_type: `${trigger}-runner`,
        service_name: '',
        service_base_url: '',
        endpoint_pattern: '',
        endpoint_method: '',
        endpoint_id: '',
        ci_platform: '',
        pr_number: 0,
        git_branch: '',
        git_commit_sha: '',
        repo_name: '',
        repo_url: ''
    };
}
module.exports = {
    buildBaseEvent,
    extractServiceInfo,
    extractEndpointPattern,
    generateEndpointId
};
