// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/ci-info.js
const _ = require('lodash'), isCI = require('is-ci'), githubActions = require('./ci-providers/github-actions'), gitlab = require('./ci-providers/gitlab'), circleci = require('./ci-providers/circleci'), jenkins = require('./ci-providers/jenkins'), travis = require('./ci-providers/travis'), bitbucket = require('./ci-providers/bitbucket'), azurePipelines = require('./ci-providers/azure-pipelines'), awsCodeBuild = require('./ci-providers/aws-codebuild'), googleCloudBuild = require('./ci-providers/google-cloud-build'), buildkite = require('./ci-providers/buildkite'), PROVIDER_NAMES = {
    GITHUB_ACTIONS: 'github-actions',
    GITLAB: 'gitlab',
    JENKINS: 'jenkins',
    CIRCLECI: 'circleci',
    TRAVIS: 'travis',
    BITBUCKET: 'bitbucket',
    AZURE_PIPELINES: 'azure-pipelines',
    AWS_CODEBUILD: 'aws-codebuild',
    GOOGLE_CLOUD_BUILD: 'google-cloud-build',
    TEAMCITY: 'teamcity',
    BAMBOO: 'bamboo',
    DRONE: 'drone',
    BUILDKITE: 'buildkite',
    UNKNOWN: 'unknown'
}, CI_PROVIDERS = [
    githubActions,
    gitlab,
    jenkins,
    circleci,
    travis,
    bitbucket,
    azurePipelines,
    awsCodeBuild,
    googleCloudBuild,
    buildkite
];
function removeUndefinedDeep(obj) {
    if (_.isNil(obj)) {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(removeUndefinedDeep).filter((item) => { return item !== undefined; });
    }
    if (typeof obj === 'object') {
        const cleaned = {};
        for (const [key, value] of Object.entries(obj)) {
            if (value !== undefined) {
                const cleanedValue = removeUndefinedDeep(value);
                if (cleanedValue !== undefined) {
                    cleaned[key] = cleanedValue;
                }
            }
        }
        return cleaned;
    }
    return obj;
}
function detectCIProvider(env) {
    for (const provider of CI_PROVIDERS) {
        if (provider.detect(env)) {
            return provider;
        }
    }
    return null;
}
function getCIInfo(env = process.env) {
    const inCI = (env === process.env) ? isCI : detectCIProvider(env) !== null, providerModule = inCI ? detectCIProvider(env) : null;
    if (!inCI) {
        return null;
    }
    if (!providerModule) {
        return { provider: PROVIDER_NAMES.UNKNOWN };
    }
    return removeUndefinedDeep(providerModule.getInfo(env));
}
function getCIProviderName(env) {
    const providerModule = detectCIProvider(env);
    return providerModule ? providerModule.name : null;
}
module.exports = {
    PROVIDER_NAMES: PROVIDER_NAMES,
    getCIInfo: getCIInfo,
    detectCIProvider: getCIProviderName
};
