// @postman/pm-bin-linux-x64/bin/lib/framework/execution-context/index.js
const _ = require('lodash'), isCI = require('is-ci'), { getSystemInfo } = require('./system-info'), { getCIInfo, detectCIProvider } = require('./ci-info'), { version } = require('../../../package.json'), DEFAULT_HEADER_PREFIX = 'x-postman-cli-', MAX_HEADER_LENGTH = 1024;
let contextCache = null;
function getCLIInfo() {
    return {
        version: version,
        userAgent: `Postman CLI/${version}`
    };
}
function flattenToHeaders(obj, headers, prefix, path = '') {
    for (const [key, value] of Object.entries(obj)) {
        const currentPath = path ? `${path}-${key}` : key;
        if (_.isNil(value)) {
            continue;
        }
        if (typeof value === 'object' && !Array.isArray(value)) {
            flattenToHeaders(value, headers, prefix, currentPath);
        }
        else {
            const headerKey = `${prefix}${currentPath}`, headerValue = Array.isArray(value) ? value.join(',') : String(value);
            if (headerValue.length <= MAX_HEADER_LENGTH) {
                headers[headerKey] = headerValue;
            }
        }
    }
}
function getAll(options = {}) {
    if (!options.noCache && contextCache) {
        return contextCache;
    }
    const context = {
        cli: getCLIInfo(),
        isCI: isCI,
        ci: getCIInfo(),
        system: getSystemInfo()
    };
    contextCache = context;
    return context;
}
function isCIEnvironment() {
    return isCI;
}
function getCIProvider() {
    return detectCIProvider(process.env);
}
function toHeaders(options = {}) {
    const { prefix = DEFAULT_HEADER_PREFIX, flatten = true } = options, context = getAll(), headers = {};
    if (flatten) {
        flattenToHeaders(context, headers, prefix);
    }
    else {
        for (const [key, value] of Object.entries(context)) {
            if (!_.isNil(value)) {
                const headerKey = `${prefix}${key}`, headerValue = typeof value === 'object' ? JSON.stringify(value) : String(value);
                if (headerValue.length <= MAX_HEADER_LENGTH) {
                    headers[headerKey] = headerValue;
                }
            }
        }
    }
    return headers;
}
function clearCache() {
    contextCache = null;
}
module.exports = {
    getAll: getAll,
    toHeaders: toHeaders,
    isCI: isCIEnvironment,
    getCIProvider: getCIProvider,
    clearCache: clearCache
};
