// @postman/pm-bin-linux-x64/bin/lib/commands/mock/runtime/pm/mock.js
const { matchRequest: matchRequestAlgorithm } = require('../matching-algorithm');
let examplesMap = new Map();
let requestsMap = new Map();
function initialize(examples = [], requests = []) {
    examplesMap.clear();
    requestsMap.clear();
    for (const example of examples) {
        examplesMap.set(example.id, example);
    }
    for (const request of requests) {
        requestsMap.set(request.id, {
            id: request.id,
            name: request.name,
            method: request.method || 'GET',
            url: request.url || '',
            path: request.path || ''
        });
    }
}
const pmMockAPI = {
    matchRequest(requestId, req) {
        if (!requestId || !req) {
            return false;
        }
        const savedRequest = requestsMap.get(requestId);
        if (!savedRequest) {
            return false;
        }
        return matchRequestAlgorithm(req, savedRequest);
    },
    sendExample(exampleId, res) {
        if (!exampleId || !res) {
            return false;
        }
        const example = examplesMap.get(exampleId);
        if (!example) {
            return false;
        }
        try {
            const statusCode = example.statusCode || 200;
            const body = example.body || '';
            res.statusCode = statusCode;
            res.setHeader('Content-Type', 'application/json');
            res.end(body);
            return true;
        }
        catch (error) {
            console.error('[pm.mock.sendExample] Error:', error.message);
            return false;
        }
    }
};
module.exports = { initialize, pmMockAPI };
