// @postman/pm-bin-linux-x64/bin/lib/framework/clickhouse/api.js
const request = require('postman-request'), util = require('../../util'), config = require('../../config'), analyticsConfig = require('./config'), console = require('../console'), DEBUG = Boolean(process.env[analyticsConfig.DEBUG_ENV_VAR]), LOG_PREFIX = analyticsConfig.LOG_PREFIX, INSERT_PATH = '/api/v1/insert', SERVICE_NAME = 'galactus-service', TIMEOUT = 30000, MAX_RETRIES = 2, RETRY_DELAY_MS = 100, RETRYABLE_ERROR_CODES = ['ECONNREFUSED', 'ENOTFOUND', 'ECONNRESET', 'ETIMEDOUT', 'EAI_AGAIN'], RETRYABLE_STATUS_CODES = [502, 503];
function debugLog(...args) {
    if (DEBUG) {
        console.log(LOG_PREFIX, ...args);
    }
}
function isRetryable(err, response) {
    if (err && RETRYABLE_ERROR_CODES.includes(err.code)) {
        return true;
    }
    if (response && RETRYABLE_STATUS_CODES.includes(response.statusCode)) {
        return true;
    }
    return false;
}
function delay(ms) {
    return new Promise((resolve) => { setTimeout(resolve, ms); });
}
function fetchAuth() {
    return new Promise((resolve, reject) => {
        config.get({}, { command: 'default' }, (err, opts) => {
            if (err) {
                return reject(err);
            }
            util.extractPostmanApiKey(opts.postmanApiKey, (extractErr, apiKey) => {
                if (extractErr) {
                    return reject(extractErr);
                }
                resolve({ apiKey: apiKey, accessToken: opts.accessToken });
            });
        });
    });
}
function doRequest(url, headers, body) {
    return new Promise((resolve) => {
        request.post({
            url: url,
            headers: headers,
            body: body,
            timeout: TIMEOUT
        }, (err, response, responseBody) => {
            if (err) {
                resolve({
                    success: false,
                    error: err.message,
                    retryable: isRetryable(err, null)
                });
                return;
            }
            if (response.statusCode >= 200 && response.statusCode < 300) {
                resolve({ success: true });
                return;
            }
            let errorMsg = `HTTP ${response.statusCode}`;
            try {
                const parsed = responseBody && JSON.parse(responseBody);
                if (parsed?.message) {
                    errorMsg += `: ${parsed.message}`;
                }
            }
            catch { }
            resolve({
                success: false,
                error: errorMsg,
                retryable: isRetryable(null, response)
            });
        });
    });
}
class ClickhouseApi {
    constructor() {
        this.auth = null;
        this.initialized = false;
    }
    async initAuth() {
        if (this.initialized) {
            return Boolean(this.auth?.apiKey || this.auth?.accessToken);
        }
        try {
            this.auth = await fetchAuth();
            debugLog('Auth initialized successfully');
        }
        catch (err) {
            debugLog('Auth initialization failed:', err?.message || err);
            this.auth = null;
        }
        this.initialized = true;
        return Boolean(this.auth?.apiKey || this.auth?.accessToken);
    }
    buildHeaders() {
        const headers = {
            ...util.defaultHeaders,
            'x-original-user-agent': util.userAgent,
            'x-pstmn-req-service': SERVICE_NAME,
            'Content-Type': 'application/json'
        };
        if (this.auth?.apiKey) {
            headers['x-api-key'] = this.auth.apiKey;
        }
        else if (this.auth?.accessToken) {
            headers['x-access-token'] = this.auth.accessToken;
        }
        return headers;
    }
    async insert(table, rows) {
        const hasAuth = await this.initAuth();
        if (!hasAuth) {
            return { success: true, skipped: true };
        }
        if (!rows || rows.length === 0) {
            return { success: true };
        }
        const url = `${util.POSTMAN_GATEWAY_BASE_URL()}${INSERT_PATH}`, body = JSON.stringify({ table: table, data: rows }), headers = this.buildHeaders();
        let lastError = null;
        for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
            if (attempt > 0) {
                await delay(RETRY_DELAY_MS * attempt);
            }
            const result = await doRequest(url, headers, body);
            if (result.success) {
                return result;
            }
            if (!result.retryable || attempt === MAX_RETRIES) {
                return { success: false, error: result.error };
            }
            lastError = result.error;
        }
        return { success: false, error: lastError };
    }
}
function create() {
    return new ClickhouseApi();
}
module.exports = {
    ClickhouseApi: ClickhouseApi,
    create: create
};
