// @postman/pm-bin-linux-x64/bin/lib/framework/clickhouse/ci-context.js
const ExecutionContext = require('../execution-context'), TABLE = 'postman_ci_context';
function build(options) {
    const ctx = ExecutionContext.getAll(), ci = ctx?.ci || {}, system = ctx?.system || {};
    return {
        timestamp: new Date().toISOString(),
        event_span: options.eventSpan,
        collection_id: options.collectionId || '',
        workspace_id: options.workspaceId || '',
        environment_id: options.environmentId || '',
        system_environment: options.systemEnvironment || 'default',
        executor_type: options.executorType,
        cli_version: options.cliVersion || '',
        ci_provider: ci.provider || '',
        branch: ci.branch || '',
        commit_sha: ci.commit?.sha || '',
        workflow_name: ci.workflow?.name || '',
        workflow_id: ci.workflow?.id || ci.pipeline?.id || '',
        job_name: ci.job?.name || '',
        job_id: ci.job?.id || '',
        pipeline_id: ci.pipeline?.id || '',
        pipeline_number: String(ci.pipeline?.number || ''),
        run_attempt: Number(ci.pipeline?.attempt || 0),
        actor: ci.actor || ci.pipeline?.triggerLogin || '',
        repo_code_owner: ci.repository?.code_owner || '',
        repo_owner: ci.repository?.owner || ci.project?.username || '',
        repo_name: ci.repository?.name || ci.project?.reponame || '',
        repo_id: String(ci.repository?.id || ''),
        repo_url: ci.repository?.url || ci.pipeline?.repoUrl || '',
        pr_number: Number(ci.pullRequest?.number || 0),
        pr_title: ci.pullRequest?.title || '',
        pr_status: ci.pullRequest?.status || '',
        pr_by: ci.pullRequest?.user || ci.actor || '',
        pr_base_branch: ci.pullRequest?.baseRef || ci.pullRequest?.targetBranch || '',
        pr_head_branch: ci.pullRequest?.headRef || ci.pullRequest?.sourceBranch || '',
        system: system || {}
    };
}
async function insert(client, options) {
    try {
        await client.insert(TABLE, [build(options)]);
    }
    catch {
    }
}
module.exports = {
    TABLE: TABLE,
    build: build,
    insert: insert
};
