// @postman/pm-bin-linux-x64/bin/lib/framework/clickhouse/client.js
const { create: createApi } = require('./api'), { validateApiInterface } = require('./types'), console = require('../../framework/console'), config = require('./config'), DEFAULTS = {
    maxEvents: config.MAX_EVENTS,
    maxBytes: config.MAX_BYTES,
    flushDelayMs: config.FLUSH_DELAY_MS
};
class Client {
    constructor(options = {}) {
        const { api, ...configOptions } = options;
        this.config = { ...DEFAULTS, ...configOptions };
        if (api) {
            validateApiInterface(api);
            this.api = api;
        }
        else {
            this.api = createApi();
        }
        this.buffers = new Map();
        this.timers = new Map();
        this._authCheck = null;
        this.debug = Boolean(process.env[config.DEBUG_ENV_VAR]);
        this._log('Client initialized', this.config);
    }
    _log(...args) {
        if (this.debug) {
            console.log(config.LOG_PREFIX, ...args);
        }
    }
    _error(...args) {
        if (this.debug) {
            console.error(config.LOG_PREFIX, ...args);
        }
    }
    checkAuth() {
        if (!this._authCheck) {
            this._log('Initializing auth...');
            this._authCheck = this.api.initAuth().then((result) => {
                this._log('Auth result:', result ? 'authenticated' : 'not authenticated');
                return result;
            });
        }
        return this._authCheck;
    }
    async flush(table) {
        if (this.timers.has(table)) {
            clearTimeout(this.timers.get(table));
            this.timers.delete(table);
        }
        const buf = this.buffers.get(table), hasAuth = await this.checkAuth();
        if (!buf || buf.events.length === 0) {
            return { success: true, count: 0 };
        }
        const events = buf.events;
        buf.events = [];
        buf.sizeBytes = 0;
        if (!hasAuth) {
            return { success: true, count: events.length, skipped: true };
        }
        this._log(`Flushing ${events.length} events to ${table}...`);
        try {
            const result = await this.api.insert(table, events);
            if (result.success) {
                this._log(`Flushed ${events.length} events to ${table}`);
            }
            else if (!result.skipped) {
                this._error(`Flush failed for ${table}:`, result.error);
            }
            return { success: result.success, count: events.length, error: result.error };
        }
        catch (err) {
            this._error(`Flush error for ${table}:`, err.message);
            return { success: false, count: events.length, error: err.message };
        }
    }
    async flushAll() {
        const tables = [...this.buffers.keys()];
        await Promise.all(tables.map((t) => { return this.flush(t); }));
    }
    buffer(table, event) {
        try {
            if (!this.buffers.has(table)) {
                this.buffers.set(table, { events: [], sizeBytes: 0 });
            }
            const buf = this.buffers.get(table);
            buf.events.push(event);
            buf.sizeBytes += Buffer.byteLength(JSON.stringify(event), 'utf8');
            if (buf.events.length >= this.config.maxEvents || buf.sizeBytes >= this.config.maxBytes) {
                this._log(`Threshold hit for ${table}: ${buf.events.length} events, ${buf.sizeBytes} bytes`);
                this.flush(table);
                return;
            }
            if (!this.timers.has(table)) {
                this._log(`Scheduling flush for ${table} in ${this.config.flushDelayMs}ms`);
                this.timers.set(table, setTimeout(() => { return this.flush(table); }, this.config.flushDelayMs));
            }
        }
        catch (err) {
            this._error('Buffer error:', err.message);
        }
    }
    async insert(table, rows) {
        if (!await this.checkAuth()) {
            this._log(`Insert skipped for ${table}: no auth`);
            return { success: true, skipped: true };
        }
        this._log(`Inserting ${rows.length} rows to ${table}`);
        try {
            const result = await this.api.insert(table, rows);
            if (!result.success && !result.skipped) {
                this._error(`Insert failed for ${table}:`, result.error);
            }
            return result;
        }
        catch (err) {
            this._error(`Insert error for ${table}:`, err.message);
            return { success: false, error: err.message };
        }
    }
    destroy() {
        this._log('Destroying client...');
        this.timers.forEach((t) => { return clearTimeout(t); });
        this.timers.clear();
        this.buffers.clear();
        this._authCheck = null;
    }
}
let instance = null;
function createClient(options) {
    return new Client(options);
}
function getClient() {
    if (!instance) {
        instance = new Client();
    }
    return instance;
}
module.exports = {
    Client: Client,
    createClient: createClient,
    getClient: getClient
};
