// @postman/pm-bin-linux-x64/bin/lib/run/analytics.js
const { v4: uuidv4 } = require('uuid'), { createClient, context } = require('../framework/clickhouse'), { extractServiceInfo, extractEndpointPattern, generateEndpointId } = require('../framework/clickhouse/event-helpers'), ExecutionContext = require('../framework/execution-context'), TABLE = 'postman_event_log_v3';
function buildCIContext() {
    const ctx = ExecutionContext.getAll(), ci = ctx?.ci || {}, system = ctx?.system || {}, platform = ci.provider || '', branch = ci.branch || '', commitSha = ci.commit?.sha || '', repoName = ci.repository?.name || ci.project?.reponame || '', repoUrl = ci.repository?.url || ci.pipeline?.repoUrl || '', prNumber = ci.pullRequest?.number ? Number(ci.pullRequest.number) : 0;
    return {
        platform: platform,
        branch: branch,
        commitSha: commitSha,
        repoName: repoName,
        repoUrl: repoUrl,
        prNumber: prNumber,
        workflow: {
            name: ci.workflow?.name || '',
            id: ci.workflow?.id || ci.pipeline?.id || ''
        },
        job: {
            name: ci.job?.name || '',
            id: ci.job?.id || ''
        },
        pipeline: {
            id: ci.pipeline?.id || '',
            number: ci.pipeline?.number || '',
            attempt: ci.pipeline?.attempt || '',
            url: ci.pipeline?.url || ''
        },
        actor: ci.actor || ci.pipeline?.triggerLogin || '',
        repository: {
            name: repoName,
            owner: ci.repository?.owner || ci.project?.username || '',
            id: ci.repository?.id || '',
            url: repoUrl
        },
        pull_request: ci.pullRequest ? {
            number: prNumber,
            title: ci.pullRequest.title || '',
            base_ref: ci.pullRequest.baseRef || '',
            head_ref: ci.pullRequest.headRef || ''
        } : undefined,
        system: system || undefined
    };
}
function buildBaseEvent(state, eventType) {
    return {
        event_type: eventType,
        event_span: state.eventSpan,
        event_index: state.eventIndex++,
        timestamp: new Date().toISOString(),
        collection_id: state.collectionId,
        workspace_id: state.workspaceId,
        environment_id: state.environmentId,
        system_environment: state.executionEnv || 'default',
        executor_type: 'cli',
        service_name: '',
        service_base_url: '',
        endpoint_pattern: '',
        endpoint_method: '',
        endpoint_id: '',
        ci_platform: state.ci.platform,
        pr_number: state.ci.prNumber,
        git_branch: state.ci.branch,
        git_commit_sha: state.ci.commitSha,
        repo_name: state.ci.repoName,
        repo_url: state.ci.repoUrl
    };
}
async function setup(emitter, options) {
    if (!options.reportEvents) {
        return {
            flush: () => { return Promise.resolve(); }
        };
    }
    const client = createClient(), cliVersion = require('../../package.json').version, hasAuth = await client.checkAuth();
    if (!hasAuth) {
        return {
            flush: () => { return Promise.resolve(); }
        };
    }
    const hasRealEnvironment = options.environment &&
        (options.environment.name || (options.environment.values && options.environment.values.count?.() > 0)), state = {
        eventSpan: uuidv4(),
        eventIndex: 0,
        collectionId: options.collection?.id || '',
        workspaceId: options.workspaceId || '',
        environmentId: hasRealEnvironment ? (options.environment?.id || '') : '',
        executionEnv: options.executionEnvironment || '',
        ci: buildCIContext()
    }, pendingResponses = new Map(), requestAssertions = new Map();
    context.insert(client, {
        eventSpan: state.eventSpan,
        executorType: 'cli',
        cliVersion: cliVersion,
        workspaceId: state.workspaceId,
        collectionId: state.collectionId,
        environmentId: state.environmentId,
        systemEnvironment: state.executionEnv
    }).catch(() => {
    });
    emitter.on('start', (_err, args) => {
        try {
            const event = buildBaseEvent(state, 'run_started');
            event.event_data = {
                collection_name: options.collection?.name || '',
                environment_id: state.environmentId || null,
                environment_name: hasRealEnvironment ? (options.environment?.name || null) : null,
                execution_environment: state.executionEnv || null,
                executor: {
                    arch: process.arch,
                    name: 'postman-cli',
                    platform: process.platform,
                    version: cliVersion
                },
                execution_plan: {
                    total_iterations: String(args.cursor?.cycles || 0),
                    total_requests: String(args.cursor?.length || 0)
                },
                ci_context: state.ci
            };
            client.buffer(TABLE, event);
        }
        catch {
        }
    });
    emitter.on('response', (_err, args) => {
        try {
            const key = `${args.cursor?.iteration || 0}:${args.cursor?.position || 0}`, svc = extractServiceInfo(args.request?.url), endpointPattern = extractEndpointPattern(args.request?.url), method = args.request?.method || 'GET', endpointId = generateEndpointId(svc.serviceName || '', method, endpointPattern);
            pendingResponses.set(key, {
                timestamp: new Date().toISOString(),
                serviceName: svc.serviceName,
                baseUrl: svc.baseUrl,
                endpointPattern: endpointPattern || '',
                endpointMethod: method,
                endpointId: endpointId || '',
                item: { id: args.item?.id || '', name: args.item?.name || '' },
                request: {
                    method: method,
                    url: args.request?.url?.toString() || ''
                },
                response: {
                    code: args.response?.code || 0,
                    status: args.response?.status || '',
                    time_ms: args.response?.responseTime || 0,
                    size_bytes: args.response?.responseSize || 0
                },
                position: {
                    iteration: args.cursor?.iteration || 0,
                    request_index: args.cursor?.position || 0
                }
            });
        }
        catch {
        }
    });
    emitter.on('assertion', (err, args) => {
        try {
            const key = `${args.cursor?.iteration || 0}:${args.cursor?.position || 0}`;
            let data = requestAssertions.get(key);
            if (!data) {
                const pattern = extractEndpointPattern(args.item?.request?.url), method = args.item?.request?.method || '', resp = pendingResponses.get(key), endpointId = generateEndpointId(resp?.serviceName || '', method, pattern);
                data = {
                    endpointPattern: pattern || '',
                    endpointMethod: method,
                    endpointId: endpointId || '',
                    total: 0,
                    passed: 0,
                    failed: 0,
                    skipped: 0
                };
                requestAssertions.set(key, data);
            }
            data.total++;
            if (args.skipped) {
                data.skipped++;
            }
            else if (err) {
                data.failed++;
            }
            else {
                data.passed++;
            }
        }
        catch {
        }
    });
    emitter.on('item', (_err, args) => {
        try {
            const key = `${args.cursor?.iteration || 0}:${args.cursor?.position || 0}`, resp = pendingResponses.get(key), asserts = requestAssertions.get(key);
            if (!resp) {
                return;
            }
            pendingResponses.delete(key);
            requestAssertions.delete(key);
            const event = buildBaseEvent(state, 'request_executed');
            event.timestamp = resp.timestamp;
            event.service_name = resp.serviceName || '';
            event.service_base_url = resp.baseUrl || '';
            event.endpoint_pattern = asserts?.endpointPattern || resp.endpointPattern || '';
            event.endpoint_method = asserts?.endpointMethod || resp.endpointMethod || '';
            event.endpoint_id = asserts?.endpointId || resp.endpointId || '';
            event.event_data = {
                item: resp.item,
                request: resp.request,
                response: resp.response,
                position: resp.position,
                execution_environment: state.executionEnv || null,
                ci_context: state.ci,
                tests: asserts ? {
                    total_assertions: asserts.total,
                    passed_assertions: asserts.passed,
                    failed_assertions: asserts.failed,
                    skipped_assertions: asserts.skipped,
                    has_failures: asserts.failed > 0,
                    has_exceptions: false
                } : undefined
            };
            client.buffer(TABLE, event);
        }
        catch {
        }
    });
    emitter.on('done', (_err, args) => {
        try {
            for (const [key, resp] of pendingResponses) {
                const asserts = requestAssertions.get(key), event = buildBaseEvent(state, 'request_executed');
                event.timestamp = resp.timestamp;
                event.service_name = resp.serviceName || '';
                event.service_base_url = resp.baseUrl || '';
                event.endpoint_pattern = asserts?.endpointPattern || resp.endpointPattern || '';
                event.endpoint_method = asserts?.endpointMethod || resp.endpointMethod || '';
                event.endpoint_id = asserts?.endpointId || resp.endpointId || '';
                event.event_data = {
                    item: resp.item,
                    request: resp.request,
                    response: resp.response,
                    position: resp.position,
                    execution_environment: state.executionEnv || null,
                    ci_context: state.ci,
                    tests: asserts ? {
                        total_assertions: asserts.total,
                        passed_assertions: asserts.passed,
                        failed_assertions: asserts.failed,
                        skipped_assertions: asserts.skipped,
                        has_failures: asserts.failed > 0,
                        has_exceptions: false
                    } : undefined
                };
                client.buffer(TABLE, event);
            }
            pendingResponses.clear();
            requestAssertions.clear();
            const stats = args.run?.stats || {}, timings = args.run?.timings || {}, event = buildBaseEvent(state, 'run_completed');
            event.event_data = {
                summary: {
                    run_passed: (stats.assertions?.failed || 0) === 0 &&
                        (stats.scripts?.failed || 0) === 0 &&
                        (stats.requests?.failed || 0) === 0,
                    total_duration_ms: (timings.completed || 0) - (timings.started || 0),
                    total_requests: stats.requests?.total || 0,
                    failed_requests: stats.requests?.failed || 0,
                    total_assertions: stats.assertions?.total || 0,
                    failed_assertions: stats.assertions?.failed || 0,
                    total_scripts: stats.scripts?.total || 0,
                    failed_scripts: stats.scripts?.failed || 0
                },
                execution_environment: state.executionEnv || null,
                ci_context: state.ci,
                stats: stats,
                timings: timings
            };
            client.buffer(TABLE, event);
        }
        catch {
        }
    });
    emitter.on('publish', (_err, args) => {
        try {
            if (!args?.workspaceId) {
                return;
            }
            state.workspaceId = args.workspaceId;
            const event = buildBaseEvent(state, 'run_published');
            event.event_data = {
                execution_environment: state.executionEnv || null,
                ci_context: state.ci,
                run_id: args.runId || ''
            };
            client.buffer(TABLE, event);
        }
        catch {
        }
    });
    emitter.on('error', (err) => {
        try {
            const event = buildBaseEvent(state, 'run_error');
            event.event_data = {
                error: {
                    message: err?.message || 'Unknown error',
                    code: err?.code || null,
                    stack: err?.stack || null
                },
                execution_environment: state.executionEnv || null,
                ci_context: state.ci
            };
            client.buffer(TABLE, event);
        }
        catch {
        }
    });
    return {
        flush: () => { return client.flush(TABLE); }
    };
}
module.exports = { setup };
