// @postman/pm-bin-linux-x64/bin/lib/run/cloud-file.js
const axios = require('axios'), util = require('../util'), _ = require('lodash'), USER_FACING_ERROR_MESSAGES = {
    'urn:error:file:planLimitExceeded': 'File couldn’t be retrieved from Postman as you have reached the limit for data file retrievals. ' +
        'To continue accessing your uploaded files, upgrade your plan by visiting https://www.postman.com/pricing/'
};
function _createFriendlyError(status, errorData = {}) {
    const err = new Error(USER_FACING_ERROR_MESSAGES[errorData.type] ||
        errorData.detail || errorData.title || 'Something went wrong while fetching the file');
    err.status = status;
    err.type = errorData.type;
    err.name = 'Axios Error';
    return err;
}
function _handleAxiosError(host, path, error) {
    const loggedError = {
        host: host,
        path: path,
        message: `Unable to send request server: ${error.message}`,
        code: error.code,
        status: _.get(error, 'response.status')
    }, responseBody = _.get(error, 'response.data'), parsedResponseError = Buffer.isBuffer(responseBody) ? JSON.parse(responseBody.toString()) : responseBody;
    if (error.response) {
        loggedError.message = `Invalid response from server: ${error.message}`;
    }
    else if (error.request) {
        loggedError.message = `No response from server": ${error.message}`;
    }
    logger.error(JSON.stringify(loggedError));
    throw _createFriendlyError(_.get(error, 'response.status', 500), parsedResponseError);
}
async function getCloudFileMeta({ fileId, postmanApiKey, accessToken }) {
    if (!fileId) {
        throw new Error('FileId is missing');
    }
    if (!postmanApiKey && !accessToken) {
        throw new Error('Postman credentials is missing');
    }
    const requestHeaders = {
        accept: 'application/vnd.postman.v2+json',
        'User-Agent': util.userAgent,
        'x-pstmn-req-service': 'file'
    }, axiosConfig = {
        method: 'get',
        baseURL: util.POSTMAN_GATEWAY_BASE_URL(),
        url: `/files/${fileId}`,
        headers: requestHeaders,
        withCredentials: true
    };
    if (postmanApiKey) {
        requestHeaders['x-api-key'] = postmanApiKey;
    }
    else {
        requestHeaders['x-access-token'] = accessToken;
    }
    const response = await axios(axiosConfig).catch((err) => {
        _handleAxiosError(axiosConfig.baseURL, axiosConfig.url, err);
    });
    return response.data && response.data.data;
}
async function fetchBlobContents({ fileId, blobId, postmanApiKey, accessToken }) {
    if (!fileId || !blobId) {
        throw new Error('Something went wrong while fetching file contents');
    }
    if (!postmanApiKey && !accessToken) {
        throw new Error('Postman credentials are missing');
    }
    const requestHeaders = {
        accept: 'application/vnd.postman.v2+json',
        'x-pstmn-req-service': 'file',
        'User-Agent': util.userAgent
    }, axiosConfig = {
        method: 'get',
        url: `blobs/${blobId}`,
        baseURL: util.POSTMAN_GATEWAY_BASE_URL(),
        headers: requestHeaders,
        withCredentials: true,
        responseType: 'arraybuffer'
    };
    if (postmanApiKey) {
        requestHeaders['x-api-key'] = postmanApiKey;
    }
    else {
        requestHeaders['x-access-token'] = accessToken;
    }
    const response = await axios(axiosConfig).catch((err) => {
        _handleAxiosError(axiosConfig.baseURL, axiosConfig.url, err);
    });
    return response.data;
}
module.exports = {
    getCloudFileMeta,
    fetchBlobContents
};
