// @postman/pm-bin-linux-x64/bin/lib/run/index.js
var _ = require('lodash'), async = require('async'), sdk = require('postman-collection'), runtime = require('postman-runtime'), request = require('postman-request'), fs = require('fs'), path = require('path'), EventEmitter = require('events'), PackageResolver = require('./package-resolver'), RequestResolver = require('./request-resolver'), print = require('../framework/print'), colors = require('colors/safe'), SecureFS = require('./secure-fs'), RunSummary = require('./summary'), { setup: setupAnalytics } = require('./analytics'), getOptions = require('./options'), exportFile = require('./export-file'), util = require('../util'), console = require('../framework/console'), runUploadWrapper = require('../run-upload/index').runUploadWrapper, getUserFacingError = require('./get-user-facing-error'), RunContext = require('./run-context'), { executeUnifiedRun } = require('./unified/run'), runtimeEvents = {
    beforeIteration: [],
    beforeItem: ['item'],
    beforePrerequest: ['events', 'item'],
    prerequest: ['executions', 'item'],
    beforeRequest: ['request', 'item'],
    request: ['response', 'request', 'item', 'cookies', 'history'],
    response: ['response', 'request', 'item', 'cookies', 'history'],
    beforeTest: ['events', 'item'],
    test: ['executions', 'item'],
    item: ['item'],
    iteration: [],
    beforeScript: ['script', 'event', 'item'],
    script: ['execution', 'script', 'event', 'item']
}, builtInReporters = {
    cli: require('../reporters/cli'),
    json: require('../reporters/json'),
    junit: require('../reporters/junit'),
    html: require('../reporters/html')
}, MULTIENTRY_WITH_ORDER_LOOKUP_STRATEGY = 'followOrder';
function resolveWorkspaceIdForCollection(options) {
    return new Promise((resolve) => {
        if (options.workspaceId) {
            return resolve();
        }
        const workingDir = options.workingDir || process.cwd(), configPath = path.join(workingDir, '.postman', 'config.json');
        fs.readFile(configPath, 'utf8', (err, data) => {
            if (err) {
                return resolve();
            }
            try {
                const parsed = JSON.parse(data), workspaceId = _.get(parsed, 'workspace.id') || _.get(parsed, 'workspace_id');
                if (workspaceId && typeof workspaceId === 'string') {
                    options.workspaceId = workspaceId;
                }
            }
            catch (e) {
            }
            return resolve();
        });
    });
}
module.exports = function (options, callback) {
    (!callback && _.isFunction(options)) && ((callback = options),
        (options = {}));
    !_.isFunction(callback) && (callback = _.noop);
    var emitter = new EventEmitter(), runner = new runtime.Runner(), stopOnFailure, entrypoint;
    emitter.reporters = {};
    try {
        var requestedReporters = _.isString(options && options.reporters) ?
            [options.reporters] : (options && options.reporters);
        _.isArray(requestedReporters) && _.forEach(requestedReporters, function (reporterName) {
            if (!_.has(emitter.reporters, reporterName)) {
                const Reporter = builtInReporters[reporterName];
                Reporter && (emitter.reporters[reporterName] =
                    reporterDispatcher(emitter, reporterName, options || {}));
            }
        });
    }
    catch (e) {
    }
    function reporterDispatcher(emitter, reporterFormat, options) {
        const Reporter = builtInReporters[reporterFormat], reporterOptions = _.get(options, ['reporter', reporterFormat], {});
        return new Reporter(emitter, reporterOptions, options);
    }
    getOptions(options, async function (err, options) {
        if (err) {
            return callback(err);
        }
        if (!_.isObject(options.collection)) {
            return callback(new Error('expecting a collection to run'));
        }
        var sslClientCertList = options.sslClientCertList || [], cookieJar = options.cookieJar || request.jar();
        options.cookieJar = cookieJar;
        if (options.sslClientCert) {
            sslClientCertList.push({
                name: 'client-cert',
                matches: [sdk.UrlMatchPattern.MATCH_ALL_URLS],
                key: { src: options.sslClientKey },
                cert: { src: options.sslClientCert },
                passphrase: options.sslClientPassphrase
            });
        }
        if (_.isArray(options.bail)) {
            options.bail = _.transform(options.bail, function (result, value) {
                result[value] = true;
            }, {});
        }
        if (options.i) {
            entrypoint = { execute: options.i };
            _.isArray(entrypoint.execute) &&
                (entrypoint.lookupStrategy = MULTIENTRY_WITH_ORDER_LOOKUP_STRATEGY);
        }
        (typeof options.bail !== 'undefined' &&
            (options.bail === true || (_.isObject(options.bail) && options.bail.failure))) ?
            stopOnFailure = true : stopOnFailure = false;
        await resolveWorkspaceIdForCollection(options);
        emitter.summary = new RunSummary(emitter, options);
        emitter.exports = [];
        emitter.runner = runner;
        const analyticsCleanup = await setupAnalytics(emitter, options);
        util.removeCloudFilesFolder(options.workingDir);
        if (options.isExtensibleCollection) {
            const runContext = new RunContext(), reporters = _.isString(options.reporters) ? [options.reporters] : options.reporters, unsupportedReporters = _.isArray(reporters) ?
                reporters.filter((r) => { return r !== 'cli'; }) : [];
            if (unsupportedReporters.length) {
                const reporterValues = unsupportedReporters.map((v) => { return `"${v}"`; }).join(', '), err = new Error(`Reporter ${reporterValues} is not supported for multi-protocol collections.`);
                err.friendly = 'Only the "cli" reporter is supported for multi-protocol collections.';
                return callback(err);
            }
            emitter.runContext = runContext;
            !emitter.reporters && (emitter.reporters = {});
            _.isArray(reporters) && _.forEach(reporters, function (reporterName) {
                if (_.has(emitter.reporters, reporterName)) {
                    return;
                }
                var Reporter = builtInReporters[reporterName];
                try {
                    Reporter && (emitter.reporters[reporterName] = reporterDispatcher(emitter, reporterName, options));
                }
                catch (error) {
                    console.warn('postman: could not load "' + reporterName + '" reporter');
                    console.warn(error);
                }
            });
            await executeUnifiedRun(options, emitter, runContext, async function (runErr, summary) {
                if (options.exportCookieJar && options.cookieJar) {
                    await new Promise(function (resolve) {
                        exportFile({
                            name: 'cookie-jar',
                            default: 'postman-cookie-jar.json',
                            path: options.exportCookieJar,
                            content: options.cookieJar.toJSON()
                        }, function () { resolve(); });
                    });
                }
                callback(runErr, summary);
            });
            return;
        }
        const JSONReporterOptions = _.get(options, 'reporter.json', {});
        if (!options.silent) {
            print.lf(colors.grey('Running your collection...'));
        }
        runner.run(options.collection, {
            stopOnFailure: stopOnFailure,
            abortOnFailure: options.abortOnFailure,
            abortOnError: _.get(options, 'bail.folder'),
            iterationCount: options.iterationCount,
            environment: options.environment,
            globals: options.globals,
            entrypoint: entrypoint,
            data: options.iterationData,
            delay: {
                item: options.delayRequest
            },
            timeout: {
                global: options.timeout || 0,
                request: options.timeoutRequest || 0,
                script: options.timeoutScript || 0
            },
            fileResolver: new SecureFS({
                workingDir: options.workingDir,
                insecureFileRead: options.insecureFileRead,
                postmanApiKey: options.postmanApiKey,
                accessToken: options.accessToken
            }),
            ...(!JSONReporterOptions.structure && {
                script: {
                    packageResolver: new PackageResolver({ postmanApiKey: options.postmanApiKey,
                        accessToken: options.accessToken }).resolvePackages,
                    requestResolver: new RequestResolver({
                        postmanApiKey: options.postmanApiKey,
                        accessToken: options.accessToken
                    }).resolveRequest
                }
            }),
            requester: {
                useWhatWGUrlParser: true,
                cookieJar: cookieJar,
                followRedirects: _.has(options, 'ignoreRedirects') ? !options.ignoreRedirects : undefined,
                strictSSL: _.has(options, 'insecure') ? !options.insecure : undefined,
                timings: Boolean(options.verbose),
                extendedRootCA: options.sslExtraCaCerts,
                agents: _.isObject(options.requestAgents) ? options.requestAgents : undefined
            },
            certificates: sslClientCertList.length && new sdk.CertificateList({}, sslClientCertList)
        }, function (err, run) {
            if (err) {
                return callback(getUserFacingError(err));
            }
            var callbacks = {}, reporters = _.isString(options.reporters) ? [options.reporters] : options.reporters, legacyAssertionIndices = {};
            _.forEach(runtimeEvents, function (definition, eventName) {
                callbacks[eventName] = function (err, cursor) {
                    var args = arguments, obj = { cursor };
                    _.forEach(definition, function (key, index) {
                        obj[key] = args[index + 2];
                    });
                    args = [eventName, err, obj];
                    emitter.emit.apply(emitter, args);
                };
            });
            _.assignIn(callbacks, {
                start(err, cursor) {
                    emitter.emit('start', err, {
                        cursor,
                        run
                    });
                },
                console(cursor, level) {
                    emitter.emit('console', null, {
                        cursor: cursor,
                        level: level,
                        messages: _.slice(arguments, 2)
                    });
                },
                exception(cursor, err) {
                    emitter.emit('exception', null, {
                        cursor: cursor,
                        error: err
                    });
                },
                assertion(cursor, assertions) {
                    _.forEach(assertions, function (assertion) {
                        var errorName = _.get(assertion, 'error.name', 'AssertionError');
                        !assertion && (assertion = {});
                        assertion.index && (legacyAssertionIndices[cursor.ref] = assertion.index);
                        emitter.emit('assertion', (assertion.passed ? null : {
                            name: errorName,
                            index: assertion.index,
                            test: assertion.name,
                            message: _.get(assertion, 'error.message', assertion.name || ''),
                            stack: errorName + ': ' + _.get(assertion, 'error.message', '') + '\n' +
                                '   at Object.eval sandbox-script.js:' + (assertion.index + 1) + ':' +
                                ((cursor && cursor.position || 0) + 1) + ')'
                        }), {
                            cursor: cursor,
                            assertion: assertion.name,
                            skipped: assertion.skipped,
                            error: assertion.error,
                            item: run.resolveCursor(cursor)
                        });
                    });
                },
                done(err, cursor) {
                    if (err) {
                        emitter.emit('done', err, emitter.summary);
                        callback(err, emitter.summary);
                        return;
                    }
                    emitter.emit('beforeDone', null, {
                        cursor: cursor,
                        summary: emitter.summary
                    });
                    _.forEach(['environment', 'globals', 'collection', 'cookie-jar'], function (item) {
                        var path = _.get(options, _.camelCase(`export-${item}`));
                        path && emitter.exports.push({
                            name: item,
                            default: `postman-${item}.json`,
                            path: path,
                            content: item === 'cookie-jar' ?
                                cookieJar.toJSON() :
                                _(emitter.summary[item].toJSON())
                                    .defaults({
                                    name: item
                                })
                                    .merge({
                                    _postman_variable_scope: item,
                                    _postman_exported_at: (new Date()).toISOString(),
                                    _postman_exported_using: util.userAgent
                                })
                                    .value()
                        });
                    });
                    async.series([
                        (next) => {
                            async.each(emitter.exports, exportFile, next);
                        },
                        (next) => {
                            emitter.emit('done', err, emitter.summary);
                            return next();
                        },
                        (next) => {
                            if (options.skipRunUpload) {
                                return next();
                            }
                            runUploadWrapper(emitter.summary, options, (err, workspaceId, runId) => {
                                if (err) {
                                    emitter.emit('publish', err);
                                }
                                else {
                                    emitter.emit('publish', null, {
                                        workspaceId,
                                        runId
                                    });
                                }
                                return next();
                            });
                        }
                    ], async (err) => {
                        if (analyticsCleanup && analyticsCleanup.flush) {
                            await analyticsCleanup.flush();
                        }
                        callback(err, emitter.summary);
                    });
                    util.removeCloudFilesFolder(options.workingDir);
                }
            });
            emitter.on('script', function (err, o) {
                o && o.event && emitter.emit(o.event.listen + 'Script', err, o);
            });
            emitter.on('beforeScript', function (err, o) {
                o && o.event && emitter.emit(_.camelCase('before-' + o.event.listen + 'Script'), err, o);
            });
            !emitter.reporters && (emitter.reporters = {});
            _.isArray(reporters) && _.forEach(reporters, function (reporterName) {
                if (_.has(emitter.reporters, reporterName)) {
                    return;
                }
                const Reporter = builtInReporters[reporterName];
                try {
                    Reporter && (emitter.reporters[reporterName] = reporterDispatcher(emitter, reporterName, options));
                }
                catch (error) {
                    console.warn(`postman: could not load "${reporterName}" reporter`);
                    console.warn(error);
                }
            });
            setImmediate(function () {
                run.start(callbacks);
            });
        });
    });
    return emitter;
};
