// @postman/pm-bin-linux-x64/bin/lib/run/options.js
var _ = require('lodash'), fs = require('fs'), async = require('async'), Collection = require('postman-collection').Collection, VariableScope = require('postman-collection').VariableScope, CookieJar = require('@postman/tough-cookie').CookieJar, transformer = require('postman-collection-transformer'), liquidJSON = require('liquid-json'), parseCsv = require('csv-parse'), requestFactory = require('@postman/platform-req/lib/request'), PostmanResource = require('../api').PostmanResource, console = require('../framework/console'), featureFlags = require('../framework/feature-flags'), util = require('../util'), { loadV3AsExtensibleCollection } = require('./v3-loader'), { isYamlFile } = require('../postman-entities/file-utils'), { readEnvironmentFile } = require('../postman-entities/environment/environment'), EXTENSIBLE_COLLECTION_OBJECT_ID_REGEX = /^[0-9a-f]{24}$/i, COLLECTION_LOAD_ERROR_MESSAGE = 'collection could not be loaded', ITERATION_DATA_LOAD_ERROR_MESSAGE = 'iteration data could not be loaded', LOAD_ERROR_MESSAGE = 'could not load ', COLLECTION_TRANSFORMER_OPTION = { inputVersion: '1.0.0', outputVersion: '2.1.0' }, isExtensibleCollectionId = function (value) {
    if (!_.isString(value)) {
        return false;
    }
    if (EXTENSIBLE_COLLECTION_OBJECT_ID_REGEX.test(value)) {
        return true;
    }
    return false;
}, isExtensibleCollection = function (collection) {
    if (!_.isObject(collection)) {
        return false;
    }
    if (isExtensibleCollectionId(collection.id)) {
        return true;
    }
    if (collection.type === 'collection' && _.isArray(collection.children)) {
        return true;
    }
    return false;
}, spreadItems = function (children) {
    if (!Array.isArray(children)) {
        return;
    }
    children.forEach(function (child) {
        if (child.item) {
            Object.assign(child, child.item);
            delete child.item;
        }
        spreadItems(child.children);
    });
}, fetchExtensibleCollectionFromCloud = function (collectionId, options, callback) {
    var gatewayBaseUrl = util.POSTMAN_GATEWAY_BASE_URL(), url = gatewayBaseUrl + '/collection-exports/' + collectionId, headers = {
        'User-Agent': util.userAgent
    }, client = requestFactory.defaults({
        timeout: 300000,
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: { GET: { enabled: true } }
        }
    });
    if (options.postmanApiKey) {
        headers['X-Api-Key'] = options.postmanApiKey;
    }
    else if (options.accessToken) {
        headers['x-access-token'] = options.accessToken;
    }
    headers['x-pstmn-req-service'] = 'collection-api';
    client.get({ url: url, headers: headers }, function (err, response, body) {
        if (err) {
            return callback(new Error(COLLECTION_LOAD_ERROR_MESSAGE +
                '\n  unable to fetch extensible collection from "' + url + '"' +
                '\n  ' + (err.message || err)));
        }
        try {
            _.isString(body) && (body = liquidJSON.parse(body.trim()));
        }
        catch (e) {
            return callback(new Error(COLLECTION_LOAD_ERROR_MESSAGE +
                '\n  the response from "' + url + '" did not contain valid JSON data'));
        }
        if (!(/2../).test(response.statusCode)) {
            return callback(new Error(COLLECTION_LOAD_ERROR_MESSAGE +
                '\n  failed to fetch extensible collection (HTTP ' + response.statusCode + ')'));
        }
        var raw = (body && body.data) || body;
        if (!raw || !raw.collection) {
            return callback(new Error(COLLECTION_LOAD_ERROR_MESSAGE +
                '\n  the response from "' + url + '" did not contain a valid collection'));
        }
        if (raw.children) {
            raw.collection.children = raw.children;
        }
        spreadItems(raw.collection.children);
        raw.collection.type = 'collection';
        return callback(null, raw.collection);
    });
}, createPostmanResources = function (options, callback) {
    return async.mapValuesSeries(_.pick(options, ['collection', 'environment']), (value, type, cb) => {
        if (!util.isPostmanResource(value)) {
            return cb(null);
        }
        if ((/https?.*apikey=.*/).test(value)) {
            let urlObj = new URL(value), apikey = urlObj.searchParams.get('apikey');
            urlObj.searchParams.delete('apikey');
            if (util.isPublicAPIUrl(value)) {
                return cb(null, new PostmanResource(type, util.extractResourceIDFromPublicAPIUrl(value), apikey));
            }
            return cb(null, new PostmanResource(type, urlObj.href, apikey));
        }
        return util.extractPostmanApiKey(options.postmanApiKey, (err, apikey) => {
            if (err) {
                return cb(err);
            }
            options.postmanApiKey = apikey;
            return cb(null, new PostmanResource(type, value, apikey, options.accessToken));
        });
    }, callback);
}, extractModel = function (source, type) {
    source = source[type] || source;
    if (!_.isObject(source)) {
        return undefined;
    }
    !source.values && _.isObject(source[type]) && (source = source[type]);
    return source;
}, externalLoader = function (type, location, cb) {
    if (location instanceof PostmanResource) {
        return location.get(cb);
    }
    if (_.isString(location) && isYamlFile(location)) {
        return readEnvironmentFile(location).then(function (data) {
            return cb(null, extractModel(data, type));
        }).catch(function (err) {
            return cb(err);
        });
    }
    return _.isString(location) ? util.fetchJson(location, type, function (err, data) {
        if (err) {
            return cb(err);
        }
        return cb(null, extractModel(data, type));
    }) : cb(null, extractModel(location, type));
}, processCollection = function (collection, callback) {
    if (util.isV1Collection(collection)) {
        console.warn('Warning: Postman CLI deprecates support for the v1 collection format');
        console.warn('  Use the Postman Native app to export collections in the v2 format\n');
        return transformer.convert(collection, COLLECTION_TRANSFORMER_OPTION, callback);
    }
    callback(null, collection);
}, loadScopes = function (type, value, callback) {
    var done = function (err, scope) {
        if (err) {
            return callback(new Error(LOAD_ERROR_MESSAGE + `${type}\n  ${err.message || err}`));
        }
        if (!_.isObject(scope)) {
            return done(new Error(LOAD_ERROR_MESSAGE + type));
        }
        callback(null, new VariableScope(VariableScope.isVariableScope(scope) ? scope.toJSON() : scope));
    };
    if (_.isString(value) || value instanceof PostmanResource) {
        return externalLoader(type, value, done);
    }
    return done(null, value);
}, csvAutoParse = function (value, context) {
    if (context.quoting) {
        return value;
    }
    if (util.isInt(value)) {
        return parseInt(value, 10);
    }
    if (util.isFloat(value)) {
        return parseFloat(value);
    }
    return value;
}, configLoaders = {
    collection: function (value, callback) {
        var done = function (err, collection) {
            if (err) {
                return callback(err);
            }
            if (!_.isObject(collection)) {
                return callback(new Error(COLLECTION_LOAD_ERROR_MESSAGE));
            }
            if (isExtensibleCollection(collection)) {
                return callback(null, collection);
            }
            collection = new Collection(Collection.isCollection(collection) ?
                collection.toJSON() : collection);
            callback(null, collection);
        };
        if (_.isString(value) || value instanceof PostmanResource) {
            return externalLoader('collection', value, function (err, data) {
                if (err) {
                    return done(new Error(COLLECTION_LOAD_ERROR_MESSAGE +
                        (err.help ? `\n  ${err.help}` : '') +
                        `\n  ${err.message || err}`));
                }
                if (!_.isObject(data)) {
                    return done(new Error(COLLECTION_LOAD_ERROR_MESSAGE));
                }
                if (isExtensibleCollection(data)) {
                    return done(null, data);
                }
                return processCollection(data, done);
            });
        }
        return processCollection(value, done);
    },
    environment: loadScopes.bind(this, 'environment'),
    globals: loadScopes.bind(this, 'globals'),
    iterationData: function (location, callback) {
        if (_.isArray(location)) {
            return callback(null, location);
        }
        util.fetch(location, function (err, data) {
            if (err) {
                return callback(new Error(ITERATION_DATA_LOAD_ERROR_MESSAGE + `\n  ${err.message || err}`));
            }
            async.waterfall([
                (cb) => {
                    try {
                        return cb(null, liquidJSON.parse(data.trim()));
                    }
                    catch (e) {
                        return cb(null, undefined);
                    }
                },
                (json, cb) => {
                    if (json) {
                        return cb(null, json);
                    }
                    parseCsv(data, {
                        columns: true,
                        escape: '"',
                        cast: csvAutoParse,
                        trim: true,
                        relax: true,
                        relax_column_count: true,
                        bom: true
                    }, cb);
                }
            ], (err, parsed) => {
                if (err) {
                    return callback(new Error(ITERATION_DATA_LOAD_ERROR_MESSAGE + `\n ${err.message || err}`));
                }
                callback(null, parsed);
            });
        });
    },
    sslClientCertList: function (location, callback) {
        if (Array.isArray(location)) {
            return callback(null, location);
        }
        if (typeof location !== 'string') {
            return callback(new Error('path for ssl client certificates list file must be a string'));
        }
        fs.readFile(location, function (err, value) {
            if (err) {
                return callback(new Error(`unable to read the ssl client certificates file "${location}"`));
            }
            try {
                value = liquidJSON.parse(value.toString(util.detectEncoding(value)).trim());
            }
            catch (e) {
                return callback(new Error(`the file at ${location} does not contain valid JSON data.`));
            }
            if (!Array.isArray(value)) {
                return callback(new Error('expected ssl client certificates list to be an array.'));
            }
            return callback(null, value);
        });
    },
    cookieJar: function (location, callback) {
        if (_.isObject(location) && location.constructor.name === 'CookieJar') {
            return callback(null, location);
        }
        if (typeof location !== 'string') {
            return callback(new Error('cookieJar must be a path to a JSON file or a CookieJar instance'));
        }
        fs.readFile(location, function (err, value) {
            if (err) {
                return callback(new Error(`unable to read the cookie jar file "${location}"`));
            }
            try {
                value = CookieJar.fromJSON(value.toString());
            }
            catch (e) {
                return callback(new Error(`the file at ${location} does not contain valid JSON data.`));
            }
            return callback(null, value);
        });
    },
    i: function (location, callback) {
        if (!location.length) {
            return callback();
        }
        const POSTMAN_UID_REGEX = /^([0-9A-Z]+)-([0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12})$/i;
        let parsedLocation = [];
        _.forEach(location, (value) => {
            let parsed = POSTMAN_UID_REGEX.exec(value);
            if (!parsed) {
                parsedLocation.push(value);
                return true;
            }
            parsedLocation.push(parsed[2]);
        });
        return callback(null, parsedLocation);
    }
};
module.exports = function (options, callback) {
    options.postmanVersion = util.version;
    options.workingDir = options.workingDir || process.cwd();
    options.insecureFileRead = Boolean(_.get(options, 'insecureFileRead', true));
    options.skipRunUpload = Boolean(_.get(options, 'skipRunUpload', false));
    !options.environment && (options.environment = {});
    !options.globals && (options.globals = {});
    let postmanResources = {};
    async.waterfall([
        (next) => {
            if (!isExtensibleCollectionId(options.collection)) {
                return next(null);
            }
            return util.extractPostmanApiKey(options.postmanApiKey, (err, apikey) => {
                if (err || (!apikey && !options.accessToken)) {
                    return next(null);
                }
                options.postmanApiKey = apikey;
                return fetchExtensibleCollectionFromCloud(options.collection, options, (ecErr, ecCollection) => {
                    if (!ecErr && ecCollection) {
                        options.collection = ecCollection;
                    }
                    else {
                        console.warn('Could not fetch extensible collection from cloud, ' +
                            'falling back to standard collection fetch.' + ecErr);
                    }
                    return next(null);
                });
            });
        },
        (next) => {
            if (!_.isString(options.collection)) {
                return next(null);
            }
            return loadV3AsExtensibleCollection(options.collection).then((ec) => {
                if (ec) {
                    options.collection = ec;
                }
                return next(null);
            }).catch(() => {
                return next(null);
            });
        },
        (next) => {
            return featureFlags.init(() => {
                return next(null);
            });
        },
        (next) => {
            return createPostmanResources(options, next);
        },
        (_postmanResources, next) => {
            postmanResources = _postmanResources;
            _.forEach(postmanResources, (postmanResource, type) => {
                postmanResource && (options[type] = postmanResource);
            });
            return async.mapValues(options, (value, name, cb) => {
                if (value && _.isFunction(configLoaders[name])) {
                    return configLoaders[name](value, cb);
                }
                return cb(null, value);
            }, next);
        }
    ], (err, result) => {
        if (err) {
            return callback(err);
        }
        result.isExtensibleCollection = isExtensibleCollection(result.collection);
        !_.isEmpty(options.globalVar) && _.forEach(options.globalVar, function (variable) {
            variable && (result.globals.set(variable.key, variable.value));
        });
        !_.isEmpty(options.envVar) && _.forEach(options.envVar, function (variable) {
            variable && (result.environment.set(variable.key, variable.value));
        });
        return callback(null, result, postmanResources);
    });
};
