// @postman/pm-bin-linux-x64/bin/lib/run/package-resolver.js
const { Packages } = require('../api'), util = require('../util'), print = require('../framework/print'), colors = require('colors/safe'), CACHE_SIZE = 100;
class PackageResolver {
    constructor({ postmanApiKey, accessToken, packageLoader, logger = null }) {
        this.cache = new Map();
        this.postmanApiKey = postmanApiKey;
        this.accessToken = accessToken;
        this.packageLoader = packageLoader;
        this.logger = logger;
        this.resolvePackages = this.resolvePackages.bind(this);
        this._cachePackage = this._cachePackage.bind(this);
        this._getPackageFromCache = this._getPackageFromCache.bind(this);
    }
    _cachePackage(packageObj) {
        if (this.cache.size >= CACHE_SIZE) {
            const lruElement = this.cache.keys().next().value;
            this.cache.delete(lruElement);
        }
        this.cache.set(packageObj.id, packageObj);
    }
    _getPackageFromCache(packageId) {
        const packageObj = this.cache.get(packageId);
        if (!packageObj) {
            return undefined;
        }
        this.cache.delete(packageId);
        this._cachePackage(packageObj);
        return packageObj;
    }
    async resolvePackages(input, callback) {
        const { packages } = input, packagesToLoad = [], loadedPackages = {}, postmanApiKey = util.getValidPostmanAPIKey({ postmanApiKey: this.postmanApiKey }), accessToken = this.accessToken;
        if (!packages || !Object.keys(packages).length) {
            return callback(null, {});
        }
        if (this.logger) {
            this.logger.info('Loading packages...');
        }
        else {
            print.print(colors.gray('  %s'), 'Loading packages...');
            print.wait();
        }
        Object.keys(packages).forEach((name) => {
            const cachedPackage = this._getPackageFromCache(packages[name].id);
            if (cachedPackage) {
                loadedPackages[name] = {
                    data: cachedPackage.sourceCode,
                    error: cachedPackage.error && cachedPackage.error.message
                };
            }
            else {
                packagesToLoad.push({ name: name, id: packages[name].id });
            }
        });
        try {
            if (packagesToLoad.length) {
                const resolvedPackages = this.packageLoader ?
                    await this.packageLoader.loadPackageContentsInBulk({ packages: packagesToLoad }) :
                    await Packages.loadPackageContentsInBulk({
                        packages: packagesToLoad, postmanApiKey: postmanApiKey, accessToken: accessToken
                    });
                resolvedPackages.forEach((resolvedPackage) => {
                    loadedPackages[resolvedPackage.name] = {
                        data: resolvedPackage.sourceCode,
                        error: resolvedPackage.error && resolvedPackage.error.message
                    };
                    this._cachePackage(resolvedPackage);
                });
            }
            const successFullyLoadedCount = Object.values(loadedPackages).filter((item) => {
                return !item.error;
            }).length, requestedPackageCount = Object.keys(packages).length;
            if (!this.logger) {
                print.lf(colors.gray(`[${successFullyLoadedCount}` +
                    `/${requestedPackageCount}]`), successFullyLoadedCount !== requestedPackageCount ? colors.red('[Error]') : '');
                print.unwait();
            }
            return callback(null, {
                ...loadedPackages
            });
        }
        catch (error) {
            if (this.logger) {
                this.logger.error('Error loading packages', { error });
            }
            else {
                print.lf(colors.red('[Error]'));
            }
            return callback(error);
        }
    }
}
module.exports = PackageResolver;
