// @postman/pm-bin-linux-x64/bin/lib/run/request-resolver.js
const { PostmanResource } = require('../api'), util = require('../util'), print = require('../framework/print'), colors = require('colors/safe'), CollectionTransformer = require('postman-collection-transformer'), CACHE_SIZE = 500;
function sanitizeRequestData(request) {
    request.uid = `${request.owner}-${request.id}`;
    request.id = request.uid;
    if (request.dataMode === 'raw' || request.dataMode === 'binary') {
        request.rawModeData = request.data;
    }
    if (request.dataMode === 'graphql') {
        request.graphqlModeData = request.data;
    }
}
function setDefaultFields(request) {
    if (!request.protocolProfileBehavior) {
        request.protocolProfileBehavior = {};
    }
    typeof request.protocolProfileBehavior.disableBodyPruning === 'undefined' &&
        (request.protocolProfileBehavior.disableBodyPruning = true);
}
function getRootFolderInAncestorTree(request) {
    let rootFolderId = null, { folder: currentFolder } = request;
    while (currentFolder) {
        rootFolderId = currentFolder.id;
        currentFolder = currentFolder.folder;
    }
    return rootFolderId;
}
function constructFoldersForRequest(folder, request) {
    const constructedFolders = [], initialFolderItem = folder;
    constructedFolders.push({
        ...initialFolderItem,
        id: folder.id,
        order: [request.uid || request.id]
    });
    let { folder: parentFolder } = initialFolderItem, currentFolderId = folder.id;
    while (parentFolder) {
        constructedFolders.push({
            ...parentFolder,
            id: parentFolder.id,
            folders_order: [currentFolderId]
        });
        currentFolderId = parentFolder.id;
        parentFolder = parentFolder.folder;
    }
    return constructedFolders;
}
function getRequestExecutionObject(request) {
    const constructedFolders = request.folder && constructFoldersForRequest(request.folder, request), isRequestDirectlyInsideCollection = !request.folder, partialCollectionItem = request.collection, parentFolderIdForRequest = getRootFolderInAncestorTree(request), requestExecutionObject = {
        ...partialCollectionItem,
        order: isRequestDirectlyInsideCollection ? [request.uid || request.id] : [],
        requests: [request],
        folders_order: parentFolderIdForRequest ? [parentFolderIdForRequest] : [],
        folders: constructedFolders
    };
    return requestExecutionObject;
}
function transformToV2Collection(collection) {
    return new Promise(((resolve, reject) => {
        CollectionTransformer.convert(collection, { inputVersion: '1.0.0', outputVersion: '2.0.0', retainIds: true }, (err, transformedCollection) => {
            if (err) {
                return reject(err);
            }
            resolve(transformedCollection);
        });
    }));
}
class RequestResolver {
    constructor({ postmanApiKey, accessToken, requestLoader }) {
        this.cache = new Map();
        this.postmanApiKey = postmanApiKey;
        this.accessToken = accessToken;
        this.requestLoader = requestLoader;
        this._cacheRequest = this._cacheRequest.bind(this);
        this._getRequestFromCache = this._getRequestFromCache.bind(this);
        this.resolveRequest = this.resolveRequest.bind(this);
    }
    _cacheRequest(requestObj) {
        if (this.cache.size >= CACHE_SIZE) {
            const lruElement = this.cache.keys().next().value;
            this.cache.delete(lruElement);
        }
        this.cache.set(requestObj.id, requestObj);
    }
    _getRequestFromCache(requestId) {
        const requestObj = this.cache.get(requestId);
        if (!requestObj) {
            return undefined;
        }
        this.cache.delete(requestId);
        this._cacheRequest(requestObj);
        return requestObj;
    }
    async resolveRequest(requestId, referencedRequestContext, callback) {
        if (!requestId) {
            return callback(new Error('Invalid Request ID passed to pm.execution.runRequest'));
        }
        const postmanApiKey = util.getValidPostmanAPIKey({ postmanApiKey: this.postmanApiKey }), accessToken = this.accessToken, cachedRequest = this._getRequestFromCache(requestId), { nestedRequestMeta } = referencedRequestContext || {};
        if (cachedRequest) {
            return callback(cachedRequest.error && cachedRequest.error.message ? cachedRequest.error.message : null, cachedRequest.requestData);
        }
        let requestNameFromMeta = '', requestPathFromMeta = '';
        if (nestedRequestMeta && nestedRequestMeta.location && nestedRequestMeta.location.length) {
            requestNameFromMeta = nestedRequestMeta.location[nestedRequestMeta.location.length - 1];
            requestPathFromMeta = nestedRequestMeta.location.join(' / ');
        }
        print.print(colors.gray('  %s'), `Loading referenced request: ${requestNameFromMeta || requestId}.\n`);
        try {
            let data;
            if (this.requestLoader) {
                data = await this.requestLoader.loadRequest(requestId);
            }
            else {
                const requestResource = new PostmanResource('request', requestId, postmanApiKey, accessToken);
                data = await new Promise((resolve, reject) => {
                    requestResource.get((err, requestData) => {
                        if (err || !requestData || !requestData.id) {
                            const error = new Error(`Failed to find request with ID: ${requestId}. ` +
                                'Please ensure the request exists.');
                            if (requestPathFromMeta) {
                                error.message = (`Failed to find request at ${requestPathFromMeta}. ` +
                                    'Please ensure the request exists.');
                            }
                            return reject(error);
                        }
                        resolve(requestData);
                    });
                });
            }
            if (!data || !data.id) {
                const error = new Error(`Failed to find request with ID: ${requestId}. ` +
                    'Please ensure the request exists.');
                if (requestPathFromMeta) {
                    error.message = (`Failed to find request at ${requestPathFromMeta}. ` +
                        'Please ensure the request exists.');
                }
                throw error;
            }
            sanitizeRequestData(data);
            setDefaultFields(data);
            const loadedRequest = await transformToV2Collection(getRequestExecutionObject(data));
            this._cacheRequest({ id: requestId, requestData: loadedRequest, error: null });
            return callback(null, loadedRequest);
        }
        catch (error) {
            this._cacheRequest({ id: requestId, requestData: null, error: error });
            print.lf(colors.red('[Error] Something went wrong while loading referenced request: ' + error.message));
            return callback(error);
        }
    }
}
module.exports = RequestResolver;
