// @postman/pm-bin-linux-x64/bin/lib/run/run-context.js
var _ = require('lodash');
function RunContext() {
    this.itemRegistry = {};
    this.parentMap = {};
    this.responseBuffer = {};
    this.itemPlan = [];
}
_.assign(RunContext.prototype, {
    registerItems: function (itemTree, parentId) {
        if (!itemTree || !itemTree.id) {
            return;
        }
        this.itemRegistry[itemTree.id] = itemTree;
        if (parentId) {
            this.parentMap[itemTree.id] = parentId;
        }
        if (Array.isArray(itemTree.children)) {
            itemTree.children.forEach((child) => {
                this.registerItems(child, itemTree.id);
            });
        }
    },
    getParentFolder: function (itemId) {
        var parentId = this.parentMap[itemId], parent = parentId && this.itemRegistry[parentId];
        if (parent && parent.type === 'folder') {
            return parent;
        }
        return undefined;
    },
    getFullName: function (itemId) {
        var item = this.itemRegistry[itemId], chain = [], currentId;
        if (!item) {
            return undefined;
        }
        chain.push(item.title || item.id);
        currentId = itemId;
        while (this.parentMap[currentId]) {
            currentId = this.parentMap[currentId];
            var ancestor = this.itemRegistry[currentId];
            if (ancestor && ancestor.type === 'folder') {
                chain.unshift(ancestor.title || ancestor.id);
            }
        }
        return chain.join(' / ');
    },
    resolveItemRef: function (ref) {
        if (!ref) {
            return undefined;
        }
        var id = ref.includes('#') ? ref.split('#')[1] : ref;
        return this.itemRegistry[id];
    },
    bufferResponseStart: function (correlationId, payload) {
        if (!correlationId) {
            return;
        }
        this.responseBuffer[correlationId] = {
            start: payload,
            chunks: [],
            end: null
        };
    },
    bufferResponseData: function (correlationId, payload) {
        if (!correlationId || !this.responseBuffer[correlationId]) {
            return;
        }
        if (payload && payload.data) {
            var chunk = Buffer.isBuffer(payload.data) ? payload.data : Buffer.from(payload.data);
            this.responseBuffer[correlationId].chunks.push(chunk);
        }
    },
    clearResponseBuffer: function (correlationId) {
        if (correlationId && this.responseBuffer[correlationId]) {
            delete this.responseBuffer[correlationId];
        }
    },
    assembleResponse: function (correlationId, endPayload) {
        if (!correlationId || !this.responseBuffer[correlationId]) {
            return null;
        }
        var buffer = this.responseBuffer[correlationId], start = buffer.start || {}, body = buffer.chunks.length > 0 ? Buffer.concat(buffer.chunks) : Buffer.alloc(0);
        delete this.responseBuffer[correlationId];
        return {
            statusCode: start.statusCode,
            statusText: start.statusText,
            headers: start.headers || [],
            body: body,
            timings: (endPayload && endPayload.timings) || (start.timings) || {},
            sizes: (endPayload && endPayload.sizes) || {}
        };
    }
});
module.exports = RunContext;
