// @postman/pm-bin-linux-x64/bin/lib/run/secure-fs.js
require('../logger').init();
const fs = require('fs'), _ = require('lodash'), path = require('path'), ShortUniqueId = require('short-unique-id'), { Readable, PassThrough } = require('stream'), { getCloudFileMeta, fetchBlobContents } = require('./cloud-file'), print = require('../framework/print'), util = require('../util'), PPERM_ERR = 'PPERM: insecure file access outside working directory', FUNCTION = 'function', DEPRECATED_SYNC_WRITE_STREAM = 'SyncWriteStream', EXPERIMENTAL_PROMISE = 'promises', REMOTE_FILE_PREFIX = 'postman-cloud:///', uid = new ShortUniqueId({ length: 10 }), isWindowsRoot = function (path) {
    const drive = path.charAt(1);
    return ((path.charAt(0) === '/') &&
        ((drive >= 'A' && drive <= 'Z') || (drive >= 'a' && drive <= 'z')) &&
        (path.charAt(2) === ':')) ||
        path.slice(0, 3) === '///';
}, stripTrailingSep = function (thePath) {
    if (thePath[thePath.length - 1] === path.sep) {
        return thePath.slice(0, -1);
    }
    return thePath;
}, pathIsInside = function (thePath, potentialParent) {
    thePath = stripTrailingSep(thePath);
    potentialParent = stripTrailingSep(potentialParent);
    if (global.process.platform === 'win32') {
        thePath = thePath.toLowerCase();
        potentialParent = potentialParent.toLowerCase();
    }
    return thePath.lastIndexOf(potentialParent, 0) === 0 &&
        (thePath[potentialParent.length] === path.sep ||
            thePath[potentialParent.length] === undefined);
};
function SecureFS({ workingDir, insecureFileRead = false, fileWhitelist = [], postmanApiKey, accessToken, fileLoader } = {}) {
    this._fs = fs;
    this._path = path;
    this.constants = this._fs.constants;
    this.workingDir = workingDir;
    this.insecureFileRead = insecureFileRead;
    this.fileWhitelist = fileWhitelist;
    this.postmanApiKey = postmanApiKey;
    this.accessToken = accessToken;
    this.fileLoader = fileLoader;
    this.isWindows = global.process.platform === 'win32';
    this.maxCacheSize = 100 * 1024 * 1024;
    this.currentCacheSize = 0;
    this._fileIdNameMap = new Map();
}
SecureFS.prototype._resolve = function (relOrAbsPath, whiteList) {
    this.isWindows && isWindowsRoot(relOrAbsPath) && (relOrAbsPath = relOrAbsPath.substring(1));
    let resolvedPath = this._path.resolve(this.workingDir, relOrAbsPath);
    if (!this.insecureFileRead &&
        !pathIsInside(resolvedPath, this.workingDir) &&
        !_.includes(whiteList, resolvedPath)) {
        return undefined;
    }
    return resolvedPath;
};
SecureFS.prototype.resolvePath = function (relOrAbsPath, whiteList, callback) {
    if (!callback && typeof whiteList === FUNCTION) {
        callback = whiteList;
        whiteList = [];
    }
    let resolvedPath = this._resolve(relOrAbsPath, _.concat(this.fileWhitelist, whiteList));
    if (!resolvedPath) {
        return callback(new Error(PPERM_ERR));
    }
    return callback(null, resolvedPath);
};
SecureFS.prototype.resolvePathSync = function (relOrAbsPath, whiteList) {
    const resolvedPath = this._resolve(relOrAbsPath, _.concat(this.fileWhitelist, whiteList));
    if (!resolvedPath) {
        throw new Error(PPERM_ERR);
    }
    return resolvedPath;
};
Object.getOwnPropertyNames(fs).map((prop) => {
    if (prop === DEPRECATED_SYNC_WRITE_STREAM || prop === EXPERIMENTAL_PROMISE || typeof fs[prop] !== FUNCTION) {
        return;
    }
    SecureFS.prototype[prop] = fs[prop];
});
SecureFS.prototype._isFileRemote = function (path) {
    return path.includes(REMOTE_FILE_PREFIX);
};
SecureFS.prototype.stat = function (fileSrc, callback) {
    if (this._isFileRemote(fileSrc)) {
        const fileId = fileSrc.substring(REMOTE_FILE_PREFIX.length);
        if (!this._fileIdNameMap.has(fileId)) {
            this._fetchFileAndSave(fileId).then((tempFileDetails) => {
                fs.stat(tempFileDetails.tempFileSrc, callback);
            }).catch((err) => { callback(err); });
        }
        else {
            const tempFileDetails = this._fileIdNameMap.get(fileId);
            this._fileIdNameMap.delete(fileId);
            this._fileIdNameMap.set(fileId, tempFileDetails);
            return fs.stat(tempFileDetails.tempFileSrc, callback);
        }
    }
    else {
        this.resolvePath(fileSrc, (err, resolvedPath) => {
            if (err) {
                return callback(err);
            }
            return this._fs.stat(resolvedPath, callback);
        });
    }
};
SecureFS.prototype.createReadStream = function (fileSrc, options) {
    try {
        if (this._isFileRemote(fileSrc)) {
            const passThroughStream = new PassThrough(), fileId = fileSrc.substring(REMOTE_FILE_PREFIX.length);
            if (!this._fileIdNameMap.has(fileId)) {
                this._fetchFileAndSave(fileId).then((tempFileDetails) => {
                    const { tempFileSrc, fileName } = tempFileDetails;
                    passThroughStream.path = fileName;
                    passThroughStream.mode = 0o400;
                    fs.createReadStream(tempFileSrc, options).pipe(passThroughStream);
                }).catch((err) => {
                    passThroughStream.emit('error', err);
                    passThroughStream.destroy();
                });
                return passThroughStream;
            }
            const tempFileDetails = this._fileIdNameMap.get(fileId), { tempFileSrc } = tempFileDetails;
            this._fileIdNameMap.delete(fileId);
            this._fileIdNameMap.set(fileId, tempFileDetails);
            return fs.createReadStream(tempFileSrc, { ...options, mode: 0o400 });
        }
        return fs.createReadStream(this.resolvePathSync(fileSrc), { ...options, mode: 0o400 });
    }
    catch (err) {
        class ErrorReadStream extends Readable {
            constructor() {
                super({ emitClose: false });
                this.emit('error', err);
                (options && !options.autoClose) || this.destroy();
            }
        }
        return new ErrorReadStream();
    }
};
SecureFS.prototype._saveFile = function (path, stream) {
    const directoryPath = path.substring(0, path.lastIndexOf('/'));
    if (!fs.existsSync(directoryPath)) {
        fs.mkdirSync(directoryPath, { recursive: true });
    }
    fs.writeFileSync(path, stream);
    fs.chmodSync(path, 0o400);
};
SecureFS.prototype._fetchFileAndSave = async function (fileId) {
    if (this._fileIdNameMap.has(fileId)) {
        return this._fileIdNameMap.get(fileId);
    }
    const postmanApiKey = util.getValidPostmanAPIKey({ postmanApiKey: this.postmanApiKey }), accessToken = this.accessToken;
    try {
        let fileMeta, fileContents;
        if (this.fileLoader) {
            const result = await this.fileLoader.loadFile({ fileId });
            fileMeta = result.fileMeta;
            fileContents = result.fileContents;
        }
        else {
            if (!postmanApiKey && !accessToken) {
                print.lf('Postman credentials is required for getting files from postman cloud.\n' +
                    'Please use "postman login" command to login');
                throw new Error('Postman API Key is missing');
            }
            fileMeta = await getCloudFileMeta({ fileId, postmanApiKey, accessToken });
            const blobId = _.get(fileMeta, 'contentMetadata.blob.id');
            fileContents = await fetchBlobContents({ fileId, blobId, postmanApiKey, accessToken });
        }
        const filePath = `${this.workingDir}/.cloudFiles/${uid.rnd()}/${_.get(fileMeta, 'name')}`, tempFileDetails = {
            tempFileSrc: filePath,
            fileName: _.get(fileMeta, 'name'),
            fileSize: _.get(fileMeta, 'size', 0)
        };
        this._saveFile(filePath, fileContents);
        this._fileIdNameMap.set(fileId, tempFileDetails);
        this.currentCacheSize += tempFileDetails.fileSize;
        while (this.currentCacheSize > this.maxCacheSize && this._fileIdNameMap.size > 1) {
            const firstFileId = this._fileIdNameMap.keys().next().value, { tempFileSrc: firstFilePath, fileSize: firstFileSize } = this._fileIdNameMap.get(firstFileId);
            this._fileIdNameMap.delete(firstFileId);
            this.currentCacheSize -= firstFileSize;
            fs.unlink(firstFilePath, (err) => {
                if (err) {
                    logger.error(`Error deleting file ${firstFilePath}: ${err}`);
                }
            });
        }
        return tempFileDetails;
    }
    catch (error) {
        logger.error(`Error while fetching cloud File, ${error}`);
        if (error.name === 'Axios Error') {
            throw new Error(`Could not fetch cloud file - ${fileId}: ${error.message}`);
        }
        throw new Error(`Could not fetch cloud file - ${fileId}`);
    }
};
module.exports = SecureFS;
