// @postman/pm-bin-linux-x64/bin/lib/run/summary.js
var _ = require('lodash'), sdk = require('postman-collection'), SerialiseError = require('serialised-error'), { attachUnifiedTrackers } = require('./unified/summary'), RunSummary;
RunSummary = function RunSummary(emitter, options) {
    var summary = this;
    _.assign(summary, {
        collection: _.get(options, 'collection'),
        environment: _.get(options, 'environment'),
        globals: _.get(options, 'globals'),
        run: {
            stats: {
                iterations: {},
                items: {},
                scripts: {},
                prerequests: {},
                requests: {},
                tests: {},
                assertions: {},
                testScripts: {},
                prerequestScripts: {},
                responses: {}
            },
            timings: {
                responseAverage: 0,
                responseMin: 0,
                responseMax: 0,
                responseSd: 0,
                dnsAverage: 0,
                dnsMin: 0,
                dnsMax: 0,
                dnsSd: 0,
                firstByteAverage: 0,
                firstByteMin: 0,
                firstByteMax: 0,
                firstByteSd: 0
            },
            executions: [],
            transfers: {
                responseTotal: 0
            },
            failures: [],
            error: null
        }
    });
    RunSummary.attachTimingTrackers(this, emitter);
    RunSummary.attachStatisticTrackers(this, emitter);
    RunSummary.attachRequestTracker(this, emitter);
    RunSummary.attachFailureTrackers(this, emitter);
    RunSummary.attachReportingTrackers(this, emitter);
    RunSummary.attachUnifiedTrackers(this, emitter, options);
};
_.assign(RunSummary, {
    attachReportingTrackers(summary, emitter) {
        var cache = {}, executions = summary.run.executions;
        emitter.on('beforeItem', function (err, o) {
            if (err || !_.get(o, 'cursor.ref')) {
                return;
            }
            cache[o.cursor.ref] = _.assignIn(cache[o.cursor.ref] || {}, {
                cursor: o.cursor,
                item: o.item,
                itemLegacy: o.item,
                request: _.get(o, 'item.request'),
                id: _.get(o, 'item.id')
            });
            executions.push(cache[o.cursor.ref]);
        });
        emitter.on('request', function (err, o) {
            if (!_.get(o, 'cursor.ref')) {
                return;
            }
            if (_.get(o, 'cursor.scriptId')) {
                return;
            }
            var execution = cache[o.cursor.ref] = (cache[o.cursor.ref] || {});
            if (_.isEmpty(execution)) {
                executions.push(execution);
            }
            _.assignIn(execution, {
                cursor: o.cursor,
                request: o.request,
                item: o.item,
                response: o.response,
                id: _.get(o, 'item.id')
            }, err && {
                requestError: new SerialiseError(err, true)
            });
        });
        emitter.on('script', function (err, o) {
            if (!_.get(o, 'cursor.ref')) {
                return;
            }
            var execution = cache[o.cursor.ref] = (cache[o.cursor.ref] || {}), eventName = o && o.event && (o.event.listen + 'Script');
            err && (execution && eventName) && (execution[eventName] || (execution[eventName] = [])).push({
                error: err
            });
        });
        emitter.on('assertion', function (err, o) {
            if (!_.get(o, 'cursor.ref')) {
                return;
            }
            var execution = cache[o.cursor.ref] = (cache[o.cursor.ref] || {});
            if (!execution) {
                return;
            }
            (execution.assertions || (execution.assertions = [])).push({
                assertion: o.assertion,
                skipped: o.skipped,
                error: err || undefined
            });
        });
    },
    attachTimingTrackers(summary, emitter) {
        emitter.on('start', function () { summary.run.timings.started = Date.now(); });
        emitter.on('beforeDone', function () {
            summary.run.timings.completed = Date.now();
        });
        emitter.on('done', function (err) {
            err && (summary.error = err);
        });
    },
    attachStatisticTrackers(summary, emitter) {
        _.forEach(summary.run.stats, function (tracker, name) {
            name = name.slice(0, -1);
            _.assign(tracker, { total: 0, pending: 0, failed: 0 });
            emitter.on(_.camelCase('before-' + name), function () {
                tracker.pending += 1;
            });
            if (name === 'response') {
                _.assign(tracker, { totalResponseTime: 0 });
                emitter.on('response', function (err, o) {
                    if (err || !(o && o.response)) {
                        tracker.failed += 1;
                        return;
                    }
                    tracker.total += 1;
                    tracker.totalResponseTime += o.response.responseTime;
                });
            }
            else {
                emitter.on(name, function (err) {
                    tracker.pending && (tracker.pending -= 1);
                    err && (tracker.failed += 1);
                    tracker.total += 1;
                });
            }
        });
    },
    attachRequestTracker(summary, emitter) {
        emitter.on('request', function (err, o) {
            if (err || !(o && o.response)) {
                return;
            }
            var size = _.isFunction(o.response.size) && o.response.size(), time = o.response.responseTime, requestCount = summary.run.stats.requests.total, timings, timingPhases;
            size && (summary.run.transfers.responseTotal += (size.body || 0 + size.headers || 0));
            timings = _.last(_.get(o, 'history.execution.data'));
            timings = timings && timings.timings;
            timingPhases = timings && sdk.Response.timingPhases(timings);
            (timingPhases || time) && _.forEach([
                'dns',
                'firstByte',
                'response'
            ], (value) => {
                var currentValue = (value === 'response') ? time : (timingPhases && timingPhases[value]), previousAverage = summary.run.timings[`${value}Average`], previousVariance = summary.run.timings[`${value}Sd`] ** 2, delta1 = currentValue - previousAverage, delta2, currentVariance;
                if (!currentValue) {
                    return;
                }
                summary.run.timings[`${value}Average`] =
                    (previousAverage * (requestCount - 1) + currentValue) / requestCount;
                if (!summary.run.timings[`${value}Min`]) {
                    summary.run.timings[`${value}Min`] = currentValue;
                }
                else {
                    summary.run.timings[`${value}Min`] =
                        Math.min(summary.run.timings[`${value}Min`], currentValue);
                }
                summary.run.timings[`${value}Max`] = Math.max(summary.run.timings[`${value}Max`], currentValue);
                delta2 = currentValue - summary.run.timings[`${value}Average`];
                currentVariance = (previousVariance * (requestCount - 1) + (delta1 * delta2)) / requestCount;
                summary.run.timings[`${value}Sd`] = Math.sqrt(currentVariance);
            });
        });
    },
    attachFailureTrackers(summary, emitter) {
        var eventsToTrack = ['beforeIteration', 'iteration', 'beforeItem', 'item', 'beforeScript', 'script',
            'beforePrerequest', 'prerequest', 'beforeRequest', 'request', 'beforeTest', 'test', 'beforeAssertion',
            'assertion'];
        _.forEach(eventsToTrack, function (event) {
            let parentItem;
            emitter.on(event, function (err, o) {
                if (!err) {
                    return;
                }
                var item = o && o.item, source = event;
                if (event === 'request') {
                    const scriptId = o.cursor && o.cursor.scriptId;
                    scriptId && summary.run.executions.map((execution) => {
                        execution.item.events.map((eventData) => {
                            if (eventData.script.id === scriptId) {
                                parentItem = _.omit(execution, 'cursor');
                            }
                        });
                    });
                }
                if (event === 'script') {
                    o.event && (source = o.event.listen + '-script');
                    if (err.stacktrace && err.stacktrace[0] && err.stacktrace[0].lineNumber) {
                        source += (':' + (err.stacktrace[0].lineNumber - 2));
                        err.stacktrace[0].columnNumber && (source += (':' + err.stacktrace[0].columnNumber));
                    }
                }
                else if (event === 'assertion') {
                    _.has(err, 'index') && (source += (':' + err.index));
                    source += ' in test-script';
                }
                if (err.stack && !err.stacktrace) {
                    err = new SerialiseError(err, true);
                }
                summary.run.failures.push({
                    error: err,
                    at: source,
                    source: item || undefined,
                    parent: item && item.__parent && item.__parent.__parent || parentItem || undefined,
                    cursor: o.cursor || {}
                });
            });
        });
    },
    attachUnifiedTrackers
});
module.exports = RunSummary;
