// @postman/pm-bin-linux-x64/bin/lib/commands/runner/utils/working-directory.js
const fs = require('fs'), path = require('path'), os = require('os'), cliEnvironment = require('../../../config/cli-environment');
function isDirectoryWritable(dirPath) {
    try {
        fs.mkdirSync(dirPath, { recursive: true });
        fs.accessSync(dirPath, fs.constants.W_OK);
        return true;
    }
    catch (error) {
        return false;
    }
}
function getValidHomeDirectory() {
    const homeDir = os.homedir();
    return (!homeDir || homeDir === '/') ? process.cwd() : homeDir;
}
function getDirectoryCandidates() {
    const homeDir = getValidHomeDirectory();
    const postmanHomeDir = cliEnvironment.getSettings().postmanHomeDir || '.postman';
    return [
        {
            path: path.join(homeDir, postmanHomeDir, 'cli'),
            name: 'Postman home directory'
        },
        {
            path: path.join(process.cwd(), postmanHomeDir, 'cli'),
            name: 'Current directory'
        },
        {
            path: path.join(os.tmpdir(), postmanHomeDir, 'cli'),
            name: 'Temporary directory'
        }
    ];
}
function getWorkingDirectory() {
    const candidates = getDirectoryCandidates();
    for (const candidate of candidates) {
        if (isDirectoryWritable(candidate.path)) {
            return candidate.path;
        }
    }
    return null;
}
module.exports = {
    getWorkingDirectory,
    isDirectoryWritable
};
