// @postman/pm-bin-linux-x64/bin/lib/config/rc-file.js
var _ = require('lodash'), fs = require('fs'), waterfall = require('async/waterfall'), join = require('path').join, async = require('async'), util = require('../util'), liquidJSON = require('liquid-json'), os = require('os'), cliEnvironment = require('./cli-environment'), CONFIG_DIR_MODE = 0o700, getHomeConfigDir = function () {
    const settings = cliEnvironment.getSettings();
    return join(os.homedir(), settings.postmanHomeDir);
}, FILE_NAME = 'postmanrc', CONFIG_FILE_MODE = 0o600, getConfigFiles = function () {
    return {
        home: join(getHomeConfigDir(), FILE_NAME),
        cwd: join(process.cwd(), '.' + FILE_NAME)
    };
}, DIR_CREATE_ERROR = (dir) => { return `Couldn't create the config directory at ${dir}.`; }, INVALID_DATA_ERROR = (file) => { return `The config file at ${file} contains invalid data.`; }, FILE_WRITE_ERROR = (file) => { return `Couldn't write the config file at ${file}.`; };
module.exports = {
    get HOME_CONFIG_DIR() { return getHomeConfigDir(); },
    CONFIG_DIR_MODE: CONFIG_DIR_MODE,
    CONFIG_FILE_MODE: CONFIG_FILE_MODE,
    load: (types, callback) => {
        if (!callback && _.isFunction(types)) {
            callback = types;
            types = ['home'];
        }
        var files = [], configFiles = getConfigFiles();
        _.forEach(types, (type) => {
            configFiles[type] && files.push(configFiles[type]);
        });
        async.mapSeries(files, (path, cb) => {
            fs.readFile(path, (err, data) => {
                if (err) {
                    return cb(null, {});
                }
                data && data.toString && (data = data.toString(util.detectEncoding(data)).trim());
                try {
                    return cb(null, liquidJSON.parse(data));
                }
                catch (e) {
                    return cb(new Error(INVALID_DATA_ERROR(path)));
                }
            });
        }, (err, fileData) => {
            if (err) {
                return callback(err);
            }
            return callback(null, _.merge({}, ...fileData));
        });
    },
    store: (data, type, callback) => {
        if (!callback && _.isFunction(type)) {
            callback = type;
            type = 'home';
        }
        let file = _.get(getConfigFiles(), type);
        waterfall([
            (next) => {
                const homeConfigDir = getHomeConfigDir();
                if (type === 'home' && !fs.existsSync(homeConfigDir)) {
                    fs.mkdir(homeConfigDir, { mode: CONFIG_DIR_MODE }, (err) => {
                        if (err) {
                            return next(new Error(DIR_CREATE_ERROR(homeConfigDir)));
                        }
                        return next(null);
                    });
                }
                else {
                    return next(null);
                }
            },
            (next) => {
                fs.writeFile(file, JSON.stringify(data, null, 2), { mode: CONFIG_FILE_MODE }, (err) => {
                    if (err) {
                        return next(new Error(FILE_WRITE_ERROR(file)));
                    }
                    return next(null);
                });
            }
        ], (err) => {
            callback(err);
        });
    },
    loadSync: (types = ['home']) => {
        var files = [];
        const configFiles = getConfigFiles();
        _.forEach(types, (type) => {
            if (configFiles[type]) {
                files.push(configFiles[type]);
            }
        });
        const fileData = files.map((path) => {
            try {
                if (!fs.existsSync(path)) {
                    return {};
                }
                let data = fs.readFileSync(path);
                data && data.toString && (data = data.toString(util.detectEncoding(data)).trim());
                return liquidJSON.parse(data);
            }
            catch (e) {
                if (e instanceof SyntaxError) {
                    throw new Error(INVALID_DATA_ERROR(path));
                }
                return {};
            }
        });
        return _.merge({}, ...fileData);
    }
};
