// @postman/pm-bin-linux-x64/bin/lib/config/validateAccessToken.js
const requestFactory = require('@postman/platform-req/lib/request');
const util = require('../util');
function getClient() {
    return requestFactory.defaults({
        retry: {
            enabled: true,
            delay: 50,
            backOffNoise: 100,
            maxAttempts: 2,
            maxDelay: 1000,
            statusCodes: [500, 502, 503, 504],
            methods: {
                GET: {
                    enabled: true
                }
            }
        }
    });
}
function validateAccessToken(accessToken, callback) {
    if (!accessToken) {
        return callback(new Error('Access token is required'));
    }
    const client = getClient(), regionUtil = require('../region-util'), region = regionUtil.getCurrentRegion(), iapubBaseUrl = util.POSTMAN_IAPUB_BASE_URL(region), url = `${iapubBaseUrl}/api/sessions/current`, requestOptions = {
        url: url,
        method: 'GET',
        headers: {
            'x-access-token': accessToken,
            'User-Agent': util.userAgent
        },
        json: true,
        timeout: 10000
    };
    client.get(requestOptions, (error, response) => {
        if (error) {
            return callback(new Error(`Failed to validate access token: ${error.message}`));
        }
        if (response && response.statusCode === 200 &&
            response.body && response.body.session && response.body.session.status === 'active') {
            return callback(null, {
                postmanApiKey: accessToken,
                valid: true,
                session: response.body.session
            });
        }
        const statusMessage = response && response.statusCode ?
            `Access token validation failed with status: ${response.statusCode}` :
            'Access token validation failed';
        return callback(new Error(`${statusMessage}. Please run \`postman login\` to authenticate again.`));
    });
}
module.exports = validateAccessToken;
